// @takeoff-ui/core@0.11.9/dist/collection/components/tk-radio/tk-radio-group/tk-radio-group.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { isEqual } from "lodash-es";
import { renderHint } from "../../../utils/hint-utils";
import { getDataTestId } from "../../../utils/test-id-utils";
export class TkRadioGroup {
    constructor() {
        /**
         * The direction of the radio buttons.
         */
        this.direction = 'horizontal';
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Determines the position of the radio group and label.
         */
        this.position = 'left';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * Determines whether the radios will spread evenly across the horizontal space.
         */
        this.spread = false;
        /**
         * Determines the appearance types of radios.
         */
        this.type = 'default';
    }
    invalidChanged() {
        var _a;
        if (((_a = this.slottedItems) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.slottedItems.forEach(item => {
                item.invalid = this.invalid;
            });
        }
    }
    /**
     * Watches for changes in the selected value and emits a custom event when the value changes.
     */
    valueChanged() {
        this.updateTkRadio();
    }
    formResetCallback() {
        this.handleFormReset();
    }
    updateTkRadio() {
        if (this.slottedItems.length > 0) {
            this.slottedItems.forEach(item => {
                item.checked = isEqual(this.value, item.value);
            });
        }
    }
    handleFormReset() {
        this.value = null;
        this.tkChange.emit(this.value);
        this.updateTkRadio();
    }
    handleChange(e) {
        this.value = e.detail;
        this.tkChange.emit(this.value);
        this.updateTkRadio();
    }
    handleSlotChange() {
        this.slottedItems = this.el.querySelectorAll('tk-radio');
        if (this.slottedItems.length > 0) {
            this.slottedItems.forEach(item => {
                item.addEventListener('tk-change', e => {
                    e.stopPropagation();
                    this.handleChange(e);
                });
                item.checked = isEqual(this.value, item.value);
                item.invalid = this.invalid;
                if (this.spread)
                    item.style.flex = '1';
                item.setAttribute('data-type', this.type);
                // radio componentlerinde name yoksa ve radi group componentinde name var ise
                // group'a verilen name radio componentlerine setlenmesi için yazılmıştır.
                if (!item.name && this.name)
                    item.name = this.name;
            });
        }
    }
    render() {
        var _a;
        let _label;
        const rootClasses = classNames('tk-radio-group-container', {
            vertical: this.direction === 'vertical',
        });
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            const asterisk = (h("span", { key: '43ac6ce9cd0f2cc9f50dc619f011a1ca0ff27d67', class: "asterisk", "data-testid": getDataTestId(this.dataTestid, 'label-asterisk') }, "*"));
            _label = (h("label", { key: 'bcbff37d9ee0a14424bae6b954d3f19f27b42ed6', class: "label", "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label, this.showAsterisk && asterisk));
        }
        return (h("div", { key: 'bed985255a6a99c6e9490f6b880778489240761d', class: rootClasses, "aria-invalid": this.invalid, "data-testid": getDataTestId(this.dataTestid, 'container') }, _label, h("div", { key: 'daba64884f66d2217297a06f517f9eca2cef4ada', class: classNames('tk-radio-holder', this.type, { spread: this.spread }), "data-testid": getDataTestId(this.dataTestid, 'holder') }, h("slot", { key: '8f21682f7b8c2be6566bd5b1a612dc4feb4e3ec3', onSlotchange: this.handleSlotChange.bind(this) })), renderHint(this.hint, this.error, this.invalid, this.dataTestid)));
    }
    static get is() { return "tk-radio-group"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-radio-group.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-radio-group.css"]
        };
    }
    static get properties() {
        return {
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the label for the element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "direction": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'vertical' | 'horizontal'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The direction of the radio buttons."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "direction",
                "defaultValue": "'horizontal'"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "position": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines the position of the radio group and label."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "position",
                "defaultValue": "'left'"
            },
            "value": {
                "type": "any",
                "mutable": true,
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "spread": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines whether the radios will spread evenly across the horizontal space."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "spread",
                "defaultValue": "false"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'default' | 'card'",
                    "resolved": "\"card\" | \"default\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines the appearance types of radios."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'default'"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "This is the hint message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This is the error message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The name attribute for the input element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "slottedItems": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the value has changed."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "invalid",
                "methodName": "invalidChanged"
            }, {
                "propName": "value",
                "methodName": "valueChanged"
            }];
    }
    static get attachInternalsMemberName() { return "internals"; }
}
