// @takeoff-ui/core@0.11.9/components/index.js
export { g as getAssetPath, r as render, s as setAssetPath, a as setNonce, b as setPlatformOptions } from './p-BfLQgzDX.js';
const t = new Map(),
  o = t => {
    var o, e, n, i, s, d;
    (['top-right', 'bottom-right', 'bottom-left', 'top-left'].includes(t.position) || (t.position = 'top-right'), null == (null == t ? void 0 : t.timeout) && (t.timeout = 6e3));
    let l = document.querySelector('.tk-toaster.' + t.position);
    l || ((l = document.createElement('div')), l.classList.add('tk-toaster'), l.classList.add(t.position), document.querySelector('body').append(l));
    const r = document.createElement('tk-alert');
    if (
      ((null === (o = t.header) || void 0 === o ? void 0 : o.length) > 0 && (r.header = t.header),
      (null === (e = t.message) || void 0 === e ? void 0 : e.length) > 0 && (r.message = t.message),
      r.setAttribute('type', t.type),
      r.setAttribute('variant', t.variant),
      (null === (n = t.alignItems) || void 0 === n ? void 0 : n.length) > 0 && (r.alignItems = t.alignItems),
      (null === (i = t.icon) || void 0 === i ? void 0 : i.length) > 0 && (r.icon = t.icon),
      (null === (s = t.iconSize) || void 0 === s ? void 0 : s.length) > 0 && (r.iconSize = t.iconSize),
      t.removable && (r.removable = t.removable),
      (null === (d = t.actions) || void 0 === d ? void 0 : d.length) > 0)
    ) {
      const o = document.createElement('div');
      (o.setAttribute('slot', 'footer-action'),
        (o.style.display = 'flex'),
        (o.style.gap = '8px'),
        t.actions.forEach(t => {
          const e = document.createElement('tk-button');
          ((e.label = t.label),
            (e.icon = t.icon),
            (e.variant = t.variant || 'primary'),
            (e.type = t.type || 'filled'),
            (e.size = t.size || 'base'),
            'function' == typeof t.action && e.addEventListener('tk-click', () => t.action()),
            o.appendChild(e));
        }),
        r.appendChild(o));
    }
    t.containerStyle && (r.containerStyle = t.containerStyle);
    const c = document.createElement('div');
    (c.classList.add('tk-toast'),
      l.append(c),
      c.append(r),
      setTimeout(() => {
        c.classList.add('open');
      }, 1),
      t.persistent ||
        setTimeout(() => {
          (c.classList.remove('open'),
            setTimeout(() => {
              (r.remove(), c.remove());
            }, 300));
        }, t.timeout));
  },
  e = o => {
    var e, n, i, s, d, l;
    const { persistentId: r } = o,
      c = t.get(r);
    if (c) {
      if (c.isRemoving) return !1;
      c.isRemoving = !0;
      const { container: o } = c;
      return (
        o.classList.remove('open'),
        setTimeout(() => {
          (o.parentNode && o.remove(), t.delete(r));
        }, 300),
        !1
      );
    }
    const u = Object.assign(Object.assign({}, o), { persistent: !0, removable: !0 });
    ['top-right', 'bottom-right', 'bottom-left', 'top-left'].includes(u.position) || (u.position = 'top-right');
    let v = document.querySelector('.tk-toaster.' + u.position);
    v || ((v = document.createElement('div')), v.classList.add('tk-toaster'), v.classList.add(u.position), document.querySelector('body').append(v));
    const a = document.createElement('tk-alert');
    if (
      ((null === (e = u.header) || void 0 === e ? void 0 : e.length) > 0 && (a.header = u.header),
      (null === (n = u.message) || void 0 === n ? void 0 : n.length) > 0 && (a.message = u.message),
      a.setAttribute('type', u.type),
      a.setAttribute('variant', u.variant),
      (null === (i = u.alignItems) || void 0 === i ? void 0 : i.length) > 0 && (a.alignItems = u.alignItems),
      (null === (s = u.icon) || void 0 === s ? void 0 : s.length) > 0 && (a.icon = u.icon),
      (null === (d = u.iconSize) || void 0 === d ? void 0 : d.length) > 0 && (a.iconSize = u.iconSize),
      u.removable && (a.removable = u.removable),
      (null === (l = u.actions) || void 0 === l ? void 0 : l.length) > 0)
    ) {
      const t = document.createElement('div');
      (t.setAttribute('slot', 'footer-action'),
        (t.style.display = 'flex'),
        (t.style.gap = '8px'),
        u.actions.forEach(o => {
          const e = document.createElement('tk-button');
          ((e.label = o.label),
            (e.icon = o.icon),
            (e.variant = o.variant || 'primary'),
            (e.type = o.type || 'filled'),
            (e.size = o.size || 'base'),
            'function' == typeof o.action && e.addEventListener('tk-click', () => o.action()),
            t.appendChild(e));
        }),
        a.appendChild(t));
    }
    const m = document.createElement('div');
    return (
      m.classList.add('tk-toast'),
      v.append(m),
      m.append(a),
      t.set(r, { element: a, container: m, isRemoving: !1 }),
      a.addEventListener('remove', () => {
        const o = t.get(r);
        o && !o.isRemoving && t.delete(r);
      }),
      setTimeout(() => {
        m.classList.add('open');
      }, 1),
      !0
    );
  },
  n = o => {
    const e = t.get(o);
    return !!e && !e.isRemoving;
  },
  i = () => {
    t.forEach(({ container: o, isRemoving: e }, n) => {
      if (!e) {
        const e = t.get(n);
        (e && (e.isRemoving = !0),
          o.classList.remove('open'),
          setTimeout(() => {
            (o.parentNode && o.remove(), t.delete(n));
          }, 300));
      }
    });
  };
export { o as createToast, i as dismissAllPersistentToasts, n as isPersistentToastVisible, e as showPersistentToast };
