// @takeoff-ui/core@0.11.9/components/p-5F1gBr3j.js
import { p as e, H as t, c as a, h as i, t as n } from './p-BfLQgzDX.js';
import { c as o } from './p-D7YaJBkB.js';
import { g as s } from './p-Dpaq_7dX.js';
import { d as r } from './p-BpML1w-i.js';
import { v as l } from './p-CK3_k8jD.js';
const c = e(
  class extends t {
    constructor(e) {
      (super(),
        !1 !== e && this.__registerHost(),
        (this.tkChange = a(this, 'tk-change', 7)),
        (this.internals = this.attachInternals()),
        (this.uniqueId = l()),
        (this.hasContentSlot = !1),
        (this.disabled = !1),
        (this.invalid = !1),
        (this.type = 'default'),
        (this.size = 'base'),
        (this.indeterminate = !1),
        (this.value = !1));
    }
    indeterminateChanged(e, t) {
      e !== t && ((this.inputElement.indeterminate = e), e && (this.value = null));
    }
    valueChanged(e, t) {
      e !== t && null !== e && (this.indeterminate = !1);
    }
    componentWillLoad() {
      (this.indeterminate && (this.value = null), (this.hasContentSlot = !!this.el.querySelector('[slot="content"]')));
    }
    formResetCallback() {
      this.handleFormReset();
    }
    handleFormReset() {
      ((this.value = !1), (this.indeterminate = !1), (this.inputElement.checked = !1), this.tkChange.emit(this.value));
    }
    handleChange() {
      this.disabled ||
        (this.inputElement.indeterminate ? ((this.value = null), (this.indeterminate = !1)) : (this.value = this.inputElement.checked), this.tkChange.emit(this.value));
    }
    renderInput() {
      return i('input', {
        id: this.uniqueId,
        type: 'checkbox',
        ref: e => (this.inputElement = e),
        checked: this.value,
        indeterminate: this.indeterminate,
        disabled: this.disabled,
        onChange: this.handleChange.bind(this),
        name: this.name,
      });
    }
    render() {
      return i(
        'div',
        {
          'key': '36149a307c16ce4734f583e625bcfa544e6f9fec',
          'class': o('tk-checkbox-container', this.type, this.size),
          'aria-disabled': this.disabled,
          'aria-invalid': this.invalid,
        },
        i(
          'label',
          { key: '141ecdcaf29c7fe34b92c8469dc42ca667e78f0c', htmlFor: this.uniqueId },
          this.renderInput(),
          i(
            'div',
            { key: 'd80febfa0909b18a66f06de79219d4874eb712be', class: 'mask' },
            i(
              'tk-icon',
              Object.assign(
                { key: '9c00972114617557ac5e0878f3ad4dfcd7d206fd', class: 'inner-mask' },
                s(this.indeterminate ? 'remove' : this.value ? 'check' : null, { color: 'var(--static-light)', size: 'small' === this.size ? 'xxsmall' : 'xsmall' }),
              ),
            ),
          ),
          this.hasContentSlot
            ? i('slot', { name: 'content' })
            : i(
                'div',
                { class: 'tk-checkbox-text-holder' },
                i('div', { class: 'tk-checkbox-label' }, this.label),
                i('div', { class: 'tk-checkbox-description' }, this.description),
              ),
        ),
      );
    }
    static get formAssociated() {
      return !0;
    }
    get el() {
      return this;
    }
    static get watchers() {
      return { indeterminate: [{ indeterminateChanged: 0 }], value: [{ valueChanged: 0 }] };
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-checkbox{display:inline-block}.tk-checkbox-container{font-family:var(--family-body)}.tk-checkbox-container.card{border-radius:var(--radius-m-base);border:1px solid var(--border-light);background:var(--static-Light);gap:var(--card-switcher-container-gap);padding:var(--card-switcher-container-v-padding) var(--card-switcher-container-h-padding)}.tk-checkbox-container.card:hover{background-color:var(--background-lightest)}.tk-checkbox-container.card .mask{align-self:center}.tk-checkbox-container label{padding:2px;display:flex;flex-direction:row;align-items:center;justify-content:flex-start;gap:var(--spacing-s);color:var(--text-darkest);font-size:var(--size-sm);cursor:pointer;user-select:none}.tk-checkbox-container label input[type=checkbox]{margin:0;appearance:none;outline:none;display:none}.tk-checkbox-container label .mask{background-color:var(--static-light);border:2px solid var(--border-sub-base);border-radius:var(--radius-s);box-sizing:border-box;-moz-box-sizing:border-box;-webkit-box-sizing:border-box}.tk-checkbox-container label .mask .inner-mask{visibility:hidden;background-color:var(--states-info-base);display:flex;border-radius:2px}.tk-checkbox-container label .tk-checkbox-text-holder .tk-checkbox-description{color:var(--text-base);font-size:var(--size-xs);font-weight:var(--weight-regular)}.tk-checkbox-container:hover label .mask{background-color:var(--background-lightest)}.tk-checkbox-container label input:focus-visible+.mask{border-color:var(--states-info-base)}.tk-checkbox-container label input:checked+.mask,.tk-checkbox-container label input:indeterminate+.mask{border-color:var(--states-info-light);background-color:var(--static-light)}.tk-checkbox-container label input:checked+.mask .inner-mask,.tk-checkbox-container label input:indeterminate+.mask .inner-mask{visibility:visible}.tk-checkbox-container label:hover input:checked+.mask,.tk-checkbox-container label:hover input:indeterminate+.mask{border-color:var(--states-info-sub-base);background-color:var(--states-info-light)}.tk-checkbox-container[aria-disabled],.tk-checkbox-container[disabled]{pointer-events:none}.tk-checkbox-container[aria-disabled] label,.tk-checkbox-container[disabled] label{color:var(--text-sub-base)}.tk-checkbox-container[aria-disabled] label input+.mask,.tk-checkbox-container[disabled] label input+.mask{border:2px solid var(--border-sub-base);background-color:var(--background-light)}.tk-checkbox-container[aria-disabled] label input+.mask .inner-mask,.tk-checkbox-container[disabled] label input+.mask .inner-mask{background-color:var(--background-sub-base)}.tk-checkbox-container[aria-invalid] label .mask{border:2px solid var(--states-danger-base);background:var(--static-light)}.tk-checkbox-container[aria-invalid] label input:checked+.mask,.tk-checkbox-container[aria-invalid] label input:indeterminate+.mask{border-color:var(--states-danger-light)}.tk-checkbox-container[aria-invalid] label input:checked+.mask .inner-mask,.tk-checkbox-container[aria-invalid] label input:indeterminate+.mask .inner-mask{background-color:var(--states-danger-base)}.tk-checkbox-container[aria-invalid] label:hover input:checked+.mask,.tk-checkbox-container[aria-invalid] label:hover input:indeterminate+.mask{border-color:var(--states-danger-sub-base);background-color:var(--states-danger-light)}.tk-checkbox-container.base label .mask{border-radius:6px;padding:2px}.tk-checkbox-container.base label .mask .inner-mask{width:12px;height:12px}.tk-checkbox-container.small label .mask{border-radius:4px;padding:1px}.tk-checkbox-container.small label .mask .inner-mask{width:10px;height:10px}';
    }
  },
  [
    324,
    'tk-checkbox',
    { disabled: [4], description: [1], invalid: [4], label: [1], name: [1], type: [1], size: [1], indeterminate: [1028], value: [1028], hasContentSlot: [32] },
    void 0,
    { indeterminate: [{ indeterminateChanged: 0 }], value: [{ valueChanged: 0 }] },
  ],
);
function d() {
  'undefined' != typeof customElements &&
    ['tk-checkbox', 'tk-icon'].forEach(e => {
      switch (e) {
        case 'tk-checkbox':
          customElements.get(n(e)) || customElements.define(n(e), c);
          break;
        case 'tk-icon':
          customElements.get(n(e)) || r();
      }
    });
}
export { c as T, d };
