// @takeoff-ui/core@0.11.9/components/p-BiHzhysf.js
const t = ['top', 'right', 'bottom', 'left'],
  n = Math.min,
  e = Math.max,
  o = Math.round,
  r = Math.floor,
  i = t => ({ x: t, y: t }),
  a = { left: 'right', right: 'left', bottom: 'top', top: 'bottom' };
function l(t, o, r) {
  return e(t, n(o, r));
}
function c(t, n) {
  return 'function' == typeof t ? t(n) : t;
}
function s(t) {
  return t.split('-')[0];
}
function u(t) {
  return t.split('-')[1];
}
function f(t) {
  return 'x' === t ? 'y' : 'x';
}
function d(t) {
  return 'y' === t ? 'height' : 'width';
}
function m(t) {
  const n = t[0];
  return 't' === n || 'b' === n ? 'y' : 'x';
}
function p(t) {
  return f(m(t));
}
function h(t) {
  return t.includes('start') ? t.replace('start', 'end') : t.replace('end', 'start');
}
const y = ['left', 'right'],
  g = ['right', 'left'],
  x = ['top', 'bottom'],
  w = ['bottom', 'top'];
function b(t) {
  const n = s(t);
  return a[n] + t.slice(n.length);
}
function v(t) {
  return 'number' != typeof t
    ? (function (t) {
        return { top: 0, right: 0, bottom: 0, left: 0, ...t };
      })(t)
    : { top: t, right: t, bottom: t, left: t };
}
function A(t) {
  const { x: n, y: e, width: o, height: r } = t;
  return { width: o, height: r, top: e, left: n, right: n + o, bottom: e + r, x: n, y: e };
}
function S(t, n, e) {
  let { reference: o, floating: r } = t;
  const i = m(n),
    a = p(n),
    l = d(a),
    c = s(n),
    f = 'y' === i,
    h = o.x + o.width / 2 - r.width / 2,
    y = o.y + o.height / 2 - r.height / 2,
    g = o[l] / 2 - r[l] / 2;
  let x;
  switch (c) {
    case 'top':
      x = { x: h, y: o.y - r.height };
      break;
    case 'bottom':
      x = { x: h, y: o.y + o.height };
      break;
    case 'right':
      x = { x: o.x + o.width, y };
      break;
    case 'left':
      x = { x: o.x - r.width, y };
      break;
    default:
      x = { x: o.x, y: o.y };
  }
  switch (u(n)) {
    case 'start':
      x[a] -= g * (e && f ? -1 : 1);
      break;
    case 'end':
      x[a] += g * (e && f ? -1 : 1);
  }
  return x;
}
async function R(t, n) {
  var e;
  void 0 === n && (n = {});
  const { x: o, y: r, platform: i, rects: a, elements: l, strategy: s } = t,
    { boundary: u = 'clippingAncestors', rootBoundary: f = 'viewport', elementContext: d = 'floating', altBoundary: m = !1, padding: p = 0 } = c(n, t),
    h = v(p),
    y = l[m ? ('floating' === d ? 'reference' : 'floating') : d],
    g = A(
      await i.getClippingRect({
        element:
          null == (e = await (null == i.isElement ? void 0 : i.isElement(y))) || e
            ? y
            : y.contextElement || (await (null == i.getDocumentElement ? void 0 : i.getDocumentElement(l.floating))),
        boundary: u,
        rootBoundary: f,
        strategy: s,
      }),
    ),
    x = 'floating' === d ? { x: o, y: r, width: a.floating.width, height: a.floating.height } : a.reference,
    w = await (null == i.getOffsetParent ? void 0 : i.getOffsetParent(l.floating)),
    b = ((await (null == i.isElement ? void 0 : i.isElement(w))) && (await (null == i.getScale ? void 0 : i.getScale(w)))) || { x: 1, y: 1 },
    S = A(
      i.convertOffsetParentRelativeRectToViewportRelativeRect
        ? await i.convertOffsetParentRelativeRectToViewportRelativeRect({ elements: l, rect: x, offsetParent: w, strategy: s })
        : x,
    );
  return { top: (g.top - S.top + h.top) / b.y, bottom: (S.bottom - g.bottom + h.bottom) / b.y, left: (g.left - S.left + h.left) / b.x, right: (S.right - g.right + h.right) / b.x };
}
function F(t, n) {
  return { top: t.top - n.height, right: t.right - n.width, bottom: t.bottom - n.height, left: t.left - n.width };
}
function O(n) {
  return t.some(t => n[t] >= 0);
}
const k = new Set(['left', 'top']);
function C() {
  return 'undefined' != typeof window;
}
function T(t) {
  return L(t) ? (t.nodeName || '').toLowerCase() : '#document';
}
function z(t) {
  var n;
  return (null == t || null == (n = t.ownerDocument) ? void 0 : n.defaultView) || window;
}
function D(t) {
  var n;
  return null == (n = (L(t) ? t.ownerDocument : t.document) || window.document) ? void 0 : n.documentElement;
}
function L(t) {
  return !!C() && (t instanceof Node || t instanceof z(t).Node);
}
function M(t) {
  return !!C() && (t instanceof Element || t instanceof z(t).Element);
}
function P(t) {
  return !!C() && (t instanceof HTMLElement || t instanceof z(t).HTMLElement);
}
function $(t) {
  return !(!C() || 'undefined' == typeof ShadowRoot) && (t instanceof ShadowRoot || t instanceof z(t).ShadowRoot);
}
function B(t) {
  const { overflow: n, overflowX: e, overflowY: o, display: r } = Y(t);
  return /auto|scroll|overlay|hidden|clip/.test(n + o + e) && 'inline' !== r && 'contents' !== r;
}
function H(t) {
  return /^(table|td|th)$/.test(T(t));
}
function E(t) {
  try {
    if (t.matches(':popover-open')) return !0;
  } catch (t) {}
  try {
    return t.matches(':modal');
  } catch (t) {
    return !1;
  }
}
const I = /transform|translate|scale|rotate|perspective|filter/,
  j = /paint|layout|strict|content/,
  N = t => !!t && 'none' !== t;
let q;
function W(t) {
  const n = M(t) ? Y(t) : t;
  return (
    N(n.transform) ||
    N(n.translate) ||
    N(n.scale) ||
    N(n.rotate) ||
    N(n.perspective) ||
    (!V() && (N(n.backdropFilter) || N(n.filter))) ||
    I.test(n.willChange || '') ||
    j.test(n.contain || '')
  );
}
function V() {
  return (null == q && (q = 'undefined' != typeof CSS && CSS.supports && CSS.supports('-webkit-backdrop-filter', 'none')), q);
}
function X(t) {
  return /^(html|body|#document)$/.test(T(t));
}
function Y(t) {
  return z(t).getComputedStyle(t);
}
function _(t) {
  return M(t) ? { scrollLeft: t.scrollLeft, scrollTop: t.scrollTop } : { scrollLeft: t.scrollX, scrollTop: t.scrollY };
}
function G(t) {
  if ('html' === T(t)) return t;
  const n = t.assignedSlot || t.parentNode || ($(t) && t.host) || D(t);
  return $(n) ? n.host : n;
}
function J(t) {
  const n = G(t);
  return X(n) ? (t.ownerDocument ? t.ownerDocument.body : t.body) : P(n) && B(n) ? n : J(n);
}
function K(t, n, e) {
  var o;
  (void 0 === n && (n = []), void 0 === e && (e = !0));
  const r = J(t),
    i = r === (null == (o = t.ownerDocument) ? void 0 : o.body),
    a = z(r);
  if (i) {
    const t = Q(a);
    return n.concat(a, a.visualViewport || [], B(r) ? r : [], t && e ? K(t) : []);
  }
  return n.concat(r, K(r, [], e));
}
function Q(t) {
  return t.parent && Object.getPrototypeOf(t.parent) ? t.frameElement : null;
}
function U(t) {
  const n = Y(t);
  let e = parseFloat(n.width) || 0,
    r = parseFloat(n.height) || 0;
  const i = P(t),
    a = i ? t.offsetWidth : e,
    l = i ? t.offsetHeight : r,
    c = o(e) !== a || o(r) !== l;
  return (c && ((e = a), (r = l)), { width: e, height: r, $: c });
}
function Z(t) {
  return M(t) ? t : t.contextElement;
}
function tt(t) {
  const n = Z(t);
  if (!P(n)) return i(1);
  const e = n.getBoundingClientRect(),
    { width: r, height: a, $: l } = U(n);
  let c = (l ? o(e.width) : e.width) / r,
    s = (l ? o(e.height) : e.height) / a;
  return ((c && Number.isFinite(c)) || (c = 1), (s && Number.isFinite(s)) || (s = 1), { x: c, y: s });
}
const nt = i(0);
function et(t) {
  const n = z(t);
  return V() && n.visualViewport ? { x: n.visualViewport.offsetLeft, y: n.visualViewport.offsetTop } : nt;
}
function ot(t, n, e, o) {
  (void 0 === n && (n = !1), void 0 === e && (e = !1));
  const r = t.getBoundingClientRect(),
    a = Z(t);
  let l = i(1);
  n && (o ? M(o) && (l = tt(o)) : (l = tt(t)));
  const c = (function (t, n, e) {
    return (void 0 === n && (n = !1), !(!e || (n && e !== z(t))) && n);
  })(a, e, o)
    ? et(a)
    : i(0);
  let s = (r.left + c.x) / l.x,
    u = (r.top + c.y) / l.y,
    f = r.width / l.x,
    d = r.height / l.y;
  if (a) {
    const t = z(a),
      n = o && M(o) ? z(o) : o;
    let e = t,
      r = Q(e);
    for (; r && o && n !== e; ) {
      const t = tt(r),
        n = r.getBoundingClientRect(),
        o = Y(r),
        i = n.left + (r.clientLeft + parseFloat(o.paddingLeft)) * t.x,
        a = n.top + (r.clientTop + parseFloat(o.paddingTop)) * t.y;
      ((s *= t.x), (u *= t.y), (f *= t.x), (d *= t.y), (s += i), (u += a), (e = z(r)), (r = Q(e)));
    }
  }
  return A({ width: f, height: d, x: s, y: u });
}
function rt(t, n) {
  const e = _(t).scrollLeft;
  return n ? n.left + e : ot(D(t)).left + e;
}
function it(t, n) {
  const e = t.getBoundingClientRect();
  return { x: e.left + n.scrollLeft - rt(t, e), y: e.top + n.scrollTop };
}
function at(t, n, o) {
  let r;
  if ('viewport' === n)
    r = (function (t, n) {
      const e = z(t),
        o = D(t),
        r = e.visualViewport;
      let i = o.clientWidth,
        a = o.clientHeight,
        l = 0,
        c = 0;
      if (r) {
        ((i = r.width), (a = r.height));
        const t = V();
        (!t || (t && 'fixed' === n)) && ((l = r.offsetLeft), (c = r.offsetTop));
      }
      const s = rt(o);
      if (s <= 0) {
        const t = o.ownerDocument,
          n = t.body,
          e = getComputedStyle(n),
          r = ('CSS1Compat' === t.compatMode && parseFloat(e.marginLeft) + parseFloat(e.marginRight)) || 0,
          a = Math.abs(o.clientWidth - n.clientWidth - r);
        a <= 25 && (i -= a);
      } else s <= 25 && (i += s);
      return { width: i, height: a, x: l, y: c };
    })(t, o);
  else if ('document' === n)
    r = (function (t) {
      const n = D(t),
        o = _(t),
        r = t.ownerDocument.body,
        i = e(n.scrollWidth, n.clientWidth, r.scrollWidth, r.clientWidth),
        a = e(n.scrollHeight, n.clientHeight, r.scrollHeight, r.clientHeight);
      let l = -o.scrollLeft + rt(t);
      const c = -o.scrollTop;
      return ('rtl' === Y(r).direction && (l += e(n.clientWidth, r.clientWidth) - i), { width: i, height: a, x: l, y: c });
    })(D(t));
  else if (M(n))
    r = (function (t, n) {
      const e = ot(t, !0, 'fixed' === n),
        o = e.top + t.clientTop,
        r = e.left + t.clientLeft,
        a = P(t) ? tt(t) : i(1);
      return { width: t.clientWidth * a.x, height: t.clientHeight * a.y, x: r * a.x, y: o * a.y };
    })(n, o);
  else {
    const e = et(t);
    r = { x: n.x - e.x, y: n.y - e.y, width: n.width, height: n.height };
  }
  return A(r);
}
function lt(t, n) {
  const e = G(t);
  return !(e === n || !M(e) || X(e)) && ('fixed' === Y(e).position || lt(e, n));
}
function ct(t, n, e) {
  const o = P(n),
    r = D(n),
    a = 'fixed' === e,
    l = ot(t, !0, a, n);
  let c = { scrollLeft: 0, scrollTop: 0 };
  const s = i(0);
  function u() {
    s.x = rt(r);
  }
  if (o || (!o && !a))
    if ((('body' !== T(n) || B(r)) && (c = _(n)), o)) {
      const t = ot(n, !0, a, n);
      ((s.x = t.x + n.clientLeft), (s.y = t.y + n.clientTop));
    } else r && u();
  a && !o && r && u();
  const f = !r || o || a ? i(0) : it(r, c);
  return { x: l.left + c.scrollLeft - s.x - f.x, y: l.top + c.scrollTop - s.y - f.y, width: l.width, height: l.height };
}
function st(t) {
  return 'static' === Y(t).position;
}
function ut(t, n) {
  if (!P(t) || 'fixed' === Y(t).position) return null;
  if (n) return n(t);
  let e = t.offsetParent;
  return (D(t) === e && (e = e.ownerDocument.body), e);
}
function ft(t, n) {
  const e = z(t);
  if (E(t)) return e;
  if (!P(t)) {
    let n = G(t);
    for (; n && !X(n); ) {
      if (M(n) && !st(n)) return n;
      n = G(n);
    }
    return e;
  }
  let o = ut(t, n);
  for (; o && H(o) && st(o); ) o = ut(o, n);
  return o && X(o) && st(o) && !W(o)
    ? e
    : o ||
        (function (t) {
          let n = G(t);
          for (; P(n) && !X(n); ) {
            if (W(n)) return n;
            if (E(n)) return null;
            n = G(n);
          }
          return null;
        })(t) ||
        e;
}
const dt = {
  convertOffsetParentRelativeRectToViewportRelativeRect: function (t) {
    let { elements: n, rect: e, offsetParent: o, strategy: r } = t;
    const a = 'fixed' === r,
      l = D(o),
      c = !!n && E(n.floating);
    if (o === l || (c && a)) return e;
    let s = { scrollLeft: 0, scrollTop: 0 },
      u = i(1);
    const f = i(0),
      d = P(o);
    if ((d || (!d && !a)) && (('body' !== T(o) || B(l)) && (s = _(o)), d)) {
      const t = ot(o);
      ((u = tt(o)), (f.x = t.x + o.clientLeft), (f.y = t.y + o.clientTop));
    }
    const m = !l || d || a ? i(0) : it(l, s);
    return { width: e.width * u.x, height: e.height * u.y, x: e.x * u.x - s.scrollLeft * u.x + f.x + m.x, y: e.y * u.y - s.scrollTop * u.y + f.y + m.y };
  },
  getDocumentElement: D,
  getClippingRect: function (t) {
    let { element: o, boundary: r, rootBoundary: i, strategy: a } = t;
    const l = [
        ...('clippingAncestors' === r
          ? E(o)
            ? []
            : (function (t, n) {
                const e = n.get(t);
                if (e) return e;
                let o = K(t, [], !1).filter(t => M(t) && 'body' !== T(t)),
                  r = null;
                const i = 'fixed' === Y(t).position;
                let a = i ? G(t) : t;
                for (; M(a) && !X(a); ) {
                  const n = Y(a),
                    e = W(a);
                  (e || 'fixed' !== n.position || (r = null),
                    (i ? !e && !r : (!e && 'static' === n.position && r && ('absolute' === r.position || 'fixed' === r.position)) || (B(a) && !e && lt(t, a)))
                      ? (o = o.filter(t => t !== a))
                      : (r = n),
                    (a = G(a)));
                }
                return (n.set(t, o), o);
              })(o, this._c)
          : [].concat(r)),
        i,
      ],
      c = at(o, l[0], a);
    let s = c.top,
      u = c.right,
      f = c.bottom,
      d = c.left;
    for (let t = 1; t < l.length; t++) {
      const r = at(o, l[t], a);
      ((s = e(r.top, s)), (u = n(r.right, u)), (f = n(r.bottom, f)), (d = e(r.left, d)));
    }
    return { width: u - d, height: f - s, x: d, y: s };
  },
  getOffsetParent: ft,
  getElementRects: async function (t) {
    const n = this.getOffsetParent || ft,
      e = this.getDimensions,
      o = await e(t.floating);
    return { reference: ct(t.reference, await n(t.floating), t.strategy), floating: { x: 0, y: 0, width: o.width, height: o.height } };
  },
  getClientRects: function (t) {
    return Array.from(t.getClientRects());
  },
  getDimensions: function (t) {
    const { width: n, height: e } = U(t);
    return { width: n, height: e };
  },
  getScale: tt,
  isElement: M,
  isRTL: function (t) {
    return 'rtl' === Y(t).direction;
  },
};
function mt(t, n) {
  return t.x === n.x && t.y === n.y && t.width === n.width && t.height === n.height;
}
const pt = function (t) {
    return (
      void 0 === t && (t = 0),
      {
        name: 'offset',
        options: t,
        async fn(n) {
          var e, o;
          const { x: r, y: i, placement: a, middlewareData: l } = n,
            f = await (async function (t, n) {
              const { placement: e, platform: o, elements: r } = t,
                i = await (null == o.isRTL ? void 0 : o.isRTL(r.floating)),
                a = s(e),
                l = u(e),
                f = 'y' === m(e),
                d = k.has(a) ? -1 : 1,
                p = i && f ? -1 : 1,
                h = c(n, t);
              let {
                mainAxis: y,
                crossAxis: g,
                alignmentAxis: x,
              } = 'number' == typeof h
                ? { mainAxis: h, crossAxis: 0, alignmentAxis: null }
                : { mainAxis: h.mainAxis || 0, crossAxis: h.crossAxis || 0, alignmentAxis: h.alignmentAxis };
              return (l && 'number' == typeof x && (g = 'end' === l ? -1 * x : x), f ? { x: g * p, y: y * d } : { x: y * d, y: g * p });
            })(n, t);
          return a === (null == (e = l.offset) ? void 0 : e.placement) && null != (o = l.arrow) && o.alignmentOffset
            ? {}
            : { x: r + f.x, y: i + f.y, data: { ...f, placement: a } };
        },
      }
    );
  },
  ht = function (t) {
    return (
      void 0 === t && (t = {}),
      {
        name: 'shift',
        options: t,
        async fn(n) {
          const { x: e, y: o, placement: r, platform: i } = n,
            {
              mainAxis: a = !0,
              crossAxis: u = !1,
              limiter: d = {
                fn: t => {
                  let { x: n, y: e } = t;
                  return { x: n, y: e };
                },
              },
              ...p
            } = c(t, n),
            h = { x: e, y: o },
            y = await i.detectOverflow(n, p),
            g = m(s(r)),
            x = f(g);
          let w = h[x],
            b = h[g];
          (a && (w = l(w + y['y' === x ? 'top' : 'left'], w, w - y['y' === x ? 'bottom' : 'right'])),
            u && (b = l(b + y['y' === g ? 'top' : 'left'], b, b - y['y' === g ? 'bottom' : 'right'])));
          const v = d.fn({ ...n, [x]: w, [g]: b });
          return { ...v, data: { x: v.x - e, y: v.y - o, enabled: { [x]: a, [g]: u } } };
        },
      }
    );
  },
  yt = function (t) {
    return (
      void 0 === t && (t = {}),
      {
        name: 'flip',
        options: t,
        async fn(n) {
          var e, o;
          const { placement: r, middlewareData: i, rects: a, initialPlacement: l, platform: f, elements: v } = n,
            {
              mainAxis: A = !0,
              crossAxis: S = !0,
              fallbackPlacements: R,
              fallbackStrategy: F = 'bestFit',
              fallbackAxisSideDirection: O = 'none',
              flipAlignment: k = !0,
              ...C
            } = c(t, n);
          if (null != (e = i.arrow) && e.alignmentOffset) return {};
          const T = s(r),
            z = m(l),
            D = s(l) === l,
            L = await (null == f.isRTL ? void 0 : f.isRTL(v.floating)),
            M =
              R ||
              (D || !k
                ? [b(l)]
                : (function (t) {
                    const n = b(t);
                    return [h(t), n, h(n)];
                  })(l)),
            P = 'none' !== O;
          !R &&
            P &&
            M.push(
              ...(function (t, n, e, o) {
                const r = u(t);
                let i = (function (t, n, e) {
                  switch (t) {
                    case 'top':
                    case 'bottom':
                      return e ? (n ? g : y) : n ? y : g;
                    case 'left':
                    case 'right':
                      return n ? x : w;
                    default:
                      return [];
                  }
                })(s(t), 'start' === e, o);
                return (r && ((i = i.map(t => t + '-' + r)), n && (i = i.concat(i.map(h)))), i);
              })(l, k, O, L),
            );
          const $ = [l, ...M],
            B = await f.detectOverflow(n, C),
            H = [];
          let E = (null == (o = i.flip) ? void 0 : o.overflows) || [];
          if ((A && H.push(B[T]), S)) {
            const t = (function (t, n, e) {
              void 0 === e && (e = !1);
              const o = u(t),
                r = p(t),
                i = d(r);
              let a = 'x' === r ? (o === (e ? 'end' : 'start') ? 'right' : 'left') : 'start' === o ? 'bottom' : 'top';
              return (n.reference[i] > n.floating[i] && (a = b(a)), [a, b(a)]);
            })(r, a, L);
            H.push(B[t[0]], B[t[1]]);
          }
          if (((E = [...E, { placement: r, overflows: H }]), !H.every(t => t <= 0))) {
            var I, j;
            const t = ((null == (I = i.flip) ? void 0 : I.index) || 0) + 1,
              n = $[t];
            if (n && ('alignment' !== S || z === m(n) || E.every(t => m(t.placement) !== z || t.overflows[0] > 0)))
              return { data: { index: t, overflows: E }, reset: { placement: n } };
            let e = null == (j = E.filter(t => t.overflows[0] <= 0).sort((t, n) => t.overflows[1] - n.overflows[1])[0]) ? void 0 : j.placement;
            if (!e)
              switch (F) {
                case 'bestFit': {
                  var N;
                  const t =
                    null ==
                    (N = E.filter(t => {
                      if (P) {
                        const n = m(t.placement);
                        return n === z || 'y' === n;
                      }
                      return !0;
                    })
                      .map(t => [t.placement, t.overflows.filter(t => t > 0).reduce((t, n) => t + n, 0)])
                      .sort((t, n) => t[1] - n[1])[0])
                      ? void 0
                      : N[0];
                  t && (e = t);
                  break;
                }
                case 'initialPlacement':
                  e = l;
              }
            if (r !== e) return { reset: { placement: e } };
          }
          return {};
        },
      }
    );
  },
  gt = function (t) {
    return (
      void 0 === t && (t = {}),
      {
        name: 'size',
        options: t,
        async fn(o) {
          var r, i;
          const { placement: a, rects: l, platform: f, elements: d } = o,
            { apply: p = () => {}, ...h } = c(t, o),
            y = await f.detectOverflow(o, h),
            g = s(a),
            x = u(a),
            w = 'y' === m(a),
            { width: b, height: v } = l.floating;
          let A, S;
          'top' === g || 'bottom' === g
            ? ((A = g), (S = x === ((await (null == f.isRTL ? void 0 : f.isRTL(d.floating))) ? 'start' : 'end') ? 'left' : 'right'))
            : ((S = g), (A = 'end' === x ? 'top' : 'bottom'));
          const R = v - y.top - y.bottom,
            F = b - y.left - y.right;
          let O = n(v - y[A], R),
            k = n(b - y[S], F);
          if ((null != (r = o.middlewareData.shift) && r.enabled.x && (k = F), null != (i = o.middlewareData.shift) && i.enabled.y && (O = R), !o.middlewareData.shift && !x)) {
            const t = e(y.left, 0),
              n = e(y.right, 0),
              o = e(y.top, 0),
              r = e(y.bottom, 0);
            w ? (k = b - 2 * (0 !== t || 0 !== n ? t + n : e(y.left, y.right))) : (O = v - 2 * (0 !== o || 0 !== r ? o + r : e(y.top, y.bottom)));
          }
          await p({ ...o, availableWidth: k, availableHeight: O });
          const C = await f.getDimensions(d.floating);
          return b !== C.width || v !== C.height ? { reset: { rects: !0 } } : {};
        },
      }
    );
  },
  xt = function (t) {
    return (
      void 0 === t && (t = {}),
      {
        name: 'hide',
        options: t,
        async fn(n) {
          const { rects: e, platform: o } = n,
            { strategy: r = 'referenceHidden', ...i } = c(t, n);
          switch (r) {
            case 'referenceHidden': {
              const t = F(await o.detectOverflow(n, { ...i, elementContext: 'reference' }), e.reference);
              return { data: { referenceHiddenOffsets: t, referenceHidden: O(t) } };
            }
            case 'escaped': {
              const t = F(await o.detectOverflow(n, { ...i, altBoundary: !0 }), e.floating);
              return { data: { escapedOffsets: t, escaped: O(t) } };
            }
            default:
              return {};
          }
        },
      }
    );
  },
  wt = t => ({
    name: 'arrow',
    options: t,
    async fn(e) {
      const { x: o, y: r, placement: i, rects: a, platform: s, elements: f, middlewareData: m } = e,
        { element: h, padding: y = 0 } = c(t, e) || {};
      if (null == h) return {};
      const g = v(y),
        x = { x: o, y: r },
        w = p(i),
        b = d(w),
        A = await s.getDimensions(h),
        S = 'y' === w,
        R = S ? 'top' : 'left',
        F = S ? 'bottom' : 'right',
        O = S ? 'clientHeight' : 'clientWidth',
        k = a.reference[b] + a.reference[w] - x[w] - a.floating[b],
        C = x[w] - a.reference[w],
        T = await (null == s.getOffsetParent ? void 0 : s.getOffsetParent(h));
      let z = T ? T[O] : 0;
      (z && (await (null == s.isElement ? void 0 : s.isElement(T)))) || (z = f.floating[O] || a.floating[b]);
      const D = k / 2 - C / 2,
        L = z / 2 - A[b] / 2 - 1,
        M = n(g[R], L),
        P = n(g[F], L),
        $ = M,
        B = z - A[b] - P,
        H = z / 2 - A[b] / 2 + D,
        E = l($, H, B),
        I = !m.arrow && null != u(i) && H !== E && a.reference[b] / 2 - (H < $ ? M : P) - A[b] / 2 < 0,
        j = I ? (H < $ ? H - $ : H - B) : 0;
      return { [w]: x[w] + j, data: { [w]: E, centerOffset: H - E - j, ...(I && { alignmentOffset: j }) }, reset: I };
    },
  }),
  bt = (t, n) => {
    Object.assign(t.style, n);
  },
  vt = (t, n = 'block') => {
    bt(t, { display: n, visibility: 'visible', opacity: '1' });
  },
  At = t => {
    bt(t, { display: 'none', visibility: 'hidden', opacity: '0' });
  };
function St(t, o, i, a) {
  return (function (t, o, i, a) {
    void 0 === a && (a = {});
    const {
        ancestorScroll: l = !0,
        ancestorResize: c = !0,
        elementResize: s = 'function' == typeof ResizeObserver,
        layoutShift: u = 'function' == typeof IntersectionObserver,
        animationFrame: f = !1,
      } = a,
      d = Z(t),
      m = l || c ? [...(d ? K(d) : []), ...(o ? K(o) : [])] : [];
    m.forEach(t => {
      (l && t.addEventListener('scroll', i, { passive: !0 }), c && t.addEventListener('resize', i));
    });
    const p =
      d && u
        ? (function (t, o) {
            let i,
              a = null;
            const l = D(t);
            function c() {
              var t;
              (clearTimeout(i), null == (t = a) || t.disconnect(), (a = null));
            }
            return (
              (function s(u, f) {
                (void 0 === u && (u = !1), void 0 === f && (f = 1), c());
                const d = t.getBoundingClientRect(),
                  { left: m, top: p, width: h, height: y } = d;
                if ((u || o(), !h || !y)) return;
                const g = { rootMargin: -r(p) + 'px ' + -r(l.clientWidth - (m + h)) + 'px ' + -r(l.clientHeight - (p + y)) + 'px ' + -r(m) + 'px', threshold: e(0, n(1, f)) || 1 };
                let x = !0;
                function w(n) {
                  const e = n[0].intersectionRatio;
                  if (e !== f) {
                    if (!x) return s();
                    e
                      ? s(!1, e)
                      : (i = setTimeout(() => {
                          s(!1, 1e-7);
                        }, 1e3));
                  }
                  (1 !== e || mt(d, t.getBoundingClientRect()) || s(), (x = !1));
                }
                try {
                  a = new IntersectionObserver(w, { ...g, root: l.ownerDocument });
                } catch (t) {
                  a = new IntersectionObserver(w, g);
                }
                a.observe(t);
              })(!0),
              c
            );
          })(d, i)
        : null;
    let h,
      y = -1,
      g = null;
    s &&
      ((g = new ResizeObserver(t => {
        let [n] = t;
        (n &&
          n.target === d &&
          g &&
          o &&
          (g.unobserve(o),
          cancelAnimationFrame(y),
          (y = requestAnimationFrame(() => {
            var t;
            null == (t = g) || t.observe(o);
          }))),
          i());
      })),
      d && !f && g.observe(d),
      o && g.observe(o));
    let x = f ? ot(t) : null;
    return (
      f &&
        (function n() {
          const e = ot(t);
          (x && !mt(x, e) && i(), (x = e), (h = requestAnimationFrame(n)));
        })(),
      i(),
      () => {
        var t;
        (m.forEach(t => {
          (l && t.removeEventListener('scroll', i), c && t.removeEventListener('resize', i));
        }),
          null == p || p(),
          null == (t = g) || t.disconnect(),
          (g = null),
          f && cancelAnimationFrame(h));
      }
    );
  })(
    t,
    o,
    () => {
      (function (t, n, e, o) {
        const { placement: r, offset: i = 8, size: a } = o || {};
        e &&
          bt(e, {
            position: 'absolute',
            width: '9px',
            height: '9px',
            background: 'inherit',
            border: '1px solid',
            borderColor: 'inherit',
            transform: 'rotate(45deg)',
            zIndex: '1300',
          });
        const l = [pt(i), yt(), ht({ padding: 5 }), ...(e ? [wt({ element: e, padding: 8 })] : []), ...(a ? [gt(a)] : []), xt()];
        return ((t, n, e) => {
          const o = new Map(),
            r = { platform: dt, ...e },
            i = { ...r.platform, _c: o };
          return (async (t, n, e) => {
            const { placement: o = 'bottom', strategy: r = 'absolute', middleware: i = [], platform: a } = e,
              l = a.detectOverflow ? a : { ...a, detectOverflow: R },
              c = await (null == a.isRTL ? void 0 : a.isRTL(n));
            let s = await a.getElementRects({ reference: t, floating: n, strategy: r }),
              { x: u, y: f } = S(s, o, c),
              d = o,
              m = 0;
            const p = {};
            for (let e = 0; e < i.length; e++) {
              const h = i[e];
              if (!h) continue;
              const { name: y, fn: g } = h,
                {
                  x,
                  y: w,
                  data: b,
                  reset: v,
                } = await g({ x: u, y: f, initialPlacement: o, placement: d, strategy: r, middlewareData: p, rects: s, platform: l, elements: { reference: t, floating: n } });
              ((u = null != x ? x : u),
                (f = null != w ? w : f),
                (p[y] = { ...p[y], ...b }),
                v &&
                  m < 50 &&
                  (m++,
                  'object' == typeof v &&
                    (v.placement && (d = v.placement),
                    v.rects && (s = !0 === v.rects ? await a.getElementRects({ reference: t, floating: n, strategy: r }) : v.rects),
                    ({ x: u, y: f } = S(s, d, c))),
                  (e = -1)));
            }
            return { x: u, y: f, placement: d, strategy: r, middlewareData: p };
          })(t, n, { ...r, platform: i });
        })(t, n, { strategy: 'fixed', placement: r, middleware: l }).then(({ x: t, y: o, middlewareData: r, placement: i }) => {
          ((n.style.left = `${t}px`), (n.style.top = `${o}px`));
          const a = i.split('-')[0],
            l = { top: 'bottom', right: 'left', bottom: 'top', left: 'right' }[a];
          if (e && r.arrow) {
            const { x: t, y: n } = r.arrow;
            ((e.dataset.side = a),
              bt(e, { left: null != t ? `${t}px` : '', top: null != n ? `${n}px` : '', right: '', bottom: '', [l]: '-5.7px' }),
              bt(
                e,
                {
                  top: { borderTop: 'none', borderLeft: 'none' },
                  bottom: { borderBottom: 'none', borderRight: 'none' },
                  left: { borderLeft: 'none', borderBottom: 'none' },
                  right: { borderRight: 'none', borderTop: 'none' },
                }[a],
              ));
          }
          return (r.hide && (r.hide.referenceHidden ? n.classList.add('floating-hidden') : n.classList.remove('floating-hidden')), i);
        });
      })(t, o, i, a).then(() => {});
    },
    { animationFrame: !0 },
  );
}
export { bt as a, St as f, At as h, vt as s };
