// @takeoff-ui/core@0.11.9/components/p-BmF3Oegw.js
var t = 'object' == typeof global && global && global.Object === Object && global,
  n = 'object' == typeof self && self && self.Object === Object && self,
  r = t || n || Function('return this')(),
  e = r.Symbol,
  o = Object.prototype,
  i = o.hasOwnProperty,
  u = o.toString,
  c = e ? e.toStringTag : void 0,
  s = Object.prototype.toString,
  a = e ? e.toStringTag : void 0;
function f(t) {
  return null == t
    ? void 0 === t
      ? '[object Undefined]'
      : '[object Null]'
    : a && a in Object(t)
      ? (function (t) {
          var n = i.call(t, c),
            r = t[c];
          try {
            t[c] = void 0;
            var e = !0;
          } catch (t) {}
          var o = u.call(t);
          return (e && (n ? (t[c] = r) : delete t[c]), o);
        })(t)
      : (function (t) {
          return s.call(t);
        })(t);
}
function h(t) {
  return null != t && 'object' == typeof t;
}
var l = Array.isArray;
function v(t) {
  var n = typeof t;
  return null != t && ('object' == n || 'function' == n);
}
function b(t) {
  if (!v(t)) return !1;
  var n = f(t);
  return '[object Function]' == n || '[object GeneratorFunction]' == n || '[object AsyncFunction]' == n || '[object Proxy]' == n;
}
var d,
  j = r['__core-js_shared__'],
  p = (d = /[^.]+$/.exec((j && j.keys && j.keys.IE_PROTO) || '')) ? 'Symbol(src)_1.' + d : '',
  y = Function.prototype.toString;
function m(t) {
  if (null != t) {
    try {
      return y.call(t);
    } catch (t) {}
    try {
      return t + '';
    } catch (t) {}
  }
  return '';
}
var _ = /^\[object .+?Constructor\]$/,
  g = RegExp(
    '^' +
      Function.prototype.toString
        .call(Object.prototype.hasOwnProperty)
        .replace(/[\\^$.*+?()[\]{}|]/g, '\\$&')
        .replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, '$1.*?') +
      '$',
  );
function O(t, n) {
  var r = (function (t, n) {
    return null == t ? void 0 : t[n];
  })(t, n);
  return (function (t) {
    return !(!v(t) || ((n = t), p && p in n)) && (b(t) ? g : _).test(m(t));
    var n;
  })(r)
    ? r
    : void 0;
}
var x = /^(?:0|[1-9]\d*)$/;
function w(t, n) {
  var r = typeof t;
  return !!(n = null == n ? 9007199254740991 : n) && ('number' == r || ('symbol' != r && x.test(t))) && t > -1 && t % 1 == 0 && t < n;
}
function $(t, n) {
  return t === n || (t != t && n != n);
}
function F(t) {
  return 'number' == typeof t && t > -1 && t % 1 == 0 && t <= 9007199254740991;
}
function A(t) {
  return null != t && F(t.length) && !b(t);
}
var S = Object.prototype;
function M(t) {
  var n = t && t.constructor;
  return t === (('function' == typeof n && n.prototype) || S);
}
function P(t) {
  return h(t) && '[object Arguments]' == f(t);
}
var U = Object.prototype,
  k = U.hasOwnProperty,
  q = U.propertyIsEnumerable,
  C = P(
    (function () {
      return arguments;
    })(),
  )
    ? P
    : function (t) {
        return h(t) && k.call(t, 'callee') && !q.call(t, 'callee');
      },
  E = 'object' == typeof exports && exports && !exports.nodeType && exports,
  G = E && 'object' == typeof module && module && !module.nodeType && module,
  L = G && G.exports === E ? r.Buffer : void 0,
  N =
    (L ? L.isBuffer : void 0) ||
    function () {
      return !1;
    },
  R = {};
((R['[object Float32Array]'] =
  R['[object Float64Array]'] =
  R['[object Int8Array]'] =
  R['[object Int16Array]'] =
  R['[object Int32Array]'] =
  R['[object Uint8Array]'] =
  R['[object Uint8ClampedArray]'] =
  R['[object Uint16Array]'] =
  R['[object Uint32Array]'] =
    !0),
  (R['[object Arguments]'] =
    R['[object Array]'] =
    R['[object ArrayBuffer]'] =
    R['[object Boolean]'] =
    R['[object DataView]'] =
    R['[object Date]'] =
    R['[object Error]'] =
    R['[object Function]'] =
    R['[object Map]'] =
    R['[object Number]'] =
    R['[object Object]'] =
    R['[object RegExp]'] =
    R['[object Set]'] =
    R['[object String]'] =
    R['[object WeakMap]'] =
      !1));
var z,
  B = 'object' == typeof exports && exports && !exports.nodeType && exports,
  D = B && 'object' == typeof module && module && !module.nodeType && module,
  H = D && D.exports === B && t.process,
  I = (function () {
    try {
      return (D && D.require && D.require('util').types) || (H && H.binding && H.binding('util'));
    } catch (t) {}
  })(),
  J = I && I.isTypedArray,
  K = J
    ? ((z = J),
      function (t) {
        return z(t);
      })
    : function (t) {
        return h(t) && F(t.length) && !!R[f(t)];
      },
  Q = Object.prototype.hasOwnProperty;
function T(t, n) {
  var r = l(t),
    e = !r && C(t),
    o = !r && !e && N(t),
    i = !r && !e && !o && K(t),
    u = r || e || o || i,
    c = u
      ? (function (t, n) {
          for (var r = -1, e = Array(t); ++r < t; ) e[r] = n(r);
          return e;
        })(t.length, String)
      : [],
    s = c.length;
  for (var a in t)
    (!n && !Q.call(t, a)) ||
      (u && ('length' == a || (o && ('offset' == a || 'parent' == a)) || (i && ('buffer' == a || 'byteLength' == a || 'byteOffset' == a)) || w(a, s))) ||
      c.push(a);
  return c;
}
function V(t, n) {
  return function (r) {
    return t(n(r));
  };
}
var W = O(Object, 'create'),
  X = Object.prototype.hasOwnProperty,
  Y = Object.prototype.hasOwnProperty;
function Z(t) {
  var n = -1,
    r = null == t ? 0 : t.length;
  for (this.clear(); ++n < r; ) {
    var e = t[n];
    this.set(e[0], e[1]);
  }
}
function tt(t, n) {
  for (var r = t.length; r--; ) if ($(t[r][0], n)) return r;
  return -1;
}
((Z.prototype.clear = function () {
  ((this.__data__ = W ? W(null) : {}), (this.size = 0));
}),
  (Z.prototype.delete = function (t) {
    var n = this.has(t) && delete this.__data__[t];
    return ((this.size -= n ? 1 : 0), n);
  }),
  (Z.prototype.get = function (t) {
    var n = this.__data__;
    if (W) {
      var r = n[t];
      return '__lodash_hash_undefined__' === r ? void 0 : r;
    }
    return X.call(n, t) ? n[t] : void 0;
  }),
  (Z.prototype.has = function (t) {
    var n = this.__data__;
    return W ? void 0 !== n[t] : Y.call(n, t);
  }),
  (Z.prototype.set = function (t, n) {
    var r = this.__data__;
    return ((this.size += this.has(t) ? 0 : 1), (r[t] = W && void 0 === n ? '__lodash_hash_undefined__' : n), this);
  }));
var nt = Array.prototype.splice;
function rt(t) {
  var n = -1,
    r = null == t ? 0 : t.length;
  for (this.clear(); ++n < r; ) {
    var e = t[n];
    this.set(e[0], e[1]);
  }
}
((rt.prototype.clear = function () {
  ((this.__data__ = []), (this.size = 0));
}),
  (rt.prototype.delete = function (t) {
    var n = this.__data__,
      r = tt(n, t);
    return !(r < 0 || (r == n.length - 1 ? n.pop() : nt.call(n, r, 1), --this.size, 0));
  }),
  (rt.prototype.get = function (t) {
    var n = this.__data__,
      r = tt(n, t);
    return r < 0 ? void 0 : n[r][1];
  }),
  (rt.prototype.has = function (t) {
    return tt(this.__data__, t) > -1;
  }),
  (rt.prototype.set = function (t, n) {
    var r = this.__data__,
      e = tt(r, t);
    return (e < 0 ? (++this.size, r.push([t, n])) : (r[e][1] = n), this);
  }));
var et = O(r, 'Map');
function ot(t, n) {
  var r,
    e,
    o = t.__data__;
  return ('string' == (e = typeof (r = n)) || 'number' == e || 'symbol' == e || 'boolean' == e ? '__proto__' !== r : null === r)
    ? o['string' == typeof n ? 'string' : 'hash']
    : o.map;
}
function it(t) {
  var n = -1,
    r = null == t ? 0 : t.length;
  for (this.clear(); ++n < r; ) {
    var e = t[n];
    this.set(e[0], e[1]);
  }
}
function ut(t) {
  var n = (this.__data__ = new rt(t));
  this.size = n.size;
}
((it.prototype.clear = function () {
  ((this.size = 0), (this.__data__ = { hash: new Z(), map: new (et || rt)(), string: new Z() }));
}),
  (it.prototype.delete = function (t) {
    var n = ot(this, t).delete(t);
    return ((this.size -= n ? 1 : 0), n);
  }),
  (it.prototype.get = function (t) {
    return ot(this, t).get(t);
  }),
  (it.prototype.has = function (t) {
    return ot(this, t).has(t);
  }),
  (it.prototype.set = function (t, n) {
    var r = ot(this, t),
      e = r.size;
    return (r.set(t, n), (this.size += r.size == e ? 0 : 1), this);
  }),
  (ut.prototype.clear = function () {
    ((this.__data__ = new rt()), (this.size = 0));
  }),
  (ut.prototype.delete = function (t) {
    var n = this.__data__,
      r = n.delete(t);
    return ((this.size = n.size), r);
  }),
  (ut.prototype.get = function (t) {
    return this.__data__.get(t);
  }),
  (ut.prototype.has = function (t) {
    return this.__data__.has(t);
  }),
  (ut.prototype.set = function (t, n) {
    var r = this.__data__;
    if (r instanceof rt) {
      var e = r.__data__;
      if (!et || e.length < 199) return (e.push([t, n]), (this.size = ++r.size), this);
      r = this.__data__ = new it(e);
    }
    return (r.set(t, n), (this.size = r.size), this);
  }));
var ct = r.Uint8Array;
export {
  et as M,
  ut as S,
  ct as U,
  A as a,
  w as b,
  M as c,
  T as d,
  $ as e,
  h as f,
  O as g,
  f as h,
  v as i,
  N as j,
  K as k,
  l,
  C as m,
  b as n,
  V as o,
  it as p,
  e as q,
  r,
  F as s,
  m as t,
};
