// @takeoff-ui/core@0.11.9/components/p-BTL1oIcv.js
import { p as t, H as e, c as o, h as a, t as l } from './p-BfLQgzDX.js';
import { c as s } from './p-D7YaJBkB.js';
import { r } from './p-Dpaq_7dX.js';
import { g as i } from './p-BDMnqEGL.js';
import { d as n } from './p-BpML1w-i.js';
const g = t(
  class extends e {
    constructor(t) {
      (super(),
        !1 !== t && this.__registerHost(),
        this.__attachShadow(),
        (this.tkToggle = o(this, 'tk-toggle', 7)),
        (this.iconPosition = 'left'),
        (this.label = ''),
        (this.size = 'base'),
        (this.variant = 'neutral'),
        (this.type = 'filled'),
        (this.selected = !1),
        (this.handleClick = t => {
          if (this.disabled) return (t.preventDefault(), void t.stopPropagation());
          this.tkToggle.emit({ value: this.value, selected: !this.selected });
        }));
    }
    getIconColor() {
      if (this.disabled) return 'var(--icon-sub-base)';
      if (this.selected) {
        if ('filled' === this.type || 'raised' === this.type) return 'var(--static-white)';
        if ('neutral' === this.variant) return 'var(--icon-darkest)';
        if ('primary' === this.variant) return 'var(--primary-base)';
      }
      return null;
    }
    createIcons() {
      return r(this.icon, { size: this.size, variant: 'neutral', additionalProps: { color: this.getIconColor() }, dataTestid: this.dataTestid }, this.iconPosition);
    }
    renderLabel() {
      var t;
      return (null === (t = this.label) || void 0 === t ? void 0 : t.length) > 0
        ? a('span', { 'class': s('tk-toggle-button-label', this.size), 'data-testid': i(this.dataTestid, 'label') }, this.label)
        : null;
    }
    render() {
      const t = s('tk-toggle-button', this.variant, this.type, this.size, { rounded: this.rounded, selected: this.selected && !this.disabled, disabled: this.disabled }),
        e = this.createIcons();
      return a(
        'button',
        { 'key': 'eb6b89c8ba9d6df6d97625d0142a0e9e73d827ae', 'class': t, 'disabled': this.disabled, 'onClick': this.handleClick, 'data-testid': i(this.dataTestid, 'button') },
        e.leftIcon,
        this.renderLabel(),
        e.rightIcon,
      );
    }
    get el() {
      return this;
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:flex;width:100%}.tk-toggle-button{display:flex;align-items:center;justify-content:center;flex:1 0 auto;border:0px;padding:var(--button-sizes-large-v-padding) var(--button-sizes-base-h-padding);gap:var(--button-sizes-base-gap);cursor:pointer;border-radius:var(--button-sizes-base-radius);background-color:transparent}.tk-toggle-button.selected{box-shadow:0 2px 2px -1px var(--shadow-black-alpha-base, rgba(82, 88, 102, 0.08)), 0 6px 12px -2px var(--shadow-black-alpha-base, rgba(82, 88, 102, 0.08)), 0 8px 20px 0 var(--shadow-black-alpha-base, rgba(82, 88, 102, 0.08))}.tk-toggle-button.neutral.filled.selected{background-color:var(--background-darkest)}.tk-toggle-button.neutral.outlined.selected{background-color:var(--background-lightest);border:1px solid var(--border-sub-base)}.tk-toggle-button.neutral.text.selected{background-color:var(--static-light)}.tk-toggle-button.neutral.raised.selected{background-color:var(--background-darkest);border:1px solid var(--static-black)}.tk-toggle-button.neutral.filled-light.selected{background-color:var(--background-light)}.tk-toggle-button.neutral.filled.selected .tk-toggle-button-label,.tk-toggle-button.neutral.raised.selected .tk-toggle-button-label{color:var(--text-lightest)}.tk-toggle-button.neutral.outlined.selected .tk-toggle-button-label,.tk-toggle-button.neutral.text.selected .tk-toggle-button-label,.tk-toggle-button.neutral.filled-light.selected .tk-toggle-button-label{color:var(--background-darkest)}.tk-toggle-button.primary.filled.selected{background-color:var(--primary-base)}.tk-toggle-button.primary.outlined.selected{background-color:var(--background-lightest);border:1px solid var(--primary-sub-base)}.tk-toggle-button.primary.text.selected{background-color:var(--static-light)}.tk-toggle-button.primary.raised.selected{background-color:var(--primary-base);border:1px solid var(--primary-sub-base)}.tk-toggle-button.primary.filled-light.selected{background-color:var(--primary-lightest)}.tk-toggle-button.primary.filled.selected .tk-toggle-button-label,.tk-toggle-button.primary.raised.selected .tk-toggle-button-label{color:var(--static-white)}.tk-toggle-button.primary.outlined.selected .tk-toggle-button-label,.tk-toggle-button.primary.text.selected .tk-toggle-button-label,.tk-toggle-button.primary.filled-light.selected .tk-toggle-button-label{color:var(--primary-base)}.tk-toggle-button .tk-toggle-button-label{color:var(--text-base);font-family:var(--desktop-body-m-base-font);font-weight:400;font-style:normal}.tk-toggle-button .tk-toggle-button-label.large{line-height:var(--desktop-label-l-line-height);font-size:var(--desktop-label-l-size)}.tk-toggle-button .tk-toggle-button-label.base{line-height:var(--desktop-label-m-base-line-height);font-size:var(--desktop-label-m-base-size)}.tk-toggle-button .tk-toggle-button-label.small{line-height:var(--desktop-label-s-line-height);font-size:var(--desktop-label-s-size)}.tk-toggle-button.small{height:0px;padding:var(--button-sizes-large-v-padding) var(--button-sizes-small-h-padding)}.tk-toggle-button.base{height:8px}.tk-toggle-button.large{height:16px;padding:var(--button-sizes-large-v-padding) var(--button-sizes-large-h-padding);gap:var(--button-sizes-large-gap)}.tk-toggle-button.rounded{border-radius:999px}.tk-toggle-button:hover .tk-toggle-button-label,.tk-toggle-button:hover tk-icon i{color:var(--text-darkest)}.tk-toggle-button.disabled{pointer-events:none;user-select:none}.tk-toggle-button.disabled .tk-toggle-button-label{color:var(--text-sub-base)}';
    }
  },
  [
    1,
    'tk-toggle-button',
    {
      disabled: [4],
      icon: [1],
      iconPosition: [1, 'icon-position'],
      label: [1],
      rounded: [4],
      size: [1],
      variant: [1],
      type: [1],
      value: [8],
      selected: [1540],
      dataTestid: [513, 'data-testid'],
    },
  ],
);
function b() {
  'undefined' != typeof customElements &&
    ['tk-toggle-button', 'tk-icon'].forEach(t => {
      switch (t) {
        case 'tk-toggle-button':
          customElements.get(l(t)) || customElements.define(l(t), g);
          break;
        case 'tk-icon':
          customElements.get(l(t)) || n();
      }
    });
}
export { g as T, b as d };
