// @takeoff-ui/core@0.11.9/components/p-BU1BjIOv.js
import { p as e, H as t, c as i, h as s, t as r } from './p-BfLQgzDX.js';
import { c as n } from './p-D7YaJBkB.js';
import { d as a } from './p-u1eXbiPL.js';
import { d as l } from './p-5F1gBr3j.js';
import { d as o } from './p-BpML1w-i.js';
const h = e(
  class extends t {
    constructor(e) {
      (super(),
        !1 !== e && this.__registerHost(),
        (this.tkItemClick = i(this, 'tk-item-click', 7)),
        (this.tkChange = i(this, 'tk-change', 7)),
        (this.tkExpandChange = i(this, 'tk-expand-change', 7)),
        (this.isAllSelected = !1),
        (this.expandedPaths = new Set()),
        (this.selectedPath = null),
        (this.isInitialLoad = !0),
        (this.items = []),
        (this.mode = 'basic'),
        (this.type = 'basic'),
        (this.size = 'base'),
        (this.disabled = !1),
        (this.branchIcon = ''),
        (this.leafIcon = ''),
        (this.showBadge = !0),
        (this.showZeroCountBadges = !0),
        (this.selectable = !1),
        (this.showPointer = !0),
        (this.selectionStrategy = 'all'),
        (this.containerStyle = null),
        (this.stepStyle = null),
        (this.selectAll = !1),
        (this.selectAllLabel = 'Select All'),
        (this.expandAll = !1),
        (this.handleToggleUnified = (e, t) => {
          const i = this.collectDescendantPaths(e, t);
          if ('stepper' === this.mode) {
            const s = e.split('-'),
              r = [];
            for (let e = 1; e <= s.length; e++) r.push(s.slice(0, e).join('-'));
            if (!this.expandedPaths.has(e)) {
              if ((this.isControlled() || (this.expandedPaths = new Set(r)), this.handleSelect(e, t), this.isControlled())) {
                const e = new Set(r),
                  t = this.getDeepestPaths(e)
                    .map(e => this.indexPathToKey(e))
                    .filter(e => null !== e);
                this.tkExpandChange.emit(t);
              }
              return;
            }
            if (
              (this.isControlled() || (this.expandedPaths.delete(e), i.forEach(e => this.expandedPaths.delete(e))),
              (e === this.selectedPath || i.includes(this.selectedPath)) && (this.selectedPath = null),
              this.isControlled())
            ) {
              const t = new Set(this.expandedPaths);
              (t.delete(e), i.forEach(e => t.delete(e)));
              const s = this.getDeepestPaths(t)
                .map(e => this.indexPathToKey(e))
                .filter(e => null !== e);
              this.tkExpandChange.emit(s);
            }
          } else if (this.expandedPaths.has(e)) {
            if (
              (this.isControlled() || (this.expandedPaths.delete(e), i.forEach(e => this.expandedPaths.delete(e))),
              (this.selectedPath === e || i.includes(this.selectedPath)) && (this.selectedPath = null),
              this.isControlled())
            ) {
              const t = new Set(this.expandedPaths);
              (t.delete(e), i.forEach(e => t.delete(e)));
              const s = this.getDeepestPaths(t)
                .map(e => this.indexPathToKey(e))
                .filter(e => null !== e);
              this.tkExpandChange.emit(s);
            }
          } else if ((this.isControlled() || this.expandedPaths.add(e), this.handleSelect(e, t), this.isControlled())) {
            const t = new Set(this.expandedPaths);
            t.add(e);
            const i = this.getDeepestPaths(t)
              .map(e => this.indexPathToKey(e))
              .filter(e => null !== e);
            this.tkExpandChange.emit(i);
          }
        }),
        (this.handleSelect = (e, t) => {
          ((this.selectedPath = e), (this.isInitialLoad = !1), this.tkItemClick.emit(t));
        }),
        (this.getAllKeys = e => {
          const t = [];
          for (const i of e) (i.key && t.push(i.key), i.children && i.children.length > 0 && t.push(...this.getAllKeys(i.children)));
          return t;
        }),
        (this.getAllChildKeys = e => {
          const t = [];
          return (
            e.children &&
              e.children.length > 0 &&
              e.children.forEach(e => {
                (t.push(e.key), t.push(...this.getAllChildKeys(e)));
              }),
            t
          );
        }),
        (this.getLeafKeys = e => {
          const t = [];
          if (!e.children || 0 === e.children.length) return (e.key && t.push(e.key), t);
          for (const i of e.children) t.push(...this.getLeafKeys(i));
          return t;
        }),
        (this.handleCheckboxChange = (e, t) => {
          var i;
          if ('all' === this.selectionStrategy) {
            const s = this.getAllChildKeys(t);
            this.value = e ? [...(this.value || []), t.key, ...s] : null === (i = this.value) || void 0 === i ? void 0 : i.filter(e => e !== t.key && !s.includes(e));
          } else {
            const i = this.getLeafKeys(t);
            this.value = e ? [...(this.value || []), ...i] : (this.value || []).filter(e => !i.includes(e));
            const s = this.items.flatMap(e => this.getLeafKeys(e));
            this.value = (this.value || []).filter(e => s.includes(e));
          }
          ((this.value = Array.from(new Set((this.value || []).filter(e => !!e)))), this.tkChange.emit(this.value));
        }),
        (this.getCheckboxState = e => {
          var t;
          const i = !e.children || 0 === e.children.length,
            s = (null === (t = this.value) || void 0 === t ? void 0 : t.includes(e.key)) || !1;
          if (i) return { checked: s, indeterminate: !1 };
          const r = e.children.map(e => this.getCheckboxState(e)),
            n = r.filter(e => e.checked || e.indeterminate),
            a = r.every(e => e.checked);
          return 'all' === this.selectionStrategy
            ? a && n.length === e.children.length
              ? { checked: !0, indeterminate: !1 }
              : n.length > 0
                ? { checked: !1, indeterminate: !0 }
                : { checked: s, indeterminate: !1 }
            : a && n.length === e.children.length
              ? { checked: !0, indeterminate: !1 }
              : n.length > 0
                ? { checked: !1, indeterminate: !0 }
                : { checked: !1, indeterminate: !1 };
        }),
        (this.getSelectedCount = e => {
          var t;
          let i = 0;
          if ((e.key && (null === (t = this.value) || void 0 === t ? void 0 : t.includes(e.key)) && (i += 1), e.children && e.children.length > 0))
            for (const t of e.children) i += this.getSelectedCount(t);
          return i;
        }),
        (this.handleItemClick = (e, t, i, s) => {
          if (!this.disabled && !i)
            if (s) this.handleToggleUnified(e, t);
            else {
              if ('stepper' === this.mode) {
                const t = e.includes('-') ? e.split('-').slice(0, -1).join('-') : '',
                  i = [];
                if (t) {
                  const e = t.split('-');
                  for (let t = 1; t <= e.length; t++) i.push(e.slice(0, t).join('-'));
                }
                if (this.isControlled()) {
                  const e = new Set(i),
                    t = this.getDeepestPaths(e)
                      .map(e => this.indexPathToKey(e))
                      .filter(e => null !== e);
                  this.tkExpandChange.emit(t);
                } else this.expandedPaths = new Set(i);
              }
              this.handleSelect(e, t);
            }
        }),
        (this.handleSelectAllClick = () => {
          this.disabled ||
            ((this.value = this.isAllSelected ? [] : 'leaf' === this.selectionStrategy ? this.items.flatMap(e => this.getLeafKeys(e)) : this.getAllKeys(this.items)),
            this.tkChange.emit(this.value));
        }),
        (this.renderItem = (e, t = '', i, r = 0) => {
          var a;
          const l = t ? `${t}-${i}` : `${i}`,
            o = !!(e.children && e.children.length > 0),
            h = this.expandedPaths.has(l),
            d = this.selectedPath === l || (this.isInitialLoad && (null === (a = this.expandedKeys) || void 0 === a ? void 0 : a.includes(e.key))),
            c = this.disabled || e.disabled,
            p = n('tk-tree-view', 'node', { directory: o, file: !o, expanded: h, selected: d, disabled: c }),
            v = this.getSelectedCount(e);
          return s(
            'div',
            { class: p },
            this.showPointer && (h || d) && s('span', { class: n('tk-tree-view', 'pointer', this.size) }),
            s(
              'div',
              {
                class: n('tk-tree-view', 'label', { selected: d, disabled: c }, this.size),
                onClick: () => {
                  this.handleItemClick(l, e, c, o);
                },
              },
              o && 'basic' === this.mode && s('tk-icon', { variant: d ? 'primary' : 'neutral', icon: h ? 'arrow_drop_down' : 'arrow_right', size: this.size }),
              this.selectable &&
                s('tk-checkbox', {
                  'onClick': e => {
                    e.stopPropagation();
                  },
                  'value': this.getCheckboxState(e).checked,
                  'indeterminate': this.getCheckboxState(e).indeterminate,
                  'disabled': c,
                  'onTk-change': t => {
                    (t.stopPropagation(), this.handleCheckboxChange(t.detail, e));
                  },
                }),
              o && this.branchIcon && s('tk-icon', { icon: this.branchIcon, variant: d ? 'primary' : 'neutral', size: this.size }),
              !o && this.leafIcon && s('tk-icon', { icon: this.leafIcon, variant: d ? 'primary' : 'neutral', size: this.size }),
              s(
                'div',
                { class: n('tk-tree-view', 'text-container', this.size) },
                s('span', { class: n('tk-tree-view', 'text', this.size) }, e.label),
                (() => {
                  var t, i, r, n, a, l, h, d, c;
                  const p = this.selectable ? v : null === (t = e.children) || void 0 === t ? void 0 : t.length;
                  return (
                    o &&
                    this.showBadge &&
                    (this.showZeroCountBadges || p > 0) &&
                    s('tk-badge', {
                      count: p,
                      size: this.size,
                      type: null !== (r = null === (i = this.badgeOptions) || void 0 === i ? void 0 : i.type) && void 0 !== r ? r : 'filledlight',
                      variant: null !== (a = null === (n = this.badgeOptions) || void 0 === n ? void 0 : n.variant) && void 0 !== a ? a : 'neutral',
                      rounded: null === (h = null === (l = this.badgeOptions) || void 0 === l ? void 0 : l.rounded) || void 0 === h || h,
                      icon: null === (d = this.badgeOptions) || void 0 === d ? void 0 : d.icon,
                      iconPosition: null === (c = this.badgeOptions) || void 0 === c ? void 0 : c.iconPosition,
                    })
                  );
                })(),
              ),
              'stepper' === this.mode &&
                o &&
                e.children &&
                e.children.length > 0 &&
                s('tk-icon', { variant: d ? 'primary' : 'neutral', icon: h ? 'keyboard_arrow_right' : 'keyboard_arrow_down', size: this.size }),
            ),
            'basic' === this.mode &&
              o &&
              h &&
              e.children &&
              e.children.length > 0 &&
              s(
                'div',
                { class: n('tk-tree-view', 'children') },
                e.children.map((e, t) => this.renderItem(e, l, t, r + 1)),
              ),
          );
        }));
    }
    itemsChanged() {
      this.initializeExpandedPaths();
    }
    modeChanged() {
      this.initializeExpandedPaths();
    }
    expandAllChanged(e, t) {
      e !== t && this.initializeExpandedPaths();
    }
    expandedKeysChanged(e) {
      if (Array.isArray(e)) {
        const t = [],
          i = [];
        if (
          (e.forEach((e, s) => {
            if ('stepper' === this.mode && 0 !== s) return;
            const r = this.findIndexPath(e);
            r ? t.push(r) : i.push(e);
          }),
          i.length > 0 && console.error('Invalid keys given to expandedKeys prop:', i),
          (this.expandedPaths = this.expandKeysWithAncestors(t)),
          null !== this.selectedPath)
        ) {
          const e = this.selectedPath.split('-');
          let t = !0;
          for (let i = 1; i <= e.length; i++) {
            const s = e.slice(0, i).join('-');
            if (i < e.length && !this.expandedPaths.has(s)) {
              t = !1;
              break;
            }
          }
          t || (this.selectedPath = null);
        }
      }
    }
    componentWillLoad() {
      this.initializeExpandedPaths();
    }
    componentWillRender() {
      this.isAllSelected = this.computeIsAllSelected();
    }
    isControlled() {
      return Array.isArray(this.expandedKeys);
    }
    findIndexPath(e, t = this.items, i = []) {
      for (let s = 0; s < t.length; s++) {
        const r = t[s],
          n = [...i, s];
        if (r.key === e) return n.join('-');
        if (r.children && r.children.length > 0) {
          const t = this.findIndexPath(e, r.children, n);
          if (t) return t;
        }
      }
      return null;
    }
    indexPathToKey(e) {
      const t = e.split('-').map(Number);
      let i = this.items;
      for (let e = 0; e < t.length; e++) {
        const s = t[e];
        if (!i || s < 0 || s >= i.length) return null;
        const r = i[s];
        if (e === t.length - 1) return r.key || null;
        i = r.children;
      }
      return null;
    }
    getDeepestPaths(e) {
      const t = Array.from(e);
      return t.filter(e => !t.some(t => t.startsWith(e + '-')));
    }
    expandKeysWithAncestors(e) {
      const t = new Set();
      return (
        e.forEach(e => {
          t.add(e);
          const i = e.split('-');
          for (let e = 1; e < i.length; e++) {
            const s = i.slice(0, e).join('-');
            t.add(s);
          }
        }),
        t
      );
    }
    collectDescendantPaths(e, t) {
      const i = [];
      return (
        t.children &&
          t.children.length > 0 &&
          t.children.forEach((t, s) => {
            const r = `${e}-${s}`;
            (i.push(r), i.push(...this.collectDescendantPaths(r, t)));
          }),
        i
      );
    }
    initializeExpandedPaths() {
      if (!this.items || 0 === this.items.length) return void (this.expandedPaths = new Set());
      if (this.isControlled()) {
        const e = [],
          t = [];
        return (
          this.expandedKeys.forEach((i, s) => {
            if ('stepper' === this.mode && 0 !== s) return;
            const r = this.findIndexPath(i);
            r ? e.push(r) : t.push(i);
          }),
          t.length > 0 && console.error('Invalid keys given to expandedKeys prop:', t),
          void (this.expandedPaths = this.expandKeysWithAncestors(e))
        );
      }
      if (!this.expandAll) return;
      const e = new Set();
      if ('basic' === this.mode) {
        const t = (i, s = '') => {
          i.forEach((i, r) => {
            const n = s ? `${s}-${r}` : `${r}`;
            i.children && i.children.length > 0 && (e.add(n), t(i.children, n));
          });
        };
        t(this.items);
      } else if ('stepper' === this.mode) {
        let t = this.items,
          i = '';
        for (; t && t.length > 0; ) {
          const s = t.findIndex(e => e.children && e.children.length > 0);
          if (-1 === s) break;
          const r = i ? `${i}-${s}` : `${s}`;
          (e.add(r), (i = r), (t = t[s].children));
        }
      }
      this.expandedPaths = e;
    }
    getStepperColumns() {
      const e = [];
      let t = this.items,
        i = '';
      for (; t && t.length > 0; ) {
        e.push({ items: t, basePath: i });
        const s = t.find((e, t) => this.expandedPaths.has(i ? `${i}-${t}` : `${t}`));
        if (!(s && s.children && s.children.length > 0)) break;
        {
          const e = t.indexOf(s);
          ((i = i ? `${i}-${e}` : `${e}`), (t = s.children));
        }
      }
      return e;
    }
    computeIsAllSelected() {
      const e = 'leaf' === this.selectionStrategy ? this.items.flatMap(e => this.getLeafKeys(e)) : this.getAllKeys(this.items);
      return (
        e.length > 0 &&
        e.every(e => {
          var t;
          return null === (t = this.value) || void 0 === t ? void 0 : t.includes(e);
        })
      );
    }
    renderSelectAll() {
      var e, t, i, r;
      if (!this.selectAll || !this.selectable) return null;
      const a = !this.isAllSelected && (null !== (t = null === (e = this.value) || void 0 === e ? void 0 : e.length) && void 0 !== t ? t : 0) > 0,
        l = null !== (r = null === (i = this.value) || void 0 === i ? void 0 : i.length) && void 0 !== r ? r : 0;
      return s(
        'div',
        { class: n('select-all', this.size, { disabled: this.disabled }), onClick: this.handleSelectAllClick },
        s(
          'div',
          { class: n('select-all-content') },
          s('tk-checkbox', {
            'value': this.isAllSelected,
            'indeterminate': a,
            'disabled': this.disabled,
            'onClick': e => e.stopPropagation(),
            'onTk-change': e => {
              (e.stopPropagation(), this.handleSelectAllClick());
            },
          }),
          s('div', null, this.selectAllLabel),
        ),
        l > 0 && s('tk-badge', { count: l, size: this.size, type: 'filledlight', variant: 'neutral', rounded: !0 }),
      );
    }
    render() {
      if (!this.items || !this.items.length) return null;
      if ('stepper' === this.mode) {
        const e = this.getStepperColumns();
        return s(
          'div',
          { class: n('tk-tree-view', 'stepper'), style: this.containerStyle },
          this.renderSelectAll(),
          e.map((e, t) =>
            s(
              'div',
              { class: n('tk-tree-view', 'column', 'divided'), key: t, style: this.stepStyle },
              e.items.map((t, i) => this.renderItem(t, e.basePath, i)),
            ),
          ),
        );
      }
      {
        const e = n('tk-tree-view', this.type, { disabled: this.disabled });
        return s(
          'div',
          { class: e, style: this.containerStyle },
          this.renderSelectAll(),
          s(
            'div',
            { class: n('tk-tree-view', 'content') },
            this.items.map((e, t) => this.renderItem(e, '', t)),
          ),
        );
      }
    }
    get el() {
      return this;
    }
    static get watchers() {
      return { items: [{ itemsChanged: 0 }], mode: [{ modeChanged: 0 }], expandAll: [{ expandAllChanged: 0 }], expandedKeys: [{ expandedKeysChanged: 0 }] };
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-tree-view{display:inline-flex;flex-direction:column;align-items:flex-start;background:transparent}.tk-tree-view::-webkit-scrollbar{width:4px;height:4px}.tk-tree-view::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-tree-view::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-tree-view::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-tree-view::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-tree-view.stepper{display:flex;flex-direction:row;align-items:flex-start;gap:var(--tree-gap, 12px);overflow-x:auto}.tk-tree-view.stepper .select-all{width:auto}.tk-tree-view.stepper .column .node .tk-tree-view.pointer{height:100%}.tk-tree-view.divided{padding:var(--tree-v-padding, 8px) var(--tree-h-padding, 10px);border-radius:8px;border:1px solid var(--border-light, #e1e4ea)}.tk-tree-view.light{padding:var(--tree-v-padding, 8px) var(--tree-h-padding, 10px);border-radius:8px;border:1px solid var(--border-light, #e1e4ea);background:var(--background-lightest, #f9fafc)}.tk-tree-view.content{width:100%;display:flex;flex-direction:column;gap:var(--tree-gap, 2px);font-family:var(--family-body)}.tk-tree-view .select-all{display:flex;flex-direction:row;align-items:center;justify-content:space-between;gap:4px;cursor:pointer;color:var(--text-dark);user-select:none;width:100%}.tk-tree-view .select-all.large{padding:var(--items-tree-large-v-padding, 10px) var(--items-tree-large-h-padding, 12px);font-size:var(--desktop-body-m-base-size, 16px);line-height:var(--desktop-body-m-base-line-height, 24px)}.tk-tree-view .select-all.base{padding:var(--items-tree-base-v-padding, 6px) var(--items-tree-base-h-padding, 8px);font-size:var(--desktop-body-s-size, 14px);line-height:var(--desktop-body-s-line-height, 20px)}.tk-tree-view .select-all.small{padding:var(--items-tree-small-v-padding, 4px) var(--items-tree-small-h-padding, 6px);font-size:var(--desktop-body-2xs-size, 11px);line-height:var(--line-height-xxs-normal, 16px)}.tk-tree-view .select-all.disabled{color:var(--text-sub-base);cursor:not-allowed}.tk-tree-view .select-all:not(.disabled):hover{color:var(--text-darkest)}.tk-tree-view .select-all .select-all-content{display:flex;flex-direction:row;align-items:center;gap:4px}.tk-tree-view.node{display:flex;align-items:flex-start;border-radius:4px;flex-direction:column;position:relative;width:100%}.tk-tree-view.node.expanded{font-weight:500;gap:var(--tree-gap, 4px)}.tk-tree-view.node.directory{font-weight:500}.tk-tree-view.node.file{font-weight:400}.tk-tree-view.node .tk-tree-view.pointer{background-color:var(--primary-base);width:4px;border-radius:0 4px 4px 0;position:absolute;left:0;top:0;z-index:1}.tk-tree-view.node .tk-tree-view.pointer.base{height:32px}.tk-tree-view.node .tk-tree-view.pointer.small{height:24px}.tk-tree-view.node .tk-tree-view.pointer.large{height:44px}.tk-tree-view.text-container{display:flex;flex-direction:row;width:100%}.tk-tree-view.text-container.large{gap:var(--items-tree-large-gap, 10px)}.tk-tree-view.text-container.base{gap:var(--items-tree-base-gap, 8px)}.tk-tree-view.text-container.small{gap:var(--items-tree-small-gap, 4px)}.tk-tree-view.label{display:flex;flex-direction:row;align-items:center;gap:4px;cursor:pointer;color:var(--text-dark, #525866);user-select:none;border-radius:4px;width:100%;justify-content:space-between}.tk-tree-view.label.large{padding:var(--items-tree-large-v-padding, 10px) var(--items-tree-large-h-padding, 12px)}.tk-tree-view.label.base{padding:var(--items-tree-base-v-padding, 6px) var(--items-tree-base-h-padding, 8px)}.tk-tree-view.label.small{padding:var(--items-tree-small-v-padding, 4px) var(--items-tree-small-h-padding, 6px)}.tk-tree-view.label.selected{background:var(--primary-lightest, #fae6e8);color:var(--text-darkest, #222530)}.tk-tree-view.label.disabled{color:var(--text-sub-base, #99a0ae);cursor:not-allowed}.tk-tree-view.label:not(.disabled,.selected):hover{color:var(--text-darkest, #222530);background:var(--background-lightest, #f9fafc)}.tk-tree-view.text{font-family:var(--desktop-body-s-font);font-style:normal;font-weight:400}.tk-tree-view.text.base{font-size:var(--body-s-size, 14px);line-height:var(--body-s-line-height, 20px)}.tk-tree-view.text.small{font-size:var(--body-2xs-size, 11px);line-height:var(--line-height-xxs-normal, 16px)}.tk-tree-view.text.large{font-size:var(--body-m-base-size, 16px);line-height:var(--body-m-base-line-height, 24px)}.tk-tree-view.children{margin-left:18px;border-left:1px solid var(--border-light, #e1e4ea);padding-left:10px;gap:2px;width:calc(100% - 18px);box-sizing:border-box;}';
    }
  },
  [
    0,
    'tk-tree-view',
    {
      items: [16],
      mode: [1],
      type: [1],
      size: [1],
      disabled: [4],
      branchIcon: [1, 'branch-icon'],
      leafIcon: [1, 'leaf-icon'],
      showBadge: [4, 'show-badge'],
      showZeroCountBadges: [4, 'show-zero-count-badges'],
      badgeOptions: [16],
      value: [16],
      selectable: [4],
      showPointer: [4, 'show-pointer'],
      selectionStrategy: [1, 'selection-strategy'],
      containerStyle: [16],
      stepStyle: [16],
      selectAll: [4, 'select-all'],
      selectAllLabel: [1, 'select-all-label'],
      expandAll: [4, 'expand-all'],
      expandedKeys: [1040],
      expandedPaths: [32],
      selectedPath: [32],
      isInitialLoad: [32],
    },
    void 0,
    { items: [{ itemsChanged: 0 }], mode: [{ modeChanged: 0 }], expandAll: [{ expandAllChanged: 0 }], expandedKeys: [{ expandedKeysChanged: 0 }] },
  ],
);
function d() {
  'undefined' != typeof customElements &&
    ['tk-tree-view', 'tk-badge', 'tk-checkbox', 'tk-icon'].forEach(e => {
      switch (e) {
        case 'tk-tree-view':
          customElements.get(r(e)) || customElements.define(r(e), h);
          break;
        case 'tk-badge':
          customElements.get(r(e)) || a();
          break;
        case 'tk-checkbox':
          customElements.get(r(e)) || l();
          break;
        case 'tk-icon':
          customElements.get(r(e)) || o();
      }
    });
}
export { h as T, d };
