// @takeoff-ui/core@0.11.9/components/p-C0U9wSBw.js
class t {
  checkConfig() {
    if (!this.config.referenceElement) throw new Error('Reference element must be provided');
    if (!this.config.handler) throw new Error('Handler must be provided');
  }
  constructor(t = {}) {
    ((this.config = { useCapture: !0, ignoredElements: [], disabled: !1 }),
      (this._isListenerActive = !1),
      (this._handleWindowClick = t => {
        var i;
        const s = t.composedPath(),
          e = s.some(t => t === this.config.referenceElement),
          h = null === (i = this.config.ignoredElements) || void 0 === i ? void 0 : i.some(t => s.some(i => i === t));
        e || h || this.config.handler(t);
      }),
      this.updateConfig(t),
      this.checkConfig());
  }
  evaluateListenerState() {
    this.config.disabled || this._isListenerActive ? this.config.disabled && this._isListenerActive && this.unbindListener() : this.bindListener();
  }
  updateConfig(t) {
    ((this.config = Object.assign(Object.assign({}, this.config), t)), this.evaluateListenerState());
  }
  bindListener() {
    this.config.disabled || this._isListenerActive || (window.addEventListener('click', this._handleWindowClick, this.config.useCapture), (this._isListenerActive = !0));
  }
  unbindListener() {
    this._isListenerActive && (window.removeEventListener('click', this._handleWindowClick, this.config.useCapture), (this._isListenerActive = !1));
  }
  disconnectedCallback() {
    this.unbindListener();
  }
}
export { t as C };
