// @takeoff-ui/core@0.11.9/components/p-CW-wkNm5.js
import { p as t, H as o, c as e, h as a, t as i } from './p-BfLQgzDX.js';
import { c as n } from './p-D7YaJBkB.js';
import { g as s } from './p-BDMnqEGL.js';
import { i as r } from './p-DzS7aqZe.js';
const l = t(
  class extends o {
    constructor(t) {
      (super(),
        !1 !== t && this.__registerHost(),
        this.__attachShadow(),
        (this.tkChange = e(this, 'tk-change', 7)),
        (this.type = 'basic'),
        (this.rounded = !1),
        (this.direction = 'horizontal'));
    }
    roundedChanged() {
      this.updateSlottedItems();
    }
    valueChanged() {
      (this.updateSelected(), this.tkChange.emit(this.value));
    }
    componentWillLoad() {
      this.updateSlottedItems();
    }
    updateSlottedItems() {
      ((this.slottedItems = this.el.querySelectorAll('tk-toggle-button')),
        this.updateSelected(),
        this.slottedItems.forEach(t => {
          ((t.rounded = this.rounded), t.addEventListener('tk-toggle', this.handleToggle.bind(this)));
        }));
    }
    updateSelected() {
      var t;
      (null === (t = this.slottedItems) || void 0 === t ? void 0 : t.length) > 0 &&
        this.slottedItems.forEach(t => {
          t.disabled ? t.selected && ((t.selected = !1), (t.value = void 0)) : (t.selected = r(this.value, t.value));
        });
    }
    handleToggle(t) {
      const { value: o } = t.detail;
      ((this.value = o), this.updateSelected());
    }
    render() {
      const t = n('tk-toggle-button-group', this.type, { rounded: this.rounded, vertical: 'vertical' === this.direction, horizontal: 'horizontal' === this.direction });
      return a(
        'div',
        { 'key': '3bb7ea989256d18ee0e94fbcaf825e0c64d4ea69', 'class': t, 'data-testid': s(this.dataTestid, 'container') },
        a('slot', { key: '88e50487b5e744a4082df70f626a41259ab41c18', onSlotchange: () => this.updateSlottedItems() }),
      );
    }
    get el() {
      return this;
    }
    static get watchers() {
      return { rounded: [{ roundedChanged: 0 }], value: [{ valueChanged: 0 }] };
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-toggle-button-group{display:flex;align-items:center;gap:8px;padding:4px;align-self:stretch;background:var(--static-light);border-radius:12px}.tk-toggle-button-group.basic{background-color:transparent}.tk-toggle-button-group.light{border:1px solid var(--border-light, #e1e4ea);background:var(--background-lightest)}.tk-toggle-button-group.divided{border:1px solid var(--border-light, #e1e4ea)}.tk-toggle-button-group.horizontal{flex-direction:row}.tk-toggle-button-group.horizontal.rounded{border-radius:999px}.tk-toggle-button-group.vertical{flex-direction:column}.tk-toggle-button-group.vertical.rounded{border-radius:24px}';
    }
  },
  [
    257,
    'tk-toggle-button-group',
    { type: [1], rounded: [4], value: [1032], direction: [1], dataTestid: [513, 'data-testid'], slottedItems: [32] },
    void 0,
    { rounded: [{ roundedChanged: 0 }], value: [{ valueChanged: 0 }] },
  ],
);
function d() {
  'undefined' != typeof customElements &&
    ['tk-toggle-button-group'].forEach(t => {
      'tk-toggle-button-group' === t && (customElements.get(i(t)) || customElements.define(i(t), l));
    });
}
export { l as T, d };
