// @takeoff-ui/core@0.11.9/components/p-DKSDLme-.js
import { p as t, H as i, c as a, h as o, d as e, t as r } from './p-BfLQgzDX.js';
import { c as n } from './p-D7YaJBkB.js';
import { v as s } from './p-CK3_k8jD.js';
const d = t(
  class extends i {
    constructor(t) {
      (super(),
        !1 !== t && this.__registerHost(),
        (this.tkChange = a(this, 'tk-change', 7)),
        (this.internals = this.attachInternals()),
        (this.uniqueId = s()),
        (this.hasContentSlot = !1),
        (this.disabled = !1),
        (this.invalid = !1),
        (this.checked = !1),
        this.el.closest('tk-radio-group') || (this.windowClickHandler = this.handleWindowClick.bind(this)));
    }
    componentWillLoad() {
      ((this.hasContentSlot = !!this.el.querySelector('[slot="content"]')),
        (this.parentEl = this.el.closest('tk-radio-group')),
        this.parentEl && !this.position && (this.position = this.parentEl.position));
    }
    componentDidRender() {
      this.bindWindowClickListener();
    }
    disconnectedCallback() {
      this.unbindWindowClickListener();
    }
    bindWindowClickListener() {
      window.addEventListener('click', this.windowClickHandler);
    }
    unbindWindowClickListener() {
      window.removeEventListener('click', this.windowClickHandler);
    }
    handleWindowClick(t) {
      var i, a;
      let o = t.target;
      ('TK-RADIO' !== o.tagName && (o = o.closest('tk-radio')),
        (null === (i = null == o ? void 0 : o.getAttribute('name')) || void 0 === i ? void 0 : i.length) > 0 &&
          (null == o ? void 0 : o.getAttribute('name')) == this.name &&
          (null === (a = null == o ? void 0 : o.getAttribute('data-tk-radio-id')) || void 0 === a ? void 0 : a.length) > 0 &&
          (null == o ? void 0 : o.getAttribute('data-tk-radio-id')) !== this.uniqueId &&
          !(null == o ? void 0 : o.getAttribute('disabled')) &&
          (this.checked = !1));
    }
    handleInputChange() {
      this.disabled || ((this.checked = !0), this.tkChange.emit(this.value));
    }
    render() {
      const t = n('tk-radio-container', this.position, { 'disabled': this.disabled, 'width-description': !!this.description });
      return o(
        e,
        { 'key': '92262fceb575a73daaccdaec498a70dd9073ede3', 'data-tk-radio-id': this.uniqueId, 'invalid': this.invalid, 'disabled': this.disabled },
        o(
          'label',
          { 'key': 'e2f683494e14e42f333b449037471d6f0f24e865', 'htmlFor': this.uniqueId, 'class': t, 'aria-disabled': this.disabled, 'aria-invalid': this.invalid },
          o('input', {
            key: '7a15a9f05c0ee1c08dc95803404caa1ee6008d59',
            id: this.uniqueId,
            type: 'radio',
            name: this.name,
            value: this.value,
            checked: this.checked,
            disabled: this.disabled,
            onChange: () => this.handleInputChange(),
          }),
          o('div', { key: 'bc1d16c1d7007e67848a67ed652931ca4b218db7', class: 'mask' }, o('div', { key: '282040082c5e42618c7bf7a28f7d43d088ef8dbf' })),
          this.hasContentSlot
            ? o('slot', { name: 'content' })
            : o(
                'div',
                { class: 'tk-radio-text-holder' },
                this.label && o('div', { class: 'tk-radio-label' }, this.label),
                this.description && o('div', { class: 'tk-radio-description' }, this.description),
              ),
        ),
      );
    }
    static get formAssociated() {
      return !0;
    }
    get el() {
      return this;
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-radio{display:inline-block}.tk-radio-container{display:flex;flex-direction:row;align-items:center;justify-content:flex-start;gap:var(--spacing-s);color:var(--text-darkest);font-family:var(--family-body);font-size:var(--size-sm);font-weight:var(--weight-medium);cursor:pointer;user-select:none}.tk-radio-container.right{flex-direction:row-reverse}.tk-radio-container.width-description{align-items:start}.tk-radio-container input{margin:0;appearance:none;outline:none;display:none}.tk-radio-container .mask{width:20px;min-width:20px;height:20px;background-color:var(--static-light);border:2px solid var(--border-light);border-radius:var(--radius-full);box-sizing:border-box;-moz-box-sizing:border-box;-webkit-box-sizing:border-box;padding:2px}.tk-radio-container .mask div{display:none;width:12px;height:12px;color:var(--static-light);border-radius:var(--radius-full);background-color:var(--states-info-base)}.tk-radio-container .tk-radio-text-holder .tk-radio-description{color:var(--text-base);font-size:var(--size-xs);font-weight:var(--weight-regular)}.tk-radio-container input:checked+.mask{border-color:var(--states-info-lightest);background-color:var(--static-light)}.tk-radio-container input:checked+.mask div{display:block}tk-radio:not([disabled]) .tk-radio-container:hover .mask{background-color:var(--background-lightest)}tk-radio:not([disabled]) .tk-radio-container input:focus-visible+.mask{border-color:var(--states-info-base)}tk-radio:not([disabled]) .tk-radio-container:hover input:checked+.mask{border-color:var(--states-info-light);background-color:var(--states-info-lightest)}tk-radio[disabled] .tk-radio-container{cursor:not-allowed;color:var(--text-sub-base);pointer-events:none}tk-radio[disabled] .tk-radio-container input+.mask{border:2px solid var(--border-light);background-color:var(--background-light)}tk-radio[disabled] .tk-radio-container input+.mask div{background-color:var(--background-sub-base)}tk-radio[invalid]:not([disabled]) .tk-radio-container .mask{border:2px solid var(--states-danger-base);background:var(--static-light)}tk-radio[invalid]:not([disabled]) .tk-radio-container input:checked+.mask{border-color:var(--states-danger-light)}tk-radio[invalid]:not([disabled]) .tk-radio-container input:checked+.mask div{background-color:var(--states-danger-base)}tk-radio[data-type=card]{border-radius:var(--radius-m-base);border:1px solid var(--border-light)}tk-radio[data-type=card] .tk-radio-container{padding:var(--card-switcher-container-v-padding) var(--card-switcher-container-h-padding)}tk-radio[data-type=card]:not([disabled]):hover{background-color:var(--background-lightest)}';
    }
  },
  [324, 'tk-radio', { disabled: [4], description: [1], invalid: [4], label: [1], position: [1], checked: [1028], name: [513], value: [8], hasContentSlot: [32] }],
);
function l() {
  'undefined' != typeof customElements &&
    ['tk-radio'].forEach(t => {
      'tk-radio' === t && (customElements.get(r(t)) || customElements.define(r(t), d));
    });
}
export { d as T, l as d };
