// @takeoff-ui/core@0.11.9/components/p-Do1x4WaI.js
import { p as t, H as a, h as e, d as n, t as o } from './p-BfLQgzDX.js';
const i = t(
  class extends a {
    constructor(t) {
      (super(), !1 !== t && this.__registerHost(), this.__attachShadow(), (this.badged = !1));
    }
    iconChanged() {
      this.emitUpdate();
    }
    labelChanged() {
      this.emitUpdate();
    }
    disabledChanged() {
      this.emitUpdate();
    }
    badgedChanged() {
      this.emitUpdate();
    }
    badgeLabelChanged() {
      this.emitUpdate();
    }
    badgeCountChanged() {
      this.emitUpdate();
    }
    badgeOptionsChanged() {
      this.emitUpdate();
    }
    tooltipOptionsChanged() {
      this.emitUpdate();
    }
    emitUpdate() {
      this.el.dispatchEvent(
        new CustomEvent('tk-update', {
          bubbles: !0,
          composed: !0,
          detail: {
            label: this.label,
            icon: this.icon,
            disabled: this.disabled,
            badged: this.badged,
            badgeCount: this.badgeCount,
            badgeLabel: this.badgeLabel,
            badgeOptions: this.badgeOptions,
            tooltipOptions: this.tooltipOptions,
          },
        }),
      );
    }
    render() {
      return e(n, { key: '633f113bba8aa0ab19534159b6654cd6b60db20f' }, e('slot', { key: '1a861b6242ada6cdc15071f1187d00480de7037b' }));
    }
    get el() {
      return this;
    }
    static get watchers() {
      return {
        icon: [{ iconChanged: 0 }],
        label: [{ labelChanged: 0 }],
        disabled: [{ disabledChanged: 0 }],
        badged: [{ badgedChanged: 0 }],
        badgeLabel: [{ badgeLabelChanged: 0 }],
        badgeCount: [{ badgeCountChanged: 0 }],
        badgeOptions: [{ badgeOptionsChanged: 0 }],
        tooltipOptions: [{ tooltipOptionsChanged: 0 }],
      };
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}';
    }
  },
  [
    257,
    'tk-tabs-item',
    { icon: [1], label: [1], disabled: [4], badged: [4], badgeLabel: [1, 'badge-label'], badgeCount: [8, 'badge-count'], badgeOptions: [16], tooltipOptions: [16] },
    void 0,
    {
      icon: [{ iconChanged: 0 }],
      label: [{ labelChanged: 0 }],
      disabled: [{ disabledChanged: 0 }],
      badged: [{ badgedChanged: 0 }],
      badgeLabel: [{ badgeLabelChanged: 0 }],
      badgeCount: [{ badgeCountChanged: 0 }],
      badgeOptions: [{ badgeOptionsChanged: 0 }],
      tooltipOptions: [{ tooltipOptionsChanged: 0 }],
    },
  ],
);
function s() {
  'undefined' != typeof customElements &&
    ['tk-tabs-item'].forEach(t => {
      'tk-tabs-item' === t && (customElements.get(o(t)) || customElements.define(o(t), i));
    });
}
export { i as T, s as d };
