// @takeoff-ui/core@0.11.9/components/p-Dpaq_7dX.js
import { h as t } from './p-BfLQgzDX.js';
import { g as i } from './p-BDMnqEGL.js';
const e = (t, i = {}, e = 'outlined', n = 'i') => {
    if (!t) return null;
    let c = Object.assign({ iconType: e, iconTag: n }, i);
    return (
      'string' == typeof t
        ? (c.icon = t)
        : ((c = Object.assign(
            Object.assign(Object.assign(Object.assign({}, c), t), {
              icon: t.name,
              iconType: t.style || i.iconType || e,
              iconTag: i.iconTag || n,
              color: t.color || i.color,
              fill: void 0 !== t.fill ? t.fill : i.fill,
            }),
            (null == t ? void 0 : t.click) ? { onClick: t.click } : {},
          )),
          null == c || delete c.style,
          null == c || delete c.name),
      c
    );
  },
  n = t => t && 'object' == typeof t && ('left' in t || 'right' in t),
  c = (c, o = {}, l = 'left') => {
    const { variant: a, sign: r = !1, fill: s = !1, size: f = 'base', iconStyle: j = 'outlined', iconTag: b = 'i', additionalProps: d = {}, dataTestid: g } = o;
    let O, p;
    if (!c) return { leftIcon: O, rightIcon: p };
    if (n(c)) {
      const n = c.left,
        o = c.right;
      (n && (O = t('tk-icon', Object.assign({}, e(n, Object.assign(Object.assign({ variant: a, sign: r, size: f, fill: s }, d), { 'data-testid': i(g, 'left-icon') }), j, b)))),
        o && (p = t('tk-icon', Object.assign({}, e(o, Object.assign(Object.assign({ variant: a, sign: r, size: f, fill: s }, d), { 'data-testid': i(g, 'right-icon') }), j, b)))));
    } else {
      const n = t('tk-icon', Object.assign({}, e(c, Object.assign(Object.assign({ variant: a, sign: r, size: f, fill: s }, d), { 'data-testid': i(g, `${l}-icon`) }), j, b)));
      'left' === l ? (O = n) : 'right' === l && (p = n);
    }
    return { leftIcon: O, rightIcon: p };
  };
export { e as g, n as i, c as r };
