// @takeoff-ui/core@0.11.9/components/p-dR-hNm6x.js
import { p as t, H as i, h as e, t as o } from './p-BfLQgzDX.js';
import { c as a } from './p-D7YaJBkB.js';
import { g as n } from './p-BDMnqEGL.js';
const r = t(
  class extends i {
    constructor(t) {
      (super(), !1 !== t && this.__registerHost(), this.__attachShadow(), (this.orientation = 'horizontal'));
    }
    componentWillLoad() {
      this.hasDefaultSlot = Array.from(this.el.childNodes).some(t => {
        var i;
        return (
          (t.nodeType === Node.ELEMENT_NODE && !t.hasAttribute('slot')) ||
          (t.nodeType === Node.TEXT_NODE && (null === (i = t.textContent) || void 0 === i ? void 0 : i.trim().length) > 0)
        );
      });
    }
    getRootStyles() {
      const t = {};
      if (this.my) {
        const i = 'number' == typeof this.my ? `${this.my}px` : this.my;
        ((t.marginTop = i), (t.marginBottom = i));
      }
      if (this.mx) {
        const i = 'number' == typeof this.mx ? `${this.mx}px` : this.mx;
        ((t.marginLeft = i), (t.marginRight = i));
      }
      return t;
    }
    render() {
      return e(
        'div',
        {
          'key': '1f2c3c461d6b7a9483bd112c4c0689248009469c',
          'class': a('tk-divider', `tk-divider-${this.orientation}`),
          'role': 'separator',
          'aria-orientation': this.orientation,
          'style': this.getRootStyles(),
          'data-testid': n(this.dataTestid, 'container'),
        },
        this.hasDefaultSlot &&
          e(
            'div',
            { 'key': 'b0f6cc048d7a580c92dd311ec7021ec90c0830d7', 'class': 'tk-divider-content', 'data-testid': n(this.dataTestid, 'content') },
            e('slot', { key: 'dde0629cd11c49f689f983241215c8212bae7ded' }),
          ),
      );
    }
    get el() {
      return this;
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-divider.tk-divider-horizontal{margin:16px 0;border-top:var(--spacing-px) solid var(--background-light)}.tk-divider.tk-divider-horizontal:has(.tk-divider-content){border-top:0 !important}.tk-divider.tk-divider-horizontal .tk-divider-content{display:flex;align-items:center;color:var(--text-base);font-family:var(--desktop-body-2xs-font);font-size:var(--desktop-body-2xs-size, 11px);font-style:normal;font-weight:400;line-height:var(--line-height-xxs-normal, 16px)}.tk-divider.tk-divider-horizontal .tk-divider-content::before{flex:1 1 0%;height:1px;content:"";border-top:var(--spacing-px) solid var(--background-light);margin-right:8px}.tk-divider.tk-divider-horizontal .tk-divider-content::after{flex:1 1 0%;content:"";border-top:var(--spacing-px) solid var(--background-light);margin-left:8px}.tk-divider.tk-divider-vertical{position:relative;display:flex;justify-content:center;min-height:100%;margin:0 16px}.tk-divider.tk-divider-vertical::before{content:"";position:absolute;display:block;height:100%;inset-block-start:0;inset-inline-start:50%;border-inline-start:var(--spacing-px) solid var(--background-light);box-sizing:border-box}';
    }
  },
  [257, 'tk-divider', { orientation: [1], mx: [8], my: [8], dataTestid: [513, 'data-testid'], hasDefaultSlot: [32] }],
);
function s() {
  'undefined' != typeof customElements &&
    ['tk-divider'].forEach(t => {
      'tk-divider' === t && (customElements.get(o(t)) || customElements.define(o(t), r));
    });
}
export { r as T, s as d };
