// @takeoff-ui/core@0.11.9/components/p-DXMmAH3A.js
import { i as r, g as n, e as t, a as e, b as o, c as u, d as i, o as a, f, h as c, U as v, j as s, k as b, l as d, m as l, n as p, S as j } from './p-BmF3Oegw.js';
import { i as m, b as y } from './p-BvYuaZU5.js';
var O,
  w,
  _,
  g = Object.create,
  h = (function () {
    function n() {}
    return function (t) {
      if (!r(t)) return {};
      if (g) return g(t);
      n.prototype = t;
      var e = new n();
      return ((n.prototype = void 0), e);
    };
  })(),
  x = Date.now,
  A = (function () {
    try {
      var r = n(Object, 'defineProperty');
      return (r({}, '', {}), r);
    } catch (r) {}
  })(),
  B = A
    ? function (r, n) {
        return A(r, 'toString', {
          configurable: !0,
          enumerable: !1,
          value:
            ((t = n),
            function () {
              return t;
            }),
          writable: !0,
        });
        var t;
      }
    : m,
  F =
    ((O = B),
    (w = 0),
    (_ = 0),
    function () {
      var r = x(),
        n = 16 - (r - _);
      if (((_ = r), n > 0)) {
        if (++w >= 800) return arguments[0];
      } else w = 0;
      return O.apply(void 0, arguments);
    });
function S(r, n, t) {
  '__proto__' == n && A ? A(r, n, { configurable: !0, enumerable: !0, value: t, writable: !0 }) : (r[n] = t);
}
var U = Object.prototype.hasOwnProperty;
function k(r, n, e) {
  var o = r[n];
  (U.call(r, n) && t(o, e) && (void 0 !== e || n in r)) || S(r, n, e);
}
var D = Math.max;
var M = Object.prototype.hasOwnProperty;
function P(n) {
  return e(n)
    ? i(n, !0)
    : (function (n) {
        if (!r(n))
          return (function (r) {
            var n = [];
            if (null != r) for (var t in Object(r)) n.push(t);
            return n;
          })(n);
        var t = u(n),
          e = [];
        for (var o in n) ('constructor' != o || (!t && M.call(n, o))) && e.push(o);
        return e;
      })(n);
}
var Y = a(Object.getPrototypeOf, Object),
  Z = Function.prototype.toString,
  q = Object.prototype.hasOwnProperty,
  z = Z.call(Object);
function C(r, n, e) {
  ((void 0 !== e && !t(r[n], e)) || (void 0 === e && !(n in r))) && S(r, n, e);
}
function E(r, n) {
  if (('constructor' !== n || 'function' != typeof r[n]) && '__proto__' != n) return r[n];
}
function G(n, t, o, i, a, j, m) {
  var y,
    O,
    w,
    _ = E(n, o),
    g = E(t, o),
    x = m.get(g);
  if (x) C(n, o, x);
  else {
    var A,
      B = j ? j(_, g, o + '', n, t, m) : void 0,
      F = void 0 === B;
    if (F) {
      var U = d(g),
        D = !U && s(g),
        M = !U && !D && b(g);
      ((B = g),
        U || D || M
          ? d(_)
            ? (B = _)
            : f((A = _)) && e(A)
              ? (B = (function (r, n) {
                  var t = -1,
                    e = r.length;
                  for (n || (n = Array(e)); ++t < e; ) n[t] = r[t];
                  return n;
                })(_))
              : D
                ? ((F = !1), (B = g.slice()))
                : M
                  ? ((F = !1), (w = new (O = (y = g).buffer).constructor(O.byteLength)), new v(w).set(new v(O)), (B = new y.constructor(w, y.byteOffset, y.length)))
                  : (B = [])
          : (function (r) {
                if (!f(r) || '[object Object]' != c(r)) return !1;
                var n = Y(r);
                if (null === n) return !0;
                var t = q.call(n, 'constructor') && n.constructor;
                return 'function' == typeof t && t instanceof t && Z.call(t) == z;
              })(g) || l(g)
            ? ((B = _),
              l(_)
                ? (B = (function (r) {
                    return (function (r, n, t) {
                      var e = !t;
                      t || (t = {});
                      for (var o = -1, u = n.length; ++o < u; ) {
                        var i = n[o],
                          a = void 0;
                        (void 0 === a && (a = r[i]), e ? S(t, i, a) : k(t, i, a));
                      }
                      return t;
                    })(r, P(r));
                  })(_))
                : (r(_) && !p(_)) ||
                  (B = (function (r) {
                    return 'function' != typeof r.constructor || u(r) ? {} : h(Y(r));
                  })(g)))
            : (F = !1));
    }
    (F && (m.set(g, B), a(B, g, i, j, m), m.delete(g)), C(n, o, B));
  }
}
function H(n, t, e, o, u) {
  n !== t &&
    y(
      t,
      function (i, a) {
        if ((u || (u = new j()), r(i))) G(n, t, a, e, H, o, u);
        else {
          var f = o ? o(E(n, a), i, a + '', n, t, u) : void 0;
          (void 0 === f && (f = i), C(n, a, f));
        }
      },
      P,
    );
}
'object' == typeof exports && exports && !exports.nodeType && exports && 'object' == typeof module && module && !module.nodeType && module;
var I,
  J =
    ((I = function (r, n, t) {
      H(r, n, t);
    }),
    (function (r, n) {
      return F(
        (function (r, n, t) {
          return (
            (n = D(void 0 === n ? r.length - 1 : n, 0)),
            function () {
              for (var e = arguments, o = -1, u = D(e.length - n, 0), i = Array(u); ++o < u; ) i[o] = e[n + o];
              o = -1;
              for (var a = Array(n + 1); ++o < n; ) a[o] = e[o];
              return (
                (a[n] = t(i)),
                (function (r, n, t) {
                  switch (t.length) {
                    case 0:
                      return r.call(n);
                    case 1:
                      return r.call(n, t[0]);
                    case 2:
                      return r.call(n, t[0], t[1]);
                    case 3:
                      return r.call(n, t[0], t[1], t[2]);
                  }
                  return r.apply(n, t);
                })(r, this, a)
              );
            }
          );
        })(r, n, m),
        r + '',
      );
    })(function (n, u) {
      var i = -1,
        a = u.length,
        f = a > 1 ? u[a - 1] : void 0,
        c = a > 2 ? u[2] : void 0;
      for (
        f = I.length > 3 && 'function' == typeof f ? (a--, f) : void 0,
          c &&
            (function (n, u, i) {
              if (!r(i)) return !1;
              var a = typeof u;
              return !!('number' == a ? e(i) && o(u, i.length) : 'string' == a && (u in i)) && t(i[u], n);
            })(u[0], u[1], c) &&
            ((f = a < 3 ? void 0 : f), (a = 1)),
          n = Object(n);
        ++i < a;
      ) {
        var v = u[i];
        v && I(n, v, i);
      }
      return n;
    }));
export { J as m };
