// @takeoff-ui/core@0.11.9/components/p-DzS7aqZe.js
import { g as r, r as e, o as n, c as t, d as a, a as c, l as u, h as o, t as i, M as f, p as s, q as v, e as b, U as j, j as w, S as l, k as p, f as _ } from './p-BmF3Oegw.js';
var d = r(e, 'WeakMap'),
  O = n(Object.keys, Object),
  h = Object.prototype.hasOwnProperty;
function k(r) {
  return c(r)
    ? a(r)
    : (function (r) {
        if (!t(r)) return O(r);
        var e = [];
        for (var n in Object(r)) h.call(r, n) && 'constructor' != n && e.push(n);
        return e;
      })(r);
}
var m = Object.prototype.propertyIsEnumerable,
  y = Object.getOwnPropertySymbols,
  A = y
    ? function (r) {
        return null == r
          ? []
          : ((r = Object(r)),
            (function (e) {
              for (var n = -1, t = null == e ? 0 : e.length, a = 0, c = []; ++n < t; ) {
                var u = e[n];
                m.call(r, u) && (c[a++] = u);
              }
              return c;
            })(y(r)));
      }
    : function () {
        return [];
      };
function S(r) {
  return (function (r, e, n) {
    var t = k(r);
    return u(r)
      ? t
      : (function (r, e) {
          for (var n = -1, t = e.length, a = r.length; ++n < t; ) r[a + n] = e[n];
          return r;
        })(t, n(r));
  })(r, 0, A);
}
var g = r(e, 'DataView'),
  M = r(e, 'Promise'),
  B = r(e, 'Set'),
  D = '[object Map]',
  V = '[object Promise]',
  x = '[object Set]',
  E = '[object WeakMap]',
  P = '[object DataView]',
  W = i(g),
  q = i(f),
  F = i(M),
  N = i(B),
  R = i(d),
  U = o;
function z(r) {
  var e = -1,
    n = null == r ? 0 : r.length;
  for (this.__data__ = new s(); ++e < n; ) this.add(r[e]);
}
function C(r, e) {
  for (var n = -1, t = null == r ? 0 : r.length; ++n < t; ) if (e(r[n], n, r)) return !0;
  return !1;
}
function G(r, e, n, t, a, c) {
  var u = 1 & n,
    o = r.length,
    i = e.length;
  if (o != i && !(u && i > o)) return !1;
  var f = c.get(r),
    s = c.get(e);
  if (f && s) return f == e && s == r;
  var v = -1,
    b = !0,
    j = 2 & n ? new z() : void 0;
  for (c.set(r, e), c.set(e, r); ++v < o; ) {
    var w = r[v],
      l = e[v];
    if (t) var p = u ? t(l, w, v, e, r, c) : t(w, l, v, r, e, c);
    if (void 0 !== p) {
      if (p) continue;
      b = !1;
      break;
    }
    if (j) {
      if (
        !C(e, function (r, e) {
          if (!j.has(e) && (w === r || a(w, r, n, t, c))) return j.push(e);
        })
      ) {
        b = !1;
        break;
      }
    } else if (w !== l && !a(w, l, n, t, c)) {
      b = !1;
      break;
    }
  }
  return (c.delete(r), c.delete(e), b);
}
function H(r) {
  var e = -1,
    n = Array(r.size);
  return (
    r.forEach(function (r, t) {
      n[++e] = [t, r];
    }),
    n
  );
}
function I(r) {
  var e = -1,
    n = Array(r.size);
  return (
    r.forEach(function (r) {
      n[++e] = r;
    }),
    n
  );
}
(((g && U(new g(new ArrayBuffer(1))) != P) || (f && U(new f()) != D) || (M && U(M.resolve()) != V) || (B && U(new B()) != x) || (d && U(new d()) != E)) &&
  (U = function (r) {
    var e = o(r),
      n = '[object Object]' == e ? r.constructor : void 0,
      t = n ? i(n) : '';
    if (t)
      switch (t) {
        case W:
          return P;
        case q:
          return D;
        case F:
          return V;
        case N:
          return x;
        case R:
          return E;
      }
    return e;
  }),
  (z.prototype.add = z.prototype.push =
    function (r) {
      return (this.__data__.set(r, '__lodash_hash_undefined__'), this);
    }),
  (z.prototype.has = function (r) {
    return this.__data__.has(r);
  }));
var J = v ? v.prototype : void 0,
  K = J ? J.valueOf : void 0,
  L = Object.prototype.hasOwnProperty,
  Q = '[object Arguments]',
  T = '[object Array]',
  X = '[object Object]',
  Y = Object.prototype.hasOwnProperty;
function Z(r, e, n, t, a) {
  return (
    r === e ||
    (null == r || null == e || (!_(r) && !_(e))
      ? r != r && e != e
      : (function (r, e, n, t, a, c) {
          var o = u(r),
            i = u(e),
            f = o ? T : U(r),
            s = i ? T : U(e),
            v = (f = f == Q ? X : f) == X,
            _ = (s = s == Q ? X : s) == X,
            d = f == s;
          if (d && w(r)) {
            if (!w(e)) return !1;
            ((o = !0), (v = !1));
          }
          if (d && !v)
            return (
              c || (c = new l()),
              o || p(r)
                ? G(r, e, n, t, a, c)
                : (function (r, e, n, t, a, c, u) {
                    switch (n) {
                      case '[object DataView]':
                        if (r.byteLength != e.byteLength || r.byteOffset != e.byteOffset) return !1;
                        ((r = r.buffer), (e = e.buffer));
                      case '[object ArrayBuffer]':
                        return !(r.byteLength != e.byteLength || !c(new j(r), new j(e)));
                      case '[object Boolean]':
                      case '[object Date]':
                      case '[object Number]':
                        return b(+r, +e);
                      case '[object Error]':
                        return r.name == e.name && r.message == e.message;
                      case '[object RegExp]':
                      case '[object String]':
                        return r == e + '';
                      case '[object Map]':
                        var o = H;
                      case '[object Set]':
                        if ((o || (o = I), r.size != e.size && !(1 & t))) return !1;
                        var i = u.get(r);
                        if (i) return i == e;
                        ((t |= 2), u.set(r, e));
                        var f = G(o(r), o(e), t, a, c, u);
                        return (u.delete(r), f);
                      case '[object Symbol]':
                        if (K) return K.call(r) == K.call(e);
                    }
                    return !1;
                  })(r, e, f, n, t, a, c)
            );
          if (!(1 & n)) {
            var O = v && Y.call(r, '__wrapped__'),
              h = _ && Y.call(e, '__wrapped__');
            if (O || h) {
              var k = O ? r.value() : r,
                m = h ? e.value() : e;
              return (c || (c = new l()), a(k, m, n, t, c));
            }
          }
          return (
            !!d &&
            (c || (c = new l()),
            (function (r, e, n, t, a, c) {
              var u = 1 & n,
                o = S(r),
                i = o.length;
              if (i != S(e).length && !u) return !1;
              for (var f = i; f--; ) {
                var s = o[f];
                if (!(u ? s in e : L.call(e, s))) return !1;
              }
              var v = c.get(r),
                b = c.get(e);
              if (v && b) return v == e && b == r;
              var j = !0;
              (c.set(r, e), c.set(e, r));
              for (var w = u; ++f < i; ) {
                var l = r[(s = o[f])],
                  p = e[s];
                if (t) var _ = u ? t(p, l, s, e, r, c) : t(l, p, s, r, e, c);
                if (!(void 0 === _ ? l === p || a(l, p, n, t, c) : _)) {
                  j = !1;
                  break;
                }
                w || (w = 'constructor' == s);
              }
              if (j && !w) {
                var d = r.constructor,
                  O = e.constructor;
                d == O || !('constructor' in r) || !('constructor' in e) || ('function' == typeof d && d instanceof d && 'function' == typeof O && O instanceof O) || (j = !1);
              }
              return (c.delete(r), c.delete(e), j);
            })(r, e, n, t, a, c))
          );
        })(r, e, n, t, Z, a))
  );
}
function $(r, e) {
  return Z(r, e);
}
export { C as a, Z as b, $ as i, k };
