// @takeoff-ui/core@0.11.9/components/p-tm7ZzNFD.js
import { p as t, H as o, h as i, F as e, t as s } from './p-BfLQgzDX.js';
import { r as n } from './p-Dpaq_7dX.js';
import { f as a } from './p-BiHzhysf.js';
import { g as l } from './p-BDMnqEGL.js';
import { d as r } from './p-BpML1w-i.js';
const d = t(
  class extends o {
    constructor(t) {
      (super(),
        !1 !== t && this.__registerHost(),
        this.__attachShadow(),
        (this.isOpen = !1),
        (this.hasContentSlot = !1),
        (this.position = 'right'),
        (this.variant = 'neutral'),
        (this.containerStyle = null),
        (this.handleMouseEnter = () => {
          this.isOpen = !0;
        }),
        (this.handleMouseLeave = () => {
          this.isOpen = !1;
        }));
    }
    positionChanged() {
      this.tooltipElement && this.updatePosition();
    }
    componentWillLoad() {
      this.hasContentSlot = !!this.el.querySelector('[slot="content"]');
    }
    componentDidLoad() {
      var t, o;
      ((this.triggerElement = this.el.querySelector('[slot="trigger"]')),
        null === (t = this.triggerElement) || void 0 === t || t.addEventListener('mouseenter', this.handleMouseEnter),
        null === (o = this.triggerElement) || void 0 === o || o.addEventListener('mouseleave', this.handleMouseLeave));
    }
    componentDidUpdate() {
      var t, o;
      this.isOpen
        ? (null === (t = this.cleanup) || void 0 === t || t.call(this), this.updatePosition())
        : (null === (o = this.cleanup) || void 0 === o || o.call(this), (this.cleanup = null));
    }
    disconnectedCallback() {
      var t;
      (null === (t = this.cleanup) || void 0 === t || t.call(this), (this.cleanup = null));
    }
    updatePosition() {
      this.cleanup = a(this.triggerElement, this.tooltipElement, this.arrowElement, { placement: this.position });
    }
    render() {
      let t, o;
      if (this.icon) {
        const { leftIcon: i, rightIcon: e } = n(this.icon, { variant: 'dark' === this.variant ? 'neutral' : this.variant, sign: !0, size: 'small', dataTestid: this.dataTestid });
        ((t = i), (o = e));
      }
      return i(
        'div',
        { 'key': '10236d0ac062c5574174974cec6da65b70489d5d', 'class': 'tk-tooltip', 'data-testid': l(this.dataTestid, 'container') },
        i('slot', { key: '2069b22ceb94133347e13dfc54d9b32c00410787', name: 'trigger' }),
        this.isOpen &&
          i(
            'div',
            {
              'key': 'cde814f061c1f107d1ca9d9e96d09ac417515281',
              'ref': t => (this.tooltipElement = t),
              'class': { 'tk-tooltip-content': !0, [`tk-tooltip-${this.variant}`]: !!this.variant },
              'style': Object.assign({}, this.containerStyle),
              'role': 'tooltip',
              'data-testid': l(this.dataTestid, 'content'),
            },
            this.hasContentSlot
              ? i('slot', { name: 'content' })
              : i(
                  e,
                  null,
                  t,
                  i(
                    'div',
                    null,
                    i('div', { 'class': 'tk-tooltip-header', 'data-testid': l(this.dataTestid, 'header') }, this.header),
                    i('div', { 'class': 'tk-tooltip-description', 'data-testid': l(this.dataTestid, 'description') }, this.description),
                  ),
                  o,
                ),
            i('div', {
              'key': 'db79b0ab1c3d5f13c081cb62d1b5646725f74349',
              'ref': t => (this.arrowElement = t),
              'class': 'tk-tooltip-arrow',
              'data-testid': l(this.dataTestid, 'arrow'),
            }),
          ),
      );
    }
    get el() {
      return this;
    }
    static get watchers() {
      return { position: [{ positionChanged: 0 }] };
    }
    static get style() {
      return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-tooltip{display:flex}.tk-tooltip-content{display:flex;gap:8px;position:fixed;color:var(--static-white);font-family:var(--family-body);padding:var(--tooltips-base-v-padding) var(--tooltips-base-h-padding);border-radius:var(--tooltips-base-radius, 8px);pointer-events:none;width:max-content;max-width:300px;z-index:1300}.tk-tooltip-content.tk-tooltip-dark{background:var(--background-darkest);line-height:20px;border:1px solid var(--border-dark)}.tk-tooltip-content.tk-tooltip-dark .tk-tooltip-header{text-align:left;color:var(--text-lightest);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-tooltip-content.tk-tooltip-dark .tk-tooltip-description{color:var(--text-sub-base);font-weight:300;font-size:var(--desktop-body-xs-size)}.tk-tooltip-content.tk-tooltip-dark .tk-tooltip-arrow{background:var(--background-darkest)}.tk-tooltip-content.tk-tooltip-white{border:1px solid var(--border-light);background:var(--background-lightest);line-height:20px}.tk-tooltip-content.tk-tooltip-white .tk-tooltip-header{text-align:left;color:var(--text-darkest);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-tooltip-content.tk-tooltip-white .tk-tooltip-description{color:var(--text-dark);font-weight:300;font-size:var(--desktop-body-xs-size)}.tk-tooltip-content.tk-tooltip-white .tk-tooltip-arrow{background:var(--background-lightest)}.tk-tooltip-content.tk-tooltip-info{border:1px solid var(--states-info-light);background-color:var(--states-info-lightest);line-height:20px}.tk-tooltip-content.tk-tooltip-info .tk-tooltip-header{text-align:left;color:var(--text-darkest);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-tooltip-content.tk-tooltip-info .tk-tooltip-description{color:var(--text-dark);font-weight:300;font-size:var(--desktop-body-xs-size)}.tk-tooltip-content.tk-tooltip-info .tk-tooltip-arrow{background-color:var(--states-info-lightest)}.tk-tooltip-content.tk-tooltip-success{border:1px solid var(--states-success-light);background-color:var(--states-success-lightest);line-height:20px}.tk-tooltip-content.tk-tooltip-success .tk-tooltip-header{text-align:left;color:var(--text-darkest);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-tooltip-content.tk-tooltip-success .tk-tooltip-description{color:var(--text-dark);font-weight:300;font-size:var(--desktop-body-xs-size)}.tk-tooltip-content.tk-tooltip-success .tk-tooltip-arrow{background-color:var(--states-success-lightest)}.tk-tooltip-content.tk-tooltip-danger{border:1px solid var(--states-danger-light);background-color:var(--states-danger-lightest);line-height:20px}.tk-tooltip-content.tk-tooltip-danger .tk-tooltip-header{text-align:left;color:var(--text-darkest);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-tooltip-content.tk-tooltip-danger .tk-tooltip-description{color:var(--text-dark);font-weight:300;font-size:var(--desktop-body-xs-size)}.tk-tooltip-content.tk-tooltip-danger .tk-tooltip-arrow{background-color:var(--states-danger-lightest)}.tk-tooltip-content.tk-tooltip-warning{border:1px solid var(--states-warning-light);background-color:var(--states-warning-lightest);line-height:20px}.tk-tooltip-content.tk-tooltip-warning .tk-tooltip-header{text-align:left;color:var(--text-darkest);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-tooltip-content.tk-tooltip-warning .tk-tooltip-description{color:var(--text-dark);font-weight:300;font-size:var(--desktop-body-xs-size)}.tk-tooltip-content.tk-tooltip-warning .tk-tooltip-arrow{background-color:var(--states-warning-lightest)}.tk-tooltip-content.tk-tooltip-neutral{border:1px solid var(--states-light);background-color:var(--background-light);line-height:20px}.tk-tooltip-content.tk-tooltip-neutral .tk-tooltip-header{text-align:left;color:var(--text-darkest);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-tooltip-content.tk-tooltip-neutral .tk-tooltip-description{color:var(--text-dark);font-weight:300;font-size:var(--desktop-body-xs-size)}.tk-tooltip-content.tk-tooltip-neutral .tk-tooltip-arrow{background-color:var(--background-light)}';
    }
  },
  [
    257,
    'tk-tooltip',
    { header: [1], description: [1], position: [1], variant: [1], icon: [1], containerStyle: [16], dataTestid: [513, 'data-testid'], isOpen: [32], hasContentSlot: [32] },
    void 0,
    { position: [{ positionChanged: 0 }] },
  ],
);
function p() {
  'undefined' != typeof customElements &&
    ['tk-tooltip', 'tk-icon'].forEach(t => {
      switch (t) {
        case 'tk-tooltip':
          customElements.get(s(t)) || customElements.define(s(t), d);
          break;
        case 'tk-icon':
          customElements.get(s(t)) || r();
      }
    });
}
export { d as T, p as d };
