// @takeoff-ui/core@0.11.9/components/p-yDaWUJCO.js
/*! @license DOMPurify 3.4.0 | (c) Cure53 and other contributors | Released under the Apache license 2.0 and Mozilla Public License 2.0 | github.com/cure53/DOMPurify/blob/3.4.0/LICENSE */
const { entries: e, setPrototypeOf: t, isFrozen: n, getPrototypeOf: i, getOwnPropertyDescriptor: o } = Object;
let { freeze: r, seal: a, create: l } = Object,
  { apply: s, construct: c } = 'undefined' != typeof Reflect && Reflect;
(r ||
  (r = function (e) {
    return e;
  }),
  a ||
    (a = function (e) {
      return e;
    }),
  s ||
    (s = function (e, t) {
      for (var n = arguments.length, i = new Array(n > 2 ? n - 2 : 0), o = 2; o < n; o++) i[o - 2] = arguments[o];
      return e.apply(t, i);
    }),
  c ||
    (c = function (e) {
      for (var t = arguments.length, n = new Array(t > 1 ? t - 1 : 0), i = 1; i < t; i++) n[i - 1] = arguments[i];
      return new e(...n);
    }));
const u = A(Array.prototype.forEach),
  f = A(Array.prototype.lastIndexOf),
  p = A(Array.prototype.pop),
  m = A(Array.prototype.push),
  d = A(Array.prototype.splice),
  g = A(String.prototype.toLowerCase),
  h = A(String.prototype.toString),
  y = A(String.prototype.match),
  b = A(String.prototype.replace),
  w = A(String.prototype.indexOf),
  x = A(String.prototype.trim),
  v = A(Object.prototype.hasOwnProperty),
  T = A(RegExp.prototype.test),
  k =
    ((S = TypeError),
    function () {
      for (var e = arguments.length, t = new Array(e), n = 0; n < e; n++) t[n] = arguments[n];
      return c(S, t);
    });
var S;
function A(e) {
  return function (t) {
    t instanceof RegExp && (t.lastIndex = 0);
    for (var n = arguments.length, i = new Array(n > 1 ? n - 1 : 0), o = 1; o < n; o++) i[o - 1] = arguments[o];
    return s(e, t, i);
  };
}
function E(e, i) {
  let o = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : g;
  t && t(e, null);
  let r = i.length;
  for (; r--; ) {
    let t = i[r];
    if ('string' == typeof t) {
      const e = o(t);
      e !== t && (n(i) || (i[r] = e), (t = e));
    }
    e[t] = !0;
  }
  return e;
}
function R(e) {
  for (let t = 0; t < e.length; t++) v(e, t) || (e[t] = null);
  return e;
}
function _(t) {
  const n = l(null);
  for (const [i, o] of e(t)) v(t, i) && (n[i] = Array.isArray(o) ? R(o) : o && 'object' == typeof o && o.constructor === Object ? _(o) : o);
  return n;
}
function z(e, t) {
  for (; null !== e; ) {
    const n = o(e, t);
    if (n) {
      if (n.get) return A(n.get);
      if ('function' == typeof n.value) return A(n.value);
    }
    e = i(e);
  }
  return function () {
    return null;
  };
}
const O = r([
    'a',
    'abbr',
    'acronym',
    'address',
    'area',
    'article',
    'aside',
    'audio',
    'b',
    'bdi',
    'bdo',
    'big',
    'blink',
    'blockquote',
    'body',
    'br',
    'button',
    'canvas',
    'caption',
    'center',
    'cite',
    'code',
    'col',
    'colgroup',
    'content',
    'data',
    'datalist',
    'dd',
    'decorator',
    'del',
    'details',
    'dfn',
    'dialog',
    'dir',
    'div',
    'dl',
    'dt',
    'element',
    'em',
    'fieldset',
    'figcaption',
    'figure',
    'font',
    'footer',
    'form',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'head',
    'header',
    'hgroup',
    'hr',
    'html',
    'i',
    'img',
    'input',
    'ins',
    'kbd',
    'label',
    'legend',
    'li',
    'main',
    'map',
    'mark',
    'marquee',
    'menu',
    'menuitem',
    'meter',
    'nav',
    'nobr',
    'ol',
    'optgroup',
    'option',
    'output',
    'p',
    'picture',
    'pre',
    'progress',
    'q',
    'rp',
    'rt',
    'ruby',
    's',
    'samp',
    'search',
    'section',
    'select',
    'shadow',
    'slot',
    'small',
    'source',
    'spacer',
    'span',
    'strike',
    'strong',
    'style',
    'sub',
    'summary',
    'sup',
    'table',
    'tbody',
    'td',
    'template',
    'textarea',
    'tfoot',
    'th',
    'thead',
    'time',
    'tr',
    'track',
    'tt',
    'u',
    'ul',
    'var',
    'video',
    'wbr',
  ]),
  D = r([
    'svg',
    'a',
    'altglyph',
    'altglyphdef',
    'altglyphitem',
    'animatecolor',
    'animatemotion',
    'animatetransform',
    'circle',
    'clippath',
    'defs',
    'desc',
    'ellipse',
    'enterkeyhint',
    'exportparts',
    'filter',
    'font',
    'g',
    'glyph',
    'glyphref',
    'hkern',
    'image',
    'inputmode',
    'line',
    'lineargradient',
    'marker',
    'mask',
    'metadata',
    'mpath',
    'part',
    'path',
    'pattern',
    'polygon',
    'polyline',
    'radialgradient',
    'rect',
    'stop',
    'style',
    'switch',
    'symbol',
    'text',
    'textpath',
    'title',
    'tref',
    'tspan',
    'view',
    'vkern',
  ]),
  L = r([
    'feBlend',
    'feColorMatrix',
    'feComponentTransfer',
    'feComposite',
    'feConvolveMatrix',
    'feDiffuseLighting',
    'feDisplacementMap',
    'feDistantLight',
    'feDropShadow',
    'feFlood',
    'feFuncA',
    'feFuncB',
    'feFuncG',
    'feFuncR',
    'feGaussianBlur',
    'feImage',
    'feMerge',
    'feMergeNode',
    'feMorphology',
    'feOffset',
    'fePointLight',
    'feSpecularLighting',
    'feSpotLight',
    'feTile',
    'feTurbulence',
  ]),
  M = r([
    'animate',
    'color-profile',
    'cursor',
    'discard',
    'font-face',
    'font-face-format',
    'font-face-name',
    'font-face-src',
    'font-face-uri',
    'foreignobject',
    'hatch',
    'hatchpath',
    'mesh',
    'meshgradient',
    'meshpatch',
    'meshrow',
    'missing-glyph',
    'script',
    'set',
    'solidcolor',
    'unknown',
    'use',
  ]),
  F = r([
    'math',
    'menclose',
    'merror',
    'mfenced',
    'mfrac',
    'mglyph',
    'mi',
    'mlabeledtr',
    'mmultiscripts',
    'mn',
    'mo',
    'mover',
    'mpadded',
    'mphantom',
    'mroot',
    'mrow',
    'ms',
    'mspace',
    'msqrt',
    'mstyle',
    'msub',
    'msup',
    'msubsup',
    'mtable',
    'mtd',
    'mtext',
    'mtr',
    'munder',
    'munderover',
    'mprescripts',
  ]),
  I = r([
    'maction',
    'maligngroup',
    'malignmark',
    'mlongdiv',
    'mscarries',
    'mscarry',
    'msgroup',
    'mstack',
    'msline',
    'msrow',
    'semantics',
    'annotation',
    'annotation-xml',
    'mprescripts',
    'none',
  ]),
  P = r(['#text']),
  N = r([
    'accept',
    'action',
    'align',
    'alt',
    'autocapitalize',
    'autocomplete',
    'autopictureinpicture',
    'autoplay',
    'background',
    'bgcolor',
    'border',
    'capture',
    'cellpadding',
    'cellspacing',
    'checked',
    'cite',
    'class',
    'clear',
    'color',
    'cols',
    'colspan',
    'controls',
    'controlslist',
    'coords',
    'crossorigin',
    'datetime',
    'decoding',
    'default',
    'dir',
    'disabled',
    'disablepictureinpicture',
    'disableremoteplayback',
    'download',
    'draggable',
    'enctype',
    'enterkeyhint',
    'exportparts',
    'face',
    'for',
    'headers',
    'height',
    'hidden',
    'high',
    'href',
    'hreflang',
    'id',
    'inert',
    'inputmode',
    'integrity',
    'ismap',
    'kind',
    'label',
    'lang',
    'list',
    'loading',
    'loop',
    'low',
    'max',
    'maxlength',
    'media',
    'method',
    'min',
    'minlength',
    'multiple',
    'muted',
    'name',
    'nonce',
    'noshade',
    'novalidate',
    'nowrap',
    'open',
    'optimum',
    'part',
    'pattern',
    'placeholder',
    'playsinline',
    'popover',
    'popovertarget',
    'popovertargetaction',
    'poster',
    'preload',
    'pubdate',
    'radiogroup',
    'readonly',
    'rel',
    'required',
    'rev',
    'reversed',
    'role',
    'rows',
    'rowspan',
    'spellcheck',
    'scope',
    'selected',
    'shape',
    'size',
    'sizes',
    'slot',
    'span',
    'srclang',
    'start',
    'src',
    'srcset',
    'step',
    'style',
    'summary',
    'tabindex',
    'title',
    'translate',
    'type',
    'usemap',
    'valign',
    'value',
    'width',
    'wrap',
    'xmlns',
    'slot',
  ]),
  C = r([
    'accent-height',
    'accumulate',
    'additive',
    'alignment-baseline',
    'amplitude',
    'ascent',
    'attributename',
    'attributetype',
    'azimuth',
    'basefrequency',
    'baseline-shift',
    'begin',
    'bias',
    'by',
    'class',
    'clip',
    'clippathunits',
    'clip-path',
    'clip-rule',
    'color',
    'color-interpolation',
    'color-interpolation-filters',
    'color-profile',
    'color-rendering',
    'cx',
    'cy',
    'd',
    'dx',
    'dy',
    'diffuseconstant',
    'direction',
    'display',
    'divisor',
    'dur',
    'edgemode',
    'elevation',
    'end',
    'exponent',
    'fill',
    'fill-opacity',
    'fill-rule',
    'filter',
    'filterunits',
    'flood-color',
    'flood-opacity',
    'font-family',
    'font-size',
    'font-size-adjust',
    'font-stretch',
    'font-style',
    'font-variant',
    'font-weight',
    'fx',
    'fy',
    'g1',
    'g2',
    'glyph-name',
    'glyphref',
    'gradientunits',
    'gradienttransform',
    'height',
    'href',
    'id',
    'image-rendering',
    'in',
    'in2',
    'intercept',
    'k',
    'k1',
    'k2',
    'k3',
    'k4',
    'kerning',
    'keypoints',
    'keysplines',
    'keytimes',
    'lang',
    'lengthadjust',
    'letter-spacing',
    'kernelmatrix',
    'kernelunitlength',
    'lighting-color',
    'local',
    'marker-end',
    'marker-mid',
    'marker-start',
    'markerheight',
    'markerunits',
    'markerwidth',
    'maskcontentunits',
    'maskunits',
    'max',
    'mask',
    'mask-type',
    'media',
    'method',
    'mode',
    'min',
    'name',
    'numoctaves',
    'offset',
    'operator',
    'opacity',
    'order',
    'orient',
    'orientation',
    'origin',
    'overflow',
    'paint-order',
    'path',
    'pathlength',
    'patterncontentunits',
    'patterntransform',
    'patternunits',
    'points',
    'preservealpha',
    'preserveaspectratio',
    'primitiveunits',
    'r',
    'rx',
    'ry',
    'radius',
    'refx',
    'refy',
    'repeatcount',
    'repeatdur',
    'restart',
    'result',
    'rotate',
    'scale',
    'seed',
    'shape-rendering',
    'slope',
    'specularconstant',
    'specularexponent',
    'spreadmethod',
    'startoffset',
    'stddeviation',
    'stitchtiles',
    'stop-color',
    'stop-opacity',
    'stroke-dasharray',
    'stroke-dashoffset',
    'stroke-linecap',
    'stroke-linejoin',
    'stroke-miterlimit',
    'stroke-opacity',
    'stroke',
    'stroke-width',
    'style',
    'surfacescale',
    'systemlanguage',
    'tabindex',
    'tablevalues',
    'targetx',
    'targety',
    'transform',
    'transform-origin',
    'text-anchor',
    'text-decoration',
    'text-rendering',
    'textlength',
    'type',
    'u1',
    'u2',
    'unicode',
    'values',
    'viewbox',
    'visibility',
    'version',
    'vert-adv-y',
    'vert-origin-x',
    'vert-origin-y',
    'width',
    'word-spacing',
    'wrap',
    'writing-mode',
    'xchannelselector',
    'ychannelselector',
    'x',
    'x1',
    'x2',
    'xmlns',
    'y',
    'y1',
    'y2',
    'z',
    'zoomandpan',
  ]),
  j = r([
    'accent',
    'accentunder',
    'align',
    'bevelled',
    'close',
    'columnalign',
    'columnlines',
    'columnspacing',
    'columnspan',
    'denomalign',
    'depth',
    'dir',
    'display',
    'displaystyle',
    'encoding',
    'fence',
    'frame',
    'height',
    'href',
    'id',
    'largeop',
    'length',
    'linethickness',
    'lquote',
    'lspace',
    'mathbackground',
    'mathcolor',
    'mathsize',
    'mathvariant',
    'maxsize',
    'minsize',
    'movablelimits',
    'notation',
    'numalign',
    'open',
    'rowalign',
    'rowlines',
    'rowspacing',
    'rowspan',
    'rspace',
    'rquote',
    'scriptlevel',
    'scriptminsize',
    'scriptsizemultiplier',
    'selection',
    'separator',
    'separators',
    'stretchy',
    'subscriptshift',
    'supscriptshift',
    'symmetric',
    'voffset',
    'width',
    'xmlns',
  ]),
  U = r(['xlink:href', 'xml:id', 'xlink:title', 'xml:space', 'xmlns:xlink']),
  B = a(/\{\{[\w\W]*|[\w\W]*\}\}/gm),
  W = a(/<%[\w\W]*|[\w\W]*%>/gm),
  H = a(/\$\{[\w\W]*/gm),
  q = a(/^data-[\-\w.\u00B7-\uFFFF]+$/),
  Y = a(/^aria-[\-\w]+$/),
  X = a(/^(?:(?:(?:f|ht)tps?|mailto|tel|callto|sms|cid|xmpp|matrix):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i),
  $ = a(/^(?:\w+script|data):/i),
  G = a(/[\u0000-\u0020\u00A0\u1680\u180E\u2000-\u2029\u205F\u3000]/g),
  K = a(/^html$/i),
  V = a(/^[a-z][.\w]*(-[.\w]+)+$/i);
var J = Object.freeze({
  __proto__: null,
  ARIA_ATTR: Y,
  ATTR_WHITESPACE: G,
  CUSTOM_ELEMENT: V,
  DATA_ATTR: q,
  DOCTYPE_NAME: K,
  ERB_EXPR: W,
  IS_ALLOWED_URI: X,
  IS_SCRIPT_OR_DATA: $,
  MUSTACHE_EXPR: B,
  TMPLIT_EXPR: H,
});
const Q = function () {
  return 'undefined' == typeof window ? null : window;
};
var Z = (function t() {
  let n = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : Q();
  const i = e => t(e);
  if (((i.version = '3.4.0'), (i.removed = []), !n || !n.document || 9 !== n.document.nodeType || !n.Element)) return ((i.isSupported = !1), i);
  let { document: o } = n;
  const a = o,
    s = a.currentScript,
    {
      DocumentFragment: c,
      HTMLTemplateElement: S,
      Node: A,
      Element: R,
      NodeFilter: B,
      NamedNodeMap: W = n.NamedNodeMap || n.MozNamedAttrMap,
      HTMLFormElement: H,
      DOMParser: q,
      trustedTypes: Y,
    } = n,
    $ = R.prototype,
    G = z($, 'cloneNode'),
    V = z($, 'remove'),
    Z = z($, 'nextSibling'),
    ee = z($, 'childNodes'),
    te = z($, 'parentNode');
  if ('function' == typeof S) {
    const e = o.createElement('template');
    e.content && e.content.ownerDocument && (o = e.content.ownerDocument);
  }
  let ne,
    ie = '';
  const { implementation: oe, createNodeIterator: re, createDocumentFragment: ae, getElementsByTagName: le } = o,
    { importNode: se } = a;
  let ce = {
    afterSanitizeAttributes: [],
    afterSanitizeElements: [],
    afterSanitizeShadowDOM: [],
    beforeSanitizeAttributes: [],
    beforeSanitizeElements: [],
    beforeSanitizeShadowDOM: [],
    uponSanitizeAttribute: [],
    uponSanitizeElement: [],
    uponSanitizeShadowNode: [],
  };
  i.isSupported = 'function' == typeof e && 'function' == typeof te && oe && void 0 !== oe.createHTMLDocument;
  const { MUSTACHE_EXPR: ue, ERB_EXPR: fe, TMPLIT_EXPR: pe, DATA_ATTR: me, ARIA_ATTR: de, IS_SCRIPT_OR_DATA: ge, ATTR_WHITESPACE: he, CUSTOM_ELEMENT: ye } = J;
  let { IS_ALLOWED_URI: be } = J,
    we = null;
  const xe = E({}, [...O, ...D, ...L, ...F, ...P]);
  let ve = null;
  const Te = E({}, [...N, ...C, ...j, ...U]);
  let ke = Object.seal(
      l(null, {
        tagNameCheck: { writable: !0, configurable: !1, enumerable: !0, value: null },
        attributeNameCheck: { writable: !0, configurable: !1, enumerable: !0, value: null },
        allowCustomizedBuiltInElements: { writable: !0, configurable: !1, enumerable: !0, value: !1 },
      }),
    ),
    Se = null,
    Ae = null;
  const Ee = Object.seal(
    l(null, { tagCheck: { writable: !0, configurable: !1, enumerable: !0, value: null }, attributeCheck: { writable: !0, configurable: !1, enumerable: !0, value: null } }),
  );
  let Re = !0,
    _e = !0,
    ze = !1,
    Oe = !0,
    De = !1,
    Le = !0,
    Me = !1,
    Fe = !1,
    Ie = !1,
    Pe = !1,
    Ne = !1,
    Ce = !1,
    je = !0,
    Ue = !1,
    Be = !0,
    We = !1,
    He = {},
    qe = null;
  const Ye = E({}, [
    'annotation-xml',
    'audio',
    'colgroup',
    'desc',
    'foreignobject',
    'head',
    'iframe',
    'math',
    'mi',
    'mn',
    'mo',
    'ms',
    'mtext',
    'noembed',
    'noframes',
    'noscript',
    'plaintext',
    'script',
    'style',
    'svg',
    'template',
    'thead',
    'title',
    'video',
    'xmp',
  ]);
  let Xe = null;
  const $e = E({}, ['audio', 'video', 'img', 'source', 'image', 'track']);
  let Ge = null;
  const Ke = E({}, ['alt', 'class', 'for', 'id', 'label', 'name', 'pattern', 'placeholder', 'role', 'summary', 'title', 'value', 'style', 'xmlns']),
    Ve = 'http://www.w3.org/1998/Math/MathML',
    Je = 'http://www.w3.org/2000/svg',
    Qe = 'http://www.w3.org/1999/xhtml';
  let Ze = Qe,
    et = !1,
    tt = null;
  const nt = E({}, [Ve, Je, Qe], h);
  let it = E({}, ['mi', 'mo', 'mn', 'ms', 'mtext']),
    ot = E({}, ['annotation-xml']);
  const rt = E({}, ['title', 'style', 'font', 'a', 'script']);
  let at = null;
  const lt = ['application/xhtml+xml', 'text/html'];
  let st = null,
    ct = null;
  const ut = o.createElement('form'),
    ft = function (e) {
      return e instanceof RegExp || e instanceof Function;
    },
    pt = function () {
      let e = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
      if (!ct || ct !== e) {
        if (
          ((e && 'object' == typeof e) || (e = {}),
          (e = _(e)),
          (at = -1 === lt.indexOf(e.PARSER_MEDIA_TYPE) ? 'text/html' : e.PARSER_MEDIA_TYPE),
          (st = 'application/xhtml+xml' === at ? h : g),
          (we = v(e, 'ALLOWED_TAGS') ? E({}, e.ALLOWED_TAGS, st) : xe),
          (ve = v(e, 'ALLOWED_ATTR') ? E({}, e.ALLOWED_ATTR, st) : Te),
          (tt = v(e, 'ALLOWED_NAMESPACES') ? E({}, e.ALLOWED_NAMESPACES, h) : nt),
          (Ge = v(e, 'ADD_URI_SAFE_ATTR') ? E(_(Ke), e.ADD_URI_SAFE_ATTR, st) : Ke),
          (Xe = v(e, 'ADD_DATA_URI_TAGS') ? E(_($e), e.ADD_DATA_URI_TAGS, st) : $e),
          (qe = v(e, 'FORBID_CONTENTS') ? E({}, e.FORBID_CONTENTS, st) : Ye),
          (Se = v(e, 'FORBID_TAGS') ? E({}, e.FORBID_TAGS, st) : _({})),
          (Ae = v(e, 'FORBID_ATTR') ? E({}, e.FORBID_ATTR, st) : _({})),
          (He = !!v(e, 'USE_PROFILES') && e.USE_PROFILES),
          (Re = !1 !== e.ALLOW_ARIA_ATTR),
          (_e = !1 !== e.ALLOW_DATA_ATTR),
          (ze = e.ALLOW_UNKNOWN_PROTOCOLS || !1),
          (Oe = !1 !== e.ALLOW_SELF_CLOSE_IN_ATTR),
          (De = e.SAFE_FOR_TEMPLATES || !1),
          (Le = !1 !== e.SAFE_FOR_XML),
          (Me = e.WHOLE_DOCUMENT || !1),
          (Pe = e.RETURN_DOM || !1),
          (Ne = e.RETURN_DOM_FRAGMENT || !1),
          (Ce = e.RETURN_TRUSTED_TYPE || !1),
          (Ie = e.FORCE_BODY || !1),
          (je = !1 !== e.SANITIZE_DOM),
          (Ue = e.SANITIZE_NAMED_PROPS || !1),
          (Be = !1 !== e.KEEP_CONTENT),
          (We = e.IN_PLACE || !1),
          (be = e.ALLOWED_URI_REGEXP || X),
          (Ze = e.NAMESPACE || Qe),
          (it = e.MATHML_TEXT_INTEGRATION_POINTS || it),
          (ot = e.HTML_INTEGRATION_POINTS || ot),
          (ke = e.CUSTOM_ELEMENT_HANDLING || l(null)),
          e.CUSTOM_ELEMENT_HANDLING && ft(e.CUSTOM_ELEMENT_HANDLING.tagNameCheck) && (ke.tagNameCheck = e.CUSTOM_ELEMENT_HANDLING.tagNameCheck),
          e.CUSTOM_ELEMENT_HANDLING && ft(e.CUSTOM_ELEMENT_HANDLING.attributeNameCheck) && (ke.attributeNameCheck = e.CUSTOM_ELEMENT_HANDLING.attributeNameCheck),
          e.CUSTOM_ELEMENT_HANDLING &&
            'boolean' == typeof e.CUSTOM_ELEMENT_HANDLING.allowCustomizedBuiltInElements &&
            (ke.allowCustomizedBuiltInElements = e.CUSTOM_ELEMENT_HANDLING.allowCustomizedBuiltInElements),
          De && (_e = !1),
          Ne && (Pe = !0),
          He &&
            ((we = E({}, P)),
            (ve = l(null)),
            !0 === He.html && (E(we, O), E(ve, N)),
            !0 === He.svg && (E(we, D), E(ve, C), E(ve, U)),
            !0 === He.svgFilters && (E(we, L), E(ve, C), E(ve, U)),
            !0 === He.mathMl && (E(we, F), E(ve, j), E(ve, U))),
          (Ee.tagCheck = null),
          (Ee.attributeCheck = null),
          e.ADD_TAGS && ('function' == typeof e.ADD_TAGS ? (Ee.tagCheck = e.ADD_TAGS) : (we === xe && (we = _(we)), E(we, e.ADD_TAGS, st))),
          e.ADD_ATTR && ('function' == typeof e.ADD_ATTR ? (Ee.attributeCheck = e.ADD_ATTR) : (ve === Te && (ve = _(ve)), E(ve, e.ADD_ATTR, st))),
          e.ADD_URI_SAFE_ATTR && E(Ge, e.ADD_URI_SAFE_ATTR, st),
          e.FORBID_CONTENTS && (qe === Ye && (qe = _(qe)), E(qe, e.FORBID_CONTENTS, st)),
          e.ADD_FORBID_CONTENTS && (qe === Ye && (qe = _(qe)), E(qe, e.ADD_FORBID_CONTENTS, st)),
          Be && (we['#text'] = !0),
          Me && E(we, ['html', 'head', 'body']),
          we.table && (E(we, ['tbody']), delete Se.tbody),
          e.TRUSTED_TYPES_POLICY)
        ) {
          if ('function' != typeof e.TRUSTED_TYPES_POLICY.createHTML) throw k('TRUSTED_TYPES_POLICY configuration option must provide a "createHTML" hook.');
          if ('function' != typeof e.TRUSTED_TYPES_POLICY.createScriptURL) throw k('TRUSTED_TYPES_POLICY configuration option must provide a "createScriptURL" hook.');
          ((ne = e.TRUSTED_TYPES_POLICY), (ie = ne.createHTML('')));
        } else
          (void 0 === ne &&
            (ne = (function (e, t) {
              if ('object' != typeof e || 'function' != typeof e.createPolicy) return null;
              let n = null;
              const i = 'data-tt-policy-suffix';
              t && t.hasAttribute(i) && (n = t.getAttribute(i));
              const o = 'dompurify' + (n ? '#' + n : '');
              try {
                return e.createPolicy(o, { createHTML: e => e, createScriptURL: e => e });
              } catch (e) {
                return (console.warn('TrustedTypes policy ' + o + ' could not be created.'), null);
              }
            })(Y, s)),
            null !== ne && 'string' == typeof ie && (ie = ne.createHTML('')));
        (r && r(e), (ct = e));
      }
    },
    mt = E({}, [...D, ...L, ...M]),
    dt = E({}, [...F, ...I]),
    gt = function (e) {
      m(i.removed, { element: e });
      try {
        te(e).removeChild(e);
      } catch (t) {
        V(e);
      }
    },
    ht = function (e, t) {
      try {
        m(i.removed, { attribute: t.getAttributeNode(e), from: t });
      } catch (e) {
        m(i.removed, { attribute: null, from: t });
      }
      if ((t.removeAttribute(e), 'is' === e))
        if (Pe || Ne)
          try {
            gt(t);
          } catch (e) {}
        else
          try {
            t.setAttribute(e, '');
          } catch (e) {}
    },
    yt = function (e) {
      let t = null,
        n = null;
      if (Ie) e = '<remove></remove>' + e;
      else {
        const t = y(e, /^[\r\n\t ]+/);
        n = t && t[0];
      }
      'application/xhtml+xml' === at && Ze === Qe && (e = '<html xmlns="http://www.w3.org/1999/xhtml"><head></head><body>' + e + '</body></html>');
      const i = ne ? ne.createHTML(e) : e;
      if (Ze === Qe)
        try {
          t = new q().parseFromString(i, at);
        } catch (e) {}
      if (!t || !t.documentElement) {
        t = oe.createDocument(Ze, 'template', null);
        try {
          t.documentElement.innerHTML = et ? ie : i;
        } catch (e) {}
      }
      const r = t.body || t.documentElement;
      return (e && n && r.insertBefore(o.createTextNode(n), r.childNodes[0] || null), Ze === Qe ? le.call(t, Me ? 'html' : 'body')[0] : Me ? t.documentElement : r);
    },
    bt = function (e) {
      return re.call(e.ownerDocument || e, e, B.SHOW_ELEMENT | B.SHOW_COMMENT | B.SHOW_TEXT | B.SHOW_PROCESSING_INSTRUCTION | B.SHOW_CDATA_SECTION, null);
    },
    wt = function (e) {
      return (
        e instanceof H &&
        ('string' != typeof e.nodeName ||
          'string' != typeof e.textContent ||
          'function' != typeof e.removeChild ||
          !(e.attributes instanceof W) ||
          'function' != typeof e.removeAttribute ||
          'function' != typeof e.setAttribute ||
          'string' != typeof e.namespaceURI ||
          'function' != typeof e.insertBefore ||
          'function' != typeof e.hasChildNodes)
      );
    },
    xt = function (e) {
      return 'function' == typeof A && e instanceof A;
    };
  function vt(e, t, n) {
    u(e, e => {
      e.call(i, t, n, ct);
    });
  }
  const Tt = function (e) {
      let t = null;
      if ((vt(ce.beforeSanitizeElements, e, null), wt(e))) return (gt(e), !0);
      const n = st(e.nodeName);
      if (
        (vt(ce.uponSanitizeElement, e, { tagName: n, allowedTags: we }),
        Le && e.hasChildNodes() && !xt(e.firstElementChild) && T(/<[/\w!]/g, e.innerHTML) && T(/<[/\w!]/g, e.textContent))
      )
        return (gt(e), !0);
      if (Le && e.namespaceURI === Qe && 'style' === n && xt(e.firstElementChild)) return (gt(e), !0);
      if (7 === e.nodeType) return (gt(e), !0);
      if (Le && 8 === e.nodeType && T(/<[/\w]/g, e.data)) return (gt(e), !0);
      if (Se[n] || (!(Ee.tagCheck instanceof Function && Ee.tagCheck(n)) && !we[n])) {
        if (!Se[n] && St(n)) {
          if (ke.tagNameCheck instanceof RegExp && T(ke.tagNameCheck, n)) return !1;
          if (ke.tagNameCheck instanceof Function && ke.tagNameCheck(n)) return !1;
        }
        if (Be && !qe[n]) {
          const t = te(e) || e.parentNode,
            n = ee(e) || e.childNodes;
          if (n && t)
            for (let i = n.length - 1; i >= 0; --i) {
              const o = G(n[i], !0);
              ((o.__removalCount = (e.__removalCount || 0) + 1), t.insertBefore(o, Z(e)));
            }
        }
        return (gt(e), !0);
      }
      return e instanceof R &&
        !(function (e) {
          let t = te(e);
          (t && t.tagName) || (t = { namespaceURI: Ze, tagName: 'template' });
          const n = g(e.tagName),
            i = g(t.tagName);
          return (
            !!tt[e.namespaceURI] &&
            (e.namespaceURI === Je
              ? t.namespaceURI === Qe
                ? 'svg' === n
                : t.namespaceURI === Ve
                  ? 'svg' === n && ('annotation-xml' === i || it[i])
                  : Boolean(mt[n])
              : e.namespaceURI === Ve
                ? t.namespaceURI === Qe
                  ? 'math' === n
                  : t.namespaceURI === Je
                    ? 'math' === n && ot[i]
                    : Boolean(dt[n])
                : e.namespaceURI === Qe
                  ? !(t.namespaceURI === Je && !ot[i]) && !(t.namespaceURI === Ve && !it[i]) && !dt[n] && (rt[n] || !mt[n])
                  : !('application/xhtml+xml' !== at || !tt[e.namespaceURI]))
          );
        })(e)
        ? (gt(e), !0)
        : ('noscript' !== n && 'noembed' !== n && 'noframes' !== n) || !T(/<\/no(script|embed|frames)/i, e.innerHTML)
          ? (De &&
              3 === e.nodeType &&
              ((t = e.textContent),
              u([ue, fe, pe], e => {
                t = b(t, e, ' ');
              }),
              e.textContent !== t && (m(i.removed, { element: e.cloneNode() }), (e.textContent = t))),
            vt(ce.afterSanitizeElements, e, null),
            !1)
          : (gt(e), !0);
    },
    kt = function (e, t, n) {
      if (Ae[t]) return !1;
      if (je && ('id' === t || 'name' === t) && (n in o || n in ut)) return !1;
      if (_e && !Ae[t] && T(me, t));
      else if (Re && T(de, t));
      else if (Ee.attributeCheck instanceof Function && Ee.attributeCheck(t, e));
      else if (!ve[t] || Ae[t]) {
        if (
          !(
            (St(e) &&
              ((ke.tagNameCheck instanceof RegExp && T(ke.tagNameCheck, e)) || (ke.tagNameCheck instanceof Function && ke.tagNameCheck(e))) &&
              ((ke.attributeNameCheck instanceof RegExp && T(ke.attributeNameCheck, t)) || (ke.attributeNameCheck instanceof Function && ke.attributeNameCheck(t, e)))) ||
            ('is' === t &&
              ke.allowCustomizedBuiltInElements &&
              ((ke.tagNameCheck instanceof RegExp && T(ke.tagNameCheck, n)) || (ke.tagNameCheck instanceof Function && ke.tagNameCheck(n))))
          )
        )
          return !1;
      } else if (Ge[t]);
      else if (T(be, b(n, he, '')));
      else if (('src' !== t && 'xlink:href' !== t && 'href' !== t) || 'script' === e || 0 !== w(n, 'data:') || !Xe[e])
        if (ze && !T(ge, b(n, he, '')));
        else if (n) return !1;
      return !0;
    },
    St = function (e) {
      return 'annotation-xml' !== e && y(e, ye);
    },
    At = function (e) {
      vt(ce.beforeSanitizeAttributes, e, null);
      const { attributes: t } = e;
      if (!t || wt(e)) return;
      const n = { attrName: '', attrValue: '', keepAttr: !0, allowedAttributes: ve, forceKeepAttr: void 0 };
      let o = t.length;
      for (; o--; ) {
        const r = t[o],
          { name: a, namespaceURI: l, value: s } = r,
          c = st(a),
          f = s;
        let m = 'value' === a ? f : x(f);
        if (
          ((n.attrName = c),
          (n.attrValue = m),
          (n.keepAttr = !0),
          (n.forceKeepAttr = void 0),
          vt(ce.uponSanitizeAttribute, e, n),
          (m = n.attrValue),
          !Ue || ('id' !== c && 'name' !== c) || (ht(a, e), (m = 'user-content-' + m)),
          Le && T(/((--!?|])>)|<\/(style|script|title|xmp|textarea|noscript|iframe|noembed|noframes)/i, m))
        ) {
          ht(a, e);
          continue;
        }
        if ('attributename' === c && y(m, 'href')) {
          ht(a, e);
          continue;
        }
        if (n.forceKeepAttr) continue;
        if (!n.keepAttr) {
          ht(a, e);
          continue;
        }
        if (!Oe && T(/\/>/i, m)) {
          ht(a, e);
          continue;
        }
        De &&
          u([ue, fe, pe], e => {
            m = b(m, e, ' ');
          });
        const d = st(e.nodeName);
        if (kt(d, c, m)) {
          if (ne && 'object' == typeof Y && 'function' == typeof Y.getAttributeType)
            if (l);
            else
              switch (Y.getAttributeType(d, c)) {
                case 'TrustedHTML':
                  m = ne.createHTML(m);
                  break;
                case 'TrustedScriptURL':
                  m = ne.createScriptURL(m);
              }
          if (m !== f)
            try {
              (l ? e.setAttributeNS(l, a, m) : e.setAttribute(a, m), wt(e) ? gt(e) : p(i.removed));
            } catch (t) {
              ht(a, e);
            }
        } else ht(a, e);
      }
      vt(ce.afterSanitizeAttributes, e, null);
    },
    Et = function (e) {
      let t = null;
      const n = bt(e);
      for (vt(ce.beforeSanitizeShadowDOM, e, null); (t = n.nextNode()); ) (vt(ce.uponSanitizeShadowNode, t, null), Tt(t), At(t), t.content instanceof c && Et(t.content));
      vt(ce.afterSanitizeShadowDOM, e, null);
    };
  return (
    (i.sanitize = function (e) {
      let t = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : {},
        n = null,
        o = null,
        r = null,
        l = null;
      if (((et = !e), et && (e = '\x3c!--\x3e'), 'string' != typeof e && !xt(e))) {
        if ('function' != typeof e.toString) throw k('toString is not a function');
        if ('string' != typeof (e = e.toString())) throw k('dirty is not a string, aborting');
      }
      if (!i.isSupported) return e;
      if ((Fe || pt(t), (i.removed = []), 'string' == typeof e && (We = !1), We)) {
        if (e.nodeName) {
          const t = st(e.nodeName);
          if (!we[t] || Se[t]) throw k('root node is forbidden and cannot be sanitized in-place');
        }
      } else if (e instanceof A)
        ((n = yt('\x3c!----\x3e')), (o = n.ownerDocument.importNode(e, !0)), (1 === o.nodeType && 'BODY' === o.nodeName) || 'HTML' === o.nodeName ? (n = o) : n.appendChild(o));
      else {
        if (!Pe && !De && !Me && -1 === e.indexOf('<')) return ne && Ce ? ne.createHTML(e) : e;
        if (((n = yt(e)), !n)) return Pe ? null : Ce ? ie : '';
      }
      n && Ie && gt(n.firstChild);
      const s = bt(We ? e : n);
      for (; (r = s.nextNode()); ) (Tt(r), At(r), r.content instanceof c && Et(r.content));
      if (We) return e;
      if (Pe) {
        if (De) {
          n.normalize();
          let e = n.innerHTML;
          (u([ue, fe, pe], t => {
            e = b(e, t, ' ');
          }),
            (n.innerHTML = e));
        }
        if (Ne) for (l = ae.call(n.ownerDocument); n.firstChild; ) l.appendChild(n.firstChild);
        else l = n;
        return ((ve.shadowroot || ve.shadowrootmode) && (l = se.call(a, l, !0)), l);
      }
      let f = Me ? n.outerHTML : n.innerHTML;
      return (
        Me &&
          we['!doctype'] &&
          n.ownerDocument &&
          n.ownerDocument.doctype &&
          n.ownerDocument.doctype.name &&
          T(K, n.ownerDocument.doctype.name) &&
          (f = '<!DOCTYPE ' + n.ownerDocument.doctype.name + '>\n' + f),
        De &&
          u([ue, fe, pe], e => {
            f = b(f, e, ' ');
          }),
        ne && Ce ? ne.createHTML(f) : f
      );
    }),
    (i.setConfig = function () {
      (pt(arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {}), (Fe = !0));
    }),
    (i.clearConfig = function () {
      ((ct = null), (Fe = !1));
    }),
    (i.isValidAttribute = function (e, t, n) {
      ct || pt({});
      const i = st(e),
        o = st(t);
      return kt(i, o, n);
    }),
    (i.addHook = function (e, t) {
      'function' == typeof t && m(ce[e], t);
    }),
    (i.removeHook = function (e, t) {
      if (void 0 !== t) {
        const n = f(ce[e], t);
        return -1 === n ? void 0 : d(ce[e], n, 1)[0];
      }
      return p(ce[e]);
    }),
    (i.removeHooks = function (e) {
      ce[e] = [];
    }),
    (i.removeAllHooks = function () {
      ce = {
        afterSanitizeAttributes: [],
        afterSanitizeElements: [],
        afterSanitizeShadowDOM: [],
        beforeSanitizeAttributes: [],
        beforeSanitizeElements: [],
        beforeSanitizeShadowDOM: [],
        uponSanitizeAttribute: [],
        uponSanitizeElement: [],
        uponSanitizeShadowNode: [],
      };
    }),
    i
  );
})();
export { Z as default };
