// @takeoff-ui/core@0.11.9/components/tk-alert.js
import { p as t, H as e, h as a, F as r, t as n } from './p-BfLQgzDX.js';
import { c as o } from './p-D7YaJBkB.js';
import { r as s } from './p-Dpaq_7dX.js';
import { g as i } from './p-BDMnqEGL.js';
import { d as l } from './p-BY6PxsS2.js';
import { d as c } from './p-BpML1w-i.js';
import { d } from './p-C6qt54JA.js';
const g = t(
    class extends e {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.hasRightActionSlot = !1),
          (this.hasFooterActionSlot = !1),
          (this.hasContentSlot = !1),
          (this.variant = 'neutral'),
          (this.iconSize = 'large'),
          (this.alignItems = 'center'),
          (this.type = 'filled'),
          (this.removable = !1),
          (this.containerStyle = null));
      }
      handleCloseButtonClick() {
        this.el.remove();
      }
      createIcons() {
        let t = this.icon;
        return (
          null == t &&
            ('success' == this.variant
              ? (t = 'check_circle')
              : 'info' == this.variant
                ? (t = 'info')
                : 'danger' == this.variant
                  ? (t = 'error')
                  : 'warning' == this.variant && (t = 'warning')),
          s(t, { variant: this.variant, sign: !0, size: this.iconSize, iconStyle: 'rounded', fill: !0, dataTestid: this.dataTestid })
        );
      }
      renderContent() {
        var t, e, r;
        let n, s;
        return (
          (null === (t = this.header) || void 0 === t ? void 0 : t.length) > 0 &&
            (n = a('div', { 'class': 'tk-alert-header', 'style': this.headerStyle, 'data-testid': i(this.dataTestid, 'header') }, this.header)),
          'string' == typeof this.message
            ? (s = a('div', { 'class': 'tk-alert-message', 'data-testid': i(this.dataTestid, 'message') }, this.message))
            : (null === (e = this.message) || void 0 === e ? void 0 : e.every(t => 'string' == typeof t)) &&
              (s = a(
                'div',
                { 'class': 'tk-alert-message-holder', 'data-testid': i(this.dataTestid, 'message-holder') },
                null === (r = this.message) || void 0 === r
                  ? void 0
                  : r.map((t, e) => a('div', { 'class': 'tk-alert-message', 'data-testid': i(this.dataTestid, 'message', '' + e) }, t)),
              )),
          a(
            'div',
            { 'class': o('tk-alert-content', !this.header && 'message-content'), 'data-testid': i(this.dataTestid, 'content') },
            a('div', null, n, s),
            this.hasFooterActionSlot && a('slot', { name: 'footer-action' }),
          )
        );
      }
      renderCloseButton() {
        return this.removable
          ? a('tk-button', {
              'icon': 'close',
              'size': 'small',
              'variant': 'filled' == this.type ? 'white' : 'neutral',
              'type': 'text',
              'onTk-click': () => this.handleCloseButtonClick(),
              'dataTestid': i(this.dataTestid, 'close-button'),
            })
          : null;
      }
      render() {
        ((this.hasContentSlot = !!this.el.querySelector('[slot="content"]')),
          (this.hasRightActionSlot = !!this.el.querySelector('[slot="right-action"]')),
          (this.hasFooterActionSlot = !!this.el.querySelector('[slot="footer-action"]')));
        const t = o('tk-alert-container', this.variant, this.type, 'tk-alert-alignment-' + this.alignItems),
          e = this.renderContent(),
          n = this.renderCloseButton(),
          s = this.createIcons();
        return a(
          'div',
          { 'key': 'bf2fd9e17873ac11dcddf14d4850ab40e835b3be', 'class': t, 'style': this.containerStyle, 'data-testid': i(this.dataTestid, 'container') },
          this.hasContentSlot ? a('slot', { name: 'content' }) : a(r, null, s.leftIcon, e, s.rightIcon),
          this.hasRightActionSlot && a('slot', { key: '6f6835fcfe09b87d89e40d515cbfba1fb2eb1e45', name: 'right-action' }),
          n,
        );
      }
      get el() {
        return this;
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-alert-container{display:flex;justify-content:space-between;align-items:center;gap:var(--toast-base-gap);border-radius:var(--toast-base-radius);box-shadow:var(--effect-1-default-lg);padding:var(--toast-base-v-padding) var(--toast-base-h-padding);border-width:1px;border-style:solid;font-family:var(--family-body)}.tk-alert-container.tk-alert-alignment-center{align-items:center}.tk-alert-container.tk-alert-alignment-start{align-items:flex-start}.tk-alert-container.tk-alert-alignment-end{align-items:flex-end}.tk-alert-container .tk-alert-content{display:flex;justify-content:flex-start;flex-direction:column;gap:var(--toast-base-content-gap);width:100%}.tk-alert-container .tk-alert-content.message-content{justify-content:center}.tk-alert-container .tk-alert-content .tk-alert-header{font-size:var(--desktop-body-s-size);font-weight:400;line-height:20px}.tk-alert-container .tk-alert-content .tk-alert-message{font-size:var(--desktop-body-xs-size);font-weight:300;line-height:18px}.tk-alert-container.success{border-color:var(--states-success-sub-base)}.tk-alert-container.success.filled{background-color:var(--states-success-base);border-color:var(--states-success-base)}.tk-alert-container.success.filledlight{background-color:var(--states-success-lightest)}.tk-alert-container.success.outlined{background-color:var(--background-lightest)}.tk-alert-container.success.gradient{background:var(--gradient-success-gradient-tozero)}.tk-alert-container.info{border-color:var(--states-info-sub-base)}.tk-alert-container.info.filled{background-color:var(--states-info-base);border-color:var(--states-info-base)}.tk-alert-container.info.filledlight{background-color:var(--states-info-light)}.tk-alert-container.info.outlined{background-color:var(--background-lightest)}.tk-alert-container.info.gradient{background:var(--gradient-info-gradient-tozero)}.tk-alert-container.danger{border-color:var(--states-danger-sub-base)}.tk-alert-container.danger.filled{background-color:var(--states-danger-base);border-color:var(--states-danger-base)}.tk-alert-container.danger.filledlight{background-color:var(--states-danger-light)}.tk-alert-container.danger.outlined{background-color:var(--background-lightest)}.tk-alert-container.danger.gradient{background:var(--gradient-danger-gradient-tozero)}.tk-alert-container.warning{border-color:var(--states-warning-sub-base)}.tk-alert-container.warning.filled{background-color:var(--states-warning-base);border-color:var(--states-warning-base)}.tk-alert-container.warning.filledlight{background-color:var(--states-warning-light)}.tk-alert-container.warning.outlined{background-color:var(--background-lightest)}.tk-alert-container.warning.gradient{background:var(--gradient-warning-gradient-tozero)}.tk-alert-container.neutral{border-color:var(--border-sub-base)}.tk-alert-container.neutral.filled{background-color:var(--background-sub-base)}.tk-alert-container.neutral.filled .tk-alert-content .tk-alert-header{color:var(--text-darkest)}.tk-alert-container.neutral.filled .tk-alert-content .tk-alert-message{color:var(--text-dark)}.tk-alert-container.neutral.filledlight{background-color:var(--background-light)}.tk-alert-container.neutral.outlined{background-color:var(--background-lightest)}.tk-alert-container.neutral.gradient{background:var(--gradient-gradient-gray)}.tk-alert-container.filled .tk-alert-content .tk-alert-header{color:var(--text-lightest)}.tk-alert-container.filled .tk-alert-content .tk-alert-message{color:var(--text-lightest)}.tk-alert-container.filledlight .tk-alert-content .tk-alert-header,.tk-alert-container.outlined .tk-alert-content .tk-alert-header,.tk-alert-container.gradient .tk-alert-content .tk-alert-header{color:var(--text-darkest)}.tk-alert-container.filledlight .tk-alert-content .tk-alert-message,.tk-alert-container.outlined .tk-alert-content .tk-alert-message,.tk-alert-container.gradient .tk-alert-content .tk-alert-message{color:var(--text-dark)}';
      }
    },
    [
      257,
      'tk-alert',
      {
        variant: [1],
        header: [1],
        message: [1],
        icon: [1],
        iconSize: [1, 'icon-size'],
        alignItems: [1, 'align-items'],
        type: [1],
        removable: [4],
        containerStyle: [16],
        headerStyle: [16],
        dataTestid: [513, 'data-testid'],
      },
    ],
  ),
  f = g,
  k = function () {
    'undefined' != typeof customElements &&
      ['tk-alert', 'tk-button', 'tk-icon', 'tk-spinner'].forEach(t => {
        switch (t) {
          case 'tk-alert':
            customElements.get(n(t)) || customElements.define(n(t), g);
            break;
          case 'tk-button':
            customElements.get(n(t)) || l();
            break;
          case 'tk-icon':
            customElements.get(n(t)) || c();
            break;
          case 'tk-spinner':
            customElements.get(n(t)) || d();
        }
      });
  };
export { f as TkAlert, k as defineCustomElement };
