// @takeoff-ui/core@0.11.9/components/tk-accordion-item.js
import { p as t, H as e, c as a, h as i, d as o, t as r } from './p-BfLQgzDX.js';
import { c as n } from './p-D7YaJBkB.js';
import { g as s, r as d } from './p-Dpaq_7dX.js';
import { g as c } from './p-BDMnqEGL.js';
import { d as l } from './p-BpML1w-i.js';
const m = t(
    class extends e {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkActiveChange = a(this, 'tk-active-change', 7)),
          (this.type = 'grouped'),
          (this.arrowPosition = 'right'),
          (this.hideArrows = !1),
          (this.hasHeaderSlot = !1),
          (this.mode = 'default'),
          (this.size = 'base'));
      }
      activeChanged(t, e) {
        t !== e && this.tkActiveChange.emit(t);
      }
      componentWillLoad() {
        ((this.parentEl = this.el.closest('tk-accordion')),
          this.parentEl &&
            ((this.type = this.parentEl.type),
            (this.arrowPosition = this.parentEl.arrowPosition),
            (this.expandIcon = this.parentEl.expandIcon),
            (this.collapseIcon = this.parentEl.collapseIcon),
            (this.hideArrows = this.parentEl.hideArrows),
            (this.mode = this.parentEl.mode)),
          (this.hasHeaderSlot = !!this.el.querySelector(':scope > [slot="header"]')));
      }
      renderCollapseIcon() {
        if (this.hideArrows) return null;
        let t;
        return (
          (t = this.active ? this.collapseIcon : this.expandIcon),
          i(
            'tk-icon',
            Object.assign({}, s(t, { size: 'large', variant: this.active ? 'primary' : 'neutral' }), {
              dataTestid: c(this.dataTestid, this.active ? 'collapse-icon' : 'expand-icon'),
            }),
          )
        );
      }
      createHeader() {
        return this.hasHeaderSlot ? i('slot', { name: 'header' }) : this.header || '';
      }
      render() {
        const t = n('tk-accordion-item', this.size, this.type, this.mode, { open: this.active }),
          e = d(this.icon, { sign: !0, variant: 'neutral', additionalProps: { class: 'tk-accordion-item-icon' }, dataTestid: this.dataTestid });
        return i(
          o,
          { key: '2fc3c3020a58ce6d98746361c01a7f292c76733c' },
          i(
            'div',
            { 'key': '4b3210c211ac98bc7f22e15bccc53f80d02a9c19', 'class': t, 'data-testid': c(this.dataTestid, 'item') },
            i(
              'div',
              {
                'key': 'f7a15c080331b7738b27b506432b76e998c06996',
                'class': 'header',
                'onClick': () => this.tkActiveChange.emit(!this.active),
                'data-testid': c(this.dataTestid, 'header'),
              },
              'left' === this.arrowPosition && this.renderCollapseIcon(),
              e.leftIcon,
              i('span', { 'key': 'de3161a01072ebd9c7160580a01c75ea4ef51358', 'class': 'title', 'data-testid': c(this.dataTestid, 'title') }, this.createHeader()),
              e.rightIcon,
              'right' === this.arrowPosition && this.renderCollapseIcon(),
            ),
            i(
              'div',
              { 'key': '895dcbb16edd696ee3fb5560bc260564f1703c5e', 'class': 'content ' + (this.active ? 'open' : ''), 'data-testid': c(this.dataTestid, 'content') },
              i('slot', { key: '74559f5925eaa3232bd5f1d7b9233c53b3bcc6e3', name: 'content' }),
            ),
          ),
        );
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { active: [{ activeChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-accordion-item{margin:10px auto;display:flex;flex-direction:column;transition:gap 0.2s ease-in-out;gap:0;font-family:var(--family-body)}.tk-accordion-item.open{gap:var(--accordion-base-content-gap)}.tk-accordion-item .header{display:flex;align-items:center;cursor:pointer;color:var(--text-darkest)}.tk-accordion-item .header:hover{background:var(--background-lightest) !important}.tk-accordion-item .title{flex-grow:1}.tk-accordion-item .content{color:var(--text-dark);max-height:0;overflow:hidden;font-weight:300}.tk-accordion-item .content.open{max-height:100%}.tk-accordion-item.grouped{border-radius:var(--accordion-base-radius);border:1px solid var(--border-light);background:var(--static-light)}.tk-accordion-item.grouped.open .header{padding-bottom:0}.tk-accordion-item.grouped.open :hover{border-bottom-left-radius:0;border-bottom-right-radius:0}.tk-accordion-item.grouped .content.open{padding:0 var(--accordion-base-v-padding) var(--accordion-base-h-padding)}.tk-accordion-item.divided .header{border:1px solid var(--border-light);background:var(--static-light)}.tk-accordion-item.divided .content.open{padding:var(--accordion-base-v-padding) var(--accordion-base-h-padding);border-radius:var(--accordion-base-radius);border:1px solid var(--border-light);background:var(--background-lightest)}.tk-accordion-item.base.open{gap:var(--accordion-base-content-gap)}.tk-accordion-item.base .header{font-size:var(--desktop-body-s-size);line-height:var(--desktop-body-s-line-height);padding:var(--accordion-base-v-padding) var(--accordion-base-h-padding);border-radius:var(--accordion-base-radius)}.tk-accordion-item.base .header:has(.tk-accordion-item-icon){gap:var(--accordion-base-gap)}.tk-accordion-item.base .content{font-size:var(--desktop-body-xs-size);line-height:var(--desktop-body-xs-line-height)}.tk-accordion-item.large.open{gap:var(--accordion-large-gap)}.tk-accordion-item.large .header{font-size:var(--desktop-body-m-size);line-height:var(--desktop-body-m-line-height);padding:var(--accordion-large-v-padding) var(--accordion-large-h-padding);border-radius:var(--accordion-large-radius)}.tk-accordion-item.large .header:has(.tk-accordion-item-icon){gap:var(--accordion-large-gap)}.tk-accordion-item.large .content{font-size:var(--desktop-body-s-size);line-height:var(--desktop-body-s-line-height)}.tk-accordion-item.compact{margin:0;border-radius:0;border-bottom:0}.tk-accordion-item.compact .header:hover{border-radius:0}.tk-accordion-item.compact.divided{gap:0}.tk-accordion-item.compact.divided .header,.tk-accordion-item.compact.divided .content{border-radius:0;border-bottom:0}:host(:last-of-type) .tk-accordion-item.compact.grouped,:host(:last-of-type) .tk-accordion-item.compact.divided:not(.open) .header,:host(:last-of-type) .tk-accordion-item.compact.divided .content.open{border-bottom:1px solid var(--border-light)}';
      }
    },
    [
      257,
      'tk-accordion-item',
      {
        active: [4],
        itemKey: [520, 'item-key'],
        header: [1],
        size: [1],
        icon: [1],
        dataTestid: [513, 'data-testid'],
        type: [32],
        arrowPosition: [32],
        expandIcon: [32],
        collapseIcon: [32],
        hideArrows: [32],
        hasHeaderSlot: [32],
        mode: [32],
      },
      void 0,
      { active: [{ activeChanged: 0 }] },
    ],
  ),
  h = m,
  p = function () {
    'undefined' != typeof customElements &&
      ['tk-accordion-item', 'tk-icon'].forEach(t => {
        switch (t) {
          case 'tk-accordion-item':
            customElements.get(r(t)) || customElements.define(r(t), m);
            break;
          case 'tk-icon':
            customElements.get(r(t)) || l();
        }
      });
  };
export { h as TkAccordionItem, p as defineCustomElement };
