// @takeoff-ui/core@0.11.9/components/tk-accordion.js
import { p as t, t as i, H as e, c as s, h as n } from './p-BfLQgzDX.js';
import { g as o } from './p-BDMnqEGL.js';
import { i as a } from './p-DzS7aqZe.js';
const r = t(
    class extends e {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkActiveIndexChange = s(this, 'tk-active-index-change', 7)),
          (this.tkAccordionItemSelected = s(this, 'tk-accordion-item-selected', 7)),
          (this.internalActiveIndex = []),
          (this.allowMultiple = !1),
          (this.arrowPosition = 'right'),
          (this.expandIcon = 'keyboard_arrow_down'),
          (this.collapseIcon = 'keyboard_arrow_up'),
          (this.hideArrows = !1),
          (this.type = 'grouped'),
          (this.mode = 'default'));
      }
      internalActiveIndexChanged(t, i) {
        a(t, i) ||
          (this.updateActiveIndex(),
          this.getAccordionItems().forEach((i, e) => {
            const s = this.getItemKey(i, e),
              n = t.some(t => t === s);
            i.active = n;
          }));
      }
      activeIndexChanged() {
        const t = this.normalizeActiveIndex();
        a(t, this.internalActiveIndex) || (this.internalActiveIndex = t);
      }
      componentDidLoad() {
        (this.validateItemKeylessActiveIndex(),
          this.validateActiveProps(),
          this.validateControlType(),
          this.validateKeyType(),
          this.initInternalActiveIndex(),
          this.initEventListeners());
      }
      initInternalActiveIndex() {
        if (this.hasActiveIndex()) return (this.internalActiveIndex = this.normalizeActiveIndex());
        this.getAccordionItems().forEach((t, i) => {
          t.active && (this.internalActiveIndex = [...this.internalActiveIndex, this.getItemKey(t, i)]);
        });
      }
      initEventListeners() {
        this.getAccordionItems().forEach((t, i) => {
          const e = this.getItemKey(t, i);
          t.addEventListener('tk-active-change', i => {
            (this.handleItemActiveChange(e, i.detail),
              (t.active = i.detail),
              !this.allowMultiple &&
                i.detail &&
                this.getAccordionItems()
                  .filter(i => i !== t)
                  .forEach(t => {
                    t.active = !1;
                  }));
          });
        });
      }
      validateActiveProps() {
        this.getAccordionItems().reduce((t, i) => (i.active ? t + 1 : t), 0) > 1 &&
          !this.allowMultiple &&
          console.error('TkAccordion: Multiple accordion items are set to active while allowMultiple is false. This may lead to unexpected behavior.');
      }
      validateControlType() {
        const t = this.getAccordionItems().some(t => void 0 !== t.active);
        this.hasActiveIndex() && t && console.error('TkAccordion: Accordion cannot have both activeIndex and active accordion items');
      }
      validateKeyType() {
        var t, i;
        const e = typeof (null !== (i = null === (t = this.getAccordionItems().find(t => void 0 !== t.itemKey)) || void 0 === t ? void 0 : t.itemKey) && void 0 !== i ? i : 0),
          s = this.getAccordionItems().every(t => void 0 === t.itemKey || typeof t.itemKey === e);
        let n = !0;
        (this.hasActiveIndex() && (n = Array.isArray(this.activeIndex) ? this.activeIndex.every(t => typeof t === e) : typeof this.activeIndex === e),
          (s && n) || console.error('TkAccordion: Accordion item keys must be of the same type'));
      }
      validateItemKeylessActiveIndex() {
        if (!this.hasActiveIndex()) return;
        const t = this.getAccordionItems().every(t => void 0 === t.itemKey),
          i = Array.isArray(this.activeIndex) ? this.activeIndex.map(t => typeof t) : [typeof this.activeIndex];
        t && i.some(t => 'number' !== t) && console.error('TkAccordion: When using keyless accordion items, activeIndex must be of type number or number array.');
      }
      hasActiveIndex() {
        return this.activeIndex || 0 === this.activeIndex;
      }
      normalizeActiveIndex() {
        if (!this.activeIndex && 0 !== this.activeIndex) return [];
        if (!Array.isArray(this.activeIndex)) return [this.activeIndex];
        if (this.allowMultiple) return this.activeIndex;
        const t = this.activeIndex.at(-1);
        return t ? [t] : [];
      }
      getActiveIndex() {
        return this.allowMultiple ? this.internalActiveIndex : this.internalActiveIndex[this.internalActiveIndex.length - 1];
      }
      updateActiveIndex() {
        const t = this.getActiveIndex();
        a(t, this.activeIndex) || this.tkActiveIndexChange.emit(t);
      }
      handleItemActiveChange(t, i) {
        (i && !this.internalActiveIndex.includes(t) && (this.internalActiveIndex = [...this.internalActiveIndex, t]),
          !i && this.internalActiveIndex.includes(t) && (this.internalActiveIndex = this.internalActiveIndex.filter(i => i !== t)),
          this.tkAccordionItemSelected.emit({ index: t, active: i }));
      }
      getAccordionItems() {
        return Array.from(this.el.querySelectorAll('tk-accordion-item')).filter(t => t.parentElement === this.el);
      }
      getItemKey(t, i) {
        var e;
        return null !== (e = t.itemKey) && void 0 !== e ? e : i;
      }
      render() {
        return n(
          'div',
          { 'key': '8a88bf333637906c2bd4218f3015901c5cdbd7aa', 'class': 'tk-accordion', 'data-testid': o(this.dataTestid, 'container') },
          n('slot', { key: '6bbb5335b02b07138985ed6b883a03a5cd1c4254' }),
        );
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { internalActiveIndex: [{ internalActiveIndexChanged: 0 }], activeIndex: [{ activeIndexChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-accordion{display:block}';
      }
    },
    [
      257,
      'tk-accordion',
      {
        activeIndex: [8, 'active-index'],
        allowMultiple: [4, 'allow-multiple'],
        arrowPosition: [1, 'arrow-position'],
        expandIcon: [1, 'expand-icon'],
        collapseIcon: [1, 'collapse-icon'],
        hideArrows: [4, 'hide-arrows'],
        type: [1],
        mode: [1],
        dataTestid: [513, 'data-testid'],
        internalActiveIndex: [32],
      },
      void 0,
      { internalActiveIndex: [{ internalActiveIndexChanged: 0 }], activeIndex: [{ activeIndexChanged: 0 }] },
    ],
  ),
  l = r,
  d = function () {
    'undefined' != typeof customElements &&
      ['tk-accordion'].forEach(t => {
        'tk-accordion' === t && (customElements.get(i(t)) || customElements.define(i(t), r));
      });
  };
export { l as TkAccordion, d as defineCustomElement };
