// @takeoff-ui/core@0.11.9/components/tk-breadcrumb.js
import { p as t, H as a, h as r, F as e, t as s } from './p-BfLQgzDX.js';
import { c as o } from './p-D7YaJBkB.js';
import { g as i } from './p-Dpaq_7dX.js';
import { g as n } from './p-BDMnqEGL.js';
import { d as l } from './p-CYt7G91Z.js';
import { d as b } from './p-BpML1w-i.js';
const m = t(
    class extends a {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.hasSlottedItems = !1),
          (this.separator = 'icon'),
          (this.separatorIcon = 'chevron_right'),
          (this.type = 'basic'));
      }
      componentWillLoad() {
        this.hasSlottedItems = !!this.el.querySelector('tk-breadcrumb-item');
      }
      getBreadcrumbItemProps(t, a) {
        var r, e;
        const { label: s, href: o, isExternal: i, icon: l } = t;
        return {
          key: a,
          label: s,
          href: o,
          icon: l,
          isExternal: i,
          isCurrent: a === (null !== (e = null === (r = this.model) || void 0 === r ? void 0 : r.length) && void 0 !== e ? e : 0) - 1,
          dataTestid: n(this.dataTestid, 'item', '' + a),
        };
      }
      renderSeparator(t) {
        const a = n(this.dataTestid, 'separator', '' + t),
          e = o('tk-breadcrumb-separator-icon', {
            'tk-breadcrumb-dot-separator': 'dot' === this.separator,
            'tk-breadcrumb-slash-separator': 'slash' === this.separator,
            'tk-breadcrumb-vertical-separator': 'vertical' === this.separator,
          });
        return 'icon' === this.separator
          ? r('tk-icon', Object.assign({}, i(this.separatorIcon, { class: e, color: 'var(--icon-sub-base)' }, 'rounded', 'span'), { dataTestid: n(a, 'icon') }))
          : r('span', { 'class': e, 'data-testid': n(a, 'item') });
      }
      render() {
        var t;
        const a = o('tk-breadcrumb', 'tk-breadcrumb-' + this.type);
        return r(
          'nav',
          { 'key': '266cb8b62ff2feda77d3bc749f83554f5bc7f8ae', 'class': a, 'aria-label': 'breadcrumb', 'data-testid': n(this.dataTestid, 'container') },
          r(
            'ol',
            { 'key': '8496e0dc3f511daa1c659cf74c4e9996fa9f6b7b', 'class': 'tk-breadcrumb-list', 'data-testid': n(this.dataTestid, 'list') },
            this.hasSlottedItems
              ? r('slot', null)
              : null === (t = this.model) || void 0 === t
                ? void 0
                : t.map((t, a) =>
                    r(
                      e,
                      null,
                      r('tk-breadcrumb-item', Object.assign({}, this.getBreadcrumbItemProps(t, a))),
                      a < this.model.length - 1 && r('li', { 'class': 'tk-breadcrumb-separator', 'data-testid': n(this.dataTestid, 'separator', '' + a) }, this.renderSeparator(a)),
                    ),
                  ),
          ),
        );
      }
      get el() {
        return this;
      }
      static get style() {
        return '@charset "UTF-8";@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-breadcrumb{display:flex;align-items:center;overflow-x:auto;overflow-y:hidden;font-family:var(--family-body)}.tk-breadcrumb::-webkit-scrollbar{width:4px;height:4px}.tk-breadcrumb::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-breadcrumb::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-breadcrumb::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-breadcrumb::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-breadcrumb .tk-breadcrumb-list{margin:0;padding:0;list-style-type:none;display:flex;align-items:center;justify-content:center;flex-wrap:nowrap;gap:var(--breadcrumb-container-gap)}.tk-breadcrumb .tk-breadcrumb-list .tk-breadcrumb-separator{display:flex;width:var(--breadcrumb-symbol-size);height:var(--breadcrumb-symbol-size);justify-content:center;align-items:center}.tk-breadcrumb .tk-breadcrumb-list .tk-breadcrumb-separator .tk-breadcrumb-dot-separator::before{content:"•"}.tk-breadcrumb .tk-breadcrumb-list .tk-breadcrumb-separator .tk-breadcrumb-slash-separator::before{content:"/"}.tk-breadcrumb .tk-breadcrumb-list .tk-breadcrumb-separator .tk-breadcrumb-vertical-separator::before{content:"|"}';
      }
    },
    [257, 'tk-breadcrumb', { model: [16], separator: [1], separatorIcon: [1, 'separator-icon'], type: [1], dataTestid: [513, 'data-testid'], hasSlottedItems: [32] }],
  ),
  c = m,
  d = function () {
    'undefined' != typeof customElements &&
      ['tk-breadcrumb', 'tk-breadcrumb-item', 'tk-icon'].forEach(t => {
        switch (t) {
          case 'tk-breadcrumb':
            customElements.get(s(t)) || customElements.define(s(t), m);
            break;
          case 'tk-breadcrumb-item':
            customElements.get(s(t)) || l();
            break;
          case 'tk-icon':
            customElements.get(s(t)) || b();
        }
      });
  };
export { c as TkBreadcrumb, d as defineCustomElement };
