// @takeoff-ui/core@0.11.9/components/tk-card.js
import { p as t, H as a, h as r, F as e, t as o } from './p-BfLQgzDX.js';
import { c as i } from './p-D7YaJBkB.js';
import { g as d } from './p-BDMnqEGL.js';
import { d as s } from './p-BVlaGNiB.js';
import { d as n } from './p-BY6PxsS2.js';
import { d as c } from './p-BpML1w-i.js';
import { d as l } from './p-C6qt54JA.js';
const h = t(
    class extends a {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.imageOptions = { position: 'top', background: !1, windowed: !0 }),
          (this.showAvatar = !1),
          (this.avatarProps = { severity: 'light', background: 'solid', rounded: !0, size: 'small' }),
          (this.hideHeader = !1),
          (this.showMenuButton = !1),
          (this.image = null),
          (this.headerPosition = 'top'),
          (this.headerType = 'basic'),
          (this.horizontal = !1),
          (this.footerType = 'basic'),
          (this.enableHoverShadow = !1),
          (this.containerStyle = null),
          (this.contentStyle = null));
      }
      componentWillLoad() {
        ((this.hasHeaderSlot = !!this.el.querySelector(':scope > [slot="header"]')),
          (this.hasAvatarSlot = !!this.el.querySelector(':scope > [slot="avatar"]')),
          (this.hasContentSlot = !!this.el.querySelector(':scope > [slot="content"]')),
          (this.hasFooterSlot = !!this.el.querySelector(':scope > [slot="footer"]')),
          (this.hasFooterActionsSlot = !!this.el.querySelector(':scope > [slot="footer-actions"]')),
          (this.hasHeaderActionSlot = !!this.el.querySelector(':scope > [slot="header-action"]')),
          (this.hasDefaultSlotBody = Array.from(this.el.childNodes).some(t => t.nodeType === Node.ELEMENT_NODE && !t.hasAttribute('slot'))));
      }
      createHeader() {
        return this.hideHeader
          ? null
          : this.hasHeaderSlot
            ? r('slot', { name: 'header' })
            : this.header
              ? r(
                  'div',
                  {
                    'data-testid': d(this.dataTestid, 'header-container'),
                    'class': i('tk-card-header', 'tk-card-header-' + this.headerPosition, { ['tk-card-header-' + this.headerType]: !this.imageOptions.background }),
                  },
                  r(
                    'div',
                    { 'class': 'tk-card-header-content', 'data-testid': d(this.dataTestid, 'header') },
                    this.hasAvatarSlot
                      ? r('slot', { name: 'avatar' })
                      : r(
                          e,
                          null,
                          this.showAvatar &&
                            r(
                              'div',
                              { 'class': 'tk-card-avatar', 'data-testid': d(this.dataTestid, 'avatar-container') },
                              ' ',
                              r('tk-avatar', Object.assign({}, this.avatarProps, { 'data-testid': d(this.dataTestid, 'avatar') })),
                            ),
                        ),
                    r(
                      'div',
                      { 'class': 'tk-card-title-container', 'data-testid': d(this.dataTestid, 'title-container') },
                      this.subheader && r('span', { 'class': 'tk-card-subtitle', 'data-testid': d(this.dataTestid, 'subtitle') }, this.subheader),
                      this.header && r('span', { 'class': 'tk-card-title', 'data-testid': d(this.dataTestid, 'title') }, this.header),
                    ),
                  ),
                  this.hasHeaderActionSlot
                    ? r('slot', { name: 'header-action' })
                    : this.showMenuButton &&
                        r('tk-button', {
                          'variant': 'neutral',
                          'icon': 'more_vert',
                          'size': 'base',
                          'type': 'text',
                          'aria-label': 'TkCard Header Menu Button',
                          'dataTestid': d(this.dataTestid, 'header-menu-button'),
                        }),
                )
              : null;
      }
      createImage() {
        if (!this.image || this.imageOptions.background) return null;
        const t = i('tk-card-image', 'tk-card-image-' + this.imageOptions.position, { 'tk-card-windowed-image': this.imageOptions.windowed });
        return r(
          'div',
          { 'class': t, 'data-testid': d(this.dataTestid, 'image-container') },
          r('img', { 'src': this.image, 'alt': 'Card image', 'data-testid': d(this.dataTestid, 'image') }),
        );
      }
      createContent() {
        return this.hasContentSlot
          ? r('slot', { name: 'content' })
          : this.hasDefaultSlotBody
            ? r('div', { 'class': 'tk-card-content', 'style': this.contentStyle, 'data-testid': d(this.dataTestid, 'content') }, r('slot', null))
            : null;
      }
      createFooter() {
        return this.hasFooterSlot
          ? r('slot', { name: 'footer' })
          : this.hasFooterActionsSlot
            ? r(
                'div',
                { 'data-testid': d(this.dataTestid, 'footer'), 'class': i('tk-card-footer', { ['tk-card-footer-' + this.footerType]: !this.imageOptions.background }) },
                r('slot', { name: 'footer-actions' }),
              )
            : null;
      }
      renderCardContent() {
        const t = this.createHeader(),
          a = this.createImage(),
          o = this.createContent(),
          i = this.createFooter(),
          s = this.imageOptions.position;
        return 'top' === s
          ? r(e, null, 'top' === this.headerPosition && t, a, 'bottom' === this.headerPosition && t, o, i)
          : !this.horizontal || ('left' !== s && 'right' !== s)
            ? void 0
            : r(
                e,
                null,
                'left' === s && a,
                r('div', { 'class': 'tk-card-horizontal-has-image-container', 'data-testid': d(this.dataTestid, 'horizontal-has-image-container') }, t, o, i),
                'right' === s && a,
              );
      }
      render() {
        const t = {
          'class': i('tk-card', {
            'tk-card-vertical': !this.horizontal,
            'tk-card-horizontal': this.horizontal,
            'tk-card-image-background': this.imageOptions.background,
            'has-hover-shadow': this.enableHoverShadow,
          }),
          'style': Object.assign(Object.assign({}, this.imageOptions.backgroundUrl && { background: `url(${this.imageOptions.backgroundUrl})` }), this.containerStyle),
          'data-testid': d(this.dataTestid, 'container'),
        };
        return r('div', Object.assign({ key: '38196aa981f2f98e84e4d4ab726682f27a492112' }, t), this.renderCardContent());
      }
      get el() {
        return this;
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-card{display:flex;background:var(--static-light);border-radius:16px;border:1px solid var(--border-light);box-shadow:var(--effect-1-default-base);font-family:var(--family-body)}.tk-card.has-hover-shadow:hover{box-shadow:var(--effect-1-hover-base)}.tk-card.tk-card-vertical{flex-direction:column}.tk-card.tk-card-horizontal{flex-direction:row;align-items:stretch}.tk-card.tk-card-horizontal:has(.tk-card-image-top){flex-direction:column;justify-content:stretch}.tk-card.tk-card-horizontal .tk-card-horizontal-has-image-container{display:flex;flex-direction:column;flex:1 0 0;align-self:stretch}.tk-card.tk-card-horizontal .tk-card-image:not(.tk-card-windowed-image).tk-card-image-left img{border-radius:8px 0 0 8px}.tk-card.tk-card-horizontal .tk-card-image:not(.tk-card-windowed-image).tk-card-image-right img{border-radius:0 8px 8px 0}.tk-card .tk-card-header{display:flex;padding:var(--card-header-base-v-padding, 12px) var(--card-header-base-h-padding, 16px);justify-content:space-between;align-items:center;align-self:stretch}.tk-card .tk-card-header .tk-card-header-content{display:flex;align-items:center;gap:var(--dialog-header-base-gap)}.tk-card .tk-card-header .tk-card-header-content .tk-card-title-container{display:flex;flex-direction:column;align-items:flex-start}.tk-card .tk-card-header .tk-card-header-content .tk-card-title-container .tk-card-title{color:var(--text-darkest);font-family:var(--desktop-title-h6-font);font-size:var(--desktop-title-h6-size, 20px);font-weight:400;line-height:var(--desktop-title-h6-line-height)}.tk-card .tk-card-header .tk-card-header-content .tk-card-title-container .tk-card-subtitle{color:var(--text-base);font-family:var(--desktop-subheading-2xs-font);font-size:var(--desktop-subheading-2xs-size, 11px);font-weight:400;line-height:var(--desktop-subheading-2xs-line-height, 16px)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header){border-radius:16px 16px 0 0}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-divided,.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-light{border-bottom:var(--spacing-px) solid var(--border-light)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-light{background:var(--background-lightest);border-top-left-radius:var(--radius-xl);border-top-right-radius:var(--radius-xl)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-dark{background:var(--background-darkest)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-dark .tk-card-title-container .tk-card-subtitle{color:var(--text-sub-base)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-dark .tk-card-title-container .tk-card-title{color:var(--text-lightest)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-dark .tk-card-header-menu-button{color:var(--icon-base)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-primary{background:var(--primary-base)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-primary .tk-card-title-container .tk-card-subtitle{color:var(--primary-lightest)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-primary .tk-card-title-container .tk-card-title{color:var(--text-lightest)}.tk-card .tk-card-header.tk-card-header-top:not(.tk-card-horizontal .tk-card-horizontal-has-image-container .tk-card-header).tk-card-header-primary .tk-card-header-menu-button{color:var(--primary-lightest)}.tk-card .tk-card-image{display:flex;flex-direction:column;align-items:flex-start;align-self:stretch}.tk-card .tk-card-image.tk-card-windowed-image{padding:var(--card-image-windowed-v-padding, 16px) var(--card-image-windowed-h-padding, 16px)}.tk-card .tk-card-image.tk-card-windowed-image img{border-radius:8px}.tk-card .tk-card-image img{width:100%;height:100%;flex:1 0 0;align-self:stretch}.tk-card .tk-card-content{display:flex;flex-direction:column;align-self:stretch;padding:var(--card-body-relaxed-v-padding, 16px) var(--card-body-relaxed-h-padding, 16px);color:var(--text-base, #717784);gap:8px}.tk-card .tk-card-footer{display:flex;padding:var(--card-footer-base-v-padding, 16px) var(--card-footer-base-h-padding, 16px);justify-content:flex-end;align-self:stretch}.tk-card .tk-card-footer ::slotted([slot=footer-actions]){display:flex;align-items:center;gap:var(--dialog-footer-base-gap)}.tk-card .tk-card-footer:not(.tk-card-horizontal .tk-card-footer).tk-card-footer-divided,.tk-card .tk-card-footer:not(.tk-card-horizontal .tk-card-footer).tk-card-footer-light{border-top:var(--spacing-px, 1px) solid var(--border-light, #e1e4ea)}.tk-card .tk-card-footer:not(.tk-card-horizontal .tk-card-footer).tk-card-footer-light{background:var(--background-lightest, #f9fafc);border-bottom-left-radius:var(--radius-xl);border-bottom-right-radius:var(--radius-xl)}.tk-card.tk-card-image-background .tk-card-header .tk-card-header-content .tk-card-title-container .tk-card-subtitle{color:var(--text-light, #e1e4ea)}.tk-card.tk-card-image-background .tk-card-header .tk-card-header-content .tk-card-title-container .tk-card-title{color:var(--text-lightest, #f9fafc)}.tk-card.tk-card-image-background .tk-card-content{color:var(--text-light, #e1e4ea)}';
      }
    },
    [
      257,
      'tk-card',
      {
        imageOptions: [16],
        showAvatar: [4, 'show-avatar'],
        avatarProps: [8, 'avatar-props'],
        hideHeader: [4, 'hide-header'],
        showMenuButton: [4, 'show-menu-button'],
        image: [1],
        headerPosition: [1, 'header-position'],
        header: [1],
        subheader: [1],
        headerType: [1, 'header-type'],
        horizontal: [4],
        footerType: [1, 'footer-type'],
        enableHoverShadow: [4, 'enable-hover-shadow'],
        containerStyle: [16],
        contentStyle: [16],
        dataTestid: [513, 'data-testid'],
        hasHeaderSlot: [32],
        hasAvatarSlot: [32],
        hasContentSlot: [32],
        hasFooterSlot: [32],
        hasFooterActionsSlot: [32],
        hasHeaderActionSlot: [32],
        hasDefaultSlotBody: [32],
      },
    ],
  ),
  k = h,
  m = function () {
    'undefined' != typeof customElements &&
      ['tk-card', 'tk-avatar', 'tk-button', 'tk-icon', 'tk-spinner'].forEach(t => {
        switch (t) {
          case 'tk-card':
            customElements.get(o(t)) || customElements.define(o(t), h);
            break;
          case 'tk-avatar':
            customElements.get(o(t)) || s();
            break;
          case 'tk-button':
            customElements.get(o(t)) || n();
            break;
          case 'tk-icon':
            customElements.get(o(t)) || c();
            break;
          case 'tk-spinner':
            customElements.get(o(t)) || l();
        }
      });
  };
export { k as TkCard, m as defineCustomElement };
