// @takeoff-ui/core@0.11.9/components/tk-carousel.js
import { p as t, H as i, c as s, h as o, t as e } from './p-BfLQgzDX.js';
import { c as a } from './p-D7YaJBkB.js';
import { g as r } from './p-BDMnqEGL.js';
import { d as n } from './p-BY6PxsS2.js';
import { d as l } from './p-BpML1w-i.js';
import { d as c } from './p-C6qt54JA.js';
const h = t(
    class extends i {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkChange = s(this, 'tk-change', 7)),
          (this.items = []),
          (this.activeIndex = 0),
          (this.totalItems = 0),
          (this.showIndicators = !0),
          (this.showArrows = !0),
          (this.autoplay = !1),
          (this.autoplayDelay = 3e3),
          (this.circular = !0),
          (this.navigationPlacement = 'inside'),
          (this.navigationPosition = 'distributed'),
          (this.showPlayerButton = !1),
          (this.itemsPerView = 1),
          (this.orientation = 'horizontal'),
          (this.verticalViewHeight = '300px'),
          (this.startAutoplay = () => {
            !this.autoplay ||
              this.totalItems <= this.itemsPerView ||
              (this.stopAutoplay(),
              (this.autoplayTimer = window.setInterval(() => {
                this.activeIndex >= this.totalItems - this.itemsPerView ? (this.circular ? this.changeItem(0) : this.stopAutoplay()) : this.changeItem(this.activeIndex + 1);
              }, this.autoplayDelay)));
          }),
          (this.stopAutoplay = () => {
            this.autoplayTimer && (clearInterval(this.autoplayTimer), (this.autoplayTimer = void 0));
          }),
          (this.handlePrevClick = () => {
            (this.changeItem(this.activeIndex - 1), this.tkChange.emit(this.activeIndex));
          }),
          (this.handleNextClick = () => {
            (this.changeItem(this.activeIndex + 1), this.tkChange.emit(this.activeIndex));
          }),
          (this.handleIndicatorClick = t => {
            this.changeItem(t);
          }),
          (this.handleKeyDown = t => {
            'vertical' === this.orientation
              ? 'ArrowUp' === t.key
                ? (t.preventDefault(), this.handlePrevClick())
                : 'ArrowDown' === t.key && (t.preventDefault(), this.handleNextClick())
              : 'ArrowLeft' === t.key
                ? (t.preventDefault(), this.handlePrevClick())
                : 'ArrowRight' === t.key && (t.preventDefault(), this.handleNextClick());
          }));
      }
      componentDidLoad() {
        var t, i;
        (this.slotElement || (this.slotElement = null === (t = this.el.shadowRoot) || void 0 === t ? void 0 : t.querySelector('slot')),
          (this.itemsContainer = null === (i = this.el.shadowRoot) || void 0 === i ? void 0 : i.querySelector('.tk-carousel-items-container')),
          (this.el.tabIndex = 0),
          this.updateItemPosition(),
          this.updateItems(),
          this.startAutoplay(),
          this.el.addEventListener('keydown', this.handleKeyDown));
      }
      disconnectedCallback() {
        (this.stopAutoplay(), this.el.removeEventListener('keydown', this.handleKeyDown));
      }
      get validNavigationPosition() {
        return 'vertical' === this.orientation
          ? ['left', 'right'].includes(this.navigationPosition)
            ? this.navigationPosition
            : 'right'
          : ['distributed', 'top', 'bottom'].includes(this.navigationPosition)
            ? this.navigationPosition
            : 'distributed';
      }
      updateItems() {
        var t;
        const i = (null === (t = this.slotElement) || void 0 === t ? void 0 : t.assignedElements()) || [];
        ((this.items = i.filter(t => !t.hasAttribute('slot'))), (this.totalItems = this.items.length));
      }
      updateItemPosition() {
        var t;
        if (!this.itemsContainer) return;
        const i = null === (t = this.el.shadowRoot) || void 0 === t ? void 0 : t.querySelector('.tk-carousel-overlay');
        if (!i) return;
        const s = 16 * Math.max(0, this.itemsPerView - 1);
        let o;
        'horizontal' === this.orientation
          ? (o = i.getBoundingClientRect().width)
          : ((this.el.style.height = this.verticalViewHeight + ''), (o = Number.parseFloat(this.verticalViewHeight)));
        const e = this.activeIndex * ((o - s) / this.itemsPerView + 16);
        this.itemsContainer.style.transform = 'vertical' === this.orientation ? `translateY(-${e}px)` : `translateX(-${e}px)`;
      }
      changeItem(t) {
        if (!this.totalItems) return;
        const i = Math.max(0, this.totalItems - this.itemsPerView);
        (this.circular ? (t < 0 ? (t = i) : t > i && (t = 0)) : (t < 0 && (t = 0), t > i && (t = i)),
          (this.activeIndex = t),
          this.updateItemPosition(),
          this.tkChange.emit(this.activeIndex));
      }
      createPrevButton() {
        return !this.showArrows || this.items.length <= this.itemsPerView
          ? null
          : this.circular || 0 !== this.activeIndex
            ? o('tk-button', {
                'size': 'small',
                'class': 'prev-button',
                'icon': 'vertical' === this.orientation ? 'keyboard_arrow_up' : 'chevron_left',
                'onTk-click': this.handlePrevClick,
                'dataTestid': r(this.dataTestid, 'prev-button'),
              })
            : null;
      }
      createNextButton() {
        return !this.showArrows || this.items.length <= this.itemsPerView || (!this.circular && this.activeIndex >= Math.max(0, this.totalItems - this.itemsPerView))
          ? null
          : o('tk-button', {
              'size': 'small',
              'class': 'next-button',
              'icon': 'vertical' === this.orientation ? 'keyboard_arrow_down' : 'chevron_right',
              'onTk-click': this.handleNextClick,
              'dataTestid': r(this.dataTestid, 'next-button'),
            });
      }
      createPlayerButtons() {
        return this.showPlayerButton && this.autoplay
          ? o(
              'tk-icon',
              Object.assign(
                { class: 'player-button', size: 'small' },
                {
                  color: 'var(--background-light)',
                  icon: this.autoplayTimer ? 'pause_circle' : 'play_circle',
                  onClick: this.autoplayTimer ? this.stopAutoplay : this.startAutoplay,
                },
                { dataTestid: r(this.dataTestid, 'player-icon') },
              ),
            )
          : null;
      }
      createIndicators() {
        if (!this.showIndicators || this.items.length <= this.itemsPerView) return null;
        const t = Math.max(1, this.totalItems - this.itemsPerView + 1);
        return o(
          'div',
          { 'class': 'tk-carousel-indicators', 'data-testid': r(this.dataTestid, 'indicators') },
          this.createPlayerButtons(),
          Array.from({ length: t }).map((t, i) => {
            const s = r(this.dataTestid, 'indicator', '' + i);
            return o(
              'div',
              { 'class': 'tk-carousel-indicator', 'onClick': () => this.handleIndicatorClick(i), 'data-testid': r(s) },
              o('div', { 'class': a('tk-carousel-indicator-dot', { active: i === this.activeIndex }), 'data-testid': r(s, 'dot') }),
            );
          }),
        );
      }
      renderNavigationElement() {
        const t = this.validNavigationPosition;
        return o(
          'div',
          {
            'class': a('tk-carousel-navigation-holder', { 'tk-carousel-navigation': 'distributed' !== t }, { 'vertical-navigation': 'left' === t || 'right' === t }),
            'data-testid': r(this.dataTestid, 'navigation'),
          },
          this.createPrevButton(),
          this.createIndicators(),
          this.createNextButton(),
        );
      }
      render() {
        const t = a(
          'tk-carousel',
          this.navigationPlacement,
          this.validNavigationPosition,
          { vertical: 'vertical' === this.orientation },
          { 'show-arrows': this.showArrows },
          { 'show-indicators': this.showIndicators },
        );
        return (
          this.el.style.setProperty('--items-per-view', this.itemsPerView + ''),
          o(
            'div',
            { 'key': '77f46c60f3210b667dd41a7824f2193ffbff8f14', 'class': t, 'data-testid': r(this.dataTestid, 'container') },
            o(
              'div',
              { 'key': '2cd9391d7a28317ce9f6fdbd4a5c53ce001b9c9b', 'class': 'tk-carousel-overlay', 'data-testid': r(this.dataTestid, 'overlay') },
              o(
                'div',
                { 'key': 'c34ff3931692c586dbfdad306f11528fa33ece41', 'class': 'tk-carousel-items-container', 'data-testid': r(this.dataTestid, 'items') },
                o('slot', {
                  key: '28b8ab6ad330463aa8a47161de039af37ae8aa3d',
                  onSlotchange: () => {
                    (this.updateItems(), this.updateItemPosition());
                  },
                }),
              ),
            ),
            this.renderNavigationElement(),
          )
        );
      }
      get el() {
        return this;
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block;width:100%;height:100%}.tk-carousel-overlay{overflow:hidden}.tk-carousel-overlay .tk-carousel-items-container{display:flex;gap:16px;transition:transform 0.3s ease}.tk-carousel-overlay .tk-carousel-items-container ::slotted(*){flex:0 0 calc((100% - (var(--items-per-view, 1) - 1) * 16px) / var(--items-per-view, 1));min-width:0;object-fit:cover;box-sizing:border-box}.tk-carousel{position:relative;box-sizing:border-box}.tk-carousel.vertical{height:100%}.tk-carousel.vertical .tk-carousel-overlay{height:100%}.tk-carousel.vertical .tk-carousel-items-container{flex-direction:column;width:100%;height:100%}.tk-carousel.vertical .tk-carousel-items-container ::slotted(*){max-height:calc((100% - (var(--items-per-view, 1) - 1) * 16px) / var(--items-per-view, 1))}.tk-carousel .tk-carousel-navigation{display:flex;align-items:center;justify-content:center;position:absolute;gap:20px;z-index:10}.tk-carousel .tk-carousel-navigation.vertical-navigation{flex-direction:column}.tk-carousel .tk-carousel-navigation.vertical-navigation .tk-carousel-indicators{flex-direction:column}.tk-carousel .tk-carousel-indicators{display:flex;align-items:center;gap:2px}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator{padding:4px;cursor:pointer}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator .tk-carousel-indicator-dot{width:6px;height:6px;background-color:var(--background-light);border-radius:50%;transition:all 0.3s ease}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator .tk-carousel-indicator-dot:hover{background-color:var(--primary-light)}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator .tk-carousel-indicator-dot.active{background-color:var(--primary-base);width:8px;height:8px}.tk-carousel .tk-carousel-indicators .player-button{cursor:pointer}.tk-carousel .prev-button,.tk-carousel .next-button{z-index:10;display:flex;align-items:center;justify-content:center}.tk-carousel.inside.bottom .tk-carousel-navigation{bottom:16px;left:50%;transform:translateX(-50%)}.tk-carousel.inside.top .tk-carousel-navigation{top:16px;left:50%;transform:translateX(-50%)}.tk-carousel.inside.left .tk-carousel-navigation{left:16px;top:50%;transform:translateY(-50%)}.tk-carousel.inside.right .tk-carousel-navigation{right:16px;top:50%;transform:translateY(-50%)}.tk-carousel.inside.distributed .prev-button,.tk-carousel.inside.distributed .next-button{position:absolute;top:50%;transform:translateY(-50%)}.tk-carousel.inside.distributed .prev-button{left:16px}.tk-carousel.inside.distributed .next-button{right:16px}.tk-carousel.inside.distributed .tk-carousel-indicators{position:absolute;bottom:16px;left:50%;transform:translateX(-50%)}.tk-carousel.outside.bottom.show-arrows,.tk-carousel.outside.bottom.show-indicators{padding-bottom:48px}.tk-carousel.outside.bottom .tk-carousel-navigation{bottom:0px;left:50%;transform:translateX(-50%)}.tk-carousel.outside.top.show-arrows,.tk-carousel.outside.top.show-indicators{padding-top:48px}.tk-carousel.outside.top .tk-carousel-navigation{top:0px;left:50%;transform:translateX(-50%)}.tk-carousel.outside.left.show-arrows,.tk-carousel.outside.left.show-indicators{padding-left:48px}.tk-carousel.outside.left .tk-carousel-navigation{left:0px;top:50%;transform:translateY(-50%)}.tk-carousel.outside.right.show-arrows,.tk-carousel.outside.right.show-indicators{padding-right:48px}.tk-carousel.outside.right .tk-carousel-navigation{right:0px;top:50%;transform:translateY(-50%)}.tk-carousel.outside.distributed.show-arrows{padding-inline:48px}.tk-carousel.outside.distributed.show-indicators{padding-bottom:32px}.tk-carousel.outside.distributed .prev-button,.tk-carousel.outside.distributed .next-button{position:absolute;top:50%;transform:translateY(-50%)}.tk-carousel.outside.distributed .prev-button{left:0px}.tk-carousel.outside.distributed .next-button{right:0px}.tk-carousel.outside.distributed .tk-carousel-indicators{position:absolute;bottom:0px;left:50%;transform:translateX(-50%)}';
      }
    },
    [
      257,
      'tk-carousel',
      {
        showIndicators: [4, 'show-indicators'],
        showArrows: [4, 'show-arrows'],
        autoplay: [4],
        autoplayDelay: [2, 'autoplay-delay'],
        circular: [4],
        navigationPlacement: [1, 'navigation-placement'],
        navigationPosition: [1, 'navigation-position'],
        showPlayerButton: [4, 'show-player-button'],
        itemsPerView: [2, 'items-per-view'],
        orientation: [1],
        verticalViewHeight: [1, 'vertical-view-height'],
        dataTestid: [513, 'data-testid'],
        autoplayTimer: [32],
        activeIndex: [32],
        totalItems: [32],
      },
    ],
  ),
  d = h,
  u = function () {
    'undefined' != typeof customElements &&
      ['tk-carousel', 'tk-button', 'tk-icon', 'tk-spinner'].forEach(t => {
        switch (t) {
          case 'tk-carousel':
            customElements.get(e(t)) || customElements.define(e(t), h);
            break;
          case 'tk-button':
            customElements.get(e(t)) || n();
            break;
          case 'tk-icon':
            customElements.get(e(t)) || l();
            break;
          case 'tk-spinner':
            customElements.get(e(t)) || c();
        }
      });
  };
export { d as TkCarousel, u as defineCustomElement };
