// @takeoff-ui/core@0.11.9/components/tk-dropdown.js
import { p as t, H as o, c as e, h as n, t as i } from './p-BfLQgzDX.js';
import { c as d } from './p-D7YaJBkB.js';
import { C as r } from './p-C0U9wSBw.js';
import { f as a } from './p-BiHzhysf.js';
import { v as s } from './p-CK3_k8jD.js';
const l = t(
    class extends o {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          (this.tkItemClick = e(this, 'tk-item-click', 7)),
          (this.hasEmptyDataSlot = !1),
          (this.uniqueId = s()),
          (this.isOpen = !1),
          (this.disabled = !1),
          (this.emptyMessage = 'No options available'),
          (this.optionLabelKey = 'label'),
          (this.optionValueKey = 'value'),
          (this.groupNameKey = 'groupName'),
          (this.groupOptionsKey = 'options'),
          (this.optionsAlign = 'left'),
          (this.position = 'bottom'),
          (this.size = 'base'),
          (this.closeHandler = () => {
            this.isOpen = !1;
          }));
      }
      disabledChanged(t) {
        t && (this.isOpen = !1);
      }
      componentWillLoad() {
        ((this.hasEmptyDataSlot = !!this.el.querySelector('[slot="empty-data"]')),
          (this.triggerRef = this.el.querySelector('[slot="trigger"]')),
          this.triggerRef &&
            !this.disabled &&
            ((this.triggerRef.style.cursor = 'pointer'),
            this.triggerRef.addEventListener('click', () => {
              this.isOpen = !this.isOpen;
            })));
      }
      componentDidLoad() {
        this.clickOutsideMixin = new r({ referenceElement: this.el, handler: this.closeHandler, disabled: this.disabled || !this.isOpen });
      }
      componentDidUpdate() {
        var t, o;
        (this.clickOutsideMixin.updateConfig({ disabled: this.disabled || !this.isOpen }),
          this.isOpen
            ? (null === (t = this.cleanup) || void 0 === t || t.call(this), this.updatePosition())
            : (null === (o = this.cleanup) || void 0 === o || o.call(this), (this.cleanup = null)));
      }
      disconnectedCallback() {
        var t, o;
        (null === (t = this.cleanup) || void 0 === t || t.call(this), (this.cleanup = null), null === (o = this.clickOutsideMixin) || void 0 === o || o.disconnectedCallback());
      }
      updatePosition() {
        this.cleanup = a(this.triggerRef, this.panelRef, void 0, { placement: this.position });
      }
      isGrouped() {
        var t;
        return (null === (t = this.options) || void 0 === t ? void 0 : t.length) > 0 && this.options[0].hasOwnProperty(this.groupNameKey);
      }
      getOptionLabel(t) {
        return 'object' == typeof t ? t[this.optionLabelKey] : t;
      }
      handleItemClick(t) {
        ((this.isOpen = !1), this.tkItemClick.emit(t));
      }
      createOptionItem(t) {
        return null == t
          ? void 0
          : t.map((t, o) => {
              let e;
              return (
                (e = null != this.optionHtml ? this.optionHtml(t) : this.getOptionLabel(t)),
                n('div', {
                  'class': d('tk-dropdown-item', this.optionsAlign, { disabled: null == t ? void 0 : t.disabled }),
                  'data-option-index': o,
                  'data-active': 0 == o ? 'true' : 'false',
                  'onClick': () => this.handleItemClick(t),
                  'innerHTML': e,
                })
              );
            });
      }
      createOptions() {
        var t;
        return this.isGrouped()
          ? null === (t = this.options) || void 0 === t
            ? void 0
            : t.map(t =>
                n(
                  'div',
                  { class: 'tk-dropdown-group' },
                  n('div', { class: 'tk-dropdown-group-label' }, n('label', null, t[this.groupNameKey]), n('div', { class: 'line' })),
                  this.createOptionItem(t[this.groupOptionsKey]),
                ),
              )
          : this.createOptionItem(this.options);
      }
      renderDropdown() {
        var t;
        return this.isOpen
          ? n(
              'div',
              { 'class': 'tk-dropdown-panel', 'ref': t => (this.panelRef = t), 'data-tk-dropdown-id': this.uniqueId },
              n(
                'div',
                { class: 'tk-dropdown-item-holder' },
                (null === (t = this.options) || void 0 === t ? void 0 : t.length) > 0
                  ? this.createOptions()
                  : this.hasEmptyDataSlot
                    ? n('slot', { name: 'empty-data' })
                    : this.emptyMessage,
              ),
            )
          : null;
      }
      render() {
        const t = d('tk-dropdown-container', this.size);
        return n(
          'div',
          { 'key': '50792dda7bef333a0a0d45ed6bad04dc26dc7534', 'class': t, 'data-tk-dropdown-id': this.uniqueId },
          n('slot', { key: '49a2bf076febbac475e3c9b23a0518b6db3d6ee0', name: 'trigger' }),
          this.renderDropdown(),
        );
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { disabled: [{ disabledChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-dropdown{display:inline-block}.tk-dropdown-container{position:relative}.tk-dropdown-container .tk-dropdown-panel{position:absolute;top:0;left:0;z-index:800;width:max-content;max-height:300px;overflow:hidden;border-radius:16px;border:1px solid var(--border-light);background:var(--static-light);box-shadow:var(--effect-1-default-sm);font-family:var(--family-body)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder{margin:8px;overflow-y:auto}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar{width:4px;height:4px}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-group-label{display:flex;align-items:center;gap:16px;color:var(--text-dark);font-size:var(--desktop-body-xs-size);padding:0px 8px}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-group-label label{white-space:nowrap}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-group-label .line{width:100%;height:1px;background-color:var(--background-light)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{color:var(--text-darkest);border-radius:8px;padding:var(--dropdown-items-basic-base-v-padding) var(--dropdown-items-basic-base-h-padding, 8px)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.left{text-align:left}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.center{text-align:center}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.right{text-align:right}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item:hover,.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item[data-active=true]{cursor:pointer;background:var(--background-lightest)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.disabled{color:var(--text-dark);pointer-events:none}.tk-dropdown-container.large .tk-dropdown-panel .tk-dropdown-item-holder{max-height:240px}.tk-dropdown-container.large .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{font-family:var(--desktop-body-m-base-font);font-size:var(--desktop-body-m-base-size);line-height:var(--desktop-body-m-base-line-height);padding:8px 8px}.tk-dropdown-container.base .tk-dropdown-panel .tk-dropdown-item-holder{max-height:216px}.tk-dropdown-container.base .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);line-height:var(--desktop-body-s-line-height);padding:8px 8px}.tk-dropdown-container.small .tk-dropdown-panel .tk-dropdown-item-holder{max-height:180px}.tk-dropdown-container.small .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{font-family:var(--desktop-body-xs-font);font-size:var(--desktop-body-xs-size);line-height:var(--desktop-body-xs-line-height);padding:6px 8px}';
      }
    },
    [
      260,
      'tk-dropdown',
      {
        disabled: [4],
        emptyMessage: [1, 'empty-message'],
        optionHtml: [16],
        optionLabelKey: [1, 'option-label-key'],
        optionValueKey: [1, 'option-value-key'],
        groupNameKey: [1, 'group-name-key'],
        groupOptionsKey: [1, 'group-options-key'],
        options: [16],
        optionsAlign: [1, 'options-align'],
        position: [1],
        size: [1],
        isOpen: [32],
      },
      void 0,
      { disabled: [{ disabledChanged: 0 }] },
    ],
  ),
  p = l,
  h = function () {
    'undefined' != typeof customElements &&
      ['tk-dropdown'].forEach(t => {
        'tk-dropdown' === t && (customElements.get(i(t)) || customElements.define(i(t), l));
      });
  };
export { p as TkDropdown, h as defineCustomElement };
