// @takeoff-ui/core@0.11.9/components/tk-dialog.js
import { p as t, H as a, c as i, h as e, t as o } from './p-BfLQgzDX.js';
import { c as s } from './p-D7YaJBkB.js';
import { g as l } from './p-Dpaq_7dX.js';
import { g as r } from './p-BDMnqEGL.js';
import { d } from './p-BY6PxsS2.js';
import { d as n } from './p-BpML1w-i.js';
import { d as g } from './p-C6qt54JA.js';
const h = t(
    class extends a {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          (this.tkOpen = i(this, 'tk-open', 3)),
          (this.tkClose = i(this, 'tk-close', 3)),
          (this.tkVisibleChange = i(this, 'tk-visible-change', 7)),
          (this.visible = !1),
          (this.headerType = 'basic'),
          (this.showCloseButton = !0),
          (this.showHeader = !0),
          (this.showVariantSign = !0),
          (this.variant = 'info'),
          (this.hideBackdrop = !1),
          (this.maskVariant = 'base'),
          (this.isMaskBlur = !1),
          (this.containerStyle = null),
          (this.preventDismiss = !1),
          (this.handleCloseButtonClick = () => {
            (this.tkClose.emit(), this.tkVisibleChange.emit(!1));
          }),
          (this.handleOverlayClick = () => {
            this.preventDismiss || (this.tkClose.emit(), this.tkVisibleChange.emit(!1));
          }));
      }
      visibleChanged(t) {
        t ? (this.tkOpen.emit(), this.hideBackdrop || this.blockBodyScroll()) : (this.tkClose.emit(), this.hideBackdrop || this.unblockBodyScroll());
      }
      componentWillLoad() {
        ((this.hasContainerSlot = !!this.el.querySelector(':scope > [slot="container"]')),
          (this.hasHeaderSlot = !!this.el.querySelector(':scope > [slot="header"]')),
          (this.hasContentSlot = !!this.el.querySelector(':scope > [slot="content"]')),
          (this.hasFooterSlot = !!this.el.querySelector(':scope > [slot="footer"]')),
          (this.hasFooterActionsSlot = !!this.el.querySelector(':scope > [slot="footer-actions"]')),
          (this.hasDefaultSlotContent = Array.from(this.el.childNodes).some(t => t.nodeType === Node.ELEMENT_NODE && !t.hasAttribute('slot'))));
      }
      disconnectedCallback() {
        this.hideBackdrop || this.unblockBodyScroll();
      }
      async open() {
        this.tkOpen.emit();
      }
      async close() {
        this.tkClose.emit();
      }
      blockBodyScroll() {
        if ('hidden' !== document.body.style.overflow) {
          ((this.bodyOverflow = document.body.style.overflow), (this.bodyPaddingRight = document.body.style.paddingRight));
          const t = window.innerWidth - document.documentElement.clientWidth;
          ((document.body.style.paddingRight = t + 'px'), (document.body.style.overflow = 'hidden'));
        }
      }
      unblockBodyScroll() {
        ((document.body.style.overflow = this.bodyOverflow), (document.body.style.paddingRight = this.bodyPaddingRight));
      }
      getVariantIcon() {
        switch (this.variant) {
          case 'success':
            return 'check_circle';
          case 'warning':
            return 'warning';
          case 'danger':
            return 'error';
          default:
            return 'info';
        }
      }
      createHeader() {
        if (this.showHeader) {
          if (this.hasHeaderSlot) return e('slot', { name: 'header' });
          {
            const t = s('tk-dialog-header', 'tk-dialog-header-' + this.headerType);
            return e(
              'div',
              { 'class': t, 'data-testid': r(this.dataTestid, 'header-container') },
              e(
                'div',
                { 'class': 'tk-dialog-header-content', 'data-testid': r(this.dataTestid, 'header') },
                this.showVariantSign &&
                  e(
                    'tk-icon',
                    Object.assign(
                      { sign: !0, dataTestid: r(this.dataTestid, 'sign-icon'), size: 'xlarge' },
                      l(this.getVariantIcon(), { class: s('fill tk-dialog-sign-icon'), variant: this.variant }, 'rounded', 'span'),
                    ),
                  ),
                e(
                  'div',
                  { 'class': 'tk-dialog-title-container', 'data-testid': r(this.dataTestid, 'title-container') },
                  this.subheader && e('span', { 'class': 'tk-dialog-subtitle', 'data-testid': r(this.dataTestid, 'subheader') }, this.subheader),
                  this.header && e('span', { 'class': 'tk-dialog-title', 'data-testid': r(this.dataTestid, 'title') }, this.header),
                ),
              ),
              this.showCloseButton &&
                e('tk-button', {
                  'variant': 'neutral',
                  'icon': 'close',
                  'size': 'small',
                  'type': 'text',
                  'onTk-click': this.handleCloseButtonClick,
                  'aria-label': 'Close dialog',
                  'dataTestid': r(this.dataTestid, 'close-button'),
                }),
            );
          }
        }
        return null;
      }
      createContent() {
        return this.hasContentSlot
          ? e('div', { 'class': 'tk-dialog-content', 'data-testid': r(this.dataTestid, 'content') }, e('slot', { name: 'content' }))
          : this.hasDefaultSlotContent
            ? e('div', { 'class': 'tk-dialog-content', 'data-testid': r(this.dataTestid, 'content') }, e('slot', null))
            : null;
      }
      createFooter() {
        return this.hasFooterSlot
          ? e('slot', { name: 'footer' })
          : this.hasFooterActionsSlot
            ? e('div', { 'class': 'tk-dialog-footer', 'data-testid': r(this.dataTestid, 'footer') }, e('slot', { name: 'footer-actions' }))
            : null;
      }
      createDialog() {
        const t = this.createHeader(),
          a = this.createContent(),
          i = this.createFooter(),
          o = {
            'class': s('tk-dialog', 'tk-dialog-' + this.variant),
            'style': Object.assign({ display: 'flex', flexDirection: 'column' }, this.containerStyle),
            'role': 'dialog',
            'aria-modal': !0,
            'data-testid': r(this.dataTestid, 'root'),
          };
        return e('div', Object.assign({}, o), t, a, i);
      }
      renderMask() {
        const t = this.hasContainerSlot ? e('slot', { name: 'container' }) : this.createDialog(),
          a = s('tk-dialog-mask', 'tk-dialog-mask-' + this.maskVariant, {
            'tk-dialog-visible': this.visible,
            'tk-dialog-mask-hidden': this.hideBackdrop,
            'tk-dialog-mask-blur': this.isMaskBlur,
          });
        return e(
          'div',
          { 'class': a, 'data-testid': r(this.dataTestid, 'mask') },
          e('div', { 'class': 'tk-dialog-overlay', 'onClick': () => this.handleOverlayClick(), 'data-testid': r(this.dataTestid, 'overlay') }),
          t,
        );
      }
      render() {
        return this.renderMask();
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { visible: [{ visibleChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-dialog{display:none}tk-dialog[visible],tk-dialog[visible=true]{display:block}.tk-dialog-mask{position:fixed;height:100%;width:100%;left:0;top:0;display:flex;justify-content:center;align-items:center;z-index:1500;opacity:0;transition:opacity 0.3s ease-in-out}.tk-dialog-mask.tk-dialog-mask-blur{backdrop-filter:blur(2px)}.tk-dialog-mask .tk-dialog-overlay{position:absolute;top:0;left:0;width:100%;height:100%;z-index:0}.tk-dialog-mask.tk-dialog-visible{opacity:1}.tk-dialog-mask.tk-dialog-mask-hidden{background-color:transparent}.tk-dialog-mask:not(.tk-dialog-mask-hidden).tk-dialog-mask-lightest{background-color:var(--overlay-black-lightest)}.tk-dialog-mask:not(.tk-dialog-mask-hidden).tk-dialog-mask-light{background-color:var(--overlay-black-light)}.tk-dialog-mask:not(.tk-dialog-mask-hidden).tk-dialog-mask-base{background-color:var(--overlay-black-base)}.tk-dialog-mask:not(.tk-dialog-mask-hidden).tk-dialog-mask-dark{background-color:var(--overlay-black-dark)}.tk-dialog-mask:not(.tk-dialog-mask-hidden).tk-dialog-mask-darkest{background-color:var(--overlay-black-darkest)}.tk-dialog{font-family:var(--family-body);display:flex;max-height:90%;flex-direction:column;align-items:flex-start;border-radius:var(--radius-xl);border:1px solid var(--border-lightest);background:var(--static-light);box-shadow:var(--effect-1-default-base);z-index:100}.tk-dialog .tk-dialog-header{display:flex;padding:var(--dialog-header-base-v-padding) var(--dialog-header-base-h-padding);justify-content:space-between;align-items:flex-start;align-self:stretch;border-top-left-radius:var(--radius-xl);border-top-right-radius:var(--radius-xl)}.tk-dialog .tk-dialog-header .tk-dialog-header-content{display:flex;align-items:center;gap:var(--dialog-header-base-gap)}.tk-dialog .tk-dialog-header .tk-dialog-header-content .tk-dialog-title-container{display:flex;flex-direction:column;align-items:flex-start}.tk-dialog .tk-dialog-header .tk-dialog-header-content .tk-dialog-title-container .tk-dialog-title{color:var(--text-darkest);font-family:var(--desktop-title-h5-font);font-size:var(--desktop-title-h5-size);font-style:normal;font-weight:400;line-height:var(--desktop-title-h5-line-height);}.tk-dialog .tk-dialog-header .tk-dialog-header-content .tk-dialog-title-container .tk-dialog-subtitle{color:var(--text-base);font-family:var(--desktop-subheading-xs-font);font-size:var(--desktop-subheading-xs-size);font-style:normal;font-weight:400;line-height:var(--desktop-subheading-xs-line-height);}.tk-dialog .tk-dialog-header .tk-dialog-header-close-button{display:flex;color:var(--icon-sub-base);width:var(--dialog-header-base-close-icon-size);height:var(--dialog-header-base-close-icon-size);padding:var(--button-sizes-small-full-padding);justify-content:center;align-items:center;gap:var(--dialog-header-small-gap);border-radius:var(--button-sizes-small-radius)}.tk-dialog .tk-dialog-header .tk-dialog-header-close-button:focus{box-shadow:var(--neutral-focus)}.tk-dialog .tk-dialog-header.tk-dialog-header-divided{border-bottom:var(--spacing-px) solid var(--border-light)}.tk-dialog .tk-dialog-header.tk-dialog-header-light{border-bottom:var(--spacing-px) solid var(--border-light);background:var(--background-lightest)}.tk-dialog .tk-dialog-header.tk-dialog-header-dark{background:var(--background-darkest)}.tk-dialog .tk-dialog-header.tk-dialog-header-dark .tk-dialog-title-container .tk-dialog-subtitle{color:var(--text-sub-base)}.tk-dialog .tk-dialog-header.tk-dialog-header-dark .tk-dialog-title-container .tk-dialog-title{color:var(--text-lightest)}.tk-dialog .tk-dialog-header.tk-dialog-header-dark .tk-dialog-header-close-button{color:var(--icon-base)}.tk-dialog .tk-dialog-header.tk-dialog-header-primary{background:var(--primary-base)}.tk-dialog .tk-dialog-header.tk-dialog-header-primary .tk-dialog-title-container .tk-dialog-subtitle{color:var(--primary-lightest)}.tk-dialog .tk-dialog-header.tk-dialog-header-primary .tk-dialog-title-container .tk-dialog-title{color:var(--text-lightest)}.tk-dialog .tk-dialog-header.tk-dialog-header-primary .tk-dialog-header-close-button{color:var(--primary-lightest)}.tk-dialog .tk-dialog-content{padding:var(--dialog-body-base-v-padding) var(--dialog-body-base-h-padding);align-self:stretch;overflow-y:auto;flex:1}.tk-dialog .tk-dialog-content::-webkit-scrollbar{width:4px;height:4px}.tk-dialog .tk-dialog-content::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-dialog .tk-dialog-content::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-dialog .tk-dialog-content::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-dialog .tk-dialog-content::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-dialog .tk-dialog-footer{display:flex;padding:var(--dialog-footer-base-v-padding) var(--dialog-footer-base-h-padding);justify-content:flex-end;align-self:stretch}.tk-dialog .tk-dialog-footer ::slotted([slot=footer-actions]){display:flex;align-items:center;gap:var(--dialog-footer-base-gap)}';
      }
    },
    [
      260,
      'tk-dialog',
      {
        visible: [516],
        header: [1],
        headerType: [1, 'header-type'],
        showCloseButton: [4, 'show-close-button'],
        showHeader: [4, 'show-header'],
        showVariantSign: [4, 'show-variant-sign'],
        subheader: [1],
        variant: [1],
        hideBackdrop: [4, 'hide-backdrop'],
        maskVariant: [1, 'mask-variant'],
        isMaskBlur: [4, 'is-mask-blur'],
        containerStyle: [16],
        preventDismiss: [4, 'prevent-dismiss'],
        dataTestid: [513, 'data-testid'],
        hasContainerSlot: [32],
        hasContentSlot: [32],
        hasDefaultSlotContent: [32],
        hasFooterActionsSlot: [32],
        hasFooterSlot: [32],
        hasHeaderSlot: [32],
        open: [64],
        close: [64],
      },
      void 0,
      { visible: [{ visibleChanged: 0 }] },
    ],
  ),
  k = h,
  c = function () {
    'undefined' != typeof customElements &&
      ['tk-dialog', 'tk-button', 'tk-icon', 'tk-spinner'].forEach(t => {
        switch (t) {
          case 'tk-dialog':
            customElements.get(o(t)) || customElements.define(o(t), h);
            break;
          case 'tk-button':
            customElements.get(o(t)) || d();
            break;
          case 'tk-icon':
            customElements.get(o(t)) || n();
            break;
          case 'tk-spinner':
            customElements.get(o(t)) || g();
        }
      });
  };
export { k as TkDialog, c as defineCustomElement };
