// @takeoff-ui/core@0.11.9/components/tk-drawer.js
import { p as t, H as r, c as e, h as a, t as o } from './p-BfLQgzDX.js';
import { c as s } from './p-D7YaJBkB.js';
import { g as i } from './p-BDMnqEGL.js';
import { d } from './p-BY6PxsS2.js';
import { d as n } from './p-BpML1w-i.js';
import { d as l } from './p-C6qt54JA.js';
const h = t(
    class extends r {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkDrawerOpen = e(this, 'tk-drawer-open', 3)),
          (this.tkDrawerClose = e(this, 'tk-drawer-close', 3)),
          (this.tkDrawerEnter = e(this, 'tk-drawer-enter', 7)),
          (this.tkDrawerLeave = e(this, 'tk-drawer-leave', 7)),
          (this.tkDrawerChange = e(this, 'tk-drawer-change', 7)),
          (this.containerVisible = !1),
          (this.isOpen = !1),
          (this.open = !1),
          (this.position = 'right'),
          (this.hideBackdrop = !1),
          (this.maskVariant = 'base'),
          (this.containerStyle = null),
          (this.preventDismiss = !1),
          (this.hideCloseIcon = !1),
          (this.unblockScroll = !1),
          (this.headerType = 'basic'),
          (this.footerType = 'basic'),
          (this.handleOverlayClick = () => {
            this.preventDismiss || this.tkDrawerClose.emit();
          }),
          (this.handleCloseButtonClick = () => {
            this.tkDrawerClose.emit();
          }));
      }
      openChanged(t) {
        (t
          ? requestAnimationFrame(() => {
              ((this.isOpen = !0), this.onEnter());
            })
          : ((this.isOpen = !1), this.onLeave()),
          (this.containerVisible = t),
          this.tkDrawerChange.emit(t));
      }
      componentWillLoad() {
        this.open &&
          ((this.containerVisible = !0),
          requestAnimationFrame(() => {
            ((this.isOpen = !0), this.onEnter());
          }));
      }
      async show() {
        this.tkDrawerOpen.emit();
      }
      async close() {
        this.tkDrawerClose.emit();
      }
      blockBodyScroll() {
        if ('hidden' !== document.body.style.overflow) {
          ((this.bodyOverflow = document.body.style.overflow), (this.bodyPaddingRight = document.body.style.paddingRight));
          const t = window.innerWidth - document.documentElement.clientWidth;
          ((document.body.style.paddingRight = t + 'px'), (document.body.style.overflow = 'hidden'));
        }
      }
      unblockBodyScroll() {
        ((document.body.style.overflow = this.bodyOverflow), (document.body.style.paddingRight = this.bodyPaddingRight));
      }
      getTransformValue() {
        if (!this.isOpen)
          switch (this.position) {
            case 'left':
              return 'translateX(-100%)';
            case 'right':
              return 'translateX(100%)';
            case 'top':
              return 'translateY(-100%)';
            case 'bottom':
              return 'translateY(100%)';
            default:
              return '';
          }
        return 'translate(0, 0)';
      }
      onEnter() {
        (this.unblockScroll || this.hideBackdrop || this.blockBodyScroll(),
          this.tkDrawerEnter.emit(),
          setTimeout(() => {
            this.tkDrawerOpen.emit();
          }, 300));
      }
      onLeave() {
        (this.unblockScroll || this.hideBackdrop || this.unblockBodyScroll(), this.tkDrawerLeave.emit(), (this.containerVisible = !1), this.tkDrawerClose.emit());
      }
      createHeader() {
        const t = !!this.el.querySelector('[slot="header"]'),
          r = !!this.el.querySelector('[slot="header-actions"]');
        return a(
          'div',
          { 'class': s('tk-drawer-header', 'tk-drawer-header-' + this.headerType), 'data-testid': i(this.dataTestid, 'header') },
          t ? a('slot', { name: 'header' }) : this.header ? a('span', { 'class': 'tk-drawer-header-label', 'data-testid': i(this.dataTestid, 'label') }, this.header) : null,
          r
            ? a('slot', { name: 'header-actions' })
            : !this.hideCloseIcon &&
                a('tk-button', {
                  'class': 'tk-drawer-close',
                  'variant': 'neutral',
                  'icon': 'close',
                  'type': 'text',
                  'onTk-click': this.handleCloseButtonClick,
                  'dataTestid': i(this.dataTestid, 'close-button'),
                }),
        );
      }
      createFooter() {
        return this.el.querySelector('[slot="footer"]')
          ? a('div', { 'class': s('tk-drawer-footer', 'tk-drawer-footer-' + this.footerType), 'data-testid': i(this.dataTestid, 'footer') }, a('slot', { name: 'footer' }))
          : null;
      }
      createDrawer() {
        const t = s('tk-drawer', 'tk-drawer-' + this.position, { 'tk-drawer-open': this.isOpen }),
          r = Object.assign(Object.assign({}, this.containerStyle && Object.assign({}, this.containerStyle)), { transform: this.getTransformValue() });
        return a(
          'div',
          { 'class': t, 'style': r, 'data-testid': i(this.dataTestid, 'panel') },
          a(
            'div',
            { 'class': 'tk-drawer-content', 'data-testid': i(this.dataTestid, 'content') },
            a(
              'slot',
              { name: 'container' },
              this.createHeader(),
              a('div', { 'class': 'tk-drawer-body', 'data-testid': i(this.dataTestid, 'body') }, a('slot', { name: 'content' })),
              this.createFooter(),
            ),
          ),
        );
      }
      renderMask() {
        const t = s('tk-drawer-mask', 'tk-drawer-mask-' + this.maskVariant, { 'tk-drawer-visible': this.containerVisible, 'tk-drawer-mask-hidden': this.hideBackdrop });
        return a(
          'div',
          { 'class': t, 'data-testid': i(this.dataTestid, 'mask') },
          a('div', { 'class': 'tk-drawer-overlay', 'onClick': () => this.handleOverlayClick(), 'data-testid': i(this.dataTestid, 'overlay') }),
          this.createDrawer(),
        );
      }
      render() {
        return this.containerVisible && this.renderMask();
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { open: [{ openChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-drawer-mask{position:fixed;height:100%;width:100%;left:0;top:0;display:flex;z-index:1500;opacity:0;transition:opacity 0.3s ease-in-out}.tk-drawer-mask .tk-drawer-overlay{position:absolute;top:0;left:0;width:100%;height:100%;z-index:0}.tk-drawer-mask.tk-drawer-visible{opacity:1}.tk-drawer-mask.tk-drawer-mask-hidden{background-color:transparent}.tk-drawer-mask:not(.tk-drawer-mask-hidden).tk-drawer-mask-lightest{background-color:var(--overlay-black-lightest)}.tk-drawer-mask:not(.tk-drawer-mask-hidden).tk-drawer-mask-light{background-color:var(--overlay-black-light)}.tk-drawer-mask:not(.tk-drawer-mask-hidden).tk-drawer-mask-base{background-color:var(--overlay-black-base)}.tk-drawer-mask:not(.tk-drawer-mask-hidden).tk-drawer-mask-dark{background-color:var(--overlay-black-dark)}.tk-drawer-mask:not(.tk-drawer-mask-hidden).tk-drawer-mask-darkest{background-color:var(--overlay-black-darkest)}.tk-drawer{position:fixed;border:1px solid var(--border-light);background-color:var(--static-light);box-shadow:0 0 20px rgba(0, 0, 0, 0.1);transition:transform 0.3s ease-in-out;display:flex;flex-direction:column;transition:transform 0.3s cubic-bezier(0.4, 0, 0.2, 1);will-change:transform;z-index:100}.tk-drawer.tk-drawer-full-screen{width:100vw !important;height:100vh !important;max-height:100%;top:0px !important;left:0px !important;transition:none;transform:none}.tk-drawer.tk-drawer-left,.tk-drawer.tk-drawer-right{top:0;bottom:0;height:100%;width:344px}.tk-drawer.tk-drawer-top,.tk-drawer.tk-drawer-bottom{left:0;right:0;width:100%;height:344px}.tk-drawer.tk-drawer-left{left:0;border-radius:0px 8px 8px 0px}.tk-drawer.tk-drawer-left .tk-drawer-header{border-top-right-radius:var(--radius-m-base)}.tk-drawer.tk-drawer-left .tk-drawer-footer{border-bottom-right-radius:var(--radius-m-base)}.tk-drawer.tk-drawer-right{right:0;border-radius:8px 0px 0px 8px}.tk-drawer.tk-drawer-right .tk-drawer-header{border-top-left-radius:var(--radius-m-base)}.tk-drawer.tk-drawer-right .tk-drawer-footer{border-bottom-left-radius:var(--radius-m-base)}.tk-drawer.tk-drawer-top{top:0;border-radius:0px 0px 8px 8px}.tk-drawer.tk-drawer-top .tk-drawer-header{border-radius:0}.tk-drawer.tk-drawer-top .tk-drawer-footer{border-bottom-left-radius:var(--radius-m-base);border-bottom-right-radius:var(--radius-m-base)}.tk-drawer.tk-drawer-bottom{bottom:0;border-radius:8px 8px 0px 0px}.tk-drawer.tk-drawer-bottom .tk-drawer-header{border-top-left-radius:var(--radius-m-base);border-top-right-radius:var(--radius-m-base)}.tk-drawer.tk-drawer-bottom .tk-drawer-footer{border-radius:0}.tk-drawer-content{display:flex;flex-direction:column;height:100%;font-family:var(--family-body)}.tk-drawer-header{display:flex;justify-content:space-between;align-items:center;padding:var(--notification-header-v-padding) var(--notification-header-h-padding)}.tk-drawer-header .tk-drawer-header-label{color:var(--text-darkest);font-family:var(--desktop-body-m-base-font);font-size:var(--desktop-body-m-base-size);font-style:normal;font-weight:400;line-height:var(--desktop-body-m-base-line-height)}.tk-drawer-header.tk-drawer-header-divided,.tk-drawer-header.tk-drawer-header-light{border-bottom:var(--spacing-px) solid var(--border-light)}.tk-drawer-header.tk-drawer-header-light{background:var(--background-lightest)}.tk-drawer-header.tk-drawer-header-dark{background:var(--background-darkest)}.tk-drawer-header.tk-drawer-header-dark .tk-drawer-header-label{color:var(--text-lightest)}.tk-drawer-header.tk-drawer-header-dark .tk-drawer-close{color:var(--icon-base)}.tk-drawer-header.tk-drawer-header-primary{background:var(--primary-base)}.tk-drawer-header.tk-drawer-header-primary .tk-drawer-header-label{color:var(--text-lightest)}.tk-drawer-header.tk-drawer-header-primary .tk-drawer-close{color:var(--primary-lightest)}.tk-drawer-body{flex-grow:1;overflow-y:auto;padding:0 var(--notification-body-h-padding)}.tk-drawer-body::-webkit-scrollbar{width:4px;height:4px}.tk-drawer-body::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-drawer-body::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-drawer-body::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-drawer-body::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-drawer-footer{padding:var(--notification-footer-v-padding) var(--notification-footer-h-padding);box-shadow:var(--effect-2-default-base)}.tk-drawer-footer ::slotted([slot=footer]){display:flex;justify-content:flex-end;align-items:center;gap:var(--settings-footer-gap, 8px)}.tk-drawer-footer.tk-drawer-footer-divided,.tk-drawer-footer.tk-drawer-footer-light{border-top:var(--spacing-px, 1px) solid var(--border-light)}.tk-drawer-footer.tk-drawer-footer-light{background:var(--background-lightest)}';
      }
    },
    [
      257,
      'tk-drawer',
      {
        open: [516],
        position: [1],
        hideBackdrop: [4, 'hide-backdrop'],
        maskVariant: [1, 'mask-variant'],
        containerStyle: [16],
        preventDismiss: [4, 'prevent-dismiss'],
        hideCloseIcon: [4, 'hide-close-icon'],
        unblockScroll: [4, 'unblock-scroll'],
        header: [1],
        headerType: [1, 'header-type'],
        footerType: [1, 'footer-type'],
        dataTestid: [513, 'data-testid'],
        containerVisible: [32],
        isOpen: [32],
        show: [64],
        close: [64],
      },
      void 0,
      { open: [{ openChanged: 0 }] },
    ],
  ),
  k = h,
  w = function () {
    'undefined' != typeof customElements &&
      ['tk-drawer', 'tk-button', 'tk-icon', 'tk-spinner'].forEach(t => {
        switch (t) {
          case 'tk-drawer':
            customElements.get(o(t)) || customElements.define(o(t), h);
            break;
          case 'tk-button':
            customElements.get(o(t)) || d();
            break;
          case 'tk-icon':
            customElements.get(o(t)) || n();
            break;
          case 'tk-spinner':
            customElements.get(o(t)) || l();
        }
      });
  };
export { k as TkDrawer, w as defineCustomElement };
