// @takeoff-ui/core@0.11.9/components/tk-gantt-chart.js
import { p as t, H as a, c as e, h as i, F as n, t as r } from './p-BfLQgzDX.js';
import { c as s } from './p-D7YaJBkB.js';
import { g as o } from './p-BDMnqEGL.js';
import { d as l } from './p-BpML1w-i.js';
import { d } from './p-tm7ZzNFD.js';
const h = { weekly: 40, monthly: 24, quarterly: 6, yearly: 2 },
  c = [
    { field: 'name', header: 'Task Name', width: '200px' },
    { field: 'startDate', header: 'Start', width: '100px' },
    { field: 'endDate', header: 'End', width: '100px' },
  ];
function p(t) {
  const a = new Date('string' == typeof t ? t : t.getTime());
  return (a.setHours(0, 0, 0, 0), a);
}
function g(t, a) {
  return Math.max(Math.round((a.getTime() - t.getTime()) / 864e5), 0);
}
function k(t, a) {
  const e = new Date(t.getTime());
  return (e.setDate(e.getDate() + a), e);
}
function b(t) {
  const a = t.getDay();
  return 0 === a || 6 === a;
}
function m(t) {
  const a = new Date(Date.UTC(t.getFullYear(), t.getMonth(), t.getDate()));
  a.setUTCDate(a.getUTCDate() + 4 - (a.getUTCDay() || 7));
  const e = new Date(Date.UTC(a.getUTCFullYear(), 0, 1));
  return Math.ceil(((a.getTime() - e.getTime()) / 864e5 + 1) / 7);
}
function u(t, a) {
  const e = new Date(t.getTime());
  e.setHours(0, 0, 0, 0);
  const i = (e.getDay() - a + 7) % 7;
  return (e.setDate(e.getDate() - i), e);
}
function f(t) {
  return Math.floor(t.getMonth() / 3) + 1;
}
function w(t) {
  let a = 1 / 0,
    e = -1 / 0;
  const i = t => {
    var n, r;
    for (const s of t) {
      const t = p(s.startDate).getTime(),
        o = p(s.endDate).getTime();
      if ((t < a && (a = t), o > e && (e = o), null === (n = s.segments) || void 0 === n ? void 0 : n.length))
        for (const t of s.segments) {
          const i = p(t.startDate).getTime(),
            n = p(t.endDate).getTime();
          (i < a && (a = i), n > e && (e = n));
        }
      (null === (r = s.children) || void 0 === r ? void 0 : r.length) && i(s.children);
    }
  };
  return (i(t), { min: new Date(a), max: k(new Date(e), 1) });
}
function y(t, a, e = 0) {
  var i;
  const n = [];
  for (const r of t) (n.push({ task: r, depth: e }), (null === (i = r.children) || void 0 === i ? void 0 : i.length) && a.has(r.id) && n.push(...y(r.children, a, e + 1)));
  return n;
}
function v(t) {
  return t.toLowerCase().startsWith('tr') ? 'H' : 'W';
}
function x(t, a, e) {
  return `${t.toLocaleString(e, { month: 'short', day: 'numeric' })}–${a.getDate()}`;
}
function D(t, a) {
  const e = p(t).getTime();
  return a.some(t => p(t.date).getTime() === e);
}
const S = t(
    class extends a {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkTaskClick = e(this, 'tk-task-click', 7)),
          (this.tkTaskToggle = e(this, 'tk-task-toggle', 7)),
          (this.expandedIds = new Set()),
          (this.computedViewType = 'monthly'),
          (this.tasks = []),
          (this.columns = c),
          (this.secondaryHeaderMode = 'days'),
          (this.weekStartDay = 1),
          (this.indicators = []),
          (this.showTodayIndicator = !0),
          (this.highlightWeekends = !0),
          (this.holidays = []),
          (this.rowHeight = 40),
          (this.panelWidth = 320),
          (this.hidePanel = !1),
          (this.locale = 'default'));
      }
      tasksChanged() {
        this.recalculate();
      }
      viewTypeChanged() {
        this.recalculate();
      }
      componentWillLoad() {
        this.recalculate();
      }
      recalculate() {
        var t;
        this.computedViewType =
          null !== (t = this.viewType) && void 0 !== t
            ? t
            : (function (t) {
                if (!t.length) return 'monthly';
                const { min: a, max: e } = w(t),
                  i = g(a, e);
                return i <= 14 ? 'weekly' : i <= 90 ? 'monthly' : i <= 365 ? 'quarterly' : 'yearly';
              })(this.tasks);
      }
      handleToggleRow(t, a) {
        a.stopPropagation();
        const e = new Set(this.expandedIds),
          i = e.has(t.id);
        (i ? e.delete(t.id) : e.add(t.id), (this.expandedIds = e), this.tkTaskToggle.emit({ task: t, expanded: !i }));
      }
      handleTaskClick(t) {
        this.tkTaskClick.emit(t);
      }
      handleTimelineScroll(t) {
        this.panelRef && (this.panelRef.scrollTop = t.target.scrollTop);
      }
      handlePanelScroll(t) {
        this.timelineRef && (this.timelineRef.scrollTop = t.target.scrollTop);
      }
      createTaskBar(t, a, e, n) {
        var r, l, d;
        const h = t.id + '',
          c = null != n ? `${h}-segment-${n}` : h,
          p = null !== (r = null == e ? void 0 : e.progress) && void 0 !== r ? r : t.progress,
          g = null !== (l = null == e ? void 0 : e.style) && void 0 !== l ? l : t.style,
          k = null !== (d = null == e ? void 0 : e.label) && void 0 !== d ? d : t.name,
          b = Object.assign({ left: a.left + 'px', width: a.width + 'px', height: this.rowHeight - 12 + 'px' }, g),
          m = this.createTooltipContent(t, c);
        if (this.taskBarHtml) {
          const a = this.taskBarHtml(t),
            e = i(
              'div',
              'string' == typeof a
                ? { 'class': 'custom-task-bar', 'innerHTML': a, 'data-testid': o(this.dataTestid, 'custom-task-bar', c) }
                : { 'class': 'custom-task-bar', 'ref': t => t && t.replaceChildren(a), 'data-testid': o(this.dataTestid, 'custom-task-bar', c) },
            );
          return i(
            'tk-tooltip',
            { position: 'top', variant: 'dark', dataTestid: o(this.dataTestid, 'custom-task-bar-tooltip', c) },
            i(
              'div',
              {
                'slot': 'trigger',
                'class': 'tk-gantt-chart-task-bar tk-gantt-chart-task-bar-custom',
                'style': b,
                'onClick': () => this.handleTaskClick(t),
                'data-testid': o(this.dataTestid, 'custom-task-bar-trigger', c),
              },
              e,
            ),
            i('div', { 'slot': 'content', 'data-testid': o(this.dataTestid, 'custom-task-bar-tooltip-content', c) }, m),
          );
        }
        return i(
          'tk-tooltip',
          { position: 'top', variant: 'dark', dataTestid: o(this.dataTestid, 'task-bar-tooltip', c) },
          i(
            'div',
            {
              'slot': 'trigger',
              'class': s('tk-gantt-chart-task-bar', { 'tk-gantt-chart-task-bar-segment': null != n }),
              'style': b,
              'onClick': () => this.handleTaskClick(t),
              'data-testid': o(this.dataTestid, 'task-bar', c),
            },
            null != p &&
              p > 0 &&
              i('div', { 'class': 'tk-gantt-chart-task-bar-progress', 'style': { width: Math.min(p, 100) + '%' }, 'data-testid': o(this.dataTestid, 'task-bar-progress', c) }),
            i('span', { 'class': 'tk-gantt-chart-task-bar-label', 'data-testid': o(this.dataTestid, 'task-bar-label', c) }, k),
          ),
          i('div', { 'slot': 'content', 'data-testid': o(this.dataTestid, 'task-bar-tooltip-content', c) }, m),
        );
      }
      createIndicatorLine(t, a, e, n) {
        const r = (function (t, a, e) {
          const i = h[e];
          return g(a, p(t)) * i;
        })(t.date, a, this.computedViewType);
        return i(
          'div',
          {
            'class': s('tk-gantt-chart-indicator', { 'tk-gantt-chart-indicator-dashed': 'dashed' === t.lineStyle }),
            'style': { left: r + 'px', height: e + 'px', borderLeftColor: t.color || 'var(--primary-base)', borderLeftStyle: 'dashed' === t.lineStyle ? 'dashed' : 'solid' },
            'data-testid': o(this.dataTestid, 'indicator', n),
          },
          t.label && i('span', { 'class': 'tk-gantt-chart-indicator-label', 'data-testid': o(this.dataTestid, 'indicator-label', n) }, t.label),
        );
      }
      createTooltipContent(t, a) {
        if (this.tooltipHtml) {
          const e = this.tooltipHtml(t);
          return i(
            'div',
            'string' == typeof e
              ? { 'innerHTML': e, 'data-testid': o(this.dataTestid, 'custom-tooltip', a) }
              : { 'ref': t => t && t.replaceChildren(e), 'data-testid': o(this.dataTestid, 'custom-tooltip', a) },
          );
        }
        const e = p(t.startDate).toLocaleDateString(this.locale),
          r = p(t.endDate).toLocaleDateString(this.locale);
        return i(
          n,
          null,
          i('div', { 'class': 'tk-gantt-chart-tooltip-title', 'data-testid': o(this.dataTestid, 'tooltip-title', a) }, t.name),
          i(
            'div',
            { 'class': 'tk-gantt-chart-tooltip-row', 'data-testid': o(this.dataTestid, 'tooltip-start', a) },
            i('span', { 'class': 'tk-gantt-chart-tooltip-label', 'data-testid': o(this.dataTestid, 'tooltip-start-label', a) }, 'Start:'),
            e,
          ),
          i(
            'div',
            { 'class': 'tk-gantt-chart-tooltip-row', 'data-testid': o(this.dataTestid, 'tooltip-end', a) },
            i('span', { 'class': 'tk-gantt-chart-tooltip-label', 'data-testid': o(this.dataTestid, 'tooltip-end-label', a) }, 'End:'),
            r,
          ),
          null != t.progress &&
            i(
              'div',
              { 'class': 'tk-gantt-chart-tooltip-row', 'data-testid': o(this.dataTestid, 'tooltip-progress', a) },
              i('span', { 'class': 'tk-gantt-chart-tooltip-label', 'data-testid': o(this.dataTestid, 'tooltip-progress-label', a) }, 'Progress:'),
              t.progress,
              '%',
            ),
        );
      }
      renderPanel(t) {
        return this.hidePanel
          ? null
          : i(
              'div',
              {
                'class': 'tk-gantt-chart-panel',
                'style': { width: this.panelWidth + 'px' },
                'ref': t => (this.panelRef = t),
                'onScroll': t => this.handlePanelScroll(t),
                'data-testid': o(this.dataTestid, 'panel'),
              },
              i(
                'div',
                { 'class': 'tk-gantt-chart-panel-header', 'style': { height: '72px' }, 'data-testid': o(this.dataTestid, 'panel-header') },
                this.columns.map(t =>
                  i(
                    'div',
                    {
                      'class': 'tk-gantt-chart-panel-header-cell',
                      'style': { width: t.width || 'auto', minWidth: t.width || '80px' },
                      'data-testid': o(this.dataTestid, 'panel-header-cell', t.field + ''),
                    },
                    t.header,
                  ),
                ),
              ),
              i(
                'div',
                { 'class': 'tk-gantt-chart-panel-body', 'data-testid': o(this.dataTestid, 'panel-body') },
                t.map(({ task: t, depth: a }) =>
                  i(
                    'div',
                    { 'class': 'tk-gantt-chart-panel-row', 'style': { height: this.rowHeight + 'px' }, 'data-testid': o(this.dataTestid, 'panel-row', t.id + '') },
                    this.columns.map((e, n) => {
                      var r, l;
                      return i(
                        'div',
                        {
                          'class': 'tk-gantt-chart-panel-cell',
                          'style': { width: e.width || 'auto', minWidth: e.width || '80px', paddingLeft: 0 === n ? 20 * a + 8 + 'px' : '8px' },
                          'data-testid': o(this.dataTestid, 'panel-cell', t.id + '', e.field + ''),
                        },
                        0 === n &&
                          (null !== (l = null === (r = t.children) || void 0 === r ? void 0 : r.length) && void 0 !== l ? l : 0) > 0 &&
                          i(
                            'button',
                            {
                              'class': s('tk-gantt-chart-expand-btn', { 'tk-gantt-chart-expand-btn-open': this.expandedIds.has(t.id) }),
                              'onClick': a => this.handleToggleRow(t, a),
                              'data-testid': o(this.dataTestid, 'panel-expand-button', t.id + ''),
                            },
                            i('tk-icon', { icon: 'chevron_right', size: 'small', dataTestid: o(this.dataTestid, 'panel-expand-icon', t.id + '') }),
                          ),
                        e.html
                          ? i('span', { 'innerHTML': e.html(t), 'data-testid': o(this.dataTestid, 'panel-cell-content', t.id + '', e.field + '') })
                          : i('span', { 'data-testid': o(this.dataTestid, 'panel-cell-content', t.id + '', e.field + '') }, t[e.field]),
                      );
                    }),
                  ),
                ),
              ),
            );
      }
      renderTimelineHeader(t, a, e) {
        const n = this.computedViewType,
          r = h[n],
          l = (function (t, a, e, i) {
            const n = [];
            switch (e) {
              case 'weekly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = x(e, k(e, 6), i);
                  (n.push({ label: t, span: 7, startDate: new Date(e.getTime()) }), (e = k(e, 7)));
                }
                break;
              }
              case 'monthly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = e.getFullYear(),
                    r = e.getMonth(),
                    s = new Date(t, r + 1, 1),
                    o = g(e, s < a ? s : a),
                    l = e.toLocaleString(i, { month: 'long', year: 'numeric' });
                  (n.push({ label: l, span: o, startDate: new Date(e.getTime()) }), (e = s));
                }
                break;
              }
              case 'quarterly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = f(e),
                    i = e.getFullYear(),
                    r = new Date(i, 3 * t, 1),
                    s = g(e, r < a ? r : a);
                  (n.push({ label: `Q${t} ${i}`, span: s, startDate: new Date(e.getTime()) }), (e = r));
                }
                break;
              }
              case 'yearly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = e.getFullYear(),
                    i = new Date(t + 1, 0, 1),
                    r = g(e, i < a ? i : a);
                  (n.push({ label: '' + t, span: r, startDate: new Date(e.getTime()) }), (e = i));
                }
                break;
              }
            }
            return n;
          })(t, a, n, this.locale),
          d = (function (t, a, e, i, n, r = 'days') {
            const s = [];
            if ('weeks' === r)
              return (function (t, a, e, i) {
                const n = [];
                let r = u(t, i);
                for (r < t && (r = new Date(t.getTime())); r < a; ) {
                  const t = m(r),
                    s = v(e),
                    o = k(u(r, i), 7),
                    l = g(r, o < a ? o : a);
                  (l > 0 && n.push({ label: `${s}${t}`, span: l, startDate: new Date(r.getTime()) }), (r = o));
                }
                return n;
              })(t, a, i, n);
            switch (e) {
              case 'weekly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = e.toLocaleString(i, { weekday: 'short' });
                  (s.push({ label: t, span: 1, startDate: new Date(e.getTime()) }), (e = k(e, 1)));
                }
                break;
              }
              case 'monthly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = '' + e.getDate();
                  (s.push({ label: t, span: 1, startDate: new Date(e.getTime()) }), (e = k(e, 1)));
                }
                break;
              }
              case 'quarterly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = e.getFullYear(),
                    n = e.getMonth(),
                    r = new Date(t, n + 1, 1),
                    o = g(e, r < a ? r : a),
                    l = e.toLocaleString(i, { month: 'short' });
                  (s.push({ label: l, span: o, startDate: new Date(e.getTime()) }), (e = r));
                }
                break;
              }
              case 'yearly': {
                let e = new Date(t.getTime());
                for (; e < a; ) {
                  const t = e.getFullYear(),
                    n = e.getMonth(),
                    r = new Date(t, n + 1, 1),
                    o = g(e, r < a ? r : a),
                    l = e.toLocaleString(i, { month: 'short' });
                  (s.push({ label: l, span: o, startDate: new Date(e.getTime()) }), (e = r));
                }
                break;
              }
            }
            return s;
          })(t, a, n, this.locale, this.weekStartDay, this.secondaryHeaderMode);
        return i(
          'div',
          { 'class': 'tk-gantt-chart-timeline-headers', 'style': { width: e + 'px' }, 'data-testid': o(this.dataTestid, 'timeline-headers') },
          i(
            'div',
            {
              'class': 'tk-gantt-chart-timeline-header tk-gantt-chart-timeline-header-primary',
              'style': { height: '36px' },
              'data-testid': o(this.dataTestid, 'timeline-primary'),
            },
            l.map((t, a) =>
              i(
                'div',
                { 'class': 'tk-gantt-chart-timeline-header-cell', 'style': { width: t.span * r + 'px' }, 'data-testid': o(this.dataTestid, 'timeline-primary-cell', '' + a) },
                t.label,
              ),
            ),
          ),
          i(
            'div',
            {
              'class': 'tk-gantt-chart-timeline-header tk-gantt-chart-timeline-header-secondary',
              'style': { height: '36px' },
              'data-testid': o(this.dataTestid, 'timeline-secondary'),
            },
            d.map((t, a) => {
              const e = t.startDate,
                n = 1 === t.span && b(e),
                l = 1 === t.span && D(e, this.holidays);
              return i(
                'div',
                {
                  'class': s('tk-gantt-chart-timeline-header-cell', {
                    'tk-gantt-chart-timeline-header-cell-weekend': n && this.highlightWeekends,
                    'tk-gantt-chart-timeline-header-cell-holiday': l,
                  }),
                  'style': { width: t.span * r + 'px' },
                  'data-testid': o(this.dataTestid, 'timeline-secondary-cell', '' + a),
                },
                t.label,
              );
            }),
          ),
        );
      }
      renderTimelineBody(t, a, e, n) {
        const r = h[this.computedViewType],
          l = t.length * this.rowHeight;
        return i(
          'div',
          { 'class': 'tk-gantt-chart-timeline-body', 'style': { width: e + 'px' }, 'data-testid': o(this.dataTestid, 'timeline-body') },
          (this.highlightWeekends || this.holidays.length > 0) &&
            Array.from({ length: n }).map((t, e) => {
              const n = k(a, e),
                d = this.highlightWeekends && b(n),
                h = D(n, this.holidays);
              return d || h
                ? i('div', {
                    'class': s('tk-gantt-chart-column-highlight', { 'tk-gantt-chart-weekend': d, 'tk-gantt-chart-holiday': h }),
                    'style': { left: e * r + 'px', width: r + 'px', height: l + 'px' },
                    'data-testid': o(this.dataTestid, 'column-highlight', '' + e),
                  })
                : null;
            }),
          this.showTodayIndicator && this.createIndicatorLine({ date: new Date().toISOString(), label: 'Today', color: 'var(--primary-base)', lineStyle: 'solid' }, a, l, 'today'),
          this.indicators.map((t, e) => this.createIndicatorLine(t, a, l, 'custom-' + e)),
          t.map(({ task: t }) => {
            const e = t.segments && t.segments.length > 0;
            return i(
              'div',
              { 'class': 'tk-gantt-chart-timeline-row', 'style': { height: this.rowHeight + 'px' }, 'data-testid': o(this.dataTestid, 'timeline-row', t.id + '') },
              e
                ? t.segments.map((e, i) => {
                    const n = (function (t, a, e, i) {
                      const n = h[i],
                        r = p(t),
                        s = p(a);
                      return { left: g(e, r) * n, width: Math.max((g(r, s) + 1) * n, n) };
                    })(e.startDate, e.endDate, a, this.computedViewType);
                    return this.createTaskBar(t, n, e, i);
                  })
                : this.createTaskBar(
                    t,
                    (function (t, a, e) {
                      const i = h[e],
                        n = p(t.startDate),
                        r = p(t.endDate);
                      return { left: g(a, n) * i, width: Math.max((g(n, r) + 1) * i, i) };
                    })(t, a, this.computedViewType),
                  ),
            );
          }),
        );
      }
      renderEmptyState() {
        return i(
          'div',
          { 'class': 'tk-gantt-chart-empty', 'data-testid': o(this.dataTestid, 'empty') },
          i('slot', { name: 'empty' }, i('span', { 'data-testid': o(this.dataTestid, 'empty-text') }, 'No tasks to display')),
        );
      }
      render() {
        if (!this.tasks.length)
          return i('div', { 'class': 'tk-gantt-chart', 'style': this.containerStyle, 'data-testid': o(this.dataTestid, 'container') }, this.renderEmptyState());
        const t = y(this.tasks, this.expandedIds),
          {
            start: a,
            end: e,
            totalDays: n,
          } = (function (t, a, e) {
            if (!t.length) {
              const t = p(new Date());
              return { start: t, end: k(t, 30), totalDays: 30 };
            }
            const { min: i, max: n } = w(t);
            let r, s;
            switch (a) {
              case 'weekly':
                ((r = u(i, e)), (s = k(u(n, e), 7)));
                break;
              case 'monthly':
                ((r = new Date(i.getFullYear(), i.getMonth(), 1)), (s = new Date(n.getFullYear(), n.getMonth() + 1, 1)));
                break;
              case 'quarterly': {
                const t = 3 * Math.floor(i.getMonth() / 3);
                r = new Date(i.getFullYear(), t, 1);
                const a = 3 * Math.floor(n.getMonth() / 3) + 3;
                s = new Date(n.getFullYear(), a, 1);
                break;
              }
              case 'yearly':
                ((r = new Date(i.getFullYear(), 0, 1)), (s = new Date(n.getFullYear() + 1, 0, 1)));
                break;
              default:
                ((r = i), (s = n));
            }
            return { start: r, end: s, totalDays: g(r, s) };
          })(this.tasks, this.computedViewType, this.weekStartDay),
          r = n * h[this.computedViewType];
        return i(
          'div',
          { 'class': 'tk-gantt-chart', 'style': this.containerStyle, 'data-testid': o(this.dataTestid, 'container') },
          this.renderPanel(t),
          i(
            'div',
            {
              'class': 'tk-gantt-chart-timeline',
              'ref': t => (this.timelineRef = t),
              'onScroll': t => this.handleTimelineScroll(t),
              'data-testid': o(this.dataTestid, 'timeline'),
            },
            this.renderTimelineHeader(a, e, r),
            this.renderTimelineBody(t, a, r, n),
          ),
        );
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { tasks: [{ tasksChanged: 0 }], viewType: [{ viewTypeChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-gantt-chart{display:flex;position:relative;overflow:hidden;border:1px solid var(--border-light);border-radius:var(--radius-m-base);background:var(--static-white);font-size:var(--desktop-body-s-size);color:var(--text-darkest);font-family:var(--family-body)}.tk-gantt-chart-panel{flex-shrink:0;border-right:1px solid var(--border-light);overflow-y:auto;overflow-x:auto}.tk-gantt-chart-panel::-webkit-scrollbar{width:4px;height:4px}.tk-gantt-chart-panel::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-gantt-chart-panel::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-gantt-chart-panel::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-gantt-chart-panel::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-gantt-chart-panel-header{display:flex;align-items:center;position:sticky;top:0;z-index:2}.tk-gantt-chart-panel-header-cell{display:flex;align-items:center;padding:0 8px;font-weight:600;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;height:100%;border-right:1px solid var(--border-light);background:var(--background-lightest);border-bottom:1px solid var(--border-light)}.tk-gantt-chart-panel-header-cell:last-child{border-right:none}.tk-gantt-chart-panel-body{position:relative}.tk-gantt-chart-panel-row{display:flex;align-items:center}.tk-gantt-chart-panel-row:last-child .tk-gantt-chart-panel-cell{border-bottom:none}.tk-gantt-chart-panel-row:hover{background:var(--background-lightest)}.tk-gantt-chart-panel-cell{display:flex;align-items:center;gap:4px;padding:0 8px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;height:100%;border-right:1px solid var(--border-light);border-bottom:1px solid var(--border-light)}.tk-gantt-chart-panel-cell:last-child{border-right:none}.tk-gantt-chart-expand-btn{display:inline-flex;align-items:center;justify-content:center;background:none;border:none;cursor:pointer;padding:0;margin-right:4px;transition:transform 0.2s ease;color:var(--icon-base)}.tk-gantt-chart-expand-btn:hover{color:var(--icon-darkest)}.tk-gantt-chart-expand-btn-open{transform:rotate(90deg)}.tk-gantt-chart-timeline{flex:1;overflow:auto;position:relative}.tk-gantt-chart-timeline::-webkit-scrollbar{width:4px;height:4px}.tk-gantt-chart-timeline::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-gantt-chart-timeline::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-gantt-chart-timeline::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-gantt-chart-timeline::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-gantt-chart-timeline-headers{position:sticky;top:0;z-index:2;background:var(--background-lightest);border-bottom:1px solid var(--border-light)}.tk-gantt-chart-timeline-header{display:flex}.tk-gantt-chart-timeline-header-primary{border-bottom:1px solid var(--border-light)}.tk-gantt-chart-timeline-header-cell{display:flex;align-items:center;justify-content:center;border-right:1px solid var(--border-light);font-weight:500;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;font-size:var(--desktop-body-xs-size, 11px)}.tk-gantt-chart-timeline-header-cell:last-child{border-right:none}.tk-gantt-chart-timeline-header-cell-weekend{background:var(--neutral-100);color:var(--text-light)}.tk-gantt-chart-timeline-header-cell-holiday{background:var(--primary-100, #fff0f0);color:var(--primary-base)}.tk-gantt-chart-timeline-body{position:relative}.tk-gantt-chart-timeline-row{position:relative;display:flex;align-items:center;border-bottom:1px solid var(--border-light)}.tk-gantt-chart-timeline-row .custom-task-bar{width:100%;height:100%}.tk-gantt-chart-timeline-row:last-child{border-bottom:none}.tk-gantt-chart-column-highlight{position:absolute;top:0;pointer-events:none}.tk-gantt-chart-weekend{background:rgba(0, 0, 0, 0.03)}.tk-gantt-chart-holiday{background:rgba(201, 0, 25, 0.06)}.tk-gantt-chart-task-bar{position:absolute;top:50%;transform:translateY(-50%);border-radius:var(--radius-s-base, 4px);background:var(--primary-base);cursor:pointer;display:flex;align-items:center;overflow:hidden;transition:opacity 0.15s ease}.tk-gantt-chart-task-bar:hover{opacity:0.85}.tk-gantt-chart-task-bar-custom{background:none}.tk-gantt-chart-task-bar-segment{border-radius:var(--radius-s-base, 4px)}.tk-gantt-chart-task-bar-progress{position:absolute;left:0;top:0;height:100%;background:rgba(255, 255, 255, 0.35);border-radius:var(--radius-s-base, 4px) 0 0 var(--radius-s-base, 4px);pointer-events:none}.tk-gantt-chart-task-bar-label{position:relative;z-index:1;padding:0 6px;color:var(--static-white);font-size:var(--desktop-body-xs-size, 11px);white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.tk-gantt-chart-indicator{position:absolute;top:0;width:0;border-left-width:2px;z-index:1;pointer-events:none}.tk-gantt-chart-indicator-label{position:absolute;top:-18px;left:4px;font-size:var(--desktop-body-xs-size, 10px);white-space:nowrap;color:inherit;font-weight:500}.tk-gantt-chart-tooltip{position:absolute;z-index:10;padding:8px 12px;border-radius:var(--radius-s-base, 4px);background:var(--frames-darkest);color:var(--static-white);font-size:var(--desktop-body-xs-size, 11px);pointer-events:none;box-shadow:0 2px 8px rgba(0, 0, 0, 0.15);min-width:140px}.tk-gantt-chart-tooltip-title{font-weight:600;margin-bottom:4px}.tk-gantt-chart-tooltip-row{line-height:1.6}.tk-gantt-chart-tooltip-label{font-weight:500;margin-right:4px}.tk-gantt-chart-empty{display:flex;align-items:center;justify-content:center;width:100%;min-height:200px;color:var(--text-light)}';
      }
    },
    [
      257,
      'tk-gantt-chart',
      {
        tasks: [16],
        columns: [16],
        viewType: [1, 'view-type'],
        secondaryHeaderMode: [1, 'secondary-header-mode'],
        weekStartDay: [2, 'week-start-day'],
        indicators: [16],
        showTodayIndicator: [4, 'show-today-indicator'],
        highlightWeekends: [4, 'highlight-weekends'],
        holidays: [16],
        tooltipHtml: [16],
        taskBarHtml: [16],
        rowHeight: [2, 'row-height'],
        panelWidth: [2, 'panel-width'],
        hidePanel: [4, 'hide-panel'],
        containerStyle: [16],
        locale: [1],
        dataTestid: [513, 'data-testid'],
        expandedIds: [32],
        computedViewType: [32],
      },
      void 0,
      { tasks: [{ tasksChanged: 0 }], viewType: [{ viewTypeChanged: 0 }] },
    ],
  ),
  M = S,
  T = function () {
    'undefined' != typeof customElements &&
      ['tk-gantt-chart', 'tk-icon', 'tk-tooltip'].forEach(t => {
        switch (t) {
          case 'tk-gantt-chart':
            customElements.get(r(t)) || customElements.define(r(t), S);
            break;
          case 'tk-icon':
            customElements.get(r(t)) || l();
            break;
          case 'tk-tooltip':
            customElements.get(r(t)) || d();
        }
      });
  };
export { M as TkGanttChart, T as defineCustomElement };
