// @takeoff-ui/core@0.11.9/components/tk-org-chart.js
import { p as t, H as n, c as i, h as r, t as e } from './p-BfLQgzDX.js';
import { m as o } from './p-DXMmAH3A.js';
var s = 'http://www.w3.org/1999/xhtml',
  u = { svg: 'http://www.w3.org/2000/svg', xhtml: s, xlink: 'http://www.w3.org/1999/xlink', xml: 'http://www.w3.org/XML/1998/namespace', xmlns: 'http://www.w3.org/2000/xmlns/' };
function a(t) {
  var n = (t += ''),
    i = n.indexOf(':');
  return (i >= 0 && 'xmlns' !== (n = t.slice(0, i)) && (t = t.slice(i + 1)), u.hasOwnProperty(n) ? { space: u[n], local: t } : t);
}
function h(t) {
  return function () {
    var n = this.ownerDocument,
      i = this.namespaceURI;
    return i === s && n.documentElement.namespaceURI === s ? n.createElement(t) : n.createElementNS(i, t);
  };
}
function l(t) {
  return function () {
    return this.ownerDocument.createElementNS(t.space, t.local);
  };
}
function c(t) {
  var n = a(t);
  return (n.local ? l : h)(n);
}
function f() {}
function d(t) {
  return null == t
    ? f
    : function () {
        return this.querySelector(t);
      };
}
function p() {
  return [];
}
function v(t) {
  return null == t
    ? p
    : function () {
        return this.querySelectorAll(t);
      };
}
function g(t) {
  return function () {
    return this.matches(t);
  };
}
function m(t) {
  return function (n) {
    return n.matches(t);
  };
}
var w = Array.prototype.find;
function y() {
  return this.firstElementChild;
}
var x = Array.prototype.filter;
function C() {
  return Array.from(this.children);
}
function $(t) {
  return Array(t.length);
}
function b(t, n) {
  ((this.ownerDocument = t.ownerDocument), (this.namespaceURI = t.namespaceURI), (this._next = null), (this._parent = t), (this.__data__ = n));
}
function k(t, n, i, r, e, o) {
  for (var s, u = 0, a = n.length, h = o.length; u < h; ++u) (s = n[u]) ? ((s.__data__ = o[u]), (r[u] = s)) : (i[u] = new b(t, o[u]));
  for (; u < a; ++u) (s = n[u]) && (e[u] = s);
}
function M(t, n, i, r, e, o, s) {
  var u,
    a,
    h,
    l = new Map(),
    c = n.length,
    f = o.length,
    d = Array(c);
  for (u = 0; u < c; ++u) (a = n[u]) && ((d[u] = h = s.call(a, a.__data__, u, n) + ''), l.has(h) ? (e[u] = a) : l.set(h, a));
  for (u = 0; u < f; ++u) ((h = s.call(t, o[u], u, o) + ''), (a = l.get(h)) ? ((r[u] = a), (a.__data__ = o[u]), l.delete(h)) : (i[u] = new b(t, o[u])));
  for (u = 0; u < c; ++u) (a = n[u]) && l.get(d[u]) === a && (e[u] = a);
}
function E(t) {
  return t.__data__;
}
function L(t) {
  return 'object' == typeof t && 'length' in t ? t : Array.from(t);
}
function z(t, n) {
  return t < n ? -1 : t > n ? 1 : t >= n ? 0 : NaN;
}
function A(t) {
  return function () {
    this.removeAttribute(t);
  };
}
function N(t) {
  return function () {
    this.removeAttributeNS(t.space, t.local);
  };
}
function X(t, n) {
  return function () {
    this.setAttribute(t, n);
  };
}
function T(t, n) {
  return function () {
    this.setAttributeNS(t.space, t.local, n);
  };
}
function B(t, n) {
  return function () {
    var i = n.apply(this, arguments);
    null == i ? this.removeAttribute(t) : this.setAttribute(t, i);
  };
}
function S(t, n) {
  return function () {
    var i = n.apply(this, arguments);
    null == i ? this.removeAttributeNS(t.space, t.local) : this.setAttributeNS(t.space, t.local, i);
  };
}
function Y(t) {
  return (t.ownerDocument && t.ownerDocument.defaultView) || (t.document && t) || t.defaultView;
}
function O(t) {
  return function () {
    this.style.removeProperty(t);
  };
}
function R(t, n, i) {
  return function () {
    this.style.setProperty(t, n, i);
  };
}
function H(t, n, i) {
  return function () {
    var r = n.apply(this, arguments);
    null == r ? this.style.removeProperty(t) : this.style.setProperty(t, r, i);
  };
}
function j(t, n) {
  return t.style.getPropertyValue(n) || Y(t).getComputedStyle(t, null).getPropertyValue(n);
}
function F(t) {
  return function () {
    delete this[t];
  };
}
function _(t, n) {
  return function () {
    this[t] = n;
  };
}
function D(t, n) {
  return function () {
    var i = n.apply(this, arguments);
    null == i ? delete this[t] : (this[t] = i);
  };
}
function Z(t) {
  return t.trim().split(/^|\s+/);
}
function G(t) {
  return t.classList || new I(t);
}
function I(t) {
  ((this._node = t), (this._names = Z(t.getAttribute('class') || '')));
}
function J(t, n) {
  for (var i = G(t), r = -1, e = n.length; ++r < e; ) i.add(n[r]);
}
function P(t, n) {
  for (var i = G(t), r = -1, e = n.length; ++r < e; ) i.remove(n[r]);
}
function U(t) {
  return function () {
    J(this, t);
  };
}
function V(t) {
  return function () {
    P(this, t);
  };
}
function q(t, n) {
  return function () {
    (n.apply(this, arguments) ? J : P)(this, t);
  };
}
function W() {
  this.textContent = '';
}
function Q(t) {
  return function () {
    this.textContent = t;
  };
}
function K(t) {
  return function () {
    var n = t.apply(this, arguments);
    this.textContent = null == n ? '' : n;
  };
}
function tt() {
  this.innerHTML = '';
}
function nt(t) {
  return function () {
    this.innerHTML = t;
  };
}
function it(t) {
  return function () {
    var n = t.apply(this, arguments);
    this.innerHTML = null == n ? '' : n;
  };
}
function rt() {
  this.nextSibling && this.parentNode.appendChild(this);
}
function et() {
  this.previousSibling && this.parentNode.insertBefore(this, this.parentNode.firstChild);
}
function ot() {
  return null;
}
function st() {
  var t = this.parentNode;
  t && t.removeChild(this);
}
function ut() {
  var t = this.cloneNode(!1),
    n = this.parentNode;
  return n ? n.insertBefore(t, this.nextSibling) : t;
}
function at() {
  var t = this.cloneNode(!0),
    n = this.parentNode;
  return n ? n.insertBefore(t, this.nextSibling) : t;
}
function ht(t) {
  return function () {
    var n = this.__on;
    if (n) {
      for (var i, r = 0, e = -1, o = n.length; r < o; ++r)
        ((i = n[r]), (t.type && i.type !== t.type) || i.name !== t.name ? (n[++e] = i) : this.removeEventListener(i.type, i.listener, i.options));
      ++e ? (n.length = e) : delete this.__on;
    }
  };
}
function lt(t, n, i) {
  return function () {
    var r,
      e = this.__on,
      o = (function (t) {
        return function (n) {
          t.call(this, n, this.__data__);
        };
      })(n);
    if (e)
      for (var s = 0, u = e.length; s < u; ++s)
        if ((r = e[s]).type === t.type && r.name === t.name)
          return (this.removeEventListener(r.type, r.listener, r.options), this.addEventListener(r.type, (r.listener = o), (r.options = i)), void (r.value = n));
    (this.addEventListener(t.type, o, i), (r = { type: t.type, name: t.name, value: n, listener: o, options: i }), e ? e.push(r) : (this.__on = [r]));
  };
}
function ct(t, n, i) {
  var r = Y(t),
    e = r.CustomEvent;
  ('function' == typeof e
    ? (e = new e(n, i))
    : ((e = r.document.createEvent('Event')), i ? (e.initEvent(n, i.bubbles, i.cancelable), (e.detail = i.detail)) : e.initEvent(n, !1, !1)),
    t.dispatchEvent(e));
}
function ft(t, n) {
  return function () {
    return ct(this, t, n);
  };
}
function dt(t, n) {
  return function () {
    return ct(this, t, n.apply(this, arguments));
  };
}
((b.prototype = {
  constructor: b,
  appendChild: function (t) {
    return this._parent.insertBefore(t, this._next);
  },
  insertBefore: function (t, n) {
    return this._parent.insertBefore(t, n);
  },
  querySelector: function (t) {
    return this._parent.querySelector(t);
  },
  querySelectorAll: function (t) {
    return this._parent.querySelectorAll(t);
  },
}),
  (I.prototype = {
    add: function (t) {
      this._names.indexOf(t) < 0 && (this._names.push(t), this._node.setAttribute('class', this._names.join(' ')));
    },
    remove: function (t) {
      var n = this._names.indexOf(t);
      n >= 0 && (this._names.splice(n, 1), this._node.setAttribute('class', this._names.join(' ')));
    },
    contains: function (t) {
      return this._names.indexOf(t) >= 0;
    },
  }));
var pt = [null];
function vt(t, n) {
  ((this._groups = t), (this._parents = n));
}
function gt() {
  return new vt([[document.documentElement]], pt);
}
function mt(t) {
  return 'string' == typeof t ? new vt([[document.querySelector(t)]], [document.documentElement]) : new vt([[t]], pt);
}
function wt(t, n) {
  if (
    ((t = (function (t) {
      let n;
      for (; (n = t.sourceEvent); ) t = n;
      return t;
    })(t)),
    void 0 === n && (n = t.currentTarget),
    n)
  ) {
    var i = n.ownerSVGElement || n;
    if (i.createSVGPoint) {
      var r = i.createSVGPoint();
      return ((r.x = t.clientX), (r.y = t.clientY), [(r = r.matrixTransform(n.getScreenCTM().inverse())).x, r.y]);
    }
    if (n.getBoundingClientRect) {
      var e = n.getBoundingClientRect();
      return [t.clientX - e.left - n.clientLeft, t.clientY - e.top - n.clientTop];
    }
  }
  return [t.pageX, t.pageY];
}
function yt(t) {
  var n = 0,
    i = t.children,
    r = i && i.length;
  if (r) for (; --r >= 0; ) n += i[r].value;
  else n = 1;
  t.value = n;
}
function xt(t, n) {
  t instanceof Map ? ((t = [void 0, t]), void 0 === n && (n = $t)) : void 0 === n && (n = Ct);
  for (var i, r, e, o, s, u = new Mt(t), a = [u]; (i = a.pop()); )
    if ((e = n(i.data)) && (s = (e = Array.from(e)).length))
      for (i.children = e, o = s - 1; o >= 0; --o) (a.push((r = e[o] = new Mt(e[o]))), (r.parent = i), (r.depth = i.depth + 1));
  return u.eachBefore(kt);
}
function Ct(t) {
  return t.children;
}
function $t(t) {
  return Array.isArray(t) ? t[1] : null;
}
function bt(t) {
  (void 0 !== t.data.value && (t.value = t.data.value), (t.data = t.data.data));
}
function kt(t) {
  var n = 0;
  do {
    t.height = n;
  } while ((t = t.parent) && t.height < ++n);
}
function Mt(t) {
  ((this.data = t), (this.depth = this.height = 0), (this.parent = null));
}
function Et(t) {
  return null == t
    ? null
    : (function (t) {
        if ('function' != typeof t) throw Error();
        return t;
      })(t);
}
((vt.prototype = gt.prototype =
  {
    constructor: vt,
    select: function (t) {
      'function' != typeof t && (t = d(t));
      for (var n = this._groups, i = n.length, r = Array(i), e = 0; e < i; ++e)
        for (var o, s, u = n[e], a = u.length, h = (r[e] = Array(a)), l = 0; l < a; ++l)
          (o = u[l]) && (s = t.call(o, o.__data__, l, u)) && ('__data__' in o && (s.__data__ = o.__data__), (h[l] = s));
      return new vt(r, this._parents);
    },
    selectAll: function (t) {
      t =
        'function' == typeof t
          ? (function (t) {
              return function () {
                return (function (t) {
                  return null == t ? [] : Array.isArray(t) ? t : Array.from(t);
                })(t.apply(this, arguments));
              };
            })(t)
          : v(t);
      for (var n = this._groups, i = n.length, r = [], e = [], o = 0; o < i; ++o)
        for (var s, u = n[o], a = u.length, h = 0; h < a; ++h) (s = u[h]) && (r.push(t.call(s, s.__data__, h, u)), e.push(s));
      return new vt(r, e);
    },
    selectChild: function (t) {
      return this.select(
        null == t
          ? y
          : (function (t) {
              return function () {
                return w.call(this.children, t);
              };
            })('function' == typeof t ? t : m(t)),
      );
    },
    selectChildren: function (t) {
      return this.selectAll(
        null == t
          ? C
          : (function (t) {
              return function () {
                return x.call(this.children, t);
              };
            })('function' == typeof t ? t : m(t)),
      );
    },
    filter: function (t) {
      'function' != typeof t && (t = g(t));
      for (var n = this._groups, i = n.length, r = Array(i), e = 0; e < i; ++e)
        for (var o, s = n[e], u = s.length, a = (r[e] = []), h = 0; h < u; ++h) (o = s[h]) && t.call(o, o.__data__, h, s) && a.push(o);
      return new vt(r, this._parents);
    },
    data: function (t, n) {
      if (!arguments.length) return Array.from(this, E);
      var i = n ? M : k,
        r = this._parents,
        e = this._groups;
      'function' != typeof t &&
        (t = (function (t) {
          return function () {
            return t;
          };
        })(t));
      for (var o = e.length, s = Array(o), u = Array(o), a = Array(o), h = 0; h < o; ++h) {
        var l = r[h],
          c = e[h],
          f = c.length,
          d = L(t.call(l, l && l.__data__, h, r)),
          p = d.length,
          v = (u[h] = Array(p)),
          g = (s[h] = Array(p));
        i(l, c, v, g, (a[h] = Array(f)), d, n);
        for (var m, w, y = 0, x = 0; y < p; ++y)
          if ((m = v[y])) {
            for (y >= x && (x = y + 1); !(w = g[x]) && ++x < p; );
            m._next = w || null;
          }
      }
      return (((s = new vt(s, r))._enter = u), (s._exit = a), s);
    },
    enter: function () {
      return new vt(this._enter || this._groups.map($), this._parents);
    },
    exit: function () {
      return new vt(this._exit || this._groups.map($), this._parents);
    },
    join: function (t, n, i) {
      var r = this.enter(),
        e = this,
        o = this.exit();
      return (
        'function' == typeof t ? (r = t(r)) && (r = r.selection()) : (r = r.append(t + '')),
        null != n && (e = n(e)) && (e = e.selection()),
        null == i ? o.remove() : i(o),
        r && e ? r.merge(e).order() : e
      );
    },
    merge: function (t) {
      for (var n = t.selection ? t.selection() : t, i = this._groups, r = n._groups, e = i.length, o = Math.min(e, r.length), s = Array(e), u = 0; u < o; ++u)
        for (var a, h = i[u], l = r[u], c = h.length, f = (s[u] = Array(c)), d = 0; d < c; ++d) (a = h[d] || l[d]) && (f[d] = a);
      for (; u < e; ++u) s[u] = i[u];
      return new vt(s, this._parents);
    },
    selection: function () {
      return this;
    },
    order: function () {
      for (var t = this._groups, n = -1, i = t.length; ++n < i; )
        for (var r, e = t[n], o = e.length - 1, s = e[o]; --o >= 0; ) (r = e[o]) && (s && 4 ^ r.compareDocumentPosition(s) && s.parentNode.insertBefore(r, s), (s = r));
      return this;
    },
    sort: function (t) {
      function n(n, i) {
        return n && i ? t(n.__data__, i.__data__) : !n - !i;
      }
      t || (t = z);
      for (var i = this._groups, r = i.length, e = Array(r), o = 0; o < r; ++o) {
        for (var s, u = i[o], a = u.length, h = (e[o] = Array(a)), l = 0; l < a; ++l) (s = u[l]) && (h[l] = s);
        h.sort(n);
      }
      return new vt(e, this._parents).order();
    },
    call: function () {
      var t = arguments[0];
      return ((arguments[0] = this), t.apply(null, arguments), this);
    },
    nodes: function () {
      return Array.from(this);
    },
    node: function () {
      for (var t = this._groups, n = 0, i = t.length; n < i; ++n)
        for (var r = t[n], e = 0, o = r.length; e < o; ++e) {
          var s = r[e];
          if (s) return s;
        }
      return null;
    },
    size: function () {
      let t = 0;
      for (const n of this) ++t;
      return t;
    },
    empty: function () {
      return !this.node();
    },
    each: function (t) {
      for (var n = this._groups, i = 0, r = n.length; i < r; ++i) for (var e, o = n[i], s = 0, u = o.length; s < u; ++s) (e = o[s]) && t.call(e, e.__data__, s, o);
      return this;
    },
    attr: function (t, n) {
      var i = a(t);
      if (arguments.length < 2) {
        var r = this.node();
        return i.local ? r.getAttributeNS(i.space, i.local) : r.getAttribute(i);
      }
      return this.each((null == n ? (i.local ? N : A) : 'function' == typeof n ? (i.local ? S : B) : i.local ? T : X)(i, n));
    },
    style: function (t, n, i) {
      return arguments.length > 1 ? this.each((null == n ? O : 'function' == typeof n ? H : R)(t, n, null == i ? '' : i)) : j(this.node(), t);
    },
    property: function (t, n) {
      return arguments.length > 1 ? this.each((null == n ? F : 'function' == typeof n ? D : _)(t, n)) : this.node()[t];
    },
    classed: function (t, n) {
      var i = Z(t + '');
      if (arguments.length < 2) {
        for (var r = G(this.node()), e = -1, o = i.length; ++e < o; ) if (!r.contains(i[e])) return !1;
        return !0;
      }
      return this.each(('function' == typeof n ? q : n ? U : V)(i, n));
    },
    text: function (t) {
      return arguments.length ? this.each(null == t ? W : ('function' == typeof t ? K : Q)(t)) : this.node().textContent;
    },
    html: function (t) {
      return arguments.length ? this.each(null == t ? tt : ('function' == typeof t ? it : nt)(t)) : this.node().innerHTML;
    },
    raise: function () {
      return this.each(rt);
    },
    lower: function () {
      return this.each(et);
    },
    append: function (t) {
      var n = 'function' == typeof t ? t : c(t);
      return this.select(function () {
        return this.appendChild(n.apply(this, arguments));
      });
    },
    insert: function (t, n) {
      var i = 'function' == typeof t ? t : c(t),
        r = null == n ? ot : 'function' == typeof n ? n : d(n);
      return this.select(function () {
        return this.insertBefore(i.apply(this, arguments), r.apply(this, arguments) || null);
      });
    },
    remove: function () {
      return this.each(st);
    },
    clone: function (t) {
      return this.select(t ? at : ut);
    },
    datum: function (t) {
      return arguments.length ? this.property('__data__', t) : this.node().__data__;
    },
    on: function (t, n, i) {
      var r,
        e,
        o = (function (t) {
          return t
            .trim()
            .split(/^|\s+/)
            .map(function (t) {
              var n = '',
                i = t.indexOf('.');
              return (i >= 0 && ((n = t.slice(i + 1)), (t = t.slice(0, i))), { type: t, name: n });
            });
        })(t + ''),
        s = o.length;
      if (!(arguments.length < 2)) {
        for (u = n ? lt : ht, r = 0; r < s; ++r) this.each(u(o[r], n, i));
        return this;
      }
      var u = this.node().__on;
      if (u) for (var a, h = 0, l = u.length; h < l; ++h) for (r = 0, a = u[h]; r < s; ++r) if ((e = o[r]).type === a.type && e.name === a.name) return a.value;
    },
    dispatch: function (t, n) {
      return this.each(('function' == typeof n ? dt : ft)(t, n));
    },
    [Symbol.iterator]: function* () {
      for (var t = this._groups, n = 0, i = t.length; n < i; ++n) for (var r, e = t[n], o = 0, s = e.length; o < s; ++o) (r = e[o]) && (yield r);
    },
  }),
  (Mt.prototype = xt.prototype =
    {
      constructor: Mt,
      count: function () {
        return this.eachAfter(yt);
      },
      each: function (t, n) {
        let i = -1;
        for (const r of this) t.call(n, r, ++i, this);
        return this;
      },
      eachAfter: function (t, n) {
        for (var i, r, e, o = this, s = [o], u = [], a = -1; (o = s.pop()); ) if ((u.push(o), (i = o.children))) for (r = 0, e = i.length; r < e; ++r) s.push(i[r]);
        for (; (o = u.pop()); ) t.call(n, o, ++a, this);
        return this;
      },
      eachBefore: function (t, n) {
        for (var i, r, e = this, o = [e], s = -1; (e = o.pop()); ) if ((t.call(n, e, ++s, this), (i = e.children))) for (r = i.length - 1; r >= 0; --r) o.push(i[r]);
        return this;
      },
      find: function (t, n) {
        let i = -1;
        for (const r of this) if (t.call(n, r, ++i, this)) return r;
      },
      sum: function (t) {
        return this.eachAfter(function (n) {
          for (var i = +t(n.data) || 0, r = n.children, e = r && r.length; --e >= 0; ) i += r[e].value;
          n.value = i;
        });
      },
      sort: function (t) {
        return this.eachBefore(function (n) {
          n.children && n.children.sort(t);
        });
      },
      path: function (t) {
        for (
          var n = this,
            i = (function (t, n) {
              if (t === n) return t;
              var i = t.ancestors(),
                r = n.ancestors(),
                e = null;
              for (t = i.pop(), n = r.pop(); t === n; ) ((e = t), (t = i.pop()), (n = r.pop()));
              return e;
            })(n, t),
            r = [n];
          n !== i;
        )
          r.push((n = n.parent));
        for (var e = r.length; t !== i; ) (r.splice(e, 0, t), (t = t.parent));
        return r;
      },
      ancestors: function () {
        for (var t = this, n = [t]; (t = t.parent); ) n.push(t);
        return n;
      },
      descendants: function () {
        return Array.from(this);
      },
      leaves: function () {
        var t = [];
        return (
          this.eachBefore(function (n) {
            n.children || t.push(n);
          }),
          t
        );
      },
      links: function () {
        var t = this,
          n = [];
        return (
          t.each(function (i) {
            i !== t && n.push({ source: i.parent, target: i });
          }),
          n
        );
      },
      copy: function () {
        return xt(this).eachBefore(bt);
      },
      [Symbol.iterator]: function* () {
        var t,
          n,
          i,
          r,
          e = this,
          o = [e];
        do {
          for (t = o.reverse(), o = []; (e = t.pop()); ) if ((yield e, (n = e.children))) for (i = 0, r = n.length; i < r; ++i) o.push(n[i]);
        } while (o.length);
      },
    }));
var Lt = { depth: -1 },
  zt = {},
  At = {};
function Nt(t) {
  return t.id;
}
function Xt(t) {
  return t.parentId;
}
function Tt(t) {
  let n = t.length;
  if (n < 2) return '';
  for (; --n > 1 && !Bt(t, n); );
  return t.slice(0, n);
}
function Bt(t, n) {
  if ('/' === t[n]) {
    let i = 0;
    for (; n > 0 && '\\' === t[--n]; ) ++i;
    if (!(1 & i)) return !0;
  }
  return !1;
}
function St(t, n) {
  return t.parent === n.parent ? 1 : 2;
}
function Yt(t) {
  var n = t.children;
  return n ? n[0] : t.t;
}
function Ot(t) {
  var n = t.children;
  return n ? n[n.length - 1] : t.t;
}
function Rt(t, n, i) {
  var r = i / (n.i - t.i);
  ((n.c -= r), (n.s += i), (t.c += r), (n.z += i), (n.m += i));
}
function Ht(t, n, i) {
  return t.a.parent === n.parent ? t.a : i;
}
function jt(t, n) {
  ((this._ = t),
    (this.parent = null),
    (this.children = null),
    (this.A = null),
    (this.a = this),
    (this.z = 0),
    (this.m = 0),
    (this.c = 0),
    (this.s = 0),
    (this.t = null),
    (this.i = n));
}
jt.prototype = Object.create(Mt.prototype);
var Ft = { value: () => {} };
function _t() {
  for (var t, n = 0, i = arguments.length, r = {}; n < i; ++n) {
    if (!(t = arguments[n] + '') || t in r || /[\s.]/.test(t)) throw Error('illegal type: ' + t);
    r[t] = [];
  }
  return new Dt(r);
}
function Dt(t) {
  this._ = t;
}
function Zt(t, n) {
  for (var i, r = 0, e = t.length; r < e; ++r) if ((i = t[r]).name === n) return i.value;
}
function Gt(t, n, i) {
  for (var r = 0, e = t.length; r < e; ++r)
    if (t[r].name === n) {
      ((t[r] = Ft), (t = t.slice(0, r).concat(t.slice(r + 1))));
      break;
    }
  return (null != i && t.push({ name: n, value: i }), t);
}
Dt.prototype = _t.prototype = {
  constructor: Dt,
  on: function (t, n) {
    var i,
      r = this._,
      e = (function (t, n) {
        return t
          .trim()
          .split(/^|\s+/)
          .map(function (t) {
            var i = '',
              r = t.indexOf('.');
            if ((r >= 0 && ((i = t.slice(r + 1)), (t = t.slice(0, r))), t && !n.hasOwnProperty(t))) throw Error('unknown type: ' + t);
            return { type: t, name: i };
          });
      })(t + '', r),
      o = -1,
      s = e.length;
    if (!(arguments.length < 2)) {
      if (null != n && 'function' != typeof n) throw Error('invalid callback: ' + n);
      for (; ++o < s; )
        if ((i = (t = e[o]).type)) r[i] = Gt(r[i], t.name, n);
        else if (null == n) for (i in r) r[i] = Gt(r[i], t.name, null);
      return this;
    }
    for (; ++o < s; ) if ((i = (t = e[o]).type) && (i = Zt(r[i], t.name))) return i;
  },
  copy: function () {
    var t = {},
      n = this._;
    for (var i in n) t[i] = n[i].slice();
    return new Dt(t);
  },
  call: function (t, n) {
    if ((i = arguments.length - 2) > 0) for (var i, r, e = Array(i), o = 0; o < i; ++o) e[o] = arguments[o + 2];
    if (!this._.hasOwnProperty(t)) throw Error('unknown type: ' + t);
    for (o = 0, i = (r = this._[t]).length; o < i; ++o) r[o].value.apply(n, e);
  },
  apply: function (t, n, i) {
    if (!this._.hasOwnProperty(t)) throw Error('unknown type: ' + t);
    for (var r = this._[t], e = 0, o = r.length; e < o; ++e) r[e].value.apply(n, i);
  },
};
const It = { capture: !0, passive: !1 };
function Jt(t) {
  (t.preventDefault(), t.stopImmediatePropagation());
}
function Pt(t, n, i) {
  ((t.prototype = n.prototype = i), (i.constructor = t));
}
function Ut(t, n) {
  var i = Object.create(t.prototype);
  for (var r in n) i[r] = n[r];
  return i;
}
function Vt() {}
var qt = 0.7,
  Wt = 1 / qt,
  Qt = '\\s*([+-]?\\d+)\\s*',
  Kt = '\\s*([+-]?(?:\\d*\\.)?\\d+(?:[eE][+-]?\\d+)?)\\s*',
  tn = '\\s*([+-]?(?:\\d*\\.)?\\d+(?:[eE][+-]?\\d+)?)%\\s*',
  nn = /^#([0-9a-f]{3,8})$/,
  rn = RegExp(`^rgb\\(${Qt},${Qt},${Qt}\\)$`),
  en = RegExp(`^rgb\\(${tn},${tn},${tn}\\)$`),
  on = RegExp(`^rgba\\(${Qt},${Qt},${Qt},${Kt}\\)$`),
  sn = RegExp(`^rgba\\(${tn},${tn},${tn},${Kt}\\)$`),
  un = RegExp(`^hsl\\(${Kt},${tn},${tn}\\)$`),
  an = RegExp(`^hsla\\(${Kt},${tn},${tn},${Kt}\\)$`),
  hn = {
    aliceblue: 15792383,
    antiquewhite: 16444375,
    aqua: 65535,
    aquamarine: 8388564,
    azure: 15794175,
    beige: 16119260,
    bisque: 16770244,
    black: 0,
    blanchedalmond: 16772045,
    blue: 255,
    blueviolet: 9055202,
    brown: 10824234,
    burlywood: 14596231,
    cadetblue: 6266528,
    chartreuse: 8388352,
    chocolate: 13789470,
    coral: 16744272,
    cornflowerblue: 6591981,
    cornsilk: 16775388,
    crimson: 14423100,
    cyan: 65535,
    darkblue: 139,
    darkcyan: 35723,
    darkgoldenrod: 12092939,
    darkgray: 11119017,
    darkgreen: 25600,
    darkgrey: 11119017,
    darkkhaki: 12433259,
    darkmagenta: 9109643,
    darkolivegreen: 5597999,
    darkorange: 16747520,
    darkorchid: 10040012,
    darkred: 9109504,
    darksalmon: 15308410,
    darkseagreen: 9419919,
    darkslateblue: 4734347,
    darkslategray: 3100495,
    darkslategrey: 3100495,
    darkturquoise: 52945,
    darkviolet: 9699539,
    deeppink: 16716947,
    deepskyblue: 49151,
    dimgray: 6908265,
    dimgrey: 6908265,
    dodgerblue: 2003199,
    firebrick: 11674146,
    floralwhite: 16775920,
    forestgreen: 2263842,
    fuchsia: 16711935,
    gainsboro: 14474460,
    ghostwhite: 16316671,
    gold: 16766720,
    goldenrod: 14329120,
    gray: 8421504,
    green: 32768,
    greenyellow: 11403055,
    grey: 8421504,
    honeydew: 15794160,
    hotpink: 16738740,
    indianred: 13458524,
    indigo: 4915330,
    ivory: 16777200,
    khaki: 15787660,
    lavender: 15132410,
    lavenderblush: 16773365,
    lawngreen: 8190976,
    lemonchiffon: 16775885,
    lightblue: 11393254,
    lightcoral: 15761536,
    lightcyan: 14745599,
    lightgoldenrodyellow: 16448210,
    lightgray: 13882323,
    lightgreen: 9498256,
    lightgrey: 13882323,
    lightpink: 16758465,
    lightsalmon: 16752762,
    lightseagreen: 2142890,
    lightskyblue: 8900346,
    lightslategray: 7833753,
    lightslategrey: 7833753,
    lightsteelblue: 11584734,
    lightyellow: 16777184,
    lime: 65280,
    limegreen: 3329330,
    linen: 16445670,
    magenta: 16711935,
    maroon: 8388608,
    mediumaquamarine: 6737322,
    mediumblue: 205,
    mediumorchid: 12211667,
    mediumpurple: 9662683,
    mediumseagreen: 3978097,
    mediumslateblue: 8087790,
    mediumspringgreen: 64154,
    mediumturquoise: 4772300,
    mediumvioletred: 13047173,
    midnightblue: 1644912,
    mintcream: 16121850,
    mistyrose: 16770273,
    moccasin: 16770229,
    navajowhite: 16768685,
    navy: 128,
    oldlace: 16643558,
    olive: 8421376,
    olivedrab: 7048739,
    orange: 16753920,
    orangered: 16729344,
    orchid: 14315734,
    palegoldenrod: 15657130,
    palegreen: 10025880,
    paleturquoise: 11529966,
    palevioletred: 14381203,
    papayawhip: 16773077,
    peachpuff: 16767673,
    peru: 13468991,
    pink: 16761035,
    plum: 14524637,
    powderblue: 11591910,
    purple: 8388736,
    rebeccapurple: 6697881,
    red: 16711680,
    rosybrown: 12357519,
    royalblue: 4286945,
    saddlebrown: 9127187,
    salmon: 16416882,
    sandybrown: 16032864,
    seagreen: 3050327,
    seashell: 16774638,
    sienna: 10506797,
    silver: 12632256,
    skyblue: 8900331,
    slateblue: 6970061,
    slategray: 7372944,
    slategrey: 7372944,
    snow: 16775930,
    springgreen: 65407,
    steelblue: 4620980,
    tan: 13808780,
    teal: 32896,
    thistle: 14204888,
    tomato: 16737095,
    turquoise: 4251856,
    violet: 15631086,
    wheat: 16113331,
    white: 16777215,
    whitesmoke: 16119285,
    yellow: 16776960,
    yellowgreen: 10145074,
  };
function ln() {
  return this.rgb().formatHex();
}
function cn() {
  return this.rgb().formatRgb();
}
function fn(t) {
  var n, i;
  return (
    (t = (t + '').trim().toLowerCase()),
    (n = nn.exec(t))
      ? ((i = n[1].length),
        (n = parseInt(n[1], 16)),
        6 === i
          ? dn(n)
          : 3 === i
            ? new gn(((n >> 8) & 15) | ((n >> 4) & 240), ((n >> 4) & 15) | (240 & n), ((15 & n) << 4) | (15 & n), 1)
            : 8 === i
              ? pn((n >> 24) & 255, (n >> 16) & 255, (n >> 8) & 255, (255 & n) / 255)
              : 4 === i
                ? pn(((n >> 12) & 15) | ((n >> 8) & 240), ((n >> 8) & 15) | ((n >> 4) & 240), ((n >> 4) & 15) | (240 & n), (((15 & n) << 4) | (15 & n)) / 255)
                : null)
      : (n = rn.exec(t))
        ? new gn(n[1], n[2], n[3], 1)
        : (n = en.exec(t))
          ? new gn((255 * n[1]) / 100, (255 * n[2]) / 100, (255 * n[3]) / 100, 1)
          : (n = on.exec(t))
            ? pn(n[1], n[2], n[3], n[4])
            : (n = sn.exec(t))
              ? pn((255 * n[1]) / 100, (255 * n[2]) / 100, (255 * n[3]) / 100, n[4])
              : (n = un.exec(t))
                ? $n(n[1], n[2] / 100, n[3] / 100, 1)
                : (n = an.exec(t))
                  ? $n(n[1], n[2] / 100, n[3] / 100, n[4])
                  : hn.hasOwnProperty(t)
                    ? dn(hn[t])
                    : 'transparent' === t
                      ? new gn(NaN, NaN, NaN, 0)
                      : null
  );
}
function dn(t) {
  return new gn((t >> 16) & 255, (t >> 8) & 255, 255 & t, 1);
}
function pn(t, n, i, r) {
  return (r <= 0 && (t = n = i = NaN), new gn(t, n, i, r));
}
function vn(t, n, i, r) {
  return 1 === arguments.length
    ? (function (t) {
        return (t instanceof Vt || (t = fn(t)), t ? new gn((t = t.rgb()).r, t.g, t.b, t.opacity) : new gn());
      })(t)
    : new gn(t, n, i, null == r ? 1 : r);
}
function gn(t, n, i, r) {
  ((this.r = +t), (this.g = +n), (this.b = +i), (this.opacity = +r));
}
function mn() {
  return `#${Cn(this.r)}${Cn(this.g)}${Cn(this.b)}`;
}
function wn() {
  const t = yn(this.opacity);
  return `${1 === t ? 'rgb(' : 'rgba('}${xn(this.r)}, ${xn(this.g)}, ${xn(this.b)}${1 === t ? ')' : `, ${t})`}`;
}
function yn(t) {
  return isNaN(t) ? 1 : Math.max(0, Math.min(1, t));
}
function xn(t) {
  return Math.max(0, Math.min(255, Math.round(t) || 0));
}
function Cn(t) {
  return ((t = xn(t)) < 16 ? '0' : '') + t.toString(16);
}
function $n(t, n, i, r) {
  return (r <= 0 ? (t = n = i = NaN) : i <= 0 || i >= 1 ? (t = n = NaN) : n <= 0 && (t = NaN), new kn(t, n, i, r));
}
function bn(t) {
  if (t instanceof kn) return new kn(t.h, t.s, t.l, t.opacity);
  if ((t instanceof Vt || (t = fn(t)), !t)) return new kn();
  if (t instanceof kn) return t;
  var n = (t = t.rgb()).r / 255,
    i = t.g / 255,
    r = t.b / 255,
    e = Math.min(n, i, r),
    o = Math.max(n, i, r),
    s = NaN,
    u = o - e,
    a = (o + e) / 2;
  return (
    u ? ((s = n === o ? (i - r) / u + 6 * (i < r) : i === o ? (r - n) / u + 2 : (n - i) / u + 4), (u /= a < 0.5 ? o + e : 2 - o - e), (s *= 60)) : (u = a > 0 && a < 1 ? 0 : s),
    new kn(s, u, a, t.opacity)
  );
}
function kn(t, n, i, r) {
  ((this.h = +t), (this.s = +n), (this.l = +i), (this.opacity = +r));
}
function Mn(t) {
  return (t = (t || 0) % 360) < 0 ? t + 360 : t;
}
function En(t) {
  return Math.max(0, Math.min(1, t || 0));
}
function Ln(t, n, i) {
  return 255 * (t < 60 ? n + ((i - n) * t) / 60 : t < 180 ? i : t < 240 ? n + ((i - n) * (240 - t)) / 60 : n);
}
(Pt(Vt, fn, {
  copy(t) {
    return Object.assign(new this.constructor(), this, t);
  },
  displayable() {
    return this.rgb().displayable();
  },
  hex: ln,
  formatHex: ln,
  formatHex8: function () {
    return this.rgb().formatHex8();
  },
  formatHsl: function () {
    return bn(this).formatHsl();
  },
  formatRgb: cn,
  toString: cn,
}),
  Pt(
    gn,
    vn,
    Ut(Vt, {
      brighter(t) {
        return new gn(this.r * (t = null == t ? Wt : Math.pow(Wt, t)), this.g * t, this.b * t, this.opacity);
      },
      darker(t) {
        return new gn(this.r * (t = null == t ? qt : Math.pow(qt, t)), this.g * t, this.b * t, this.opacity);
      },
      rgb() {
        return this;
      },
      clamp() {
        return new gn(xn(this.r), xn(this.g), xn(this.b), yn(this.opacity));
      },
      displayable() {
        return -0.5 <= this.r && this.r < 255.5 && -0.5 <= this.g && this.g < 255.5 && -0.5 <= this.b && this.b < 255.5 && 0 <= this.opacity && this.opacity <= 1;
      },
      hex: mn,
      formatHex: mn,
      formatHex8: function () {
        return `#${Cn(this.r)}${Cn(this.g)}${Cn(this.b)}${Cn(255 * (isNaN(this.opacity) ? 1 : this.opacity))}`;
      },
      formatRgb: wn,
      toString: wn,
    }),
  ),
  Pt(
    kn,
    function (t, n, i, r) {
      return 1 === arguments.length ? bn(t) : new kn(t, n, i, null == r ? 1 : r);
    },
    Ut(Vt, {
      brighter(t) {
        return new kn(this.h, this.s, this.l * (t = null == t ? Wt : Math.pow(Wt, t)), this.opacity);
      },
      darker(t) {
        return new kn(this.h, this.s, this.l * (t = null == t ? qt : Math.pow(qt, t)), this.opacity);
      },
      rgb() {
        var t = (this.h % 360) + 360 * (this.h < 0),
          n = this.l,
          i = n + (n < 0.5 ? n : 1 - n) * (isNaN(t) || isNaN(this.s) ? 0 : this.s),
          r = 2 * n - i;
        return new gn(Ln(t >= 240 ? t - 240 : t + 120, r, i), Ln(t, r, i), Ln(t < 120 ? t + 240 : t - 120, r, i), this.opacity);
      },
      clamp() {
        return new kn(Mn(this.h), En(this.s), En(this.l), yn(this.opacity));
      },
      displayable() {
        return ((0 <= this.s && this.s <= 1) || isNaN(this.s)) && 0 <= this.l && this.l <= 1 && 0 <= this.opacity && this.opacity <= 1;
      },
      formatHsl() {
        const t = yn(this.opacity);
        return `${1 === t ? 'hsl(' : 'hsla('}${Mn(this.h)}, ${100 * En(this.s)}%, ${100 * En(this.l)}%${1 === t ? ')' : `, ${t})`}`;
      },
    }),
  ));
var zn = t => () => t;
function An(t, n) {
  var i = n - t;
  return i
    ? (function (t, n) {
        return function (i) {
          return t + i * n;
        };
      })(t, i)
    : zn(isNaN(t) ? n : t);
}
var Nn = (function t(n) {
  var i = (function (t) {
    return 1 == (t = +t)
      ? An
      : function (n, i) {
          return i - n
            ? (function (t, n, i) {
                return (
                  (t = Math.pow(t, i)),
                  (n = Math.pow(n, i) - t),
                  (i = 1 / i),
                  function (r) {
                    return Math.pow(t + r * n, i);
                  }
                );
              })(n, i, t)
            : zn(isNaN(n) ? i : n);
        };
  })(n);
  function r(t, n) {
    var r = i((t = vn(t)).r, (n = vn(n)).r),
      e = i(t.g, n.g),
      o = i(t.b, n.b),
      s = An(t.opacity, n.opacity);
    return function (n) {
      return ((t.r = r(n)), (t.g = e(n)), (t.b = o(n)), (t.opacity = s(n)), t + '');
    };
  }
  return ((r.gamma = t), r);
})(1);
function Xn(t, n) {
  return (
    (t = +t),
    (n = +n),
    function (i) {
      return t * (1 - i) + n * i;
    }
  );
}
var Tn = /[-+]?(?:\d+\.?\d*|\.?\d+)(?:[eE][-+]?\d+)?/g,
  Bn = RegExp(Tn.source, 'g');
function Sn(t, n) {
  var i,
    r,
    e,
    o = (Tn.lastIndex = Bn.lastIndex = 0),
    s = -1,
    u = [],
    a = [];
  for (t += '', n += ''; (i = Tn.exec(t)) && (r = Bn.exec(n)); )
    ((e = r.index) > o && ((e = n.slice(o, e)), u[s] ? (u[s] += e) : (u[++s] = e)),
      (i = i[0]) === (r = r[0]) ? (u[s] ? (u[s] += r) : (u[++s] = r)) : ((u[++s] = null), a.push({ i: s, x: Xn(i, r) })),
      (o = Bn.lastIndex));
  return (
    o < n.length && ((e = n.slice(o)), u[s] ? (u[s] += e) : (u[++s] = e)),
    u.length < 2
      ? a[0]
        ? (function (t) {
            return function (n) {
              return t(n) + '';
            };
          })(a[0].x)
        : (function (t) {
            return function () {
              return t;
            };
          })(n)
      : ((n = a.length),
        function (t) {
          for (var i, r = 0; r < n; ++r) u[(i = a[r]).i] = i.x(t);
          return u.join('');
        })
  );
}
var Yn,
  On = 180 / Math.PI,
  Rn = { translateX: 0, translateY: 0, rotate: 0, skewX: 0, scaleX: 1, scaleY: 1 };
function Hn(t, n, i, r, e, o) {
  var s, u, a;
  return (
    (s = Math.sqrt(t * t + n * n)) && ((t /= s), (n /= s)),
    (a = t * i + n * r) && ((i -= t * a), (r -= n * a)),
    (u = Math.sqrt(i * i + r * r)) && ((i /= u), (r /= u), (a /= u)),
    t * r < n * i && ((t = -t), (n = -n), (a = -a), (s = -s)),
    { translateX: e, translateY: o, rotate: Math.atan2(n, t) * On, skewX: Math.atan(a) * On, scaleX: s, scaleY: u }
  );
}
function jn(t, n, i, r) {
  function e(t) {
    return t.length ? t.pop() + ' ' : '';
  }
  return function (o, s) {
    var u = [],
      a = [];
    return (
      (o = t(o)),
      (s = t(s)),
      (function (t, r, e, o, s, u) {
        if (t !== e || r !== o) {
          var a = s.push('translate(', null, n, null, i);
          u.push({ i: a - 4, x: Xn(t, e) }, { i: a - 2, x: Xn(r, o) });
        } else (e || o) && s.push('translate(' + e + n + o + i);
      })(o.translateX, o.translateY, s.translateX, s.translateY, u, a),
      (function (t, n, i, o) {
        t !== n ? (t - n > 180 ? (n += 360) : n - t > 180 && (t += 360), o.push({ i: i.push(e(i) + 'rotate(', null, r) - 2, x: Xn(t, n) })) : n && i.push(e(i) + 'rotate(' + n + r);
      })(o.rotate, s.rotate, u, a),
      (function (t, n, i, o) {
        t !== n ? o.push({ i: i.push(e(i) + 'skewX(', null, r) - 2, x: Xn(t, n) }) : n && i.push(e(i) + 'skewX(' + n + r);
      })(o.skewX, s.skewX, u, a),
      (function (t, n, i, r, o, s) {
        if (t !== i || n !== r) {
          var u = o.push(e(o) + 'scale(', null, ',', null, ')');
          s.push({ i: u - 4, x: Xn(t, i) }, { i: u - 2, x: Xn(n, r) });
        } else (1 === i && 1 === r) || o.push(e(o) + 'scale(' + i + ',' + r + ')');
      })(o.scaleX, o.scaleY, s.scaleX, s.scaleY, u, a),
      (o = s = null),
      function (t) {
        for (var n, i = -1, r = a.length; ++i < r; ) u[(n = a[i]).i] = n.x(t);
        return u.join('');
      }
    );
  };
}
var Fn = jn(
    function (t) {
      const n = new ('function' == typeof DOMMatrix ? DOMMatrix : WebKitCSSMatrix)(t + '');
      return n.isIdentity ? Rn : Hn(n.a, n.b, n.c, n.d, n.e, n.f);
    },
    'px, ',
    'px)',
    'deg)',
  ),
  _n = jn(
    function (t) {
      return null == t
        ? Rn
        : (Yn || (Yn = document.createElementNS('http://www.w3.org/2000/svg', 'g')),
          Yn.setAttribute('transform', t),
          (t = Yn.transform.baseVal.consolidate()) ? Hn((t = t.matrix).a, t.b, t.c, t.d, t.e, t.f) : Rn);
    },
    ', ',
    ')',
    ')',
  );
function Dn(t) {
  return ((t = Math.exp(t)) + 1 / t) / 2;
}
var Zn,
  Gn,
  In = (function t(n, i, r) {
    function e(t, e) {
      var o,
        s,
        u = t[0],
        a = t[1],
        h = t[2],
        l = e[2],
        c = e[0] - u,
        f = e[1] - a,
        d = c * c + f * f;
      if (d < 1e-12)
        ((s = Math.log(l / h) / n),
          (o = function (t) {
            return [u + t * c, a + t * f, h * Math.exp(n * t * s)];
          }));
      else {
        var p = Math.sqrt(d),
          v = (l * l - h * h + r * d) / (2 * h * i * p),
          g = (l * l - h * h - r * d) / (2 * l * i * p),
          m = Math.log(Math.sqrt(v * v + 1) - v);
        ((s = (Math.log(Math.sqrt(g * g + 1) - g) - m) / n),
          (o = function (t) {
            var r = t * s,
              e = Dn(m),
              o =
                (h / (i * p)) *
                (e *
                  (function (t) {
                    return ((t = Math.exp(2 * t)) - 1) / (t + 1);
                  })(n * r + m) -
                  (function (t) {
                    return ((t = Math.exp(t)) - 1 / t) / 2;
                  })(m));
            return [u + o * c, a + o * f, (h * e) / Dn(n * r + m)];
          }));
      }
      return ((o.duration = (1e3 * s * n) / Math.SQRT2), o);
    }
    return (
      (e.rho = function (n) {
        var i = Math.max(0.001, +n),
          r = i * i;
        return t(i, r, r * r);
      }),
      e
    );
  })(Math.SQRT2, 2, 4),
  Jn = 0,
  Pn = 0,
  Un = 0,
  Vn = 0,
  qn = 0,
  Wn = 0,
  Qn = 'object' == typeof performance && performance.now ? performance : Date,
  Kn =
    'object' == typeof window && window.requestAnimationFrame
      ? window.requestAnimationFrame.bind(window)
      : function (t) {
          setTimeout(t, 17);
        };
function ti() {
  return qn || (Kn(ni), (qn = Qn.now() + Wn));
}
function ni() {
  qn = 0;
}
function ii() {
  this._call = this._time = this._next = null;
}
function ri(t, n, i) {
  var r = new ii();
  return (r.restart(t, n, i), r);
}
function ei() {
  ((qn = (Vn = Qn.now()) + Wn), (Jn = Pn = 0));
  try {
    !(function () {
      (ti(), ++Jn);
      for (var t, n = Zn; n; ) ((t = qn - n._time) >= 0 && n._call.call(void 0, t), (n = n._next));
      --Jn;
    })();
  } finally {
    ((Jn = 0),
      (function () {
        for (var t, n, i = Zn, r = 1 / 0; i; )
          i._call ? (r > i._time && (r = i._time), (t = i), (i = i._next)) : ((n = i._next), (i._next = null), (i = t ? (t._next = n) : (Zn = n)));
        ((Gn = t), si(r));
      })(),
      (qn = 0));
  }
}
function oi() {
  var t = Qn.now(),
    n = t - Vn;
  n > 1e3 && ((Wn -= n), (Vn = t));
}
function si(t) {
  Jn ||
    (Pn && (Pn = clearTimeout(Pn)),
    t - qn > 24
      ? (t < 1 / 0 && (Pn = setTimeout(ei, t - Qn.now() - Wn)), Un && (Un = clearInterval(Un)))
      : (Un || ((Vn = Qn.now()), (Un = setInterval(oi, 1e3))), (Jn = 1), Kn(ei)));
}
function ui(t, n, i) {
  var r = new ii();
  return (
    r.restart(
      i => {
        (r.stop(), t(i + n));
      },
      (n = null == n ? 0 : +n),
      i,
    ),
    r
  );
}
ii.prototype = ri.prototype = {
  constructor: ii,
  restart: function (t, n, i) {
    if ('function' != typeof t) throw new TypeError('callback is not a function');
    ((i = (null == i ? ti() : +i) + (null == n ? 0 : +n)),
      this._next || Gn === this || (Gn ? (Gn._next = this) : (Zn = this), (Gn = this)),
      (this._call = t),
      (this._time = i),
      si());
  },
  stop: function () {
    this._call && ((this._call = null), (this._time = 1 / 0), si());
  },
};
var ai = _t('start', 'end', 'cancel', 'interrupt'),
  hi = [];
function li(t, n, i, r, e, o) {
  var s = t.__transition;
  if (s) {
    if (i in s) return;
  } else t.__transition = {};
  !(function (t, n, i) {
    var r,
      e = t.__transition;
    function o(a) {
      var h, l, c, f;
      if (1 !== i.state) return u();
      for (h in e)
        if ((f = e[h]).name === i.name) {
          if (3 === f.state) return ui(o);
          4 === f.state
            ? ((f.state = 6), f.timer.stop(), f.on.call('interrupt', t, t.__data__, f.index, f.group), delete e[h])
            : +h < n && ((f.state = 6), f.timer.stop(), f.on.call('cancel', t, t.__data__, f.index, f.group), delete e[h]);
        }
      if (
        (ui(function () {
          3 === i.state && ((i.state = 4), i.timer.restart(s, i.delay, i.time), s(a));
        }),
        (i.state = 2),
        i.on.call('start', t, t.__data__, i.index, i.group),
        2 === i.state)
      ) {
        for (i.state = 3, r = Array((c = i.tween.length)), h = 0, l = -1; h < c; ++h) (f = i.tween[h].value.call(t, t.__data__, i.index, i.group)) && (r[++l] = f);
        r.length = l + 1;
      }
    }
    function s(n) {
      for (var e = n < i.duration ? i.ease.call(null, n / i.duration) : (i.timer.restart(u), (i.state = 5), 1), o = -1, s = r.length; ++o < s; ) r[o].call(t, e);
      5 === i.state && (i.on.call('end', t, t.__data__, i.index, i.group), u());
    }
    function u() {
      for (var r in ((i.state = 6), i.timer.stop(), delete e[n], e)) return;
      delete t.__transition;
    }
    ((e[n] = i),
      (i.timer = ri(
        function (t) {
          ((i.state = 1), i.timer.restart(o, i.delay, i.time), i.delay <= t && o(t - i.delay));
        },
        0,
        i.time,
      )));
  })(t, i, { name: n, index: r, group: e, on: ai, tween: hi, time: o.time, delay: o.delay, duration: o.duration, ease: o.ease, timer: null, state: 0 });
}
function ci(t, n) {
  var i = di(t, n);
  if (i.state > 0) throw Error('too late; already scheduled');
  return i;
}
function fi(t, n) {
  var i = di(t, n);
  if (i.state > 3) throw Error('too late; already running');
  return i;
}
function di(t, n) {
  var i = t.__transition;
  if (!i || !(i = i[n])) throw Error('transition not found');
  return i;
}
function pi(t, n) {
  var i,
    r,
    e,
    o = t.__transition,
    s = !0;
  if (o) {
    for (e in ((n = null == n ? null : n + ''), o))
      (i = o[e]).name === n
        ? ((r = i.state > 2 && i.state < 5), (i.state = 6), i.timer.stop(), i.on.call(r ? 'interrupt' : 'cancel', t, t.__data__, i.index, i.group), delete o[e])
        : (s = !1);
    s && delete t.__transition;
  }
}
function vi(t, n) {
  var i, r;
  return function () {
    var e = fi(this, t),
      o = e.tween;
    if (o !== i)
      for (var s = 0, u = (r = i = o).length; s < u; ++s)
        if (r[s].name === n) {
          (r = r.slice()).splice(s, 1);
          break;
        }
    e.tween = r;
  };
}
function gi(t, n, i) {
  var r, e;
  if ('function' != typeof i) throw Error();
  return function () {
    var o = fi(this, t),
      s = o.tween;
    if (s !== r) {
      e = (r = s).slice();
      for (var u = { name: n, value: i }, a = 0, h = e.length; a < h; ++a)
        if (e[a].name === n) {
          e[a] = u;
          break;
        }
      a === h && e.push(u);
    }
    o.tween = e;
  };
}
function mi(t, n, i) {
  var r = t._id;
  return (
    t.each(function () {
      var t = fi(this, r);
      (t.value || (t.value = {}))[n] = i.apply(this, arguments);
    }),
    function (t) {
      return di(t, r).value[n];
    }
  );
}
function wi(t, n) {
  var i;
  return ('number' == typeof n ? Xn : n instanceof fn ? Nn : (i = fn(n)) ? ((n = i), Nn) : Sn)(t, n);
}
function yi(t) {
  return function () {
    this.removeAttribute(t);
  };
}
function xi(t) {
  return function () {
    this.removeAttributeNS(t.space, t.local);
  };
}
function Ci(t, n, i) {
  var r,
    e,
    o = i + '';
  return function () {
    var s = this.getAttribute(t);
    return s === o ? null : s === r ? e : (e = n((r = s), i));
  };
}
function $i(t, n, i) {
  var r,
    e,
    o = i + '';
  return function () {
    var s = this.getAttributeNS(t.space, t.local);
    return s === o ? null : s === r ? e : (e = n((r = s), i));
  };
}
function bi(t, n, i) {
  var r, e, o;
  return function () {
    var s,
      u,
      a = i(this);
    if (null != a) return (s = this.getAttribute(t)) === (u = a + '') ? null : s === r && u === e ? o : ((e = u), (o = n((r = s), a)));
    this.removeAttribute(t);
  };
}
function ki(t, n, i) {
  var r, e, o;
  return function () {
    var s,
      u,
      a = i(this);
    if (null != a) return (s = this.getAttributeNS(t.space, t.local)) === (u = a + '') ? null : s === r && u === e ? o : ((e = u), (o = n((r = s), a)));
    this.removeAttributeNS(t.space, t.local);
  };
}
function Mi(t, n) {
  var i, r;
  function e() {
    var e = n.apply(this, arguments);
    return (
      e !== r &&
        (i =
          (r = e) &&
          (function (t, n) {
            return function (i) {
              this.setAttributeNS(t.space, t.local, n.call(this, i));
            };
          })(t, e)),
      i
    );
  }
  return ((e._value = n), e);
}
function Ei(t, n) {
  var i, r;
  function e() {
    var e = n.apply(this, arguments);
    return (
      e !== r &&
        (i =
          (r = e) &&
          (function (t, n) {
            return function (i) {
              this.setAttribute(t, n.call(this, i));
            };
          })(t, e)),
      i
    );
  }
  return ((e._value = n), e);
}
function Li(t, n) {
  return function () {
    ci(this, t).delay = +n.apply(this, arguments);
  };
}
function zi(t, n) {
  return (
    (n = +n),
    function () {
      ci(this, t).delay = n;
    }
  );
}
function Ai(t, n) {
  return function () {
    fi(this, t).duration = +n.apply(this, arguments);
  };
}
function Ni(t, n) {
  return (
    (n = +n),
    function () {
      fi(this, t).duration = n;
    }
  );
}
var Xi = gt.prototype.constructor;
function Ti(t) {
  return function () {
    this.style.removeProperty(t);
  };
}
var Bi = 0;
function Si(t, n, i, r) {
  ((this._groups = t), (this._parents = n), (this._name = i), (this._id = r));
}
function Yi() {
  return ++Bi;
}
var Oi = gt.prototype;
Si.prototype = {
  constructor: Si,
  select: function (t) {
    var n = this._name,
      i = this._id;
    'function' != typeof t && (t = d(t));
    for (var r = this._groups, e = r.length, o = Array(e), s = 0; s < e; ++s)
      for (var u, a, h = r[s], l = h.length, c = (o[s] = Array(l)), f = 0; f < l; ++f)
        (u = h[f]) && (a = t.call(u, u.__data__, f, h)) && ('__data__' in u && (a.__data__ = u.__data__), (c[f] = a), li(c[f], n, i, f, c, di(u, i)));
    return new Si(o, this._parents, n, i);
  },
  selectAll: function (t) {
    var n = this._name,
      i = this._id;
    'function' != typeof t && (t = v(t));
    for (var r = this._groups, e = r.length, o = [], s = [], u = 0; u < e; ++u)
      for (var a, h = r[u], l = h.length, c = 0; c < l; ++c)
        if ((a = h[c])) {
          for (var f, d = t.call(a, a.__data__, c, h), p = di(a, i), g = 0, m = d.length; g < m; ++g) (f = d[g]) && li(f, n, i, g, d, p);
          (o.push(d), s.push(a));
        }
    return new Si(o, s, n, i);
  },
  selectChild: Oi.selectChild,
  selectChildren: Oi.selectChildren,
  filter: function (t) {
    'function' != typeof t && (t = g(t));
    for (var n = this._groups, i = n.length, r = Array(i), e = 0; e < i; ++e)
      for (var o, s = n[e], u = s.length, a = (r[e] = []), h = 0; h < u; ++h) (o = s[h]) && t.call(o, o.__data__, h, s) && a.push(o);
    return new Si(r, this._parents, this._name, this._id);
  },
  merge: function (t) {
    if (t._id !== this._id) throw Error();
    for (var n = this._groups, i = t._groups, r = n.length, e = Math.min(r, i.length), o = Array(r), s = 0; s < e; ++s)
      for (var u, a = n[s], h = i[s], l = a.length, c = (o[s] = Array(l)), f = 0; f < l; ++f) (u = a[f] || h[f]) && (c[f] = u);
    for (; s < r; ++s) o[s] = n[s];
    return new Si(o, this._parents, this._name, this._id);
  },
  selection: function () {
    return new Xi(this._groups, this._parents);
  },
  transition: function () {
    for (var t = this._name, n = this._id, i = Yi(), r = this._groups, e = r.length, o = 0; o < e; ++o)
      for (var s, u = r[o], a = u.length, h = 0; h < a; ++h)
        if ((s = u[h])) {
          var l = di(s, n);
          li(s, t, i, h, u, { time: l.time + l.delay + l.duration, delay: 0, duration: l.duration, ease: l.ease });
        }
    return new Si(r, this._parents, t, i);
  },
  call: Oi.call,
  nodes: Oi.nodes,
  node: Oi.node,
  size: Oi.size,
  empty: Oi.empty,
  each: Oi.each,
  on: function (t, n) {
    var i = this._id;
    return arguments.length < 2
      ? di(this.node(), i).on.on(t)
      : this.each(
          (function (t, n, i) {
            var r,
              e,
              o = (function (t) {
                return (t + '')
                  .trim()
                  .split(/^|\s+/)
                  .every(function (t) {
                    var n = t.indexOf('.');
                    return (n >= 0 && (t = t.slice(0, n)), !t || 'start' === t);
                  });
              })(n)
                ? ci
                : fi;
            return function () {
              var s = o(this, t),
                u = s.on;
              (u !== r && (e = (r = u).copy()).on(n, i), (s.on = e));
            };
          })(i, t, n),
        );
  },
  attr: function (t, n) {
    var i = a(t),
      r = 'transform' === i ? _n : wi;
    return this.attrTween(t, 'function' == typeof n ? (i.local ? ki : bi)(i, r, mi(this, 'attr.' + t, n)) : null == n ? (i.local ? xi : yi)(i) : (i.local ? $i : Ci)(i, r, n));
  },
  attrTween: function (t, n) {
    var i = 'attr.' + t;
    if (arguments.length < 2) return (i = this.tween(i)) && i._value;
    if (null == n) return this.tween(i, null);
    if ('function' != typeof n) throw Error();
    var r = a(t);
    return this.tween(i, (r.local ? Mi : Ei)(r, n));
  },
  style: function (t, n, i) {
    var r = 'transform' == (t += '') ? Fn : wi;
    return null == n
      ? this.styleTween(
          t,
          (function (t, n) {
            var i, r, e;
            return function () {
              var o = j(this, t),
                s = (this.style.removeProperty(t), j(this, t));
              return o === s ? null : o === i && s === r ? e : (e = n((i = o), (r = s)));
            };
          })(t, r),
        ).on('end.style.' + t, Ti(t))
      : 'function' == typeof n
        ? this.styleTween(
            t,
            (function (t, n, i) {
              var r, e, o;
              return function () {
                var s = j(this, t),
                  u = i(this),
                  a = u + '';
                return (null == u && (this.style.removeProperty(t), (a = u = j(this, t))), s === a ? null : s === r && a === e ? o : ((e = a), (o = n((r = s), u))));
              };
            })(t, r, mi(this, 'style.' + t, n)),
          ).each(
            (function (t, n) {
              var i,
                r,
                e,
                o,
                s = 'style.' + n,
                u = 'end.' + s;
              return function () {
                var a = fi(this, t),
                  h = a.on,
                  l = null == a.value[s] ? o || (o = Ti(n)) : void 0;
                ((h === i && e === l) || (r = (i = h).copy()).on(u, (e = l)), (a.on = r));
              };
            })(this._id, t),
          )
        : this.styleTween(
            t,
            (function (t, n, i) {
              var r,
                e,
                o = i + '';
              return function () {
                var s = j(this, t);
                return s === o ? null : s === r ? e : (e = n((r = s), i));
              };
            })(t, r, n),
            i,
          ).on('end.style.' + t, null);
  },
  styleTween: function (t, n, i) {
    var r = 'style.' + (t += '');
    if (arguments.length < 2) return (r = this.tween(r)) && r._value;
    if (null == n) return this.tween(r, null);
    if ('function' != typeof n) throw Error();
    return this.tween(
      r,
      (function (t, n, i) {
        var r, e;
        function o() {
          var o = n.apply(this, arguments);
          return (
            o !== e &&
              (r =
                (e = o) &&
                (function (t, n, i) {
                  return function (r) {
                    this.style.setProperty(t, n.call(this, r), i);
                  };
                })(t, o, i)),
            r
          );
        }
        return ((o._value = n), o);
      })(t, n, null == i ? '' : i),
    );
  },
  text: function (t) {
    return this.tween(
      'text',
      'function' == typeof t
        ? (function (t) {
            return function () {
              var n = t(this);
              this.textContent = null == n ? '' : n;
            };
          })(mi(this, 'text', t))
        : (function (t) {
            return function () {
              this.textContent = t;
            };
          })(null == t ? '' : t + ''),
    );
  },
  textTween: function (t) {
    var n = 'text';
    if (arguments.length < 1) return (n = this.tween(n)) && n._value;
    if (null == t) return this.tween(n, null);
    if ('function' != typeof t) throw Error();
    return this.tween(
      n,
      (function (t) {
        var n, i;
        function r() {
          var r = t.apply(this, arguments);
          return (
            r !== i &&
              (n =
                (i = r) &&
                (function (t) {
                  return function (n) {
                    this.textContent = t.call(this, n);
                  };
                })(r)),
            n
          );
        }
        return ((r._value = t), r);
      })(t),
    );
  },
  remove: function () {
    return this.on(
      'end.remove',
      (function (t) {
        return function () {
          var n = this.parentNode;
          for (var i in this.__transition) if (+i !== t) return;
          n && n.removeChild(this);
        };
      })(this._id),
    );
  },
  tween: function (t, n) {
    var i = this._id;
    if (((t += ''), arguments.length < 2)) {
      for (var r, e = di(this.node(), i).tween, o = 0, s = e.length; o < s; ++o) if ((r = e[o]).name === t) return r.value;
      return null;
    }
    return this.each((null == n ? vi : gi)(i, t, n));
  },
  delay: function (t) {
    var n = this._id;
    return arguments.length ? this.each(('function' == typeof t ? Li : zi)(n, t)) : di(this.node(), n).delay;
  },
  duration: function (t) {
    var n = this._id;
    return arguments.length ? this.each(('function' == typeof t ? Ai : Ni)(n, t)) : di(this.node(), n).duration;
  },
  ease: function (t) {
    var n = this._id;
    return arguments.length
      ? this.each(
          (function (t, n) {
            if ('function' != typeof n) throw Error();
            return function () {
              fi(this, t).ease = n;
            };
          })(n, t),
        )
      : di(this.node(), n).ease;
  },
  easeVarying: function (t) {
    if ('function' != typeof t) throw Error();
    return this.each(
      (function (t, n) {
        return function () {
          var i = n.apply(this, arguments);
          if ('function' != typeof i) throw Error();
          fi(this, t).ease = i;
        };
      })(this._id, t),
    );
  },
  end: function () {
    var t,
      n,
      i = this,
      r = i._id,
      e = i.size();
    return new Promise(function (o, s) {
      var u = { value: s },
        a = {
          value: function () {
            0 == --e && o();
          },
        };
      (i.each(function () {
        var i = fi(this, r),
          e = i.on;
        (e !== t && ((n = (t = e).copy())._.cancel.push(u), n._.interrupt.push(u), n._.end.push(a)), (i.on = n));
      }),
        0 === e && o());
    });
  },
  [Symbol.iterator]: Oi[Symbol.iterator],
};
var Ri = {
  time: null,
  delay: 0,
  duration: 250,
  ease: function (t) {
    return ((t *= 2) <= 1 ? t * t * t : (t -= 2) * t * t + 2) / 2;
  },
};
function Hi(t, n) {
  for (var i; !(i = t.__transition) || !(i = i[n]); ) if (!(t = t.parentNode)) throw Error(`transition ${n} not found`);
  return i;
}
((gt.prototype.interrupt = function (t) {
  return this.each(function () {
    pi(this, t);
  });
}),
  (gt.prototype.transition = function (t) {
    var n, i;
    t instanceof Si ? ((n = t._id), (t = t._name)) : ((n = Yi()), ((i = Ri).time = ti()), (t = null == t ? null : t + ''));
    for (var r = this._groups, e = r.length, o = 0; o < e; ++o) for (var s, u = r[o], a = u.length, h = 0; h < a; ++h) (s = u[h]) && li(s, t, n, h, u, i || Hi(s, n));
    return new Si(r, this._parents, t, n);
  }));
var ji = t => () => t;
function Fi(t, { sourceEvent: n, target: i, transform: r, dispatch: e }) {
  Object.defineProperties(this, {
    type: { value: t, enumerable: !0, configurable: !0 },
    sourceEvent: { value: n, enumerable: !0, configurable: !0 },
    target: { value: i, enumerable: !0, configurable: !0 },
    transform: { value: r, enumerable: !0, configurable: !0 },
    _: { value: e },
  });
}
function _i(t, n, i) {
  ((this.k = t), (this.x = n), (this.y = i));
}
_i.prototype = {
  constructor: _i,
  scale: function (t) {
    return 1 === t ? this : new _i(this.k * t, this.x, this.y);
  },
  translate: function (t, n) {
    return (0 === t) & (0 === n) ? this : new _i(this.k, this.x + this.k * t, this.y + this.k * n);
  },
  apply: function (t) {
    return [t[0] * this.k + this.x, t[1] * this.k + this.y];
  },
  applyX: function (t) {
    return t * this.k + this.x;
  },
  applyY: function (t) {
    return t * this.k + this.y;
  },
  invert: function (t) {
    return [(t[0] - this.x) / this.k, (t[1] - this.y) / this.k];
  },
  invertX: function (t) {
    return (t - this.x) / this.k;
  },
  invertY: function (t) {
    return (t - this.y) / this.k;
  },
  rescaleX: function (t) {
    return t.copy().domain(t.range().map(this.invertX, this).map(t.invert, t));
  },
  rescaleY: function (t) {
    return t.copy().domain(t.range().map(this.invertY, this).map(t.invert, t));
  },
  toString: function () {
    return 'translate(' + this.x + ',' + this.y + ') scale(' + this.k + ')';
  },
};
var Di = new _i(1, 0, 0);
function Zi(t) {
  t.stopImmediatePropagation();
}
function Gi(t) {
  (t.preventDefault(), t.stopImmediatePropagation());
}
function Ii(t) {
  return !((t.ctrlKey && 'wheel' !== t.type) || t.button);
}
function Ji() {
  var t = this;
  return t instanceof SVGElement
    ? (t = t.ownerSVGElement || t).hasAttribute('viewBox')
      ? [
          [(t = t.viewBox.baseVal).x, t.y],
          [t.x + t.width, t.y + t.height],
        ]
      : [
          [0, 0],
          [t.width.baseVal.value, t.height.baseVal.value],
        ]
    : [
        [0, 0],
        [t.clientWidth, t.clientHeight],
      ];
}
function Pi() {
  return this.__zoom || Di;
}
function Ui(t) {
  return -t.deltaY * (1 === t.deltaMode ? 0.05 : t.deltaMode ? 1 : 0.002) * (t.ctrlKey ? 10 : 1);
}
function Vi() {
  return navigator.maxTouchPoints || 'ontouchstart' in this;
}
function qi(t, n, i) {
  var r = t.invertX(n[0][0]) - i[0][0],
    e = t.invertX(n[1][0]) - i[1][0],
    o = t.invertY(n[0][1]) - i[0][1],
    s = t.invertY(n[1][1]) - i[1][1];
  return t.translate(e > r ? (r + e) / 2 : Math.min(0, r) || Math.max(0, e), s > o ? (o + s) / 2 : Math.min(0, o) || Math.max(0, s));
}
function Wi(t) {
  var n = 0,
    i = t.children,
    r = i && i.length;
  if (r) for (; --r >= 0; ) n += i[r].value;
  else n = 1;
  t.value = n;
}
function Qi(t, n) {
  var i,
    r,
    e,
    o,
    s,
    u = new ir(t),
    a = +t.value && (u.value = t.value),
    h = [u];
  for (null == n && (n = Ki); (i = h.pop()); )
    if ((a && (i.value = +i.data.value), (e = n(i.data)) && (s = e.length)))
      for (i.children = Array(s), o = s - 1; o >= 0; --o) (h.push((r = i.children[o] = new ir(e[o]))), (r.parent = i), (r.depth = i.depth + 1));
  return u.eachBefore(nr);
}
function Ki(t) {
  return t.children;
}
function tr(t) {
  t.data = t.data.data;
}
function nr(t) {
  var n = 0;
  do {
    t.height = n;
  } while ((t = t.parent) && t.height < ++n);
}
function ir(t) {
  ((this.data = t), (this.depth = this.height = 0), (this.parent = null));
}
ir.prototype = Qi.prototype = {
  constructor: ir,
  count: function () {
    return this.eachAfter(Wi);
  },
  each: function (t) {
    var n,
      i,
      r,
      e,
      o = this,
      s = [o];
    do {
      for (n = s.reverse(), s = []; (o = n.pop()); ) if ((t(o), (i = o.children))) for (r = 0, e = i.length; r < e; ++r) s.push(i[r]);
    } while (s.length);
    return this;
  },
  eachAfter: function (t) {
    for (var n, i, r, e = this, o = [e], s = []; (e = o.pop()); ) if ((s.push(e), (n = e.children))) for (i = 0, r = n.length; i < r; ++i) o.push(n[i]);
    for (; (e = s.pop()); ) t(e);
    return this;
  },
  eachBefore: function (t) {
    for (var n, i, r = this, e = [r]; (r = e.pop()); ) if ((t(r), (n = r.children))) for (i = n.length - 1; i >= 0; --i) e.push(n[i]);
    return this;
  },
  sum: function (t) {
    return this.eachAfter(function (n) {
      for (var i = +t(n.data) || 0, r = n.children, e = r && r.length; --e >= 0; ) i += r[e].value;
      n.value = i;
    });
  },
  sort: function (t) {
    return this.eachBefore(function (n) {
      n.children && n.children.sort(t);
    });
  },
  path: function (t) {
    for (
      var n = this,
        i = (function (t, n) {
          if (t === n) return t;
          var i = t.ancestors(),
            r = n.ancestors(),
            e = null;
          for (t = i.pop(), n = r.pop(); t === n; ) ((e = t), (t = i.pop()), (n = r.pop()));
          return e;
        })(n, t),
        r = [n];
      n !== i;
    )
      r.push((n = n.parent));
    for (var e = r.length; t !== i; ) (r.splice(e, 0, t), (t = t.parent));
    return r;
  },
  ancestors: function () {
    for (var t = this, n = [t]; (t = t.parent); ) n.push(t);
    return n;
  },
  descendants: function () {
    var t = [];
    return (
      this.each(function (n) {
        t.push(n);
      }),
      t
    );
  },
  leaves: function () {
    var t = [];
    return (
      this.eachBefore(function (n) {
        n.children || t.push(n);
      }),
      t
    );
  },
  links: function () {
    var t = this,
      n = [];
    return (
      t.each(function (i) {
        i !== t && n.push({ source: i.parent, target: i });
      }),
      n
    );
  },
  copy: function () {
    return Qi(this).eachBefore(tr);
  },
};
const { version: rr } = { version: '2.1.2' },
  er = Object.freeze({ children: t => t.children, nodeSize: t => t.data.size, spacing: 0 });
function or(t) {
  const n = Object.assign({}, er, t);
  function i(t) {
    const i = n[t];
    return 'function' == typeof i ? i : () => i;
  }
  function r(t) {
    const n = o(
      (function () {
        const t = e(),
          n = i('nodeSize'),
          r = i('spacing');
        return class extends t {
          constructor(t) {
            (super(t), Object.assign(this, { x: 0, y: 0, relX: 0, prelim: 0, shift: 0, change: 0, lExt: this, lExtRelX: 0, lThr: null, rExt: this, rExtRelX: 0, rThr: null }));
          }
          get size() {
            return n(this.data);
          }
          spacing(t) {
            return r(this.data, t.data);
          }
          get x() {
            return this.data.x;
          }
          set x(t) {
            this.data.x = t;
          }
          get y() {
            return this.data.y;
          }
          set y(t) {
            this.data.y = t;
          }
          update() {
            return (sr(this), ur(this), this);
          }
        };
      })(),
      t,
      t => t.children,
    );
    return (n.update(), n.data);
  }
  function e() {
    const t = i('nodeSize'),
      n = i('spacing');
    return class i extends Qi.prototype.constructor {
      constructor(t) {
        super(t);
      }
      copy() {
        const t = o(this.constructor, this, t => t.children);
        return (t.each(t => (t.data = t.data.data)), t);
      }
      get size() {
        return t(this);
      }
      spacing(t) {
        return n(this, t);
      }
      get nodes() {
        return this.descendants();
      }
      get xSize() {
        return this.size[0];
      }
      get ySize() {
        return this.size[1];
      }
      get top() {
        return this.y;
      }
      get bottom() {
        return this.y + this.ySize;
      }
      get left() {
        return this.x - this.xSize / 2;
      }
      get right() {
        return this.x + this.xSize / 2;
      }
      get root() {
        const t = this.ancestors();
        return t[t.length - 1];
      }
      get numChildren() {
        return this.hasChildren ? this.children.length : 0;
      }
      get hasChildren() {
        return !this.noChildren;
      }
      get noChildren() {
        return null === this.children;
      }
      get firstChild() {
        return this.hasChildren ? this.children[0] : null;
      }
      get lastChild() {
        return this.hasChildren ? this.children[this.numChildren - 1] : null;
      }
      get extents() {
        return (this.children || []).reduce((t, n) => i.maxExtents(t, n.extents), this.nodeExtents);
      }
      get nodeExtents() {
        return { top: this.top, bottom: this.bottom, left: this.left, right: this.right };
      }
      static maxExtents(t, n) {
        return { top: Math.min(t.top, n.top), bottom: Math.max(t.bottom, n.bottom), left: Math.min(t.left, n.left), right: Math.max(t.right, n.right) };
      }
    };
  }
  function o(t, n, i) {
    const r = (n, e) => {
      const o = new t(n);
      Object.assign(o, { parent: e, depth: null === e ? 0 : e.depth + 1, height: 0, length: 1 });
      const s = i(n) || [];
      return (
        (o.children = 0 === s.length ? null : s.map(t => r(t, o))),
        o.children &&
          Object.assign(
            o,
            o.children.reduce((t, n) => ({ height: Math.max(t.height, n.height + 1), length: t.length + n.length }), o),
          ),
        o
      );
    };
    return r(n, null);
  }
  return (
    Object.assign(r, {
      nodeSize(t) {
        return arguments.length ? ((n.nodeSize = t), r) : n.nodeSize;
      },
      spacing(t) {
        return arguments.length ? ((n.spacing = t), r) : n.spacing;
      },
      children(t) {
        return arguments.length ? ((n.children = t), r) : n.children;
      },
      hierarchy(t, i) {
        const r = void 0 === i ? n.children : i;
        return o(e(), t, r);
      },
      dump(t) {
        const n = i('nodeSize'),
          r = t => i => {
            const e = t + '  ',
              o = t + '    ',
              { x: s, y: u } = i,
              a = n(i),
              h = i.children || [],
              l = 0 === h.length ? ' ' : `,${e}children: [${o}${h.map(r(o)).join(o)}${e}],${t}`;
            return `{ size: [${a.join(', ')}],${e}x: ${s}, y: ${u}${l}},`;
          };
        return r('\n')(t);
      },
    }),
    r
  );
}
or.version = rr;
const sr = (t, n = 0) => (
    (t.y = n),
    (t.children || []).reduce(
      (n, i) => {
        const [r, e] = n;
        sr(i, t.y + t.ySize);
        const o = (0 === r ? i.lExt : i.rExt).bottom;
        return (0 !== r && hr(t, r, e), [r + 1, mr(o, r, e)]);
      },
      [0, null],
    ),
    ar(t),
    gr(t),
    t
  ),
  ur = (t, n, i) => {
    void 0 === n && ((n = -t.relX - t.prelim), (i = 0));
    const r = n + t.relX;
    return ((t.relX = r + t.prelim - i), (t.prelim = 0), (t.x = i + t.relX), (t.children || []).forEach(n => ur(n, r, t.x)), t);
  },
  ar = t => {
    (t.children || []).reduce(
      (t, n) => {
        const [i, r] = t,
          e = i + n.shift,
          o = r + e + n.change;
        return ((n.relX += o), [e, o]);
      },
      [0, 0],
    );
  },
  hr = (t, n, i) => {
    const r = t.children[n - 1],
      e = t.children[n];
    let o = r,
      s = r.relX,
      u = e,
      a = e.relX,
      h = !0;
    for (; o && u; ) {
      o.bottom > i.lowY && (i = i.next);
      const r = s + o.prelim - (a + u.prelim) + o.xSize / 2 + u.xSize / 2 + o.spacing(u);
      ((r > 0 || (r < 0 && h)) && ((a += r), lr(e, r), cr(t, n, i.index, r)), (h = !1));
      const l = o.bottom,
        c = u.bottom;
      (l <= c && ((o = dr(o)), o && (s += o.relX)), l >= c && ((u = fr(u)), u && (a += u.relX)));
    }
    !o && u ? pr(t, n, u, a) : o && !u && vr(t, n, o, s);
  },
  lr = (t, n) => {
    ((t.relX += n), (t.lExtRelX += n), (t.rExtRelX += n));
  },
  cr = (t, n, i, r) => {
    const e = t.children[n],
      o = n - i;
    if (o > 1) {
      const n = r / o;
      ((t.children[i + 1].shift += n), (e.shift -= n), (e.change -= r - n));
    }
  },
  fr = t => (t.hasChildren ? t.firstChild : t.lThr),
  dr = t => (t.hasChildren ? t.lastChild : t.rThr),
  pr = (t, n, i, r) => {
    const e = t.firstChild,
      o = e.lExt,
      s = t.children[n];
    o.lThr = i;
    const u = r - i.relX - e.lExtRelX;
    ((o.relX += u), (o.prelim -= u), (e.lExt = s.lExt), (e.lExtRelX = s.lExtRelX));
  },
  vr = (t, n, i, r) => {
    const e = t.children[n],
      o = e.rExt,
      s = t.children[n - 1];
    o.rThr = i;
    const u = r - i.relX - e.rExtRelX;
    ((o.relX += u), (o.prelim -= u), (e.rExt = s.rExt), (e.rExtRelX = s.rExtRelX));
  },
  gr = t => {
    if (t.hasChildren) {
      const n = t.firstChild,
        i = t.lastChild;
      Object.assign(t, { prelim: (n.prelim + n.relX - n.xSize / 2 + i.relX + i.prelim + i.xSize / 2) / 2, lExt: n.lExt, lExtRelX: n.lExtRelX, rExt: i.rExt, rExtRelX: i.rExtRelX });
    }
  },
  mr = (t, n, i) => {
    for (; null !== i && t >= i.lowY; ) i = i.next;
    return { lowY: t, index: n, next: i };
  };
function wr(t) {
  return function () {
    return t;
  };
}
const yr = Math.PI,
  xr = 2 * yr,
  Cr = 1e-6,
  $r = xr - Cr;
function br(t) {
  this._ += t[0];
  for (let n = 1, i = t.length; n < i; ++n) this._ += arguments[n] + t[n];
}
class kr {
  constructor(t) {
    ((this._x0 = this._y0 = this._x1 = this._y1 = null),
      (this._ = ''),
      (this._append =
        null == t
          ? br
          : (function (t) {
              let n = Math.floor(t);
              if (!(n >= 0)) throw Error('invalid digits: ' + t);
              if (n > 15) return br;
              const i = 10 ** n;
              return function (t) {
                this._ += t[0];
                for (let n = 1, r = t.length; n < r; ++n) this._ += Math.round(arguments[n] * i) / i + t[n];
              };
            })(t)));
  }
  moveTo(t, n) {
    this._append`M${(this._x0 = this._x1 = +t)},${(this._y0 = this._y1 = +n)}`;
  }
  closePath() {
    null !== this._x1 && ((this._x1 = this._x0), (this._y1 = this._y0), this._append`Z`);
  }
  lineTo(t, n) {
    this._append`L${(this._x1 = +t)},${(this._y1 = +n)}`;
  }
  quadraticCurveTo(t, n, i, r) {
    this._append`Q${+t},${+n},${(this._x1 = +i)},${(this._y1 = +r)}`;
  }
  bezierCurveTo(t, n, i, r, e, o) {
    this._append`C${+t},${+n},${+i},${+r},${(this._x1 = +e)},${(this._y1 = +o)}`;
  }
  arcTo(t, n, i, r, e) {
    if (((t = +t), (n = +n), (i = +i), (r = +r), (e = +e) < 0)) throw Error('negative radius: ' + e);
    let o = this._x1,
      s = this._y1,
      u = i - t,
      a = r - n,
      h = o - t,
      l = s - n,
      c = h * h + l * l;
    if (null === this._x1) this._append`M${(this._x1 = t)},${(this._y1 = n)}`;
    else if (c > Cr)
      if (Math.abs(l * u - a * h) > Cr && e) {
        let f = i - o,
          d = r - s,
          p = u * u + a * a,
          v = f * f + d * d,
          g = Math.sqrt(p),
          m = Math.sqrt(c),
          w = e * Math.tan((yr - Math.acos((p + c - v) / (2 * g * m))) / 2),
          y = w / m,
          x = w / g;
        (Math.abs(y - 1) > Cr && this._append`L${t + y * h},${n + y * l}`, this._append`A${e},${e},0,0,${+(l * f > h * d)},${(this._x1 = t + x * u)},${(this._y1 = n + x * a)}`);
      } else this._append`L${(this._x1 = t)},${(this._y1 = n)}`;
  }
  arc(t, n, i, r, e, o) {
    if (((t = +t), (n = +n), (o = !!o), (i = +i) < 0)) throw Error('negative radius: ' + i);
    let s = i * Math.cos(r),
      u = i * Math.sin(r),
      a = t + s,
      h = n + u,
      l = 1 ^ o,
      c = o ? r - e : e - r;
    (null === this._x1 ? this._append`M${a},${h}` : (Math.abs(this._x1 - a) > Cr || Math.abs(this._y1 - h) > Cr) && this._append`L${a},${h}`,
      i &&
        (c < 0 && (c = (c % xr) + xr),
        c > $r
          ? this._append`A${i},${i},0,1,${l},${t - s},${n - u}A${i},${i},0,1,${l},${(this._x1 = a)},${(this._y1 = h)}`
          : c > Cr && this._append`A${i},${i},0,${+(c >= yr)},${l},${(this._x1 = t + i * Math.cos(e))},${(this._y1 = n + i * Math.sin(e))}`));
  }
  rect(t, n, i, r) {
    this._append`M${(this._x0 = this._x1 = +t)},${(this._y0 = this._y1 = +n)}h${(i = +i)}v${+r}h${-i}Z`;
  }
  toString() {
    return this._;
  }
}
var Mr = Array.prototype.slice;
function Er(t) {
  return t[0];
}
function Lr(t) {
  return t[1];
}
class zr {
  constructor(t, n) {
    ((this._context = t), (this._x = n));
  }
  areaStart() {
    this._line = 0;
  }
  areaEnd() {
    this._line = NaN;
  }
  lineStart() {
    this._point = 0;
  }
  lineEnd() {
    ((this._line || (0 !== this._line && 1 === this._point)) && this._context.closePath(), (this._line = 1 - this._line));
  }
  point(t, n) {
    switch (((t = +t), (n = +n), this._point)) {
      case 0:
        ((this._point = 1), this._line ? this._context.lineTo(t, n) : this._context.moveTo(t, n));
        break;
      case 1:
        this._point = 2;
      default:
        this._x
          ? this._context.bezierCurveTo((this._x0 = (this._x0 + t) / 2), this._y0, this._x0, n, t, n)
          : this._context.bezierCurveTo(this._x0, (this._y0 = (this._y0 + n) / 2), t, this._y0, t, n);
    }
    ((this._x0 = t), (this._y0 = n));
  }
}
function Ar(t) {
  return new zr(t, !0);
}
function Nr(t) {
  return t.source;
}
function Xr(t) {
  return t.target;
}
const Tr = {
  selection: gt,
  select: mt,
  max: function (t, n) {
    let i;
    if (void 0 === n) for (const n of t) null != n && (i < n || (void 0 === i && n >= n)) && (i = n);
    else {
      let r = -1;
      for (let e of t) null != (e = n(e, ++r, t)) && (i < e || (void 0 === i && e >= e)) && (i = e);
    }
    return i;
  },
  min: function (t, n) {
    let i;
    if (void 0 === n) for (const n of t) null != n && (i > n || (void 0 === i && n >= n)) && (i = n);
    else {
      let r = -1;
      for (let e of t) null != (e = n(e, ++r, t)) && (i > e || (void 0 === i && e >= e)) && (i = e);
    }
    return i;
  },
  sum: function (t, n) {
    let i = 0;
    if (void 0 === n) for (let n of t) (n = +n) && (i += n);
    else {
      let r = -1;
      for (let e of t) (e = +n(e, ++r, t)) && (i += e);
    }
    return i;
  },
  cumsum: function (t, n) {
    var i = 0,
      r = 0;
    return Float64Array.from(t, void 0 === n ? t => (i += +t || 0) : e => (i += +n(e, r++, t) || 0));
  },
  tree: function () {
    var t = St,
      n = 1,
      i = 1,
      r = null;
    function e(e) {
      var a = (function (t) {
        for (var n, i, r, e, o, s = new jt(t, 0), u = [s]; (n = u.pop()); )
          if ((r = n._.children)) for (n.children = Array((o = r.length)), e = o - 1; e >= 0; --e) (u.push((i = n.children[e] = new jt(r[e], e))), (i.parent = n));
        return (((s.parent = new jt(null, 0)).children = [s]), s);
      })(e);
      if ((a.eachAfter(o), (a.parent.m = -a.z), a.eachBefore(s), r)) e.eachBefore(u);
      else {
        var h = e,
          l = e,
          c = e;
        e.eachBefore(function (t) {
          (t.x < h.x && (h = t), t.x > l.x && (l = t), t.depth > c.depth && (c = t));
        });
        var f = h === l ? 1 : t(h, l) / 2,
          d = f - h.x,
          p = n / (l.x + f + d),
          v = i / (c.depth || 1);
        e.eachBefore(function (t) {
          ((t.x = (t.x + d) * p), (t.y = t.depth * v));
        });
      }
      return e;
    }
    function o(n) {
      var i = n.children,
        r = n.parent.children,
        e = n.i ? r[n.i - 1] : null;
      if (i) {
        !(function (t) {
          for (var n, i = 0, r = 0, e = t.children, o = e.length; --o >= 0; ) (((n = e[o]).z += i), (n.m += i), (i += n.s + (r += n.c)));
        })(n);
        var o = (i[0].z + i[i.length - 1].z) / 2;
        e ? ((n.z = e.z + t(n._, e._)), (n.m = n.z - o)) : (n.z = o);
      } else e && (n.z = e.z + t(n._, e._));
      n.parent.A = (function (n, i, r) {
        if (i) {
          for (var e, o = n, s = n, u = i, a = o.parent.children[0], h = o.m, l = s.m, c = u.m, f = a.m; (u = Ot(u)), (o = Yt(o)), u && o; )
            ((a = Yt(a)),
              ((s = Ot(s)).a = n),
              (e = u.z + c - o.z - h + t(u._, o._)) > 0 && (Rt(Ht(u, n, r), n, e), (h += e), (l += e)),
              (c += u.m),
              (h += o.m),
              (f += a.m),
              (l += s.m));
          (u && !Ot(s) && ((s.t = u), (s.m += c - l)), o && !Yt(a) && ((a.t = o), (a.m += h - f), (r = n)));
        }
        return r;
      })(n, e, n.parent.A || r[0]);
    }
    function s(t) {
      ((t._.x = t.z + t.parent.m), (t.m += t.parent.m));
    }
    function u(t) {
      ((t.x *= n), (t.y = t.depth * i));
    }
    return (
      (e.separation = function (n) {
        return arguments.length ? ((t = n), e) : t;
      }),
      (e.size = function (t) {
        return arguments.length ? ((r = !1), (n = +t[0]), (i = +t[1]), e) : r ? null : [n, i];
      }),
      (e.nodeSize = function (t) {
        return arguments.length ? ((r = !0), (n = +t[0]), (i = +t[1]), e) : r ? [n, i] : null;
      }),
      e
    );
  },
  stratify: function () {
    var t,
      n = Nt,
      i = Xt;
    function r(r) {
      var e,
        o,
        s,
        u,
        a,
        h,
        l,
        c,
        f = Array.from(r),
        d = n,
        p = i,
        v = new Map();
      if (null != t) {
        const n = f.map((n, i) =>
            (function (t) {
              let n = (t = '' + t).length;
              return (Bt(t, n - 1) && !Bt(t, n - 2) && (t = t.slice(0, -1)), '/' === t[0] ? t : '/' + t);
            })(t(n, i, r)),
          ),
          i = n.map(Tt),
          e = new Set(n).add('');
        for (const t of i) e.has(t) || (e.add(t), n.push(t), i.push(Tt(t)), f.push(At));
        ((d = (t, i) => n[i]), (p = (t, n) => i[n]));
      }
      for (s = 0, e = f.length; s < e; ++s)
        ((h = f[s] = new Mt((o = f[s]))),
          null != (l = d(o, s, r)) && (l += '') && ((c = h.id = l), v.set(c, v.has(c) ? zt : h)),
          null != (l = p(o, s, r)) && (l += '') && (h.parent = l));
      for (s = 0; s < e; ++s)
        if ((l = (h = f[s]).parent)) {
          if (!(a = v.get(l))) throw Error('missing: ' + l);
          if (a === zt) throw Error('ambiguous: ' + l);
          (a.children ? a.children.push(h) : (a.children = [h]), (h.parent = a));
        } else {
          if (u) throw Error('multiple roots');
          u = h;
        }
      if (!u) throw Error('no root');
      if (null != t) {
        for (; u.data === At && 1 === u.children.length; ) ((u = u.children[0]), --e);
        for (let t = f.length - 1; t >= 0 && (h = f[t]).data === At; --t) h.data = null;
      }
      if (
        ((u.parent = Lt),
        u
          .eachBefore(function (t) {
            ((t.depth = t.parent.depth + 1), --e);
          })
          .eachBefore(kt),
        (u.parent = null),
        e > 0)
      )
        throw Error('cycle');
      return u;
    }
    return (
      (r.id = function (t) {
        return arguments.length ? ((n = Et(t)), r) : n;
      }),
      (r.parentId = function (t) {
        return arguments.length ? ((i = Et(t)), r) : i;
      }),
      (r.path = function (n) {
        return arguments.length ? ((t = Et(n)), r) : t;
      }),
      r
    );
  },
  zoom: function () {
    var t,
      n,
      i,
      r = Ii,
      e = Ji,
      o = qi,
      s = Ui,
      u = Vi,
      a = [0, 1 / 0],
      h = [
        [-1 / 0, -1 / 0],
        [1 / 0, 1 / 0],
      ],
      l = 250,
      c = In,
      f = _t('start', 'zoom', 'end'),
      d = 0,
      p = 10;
    function v(t) {
      t.property('__zoom', Pi)
        .on('wheel.zoom', $, { passive: !1 })
        .on('mousedown.zoom', b)
        .on('dblclick.zoom', k)
        .filter(u)
        .on('touchstart.zoom', M)
        .on('touchmove.zoom', E)
        .on('touchend.zoom touchcancel.zoom', L)
        .style('-webkit-tap-highlight-color', 'rgba(0,0,0,0)');
    }
    function g(t, n) {
      return (n = Math.max(a[0], Math.min(a[1], n))) === t.k ? t : new _i(n, t.x, t.y);
    }
    function m(t, n, i) {
      var r = n[0] - i[0] * t.k,
        e = n[1] - i[1] * t.k;
      return r === t.x && e === t.y ? t : new _i(t.k, r, e);
    }
    function w(t) {
      return [(+t[0][0] + +t[1][0]) / 2, (+t[0][1] + +t[1][1]) / 2];
    }
    function y(t, n, i, r) {
      t.on('start.zoom', function () {
        x(this, arguments).event(r).start();
      })
        .on('interrupt.zoom end.zoom', function () {
          x(this, arguments).event(r).end();
        })
        .tween('zoom', function () {
          var t = this,
            o = arguments,
            s = x(t, o).event(r),
            u = e.apply(t, o),
            a = null == i ? w(u) : 'function' == typeof i ? i.apply(t, o) : i,
            h = Math.max(u[1][0] - u[0][0], u[1][1] - u[0][1]),
            l = t.__zoom,
            f = 'function' == typeof n ? n.apply(t, o) : n,
            d = c(l.invert(a).concat(h / l.k), f.invert(a).concat(h / f.k));
          return function (t) {
            if (1 === t) t = f;
            else {
              var n = d(t),
                i = h / n[2];
              t = new _i(i, a[0] - n[0] * i, a[1] - n[1] * i);
            }
            s.zoom(null, t);
          };
        });
    }
    function x(t, n, i) {
      return (!i && t.__zooming) || new C(t, n);
    }
    function C(t, n) {
      ((this.that = t), (this.args = n), (this.active = 0), (this.sourceEvent = null), (this.extent = e.apply(t, n)), (this.taps = 0));
    }
    function $(t, ...n) {
      if (r.apply(this, arguments)) {
        var i = x(this, n).event(t),
          e = this.__zoom,
          u = Math.max(a[0], Math.min(a[1], e.k * Math.pow(2, s.apply(this, arguments)))),
          l = wt(t);
        if (i.wheel) ((i.mouse[0][0] === l[0] && i.mouse[0][1] === l[1]) || (i.mouse[1] = e.invert((i.mouse[0] = l))), clearTimeout(i.wheel));
        else {
          if (e.k === u) return;
          ((i.mouse = [l, e.invert(l)]), pi(this), i.start());
        }
        (Gi(t),
          (i.wheel = setTimeout(function () {
            ((i.wheel = null), i.end());
          }, 150)),
          i.zoom('mouse', o(m(g(e, u), i.mouse[0], i.mouse[1]), i.extent, h)));
      }
    }
    function b(t, ...n) {
      if (!i && r.apply(this, arguments)) {
        var e = t.currentTarget,
          s = x(this, n, !0).event(t),
          u = mt(t.view)
            .on(
              'mousemove.zoom',
              function (t) {
                if ((Gi(t), !s.moved)) {
                  var n = t.clientX - l,
                    i = t.clientY - c;
                  s.moved = n * n + i * i > d;
                }
                s.event(t).zoom('mouse', o(m(s.that.__zoom, (s.mouse[0] = wt(t, e)), s.mouse[1]), s.extent, h));
              },
              !0,
            )
            .on(
              'mouseup.zoom',
              function (t) {
                (u.on('mousemove.zoom mouseup.zoom', null),
                  (function (t, n) {
                    var i = t.document.documentElement,
                      r = mt(t).on('dragstart.drag', null);
                    (n &&
                      (r.on('click.drag', Jt, It),
                      setTimeout(function () {
                        r.on('click.drag', null);
                      }, 0)),
                      'onselectstart' in i ? r.on('selectstart.drag', null) : ((i.style.MozUserSelect = i.__noselect), delete i.__noselect));
                  })(t.view, s.moved),
                  Gi(t),
                  s.event(t).end());
              },
              !0,
            ),
          a = wt(t, e),
          l = t.clientX,
          c = t.clientY;
        (!(function (t) {
          var n = t.document.documentElement,
            i = mt(t).on('dragstart.drag', Jt, It);
          'onselectstart' in n ? i.on('selectstart.drag', Jt, It) : ((n.__noselect = n.style.MozUserSelect), (n.style.MozUserSelect = 'none'));
        })(t.view),
          Zi(t),
          (s.mouse = [a, this.__zoom.invert(a)]),
          pi(this),
          s.start());
      }
    }
    function k(t, ...n) {
      if (r.apply(this, arguments)) {
        var i = this.__zoom,
          s = wt(t.changedTouches ? t.changedTouches[0] : t, this),
          u = i.invert(s),
          a = o(m(g(i, i.k * (t.shiftKey ? 0.5 : 2)), s, u), e.apply(this, n), h);
        (Gi(t), l > 0 ? mt(this).transition().duration(l).call(y, a, s, t) : mt(this).call(v.transform, a, s, t));
      }
    }
    function M(i, ...e) {
      if (r.apply(this, arguments)) {
        var o,
          s,
          u,
          a,
          h = i.touches,
          l = h.length,
          c = x(this, e, i.changedTouches.length === l).event(i);
        for (Zi(i), s = 0; s < l; ++s)
          ((a = [(a = wt((u = h[s]), this)), this.__zoom.invert(a), u.identifier]),
            c.touch0 ? c.touch1 || c.touch0[2] === a[2] || ((c.touch1 = a), (c.taps = 0)) : ((c.touch0 = a), (o = !0), (c.taps = 1 + !!t)));
        (t && (t = clearTimeout(t)),
          o &&
            (c.taps < 2 &&
              ((n = a[0]),
              (t = setTimeout(function () {
                t = null;
              }, 500))),
            pi(this),
            c.start()));
      }
    }
    function E(t, ...n) {
      if (this.__zooming) {
        var i,
          r,
          e,
          s,
          u = x(this, n).event(t),
          a = t.changedTouches,
          l = a.length;
        for (Gi(t), i = 0; i < l; ++i)
          ((e = wt((r = a[i]), this)), u.touch0 && u.touch0[2] === r.identifier ? (u.touch0[0] = e) : u.touch1 && u.touch1[2] === r.identifier && (u.touch1[0] = e));
        if (((r = u.that.__zoom), u.touch1)) {
          var c = u.touch0[0],
            f = u.touch0[1],
            d = u.touch1[0],
            p = u.touch1[1],
            v = (v = d[0] - c[0]) * v + (v = d[1] - c[1]) * v,
            w = (w = p[0] - f[0]) * w + (w = p[1] - f[1]) * w;
          ((r = g(r, Math.sqrt(v / w))), (e = [(c[0] + d[0]) / 2, (c[1] + d[1]) / 2]), (s = [(f[0] + p[0]) / 2, (f[1] + p[1]) / 2]));
        } else {
          if (!u.touch0) return;
          ((e = u.touch0[0]), (s = u.touch0[1]));
        }
        u.zoom('touch', o(m(r, e, s), u.extent, h));
      }
    }
    function L(t, ...r) {
      if (this.__zooming) {
        var e,
          o,
          s = x(this, r).event(t),
          u = t.changedTouches,
          a = u.length;
        for (
          Zi(t),
            i && clearTimeout(i),
            i = setTimeout(function () {
              i = null;
            }, 500),
            e = 0;
          e < a;
          ++e
        )
          ((o = u[e]), s.touch0 && s.touch0[2] === o.identifier ? delete s.touch0 : s.touch1 && s.touch1[2] === o.identifier && delete s.touch1);
        if ((s.touch1 && !s.touch0 && ((s.touch0 = s.touch1), delete s.touch1), s.touch0)) s.touch0[1] = this.__zoom.invert(s.touch0[0]);
        else if ((s.end(), 2 === s.taps && ((o = wt(o, this)), Math.hypot(n[0] - o[0], n[1] - o[1]) < p))) {
          var h = mt(this).on('dblclick.zoom');
          h && h.apply(this, arguments);
        }
      }
    }
    return (
      (v.transform = function (t, n, i, r) {
        var e = t.selection ? t.selection() : t;
        (e.property('__zoom', Pi),
          t !== e
            ? y(t, n, i, r)
            : e.interrupt().each(function () {
                x(this, arguments)
                  .event(r)
                  .start()
                  .zoom(null, 'function' == typeof n ? n.apply(this, arguments) : n)
                  .end();
              }));
      }),
      (v.scaleBy = function (t, n, i, r) {
        v.scaleTo(
          t,
          function () {
            return this.__zoom.k * ('function' == typeof n ? n.apply(this, arguments) : n);
          },
          i,
          r,
        );
      }),
      (v.scaleTo = function (t, n, i, r) {
        v.transform(
          t,
          function () {
            var t = e.apply(this, arguments),
              r = this.__zoom,
              s = null == i ? w(t) : 'function' == typeof i ? i.apply(this, arguments) : i,
              u = r.invert(s),
              a = 'function' == typeof n ? n.apply(this, arguments) : n;
            return o(m(g(r, a), s, u), t, h);
          },
          i,
          r,
        );
      }),
      (v.translateBy = function (t, n, i, r) {
        v.transform(
          t,
          function () {
            return o(
              this.__zoom.translate('function' == typeof n ? n.apply(this, arguments) : n, 'function' == typeof i ? i.apply(this, arguments) : i),
              e.apply(this, arguments),
              h,
            );
          },
          null,
          r,
        );
      }),
      (v.translateTo = function (t, n, i, r, s) {
        v.transform(
          t,
          function () {
            var t = e.apply(this, arguments),
              s = this.__zoom,
              u = null == r ? w(t) : 'function' == typeof r ? r.apply(this, arguments) : r;
            return o(
              Di.translate(u[0], u[1])
                .scale(s.k)
                .translate('function' == typeof n ? -n.apply(this, arguments) : -n, 'function' == typeof i ? -i.apply(this, arguments) : -i),
              t,
              h,
            );
          },
          r,
          s,
        );
      }),
      (C.prototype = {
        event: function (t) {
          return (t && (this.sourceEvent = t), this);
        },
        start: function () {
          return (1 == ++this.active && ((this.that.__zooming = this), this.emit('start')), this);
        },
        zoom: function (t, n) {
          return (
            this.mouse && 'mouse' !== t && (this.mouse[1] = n.invert(this.mouse[0])),
            this.touch0 && 'touch' !== t && (this.touch0[1] = n.invert(this.touch0[0])),
            this.touch1 && 'touch' !== t && (this.touch1[1] = n.invert(this.touch1[0])),
            (this.that.__zoom = n),
            this.emit('zoom'),
            this
          );
        },
        end: function () {
          return (0 == --this.active && (delete this.that.__zooming, this.emit('end')), this);
        },
        emit: function (t) {
          var n = mt(this.that).datum();
          f.call(t, this.that, new Fi(t, { sourceEvent: this.sourceEvent, target: v, transform: this.that.__zoom, dispatch: f }), n);
        },
      }),
      (v.wheelDelta = function (t) {
        return arguments.length ? ((s = 'function' == typeof t ? t : ji(+t)), v) : s;
      }),
      (v.filter = function (t) {
        return arguments.length ? ((r = 'function' == typeof t ? t : ji(!!t)), v) : r;
      }),
      (v.touchable = function (t) {
        return arguments.length ? ((u = 'function' == typeof t ? t : ji(!!t)), v) : u;
      }),
      (v.extent = function (t) {
        return arguments.length
          ? ((e =
              'function' == typeof t
                ? t
                : ji([
                    [+t[0][0], +t[0][1]],
                    [+t[1][0], +t[1][1]],
                  ])),
            v)
          : e;
      }),
      (v.scaleExtent = function (t) {
        return arguments.length ? ((a[0] = +t[0]), (a[1] = +t[1]), v) : [a[0], a[1]];
      }),
      (v.translateExtent = function (t) {
        return arguments.length
          ? ((h[0][0] = +t[0][0]), (h[1][0] = +t[1][0]), (h[0][1] = +t[0][1]), (h[1][1] = +t[1][1]), v)
          : [
              [h[0][0], h[0][1]],
              [h[1][0], h[1][1]],
            ];
      }),
      (v.constrain = function (t) {
        return arguments.length ? ((o = t), v) : o;
      }),
      (v.duration = function (t) {
        return arguments.length ? ((l = +t), v) : l;
      }),
      (v.interpolate = function (t) {
        return arguments.length ? ((c = t), v) : c;
      }),
      (v.on = function () {
        var t = f.on.apply(f, arguments);
        return t === f ? v : t;
      }),
      (v.clickDistance = function (t) {
        return arguments.length ? ((d = (t = +t) * t), v) : Math.sqrt(d);
      }),
      (v.tapDistance = function (t) {
        return arguments.length ? ((p = +t), v) : p;
      }),
      v
    );
  },
  zoomIdentity: Di,
  linkHorizontal: function () {
    return (function (t) {
      let n = Nr,
        i = Xr,
        r = Er,
        e = Lr,
        o = null,
        s = null,
        u = (function (t) {
          let n = 3;
          return (
            (t.digits = function (i) {
              if (!arguments.length) return n;
              if (null == i) n = null;
              else {
                const t = Math.floor(i);
                if (!(t >= 0)) throw new RangeError('invalid digits: ' + i);
                n = t;
              }
              return t;
            }),
            () => new kr(n)
          );
        })(a);
      function a() {
        let a;
        const h = Mr.call(arguments),
          l = n.apply(this, h),
          c = i.apply(this, h);
        if (
          (null == o && (s = t((a = u()))),
          s.lineStart(),
          (h[0] = l),
          s.point(+r.apply(this, h), +e.apply(this, h)),
          (h[0] = c),
          s.point(+r.apply(this, h), +e.apply(this, h)),
          s.lineEnd(),
          a)
        )
          return ((s = null), a + '' || null);
      }
      return (
        (a.source = function (t) {
          return arguments.length ? ((n = t), a) : n;
        }),
        (a.target = function (t) {
          return arguments.length ? ((i = t), a) : i;
        }),
        (a.x = function (t) {
          return arguments.length ? ((r = 'function' == typeof t ? t : wr(+t)), a) : r;
        }),
        (a.y = function (t) {
          return arguments.length ? ((e = 'function' == typeof t ? t : wr(+t)), a) : e;
        }),
        (a.context = function (n) {
          return arguments.length ? (null == n ? (o = s = null) : (s = t((o = n))), a) : o;
        }),
        a
      );
    })(Ar);
  },
  flextree: or,
};
class Br {
  constructor() {
    const t = {
      id: 'ID' + Math.floor(1e6 * Math.random()),
      firstDraw: !0,
      ctx: document.createElement('canvas').getContext('2d'),
      initialExpandLevel: 1,
      nodeDefaultBackground: 'none',
      lastTransform: { x: 0, y: 0, k: 1 },
      allowedNodesCount: {},
      zoomBehavior: null,
      generateRoot: null,
      svgWidth: 800,
      svgHeight: window.innerHeight - 100,
      container: 'body',
      data: null,
      connections: [],
      defaultFont: 'Helvetica',
      nodeId: t => t.nodeId || t.id,
      parentNodeId: t => t.parentNodeId || t.parentId,
      rootMargin: 40,
      nodeWidth: () => 250,
      nodeHeight: () => 150,
      neighbourMargin: () => 80,
      siblingsMargin: () => 20,
      childrenMargin: () => 60,
      compactMarginPair: () => 100,
      compactMarginBetween: () => 20,
      nodeButtonWidth: () => 40,
      nodeButtonHeight: () => 40,
      nodeButtonX: () => -20,
      nodeButtonY: () => -20,
      linkYOffset: 30,
      pagingStep: () => 5,
      minPagingVisibleNodes: () => 2e3,
      scaleExtent: [0.001, 20],
      duration: 400,
      imageName: 'Chart',
      setActiveNodeCentered: !0,
      layout: 'top',
      compact: !0,
      createZoom: () => Tr.zoom(),
      onZoomStart: () => {},
      onZoom: () => {},
      onZoomEnd: () => {},
      onNodeClick: t => t,
      onExpandOrCollapse: t => t,
      nodeContent: t =>
        `<div style="padding:5px;font-size:10px;">Sample Node(id=${t.id}), override using <br/> \n            <code>chart.nodeContent({data}=>{ <br/>\n             &nbsp;&nbsp;&nbsp;&nbsp;return '' // Custom HTML <br/>\n             })</code>\n             <br/> \n             Or check different <a href="https://github.com/bumbeishvili/org-chart#jump-to-examples" target="_blank">layout examples</a>\n             </div>`,
      buttonContent: ({ node: t, state: n }) =>
        `<div style="border:1px solid #E4E2E9;border-radius:3px;padding:3px;font-size:9px;margin:auto auto;background-color:white"> ${{ left: n => (n ? `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                      <path d="M14.283 3.50094L6.51 11.4749C6.37348 11.615 6.29707 11.8029 6.29707 11.9984C6.29707 12.194 6.37348 12.3819 6.51 12.5219L14.283 20.4989C14.3466 20.5643 14.4226 20.6162 14.5066 20.6516C14.5906 20.6871 14.6808 20.7053 14.772 20.7053C14.8632 20.7053 14.9534 20.6871 15.0374 20.6516C15.1214 20.6162 15.1974 20.5643 15.261 20.4989C15.3918 20.365 15.4651 20.1852 15.4651 19.9979C15.4651 19.8107 15.3918 19.6309 15.261 19.4969L7.9515 11.9984L15.261 4.50144C15.3914 4.36756 15.4643 4.18807 15.4643 4.00119C15.4643 3.81431 15.3914 3.63482 15.261 3.50094C15.1974 3.43563 15.1214 3.38371 15.0374 3.34827C14.9534 3.31282 14.8632 3.29456 14.772 3.29456C14.6808 3.29456 14.5906 3.31282 14.5066 3.34827C14.4226 3.38371 14.3466 3.43563 14.283 3.50094V3.50094Z" fill="#716E7B" stroke="#716E7B"/>\n                      </svg></span><span style="color:#716E7B">${t.data._directSubordinatesPaging} </span></div>` : `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                          <path d="M7.989 3.49944C7.85817 3.63339 7.78492 3.8132 7.78492 4.00044C7.78492 4.18768 7.85817 4.36749 7.989 4.50144L15.2985 11.9999L7.989 19.4969C7.85817 19.6309 7.78492 19.8107 7.78492 19.9979C7.78492 20.1852 7.85817 20.365 7.989 20.4989C8.05259 20.5643 8.12863 20.6162 8.21261 20.6516C8.2966 20.6871 8.38684 20.7053 8.478 20.7053C8.56916 20.7053 8.6594 20.6871 8.74338 20.6516C8.82737 20.6162 8.90341 20.5643 8.967 20.4989L16.74 12.5234C16.8765 12.3834 16.9529 12.1955 16.9529 11.9999C16.9529 11.8044 16.8765 11.6165 16.74 11.4764L8.967 3.50094C8.90341 3.43563 8.82737 3.38371 8.74338 3.34827C8.6594 3.31282 8.56916 3.29456 8.478 3.29456C8.38684 3.29456 8.2966 3.31282 8.21261 3.34827C8.12863 3.38371 8.05259 3.43563 7.989 3.50094V3.49944Z" fill="#716E7B" stroke="#716E7B"/>\n                          </svg></span><span style="color:#716E7B">${t.data._directSubordinatesPaging} </span></div>`), bottom: n => (n ? `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                       <path d="M19.497 7.98903L12 15.297L4.503 7.98903C4.36905 7.85819 4.18924 7.78495 4.002 7.78495C3.81476 7.78495 3.63495 7.85819 3.501 7.98903C3.43614 8.05257 3.38462 8.12842 3.34944 8.21213C3.31427 8.29584 3.29615 8.38573 3.29615 8.47653C3.29615 8.56733 3.31427 8.65721 3.34944 8.74092C3.38462 8.82463 3.43614 8.90048 3.501 8.96403L11.4765 16.74C11.6166 16.8765 11.8044 16.953 12 16.953C12.1956 16.953 12.3834 16.8765 12.5235 16.74L20.499 8.96553C20.5643 8.90193 20.6162 8.8259 20.6517 8.74191C20.6871 8.65792 20.7054 8.56769 20.7054 8.47653C20.7054 8.38537 20.6871 8.29513 20.6517 8.21114C20.6162 8.12715 20.5643 8.05112 20.499 7.98753C20.3651 7.85669 20.1852 7.78345 19.998 7.78345C19.8108 7.78345 19.6309 7.85669 19.497 7.98753V7.98903Z" fill="#716E7B" stroke="#716E7B"/>\n                       </svg></span><span style="margin-left:1px;color:#716E7B" >${t.data._directSubordinatesPaging} </span></div>\n                       ` : `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                       <path d="M11.457 8.07005L3.49199 16.4296C3.35903 16.569 3.28485 16.7543 3.28485 16.9471C3.28485 17.1398 3.35903 17.3251 3.49199 17.4646L3.50099 17.4736C3.56545 17.5414 3.64304 17.5954 3.72904 17.6324C3.81504 17.6693 3.90765 17.6883 4.00124 17.6883C4.09483 17.6883 4.18745 17.6693 4.27344 17.6324C4.35944 17.5954 4.43703 17.5414 4.50149 17.4736L12.0015 9.60155L19.4985 17.4736C19.563 17.5414 19.6405 17.5954 19.7265 17.6324C19.8125 17.6693 19.9052 17.6883 19.9987 17.6883C20.0923 17.6883 20.1849 17.6693 20.2709 17.6324C20.3569 17.5954 20.4345 17.5414 20.499 17.4736L20.508 17.4646C20.641 17.3251 20.7151 17.1398 20.7151 16.9471C20.7151 16.7543 20.641 16.569 20.508 16.4296L12.543 8.07005C12.4729 7.99653 12.3887 7.93801 12.2954 7.89801C12.202 7.85802 12.1015 7.8374 12 7.8374C11.8984 7.8374 11.798 7.85802 11.7046 7.89801C11.6113 7.93801 11.527 7.99653 11.457 8.07005Z" fill="#716E7B" stroke="#716E7B"/>\n                       </svg></span><span style="margin-left:1px;color:#716E7B" >${t.data._directSubordinatesPaging} </span></div>\n                    `), right: n => (n ? `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                       <path d="M7.989 3.49944C7.85817 3.63339 7.78492 3.8132 7.78492 4.00044C7.78492 4.18768 7.85817 4.36749 7.989 4.50144L15.2985 11.9999L7.989 19.4969C7.85817 19.6309 7.78492 19.8107 7.78492 19.9979C7.78492 20.1852 7.85817 20.365 7.989 20.4989C8.05259 20.5643 8.12863 20.6162 8.21261 20.6516C8.2966 20.6871 8.38684 20.7053 8.478 20.7053C8.56916 20.7053 8.6594 20.6871 8.74338 20.6516C8.82737 20.6162 8.90341 20.5643 8.967 20.4989L16.74 12.5234C16.8765 12.3834 16.9529 12.1955 16.9529 11.9999C16.9529 11.8044 16.8765 11.6165 16.74 11.4764L8.967 3.50094C8.90341 3.43563 8.82737 3.38371 8.74338 3.34827C8.6594 3.31282 8.56916 3.29456 8.478 3.29456C8.38684 3.29456 8.2966 3.31282 8.21261 3.34827C8.12863 3.38371 8.05259 3.43563 7.989 3.50094V3.49944Z" fill="#716E7B" stroke="#716E7B"/>\n                       </svg></span><span style="color:#716E7B">${t.data._directSubordinatesPaging} </span></div>` : `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                       <path d="M14.283 3.50094L6.51 11.4749C6.37348 11.615 6.29707 11.8029 6.29707 11.9984C6.29707 12.194 6.37348 12.3819 6.51 12.5219L14.283 20.4989C14.3466 20.5643 14.4226 20.6162 14.5066 20.6516C14.5906 20.6871 14.6808 20.7053 14.772 20.7053C14.8632 20.7053 14.9534 20.6871 15.0374 20.6516C15.1214 20.6162 15.1974 20.5643 15.261 20.4989C15.3918 20.365 15.4651 20.1852 15.4651 19.9979C15.4651 19.8107 15.3918 19.6309 15.261 19.4969L7.9515 11.9984L15.261 4.50144C15.3914 4.36756 15.4643 4.18807 15.4643 4.00119C15.4643 3.81431 15.3914 3.63482 15.261 3.50094C15.1974 3.43563 15.1214 3.38371 15.0374 3.34827C14.9534 3.31282 14.8632 3.29456 14.772 3.29456C14.6808 3.29456 14.5906 3.31282 14.5066 3.34827C14.4226 3.38371 14.3466 3.43563 14.283 3.50094V3.50094Z" fill="#716E7B" stroke="#716E7B"/>\n                       </svg></span><span style="color:#716E7B">${t.data._directSubordinatesPaging} </span></div>`), top: n => (n ? `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                        <path d="M11.457 8.07005L3.49199 16.4296C3.35903 16.569 3.28485 16.7543 3.28485 16.9471C3.28485 17.1398 3.35903 17.3251 3.49199 17.4646L3.50099 17.4736C3.56545 17.5414 3.64304 17.5954 3.72904 17.6324C3.81504 17.6693 3.90765 17.6883 4.00124 17.6883C4.09483 17.6883 4.18745 17.6693 4.27344 17.6324C4.35944 17.5954 4.43703 17.5414 4.50149 17.4736L12.0015 9.60155L19.4985 17.4736C19.563 17.5414 19.6405 17.5954 19.7265 17.6324C19.8125 17.6693 19.9052 17.6883 19.9987 17.6883C20.0923 17.6883 20.1849 17.6693 20.2709 17.6324C20.3569 17.5954 20.4345 17.5414 20.499 17.4736L20.508 17.4646C20.641 17.3251 20.7151 17.1398 20.7151 16.9471C20.7151 16.7543 20.641 16.569 20.508 16.4296L12.543 8.07005C12.4729 7.99653 12.3887 7.93801 12.2954 7.89801C12.202 7.85802 12.1015 7.8374 12 7.8374C11.8984 7.8374 11.798 7.85802 11.7046 7.89801C11.6113 7.93801 11.527 7.99653 11.457 8.07005Z" fill="#716E7B" stroke="#716E7B"/>\n                        </svg></span><span style="margin-left:1px;color:#716E7B">${t.data._directSubordinatesPaging} </span></div>\n                        ` : `<div style="display:flex;"><span style="align-items:center;display:flex;"><svg width="8" height="8" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                        <path d="M19.497 7.98903L12 15.297L4.503 7.98903C4.36905 7.85819 4.18924 7.78495 4.002 7.78495C3.81476 7.78495 3.63495 7.85819 3.501 7.98903C3.43614 8.05257 3.38462 8.12842 3.34944 8.21213C3.31427 8.29584 3.29615 8.38573 3.29615 8.47653C3.29615 8.56733 3.31427 8.65721 3.34944 8.74092C3.38462 8.82463 3.43614 8.90048 3.501 8.96403L11.4765 16.74C11.6166 16.8765 11.8044 16.953 12 16.953C12.1956 16.953 12.3834 16.8765 12.5235 16.74L20.499 8.96553C20.5643 8.90193 20.6162 8.8259 20.6517 8.74191C20.6871 8.65792 20.7054 8.56769 20.7054 8.47653C20.7054 8.38537 20.6871 8.29513 20.6517 8.21114C20.6162 8.12715 20.5643 8.05112 20.499 7.98753C20.3651 7.85669 20.1852 7.78345 19.998 7.78345C19.8108 7.78345 19.6309 7.85669 19.497 7.98753V7.98903Z" fill="#716E7B" stroke="#716E7B"/>\n                        </svg></span><span style="margin-left:1px;color:#716E7B">${t.data._directSubordinatesPaging} </span></div>\n                    `) }[n.layout](t.children)}  </div>`,
      pagingButton: (t, n, i, r) => {
        const e = r.pagingStep(t.parent);
        return `\n                   <div style="margin-top:90px;">\n                      <div style="display:flex;width:170px;border-radius:20px;padding:5px 15px; padding-bottom:4px;;background-color:#E5E9F2">\n                      <div><svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">\n                      <path d="M5.59 7.41L10.18 12L5.59 16.59L7 18L13 12L7 6L5.59 7.41ZM16 6H18V18H16V6Z" fill="#716E7B" stroke="#716E7B"/>\n                      </svg>\n                      </div><div style="line-height:2"> Show next ${Math.min(t.parent.data._directSubordinatesPaging - t.parent.data._pagingStep, e)}  nodes </div></div>\n                   </div>\n                `;
      },
      nodeUpdate: function (t) {
        Tr.select(this)
          .select('.node-rect')
          .attr('stroke', t => (t.data._highlighted || t.data._upToTheRootHighlighted ? '#E27396' : 'none'))
          .attr('stroke-width', t.data._highlighted || t.data._upToTheRootHighlighted ? 10 : 1);
      },
      nodeEnter: t => t,
      nodeExit: t => t,
      linkUpdate: function (t) {
        (Tr.select(this)
          .attr('stroke', t => (t.data._upToTheRootHighlighted ? '#E27396' : '#E4E2E9'))
          .attr('stroke-width', t => (t.data._upToTheRootHighlighted ? 5 : 1)),
          t.data._upToTheRootHighlighted && Tr.select(this).raise());
      },
      hdiagonal: function (t, n, i) {
        const r = t.x,
          e = t.y,
          o = n.x,
          s = n.y;
        let u = i && null != i.x ? i.x : r,
          a = i && null != i.y ? i.y : e,
          h = o - r < 0 ? -1 : 1,
          l = s - e < 0 ? -1 : 1,
          c = Math.abs(o - r) / 2 < 35 ? Math.abs(o - r) / 2 : 35;
        c = Math.abs(s - e) / 2 < c ? Math.abs(s - e) / 2 : c;
        let f = Math.abs(o - r) / 2 - c;
        return `\n                          M ${u} ${a}\n                          L ${u} ${e}\n                          L ${r} ${e}\n                          L ${r + f * h} ${e}\n                          C ${r + f * h + c * h} ${e} \n                            ${r + f * h + c * h} ${e} \n                            ${r + f * h + c * h} ${e + c * l}\n                          L ${r + f * h + c * h} ${s - c * l} \n                          C ${r + f * h + c * h}  ${s} \n                            ${r + f * h + c * h}  ${s} \n                            ${o - f * h}  ${s}\n                          L ${o} ${s}\n               `;
      },
      diagonal: function (t, n, i, r = { sy: 0 }) {
        const e = t.x;
        let o = t.y;
        const s = n.x,
          u = n.y;
        let a = i && null != i.x ? i.x : e,
          h = i && null != i.y ? i.y : o,
          l = s - e < 0 ? -1 : 1,
          c = u - o < 0 ? -1 : 1;
        o += r.sy;
        let f = Math.abs(s - e) / 2 < 35 ? Math.abs(s - e) / 2 : 35;
        f = Math.abs(u - o) / 2 < f ? Math.abs(u - o) / 2 : f;
        let d = Math.abs(u - o) / 2 - f;
        return `\n                          M ${a} ${h}\n                          L ${e} ${h}\n                          L ${e} ${o}\n                          L ${e} ${o + d * c}\n                          C  ${e} ${o + d * c + f * c} ${e} ${o + d * c + f * c} ${e + f * l} ${o + d * c + f * c}\n                          L ${e + (Math.abs(s - e) - 2 * f) * l + f * l} ${o + d * c + f * c}\n                          C  ${s}  ${o + d * c + f * c} ${s}  ${o + d * c + f * c} ${s} ${u - d * c}\n                          L ${s} ${u}\n               `;
      },
      defs: function (t, n) {
        return `<defs>\n                    ${n
          .map(n => {
            const i = this.getTextWidth(n.label, { ctx: t.ctx, fontSize: 2, defaultFont: t.defaultFont });
            return `\n                       <marker id="${n.from + '_' + n.to}" refX="${n._source.x < n._target.x ? -7 : 7}" refY="5" markerWidth="500"  markerHeight="500"  orient="${n._source.x < n._target.x ? 'auto' : 'auto-start-reverse'}" >\n                       <rect rx=0.5 width=${n.label ? i + 3 : 0} height=3 y=1  fill="#E27396"></rect>\n                       <text font-size="2px" x=1 fill="white" y=3>${n.label || ''}</text>\n                       </marker>\n\n                       <marker id="arrow-${n.from + '_' + n.to}"  markerWidth="500"  markerHeight="500"  refY="2"  refX="1" orient="${n._source.x < n._target.x ? 'auto' : 'auto-start-reverse'}" >\n                       <path transform="translate(0)" d='M0,0 V4 L2,2 Z' fill='#E27396' />\n                       </marker>\n                    `;
          })
          .join('')}\n                    </defs>\n                    `;
      },
      connectionsUpdate: function () {
        Tr.select(this)
          .attr('stroke', () => '#E27396')
          .attr('stroke-linecap', 'round')
          .attr('stroke-width', () => '5')
          .attr('pointer-events', 'none')
          .attr('marker-start', t => `url(#${t.from + '_' + t.to})`)
          .attr('marker-end', t => `url(#arrow-${t.from + '_' + t.to})`);
      },
      linkGroupArc: Tr.linkHorizontal()
        .x(t => t.x)
        .y(t => t.y),
      layoutBindings: {
        left: {
          nodeLeftX: () => 0,
          nodeRightX: t => t.width,
          nodeTopY: t => -t.height / 2,
          nodeBottomY: t => t.height / 2,
          nodeJoinX: t => t.x + t.width,
          nodeJoinY: t => t.y - t.height / 2,
          linkJoinX: t => t.x + t.width,
          linkJoinY: t => t.y,
          linkX: t => t.x,
          linkY: t => t.y,
          linkCompactXStart: t => t.x + t.width / 2,
          linkCompactYStart: t => t.y + (t.compactEven ? t.height / 2 : -t.height / 2),
          compactLinkMidX: t => t.firstCompactNode.x,
          compactLinkMidY: (t, n) => t.firstCompactNode.y + t.firstCompactNode.flexCompactDim[0] / 4 + n.compactMarginPair(t) / 4,
          linkParentX: t => t.parent.x + t.parent.width,
          linkParentY: t => t.parent.y,
          buttonX: t => t.width,
          buttonY: t => t.height / 2,
          centerTransform: ({ rootMargin: t, centerY: n, scale: i }) => `translate(${t},${n}) scale(${i})`,
          compactDimension: { sizeColumn: t => t.height, sizeRow: t => t.width, reverse: t => t.slice().reverse() },
          nodeFlexSize: ({ height: t, width: n, siblingsMargin: i, childrenMargin: r, state: e, node: o }) =>
            e.compact && o.flexCompactDim ? [o.flexCompactDim[0], o.flexCompactDim[1]] : [t + i, n + r],
          zoomTransform: ({ centerY: t, scale: n }) => `translate(0,${t}) scale(${n})`,
          diagonal: this.hdiagonal.bind(this),
          swap: t => {
            const n = t.x;
            ((t.x = t.y), (t.y = n));
          },
          nodeUpdateTransform: ({ x: t, y: n, height: i }) => `translate(${t},${n - i / 2})`,
        },
        top: {
          nodeLeftX: t => -t.width / 2,
          nodeRightX: t => t.width / 2,
          nodeTopY: () => 0,
          nodeBottomY: t => t.height,
          nodeJoinX: t => t.x - t.width / 2,
          nodeJoinY: t => t.y + t.height,
          linkJoinX: t => t.x,
          linkJoinY: t => t.y + t.height,
          linkCompactXStart: t => t.x + (t.compactEven ? t.width / 2 : -t.width / 2),
          linkCompactYStart: t => t.y + t.height / 2,
          compactLinkMidX: (t, n) => t.firstCompactNode.x + t.firstCompactNode.flexCompactDim[0] / 4 + n.compactMarginPair(t) / 4,
          compactLinkMidY: t => t.firstCompactNode.y,
          compactDimension: { sizeColumn: t => t.width, sizeRow: t => t.height, reverse: t => t },
          linkX: t => t.x,
          linkY: t => t.y,
          linkParentX: t => t.parent.x,
          linkParentY: t => t.parent.y + t.parent.height,
          buttonX: t => t.width / 2,
          buttonY: t => t.height,
          centerTransform: ({ rootMargin: t, scale: n, centerX: i }) => `translate(${i},${t}) scale(${n})`,
          nodeFlexSize: ({ height: t, width: n, siblingsMargin: i, childrenMargin: r, state: e, node: o }) =>
            e.compact && o.flexCompactDim ? [o.flexCompactDim[0], o.flexCompactDim[1]] : [n + i, t + r],
          zoomTransform: ({ centerX: t, scale: n }) => `translate(${t},0}) scale(${n})`,
          diagonal: this.diagonal.bind(this),
          swap: () => {},
          nodeUpdateTransform: ({ x: t, y: n, width: i }) => `translate(${t - i / 2},${n})`,
        },
        bottom: {
          nodeLeftX: t => -t.width / 2,
          nodeRightX: t => t.width / 2,
          nodeTopY: t => -t.height,
          nodeBottomY: () => 0,
          nodeJoinX: t => t.x - t.width / 2,
          nodeJoinY: t => t.y - t.height - t.height,
          linkJoinX: t => t.x,
          linkJoinY: t => t.y - t.height,
          linkCompactXStart: t => t.x + (t.compactEven ? t.width / 2 : -t.width / 2),
          linkCompactYStart: t => t.y - t.height / 2,
          compactLinkMidX: (t, n) => t.firstCompactNode.x + t.firstCompactNode.flexCompactDim[0] / 4 + n.compactMarginPair(t) / 4,
          compactLinkMidY: t => t.firstCompactNode.y,
          linkX: t => t.x,
          linkY: t => t.y,
          compactDimension: { sizeColumn: t => t.width, sizeRow: t => t.height, reverse: t => t },
          linkParentX: t => t.parent.x,
          linkParentY: t => t.parent.y - t.parent.height,
          buttonX: t => t.width / 2,
          buttonY: () => 0,
          centerTransform: ({ rootMargin: t, scale: n, centerX: i, chartHeight: r }) => `translate(${i},${r - t}) scale(${n})`,
          nodeFlexSize: ({ height: t, width: n, siblingsMargin: i, childrenMargin: r, state: e, node: o }) =>
            e.compact && o.flexCompactDim ? [o.flexCompactDim[0], o.flexCompactDim[1]] : [n + i, t + r],
          zoomTransform: ({ centerX: t, scale: n }) => `translate(${t},0}) scale(${n})`,
          diagonal: this.diagonal.bind(this),
          swap: t => {
            t.y = -t.y;
          },
          nodeUpdateTransform: ({ x: t, y: n, width: i, height: r }) => `translate(${t - i / 2},${n - r})`,
        },
        right: {
          nodeLeftX: t => -t.width,
          nodeRightX: () => 0,
          nodeTopY: t => -t.height / 2,
          nodeBottomY: t => t.height / 2,
          nodeJoinX: t => t.x - t.width - t.width,
          nodeJoinY: t => t.y - t.height / 2,
          linkJoinX: t => t.x - t.width,
          linkJoinY: t => t.y,
          linkX: t => t.x,
          linkY: t => t.y,
          linkParentX: t => t.parent.x - t.parent.width,
          linkParentY: t => t.parent.y,
          buttonX: () => 0,
          buttonY: t => t.height / 2,
          linkCompactXStart: t => t.x - t.width / 2,
          linkCompactYStart: t => t.y + (t.compactEven ? t.height / 2 : -t.height / 2),
          compactLinkMidX: t => t.firstCompactNode.x,
          compactLinkMidY: (t, n) => t.firstCompactNode.y + t.firstCompactNode.flexCompactDim[0] / 4 + n.compactMarginPair(t) / 4,
          centerTransform: ({ rootMargin: t, centerY: n, scale: i, chartWidth: r }) => `translate(${r - t},${n}) scale(${i})`,
          nodeFlexSize: ({ height: t, width: n, siblingsMargin: i, childrenMargin: r, state: e, node: o }) =>
            e.compact && o.flexCompactDim ? [o.flexCompactDim[0], o.flexCompactDim[1]] : [t + i, n + r],
          compactDimension: { sizeColumn: t => t.height, sizeRow: t => t.width, reverse: t => t.slice().reverse() },
          zoomTransform: ({ centerY: t, scale: n }) => `translate(0,${t}) scale(${n})`,
          diagonal: this.hdiagonal.bind(this),
          swap: t => {
            const n = t.x;
            ((t.x = -t.y), (t.y = n));
          },
          nodeUpdateTransform: ({ x: t, y: n, width: i, height: r }) => `translate(${t - i},${n - r / 2})`,
        },
      },
    };
    ((this.getChartState = () => t),
      Object.keys(t).forEach(n => {
        this[n] = function (i) {
          return arguments.length ? ((t[n] = i), this) : t[n];
        };
      }),
      this.initializeEnterExitUpdatePattern());
  }
  initializeEnterExitUpdatePattern() {
    Tr.selection.prototype.patternify = function (t) {
      var n = t.selector,
        i = t.tag,
        r = t.data || [n],
        e = this.selectAll('.' + n).data(r, (t, n) => ('object' == typeof t && t.id ? t.id : n));
      return (e.exit().remove(), (e = e.enter().append(i).merge(e)).attr('class', n), e);
    };
  }
  getNodeChildren({ data: t, children: n, _children: i }, r) {
    return (
      r.push(t),
      n &&
        n.forEach(t => {
          this.getNodeChildren(t, r);
        }),
      i &&
        i.forEach(t => {
          this.getNodeChildren(t, r);
        }),
      r
    );
  }
  initialZoom(t) {
    return ((this.getChartState().lastTransform.k = t), this);
  }
  render() {
    const t = this.getChartState();
    if (!t.data || 0 == t.data.length)
      return (
        console.log('ORG CHART - Data is empty'),
        t.container &&
          (mt(t.container).select('.nodes-wrapper').remove(), mt(t.container).select('.links-wrapper').remove(), mt(t.container).select('.connections-wrapper').remove()),
        this
      );
    const n = Tr.select(t.container),
      i = n.node().getBoundingClientRect();
    i.width > 0 && (t.svgWidth = i.width);
    const r = { id: 'ID' + Math.floor(1e6 * Math.random()), chartWidth: t.svgWidth, chartHeight: t.svgHeight };
    if (((t.calc = r), (r.centerX = r.chartWidth / 2), (r.centerY = r.chartHeight / 2), t.firstDraw)) {
      const n = { zoom: null };
      ((n.zoom = t
        .createZoom()
        .clickDistance(10)
        .on('start', n => t.onZoomStart(n))
        .on('end', n => t.onZoomEnd(n))
        .on('zoom', (n, i) => {
          (t.onZoom(n), this.zoomed(n, i));
        })
        .scaleExtent(t.scaleExtent)),
        (t.zoomBehavior = n.zoom));
    }
    ((t.flexTreeLayout = or({
      nodeSize: n => {
        const i = t.nodeWidth(n),
          r = t.nodeHeight(n),
          e = t.siblingsMargin(n),
          o = t.childrenMargin(n);
        return t.layoutBindings[t.layout].nodeFlexSize({ state: t, node: n, width: i, height: r, siblingsMargin: e, childrenMargin: o });
      },
    }).spacing((n, i) => (n.parent == i.parent ? 0 : t.neighbourMargin(n, i)))),
      this.setLayouts({ expandNodesFirst: !1 }));
    const e = n.patternify({ tag: 'svg', selector: 'svg-chart-container' }).attr('width', t.svgWidth).attr('height', t.svgHeight).attr('font-family', t.defaultFont);
    (t.firstDraw && e.call(t.zoomBehavior).on('dblclick.zoom', null).attr('cursor', 'move'), (t.svg = e));
    const o = e.patternify({ tag: 'g', selector: 'chart' });
    return (
      (t.centerG = o.patternify({ tag: 'g', selector: 'center-group' })),
      (t.linksWrapper = t.centerG.patternify({ tag: 'g', selector: 'links-wrapper' })),
      (t.nodesWrapper = t.centerG.patternify({ tag: 'g', selector: 'nodes-wrapper' })),
      (t.connectionsWrapper = t.centerG.patternify({ tag: 'g', selector: 'connections-wrapper' })),
      (t.defsWrapper = e.patternify({ tag: 'g', selector: 'defs-wrapper' })),
      t.firstDraw &&
        t.centerG.attr('transform', () =>
          t.layoutBindings[t.layout].centerTransform({
            centerX: r.centerX,
            centerY: r.centerY,
            scale: t.lastTransform.k,
            rootMargin: t.rootMargin,
            root: t.root,
            chartHeight: r.chartHeight,
            chartWidth: r.chartWidth,
          }),
        ),
      (t.chart = o),
      this.update(t.root),
      Tr.select(window).on('resize.' + t.id, () => {
        const n = Tr.select(t.container).node().getBoundingClientRect();
        t.svg.attr('width', n.width);
      }),
      t.firstDraw && (t.firstDraw = !1),
      this
    );
  }
  addNode(t) {
    const n = this.getChartState();
    if (t && (null == n.parentNodeId(t) || n.parentNodeId(t) == n.nodeId(t)) && 0 == n.data.length) return (n.data.push(t), this.render(), this);
    const i = n.generateRoot(n.data).descendants(),
      r = i.filter(({ data: i }) => '' + n.nodeId(i) == '' + n.nodeId(t))[0];
    return (
      i.filter(({ data: i }) => '' + n.nodeId(i) == '' + n.parentNodeId(t)),
      r
        ? (console.log(`ORG CHART - ADD - Node with id "${n.nodeId(t)}" already exists in tree`), this)
        : (t._centered && !t._expanded && (t._expanded = !0), n.data.push(t), this.updateNodesState(), this)
    );
  }
  removeNode(t) {
    const n = this.getChartState(),
      i = n
        .generateRoot(n.data)
        .descendants()
        .filter(({ data: i }) => n.nodeId(i) == t)[0];
    return i
      ? (i.descendants().forEach(t => (t.data._filteredOut = !0)),
        (n.data = n.data.filter(t => !t._filteredOut)),
        0 == n.data.length ? this.render() : this.updateNodesState.bind(this)(),
        this)
      : (console.log(`ORG CHART - REMOVE - Node with id "${t}" not found in the tree`), this);
  }
  groupBy(t, n, i) {
    const r = {};
    return (
      t.forEach(t => {
        const i = n(t);
        (r[i] || (r[i] = []), r[i].push(t));
      }),
      Object.keys(r).forEach(t => {
        r[t] = i(r[t]);
      }),
      Object.entries(r)
    );
  }
  calculateCompactFlexDimensions(t) {
    const n = this.getChartState();
    (t.eachBefore(t => {
      ((t.firstCompact = null), (t.compactEven = null), (t.flexCompactDim = null), (t.firstCompactNode = null));
    }),
      t.eachBefore(t => {
        if (t.children && t.children.length > 1) {
          const i = t.children.filter(t => !t.children);
          if (i.length < 2) return;
          i.forEach((t, n) => {
            (n || (t.firstCompact = !0), (t.compactEven = !(n % 2)), (t.row = Math.floor(n / 2)));
          });
          const r = Tr.max(
              i.filter(t => t.compactEven),
              n.layoutBindings[n.layout].compactDimension.sizeColumn,
            ),
            e = Tr.max(
              i.filter(t => !t.compactEven),
              n.layoutBindings[n.layout].compactDimension.sizeColumn,
            ),
            o = 2 * Math.max(r, e),
            s = this.groupBy(
              i,
              t => t.row,
              t => Tr.max(t, t => n.layoutBindings[n.layout].compactDimension.sizeRow(t) + n.compactMarginBetween(t)),
            ),
            u = Tr.sum(s.map(t => t[1]));
          (i.forEach(t => {
            ((t.firstCompactNode = i[0]), (t.flexCompactDim = t.firstCompact ? [o + n.compactMarginPair(t), u - n.compactMarginBetween(t)] : [0, 0]));
          }),
            (t.flexCompactDim = null));
        }
      }));
  }
  calculateCompactFlexPositions(t) {
    const n = this.getChartState();
    t.eachBefore(t => {
      if (t.children) {
        const i = t.children.filter(t => t.flexCompactDim),
          r = i[0];
        if (!r) return;
        i.forEach((t, i) => {
          (0 == i && (r.x -= r.flexCompactDim[0] / 2),
            i & ((i % 2) - 1) ? (t.x = r.x + 0.25 * r.flexCompactDim[0] - n.compactMarginPair(t) / 4) : i && (t.x = r.x + 0.75 * r.flexCompactDim[0] + n.compactMarginPair(t) / 4));
        });
        const e = r.x + 0.5 * r.flexCompactDim[0];
        r.x = r.x + 0.25 * r.flexCompactDim[0] - n.compactMarginPair(r) / 4;
        const o = t.x - e;
        Math.abs(o) < 10 && i.forEach(t => (t.x += o));
        const s = this.groupBy(
            i,
            t => t.row,
            t => Tr.max(t, t => n.layoutBindings[n.layout].compactDimension.sizeRow(t)),
          ),
          u = Tr.cumsum(s.map(t => t[1] + n.compactMarginBetween(t)));
        i.forEach(t => {
          t.y = t.row ? r.y + u[t.row - 1] : r.y;
        });
      }
    });
  }
  update({ x0: t, y0: n, x: i = 0, y: r = 0, width: e, height: o }) {
    const s = this.getChartState();
    s.compact && this.calculateCompactFlexDimensions(s.root);
    const u = s.flexTreeLayout(s.root);
    s.compact && this.calculateCompactFlexPositions(s.root);
    const a = u.descendants(),
      h = u.descendants().slice(1);
    a.forEach(s.layoutBindings[s.layout].swap);
    const l = s.connections,
      c = {};
    s.allNodes.forEach(t => (c[s.nodeId(t.data)] = t));
    const f = {};
    (a.forEach(t => (f[s.nodeId(t.data)] = t)),
      l.forEach(t => {
        const n = c[t.to];
        ((t._source = c[t.from]), (t._target = n));
      }));
    const d = l.filter(t => f[t.from] && f[t.to]),
      p = s.defs.bind(this)(s, d);
    p !== s.defsWrapper.html() && s.defsWrapper.html(p);
    const v = s.linksWrapper.selectAll('path.link').data(h, t => s.nodeId(t.data)),
      g = v
        .enter()
        .insert('path', 'g')
        .attr('class', 'link')
        .attr('d', () => {
          const i = { x: s.layoutBindings[s.layout].linkJoinX({ x: t, y: n, width: e, height: o }), y: s.layoutBindings[s.layout].linkJoinY({ x: t, y: n, width: e, height: o }) };
          return s.layoutBindings[s.layout].diagonal(i, i, i);
        })
        .merge(v);
    (g.attr('fill', 'none'),
      this.isEdge() ? g.style('display', t => (t.data._pagingButton ? 'none' : 'auto')) : g.attr('display', t => (t.data._pagingButton ? 'none' : 'auto')),
      g.each(s.linkUpdate),
      g
        .transition()
        .duration(s.duration)
        .attr('d', t => {
          const n =
              s.compact && t.flexCompactDim
                ? { x: s.layoutBindings[s.layout].compactLinkMidX(t, s), y: s.layoutBindings[s.layout].compactLinkMidY(t, s) }
                : { x: s.layoutBindings[s.layout].linkX(t), y: s.layoutBindings[s.layout].linkY(t) },
            i = { x: s.layoutBindings[s.layout].linkParentX(t), y: s.layoutBindings[s.layout].linkParentY(t) },
            r = s.compact && t.flexCompactDim ? { x: s.layoutBindings[s.layout].linkCompactXStart(t), y: s.layoutBindings[s.layout].linkCompactYStart(t) } : n;
          return s.layoutBindings[s.layout].diagonal(n, i, r, { sy: s.linkYOffset });
        }),
      v
        .exit()
        .transition()
        .duration(s.duration)
        .attr('d', () => {
          const t = { x: s.layoutBindings[s.layout].linkJoinX({ x: i, y: r, width: e, height: o }), y: s.layoutBindings[s.layout].linkJoinY({ x: i, y: r, width: e, height: o }) };
          return s.layoutBindings[s.layout].diagonal(t, t, null, { sy: s.linkYOffset });
        })
        .remove());
    const m = s.connectionsWrapper.selectAll('path.connection').data(d),
      w = m
        .enter()
        .insert('path', 'g')
        .attr('class', 'connection')
        .attr('d', () => {
          const i = { x: s.layoutBindings[s.layout].linkJoinX({ x: t, y: n, width: e, height: o }), y: s.layoutBindings[s.layout].linkJoinY({ x: t, y: n, width: e, height: o }) };
          return s.layoutBindings[s.layout].diagonal(i, i, null, { sy: s.linkYOffset });
        })
        .merge(m);
    (w.attr('fill', 'none'),
      w
        .transition()
        .duration(s.duration)
        .attr('d', t => {
          const n = s.layoutBindings[s.layout].linkX({ x: t._source.x, y: t._source.y, width: t._source.width, height: t._source.height }),
            i = s.layoutBindings[s.layout].linkY({ x: t._source.x, y: t._source.y, width: t._source.width, height: t._source.height }),
            r = s.layoutBindings[s.layout].linkJoinX({ x: t._target.x, y: t._target.y, width: t._target.width, height: t._target.height }),
            e = s.layoutBindings[s.layout].linkJoinY({ x: t._target.x, y: t._target.y, width: t._target.width, height: t._target.height });
          return s.linkGroupArc({ source: { x: n, y: i }, target: { x: r, y: e } });
        }),
      w.each(s.connectionsUpdate),
      m.exit().transition().duration(s.duration).attr('opacity', 0).remove());
    const y = s.nodesWrapper.selectAll('g.node').data(a, ({ data: t }) => s.nodeId(t)),
      x = y
        .enter()
        .append('g')
        .attr('class', 'node')
        .attr('transform', i =>
          i == s.root
            ? `translate(${t},${n})`
            : `translate(${s.layoutBindings[s.layout].nodeJoinX({ x: t, y: n, width: e, height: o })},${s.layoutBindings[s.layout].nodeJoinY({ x: t, y: n, width: e, height: o })})`,
        )
        .attr('cursor', 'pointer')
        .on('click.node', (t, n) => {
          const { data: i } = n;
          [...t.srcElement.classList].includes('node-button-foreign-object') ||
            ([...t.srcElement.classList].includes('paging-button-wrapper')
              ? this.loadPagingNodes(n)
              : i._pagingButton
                ? console.log('event fired, no handlers')
                : s.onNodeClick(n));
        })
        .on('keydown.node', (t, n) => {
          if ('Enter' === t.key || ' ' === t.key || 'Spacebar' === t.key) {
            if ([...t.srcElement.classList].includes('node-button-foreign-object')) return;
            if ([...t.srcElement.classList].includes('paging-button-wrapper')) return void this.loadPagingNodes(n);
            ('Enter' !== t.key && ' ' !== t.key && 'Spacebar' !== t.key) || this.onButtonClick(t, n);
          }
        });
    (x.each(s.nodeEnter), x.patternify({ tag: 'rect', selector: 'node-rect', data: t => [t] }));
    const C = x.merge(y).style('font', '12px sans-serif');
    (C.patternify({ tag: 'foreignObject', selector: 'node-foreign-object', data: t => [t] })
      .style('overflow', 'visible')
      .patternify({ tag: 'xhtml:div', selector: 'node-foreign-object-div', data: t => [t] }),
      this.restyleForeignObjectElements());
    const $ = x
      .patternify({ tag: 'g', selector: 'node-button-g', data: t => [t] })
      .on('click', (t, n) => this.onButtonClick(t, n))
      .on('keydown', (t, n) => {
        ('Enter' !== t.key && ' ' !== t.key && 'Spacebar' !== t.key) || this.onButtonClick(t, n);
      });
    ($.patternify({ tag: 'rect', selector: 'node-button-rect', data: t => [t] })
      .attr('opacity', 0)
      .attr('pointer-events', 'all')
      .attr('width', t => s.nodeButtonWidth(t))
      .attr('height', t => s.nodeButtonHeight(t))
      .attr('x', t => s.nodeButtonX(t))
      .attr('y', t => s.nodeButtonY(t)),
      $.patternify({ tag: 'foreignObject', selector: 'node-button-foreign-object', data: t => [t] })
        .attr('width', t => s.nodeButtonWidth(t))
        .attr('height', t => s.nodeButtonHeight(t))
        .attr('x', t => s.nodeButtonX(t))
        .attr('y', t => s.nodeButtonY(t))
        .style('overflow', 'visible')
        .patternify({ tag: 'xhtml:div', selector: 'node-button-div', data: t => [t] })
        .style('pointer-events', 'none')
        .style('display', 'flex')
        .style('width', '100%')
        .style('height', '100%'),
      C.transition()
        .attr('opacity', 0)
        .duration(s.duration)
        .attr('transform', ({ x: t, y: n, width: i, height: r }) => s.layoutBindings[s.layout].nodeUpdateTransform({ x: t, y: n, width: i, height: r }))
        .attr('opacity', 1),
      C.select('.node-rect')
        .attr('width', ({ width: t }) => t)
        .attr('height', ({ height: t }) => t)
        .attr('x', ({}) => 0)
        .attr('y', ({}) => 0)
        .attr('cursor', 'pointer')
        .attr('rx', 3)
        .attr('fill', s.nodeDefaultBackground),
      C.select('.node-button-g')
        .attr(
          'transform',
          ({ width: t, height: n }) => `translate(${s.layoutBindings[s.layout].buttonX({ width: t, height: n })},${s.layoutBindings[s.layout].buttonY({ width: t, height: n })})`,
        )
        .attr('display', ({ data: t }) => (t._directSubordinates > 0 ? null : 'none'))
        .attr('opacity', ({ data: t, children: n, _children: i }) => (t._pagingButton ? 0 : n || i ? 1 : 0)),
      C.select('.node-button-foreign-object .node-button-div').html(t => s.buttonContent({ node: t, state: s })),
      C.select('.node-button-text')
        .attr('text-anchor', 'middle')
        .attr('alignment-baseline', 'middle')
        .attr('font-size', ({ children: t }) => (t ? 40 : 26))
        .text(({ children: t }) => (t ? '-' : '+'))
        .attr('y', this.isEdge() ? 10 : 0),
      C.each(s.nodeUpdate));
    const b = y.exit();
    b.each(s.nodeExit);
    const k = b.data().reduce((t, n) => (t.depth < n.depth ? t : n), { depth: 1 / 0 });
    (b
      .attr('opacity', 1)
      .transition()
      .duration(s.duration)
      .attr('transform', () => {
        let { x: t, y: n, width: i, height: r } = k.parent || {};
        return `translate(${s.layoutBindings[s.layout].nodeJoinX({ x: t, y: n, width: i, height: r })},${s.layoutBindings[s.layout].nodeJoinY({ x: t, y: n, width: i, height: r })})`;
      })
      .on('end', function () {
        Tr.select(this).remove();
      })
      .attr('opacity', 0),
      a.forEach(t => {
        ((t.x0 = t.x), (t.y0 = t.y));
      }));
    const M = s.allNodes.filter(t => t.data._centered)[0];
    if (M) {
      let t = [M];
      (M.data._centeredWithDescendants &&
        (t = s.compact
          ? M.descendants().filter((t, n) => n < 7)
          : M.descendants().filter((t, n, i) => {
              const r = Math.round(i.length / 2);
              return i.length % 2 ? n > r - 2 && n < r + 2 - 1 : n > r - 2 && n < r + 2;
            })),
        (M.data._centeredWithDescendants = null),
        (M.data._centered = null),
        this.fit({ animate: !0, scale: !1, nodes: t }));
    }
  }
  isEdge() {
    return window.navigator.userAgent.includes('Edge');
  }
  hdiagonal(t, n, i, r) {
    return this.getChartState().hdiagonal(t, n, i, r);
  }
  diagonal(t, n, i, r) {
    return this.getChartState().diagonal(t, n, i, r);
  }
  restyleForeignObjectElements() {
    const t = this.getChartState();
    (t.svg
      .selectAll('.node-foreign-object')
      .attr('width', ({ width: t }) => t)
      .attr('height', ({ height: t }) => t)
      .attr('x', ({}) => 0)
      .attr('y', ({}) => 0),
      t.svg
        .selectAll('.node-foreign-object-div')
        .style('width', ({ width: t }) => t + 'px')
        .style('height', ({ height: t }) => t + 'px')
        .html(function (n, i, r) {
          return n.data._pagingButton
            ? `<div class="paging-button-wrapper"><div style="pointer-events:none">${t.pagingButton(n, i, r, t)}</div></div>`
            : t.nodeContent.bind(this)(n, i, r, t);
        }));
  }
  onButtonClick(t, n) {
    const i = this.getChartState();
    n.data._pagingButton ||
      (i.setActiveNodeCentered && ((n.data._centered = !0), (n.data._centeredWithDescendants = !0)),
      n.children
        ? ((n._children = n.children), (n.children = null), this.setExpansionFlagToChildren(n, !1))
        : ((n.children = n._children), (n._children = null), n.children && n.children.forEach(({ data: t }) => (t._expanded = !0))),
      this.update(n),
      t.stopPropagation(),
      i.onExpandOrCollapse(n));
  }
  setExpansionFlagToChildren({ data: t, children: n, _children: i }, r) {
    ((t._expanded = r),
      n &&
        n.forEach(t => {
          this.setExpansionFlagToChildren(t, r);
        }),
      i &&
        i.forEach(t => {
          this.setExpansionFlagToChildren(t, r);
        }));
  }
  expandSomeNodes(t) {
    if (t.data._expanded) {
      let n = t.parent;
      for (; n && n._children; ) ((n.children = n._children), (n._children = null), (n = n.parent));
    }
    (t._children && t._children.forEach(t => this.expandSomeNodes(t)), t.children && t.children.forEach(t => this.expandSomeNodes(t)));
  }
  updateNodesState() {
    const t = this.getChartState();
    (this.setLayouts({ expandNodesFirst: !0 }), this.update(t.root));
  }
  setLayouts({ expandNodesFirst: t = !0 }) {
    const n = this.getChartState();
    ((n.generateRoot = Tr.stratify()
      .id(t => n.nodeId(t))
      .parentId(t => n.parentNodeId(t))),
      (n.root = n.generateRoot(n.data)));
    const i = n.root.descendants();
    n.initialExpandLevel > 1 &&
      i.length > 0 &&
      (i.forEach(t => {
        t.depth <= n.initialExpandLevel && (t.data._expanded = !0);
      }),
      (n.initialExpandLevel = 1));
    const r = {};
    (n.root
      .descendants()
      .filter(t => t.children)
      .filter(t => !t.data._pagingStep)
      .forEach(t => {
        t.data._pagingStep = n.minPagingVisibleNodes(t);
      }),
      n.root.eachBefore(t => {
        ((t.data._directSubordinatesPaging = t.children ? t.children.length : 0),
          t.children &&
            t.children.forEach((n, i) => {
              if (
                ((n.data._pagingButton = !1),
                i > t.data._pagingStep && (r[n.id] = !0),
                i === t.data._pagingStep && t.children.length - 1 > t.data._pagingStep && (n.data._pagingButton = !0),
                r[n.parent.id] && (r[n.id] = !0),
                n.data._expanded || n.data._centered || n.data._highlighted || n.data._upToTheRootHighlighted)
              ) {
                let t = n;
                for (; t && (r[t.id] || t.data._pagingButton); )
                  ((r[t.id] = !1),
                    t.data._pagingButton &&
                      ((t.data._pagingButton = !1),
                      t.parent.children.forEach(t => {
                        ((t.data._expanded = !0), (r[t.id] = !1));
                      })),
                    (t = t.parent));
              }
            }));
      }),
      (n.root = Tr.stratify()
        .id(t => n.nodeId(t))
        .parentId(t => n.parentNodeId(t))(n.data.filter(t => !0 !== r[t.id]))),
      n.root.each(t => {
        let i = t._hierarchyHeight || t.height,
          r = n.nodeWidth(t),
          e = n.nodeHeight(t);
        Object.assign(t, { width: r, height: e, _hierarchyHeight: i });
      }),
      (n.root.x0 = 0),
      (n.root.y0 = 0),
      (n.allNodes = n.root.descendants()),
      n.allNodes.forEach(t => {
        Object.assign(t.data, { _directSubordinates: t.children ? t.children.length : 0, _totalSubordinates: t.descendants().length - 1 });
      }),
      n.root.children &&
        (t && n.root.children.forEach(this.expand),
        n.root.children.forEach(t => this.collapse(t)),
        0 == n.initialExpandLevel && ((n.root._children = n.root.children), (n.root.children = null)),
        [n.root].forEach(t => this.expandSomeNodes(t))));
  }
  collapse(t) {
    t.children && ((t._children = t.children), t._children.forEach(t => this.collapse(t)), (t.children = null));
  }
  expand(t) {
    t._children && ((t.children = t._children), t.children.forEach(t => this.expand(t)), (t._children = null));
  }
  zoomed(t, n) {
    const i = this.getChartState(),
      r = i.chart,
      e = t.transform;
    ((i.lastTransform = e), r.attr('transform', e), this.isEdge() && this.restyleForeignObjectElements());
  }
  zoomTreeBounds({ x0: t, x1: n, y0: i, y1: r, params: e = { animate: !0, scale: !0, onCompleted: () => {} } }) {
    const { centerG: o, svgWidth: s, svgHeight: u, svg: a, zoomBehavior: h, duration: l, lastTransform: c } = this.getChartState();
    let f = Math.min(8, 0.9 / Math.max((n - t) / s, (r - i) / u)),
      d = Tr.zoomIdentity.translate(s / 2, u / 2);
    ((d = d.scale(e.scale ? f : c.k)),
      (d = d.translate(-(t + n) / 2, -(i + r) / 2)),
      a
        .transition()
        .duration(e.animate ? l : 0)
        .call(h.transform, d),
      o
        .transition()
        .duration(e.animate ? l : 0)
        .attr('transform', 'translate(0,0)')
        .on('end', function () {
          e.onCompleted && e.onCompleted();
        }));
  }
  fit({ animate: t = !0, nodes: n, scale: i = !0, onCompleted: r = () => {} } = {}) {
    const e = this.getChartState(),
      { root: o } = e;
    let s = n || o.descendants();
    const u = Tr.min(s, t => t.x + e.layoutBindings[e.layout].nodeLeftX(t)),
      a = Tr.max(s, t => t.x + e.layoutBindings[e.layout].nodeRightX(t)),
      h = Tr.min(s, t => t.y + e.layoutBindings[e.layout].nodeTopY(t)),
      l = Tr.max(s, t => t.y + e.layoutBindings[e.layout].nodeBottomY(t));
    return (this.zoomTreeBounds({ params: { animate: t, scale: i, onCompleted: r }, x0: u - 50, x1: a + 50, y0: h - 50, y1: l + 50 }), this);
  }
  loadPagingNodes(t) {
    const n = this.getChartState();
    t.data._pagingButton = !1;
    const i = t.parent.data._pagingStep,
      r = n.pagingStep(t.parent);
    ((t.parent.data._pagingStep = i + r), this.updateNodesState());
  }
  setExpanded(t, n = !0) {
    const i = this.getChartState(),
      r = i.allNodes.filter(({ data: n }) => i.nodeId(n) == t)[0];
    if (!r) return (console.log(`ORG CHART - ${n ? 'EXPAND' : 'COLLAPSE'} - Node with id (${t})  not found in the tree`), this);
    if (((r.data._expanded = n), 0 == n)) {
      const t = r.parent || { descendants: () => [] };
      t.descendants()
        .filter(n => n != t)
        .forEach(t => (t.data._expanded = !1));
    }
    return this;
  }
  setCentered(t) {
    const n = this.getChartState(),
      i = n
        .generateRoot(n.data)
        .descendants()
        .filter(({ data: i }) => '' + n.nodeId(i) == '' + t)[0];
    return i
      ? (i.ancestors().forEach(t => (t.data._expanded = !0)), (i.data._centered = !0), (i.data._expanded = !0), this)
      : (console.log(`ORG CHART - CENTER - Node with id (${t}) not found in the tree`), this);
  }
  setHighlighted(t) {
    const n = this.getChartState(),
      i = n
        .generateRoot(n.data)
        .descendants()
        .filter(i => '' + n.nodeId(i.data) == '' + t)[0];
    return i
      ? (i.ancestors().forEach(t => (t.data._expanded = !0)), (i.data._highlighted = !0), (i.data._expanded = !0), (i.data._centered = !0), this)
      : (console.log(`ORG CHART - HIGHLIGHT - Node with id (${t})  not found in the tree`), this);
  }
  setUpToTheRootHighlighted(t) {
    const n = this.getChartState(),
      i = n
        .generateRoot(n.data)
        .descendants()
        .filter(i => '' + n.nodeId(i.data) == '' + t)[0];
    return i
      ? (i.ancestors().forEach(t => (t.data._expanded = !0)),
        (i.data._upToTheRootHighlighted = !0),
        (i.data._expanded = !0),
        i.ancestors().forEach(t => (t.data._upToTheRootHighlighted = !0)),
        this)
      : (console.log(`ORG CHART - HIGHLIGHTROOT - Node with id (${t}) not found in the tree`), this);
  }
  clearHighlighting() {
    const t = this.getChartState();
    return (
      t.allNodes.forEach(t => {
        ((t.data._highlighted = !1), (t.data._upToTheRootHighlighted = !1));
      }),
      this.update(t.root),
      this
    );
  }
  fullscreen(t) {
    const n = this.getChartState(),
      i = Tr.select(t || n.container).node();
    (Tr.select(document).on('fullscreenchange.' + n.id, function () {
      (document.fullscreenElement || document.mozFullscreenElement || document.webkitFullscreenElement) == i
        ? setTimeout(() => {
            n.svg.attr('height', window.innerHeight - 40);
          }, 500)
        : n.svg.attr('height', n.svgHeight);
    }),
      i.requestFullscreen
        ? i.requestFullscreen()
        : i.mozRequestFullScreen
          ? i.mozRequestFullScreen()
          : i.webkitRequestFullscreen
            ? i.webkitRequestFullscreen()
            : i.msRequestFullscreen && i.msRequestFullscreen());
  }
  zoomIn() {
    const { svg: t, zoomBehavior: n } = this.getChartState();
    t.transition().call(n.scaleBy, 1.3);
  }
  zoomOut() {
    const { svg: t, zoomBehavior: n } = this.getChartState();
    t.transition().call(n.scaleBy, 0.78);
  }
  toDataURL(t, n) {
    var i = new XMLHttpRequest();
    ((i.onload = function () {
      var t = new FileReader();
      ((t.onloadend = function () {
        n(t.result);
      }),
        t.readAsDataURL(i.response));
    }),
      i.open('GET', t),
      (i.responseType = 'blob'),
      i.send());
  }
  exportImg({ full: t = !1, scale: n = 3, onLoad: i = t => t, save: r = !0, backgroundColor: e = '#FAFAFA' } = {}) {
    const o = this,
      s = this.getChartState(),
      { svg: u, root: a } = s;
    let h = 0;
    const l = u.selectAll('img');
    let c = l.size();
    const f = () => {
      JSON.parse(JSON.stringify(o.lastTransform()));
      const u = o.duration();
      t && o.fit();
      const { svg: h } = o.getChartState();
      setTimeout(
        () => {
          o.downloadImage({
            node: h.node(),
            scale: n,
            isSvg: !1,
            backgroundColor: e,
            onAlreadySerialized: () => {
              o.update(a);
            },
            imageName: s.imageName,
            onLoad: i,
            save: r,
          });
        },
        t ? u + 10 : 0,
      );
    };
    c > 0
      ? l.each(function () {
          o.toDataURL(this.src, t => {
            ((this.src = t), ++h == c && f());
          });
        })
      : f();
  }
  exportSvg() {
    const { svg: t, imageName: n } = this.getChartState();
    return (this.downloadImage({ imageName: n, node: t.node(), scale: 3, isSvg: !0 }), this);
  }
  expandAll() {
    const { data: t } = this.getChartState();
    return (t.forEach(t => (t._expanded = !0)), this.render(), this);
  }
  collapseAll() {
    const { allNodes: t } = this.getChartState();
    return (t.forEach(t => (t.data._expanded = !1)), this.initialExpandLevel(0), this.render(), this);
  }
  downloadImage({
    node: t,
    scale: n = 2,
    imageName: i = 'graph',
    isSvg: r = !1,
    save: e = !0,
    backgroundColor: o = '#FAFAFA',
    onAlreadySerialized: s = () => {},
    onLoad: u = () => {},
  }) {
    const a = t;
    function h(t, n) {
      var i = document.createElement('a');
      'string' == typeof i.download ? (document.body.appendChild(i), (i.download = n), (i.href = t), i.click(), document.body.removeChild(i)) : location.replace(t);
    }
    function l(t) {
      const n = 'http://www.w3.org/2000/xmlns/';
      t = t.cloneNode(!0);
      const i = window.location.href + '#',
        r = document.createTreeWalker(t, NodeFilter.SHOW_ELEMENT, null, !1);
      for (; r.nextNode(); ) for (const t of r.currentNode.attributes) t.value.includes(i) && (t.value = t.value.replace(i, '#'));
      return (
        t.setAttributeNS(n, 'xmlns', 'http://www.w3.org/2000/svg'),
        t.setAttributeNS(n, 'xmlns:xlink', 'http://www.w3.org/1999/xlink'),
        new XMLSerializer().serializeToString(t)
      );
    }
    if (r) {
      let t = l(a);
      return ((t = '<?xml version="1.0" standalone="no"?>\r\n' + t), h((d = 'data:image/svg+xml;charset=utf-8,' + encodeURIComponent(t)), i + '.svg'), void s());
    }
    const c = n,
      f = document.createElement('img');
    f.onload = function () {
      const t = document.createElement('canvas'),
        n = a.getBoundingClientRect();
      ((t.width = n.width * c), (t.height = n.height * c));
      const r = t.getContext('2d');
      ((r.fillStyle = o), r.fillRect(0, 0, n.width * c, n.height * c), r.drawImage(f, 0, 0, n.width * c, n.height * c));
      let s = t.toDataURL('image/png');
      (u && u(s), e && h(s, i + '.png'));
    };
    var d = 'data:image/svg+xml; charset=utf8, ' + encodeURIComponent(l(a));
    (s(), (f.src = d));
  }
  getTextWidth(t, { fontSize: n = 14, fontWeight: i = 400, defaultFont: r = 'Helvetice', ctx: e } = {}) {
    return ((e.font = `${i || ''} ${n}px ${r} `), e.measureText(t).width);
  }
  clear() {
    const t = this.getChartState();
    (Tr.select(window).on('resize.' + t.id, null), t.svg && t.svg.selectAll('*').remove());
  }
}
const Sr = t(
    class extends n {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          (this.tkNodeClick = i(this, 'tk-node-click', 7)),
          (this.nodeWidth = 160),
          (this.nodeHeight = 90),
          (this.nodeButtonWidth = 28.5),
          (this.nodeButtonHeight = 27),
          (this.nodeButtonX = -this.nodeButtonWidth / 2),
          (this.childrenMargin = 84),
          (this.compact = !1),
          (this.collapsible = !0),
          (this.defaultButtonHTML = () =>
            '\n   <div style="\n  box-sizing: border-box;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  padding: 6px 6px 4.5px 6px;\n  border: 0.75px solid #CACFD8;\n  border-radius: 8px;\n  background: #FFF;\n  box-shadow:\n    0px 0px 0px 1.5px #FFF,\n    0px 0px 0px 3px #DADEE3;\n">\n    <svg xmlns="http://www.w3.org/2000/svg" width="16" height="15" viewBox="0 0 16 15" fill="none">\n      <path\n        d="M10.4219 5.62507L7.99687 8.05007L5.57187 5.62507C5.32812 5.38132 4.93437 5.38132 4.69062 5.62507C4.44688 5.86882 4.44688 6.26257 4.69062 6.50632L7.55937 9.37507C7.80312 9.61882 8.19687 9.61882 8.44062 9.37507L11.3094 6.50632C11.5531 6.26257 11.5531 5.86882 11.3094 5.62507C11.0656 5.38757 10.6656 5.38132 10.4219 5.62507Z"\n        fill="#222530" />\n    </svg>\n  </div>'));
      }
      dataChanged() {
        this.updateOrgChart();
      }
      optionsChanged() {
        (this.mergeOptions(), this.updateOrgChart());
      }
      collapsibleChanged() {
        this.updateButtonsState();
      }
      componentWillLoad() {
        this.mergeOptions();
      }
      componentDidLoad() {
        if (this.orgChartContainerRef) {
          const t = this.normaliseData(this.data);
          ((this.orgChartInstance = new Br()),
            this.orgChartInstance
              .container(this.orgChartContainerRef)
              .data(t)
              .layout('top')
              .initialExpandLevel(1 / 0)
              .nodeWidth(() => this.nodeWidth)
              .nodeHeight(() => this.nodeHeight)
              .nodeButtonWidth(() => this.nodeButtonWidth)
              .nodeButtonHeight(() => this.nodeButtonHeight)
              .nodeButtonX(() => this.nodeButtonX)
              .childrenMargin(() => this.childrenMargin)
              .compact(this.compact)
              .nodeContent(t => this.defaultNodeHTML(t))
              .buttonContent(() => this.defaultButtonHTML()),
            this.orgChartInstance.onNodeClick(t => this.tkNodeClick.emit(t)));
          const n = this.orgChartInstance.layoutBindings();
          ((n.top.diagonal = (t, n) => {
            const i = t.x,
              r = t.y,
              e = n.x,
              o = n.y,
              s = (r + o) / 2;
            return `M ${i} ${r} L ${i} ${s} L ${e} ${s} L ${e} ${o}`;
          }),
            (n.top.buttonY = t => t.height + this.nodeButtonHeight),
            (n.top.linkParentY = t => t.parent.y + t.parent.height + this.nodeButtonHeight + 14),
            this.updateButtonsState(),
            this.orgChartInstance.layoutBindings(n).render(),
            this.orgChartInstance.render());
        }
      }
      disconnectedCallback() {
        (this.orgChartInstance && (this.orgChartInstance = void 0), this.buttonUpdateTimeout && (clearTimeout(this.buttonUpdateTimeout), (this.buttonUpdateTimeout = void 0)));
      }
      async getOrgChart() {
        return this.orgChartInstance;
      }
      async refresh() {
        this.updateOrgChart();
      }
      async addNode(t) {
        this.orgChartInstance && this.orgChartInstance.addNode(t);
      }
      async fit() {
        this.orgChartInstance && this.orgChartInstance.fit();
      }
      needsNormalising(t) {
        if (Array.isArray(t)) {
          if (!t.length) return !1;
          const n = t[0];
          return !('id' in n) || !!(null == n ? void 0 : n.children);
        }
        return !!(null == t ? void 0 : t.children);
      }
      flattenTree(t, n, i, r) {
        var e, o;
        const s = null !== (e = t.id) && void 0 !== e ? e : r.cur++;
        (i.push(Object.assign(Object.assign({}, t), { id: s, parentId: n })), (null !== (o = t.children) && void 0 !== o ? o : []).forEach(t => this.flattenTree(t, s, i, r)));
      }
      normaliseData(t) {
        if (!this.needsNormalising(t)) return Array.isArray(t) ? [...t] : [t];
        const n = [],
          i = { cur: 1 };
        return ((Array.isArray(t) ? t : [t]).forEach(t => this.flattenTree(t, null, n, i)), n);
      }
      mergeOptions() {
        this.internalOptions = o({}, {}, this.options);
      }
      updateOrgChart() {
        if (!this.orgChartInstance) return;
        const t = this.normaliseData(this.data);
        (this.orgChartInstance
          .data(t)
          .initialExpandLevel(1 / 0)
          .render(),
          this.updateButtonsState());
      }
      defaultNodeHTML(t) {
        var n, i;
        return `\n      <div style="\n      border: 1px solid #E1E4EA;\n      border-radius: 8px;\n      padding: 8px;\n      display: flex;\n      flex-direction: column;\n      align-items: center;\n      text-align: center;\n      font-family: var(--family-body);";\n      >\n        <div style="\n  width: 32px;\n  height: 32px;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  padding: 4.489px 3.333px 0px 3.667px;\n  border: 1px solid white;\n  border-radius: 9999px;\n  background-color: #E1E4EA;\n  box-sizing: border-box;\n  overflow: hidden;\n  position: relative;">\n  <svg xmlns="http://www.w3.org/2000/svg" width="26" height="28" viewBox="0 0 26 28" fill="none">\n    <path d="M18.8958 18.8634C16.9861 20.2523 15.0763 21.3634 13.1666 21.3634C11.2569 21.3634 9.34714 20.2523 7.43745 18.8634C9.01155 18.3843 9.81016 17.669 9.83329 16.7176C9.83329 16.548 9.83204 16.3072 9.83089 15.855C9.83089 15.7832 9.83072 15.7096 9.83037 15.6344C9.82725 14.4219 9.82214 12.8219 9.81495 11.1969C8.32204 9.26176 8.87047 6.99312 9.33829 7.04968C9.91183 7.11947 14.8809 2.40405 15.834 2.16228C16.7871 1.92051 19.2083 2.70103 19.625 4.87551C20.0416 7.04999 20.2158 12.531 18.6354 14.7193C18.1856 15.342 17.4773 15.5855 16.5104 15.45C16.509 15.9051 16.5063 16.1605 16.5 16.6759C16.5123 17.6706 17.3109 18.3848 18.8958 18.8634Z" fill="url(#paint0_linear)"/>\n    <path d="M16.5 15.4485C14.2083 15.1881 12.75 14.1985 12.75 14.1985C12.75 14.1985 14.4167 16.2818 16.5 16.6985V15.4485Z" fill="#FC9F6A"/>\n    <path d="M9.05179 11.2296C8.33409 9.50357 5.45804 5.98826 7.93721 3.80076C8.77054 -0.261737 13.3747 0.187951 16.4476 1.12545C18.5129 1.75555 20.0726 3.00045 20.406 2.06295C22.4893 3.80076 21.446 5.50045 20.0726 5.98826C18.8262 6.43097 16.7185 6.87368 13.1345 6.45701C12.4943 6.38253 12.6284 8.34326 12.2863 8.55524C11.773 8.87326 11.3747 6.87368 9.96763 7.47409C8.56054 8.07451 9.39554 10.8585 10.906 10.8585C11.4268 10.8585 11.6872 12.2904 10.281 12.9675C9.26013 13.4692 9.51815 12.3512 9.05179 11.2296Z" fill="url(#paint1_linear)"/>\n    <path d="M2.22901 22.043C1.23734 24.0593 0.66651 28.7096 0.66651 28.7096H25.6665C25.6665 28.7096 25.0955 24.0588 24.104 22.043C23.1125 20.0271 17.3228 18.1574 17.3228 18.1574C14.1235 19.4067 12.1657 19.4067 9.01328 18.1567C9.01328 18.1567 3.22068 20.0266 2.22901 22.043Z" fill="#FDF4EE"/>\n    <path d="M12.4373 18.8859L11.9165 28.4692H14.4165L13.8957 18.8859C13.8957 18.8859 13.4095 18.365 13.1665 18.365C12.9235 18.365 12.4373 18.8859 12.4373 18.8859Z" fill="#C90019"/>\n    <path d="M12.3185 20.8329L13.9565 19.8864L13.8955 18.8858C13.5714 18.5386 13.3283 18.365 13.1663 18.365C13.0043 18.365 12.7613 18.5386 12.4372 18.8858L12.3184 20.8329H12.3185Z" fill="#645050"/>\n    <path d="M13.1667 18.3642L14.2083 20.24L17.4375 18.2608L16.5 16.9067L13.1667 18.138L9.83333 16.9067L8.89583 18.2608L12.125 20.24L13.1667 18.3642Z" fill="white"/>\n    <defs>\n      <linearGradient id="paint0_linear" x1="13.6748" y1="2.12" x2="13.6748" y2="21.3634" gradientUnits="userSpaceOnUse">\n        <stop stop-color="#F7B186" />\n        <stop offset="1" stop-color="#FFC299" />\n      </linearGradient>\n      <linearGradient id="paint1_linear" x1="14.2663" y1="0.48877" x2="14.2663" y2="13.0909" gradientUnits="userSpaceOnUse">\n        <stop stop-color="#1D0024" />\n        <stop offset="1" stop-color="#100014" />\n      </linearGradient>\n    </defs>\n  </svg>\n</div>\n        <div style="\n      display: flex;\n      flex-direction: column;\n      align-items: center;\n      gap: 4px;\n      max-width: 90px;"\n\n>\n          <div style=\n          "font-size:14px;\n          font-weight:400;\n          color:#222530;\n          line-height:20px;\n          white-space: nowrap;\n          overflow: hidden;\n          display: inline-block;\n          max-width: 90px;">\n          ${null !== (n = t.data.name) && void 0 !== n ? n : ''}\n          </div>\n          <div style=\n          "font-size:11px;\n          font-weight:300;\n          color:#99A0AE;\n          line-height:16px;\n          white-space: nowrap;\n          overflow: hidden;\n          display: inline-block;\n          max-width: 90px;">\n          ${null !== (i = t.data.title) && void 0 !== i ? i : ''}\n          </div>\n        </div>\n      </div>`;
      }
      updateButtonsState() {
        const t = this.orgChartInstance.layoutBindings();
        this.orgChartInstance &&
          (this.buttonUpdateTimeout && clearTimeout(this.buttonUpdateTimeout),
          this.collapsible
            ? ((t.top.buttonY = t => t.height + this.nodeButtonHeight),
              (t.top.linkParentY = t => t.parent.y + t.parent.height + this.nodeButtonHeight + 14),
              (t.top.linkY = t => t.y - 4),
              this.orgChartInstance.layoutBindings(t).render(),
              (this.buttonUpdateTimeout = setTimeout(() => {
                this.orgChartInstance && this.orgChartInstance.buttonContent(({ node: t }) => (t.children || t._children ? this.defaultButtonHTML() : '')).render();
              }, 350)))
            : ((t.top.buttonY = t => t.height + this.nodeButtonHeight),
              (t.top.linkParentY = t => t.parent.y + t.parent.height),
              this.orgChartInstance
                .layoutBindings(t)
                .buttonContent(() => '')
                .render()));
      }
      render() {
        var t, n;
        const i =
          this.accessibilityLabel || (null === (n = null === (t = this.options) || void 0 === t ? void 0 : t.title) || void 0 === n ? void 0 : n.text) || 'Organization Chart';
        return r('div', {
          'key': '54339542466dad3f62e3807d40c7d9941194885c',
          'ref': t => (this.orgChartContainerRef = t),
          'class': 'tk-org-chart-container',
          'role': 'img',
          'aria-label': i,
        });
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { data: [{ dataChanged: 0 }], options: [{ optionsChanged: 0 }], collapsible: [{ collapsibleChanged: 0 }] };
      }
      static get style() {
        return '';
      }
    },
    [
      0,
      'tk-org-chart',
      {
        data: [16],
        options: [8],
        collapsible: [4],
        accessibilityLabel: [1, 'accessibility-label'],
        internalOptions: [32],
        getOrgChart: [64],
        refresh: [64],
        addNode: [64],
        fit: [64],
      },
      void 0,
      { data: [{ dataChanged: 0 }], options: [{ optionsChanged: 0 }], collapsible: [{ collapsibleChanged: 0 }] },
    ],
  ),
  Yr = Sr,
  Or = function () {
    'undefined' != typeof customElements &&
      ['tk-org-chart'].forEach(t => {
        'tk-org-chart' === t && (customElements.get(e(t)) || customElements.define(e(t), Sr));
      });
  };
export { Yr as TkOrgChart, Or as defineCustomElement };
