// @takeoff-ui/core@0.11.9/components/tk-popover.js
import { p as t, H as o, c as e, h as i, t as s } from './p-BfLQgzDX.js';
import { f as a } from './p-BiHzhysf.js';
import { C as n } from './p-C0U9wSBw.js';
import { g as r } from './p-BDMnqEGL.js';
const l = t(
    class extends o {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkChange = e(this, 'tk-change', 7)),
          (this.isOpen = !1),
          (this.hasContentSlot = !1),
          (this.trigger = 'click'),
          (this.type = 'basic'),
          (this.containerStyle = null),
          (this.closeHandler = () => {
            this.isOpen = !1;
          }));
      }
      isOpenChanged() {
        this.tkChange.emit(this.isOpen);
      }
      positionChanged() {
        this.popoverElement && this.updatePosition();
      }
      get isHover() {
        return 'hover' === this.trigger;
      }
      componentWillLoad() {
        this.hasContentSlot = !!this.el.querySelector('[slot="content"]');
      }
      componentDidLoad() {
        var t, o, e;
        ((this.clickOutsideMixin = new n({ referenceElement: this.el, handler: this.closeHandler, disabled: this.isHover || !this.isOpen })),
          (this.triggerElement = this.el.querySelector('[slot="trigger"]')),
          'hover' === this.trigger
            ? (null === (t = this.triggerElement) || void 0 === t || t.addEventListener('mouseenter', () => (this.isOpen = !0)),
              null === (o = this.triggerElement) || void 0 === o || o.addEventListener('mouseleave', () => (this.isOpen = !1)))
            : null === (e = this.triggerElement) || void 0 === e || e.addEventListener('click', () => (this.isOpen = !this.isOpen)));
      }
      disconnectedCallback() {
        var t, o, e, i, s;
        ('hover' === this.trigger
          ? (null === (t = this.triggerElement) || void 0 === t || t.removeEventListener('mouseenter', () => (this.isOpen = !0)),
            null === (o = this.triggerElement) || void 0 === o || o.removeEventListener('mouseleave', () => (this.isOpen = !1)))
          : null === (e = this.triggerElement) || void 0 === e || e.removeEventListener('click', () => (this.isOpen = !this.isOpen)),
          null === (i = this.cleanup) || void 0 === i || i.call(this),
          (this.cleanup = null),
          null === (s = this.clickOutsideMixin) || void 0 === s || s.disconnectedCallback());
      }
      componentDidUpdate() {
        var t, o;
        (this.clickOutsideMixin.updateConfig({ disabled: this.isHover || !this.isOpen }),
          this.isOpen
            ? (this.showInTopLayer(), null === (t = this.cleanup) || void 0 === t || t.call(this), this.updatePosition())
            : (null === (o = this.cleanup) || void 0 === o || o.call(this), (this.cleanup = null)));
      }
      async close() {
        this.isOpen = !1;
      }
      updatePosition() {
        this.cleanup = a(this.triggerElement, this.popoverElement, this.arrowElement, { placement: this.position });
      }
      showInTopLayer() {
        const t = this.popoverElement;
        if ('function' == typeof (null == t ? void 0 : t.showPopover) && t.isConnected && !t.matches(':popover-open'))
          try {
            t.showPopover();
          } catch (t) {}
      }
      render() {
        return i(
          'div',
          { 'key': 'a8102448b35b6d36483f3fc894cf3e501e65f913', 'class': 'tk-popover', 'data-testid': r(this.dataTestid, 'container') },
          i('slot', { key: 'e990eab12540553215603e8a4048f4d4e6bd0a2f', name: 'trigger' }),
          this.isOpen &&
            i(
              'div',
              {
                'key': '170c6b08b4a9abdc133513b1a55c514f6dc3f4fe',
                'ref': t => (this.popoverElement = t),
                'popover': 'manual',
                'class': { 'tk-popover-content': !0, ['tk-popover-' + this.type]: !0 },
                'style': Object.assign({}, this.containerStyle),
                'role': 'popover',
                'onClick': t => t.stopPropagation(),
                'data-testid': r(this.dataTestid, 'content'),
              },
              this.hasContentSlot && i('slot', { key: '217923e970a0c1050fafd595a1977ea3d0789b43', name: 'content' }),
              i('div', {
                'key': '0ab91df3542c594b8185fa73f4fe4c4fb0ee778f',
                'ref': t => (this.arrowElement = t),
                'class': 'tk-popover-arrow',
                'data-testid': r(this.dataTestid, 'arrow'),
              }),
            ),
        );
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { isOpen: [{ isOpenChanged: 0 }], position: [{ positionChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-popover-content{display:flex;gap:8px;position:fixed;padding:var(--tooltips-base-v-padding) var(--tooltips-base-h-padding);border-radius:var(--tooltips-base-radius, 8px);width:max-content;color:var(--text-base);font-family:var(--family-body);line-height:20px;z-index:1300;box-shadow:var(--shadow-base)}.tk-popover-content[popover]{margin:0;inset:auto;overflow:visible}.tk-popover-content.tk-popover-dark{border:1px solid var(--border-light);background:var(--background-darkest)}.tk-popover-content.tk-popover-dark .tk-popover-arrow{background:var(--background-darkest)}.tk-popover-content.tk-popover-white{border:1px solid var(--border-light);background:var(--static-light)}.tk-popover-content.tk-popover-white .tk-popover-arrow{background:var(--static-light)}.tk-popover-content.tk-popover-basic{border:1px solid var(--states-light);background:var(--background-lightest)}.tk-popover-content.tk-popover-basic .tk-popover-arrow{background:var(--background-lightest)}';
      }
    },
    [
      257,
      'tk-popover',
      { trigger: [1], position: [1], type: [1], containerStyle: [16], dataTestid: [513, 'data-testid'], isOpen: [32], hasContentSlot: [32], close: [64] },
      void 0,
      { isOpen: [{ isOpenChanged: 0 }], position: [{ positionChanged: 0 }] },
    ],
  ),
  p = l,
  h = function () {
    'undefined' != typeof customElements &&
      ['tk-popover'].forEach(t => {
        'tk-popover' === t && (customElements.get(s(t)) || customElements.define(s(t), l));
      });
  };
export { p as TkPopover, h as defineCustomElement };
