// @takeoff-ui/core@0.11.9/components/tk-radio-group.js
import { p as t, H as a, c as i, h as e, t as s } from './p-BfLQgzDX.js';
import { c as o } from './p-D7YaJBkB.js';
import { r as l } from './p-_JFHEL-E.js';
import { g as n } from './p-BDMnqEGL.js';
import { d as r } from './p-BpML1w-i.js';
import { i as d } from './p-DzS7aqZe.js';
const f = t(
    class extends a {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkChange = i(this, 'tk-change', 7)),
          (this.internals = this.attachInternals()),
          (this.direction = 'horizontal'),
          (this.invalid = !1),
          (this.position = 'left'),
          (this.showAsterisk = !1),
          (this.spread = !1),
          (this.type = 'default'));
      }
      invalidChanged() {
        var t;
        (null === (t = this.slottedItems) || void 0 === t ? void 0 : t.length) > 0 &&
          this.slottedItems.forEach(t => {
            t.invalid = this.invalid;
          });
      }
      valueChanged() {
        this.updateTkRadio();
      }
      formResetCallback() {
        this.handleFormReset();
      }
      updateTkRadio() {
        this.slottedItems.length > 0 &&
          this.slottedItems.forEach(t => {
            t.checked = d(this.value, t.value);
          });
      }
      handleFormReset() {
        ((this.value = null), this.tkChange.emit(this.value), this.updateTkRadio());
      }
      handleChange(t) {
        ((this.value = t.detail), this.tkChange.emit(this.value), this.updateTkRadio());
      }
      handleSlotChange() {
        ((this.slottedItems = this.el.querySelectorAll('tk-radio')),
          this.slottedItems.length > 0 &&
            this.slottedItems.forEach(t => {
              (t.addEventListener('tk-change', t => {
                (t.stopPropagation(), this.handleChange(t));
              }),
                (t.checked = d(this.value, t.value)),
                (t.invalid = this.invalid),
                this.spread && (t.style.flex = '1'),
                t.setAttribute('data-type', this.type),
                !t.name && this.name && (t.name = this.name));
            }));
      }
      render() {
        var t;
        let a;
        const i = o('tk-radio-group-container', { vertical: 'vertical' === this.direction });
        if ((null === (t = this.label) || void 0 === t ? void 0 : t.length) > 0) {
          const t = e('span', { 'key': '43ac6ce9cd0f2cc9f50dc619f011a1ca0ff27d67', 'class': 'asterisk', 'data-testid': n(this.dataTestid, 'label-asterisk') }, '*');
          a = e('label', { 'key': 'bcbff37d9ee0a14424bae6b954d3f19f27b42ed6', 'class': 'label', 'data-testid': n(this.dataTestid, 'label') }, this.label, this.showAsterisk && t);
        }
        return e(
          'div',
          { 'key': 'bed985255a6a99c6e9490f6b880778489240761d', 'class': i, 'aria-invalid': this.invalid, 'data-testid': n(this.dataTestid, 'container') },
          a,
          e(
            'div',
            { 'key': 'daba64884f66d2217297a06f517f9eca2cef4ada', 'class': o('tk-radio-holder', this.type, { spread: this.spread }), 'data-testid': n(this.dataTestid, 'holder') },
            e('slot', { key: '8f21682f7b8c2be6566bd5b1a612dc4feb4e3ec3', onSlotchange: this.handleSlotChange.bind(this) }),
          ),
          l(this.hint, this.error, this.invalid, this.dataTestid),
        );
      }
      static get formAssociated() {
        return !0;
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { invalid: [{ invalidChanged: 0 }], value: [{ valueChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-radio-group{display:block}.tk-radio-group-container{display:flex;align-items:flex-start;flex-direction:column;gap:4px;font-family:var(--family-body)}.tk-radio-group-container label{font-size:var(--size-sm);font-weight:400;color:var(--text-darkest);line-height:var(--desktop-body-s-line-height);padding:var(--input-external-label-base-label-v-padding) var(--input-external-label-base-label-h-padding)}.tk-radio-group-container label .asterisk{color:var(--states-danger-base);font-size:var(--desktop-label-s-size);font-weight:300;line-height:var(--desktop-label-s-line-height)}.tk-radio-group-container .tk-radio-holder{display:flex;flex-direction:row;gap:8px}.tk-radio-group-container .tk-radio-holder.spread{width:100%}.tk-radio-group-container.vertical .tk-radio-holder{flex-direction:column}.tk-radio-group-container[aria-invalid] .tk-radio-holder.card tk-radio{border-color:var(--states-danger-base)}';
      }
    },
    [
      321,
      'tk-radio-group',
      {
        label: [1],
        direction: [1],
        invalid: [4],
        position: [1],
        value: [1032],
        showAsterisk: [4, 'show-asterisk'],
        spread: [4],
        type: [1],
        hint: [1],
        error: [1],
        name: [1],
        dataTestid: [513, 'data-testid'],
        slottedItems: [32],
      },
      void 0,
      { invalid: [{ invalidChanged: 0 }], value: [{ valueChanged: 0 }] },
    ],
  ),
  h = f,
  m = function () {
    'undefined' != typeof customElements &&
      ['tk-radio-group', 'tk-icon'].forEach(t => {
        switch (t) {
          case 'tk-radio-group':
            customElements.get(s(t)) || customElements.define(s(t), f);
            break;
          case 'tk-icon':
            customElements.get(s(t)) || r();
        }
      });
  };
export { h as TkRadioGroup, m as defineCustomElement };
