// @takeoff-ui/core@0.11.9/components/tk-slider.js
import { p as t, H as i, c as e, h as s, t as a } from './p-BfLQgzDX.js';
import { c as l } from './p-D7YaJBkB.js';
import { r } from './p-_JFHEL-E.js';
import { g as o } from './p-BDMnqEGL.js';
import { d } from './p-BpML1w-i.js';
const n = t(
    class extends i {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkChange = e(this, 'tkChange', 7)),
          (this.min = 0),
          (this.max = 100),
          (this.step = 1),
          (this.disabled = !1),
          (this.range = !1),
          (this.value = 0),
          (this.rangeVisibility = !0),
          (this.type = 'labels'),
          (this.showAsterisk = !1),
          (this.invalid = !1),
          (this.draggingThumb = null),
          (this.handlePointerMove = t => {
            if (!this.draggingThumb || !this.trackRef) return;
            const i = this.trackRef.getBoundingClientRect(),
              e = this.clamp(this.min + ((t.clientX - i.left) / i.width) * (this.max - this.min)),
              s = Math.round(e / this.step) * this.step;
            ('min' === this.draggingThumb
              ? s > this.currentMax
                ? ((this.currentMin = this.currentMax), (this.currentMax = s), (this.draggingThumb = 'max'))
                : (this.currentMin = s)
              : s < this.currentMin
                ? ((this.currentMax = this.currentMin), (this.currentMin = s), (this.draggingThumb = 'min'))
                : (this.currentMax = s),
              this.tkChange.emit(this.range ? [this.currentMin, this.currentMax] : this.currentMin));
          }),
          (this.handlePointerUp = () => {
            ((this.draggingThumb = null), document.removeEventListener('pointermove', this.handlePointerMove), document.removeEventListener('pointerup', this.handlePointerUp));
          }));
      }
      componentWillLoad() {
        if (this.range && Array.isArray(this.value)) {
          const t = [...this.value].sort((t, i) => t - i);
          ((this.currentMin = t[0]), (this.currentMax = t[1]));
        } else 'number' == typeof this.value && (this.currentMin = this.clamp(this.value));
      }
      getPercentage(t) {
        return ((t - this.min) / (this.max - this.min)) * 100;
      }
      handlePointerDown(t) {
        this.disabled || ((this.draggingThumb = t), document.addEventListener('pointermove', this.handlePointerMove), document.addEventListener('pointerup', this.handlePointerUp));
      }
      clamp(t) {
        return Math.max(this.min, Math.min(this.max, t));
      }
      render() {
        const t = this.getPercentage(this.currentMin),
          i = this.range ? this.getPercentage(this.currentMax) : 0,
          e = 'min' === this.draggingThumb,
          a = 'max' === this.draggingThumb;
        return s(
          'div',
          {
            'key': '263f6234b9e87c2088c3640d6334351c5cd74e96',
            'class': 'tk-slider ' + (this.disabled ? 'tk-slider-disabled' : ''),
            'data-testid': o(this.dataTestid, 'container'),
          },
          this.label &&
            s(
              'label',
              { 'key': '92677e2d1e7e0de3bcec03795e126ca053ba7193', 'class': 'tk-slider-label', 'data-testid': o(this.dataTestid, 'label') },
              this.label,
              this.showAsterisk && s('span', { 'key': '268b382a22cc4775630d86a64678479202156c50', 'class': 'asterisk', 'data-testid': o(this.dataTestid, 'label-asterisk') }, '*'),
            ),
          s(
            'div',
            { 'key': '80de2293ac3a6e583653261f569fd84cf1e4b136', 'class': 'tk-slider-track-wrapper', 'data-testid': o(this.dataTestid, 'track-wrapper') },
            s(
              'div',
              { 'key': '6f6acce4a7aa9685cac7c71b8791b708bfbdc61f', 'class': 'tk-slider-track', 'ref': t => (this.trackRef = t), 'data-testid': o(this.dataTestid, 'track') },
              s(
                'div',
                this.range
                  ? { 'class': 'tk-slider-fill', 'style': { left: t + '%', width: i - t + '%' }, 'data-testid': o(this.dataTestid, 'fill') }
                  : { 'class': 'tk-slider-fill', 'style': { width: t + '%' }, 'data-testid': o(this.dataTestid, 'fill') },
              ),
              s(
                'div',
                {
                  'key': '29d606994463b861e11637fdb78ff57fd4c0d8f3',
                  'class': l('tk-slider-thumb', { 'tk-slider-thumb-disabled': this.disabled, 'tk-slider-thumb-active': e }),
                  'style': { left: t + '%' },
                  'onPointerDown': this.disabled ? void 0 : () => this.handlePointerDown('min'),
                  'data-testid': o(this.dataTestid, 'thumb-min'),
                },
                !this.disabled &&
                  (e || 'min' === this.draggingThumb) &&
                  s(
                    'div',
                    { 'key': '4924f18458bb00ec93bd86d9324d07d9381f29e5', 'class': 'tk-slider-tooltip', 'data-testid': o(this.dataTestid, 'tooltip-min') },
                    this.currentMin,
                    s('div', { 'key': 'f2475b42b43e84be0452daccc84647dbef2e6470', 'class': 'tk-slider-tooltip-arrow', 'data-testid': o(this.dataTestid, 'arrow-min') }),
                  ),
                s(
                  'svg',
                  {
                    'key': 'f082fac1f25094c638c5efaae7399718f5140ab8',
                    'xmlns': 'http://www.w3.org/2000/svg',
                    'width': '8',
                    'height': '8',
                    'data-testid': o(this.dataTestid, 'icon-min'),
                  },
                  s('circle', { key: '212cfe07b5d22ea14a6c922e54d7c072a42b062e', cx: '4', cy: '4', r: '4' }),
                ),
              ),
              this.range &&
                s(
                  'div',
                  {
                    'key': '0c5fc43620b17e3bff86925a6af28eca100ae69c',
                    'class': l('tk-slider-thumb', { 'tk-slider-thumb-disabled': this.disabled, 'tk-slider-thumb-active': a }),
                    'style': { left: i + '%' },
                    'onPointerDown': this.disabled ? void 0 : () => this.handlePointerDown('max'),
                    'data-testid': o(this.dataTestid, 'thumb-max'),
                  },
                  !this.disabled &&
                    (a || 'max' === this.draggingThumb) &&
                    s(
                      'div',
                      { 'key': 'e10c8559bb3fec1fb14a1d55ad69422ee76495f6', 'class': 'tk-slider-tooltip', 'data-testid': o(this.dataTestid, 'tooltip-max') },
                      this.currentMax,
                      s('div', { 'key': 'a267e096df2b16cb78702d0bf18b9ea0bff2dccb', 'class': 'tk-slider-tooltip-arrow', 'data-testid': o(this.dataTestid, 'arrow-max') }),
                    ),
                  s(
                    'svg',
                    {
                      'key': 'bdead5f1af72e4484dbf1be101c1ee02db02255a',
                      'xmlns': 'http://www.w3.org/2000/svg',
                      'width': '8',
                      'height': '8',
                      'data-testid': o(this.dataTestid, 'icon-max'),
                    },
                    s('circle', { key: '2743e44ae25c404501fa6f1a6062a366accef171', cx: '4', cy: '4', r: '4' }),
                  ),
                ),
            ),
          ),
          'labels' === this.type &&
            this.rangeVisibility &&
            s(
              'div',
              { 'key': '9c68e9db4c60d8e1ce38dd1fb2371d03d1174c88', 'class': 'tk-slider-labels', 'data-testid': o(this.dataTestid, 'labels') },
              s('span', { 'key': '01d65e3c44402a1e494ece8d64a8cbe1a1d3b63e', 'data-testid': o(this.dataTestid, 'label-min') }, this.min),
              s('span', { 'key': 'a24380dc86200166b0122c7cef4f142fbbb29639', 'data-testid': o(this.dataTestid, 'label-max') }, this.max),
            ),
          'ticks' === this.type &&
            s(
              'div',
              { 'key': '5534d40d73025e37b65c4d64370bb01490713bb3', 'class': 'tk-slider-ticks', 'data-testid': o(this.dataTestid, 'ticks') },
              s(
                'div',
                { 'key': '39ee25499f398aab68adbe668aab0c1f21e0c78d', 'class': 'tk-slider-tick-track', 'data-testid': o(this.dataTestid, 'tick-track') },
                Array.from({ length: Math.floor((this.max - this.min) / this.step) + 1 }).map((t, i) =>
                  s('div', { 'key': i, 'class': 'tk-slider-tick', 'data-testid': o(this.dataTestid, 'tick', '' + i) }),
                ),
              ),
            ),
          r(this.hint, this.error, this.invalid, this.dataTestid),
        );
      }
      get el() {
        return this;
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-slider{display:flex;flex-direction:column;gap:var(--slider-gap, 8px)}.tk-slider-label{display:flex;padding:var(--external-label-large-label-v-padding, 0px) 4px;align-items:center;align-self:stretch;font-family:var(--desktop-label-s-font);font-size:var(--label-m-base-size, 14px);font-style:normal;font-weight:400;line-height:var(--label-m-base-line-height, 20px);color:var(--text-primary, #1e1e1e)}.tk-slider-disabled .tk-slider-label{color:var(--text-sub-base, #99a0ae)}.tk-slider-label .asterisk{color:var(--states-danger-base, #c90019);margin-left:4px;font-size:14px}.tk-slider-track-wrapper{position:relative;height:6px;background:var(--background-light, #e1e4ea);border-radius:8px;display:flex;flex-direction:column;align-items:flex-start;gap:var(--slider-content-gap, 10px);align-self:stretch;padding:var(--slider-v-padding, 0px) var(--slider-h-padding, 0px)}.tk-slider-track{position:relative;width:100%;height:100%}.tk-slider-fill{position:absolute;height:6px;background:var(--primary-base, #c90019);border-radius:8px;top:0}.tk-slider-disabled .tk-slider-fill{background:var(--background-sub-base, #cacfd8)}.tk-slider-thumb{display:flex;justify-content:center;align-items:center;position:absolute;top:50%;left:0;transform:translate(-50%, -50%);width:16px;height:16px;border-radius:50%;border:1px solid var(--static-Light, #fff);background:var(--primary-base, #c90019);z-index:2;cursor:pointer;transition:box-shadow 0.2s ease}.tk-slider-thumb svg{width:8px;height:8px;display:block;transform:scale(1.2);transform-origin:center}.tk-slider-thumb svg circle{fill:white}.tk-slider-thumb-active{background:white;border:1px solid var(--primary-base, #c90019)}.tk-slider-thumb-active svg circle{fill:var(--primary-base, #c90019)}.tk-slider-thumb-disabled{background:var(--background-sub-base, #cacfd8);cursor:not-allowed}.tk-slider-thumb-disabled svg circle{fill:white}.tk-slider-thumb-active{box-shadow:0 0 0 2px var(--primary-base, #c90019)}.tk-slider-labels{display:flex;justify-content:space-between;flex-direction:row;width:100%;font-size:11px;line-height:16px;color:var(--text-base, #6b7280);text-transform:uppercase;font-family:var(--desktop-subheading-2xs-font);font-weight:400}.tk-slider-labels-disabled{color:var(--text-light, #cacfd8)}.tk-slider-ticks{display:flex;padding:4px var(--slider-h-padding, 8px);flex-direction:column;align-items:center;align-self:stretch;width:100%}.tk-slider-tick-track{display:flex;height:8px;justify-content:space-between;align-items:flex-end;align-self:stretch;width:100%}.tk-slider-tick{width:2px;height:8px;border-radius:4px;background:var(--background-sub-base, #cacfd8)}.tk-slider-tooltip{position:absolute;top:160%;left:50%;transform:translateX(-50%);background:#191c22;color:white;padding:4px 8px;font-size:12px;font-family:var(--desktop-subheading-xs-font);font-weight:400;border-radius:6px;white-space:nowrap;box-shadow:0 2px 6px rgba(0, 0, 0, 0.2);z-index:3}.tk-slider-tooltip-arrow{position:absolute;top:-4px;left:50%;transform:translateX(-50%);width:0;height:0;border-left:4px solid transparent;border-right:4px solid transparent;border-bottom:4px solid #191c22}';
      }
    },
    [
      1,
      'tk-slider',
      {
        label: [1],
        min: [2],
        max: [2],
        step: [2],
        disabled: [4],
        range: [4],
        value: [2],
        rangeVisibility: [4, 'range-visibility'],
        type: [1],
        showAsterisk: [4, 'show-asterisk'],
        invalid: [4],
        error: [1],
        hint: [1],
        dataTestid: [513, 'data-testid'],
        currentMin: [32],
        currentMax: [32],
        draggingThumb: [32],
      },
    ],
  ),
  h = n,
  c = function () {
    'undefined' != typeof customElements &&
      ['tk-slider', 'tk-icon'].forEach(t => {
        switch (t) {
          case 'tk-slider':
            customElements.get(a(t)) || customElements.define(a(t), n);
            break;
          case 'tk-icon':
            customElements.get(a(t)) || d();
        }
      });
  };
export { h as TkSlider, c as defineCustomElement };
