// @takeoff-ui/core@0.11.9/components/tk-step.js
import { p as t, H as e, h as o, d as i, t as a } from './p-BfLQgzDX.js';
const n = t(
    class extends e {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          (this.complete = !1),
          (this.error = !1),
          (this.isActive = !1),
          (this.isClickable = !0),
          (this.labelPosition = 'non-flip'),
          (this.stepMode = 'basic'),
          (this.disabled = !1));
      }
      render() {
        return o(i, { key: '45beb75797f35160514c95035cdade44b36d2099' }, o('slot', { key: '689ac301bf415c2c13ff23e27a867475b5177e7c' }));
      }
      get el() {
        return this;
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}';
      }
    },
    [
      260,
      'tk-step',
      {
        header: [1],
        subheader: [1],
        icon: [1],
        completeIcon: [1, 'complete-icon'],
        activeIcon: [1, 'active-icon'],
        inactiveIcon: [1, 'inactive-icon'],
        errorIcon: [1, 'error-icon'],
        complete: [4],
        error: [4],
        isActive: [4, 'is-active'],
        index: [2],
        isClickable: [4, 'is-clickable'],
        labelPosition: [1, 'label-position'],
        stepMode: [1, 'step-mode'],
        disabled: [4],
      },
    ],
  ),
  s = n,
  l = function () {
    'undefined' != typeof customElements &&
      ['tk-step'].forEach(t => {
        'tk-step' === t && (customElements.get(a(t)) || customElements.define(a(t), n));
      });
  };
export { s as TkStep, l as defineCustomElement };
