// @takeoff-ui/core@0.11.9/components/tk-textarea.js
import { p as t, H as a, c as e, h as i, t as r } from './p-BfLQgzDX.js';
import { c as s } from './p-D7YaJBkB.js';
import { r as n } from './p-_JFHEL-E.js';
import { g as o } from './p-BDMnqEGL.js';
import { d as l } from './p-BpML1w-i.js';
import { v as d } from './p-CK3_k8jD.js';
const h = t(
    class extends a {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.tkInput = e(this, 'tk-input', 7)),
          (this.tkChange = e(this, 'tk-change', 7)),
          (this.tkBlur = e(this, 'tk-blur', 7)),
          (this.tkFocus = e(this, 'tk-focus', 7)),
          (this.internals = this.attachInternals()),
          (this.uniqueId = d()),
          (this.hasFocus = !1),
          (this.passwordStrength = 0),
          (this.charCount = 0),
          (this.disabled = !1),
          (this.invalid = !1),
          (this.rows = 3),
          (this.readonly = !1),
          (this.size = 'base'),
          (this.showAsterisk = !1),
          (this.value = ''),
          (this.updateCharCount = () => {
            this.maxLength && (this.charCount = ('' + this.value).trim().length);
          }),
          (this.handleInput = t => {
            const a = t.target;
            ((this.value = (null == a ? void 0 : a.value) || ''), this.tkInput.emit(t));
          }),
          (this.handleBlur = () => {
            ((this.hasFocus = !1), this.tkBlur.emit());
          }),
          (this.handleFocus = () => {
            ((this.hasFocus = !0), this.tkFocus.emit());
          }));
      }
      valueChanged() {
        (this.tkChange.emit(this.value), this.updateCharCount());
      }
      componentWillLoad() {
        if (this.el.hasAttribute('tabindex')) {
          const t = this.el.getAttribute('tabindex');
          ((this.tabindex = null !== t ? t : void 0), this.el.removeAttribute('tabindex'));
        }
      }
      componentDidLoad() {
        this.nativeInput = this.el.shadowRoot.querySelector('textarea');
      }
      formResetCallback() {
        this.handleFormReset();
      }
      async setFocus() {
        var t;
        null === (t = this.nativeInput) || void 0 === t || t.focus();
      }
      handleFormReset() {
        ((this.value = null), (this.nativeInput.value = null), this.tkChange.emit(this.value));
      }
      renderLabel() {
        var t;
        let a;
        if ((null === (t = this.label) || void 0 === t ? void 0 : t.length) > 0) {
          const t = i('span', { 'class': 'asterisk', 'data-testid': o(this.dataTestid, 'label-asterisk') }, '*');
          a = i('label', { 'htmlFor': this.uniqueId, 'class': 'label', 'data-testid': o(this.dataTestid, 'label') }, this.label, this.showAsterisk ? t : '');
        }
        return a;
      }
      render() {
        const t = s('tk-textarea-container', this.size, { focus: this.hasFocus });
        let a;
        const e = s('counter', this.charCount == this.maxLength && 'maxed');
        return (
          this.maxLength &&
            (a = i('span', { 'key': '16acb14e4ef37bae5e34d28408cc6a697cc48a88', 'class': e, 'data-testid': o(this.dataTestid, 'counter') }, this.charCount, '/', this.maxLength)),
          i(
            'div',
            {
              'key': 'f5c34cc6cb7b59b835b54b8b79830b21471ef6b5',
              'aria-readonly': this.readonly,
              'aria-disabled': this.disabled,
              'aria-invalid': this.invalid,
              'class': t,
              'data-testid': o(this.dataTestid, 'container'),
            },
            this.renderLabel(),
            i(
              'div',
              { 'key': 'be29df5ed44ca9f4bc9b52903ddc267833557a24', 'class': 'tk-textarea', 'data-testid': o(this.dataTestid, 'control') },
              i('textarea', {
                'key': '5d3fc8d5d32271eb726d0cb103b99951b3f0053c',
                'id': this.uniqueId,
                'ref': t => (this.nativeInput = t),
                'disabled': this.disabled,
                'autoComplete': 'off',
                'name': this.name,
                'rows': this.rows,
                'maxLength': this.maxLength,
                'placeholder': this.placeholder || '',
                'readOnly': this.readonly,
                'tabindex': this.tabindex,
                'onInput': this.handleInput,
                'onBlur': this.handleBlur,
                'onFocus': this.handleFocus,
                'value': this.value,
                'data-testid': o(this.dataTestid, 'native-textarea'),
              }),
              a,
            ),
            n(this.hint, this.error, this.invalid, this.dataTestid),
          )
        );
      }
      static get formAssociated() {
        return !0;
      }
      get el() {
        return this;
      }
      static get watchers() {
        return { value: [{ valueChanged: 0 }] };
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-textarea-container{display:flex;align-items:flex-start;flex-direction:column;gap:var(--input-external-label-large-container-gap);width:100%}.tk-textarea-container .label{color:var(--text-darkest);font-family:var(--family-body);font-weight:400}.tk-textarea-container .label .asterisk{color:var(--states-danger-base);font-size:var(--desktop-label-s-size);font-weight:300;line-height:var(--desktop-label-s-line-height)}.tk-textarea-container .tk-textarea{display:flex;gap:var(--spacing-xs);position:relative;align-items:center;border:1px solid var(--border-light);border-radius:var(--spacing-m-base);background-color:var(--static-light);width:-webkit-fill-available}.tk-textarea-container .tk-textarea textarea{outline:none;border:none;background-color:transparent;color:var(--text-darkest);padding:0px;width:100%;line-height:var(--desktop-body-m-base-line-height);resize:vertical}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar{width:4px;height:4px}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-textarea-container .tk-textarea textarea::placeholder{color:var(--text-sub-base)}.tk-textarea-container .tk-textarea:hover:not(.tk-textarea-container[aria-disabled] .tk-textarea){background-color:var(--background-lightest)}.tk-textarea-container .tk-textarea .counter{position:absolute;right:28px;bottom:6px;color:var(--text-base);line-height:var(--desktop-label-s-line-height);font-size:var(--desktop-label-s-size);font-weight:400}.tk-textarea-container .tk-textarea .counter.maxed{color:var(--states-danger-base)}.tk-textarea-container.large .label{font-size:var(--desktop-label-l-size);line-height:var(--desktop-label-l-line-height);padding:var(--input-external-label-large-label-v-padding) var(--input-external-label-large-label-h-padding)}.tk-textarea-container.large .tk-textarea{padding:var(--input-external-label-large-input-v-padding) var(--input-external-label-large-input-h-padding)}.tk-textarea-container.large .tk-textarea textarea{font-family:var(--desktop-body-m-base-font);font-size:var(--desktop-body-m-base-size);line-height:var(--desktop-body-m-base-line-height)}.tk-textarea-container.base .label{font-size:var(--desktop-label-m-base-size);line-height:var(--desktop-label-m-base-line-height);padding:var(--input-external-label-base-label-v-padding) var(--input-external-label-base-label-h-padding)}.tk-textarea-container.base .tk-textarea{padding:var(--input-external-label-base-input-v-padding) var(--input-external-label-base-input-h-padding)}.tk-textarea-container.base .tk-textarea textarea{font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);line-height:var(--desktop-body-s-line-height)}.tk-textarea-container.small .label{font-size:var(--desktop-label-s-size);line-height:var(--desktop-label-s-line-height);padding:var(--input-external-label-small-label-v-padding) var(--input-external-label-small-label-h-padding)}.tk-textarea-container.small .tk-textarea{padding:var(--input-external-label-small-input-v-padding) var(--input-external-label-small-input-h-padding)}.tk-textarea-container.small .tk-textarea textarea{font-family:var(--desktop-body-xs-font);font-size:var(--desktop-body-xs-size);line-height:var(--desktop-body-xs-line-height)}.tk-textarea-container.focus .tk-textarea{border:var(--spacing-px) solid var(--states-info-base)}.tk-textarea-container[aria-disabled]{pointer-events:none}.tk-textarea-container[aria-disabled] .label,.tk-textarea-container[aria-disabled] .tk-textarea textarea{color:var(--text-sub-base);pointer-events:none}.tk-textarea-container[aria-disabled] .tk-textarea{background-color:var(--background-lightest)}.tk-textarea-container[aria-readonly] .label{color:var(--text-sub-base)}.tk-textarea-container[aria-readonly] .tk-textarea{background-color:var(--background-lightest)}.tk-textarea-container[aria-invalid] .tk-textarea{border-color:var(--states-danger-base);background-color:var(--background-lightest)}';
      }
    },
    [
      65,
      'tk-textarea',
      {
        disabled: [4],
        invalid: [4],
        error: [1],
        hint: [1],
        label: [1],
        name: [1],
        placeholder: [1],
        rows: [2],
        readonly: [4],
        size: [1],
        showAsterisk: [4, 'show-asterisk'],
        maxLength: [2, 'max-length'],
        value: [1032],
        dataTestid: [513, 'data-testid'],
        hasFocus: [32],
        passwordStrength: [32],
        charCount: [32],
        setFocus: [64],
      },
      void 0,
      { value: [{ valueChanged: 0 }] },
    ],
  ),
  b = h,
  c = function () {
    'undefined' != typeof customElements &&
      ['tk-textarea', 'tk-icon'].forEach(t => {
        switch (t) {
          case 'tk-textarea':
            customElements.get(r(t)) || customElements.define(r(t), h);
            break;
          case 'tk-icon':
            customElements.get(r(t)) || l();
        }
      });
  };
export { b as TkTextarea, c as defineCustomElement };
