// @takeoff-ui/core@0.11.9/components/tk-timeline.js
import { p as t, H as e, h as i, F as n, t as l } from './p-BfLQgzDX.js';
import { c as m } from './p-D7YaJBkB.js';
import { g as a } from './p-BDMnqEGL.js';
const s = t(
    class extends e {
      constructor(t) {
        (super(),
          !1 !== t && this.__registerHost(),
          this.__attachShadow(),
          (this.slottedItemsCount = 0),
          (this.items = []),
          (this.orientation = 'horizontal'),
          (this.alternate = !0));
      }
      connectedCallback() {
        'undefined' != typeof window &&
          ((this.mutationObserver = new MutationObserver(() => {
            this.updateSlottedItemsCount();
          })),
          this.mutationObserver.observe(this.el, { childList: !0, subtree: !0 }));
      }
      componentDidLoad() {
        this.updateSlottedItemsCount();
      }
      disconnectedCallback() {
        this.mutationObserver && this.mutationObserver.disconnect();
      }
      updateSlottedItemsCount() {
        const t = this.el.querySelectorAll('tk-timeline-item').length;
        t !== this.slottedItemsCount && (this.slottedItemsCount = t);
      }
      determineContentPlacement(t) {
        return this.alternate ? (t % 2 == 0 ? 'end' : 'start') : 'end';
      }
      createItemContent(t, e) {
        return i(
          n,
          null,
          i(
            'div',
            { 'class': 'tk-timeline-item-content-inner', 'data-testid': a(this.dataTestid, 'item-content', '' + e) },
            t.title && i('div', { 'class': 'tk-timeline-item-title', 'data-testid': a(this.dataTestid, 'item-title', '' + e) }, t.title),
            t.description && i('div', { 'class': 'tk-timeline-item-description', 'data-testid': a(this.dataTestid, 'item-description', '' + e) }, t.description),
          ),
          t.date && i('div', { 'class': 'tk-timeline-item-date', 'data-testid': a(this.dataTestid, 'item-date', '' + e) }, t.date),
        );
      }
      renderTimelineItem(t, e) {
        const n = this.determineContentPlacement(e),
          l = m('tk-timeline-item', 'tk-timeline-item-placement-' + n, {
            'tk-timeline-item-first': 0 === e,
            'tk-timeline-item-last': e === this.items.length - 1,
            'tk-timeline-item-alternate': this.alternate,
          });
        return i(
          'li',
          { 'class': l, 'data-testid': a(this.dataTestid, 'item', '' + e) },
          this.alternate &&
            i(
              'div',
              { 'class': 'tk-timeline-item-content tk-timeline-item-content-start', 'data-testid': a(this.dataTestid, 'item-start', '' + e) },
              'start' === n && this.createItemContent(t, e),
            ),
          i(
            'div',
            { 'class': 'tk-timeline-item-separator', 'data-testid': a(this.dataTestid, 'item-separator', '' + e) },
            i('div', { 'class': 'tk-timeline-item-point', 'data-testid': a(this.dataTestid, 'item-point', '' + e) }),
            i('div', { 'class': 'tk-timeline-item-connector', 'data-testid': a(this.dataTestid, 'item-connector', '' + e) }),
          ),
          i(
            'div',
            { 'class': 'tk-timeline-item-content tk-timeline-item-content-end', 'data-testid': a(this.dataTestid, 'item-end', '' + e) },
            'end' === n && this.createItemContent(t, e),
          ),
        );
      }
      renderSlottedTimelineItems() {
        const t = Array.from(this.el.querySelectorAll('tk-timeline-item'));
        return t.map((e, n) => {
          const l = this.determineContentPlacement(n),
            s = m('tk-timeline-item', 'tk-timeline-item-placement-' + l, {
              'tk-timeline-item-first': 0 === n,
              'tk-timeline-item-last': n === t.length - 1,
              'tk-timeline-item-alternate': this.alternate,
            });
          return (
            e.setAttribute('data-class', 'tk-timeline-item-' + n),
            e.setAttribute('data-index', '' + n),
            i(
              'li',
              { 'class': s, 'data-testid': a(this.dataTestid, 'item', '' + n) },
              this.alternate &&
                i(
                  'div',
                  { 'class': 'tk-timeline-item-content tk-timeline-item-content-start', 'data-testid': a(this.dataTestid, 'item-start', '' + n) },
                  'start' === l && i('slot', { name: 'item-' + n }),
                ),
              i(
                'div',
                { 'class': 'tk-timeline-item-separator', 'data-testid': a(this.dataTestid, 'item-separator', '' + n) },
                i('div', { 'class': 'tk-timeline-item-point', 'data-testid': a(this.dataTestid, 'item-point', '' + n) }),
                i('div', { 'class': 'tk-timeline-item-connector', 'data-testid': a(this.dataTestid, 'item-connector', '' + n) }),
              ),
              i(
                'div',
                { 'class': 'tk-timeline-item-content tk-timeline-item-content-end', 'data-testid': a(this.dataTestid, 'item-end', '' + n) },
                'end' === l && i('slot', { name: 'item-' + n }),
              ),
            )
          );
        });
      }
      render() {
        const t = m('tk-timeline', 'tk-timeline-' + this.orientation),
          e = this.slottedItemsCount > 0;
        return i(
          'div',
          { 'key': 'e06fbcdfa4a9bd0d2231dbe618e3df8f0842811a', 'class': t, 'data-testid': a(this.dataTestid, 'container') },
          i(
            'ul',
            { 'key': '89eba65c61f411ae349ee9f9a29a300e6a717dc9', 'class': 'tk-timeline-items', 'data-testid': a(this.dataTestid, 'items') },
            e ? i(n, null, this.renderSlottedTimelineItems()) : this.items.map((t, e) => this.renderTimelineItem(t, e)),
          ),
        );
      }
      get el() {
        return this;
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-timeline{position:relative;width:100%;height:100%;font-family:var(--family-body)}.tk-timeline .tk-timeline-items{position:relative;display:flex;width:100%;height:100%;padding:0;margin:0;list-style:none}.tk-timeline .tk-timeline-items .tk-timeline-item{position:relative;display:flex;gap:4px}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-content{display:flex;flex-direction:column;flex:1;width:100%;gap:var(--spacing-xs);box-sizing:border-box}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-content .tk-timeline-item-content-inner{display:flex;flex-direction:column}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-content .tk-timeline-item-content-inner .tk-timeline-item-title{font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);font-weight:var(--desktop-body-s-line-weight);line-height:var(--desktop-body-s-line-height);color:var(--text-dark)}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-content .tk-timeline-item-content-inner .tk-timeline-item-description{font-family:var(--desktop-body-xs-font);font-size:var(--desktop-body-xs-size);font-weight:var(--weight-regular);line-height:var(--desktop-body-xs-line-height);color:var(--text-base)}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-content .tk-timeline-item-date{font-family:var(--desktop-body-2xs-font);font-size:var(--desktop-body-2xs-size);font-weight:var(--desktop-body-2xs-line-weight);line-height:var(--desktop-body-2xs-line-height);color:var(--text-light)}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-separator{display:flex;z-index:100}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-separator .tk-timeline-item-point{width:10px;height:10px;background-color:var(--primary-base);border:1px solid var(--primary-base);border-radius:8px;margin:5px;padding:5px;box-sizing:border-box;position:relative;z-index:200;flex-shrink:0}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-separator .tk-timeline-item-point::after{content:"";position:absolute;top:50%;left:50%;transform:translate(-50%, -50%);width:4px;height:4px;background-color:var(--static-light);border-radius:var(--radius-full)}.tk-timeline .tk-timeline-items .tk-timeline-item .tk-timeline-item-separator .tk-timeline-item-connector{background-color:var(--border-light);z-index:100}.tk-timeline .tk-timeline-items .tk-timeline-item.tk-timeline-item-last .tk-timeline-item-connector{display:none}.tk-timeline.tk-timeline-horizontal .tk-timeline-items{flex-direction:row;align-items:stretch;padding:0 var(--spacing-l)}.tk-timeline.tk-timeline-horizontal .tk-timeline-item{flex-direction:column;flex:1;padding:0;min-width:100px}.tk-timeline.tk-timeline-horizontal .tk-timeline-item .tk-timeline-item-content{text-align:left;padding-left:8px;padding-right:24px;min-height:20px;height:100%;align-items:flex-start}.tk-timeline.tk-timeline-horizontal .tk-timeline-item .tk-timeline-item-content.tk-timeline-item-content-start{justify-content:flex-end}.tk-timeline.tk-timeline-horizontal .tk-timeline-item .tk-timeline-item-content.tk-timeline-item-content-end{justify-content:flex-start}.tk-timeline.tk-timeline-horizontal .tk-timeline-item .tk-timeline-item-separator{flex-direction:row;align-items:center;width:100%;flex-shrink:0}.tk-timeline.tk-timeline-horizontal .tk-timeline-item .tk-timeline-item-connector{width:100%;height:1px;transform:translateY(-50%)}.tk-timeline.tk-timeline-horizontal .tk-timeline-item:not(.tk-timeline-item-alternate) .tk-timeline-item-content-start{min-height:0;padding:0}.tk-timeline.tk-timeline-horizontal .tk-timeline-item:not(.tk-timeline-item-alternate) .tk-timeline-item-content-start .tk-timeline-item-content-details{visibility:hidden}.tk-timeline.tk-timeline-vertical .tk-timeline-items{flex-direction:column;width:100%;align-items:stretch}.tk-timeline.tk-timeline-vertical .tk-timeline-item{flex-direction:row;align-items:stretch;position:relative}.tk-timeline.tk-timeline-vertical .tk-timeline-item .tk-timeline-item-content{padding-bottom:24px}.tk-timeline.tk-timeline-vertical .tk-timeline-item .tk-timeline-item-content.tk-timeline-item-content-start{text-align:right;justify-content:flex-end;align-items:flex-end}.tk-timeline.tk-timeline-vertical .tk-timeline-item .tk-timeline-item-content.tk-timeline-item-content-end{text-align:left;justify-content:flex-start;align-items:flex-start}.tk-timeline.tk-timeline-vertical .tk-timeline-item .tk-timeline-item-separator{flex-direction:column;align-items:center;flex-shrink:0;box-sizing:border-box}.tk-timeline.tk-timeline-vertical .tk-timeline-item .tk-timeline-item-content-end{text-align:left;align-items:flex-start;justify-content:flex-start}.tk-timeline.tk-timeline-vertical .tk-timeline-item .tk-timeline-item-connector{width:1px;z-index:100;height:100%;transform:translateX(-50%)}';
      }
    },
    [257, 'tk-timeline', { items: [16], orientation: [1], alternate: [4], dataTestid: [513, 'data-testid'], slottedItemsCount: [32] }],
  ),
  o = s,
  r = function () {
    'undefined' != typeof customElements &&
      ['tk-timeline'].forEach(t => {
        'tk-timeline' === t && (customElements.get(l(t)) || customElements.define(l(t), s));
      });
  };
export { o as TkTimeline, r as defineCustomElement };
