// @takeoff-ui/core@0.11.9/components/tk-timeline-item.js
import { p as t, H as e, h as a, d as o, t as i } from './p-BfLQgzDX.js';
const n = t(
    class extends e {
      constructor(t) {
        (super(), !1 !== t && this.__registerHost());
      }
      componentDidLoad() {
        const t = this.el.closest('tk-timeline');
        if (t) {
          const e = Array.from(t.querySelectorAll('tk-timeline-item')).indexOf(this.el);
          this.el.slot = 'item-' + e;
        }
      }
      render() {
        return a(o, { key: 'ef999f4a728a6a7f248b3ab70f5f6ae3c24987ca' }, a('slot', { key: '0093189d42688a21ba6f75451793104e3dbd8b1f' }));
      }
      get el() {
        return this;
      }
      static get style() {
        return '@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}';
      }
    },
    [260, 'tk-timeline-item'],
  ),
  s = n,
  l = function () {
    'undefined' != typeof customElements &&
      ['tk-timeline-item'].forEach(t => {
        'tk-timeline-item' === t && (customElements.get(i(t)) || customElements.define(i(t), n));
      });
  };
export { s as TkTimelineItem, l as defineCustomElement };
