// @takeoff-ui/core@0.11.9/dist/collection/components/tk-slider/tk-slider.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { renderHint } from "../../utils/hint-utils";
import { getDataTestId } from "../../utils/test-id-utils";
export class TkSlider {
    constructor() {
        /** The minimum value the slider can take */
        this.min = 0;
        /** The maximum value the slider can take */
        this.max = 100;
        /** The increment step for the slider value (e.g., step = 5 → 0, 5, 10, ...) */
        this.step = 1;
        /** Whether the slider is disabled (non-interactive if true) */
        this.disabled = false;
        /** Whether the slider operates in range mode (true) or single value mode (false) */
        this.range = false;
        /** Current value of the slider. If `range` is true, it should be [min, max] */
        this.value = 0;
        /** Whether the bottom label/tick section should be visible */
        this.rangeVisibility = true;
        /**
         * The type of visual indicator shown below the track.
         * 'labels' shows min/max values, 'ticks' shows evenly spaced tick marks
         */
        this.type = 'labels';
        /** Whether to show a red asterisk next to the label (typically for required fields) */
        this.showAsterisk = false;
        /** Marks the slider as invalid; used to apply error styling */
        this.invalid = false;
        this.draggingThumb = null;
        this.handlePointerMove = (e) => {
            if (!this.draggingThumb || !this.trackRef)
                return;
            const rect = this.trackRef.getBoundingClientRect();
            const percent = (e.clientX - rect.left) / rect.width;
            const value = this.clamp(this.min + percent * (this.max - this.min));
            const snapped = Math.round(value / this.step) * this.step;
            if (this.draggingThumb === 'min') {
                if (snapped > this.currentMax) {
                    this.currentMin = this.currentMax;
                    this.currentMax = snapped;
                    this.draggingThumb = 'max';
                }
                else {
                    this.currentMin = snapped;
                }
            }
            else {
                if (snapped < this.currentMin) {
                    this.currentMax = this.currentMin;
                    this.currentMin = snapped;
                    this.draggingThumb = 'min';
                }
                else {
                    this.currentMax = snapped;
                }
            }
            if (this.range) {
                this.tkChange.emit([this.currentMin, this.currentMax]);
            }
            else {
                this.tkChange.emit(this.currentMin);
            }
        };
        this.handlePointerUp = () => {
            this.draggingThumb = null;
            document.removeEventListener('pointermove', this.handlePointerMove);
            document.removeEventListener('pointerup', this.handlePointerUp);
        };
    }
    componentWillLoad() {
        if (this.range && Array.isArray(this.value)) {
            const sorted = [...this.value].sort((a, b) => a - b);
            this.currentMin = sorted[0];
            this.currentMax = sorted[1];
        }
        else if (typeof this.value === 'number') {
            this.currentMin = this.clamp(this.value);
        }
    }
    getPercentage(value) {
        return ((value - this.min) / (this.max - this.min)) * 100;
    }
    handlePointerDown(thumb) {
        if (this.disabled)
            return;
        this.draggingThumb = thumb;
        document.addEventListener('pointermove', this.handlePointerMove);
        document.addEventListener('pointerup', this.handlePointerUp);
    }
    clamp(value) {
        return Math.max(this.min, Math.min(this.max, value));
    }
    render() {
        const minPercent = this.getPercentage(this.currentMin);
        const maxPercent = this.range ? this.getPercentage(this.currentMax) : 0;
        const isMinActive = this.draggingThumb === 'min';
        const isMaxActive = this.draggingThumb === 'max';
        return (h("div", { key: '263f6234b9e87c2088c3640d6334351c5cd74e96', class: `tk-slider ${this.disabled ? 'tk-slider-disabled' : ''}`, "data-testid": getDataTestId(this.dataTestid, 'container') }, this.label && (h("label", { key: '92677e2d1e7e0de3bcec03795e126ca053ba7193', class: "tk-slider-label", "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label, this.showAsterisk && (h("span", { key: '268b382a22cc4775630d86a64678479202156c50', class: "asterisk", "data-testid": getDataTestId(this.dataTestid, 'label-asterisk') }, "*")))), h("div", { key: '80de2293ac3a6e583653261f569fd84cf1e4b136', class: "tk-slider-track-wrapper", "data-testid": getDataTestId(this.dataTestid, 'track-wrapper') }, h("div", { key: '6f6acce4a7aa9685cac7c71b8791b708bfbdc61f', class: "tk-slider-track", ref: el => (this.trackRef = el), "data-testid": getDataTestId(this.dataTestid, 'track') }, this.range ? (h("div", { class: "tk-slider-fill", style: {
                left: `${minPercent}%`,
                width: `${maxPercent - minPercent}%`,
            }, "data-testid": getDataTestId(this.dataTestid, 'fill') })) : (h("div", { class: "tk-slider-fill", style: { width: `${minPercent}%` }, "data-testid": getDataTestId(this.dataTestid, 'fill') })), h("div", { key: '29d606994463b861e11637fdb78ff57fd4c0d8f3', class: classNames('tk-slider-thumb', {
                'tk-slider-thumb-disabled': this.disabled,
                'tk-slider-thumb-active': isMinActive,
            }), style: { left: `${minPercent}%` }, onPointerDown: !this.disabled ? () => this.handlePointerDown('min') : undefined, "data-testid": getDataTestId(this.dataTestid, 'thumb-min') }, !this.disabled && (isMinActive || this.draggingThumb === 'min') && (h("div", { key: '4924f18458bb00ec93bd86d9324d07d9381f29e5', class: "tk-slider-tooltip", "data-testid": getDataTestId(this.dataTestid, 'tooltip-min') }, this.currentMin, h("div", { key: 'f2475b42b43e84be0452daccc84647dbef2e6470', class: "tk-slider-tooltip-arrow", "data-testid": getDataTestId(this.dataTestid, 'arrow-min') }))), h("svg", { key: 'f082fac1f25094c638c5efaae7399718f5140ab8', xmlns: "http://www.w3.org/2000/svg", width: "8", height: "8", "data-testid": getDataTestId(this.dataTestid, 'icon-min') }, h("circle", { key: '212cfe07b5d22ea14a6c922e54d7c072a42b062e', cx: "4", cy: "4", r: "4" }))), this.range && (h("div", { key: '0c5fc43620b17e3bff86925a6af28eca100ae69c', class: classNames('tk-slider-thumb', {
                'tk-slider-thumb-disabled': this.disabled,
                'tk-slider-thumb-active': isMaxActive,
            }), style: { left: `${maxPercent}%` }, onPointerDown: !this.disabled ? () => this.handlePointerDown('max') : undefined, "data-testid": getDataTestId(this.dataTestid, 'thumb-max') }, !this.disabled && (isMaxActive || this.draggingThumb === 'max') && (h("div", { key: 'e10c8559bb3fec1fb14a1d55ad69422ee76495f6', class: "tk-slider-tooltip", "data-testid": getDataTestId(this.dataTestid, 'tooltip-max') }, this.currentMax, h("div", { key: 'a267e096df2b16cb78702d0bf18b9ea0bff2dccb', class: "tk-slider-tooltip-arrow", "data-testid": getDataTestId(this.dataTestid, 'arrow-max') }))), h("svg", { key: 'bdead5f1af72e4484dbf1be101c1ee02db02255a', xmlns: "http://www.w3.org/2000/svg", width: "8", height: "8", "data-testid": getDataTestId(this.dataTestid, 'icon-max') }, h("circle", { key: '2743e44ae25c404501fa6f1a6062a366accef171', cx: "4", cy: "4", r: "4" })))))), this.type === 'labels' && this.rangeVisibility && (h("div", { key: '9c68e9db4c60d8e1ce38dd1fb2371d03d1174c88', class: "tk-slider-labels", "data-testid": getDataTestId(this.dataTestid, 'labels') }, h("span", { key: '01d65e3c44402a1e494ece8d64a8cbe1a1d3b63e', "data-testid": getDataTestId(this.dataTestid, 'label-min') }, this.min), h("span", { key: 'a24380dc86200166b0122c7cef4f142fbbb29639', "data-testid": getDataTestId(this.dataTestid, 'label-max') }, this.max))), this.type === 'ticks' && (h("div", { key: '5534d40d73025e37b65c4d64370bb01490713bb3', class: "tk-slider-ticks", "data-testid": getDataTestId(this.dataTestid, 'ticks') }, h("div", { key: '39ee25499f398aab68adbe668aab0c1f21e0c78d', class: "tk-slider-tick-track", "data-testid": getDataTestId(this.dataTestid, 'tick-track') }, Array.from({ length: Math.floor((this.max - this.min) / this.step) + 1 }).map((_, index) => (h("div", { key: index, class: "tk-slider-tick", "data-testid": getDataTestId(this.dataTestid, 'tick', index.toString()) })))))), renderHint(this.hint, this.error, this.invalid, this.dataTestid)));
    }
    static get is() { return "tk-slider"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-slider.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-slider.css"]
        };
    }
    static get properties() {
        return {
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The label text displayed above the slider"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "min": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The minimum value the slider can take"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "min",
                "defaultValue": "0"
            },
            "max": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The maximum value the slider can take"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max",
                "defaultValue": "100"
            },
            "step": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The increment step for the slider value (e.g., step = 5 \u2192 0, 5, 10, ...)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "step",
                "defaultValue": "1"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the slider is disabled (non-interactive if true)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "range": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the slider operates in range mode (true) or single value mode (false)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "range",
                "defaultValue": "false"
            },
            "value": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number | [number, number]",
                    "resolved": "[number, number] | number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Current value of the slider. If `range` is true, it should be [min, max]"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value",
                "defaultValue": "0"
            },
            "rangeVisibility": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the bottom label/tick section should be visible"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "range-visibility",
                "defaultValue": "true"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'labels' | 'ticks'",
                    "resolved": "\"labels\" | \"ticks\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The type of visual indicator shown below the track.\n'labels' shows min/max values, 'ticks' shows evenly spaced tick marks"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'labels'"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether to show a red asterisk next to the label (typically for required fields)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Marks the slider as invalid; used to apply error styling"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Error message to display when `invalid` is true"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Informational hint message (shown when no error is present)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "currentMin": {},
            "currentMax": {},
            "draggingThumb": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tkChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the slider value changes.\nEmits a number for single mode, or a [min, max] tuple for range mode."
                },
                "complexType": {
                    "original": "number | [number, number]",
                    "resolved": "[number, number] | number",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
}
