// @takeoff-ui/core@0.11.9/dist/collection/components/tk-button/tk-button.js
import { h, Host } from "@stencil/core";
import classNames from "classnames";
import { renderIcons, isMultiIconOptions } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * TkButton is an extension to standard input element with icons and theming.
 * @react `import { TkButton } from '@takeoff-ui/react'`
 * @vue `import { TkButton } from '@takeoff-ui/vue'`
 * @angular `import { TkButton } from '@takeoff-ui/angular'`
 */
export class TkButton {
    constructor() {
        /**
         * This field specifies the design type of the component.
         * @defaultValue filled
         */
        this.type = 'filled';
        /**
         * Defines the position of the icon.
         * @defaultValue 'left'
         */
        this.iconPosition = 'left';
        /**
         * Label text displayed inside the button.
         */
        this.label = '';
        /**
         * Sets size for the component.
         * @defaultValue 'base'
         */
        this.size = 'base';
        /**
         * Sets the button type.
         * @defaultValue 'button'
         */
        this.mode = 'button';
        /**
         * Determines the button's variant for different styles.
         */
        this.variant = 'primary';
    }
    componentWillLoad() {
        if (this.mode == 'link') {
            this.type = null;
        }
    }
    handleClick(e) {
        if (this.disabled) {
            e.preventDefault();
            e.stopPropagation();
            return;
        }
        this.tkClick.emit(e);
        if (this.mode === 'submit') {
            const form = this.el.closest('form');
            if (form) {
                form.requestSubmit();
            }
        }
        else if (this.mode === 'reset') {
            const form = this.el.closest('form');
            if (form) {
                form.reset();
            }
        }
    }
    getButtonIconColor() {
        if (this.disabled) {
            return this.type === 'filled' || this.type === 'filledLight' ? 'var(--static-white)' : 'var(--icon-sub-base)';
        }
        if (this.type === 'filled') {
            return this.variant === 'white' ? 'var(--text-darkest)' : 'var(--static-white)';
        }
        return 'currentColor';
    }
    render() {
        var _a;
        const hasMultipleIcons = isMultiIconOptions(this.icon);
        const rootClasses = classNames('tk-button', this.type, {
            'loading': this.loading && !this.disabled,
            'rounded': this.rounded && this.icon && (this.label == '' || this.label == null || this.label.length <= 0),
            'link': this.mode == 'link',
            'underline': this.underline,
            'icon-only': (this.label == '' || this.label == null || this.label.length <= 0) && this.icon && !hasMultipleIcons,
        }, [this.variant], [this.size]);
        let _leftIcon;
        let _rightIcon;
        const spinnerElement = (h("tk-spinner", { key: '51da6bf5c608da3dead29abca991c5a554e994cf', dataTestid: getDataTestId(this.dataTestid, 'loading-spinner'), size: this.size === 'large' ? 'small' : this.size === 'base' ? 'xsmall' : 'xxsmall' }));
        if (this.loading) {
            _leftIcon = spinnerElement;
        }
        else if (this.icon) {
            const { leftIcon, rightIcon } = renderIcons(this.icon, {
                variant: null,
                additionalProps: {
                    color: this.getButtonIconColor(),
                },
                dataTestid: this.dataTestid,
            }, this.iconPosition);
            _leftIcon = leftIcon;
            _rightIcon = rightIcon;
        }
        let Tag;
        let props;
        if (this.mode == 'link') {
            Tag = 'a';
            props = { href: this.href, target: this.target };
        }
        else {
            Tag = 'button';
            props = { type: this.type };
        }
        let label;
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            label = h("span", { key: 'c6e9ee7e3f9bc03811af77fa14fb896d063d2f05', "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label);
        }
        return (h(Host, { key: '2fa76cd8bc85239920d21707be0b8749a8c5b2a7', class: { 'full-width': this.fullWidth } }, h(Tag, Object.assign({ key: '4aa27a8b19b1a679e84367b7c6b2d81bb27d5a1f', class: rootClasses }, props, { "data-testid": getDataTestId(this.dataTestid, 'container'), disabled: this.disabled, onClick: (e) => this.handleClick(e) }), _leftIcon, label, _rightIcon)));
    }
    static get is() { return "tk-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-button.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-button.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disables the button, preventing user interaction."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled"
            },
            "fullWidth": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets the button to full-width mode, making it span the container."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "full-width"
            },
            "type": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "'filled' | 'filledLight' | 'outlined' | 'text'",
                    "resolved": "\"filled\" | \"filledLight\" | \"outlined\" | \"text\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "filled"
                        }],
                    "text": "This field specifies the design type of the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'filled'"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon name to be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "iconPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'left'"
                        }],
                    "text": "Defines the position of the icon."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon-position",
                "defaultValue": "'left'"
            },
            "href": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets the URL the button should navigate to when clicked (for type=\"link\" buttons)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "href"
            },
            "target": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Specifies where to open the linked document (for type=\"link\" buttons)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "target"
            },
            "underline": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Applies underline styling to the button label regardless of mode."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "underline"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label text displayed inside the button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label",
                "defaultValue": "''"
            },
            "loading": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Shows a loading icon inside the button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "loading"
            },
            "rounded": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Makes the button round with an icon-only style."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rounded"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "mode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'button' | 'submit' | 'reset' | 'link'",
                    "resolved": "\"button\" | \"link\" | \"reset\" | \"submit\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'button'"
                        }],
                    "text": "Sets the button type."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mode",
                "defaultValue": "'button'"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'secondary' | 'neutral' | 'info' | 'success' | 'danger' | 'warning' | 'white' | 'black'",
                    "resolved": "\"black\" | \"danger\" | \"info\" | \"neutral\" | \"primary\" | \"secondary\" | \"success\" | \"warning\" | \"white\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines the button's variant for different styles."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'primary'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get events() {
        return [{
                "method": "tkClick",
                "name": "tk-click",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the button click."
                },
                "complexType": {
                    "original": "MouseEvent",
                    "resolved": "MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "el"; }
}
