// @takeoff-ui/core@0.11.9/dist/collection/components/tk-toggle-button/tk-toggle-button-group.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { isEqual } from "lodash-es";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * TkToggleButtonGroup is a component that allows you to create a group of toggle buttons.
 * @react `import { TkToggleButtonGroup, TkToggleButton } from '@takeoff-ui/react'`
 * @vue `import { TkToggleButtonGroup, TkToggleButton } from '@takeoff-ui/vue'`
 * @angular `import { TkToggleButtonGroup, TkToggleButton } from '@takeoff-ui/angular'`
 */
export class TkToggleButtonGroup {
    constructor() {
        /**
         * The value type of the toggle button group.
         */
        this.type = 'basic';
        /**
         * The value of the rounded toggle button group.
         */
        this.rounded = false;
        /**
         * The direction of the toggle button group.
         */
        this.direction = 'horizontal';
    }
    roundedChanged() {
        this.updateSlottedItems();
    }
    valueChanged() {
        this.updateSelected();
        this.tkChange.emit(this.value);
    }
    componentWillLoad() {
        this.updateSlottedItems();
    }
    updateSlottedItems() {
        this.slottedItems = this.el.querySelectorAll('tk-toggle-button');
        this.updateSelected();
        this.slottedItems.forEach(item => {
            item.rounded = this.rounded;
            item.addEventListener('tk-toggle', this.handleToggle.bind(this));
        });
    }
    updateSelected() {
        var _a;
        if (((_a = this.slottedItems) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.slottedItems.forEach(item => {
                if (item.disabled) {
                    if (item.selected) {
                        item.selected = false;
                        item.value = undefined;
                    }
                    return;
                }
                item.selected = isEqual(this.value, item.value);
            });
        }
    }
    handleToggle(e) {
        const { value } = e.detail;
        this.value = value;
        this.updateSelected();
    }
    render() {
        const groupClasses = classNames('tk-toggle-button-group', this.type, {
            rounded: this.rounded,
            vertical: this.direction === 'vertical',
            horizontal: this.direction === 'horizontal',
        });
        return (h("div", { key: '3bb7ea989256d18ee0e94fbcaf825e0c64d4ea69', class: groupClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("slot", { key: '88e50487b5e744a4082df70f626a41259ab41c18', onSlotchange: () => this.updateSlottedItems() })));
    }
    static get is() { return "tk-toggle-button-group"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-toggle-button-group.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-toggle-button-group.css"]
        };
    }
    static get properties() {
        return {
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light'",
                    "resolved": "\"basic\" | \"divided\" | \"light\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The value type of the toggle button group."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'basic'"
            },
            "rounded": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the rounded toggle button group."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rounded",
                "defaultValue": "false"
            },
            "value": {
                "type": "any",
                "mutable": true,
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the selected toggle button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            },
            "direction": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'horizontal' | 'vertical'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The direction of the toggle button group."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "direction",
                "defaultValue": "'horizontal'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "slottedItems": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the selected value changes."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "rounded",
                "methodName": "roundedChanged"
            }, {
                "propName": "value",
                "methodName": "valueChanged"
            }];
    }
}
