// @takeoff-ui/core@0.11.9/dist/collection/components/tk-toggle-button/tk-toggle-button.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { renderIcons } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
export class TkToggleButton {
    constructor() {
        /**
         * Defines the position of the icon.
         * @defaultValue 'left'
         */
        this.iconPosition = 'left';
        /**
         * Label text displayed inside the button.
         */
        this.label = '';
        /**
         * Sets size for the component.
         * @defaultValue 'base'
         */
        this.size = 'base';
        /**
         * Determines the button's variant for different styles.
         */
        this.variant = 'neutral';
        /**
         * The value of the type toggle button.
         */
        this.type = 'filled';
        /**
         * Whether the button is selected.
         */
        this.selected = false;
        this.handleClick = (e) => {
            if (this.disabled) {
                e.preventDefault();
                e.stopPropagation();
                return;
            }
            this.tkToggle.emit({ value: this.value, selected: !this.selected });
        };
    }
    getIconColor() {
        if (this.disabled)
            return 'var(--icon-sub-base)';
        if (this.selected) {
            if (this.type === 'filled' || this.type === 'raised')
                return 'var(--static-white)';
            if (this.variant === 'neutral')
                return 'var(--icon-darkest)';
            if (this.variant === 'primary')
                return 'var(--primary-base)';
        }
        return null;
    }
    createIcons() {
        return renderIcons(this.icon, {
            size: this.size,
            variant: 'neutral',
            additionalProps: { color: this.getIconColor() },
            dataTestid: this.dataTestid,
        }, this.iconPosition);
    }
    renderLabel() {
        var _a;
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            return (h("span", { class: classNames('tk-toggle-button-label', this.size), "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label));
        }
        return null;
    }
    render() {
        const rootClasses = classNames('tk-toggle-button', this.variant, this.type, this.size, {
            rounded: this.rounded,
            selected: this.selected && !this.disabled,
            disabled: this.disabled,
        });
        const icon = this.createIcons();
        return (h("button", { key: 'eb6b89c8ba9d6df6d97625d0142a0e9e73d827ae', class: rootClasses, disabled: this.disabled, onClick: this.handleClick, "data-testid": getDataTestId(this.dataTestid, 'button') }, icon.leftIcon, this.renderLabel(), icon.rightIcon));
    }
    static get is() { return "tk-toggle-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-toggle-button.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-toggle-button.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disables the button, preventing user interaction."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon name to be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "iconPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'left'"
                        }],
                    "text": "Defines the position of the icon."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon-position",
                "defaultValue": "'left'"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label text displayed inside the button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label",
                "defaultValue": "''"
            },
            "rounded": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Makes the button round with an icon-only style."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rounded"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'neutral'",
                    "resolved": "\"neutral\" | \"primary\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines the button's variant for different styles."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'neutral'"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'filled' | 'outlined' | 'text' | 'raised' | 'filled-light'",
                    "resolved": "\"filled\" | \"filled-light\" | \"outlined\" | \"raised\" | \"text\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The value of the type toggle button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'filled'"
            },
            "value": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the toggle button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            },
            "selected": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the button is selected."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "selected",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get events() {
        return [{
                "method": "tkToggle",
                "name": "tk-toggle",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the toggle button is toggled."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
}
