// @takeoff-ui/core@0.11.9/dist/collection/components/tk-chips/tk-chips.js
import { h, Host } from "@stencil/core";
import classNames from "classnames";
import { getIconElementProps, renderIcons } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The TkChip component is basically a simple UI block entity, representing for example more advanced underlying data, such as a contact, in a compact way. Chips can contain entities such as an avatar, text or an icon, optionally having a pointer too.
 * @react `import { TkChips } from '@takeoff-ui/react'`
 * @vue `import { TkChips } from '@takeoff-ui/vue'`
 * @angular `import { TkChips } from '@takeoff-ui/angular'`
 */
export class TkChips {
    constructor() {
        /**
         * The disabled status.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * This field specifies the design type of the component.
         * @defaultValue filled
         */
        this.type = 'filled';
        /**
         * This property determines whether the chip component is removable.
         * @defaultValue false
         */
        this.removable = false;
        /**
         * Determines whether the chip automatically removes itself when the close button is clicked.
         * @defaultValue true
         */
        this.autoSelfDestroy = true;
        /**
         * Sets size for the component.
         * @defaultValue 'base'
         */
        this.size = 'base';
        /**
         * The variant of the chip for styling.
         * @defaultValue 'primary'
         */
        this.variant = 'primary';
        /**
         * The position of the icon relative to the label.
         * @defaultValue 'left'
         */
        this.iconPosition = 'left';
        /**
         * Custom style to apply to the chip component.
         */
        this.containerStyle = null;
        /**
         * If true, the chip will take the full width of its container.
         * @defaultValue false
         */
        this.fullWidth = false;
    }
    componentWillLoad() {
        if (this.value == null)
            this.value = this.label;
    }
    handleClick() {
        var _a;
        this.tkRemove.emit(this.value);
        if (this.autoSelfDestroy)
            (_a = this.el) === null || _a === void 0 ? void 0 : _a.remove();
    }
    handleKeyDown(e) {
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            this.handleClick();
        }
    }
    render() {
        const rootClasses = classNames('tk-chips', this.variant, this.size, this.type, {
            removable: this.removable,
            disabled: this.disabled,
        });
        const { leftIcon, rightIcon } = renderIcons(this.icon, {
            variant: this.variant,
            size: this.size === 'large' ? 'medium' : this.size,
            additionalProps: { color: this.disabled ? 'var(--icon-sub-base)' : this.type === 'filled' ? 'var(--static-white)' : undefined },
            dataTestid: this.dataTestid,
        }, this.iconPosition);
        return (h(Host, { key: 'd2a7436dea2a7ac1a351ff911eeb32aaf54b74f9', class: { 'full-width': this.fullWidth } }, h("div", { key: 'e7e1394976d5ffdf30db28a55c004b18c002aea8', class: rootClasses, style: this.containerStyle, "data-testid": getDataTestId(this.dataTestid, 'container') }, leftIcon, this.label, rightIcon, this.removable && (h("tk-icon", Object.assign({ key: '5ace9ba540c583e82c288c933df7733c50ccf800' }, getIconElementProps('close', {
            variant: this.variant,
            color: this.disabled ? 'var(--icon-sub-base)' : this.type === 'filled' ? 'var(--static-white)' : undefined,
            size: this.size === 'large' ? 'medium' : this.size,
            onClick: () => this.handleClick(),
            onKeyDown: (e) => this.handleKeyDown(e),
            tabIndex: this.disabled ? -1 : 0,
        }), { dataTestid: getDataTestId(this.dataTestid, 'remove-icon') }))))));
    }
    static get is() { return "tk-chips"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-chips.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-chips.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "The disabled status."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'filled' | 'filledlight' | 'outlined' | 'avatar'",
                    "resolved": "\"avatar\" | \"filled\" | \"filledlight\" | \"outlined\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "filled"
                        }],
                    "text": "This field specifies the design type of the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'filled'"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon name to be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The label to display inside the chip."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "removable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "This property determines whether the chip component is removable."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "removable",
                "defaultValue": "false"
            },
            "autoSelfDestroy": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Determines whether the chip automatically removes itself when the close button is clicked."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "auto-self-destroy",
                "defaultValue": "true"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'secondary' | 'neutral' | 'info' | 'success' | 'danger' | 'warning' | 'verified'",
                    "resolved": "\"danger\" | \"info\" | \"neutral\" | \"primary\" | \"secondary\" | \"success\" | \"verified\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'primary'"
                        }],
                    "text": "The variant of the chip for styling."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'primary'"
            },
            "value": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "this.label"
                        }],
                    "text": "The value of the chips"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            },
            "iconPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'left'"
                        }],
                    "text": "The position of the icon relative to the label."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon-position",
                "defaultValue": "'left'"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom style to apply to the chip component."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "fullWidth": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If true, the chip will take the full width of its container."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "full-width",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get events() {
        return [{
                "method": "tkRemove",
                "name": "tk-remove",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "When an element is deleted, it is triggered. It returns the label."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
}
