// @takeoff-ui/core@0.11.9/dist/collection/components/tk-drawer/tk-drawer.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkDrawer` is a container component displayed as an overlay. It supports various features such as different header and footer types, multiple variants, and flexible positioning, making it suitable for a wide range of use cases.
 * @slot container - Custom container template.
 * @slot header - Custom header template.
 * @slot header-action - Custom actions template of header.
 * @slot content - Custom inner body template.
 * @slot footer - Custom footer template.
 * @react `import { TkDrawer } from '@takeoff-ui/react'`
 * @vue `import { TkDrawer } from '@takeoff-ui/vue'`
 * @angular `import { TkDrawer } from '@takeoff-ui/angular'`
 */
export class TkDrawer {
    constructor() {
        this.containerVisible = false;
        this.isOpen = false;
        /**
         * Controls whether the drawer is open or closed
         * @defaultValue false
         */
        this.open = false;
        /**
         * Determines the position of the drawer
         * @defaultValue right
         */
        this.position = 'right';
        /**
         * Controls whether the backdrop is shown
         * @defaultValue false
         */
        this.hideBackdrop = false;
        /**
         * Appearance of the mask
         * @defaultValue base
         */
        this.maskVariant = 'base';
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        /**
         * Prevents the drawer from being dismissed by clicking outside
         * @defaultValue false
         */
        this.preventDismiss = false;
        /**
         * Controls whether to hide the close icon
         * @defaultValue false
         */
        this.hideCloseIcon = false;
        /**
         * Controls whether to unblock scrolling when the drawer is open
         * @defaultValue false
         */
        this.unblockScroll = false;
        /**
         * The type of the header
         * @defaultValue 'basic'
         */
        this.headerType = 'basic';
        /**
         * The mode of the footer
         * @defaultValue 'basic'
         */
        this.footerType = 'basic';
        this.handleOverlayClick = () => {
            if (!this.preventDismiss) {
                this.tkDrawerClose.emit();
            }
        };
        this.handleCloseButtonClick = () => {
            this.tkDrawerClose.emit();
        };
    }
    openChanged(newValue) {
        if (newValue) {
            requestAnimationFrame(() => {
                this.isOpen = true;
                this.onEnter();
            });
        }
        else {
            this.isOpen = false;
            this.onLeave();
        }
        this.containerVisible = newValue;
        this.tkDrawerChange.emit(newValue);
    }
    componentWillLoad() {
        if (this.open) {
            this.containerVisible = true;
            requestAnimationFrame(() => {
                this.isOpen = true;
                this.onEnter();
            });
        }
    }
    /**
     * Opens the drawer by emitting a tk-drawer-open event
     * Parent components should listen for this event and update the open prop
     */
    async show() {
        this.tkDrawerOpen.emit();
    }
    /**
     * Closes the drawer by emitting a tk-drawer-close event
     * Parent components should listen for this event and update the open prop
     */
    async close() {
        this.tkDrawerClose.emit();
    }
    blockBodyScroll() {
        if (document.body.style.overflow !== 'hidden') {
            this.bodyOverflow = document.body.style.overflow;
            this.bodyPaddingRight = document.body.style.paddingRight;
            const scrollbarWidth = window.innerWidth - document.documentElement.clientWidth;
            document.body.style.paddingRight = `${scrollbarWidth}px`;
            document.body.style.overflow = 'hidden';
        }
    }
    unblockBodyScroll() {
        document.body.style.overflow = this.bodyOverflow;
        document.body.style.paddingRight = this.bodyPaddingRight;
    }
    getTransformValue() {
        if (!this.isOpen) {
            switch (this.position) {
                case 'left':
                    return 'translateX(-100%)';
                case 'right':
                    return 'translateX(100%)';
                case 'top':
                    return 'translateY(-100%)';
                case 'bottom':
                    return 'translateY(100%)';
                default:
                    return '';
            }
        }
        return 'translate(0, 0)';
    }
    onEnter() {
        if (!this.unblockScroll && !this.hideBackdrop) {
            this.blockBodyScroll();
        }
        this.tkDrawerEnter.emit();
        setTimeout(() => {
            this.tkDrawerOpen.emit();
        }, 300);
    }
    onLeave() {
        if (!this.unblockScroll && !this.hideBackdrop) {
            this.unblockBodyScroll();
        }
        this.tkDrawerLeave.emit();
        // setTimeout(() => {
        this.containerVisible = false;
        this.tkDrawerClose.emit();
        // }, 300);
    }
    createHeader() {
        const hasHeaderSlot = !!this.el.querySelector('[slot="header"]');
        const hasHeaderActionsSlot = !!this.el.querySelector('[slot="header-actions"]');
        return (h("div", { class: classNames('tk-drawer-header', `tk-drawer-header-${this.headerType}`), "data-testid": getDataTestId(this.dataTestid, 'header') }, hasHeaderSlot ? (h("slot", { name: "header" })) : this.header ? (h("span", { class: "tk-drawer-header-label", "data-testid": getDataTestId(this.dataTestid, 'label') }, this.header)) : null, hasHeaderActionsSlot ? (h("slot", { name: "header-actions" })) : (!this.hideCloseIcon && (h("tk-button", { class: "tk-drawer-close", variant: "neutral", icon: "close", type: "text", "onTk-click": this.handleCloseButtonClick, dataTestid: getDataTestId(this.dataTestid, 'close-button') })))));
    }
    createFooter() {
        const hasFooterSlot = !!this.el.querySelector('[slot="footer"]');
        if (hasFooterSlot) {
            return (h("div", { class: classNames('tk-drawer-footer', `tk-drawer-footer-${this.footerType}`), "data-testid": getDataTestId(this.dataTestid, 'footer') }, h("slot", { name: "footer" })));
        }
        return null;
    }
    createDrawer() {
        const drawerClass = classNames('tk-drawer', `tk-drawer-${this.position}`, {
            'tk-drawer-open': this.isOpen,
        });
        const style = Object.assign(Object.assign({}, (this.containerStyle && Object.assign({}, this.containerStyle))), { transform: this.getTransformValue() });
        return (h("div", { class: drawerClass, style: style, "data-testid": getDataTestId(this.dataTestid, 'panel') }, h("div", { class: "tk-drawer-content", "data-testid": getDataTestId(this.dataTestid, 'content') }, h("slot", { name: "container" }, this.createHeader(), h("div", { class: "tk-drawer-body", "data-testid": getDataTestId(this.dataTestid, 'body') }, h("slot", { name: "content" })), this.createFooter()))));
    }
    renderMask() {
        const maskClasses = classNames('tk-drawer-mask', `tk-drawer-mask-${this.maskVariant}`, {
            'tk-drawer-visible': this.containerVisible,
            'tk-drawer-mask-hidden': this.hideBackdrop,
        });
        return (h("div", { class: maskClasses, "data-testid": getDataTestId(this.dataTestid, 'mask') }, h("div", { class: "tk-drawer-overlay", onClick: () => this.handleOverlayClick(), "data-testid": getDataTestId(this.dataTestid, 'overlay') }), this.createDrawer()));
    }
    render() {
        return this.containerVisible && this.renderMask();
    }
    static get is() { return "tk-drawer"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-drawer.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-drawer.css"]
        };
    }
    static get properties() {
        return {
            "open": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the drawer is open or closed"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "open",
                "defaultValue": "false"
            },
            "position": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right' | 'top' | 'bottom' | 'full-screen'",
                    "resolved": "\"bottom\" | \"full-screen\" | \"left\" | \"right\" | \"top\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "right"
                        }],
                    "text": "Determines the position of the drawer"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "position",
                "defaultValue": "'right'"
            },
            "hideBackdrop": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the backdrop is shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-backdrop",
                "defaultValue": "false"
            },
            "maskVariant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'lightest' | 'light' | 'base' | 'dark' | 'darkest'",
                    "resolved": "\"base\" | \"dark\" | \"darkest\" | \"light\" | \"lightest\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "base"
                        }],
                    "text": "Appearance of the mask"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mask-variant",
                "defaultValue": "'base'"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of container element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "preventDismiss": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Prevents the drawer from being dismissed by clicking outside"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "prevent-dismiss",
                "defaultValue": "false"
            },
            "hideCloseIcon": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether to hide the close icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-close-icon",
                "defaultValue": "false"
            },
            "unblockScroll": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether to unblock scrolling when the drawer is open"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "unblock-scroll",
                "defaultValue": "false"
            },
            "header": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Text to display in the drawer header"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header"
            },
            "headerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light' | 'dark' | 'primary'",
                    "resolved": "\"basic\" | \"dark\" | \"divided\" | \"light\" | \"primary\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'basic'"
                        }],
                    "text": "The type of the header"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header-type",
                "defaultValue": "'basic'"
            },
            "footerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light'",
                    "resolved": "\"basic\" | \"divided\" | \"light\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'basic'"
                        }],
                    "text": "The mode of the footer"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "footer-type",
                "defaultValue": "'basic'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "containerVisible": {},
            "isOpen": {}
        };
    }
    static get events() {
        return [{
                "method": "tkDrawerOpen",
                "name": "tk-drawer-open",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the drawer is opened"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkDrawerClose",
                "name": "tk-drawer-close",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the drawer is closed"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkDrawerEnter",
                "name": "tk-drawer-enter",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the drawer starts to enter"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkDrawerLeave",
                "name": "tk-drawer-leave",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the drawer starts to leave"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkDrawerChange",
                "name": "tk-drawer-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the drawer's open state changes"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "show": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Opens the drawer by emitting a tk-drawer-open event\nParent components should listen for this event and update the open prop",
                    "tags": []
                }
            },
            "close": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Closes the drawer by emitting a tk-drawer-close event\nParent components should listen for this event and update the open prop",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "open",
                "methodName": "openChanged"
            }];
    }
}
