// @takeoff-ui/core@0.11.9/dist/collection/components/tk-chart/tk-chart.js
import { h } from "@stencil/core";
import Chart from "chart.js/auto";
import { getDefaultOptionsForType } from "./defaults";
import { merge } from "lodash-es";
import classNames from "classnames";
/**
 * The TkChart component allows users to visualize data in various chart formats using Chart.js.
 * @react `import { TkChart } from '@takeoff-ui/react'`
 * @vue `import { TkChart } from '@takeoff-ui/vue'`
 * @angular `import { TkChart } from '@takeoff-ui/angular'`
 */
export class TkChart {
    constructor() {
        /**
         * The type of chart to render
         */
        this.type = 'bar';
        /**
         * Width of the chart container
         */
        this.width = null;
        /**
         * Height of the chart container in pixels
         */
        this.height = null;
        /**
         * Custom plugins to use with chart
         */
        this.plugins = [];
    }
    handleDataChange() {
        this.updateChart();
    }
    handleOptionsChange() {
        this.mergeOptions();
        this.updateChart();
    }
    handleTypeChange() {
        this.destroyChart();
        this.mergeOptions();
        this.initChart();
    }
    componentWillLoad() {
        this.mergeOptions();
    }
    componentDidLoad() {
        this.initChart();
    }
    disconnectedCallback() {
        this.destroyChart();
    }
    /**
     * Get the chart instance
     */
    async getChart() {
        return this.chartInstance;
    }
    /**
     * Get the canvas element
     */
    async getCanvas() {
        return this.chartRef;
    }
    /**
     * Get base64 image of the chart
     */
    async getBase64Image() {
        var _a;
        return (_a = this.chartInstance) === null || _a === void 0 ? void 0 : _a.toBase64Image();
    }
    /**
     * Refresh the chart
     */
    async refresh() {
        if (this.chartInstance) {
            this.chartInstance.update();
        }
    }
    mergeOptions() {
        const typeDefaults = getDefaultOptionsForType(this.type);
        // Merge with user options
        this.internalOptions = merge({}, typeDefaults, this.options);
    }
    initChart() {
        if (this.chartRef) {
            const ctx = this.chartRef.getContext('2d');
            if (ctx) {
                this.chartInstance = new Chart(ctx, {
                    type: this.type,
                    data: this.data,
                    options: this.internalOptions,
                    plugins: this.plugins,
                });
            }
        }
    }
    updateChart() {
        if (this.chartInstance) {
            // Don't directly assign to type as it's not supported
            this.chartInstance.data = this.data;
            this.chartInstance.options = this.internalOptions;
            this.chartInstance.update();
        }
    }
    destroyChart() {
        if (this.chartInstance) {
            this.chartInstance.destroy();
            this.chartInstance = undefined;
        }
    }
    render() {
        var _a, _b, _c;
        const title = (_c = (_b = (_a = this.options) === null || _a === void 0 ? void 0 : _a.plugins) === null || _b === void 0 ? void 0 : _b.title) === null || _c === void 0 ? void 0 : _c.text;
        const accessibilityLabel = this.accessibilityLabel || title || 'Chart';
        return (h("div", { key: '11d9595eb84eb5097ea41ad7a6139ae59494ab8b', class: classNames('tk-chart-container', `tk-chart-${this.type}`), style: {
                width: this.width,
                height: `${this.height}px`,
            } }, h("canvas", { key: '55c7206b3939aa0c3a84194f4958393503f8f0df', ref: el => (this.chartRef = el), role: "img", "aria-label": accessibilityLabel })));
    }
    static get is() { return "tk-chart"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-chart.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-chart.css"]
        };
    }
    static get properties() {
        return {
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "ChartType",
                    "resolved": "\"bar\" | \"bubble\" | \"doughnut\" | \"line\" | \"pie\" | \"polarArea\" | \"radar\" | \"scatter\"",
                    "references": {
                        "ChartType": {
                            "location": "import",
                            "path": "chart.js/auto",
                            "id": "../../node_modules/.pnpm/chart.js@4.5.1/node_modules/chart.js/auto/auto.d.ts::ChartType",
                            "referenceLocation": "ChartType"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The type of chart to render"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'bar'"
            },
            "data": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ChartData",
                    "resolved": "ChartData<keyof ChartTypeRegistry, (number | [number, number] | Point | BubbleDataPoint)[], unknown>",
                    "references": {
                        "ChartData": {
                            "location": "import",
                            "path": "chart.js/auto",
                            "id": "../../node_modules/.pnpm/chart.js@4.5.1/node_modules/chart.js/auto/auto.d.ts::ChartData",
                            "referenceLocation": "ChartData"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Chart data prop is used to define chart data supported by the Chart.js library. With this prop, you can specify chart data described in the Chart.js documentation (https://www.chartjs.org/docs/latest/general/data-structures.html)."
                },
                "getter": false,
                "setter": false
            },
            "options": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ChartOptions",
                    "resolved": "CoreChartOptions<keyof ChartTypeRegistry> & ElementChartOptions<keyof ChartTypeRegistry> & PluginChartOptions<keyof ChartTypeRegistry> & DatasetChartOptions<keyof ChartTypeRegistry> & ScaleChartOptions<keyof ChartTypeRegistry>",
                    "references": {
                        "ChartOptions": {
                            "location": "import",
                            "path": "chart.js/auto",
                            "id": "../../node_modules/.pnpm/chart.js@4.5.1/node_modules/chart.js/auto/auto.d.ts::ChartOptions",
                            "referenceLocation": "ChartOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Chart options prop is used to define chart options supported by the Chart.js library. With this prop, you can specify any chart options described in the Chart.js documentation (https://www.chartjs.org/docs/latest/general/options.html)."
                },
                "getter": false,
                "setter": false
            },
            "width": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Width of the chart container"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "width",
                "defaultValue": "null"
            },
            "height": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Height of the chart container in pixels"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "height",
                "defaultValue": "null"
            },
            "plugins": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "any[]",
                    "resolved": "any[]",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom plugins to use with chart"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "accessibilityLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Accessibility label for the chart"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "accessibility-label"
            }
        };
    }
    static get states() {
        return {
            "internalOptions": {}
        };
    }
    static get methods() {
        return {
            "getChart": {
                "complexType": {
                    "signature": "() => Promise<any>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<any>"
                },
                "docs": {
                    "text": "Get the chart instance",
                    "tags": []
                }
            },
            "getCanvas": {
                "complexType": {
                    "signature": "() => Promise<HTMLCanvasElement | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLCanvasElement": {
                            "location": "global",
                            "id": "global::HTMLCanvasElement"
                        }
                    },
                    "return": "Promise<HTMLCanvasElement>"
                },
                "docs": {
                    "text": "Get the canvas element",
                    "tags": []
                }
            },
            "getBase64Image": {
                "complexType": {
                    "signature": "() => Promise<string | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<string>"
                },
                "docs": {
                    "text": "Get base64 image of the chart",
                    "tags": []
                }
            },
            "refresh": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Refresh the chart",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "data",
                "methodName": "handleDataChange"
            }, {
                "propName": "options",
                "methodName": "handleOptionsChange"
            }, {
                "propName": "type",
                "methodName": "handleTypeChange"
            }];
    }
}
