// @takeoff-ui/core@0.11.9/dist/esm/tk-accordion.entry.js
import { r as registerInstance, c as createEvent, h, d as getElement } from './index-D4ALLk7q.js';
import { g as getDataTestId } from './test-id-utils-BDMnqEGL.js';
import { i as isEqual } from './isEqual-CxgWR-W5.js';
import './_Uint8Array-B3_jBex_.js';

const tkAccordionCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-accordion{display:block}`;

const TkAccordion = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.tkActiveIndexChange = createEvent(this, "tk-active-index-change", 7);
        this.tkAccordionItemSelected = createEvent(this, "tk-accordion-item-selected", 7);
        this.internalActiveIndex = [];
        /**
         * Allows multiple accordion items to be expanded simultaneously.
         * @defaultValue false
         */
        this.allowMultiple = false;
        /**
         * Sets the position of opening and closing chevrons.
         * @defaultValue 'right'
         */
        this.arrowPosition = 'right';
        /**
         * Sets the expand icon
         * @defaultValue 'keyboard_arrow_down'
         */
        this.expandIcon = 'keyboard_arrow_down';
        /**
         * Sets the collapse icon
         * @defaultValue 'keyboard_arrow_up'
         */
        this.collapseIcon = 'keyboard_arrow_up';
        /**
         * Whether to hide the arrow icons.
         * @defaultValue false
         */
        this.hideArrows = false;
        /**
         * Sets accordion style for the component.
         * @defaultValue 'grouped'
         */
        this.type = 'grouped';
        /**
         * Controls the display mode of the accordion component.
         * @defaultValue 'default'
         */
        this.mode = 'default';
    }
    internalActiveIndexChanged(newValue, oldValue) {
        // Exit early if the active index hasn't changed
        if (isEqual(newValue, oldValue))
            return;
        this.updateActiveIndex();
        this.getAccordionItems().forEach((item, index) => {
            const itemKey = this.getItemKey(item, index);
            const isActive = newValue.some(activeIndex => activeIndex === itemKey);
            item.active = isActive;
        });
    }
    activeIndexChanged() {
        const normalized = this.normalizeActiveIndex();
        if (isEqual(normalized, this.internalActiveIndex))
            return;
        this.internalActiveIndex = normalized;
    }
    componentDidLoad() {
        this.validateItemKeylessActiveIndex();
        this.validateActiveProps();
        this.validateControlType();
        this.validateKeyType();
        this.initInternalActiveIndex();
        this.initEventListeners();
    }
    initInternalActiveIndex() {
        // if prop activeIndex is set, use it
        if (this.hasActiveIndex())
            return (this.internalActiveIndex = this.normalizeActiveIndex());
        // else if accordion items have active prop, use them
        this.getAccordionItems().forEach((item, index) => {
            if (item.active)
                this.internalActiveIndex = [...this.internalActiveIndex, this.getItemKey(item, index)];
        });
    }
    initEventListeners() {
        this.getAccordionItems().forEach((item, index) => {
            const itemKey = this.getItemKey(item, index);
            // listen to active change
            item.addEventListener('tk-active-change', (e) => {
                this.handleItemActiveChange(itemKey, e.detail);
                item.active = e.detail;
                if (!this.allowMultiple && e.detail) {
                    this.getAccordionItems()
                        .filter(child => child !== item)
                        .forEach(otherItem => {
                        otherItem.active = false;
                    });
                }
            });
        });
    }
    // Validation Logic
    validateActiveProps() {
        const activeCount = this.getAccordionItems().reduce((count, item) => (item.active ? count + 1 : count), 0);
        if (activeCount > 1 && !this.allowMultiple) {
            console.error('TkAccordion: Multiple accordion items are set to active while allowMultiple is false. This may lead to unexpected behavior.');
        }
    }
    validateControlType() {
        const hasActiveAccordionItems = this.getAccordionItems().some(item => item.active !== undefined);
        if (this.hasActiveIndex() && hasActiveAccordionItems)
            console.error('TkAccordion: Accordion cannot have both activeIndex and active accordion items');
    }
    validateKeyType() {
        var _a, _b;
        const type = typeof ((_b = (_a = this.getAccordionItems().find(item => item.itemKey !== undefined)) === null || _a === void 0 ? void 0 : _a.itemKey) !== null && _b !== void 0 ? _b : 0);
        const allItemsHaveSameType = this.getAccordionItems().every(item => item.itemKey === undefined || typeof item.itemKey === type);
        let activeIndexHasSameType = true;
        if (this.hasActiveIndex())
            activeIndexHasSameType = Array.isArray(this.activeIndex) ? this.activeIndex.every(item => typeof item === type) : typeof this.activeIndex === type;
        if (!allItemsHaveSameType || !activeIndexHasSameType)
            console.error('TkAccordion: Accordion item keys must be of the same type');
    }
    validateItemKeylessActiveIndex() {
        if (!this.hasActiveIndex())
            return;
        const hasKeylessItems = this.getAccordionItems().every(item => item.itemKey === undefined);
        const activeIndexType = Array.isArray(this.activeIndex) ? this.activeIndex.map(item => typeof item) : [typeof this.activeIndex];
        if (hasKeylessItems && activeIndexType.some(type => type !== 'number')) {
            console.error('TkAccordion: When using keyless accordion items, activeIndex must be of type number or number array.');
        }
    }
    hasActiveIndex() {
        return this.activeIndex || this.activeIndex === 0;
    }
    normalizeActiveIndex() {
        if (!this.activeIndex && this.activeIndex !== 0)
            return [];
        if (!Array.isArray(this.activeIndex))
            return [this.activeIndex];
        if (this.allowMultiple)
            return this.activeIndex;
        const lastItem = this.activeIndex.at(-1);
        return lastItem ? [lastItem] : [];
    }
    getActiveIndex() {
        return this.allowMultiple ? this.internalActiveIndex : this.internalActiveIndex[this.internalActiveIndex.length - 1];
    }
    updateActiveIndex() {
        const activeIndex = this.getActiveIndex();
        if (!isEqual(activeIndex, this.activeIndex))
            this.tkActiveIndexChange.emit(activeIndex);
    }
    handleItemActiveChange(itemKey, active) {
        if (active && !this.internalActiveIndex.includes(itemKey)) {
            this.internalActiveIndex = [...this.internalActiveIndex, itemKey];
        }
        if (!active && this.internalActiveIndex.includes(itemKey)) {
            this.internalActiveIndex = this.internalActiveIndex.filter(activeIndex => activeIndex !== itemKey);
        }
        this.tkAccordionItemSelected.emit({
            index: itemKey,
            active,
        });
    }
    getAccordionItems() {
        return Array.from(this.el.querySelectorAll('tk-accordion-item')).filter(child => child.parentElement === this.el);
    }
    getItemKey(accordionItem, index) {
        var _a;
        return (_a = accordionItem.itemKey) !== null && _a !== void 0 ? _a : index;
    }
    render() {
        return (h("div", { key: '8a88bf333637906c2bd4218f3015901c5cdbd7aa', class: "tk-accordion", "data-testid": getDataTestId(this.dataTestid, 'container') }, h("slot", { key: '6bbb5335b02b07138985ed6b883a03a5cd1c4254' })));
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "internalActiveIndex": [{
                "internalActiveIndexChanged": 0
            }],
        "activeIndex": [{
                "activeIndexChanged": 0
            }]
    }; }
};
TkAccordion.style = tkAccordionCss();

export { TkAccordion as tk_accordion };
