// @takeoff-ui/core@0.11.9/dist/esm/tk-carousel.entry.js
import { r as registerInstance, c as createEvent, h, d as getElement } from './index-D4ALLk7q.js';
import { c as classNames } from './index-D7YaJBkB.js';
import { g as getDataTestId } from './test-id-utils-BDMnqEGL.js';

const tkCarouselCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block;width:100%;height:100%}.tk-carousel-overlay{overflow:hidden}.tk-carousel-overlay .tk-carousel-items-container{display:flex;gap:16px;transition:transform 0.3s ease}.tk-carousel-overlay .tk-carousel-items-container ::slotted(*){flex:0 0 calc((100% - (var(--items-per-view, 1) - 1) * 16px) / var(--items-per-view, 1));min-width:0;object-fit:cover;box-sizing:border-box}.tk-carousel{position:relative;box-sizing:border-box}.tk-carousel.vertical{height:100%}.tk-carousel.vertical .tk-carousel-overlay{height:100%}.tk-carousel.vertical .tk-carousel-items-container{flex-direction:column;width:100%;height:100%}.tk-carousel.vertical .tk-carousel-items-container ::slotted(*){max-height:calc((100% - (var(--items-per-view, 1) - 1) * 16px) / var(--items-per-view, 1))}.tk-carousel .tk-carousel-navigation{display:flex;align-items:center;justify-content:center;position:absolute;gap:20px;z-index:10}.tk-carousel .tk-carousel-navigation.vertical-navigation{flex-direction:column}.tk-carousel .tk-carousel-navigation.vertical-navigation .tk-carousel-indicators{flex-direction:column}.tk-carousel .tk-carousel-indicators{display:flex;align-items:center;gap:2px}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator{padding:4px;cursor:pointer}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator .tk-carousel-indicator-dot{width:6px;height:6px;background-color:var(--background-light);border-radius:50%;transition:all 0.3s ease}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator .tk-carousel-indicator-dot:hover{background-color:var(--primary-light)}.tk-carousel .tk-carousel-indicators .tk-carousel-indicator .tk-carousel-indicator-dot.active{background-color:var(--primary-base);width:8px;height:8px}.tk-carousel .tk-carousel-indicators .player-button{cursor:pointer}.tk-carousel .prev-button,.tk-carousel .next-button{z-index:10;display:flex;align-items:center;justify-content:center}.tk-carousel.inside.bottom .tk-carousel-navigation{bottom:16px;left:50%;transform:translateX(-50%)}.tk-carousel.inside.top .tk-carousel-navigation{top:16px;left:50%;transform:translateX(-50%)}.tk-carousel.inside.left .tk-carousel-navigation{left:16px;top:50%;transform:translateY(-50%)}.tk-carousel.inside.right .tk-carousel-navigation{right:16px;top:50%;transform:translateY(-50%)}.tk-carousel.inside.distributed .prev-button,.tk-carousel.inside.distributed .next-button{position:absolute;top:50%;transform:translateY(-50%)}.tk-carousel.inside.distributed .prev-button{left:16px}.tk-carousel.inside.distributed .next-button{right:16px}.tk-carousel.inside.distributed .tk-carousel-indicators{position:absolute;bottom:16px;left:50%;transform:translateX(-50%)}.tk-carousel.outside.bottom.show-arrows,.tk-carousel.outside.bottom.show-indicators{padding-bottom:48px}.tk-carousel.outside.bottom .tk-carousel-navigation{bottom:0px;left:50%;transform:translateX(-50%)}.tk-carousel.outside.top.show-arrows,.tk-carousel.outside.top.show-indicators{padding-top:48px}.tk-carousel.outside.top .tk-carousel-navigation{top:0px;left:50%;transform:translateX(-50%)}.tk-carousel.outside.left.show-arrows,.tk-carousel.outside.left.show-indicators{padding-left:48px}.tk-carousel.outside.left .tk-carousel-navigation{left:0px;top:50%;transform:translateY(-50%)}.tk-carousel.outside.right.show-arrows,.tk-carousel.outside.right.show-indicators{padding-right:48px}.tk-carousel.outside.right .tk-carousel-navigation{right:0px;top:50%;transform:translateY(-50%)}.tk-carousel.outside.distributed.show-arrows{padding-inline:48px}.tk-carousel.outside.distributed.show-indicators{padding-bottom:32px}.tk-carousel.outside.distributed .prev-button,.tk-carousel.outside.distributed .next-button{position:absolute;top:50%;transform:translateY(-50%)}.tk-carousel.outside.distributed .prev-button{left:0px}.tk-carousel.outside.distributed .next-button{right:0px}.tk-carousel.outside.distributed .tk-carousel-indicators{position:absolute;bottom:0px;left:50%;transform:translateX(-50%)}`;

const TkCarousel = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.tkChange = createEvent(this, "tk-change", 7);
        this.items = [];
        this.activeIndex = 0;
        this.totalItems = 0;
        /**
         * Controls whether the carousel indicators are shown
         * @defaultValue true
         */
        this.showIndicators = true;
        /**
         * Controls whether the navigation arrows are shown
         * @defaultValue true
         */
        this.showArrows = true;
        /**
         * Controls whether the carousel should autoplay
         * @defaultValue false
         */
        this.autoplay = false;
        /**
         * Controls the interval of the autoplay in milliseconds
         * @defaultValue 3000
         */
        this.autoplayDelay = 3000;
        /**
         * Controls whether it should loop back to the start after reaching the end
         * @defaultValue true
         */
        this.circular = true;
        /**
         * Placement of the navigation indicators
         * @defaultValue 'inside'
         */
        this.navigationPlacement = 'inside';
        /**
         * Position of the navigation indicators.
         * For horizontal orientation: 'distributed' | 'top' | 'bottom'
         * For vertical orientation: 'left' | 'right'
         * @defaultValue 'distributed'
         */
        this.navigationPosition = 'distributed';
        /**
         * Controls whether the pause/play button is shown
         * @defaultValue false
         */
        this.showPlayerButton = false;
        /**
         * Number of items to show per view
         * @defaultValue 1
         */
        this.itemsPerView = 1;
        /**
         * Orientation of the carousel
         * @defaultValue 'horizontal'
         */
        this.orientation = 'horizontal';
        /**
         * Height of the carousel when orientation is vertical
         * @defaultValue '300px'
         */
        this.verticalViewHeight = '300px';
        this.startAutoplay = () => {
            if (!this.autoplay || this.totalItems <= this.itemsPerView)
                return;
            this.stopAutoplay();
            this.autoplayTimer = window.setInterval(() => {
                const lastStartingItem = this.totalItems - this.itemsPerView;
                if (this.activeIndex >= lastStartingItem) {
                    if (this.circular) {
                        this.changeItem(0);
                    }
                    else {
                        this.stopAutoplay();
                    }
                }
                else {
                    this.changeItem(this.activeIndex + 1);
                }
            }, this.autoplayDelay);
        };
        this.stopAutoplay = () => {
            if (this.autoplayTimer) {
                clearInterval(this.autoplayTimer);
                this.autoplayTimer = undefined;
            }
        };
        this.handlePrevClick = () => {
            this.changeItem(this.activeIndex - 1);
            this.tkChange.emit(this.activeIndex);
        };
        this.handleNextClick = () => {
            this.changeItem(this.activeIndex + 1);
            this.tkChange.emit(this.activeIndex);
        };
        this.handleIndicatorClick = (index) => {
            this.changeItem(index);
        };
        this.handleKeyDown = (e) => {
            if (this.orientation === 'vertical') {
                if (e.key === 'ArrowUp') {
                    e.preventDefault();
                    this.handlePrevClick();
                }
                else if (e.key === 'ArrowDown') {
                    e.preventDefault();
                    this.handleNextClick();
                }
            }
            else {
                if (e.key === 'ArrowLeft') {
                    e.preventDefault();
                    this.handlePrevClick();
                }
                else if (e.key === 'ArrowRight') {
                    e.preventDefault();
                    this.handleNextClick();
                }
            }
        };
    }
    componentDidLoad() {
        var _a, _b;
        if (!this.slotElement) {
            this.slotElement = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('slot');
        }
        this.itemsContainer = (_b = this.el.shadowRoot) === null || _b === void 0 ? void 0 : _b.querySelector('.tk-carousel-items-container');
        this.el.tabIndex = 0;
        this.updateItemPosition();
        this.updateItems();
        this.startAutoplay();
        this.el.addEventListener('keydown', this.handleKeyDown);
    }
    disconnectedCallback() {
        this.stopAutoplay();
        this.el.removeEventListener('keydown', this.handleKeyDown);
    }
    /**
     * Returns a valid navigation position based on orientation.
     * Horizontal: 'distributed' | 'top' | 'bottom'
     * Vertical: 'left' | 'right'
     */
    get validNavigationPosition() {
        const horizontalPositions = ['distributed', 'top', 'bottom'];
        const verticalPositions = ['left', 'right'];
        if (this.orientation === 'vertical') {
            return verticalPositions.includes(this.navigationPosition) ? this.navigationPosition : 'right';
        }
        return horizontalPositions.includes(this.navigationPosition) ? this.navigationPosition : 'distributed';
    }
    updateItems() {
        var _a;
        const assigned = ((_a = this.slotElement) === null || _a === void 0 ? void 0 : _a.assignedElements()) || [];
        this.items = assigned.filter(el => !el.hasAttribute('slot'));
        this.totalItems = this.items.length;
    }
    updateItemPosition() {
        var _a;
        if (!this.itemsContainer)
            return;
        const itemsContainerElement = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.tk-carousel-overlay');
        if (!itemsContainerElement)
            return;
        const gapSize = 16;
        const totalGaps = Math.max(0, this.itemsPerView - 1);
        const totalGapSpace = totalGaps * gapSize;
        let containerSize;
        if (this.orientation === 'horizontal') {
            containerSize = itemsContainerElement.getBoundingClientRect().width;
        }
        else {
            this.el.style.height = String(this.verticalViewHeight);
            containerSize = Number.parseFloat(this.verticalViewHeight);
        }
        const itemSize = (containerSize - totalGapSpace) / this.itemsPerView;
        const translateValue = this.activeIndex * (itemSize + gapSize);
        if (this.orientation === 'vertical') {
            this.itemsContainer.style.transform = `translateY(-${translateValue}px)`;
        }
        else {
            this.itemsContainer.style.transform = `translateX(-${translateValue}px)`;
        }
    }
    changeItem(index) {
        if (!this.totalItems)
            return;
        const lastStartingItem = Math.max(0, this.totalItems - this.itemsPerView);
        if (this.circular) {
            if (index < 0) {
                index = lastStartingItem;
            }
            else if (index > lastStartingItem) {
                index = 0;
            }
        }
        else {
            if (index < 0)
                index = 0;
            if (index > lastStartingItem)
                index = lastStartingItem;
        }
        this.activeIndex = index;
        this.updateItemPosition();
        this.tkChange.emit(this.activeIndex);
    }
    createPrevButton() {
        if (!this.showArrows || this.items.length <= this.itemsPerView)
            return null;
        if (!this.circular && this.activeIndex === 0)
            return null;
        return (h("tk-button", { size: "small", class: "prev-button", icon: this.orientation === 'vertical' ? 'keyboard_arrow_up' : 'chevron_left', "onTk-click": this.handlePrevClick, dataTestid: getDataTestId(this.dataTestid, 'prev-button') }));
    }
    createNextButton() {
        const lastStartingItem = Math.max(0, this.totalItems - this.itemsPerView);
        if (!this.showArrows || this.items.length <= this.itemsPerView)
            return null;
        if (!this.circular && this.activeIndex >= lastStartingItem)
            return null;
        return (h("tk-button", { size: "small", class: "next-button", icon: this.orientation === 'vertical' ? 'keyboard_arrow_down' : 'chevron_right', "onTk-click": this.handleNextClick, dataTestid: getDataTestId(this.dataTestid, 'next-button') }));
    }
    createPlayerButtons() {
        if (!this.showPlayerButton || !this.autoplay)
            return null;
        const iconProps = {
            color: 'var(--background-light)',
            icon: this.autoplayTimer ? 'pause_circle' : 'play_circle',
            onClick: this.autoplayTimer ? this.stopAutoplay : this.startAutoplay,
        };
        return h("tk-icon", Object.assign({ class: "player-button", size: "small" }, iconProps, { dataTestid: getDataTestId(this.dataTestid, 'player-icon') }));
    }
    createIndicators() {
        if (!this.showIndicators || this.items.length <= this.itemsPerView)
            return null;
        const indicatorCount = Math.max(1, this.totalItems - this.itemsPerView + 1);
        return (h("div", { class: "tk-carousel-indicators", "data-testid": getDataTestId(this.dataTestid, 'indicators') }, this.createPlayerButtons(), Array.from({ length: indicatorCount }).map((_, index) => {
            const indicatorDataTestid = getDataTestId(this.dataTestid, 'indicator', index.toString());
            return (h("div", { class: "tk-carousel-indicator", onClick: () => this.handleIndicatorClick(index), "data-testid": getDataTestId(indicatorDataTestid) }, h("div", { class: classNames('tk-carousel-indicator-dot', { active: index === this.activeIndex }), "data-testid": getDataTestId(indicatorDataTestid, 'dot') })));
        })));
    }
    renderNavigationElement() {
        const navPosition = this.validNavigationPosition;
        return (h("div", { class: classNames('tk-carousel-navigation-holder', { 'tk-carousel-navigation': navPosition !== 'distributed' }, { 'vertical-navigation': navPosition === 'left' || navPosition === 'right' }), "data-testid": getDataTestId(this.dataTestid, 'navigation') }, this.createPrevButton(), this.createIndicators(), this.createNextButton()));
    }
    render() {
        const navPosition = this.validNavigationPosition;
        const rootClasses = classNames('tk-carousel', this.navigationPlacement, navPosition, { vertical: this.orientation === 'vertical' }, { 'show-arrows': this.showArrows }, { 'show-indicators': this.showIndicators });
        this.el.style.setProperty('--items-per-view', String(this.itemsPerView));
        return (h("div", { key: '77f46c60f3210b667dd41a7824f2193ffbff8f14', class: rootClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("div", { key: '2cd9391d7a28317ce9f6fdbd4a5c53ce001b9c9b', class: "tk-carousel-overlay", "data-testid": getDataTestId(this.dataTestid, 'overlay') }, h("div", { key: 'c34ff3931692c586dbfdad306f11528fa33ece41', class: "tk-carousel-items-container", "data-testid": getDataTestId(this.dataTestid, 'items') }, h("slot", { key: '28b8ab6ad330463aa8a47161de039af37ae8aa3d', onSlotchange: () => {
                this.updateItems();
                this.updateItemPosition();
            } }))), this.renderNavigationElement()));
    }
    get el() { return getElement(this); }
};
TkCarousel.style = tkCarouselCss();

export { TkCarousel as tk_carousel };
