// @takeoff-ui/core@0.11.9/dist/esm/tk-dropdown.entry.js
import { r as registerInstance, c as createEvent, h, d as getElement } from './index-D4ALLk7q.js';
import { c as classNames } from './index-D7YaJBkB.js';
import { C as ClickOutsideMixin } from './clickoutside-mixin-C0U9wSBw.js';
import { f as floatingElementAutoUpdate } from './position-utils-BiHzhysf.js';
import { v as v4 } from './v4-CK3_k8jD.js';

const tkDropdownCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-dropdown{display:inline-block}.tk-dropdown-container{position:relative}.tk-dropdown-container .tk-dropdown-panel{position:absolute;top:0;left:0;z-index:800;width:max-content;max-height:300px;overflow:hidden;border-radius:16px;border:1px solid var(--border-light);background:var(--static-light);box-shadow:var(--effect-1-default-sm);font-family:var(--family-body)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder{margin:8px;overflow-y:auto}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar{width:4px;height:4px}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-group-label{display:flex;align-items:center;gap:16px;color:var(--text-dark);font-size:var(--desktop-body-xs-size);padding:0px 8px}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-group-label label{white-space:nowrap}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-group-label .line{width:100%;height:1px;background-color:var(--background-light)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{color:var(--text-darkest);border-radius:8px;padding:var(--dropdown-items-basic-base-v-padding) var(--dropdown-items-basic-base-h-padding, 8px)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.left{text-align:left}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.center{text-align:center}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.right{text-align:right}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item:hover,.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item[data-active=true]{cursor:pointer;background:var(--background-lightest)}.tk-dropdown-container .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item.disabled{color:var(--text-dark);pointer-events:none}.tk-dropdown-container.large .tk-dropdown-panel .tk-dropdown-item-holder{max-height:240px}.tk-dropdown-container.large .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{font-family:var(--desktop-body-m-base-font);font-size:var(--desktop-body-m-base-size);line-height:var(--desktop-body-m-base-line-height);padding:8px 8px}.tk-dropdown-container.base .tk-dropdown-panel .tk-dropdown-item-holder{max-height:216px}.tk-dropdown-container.base .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);line-height:var(--desktop-body-s-line-height);padding:8px 8px}.tk-dropdown-container.small .tk-dropdown-panel .tk-dropdown-item-holder{max-height:180px}.tk-dropdown-container.small .tk-dropdown-panel .tk-dropdown-item-holder .tk-dropdown-item{font-family:var(--desktop-body-xs-font);font-size:var(--desktop-body-xs-size);line-height:var(--desktop-body-xs-line-height);padding:6px 8px}`;

const TkDropdown = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.tkItemClick = createEvent(this, "tk-item-click", 7);
        this.hasEmptyDataSlot = false;
        this.uniqueId = v4();
        this.isOpen = false;
        /**
         * The disabled status.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * The message to display when there is no data available.
         */
        this.emptyMessage = 'No options available';
        /**
         * The key to use for option labels
         * @defaultValue label
         */
        this.optionLabelKey = 'label';
        /**
         * The key to use for option values
         * @defaultValue value
         */
        this.optionValueKey = 'value';
        /**
         * The key to use for group names when options are grouped
         * @defaultValue groupName
         */
        this.groupNameKey = 'groupName';
        /**
         * The key to use for grouped options array
         * @defaultValue options
         */
        this.groupOptionsKey = 'options';
        /**
         * Indicates the alignment of options.
         * @defaultValue left
         */
        this.optionsAlign = 'left';
        /**
         * Sets the position of the tooltip.
         * @defaultValue 'right'
         */
        this.position = 'bottom';
        /**
         * Sets size for the dropdown panel.
         * @defaultValue base
         */
        this.size = 'base';
        /**
         * Click outside handler implementation - called by the mixin
         */
        this.closeHandler = () => {
            this.isOpen = false;
        };
    }
    disabledChanged(newValue) {
        if (newValue)
            this.isOpen = false;
    }
    componentWillLoad() {
        this.hasEmptyDataSlot = !!this.el.querySelector('[slot="empty-data"]');
        this.triggerRef = this.el.querySelector('[slot="trigger"]');
        if (this.triggerRef && !this.disabled) {
            this.triggerRef.style.cursor = 'pointer';
            this.triggerRef.addEventListener('click', () => {
                this.isOpen = !this.isOpen;
            });
        }
    }
    componentDidLoad() {
        // Initialize click outside mixin
        this.clickOutsideMixin = new ClickOutsideMixin({
            referenceElement: this.el,
            handler: this.closeHandler,
            disabled: this.disabled || !this.isOpen,
        });
    }
    componentDidUpdate() {
        var _a, _b;
        // Update click outside disabled state based on disabled prop and open state
        this.clickOutsideMixin.updateConfig({ disabled: this.disabled || !this.isOpen });
        if (this.isOpen) {
            // Clean up old floating UI listeners before setting up new ones
            (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
            this.updatePosition();
        }
        else {
            // Remove floating UI listeners when dropdown closes
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
        }
    }
    disconnectedCallback() {
        var _a, _b;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        // Call mixin's disconnectedCallback for cleanup
        (_b = this.clickOutsideMixin) === null || _b === void 0 ? void 0 : _b.disconnectedCallback();
    }
    updatePosition() {
        this.cleanup = floatingElementAutoUpdate(this.triggerRef, this.panelRef, undefined, {
            placement: this.position,
        });
    }
    isGrouped() {
        var _a;
        return ((_a = this.options) === null || _a === void 0 ? void 0 : _a.length) > 0 && this.options[0].hasOwnProperty(this.groupNameKey);
    }
    getOptionLabel(item) {
        return typeof item === 'object' ? item[this.optionLabelKey] : item;
    }
    handleItemClick(item) {
        this.isOpen = false;
        this.tkItemClick.emit(item);
    }
    createOptionItem(options) {
        return options === null || options === void 0 ? void 0 : options.map((item, index) => {
            let optionItem;
            if (this.optionHtml != undefined) {
                optionItem = this.optionHtml(item);
            }
            else {
                optionItem = this.getOptionLabel(item);
            }
            return (h("div", { class: classNames('tk-dropdown-item', this.optionsAlign, { disabled: item === null || item === void 0 ? void 0 : item.disabled }), "data-option-index": index, "data-active": index == 0 ? 'true' : 'false', onClick: () => this.handleItemClick(item), innerHTML: optionItem }));
        });
    }
    createOptions() {
        var _a;
        if (this.isGrouped()) {
            return (_a = this.options) === null || _a === void 0 ? void 0 : _a.map(group => {
                return (h("div", { class: "tk-dropdown-group" }, h("div", { class: "tk-dropdown-group-label" }, h("label", null, group[this.groupNameKey]), h("div", { class: "line" })), this.createOptionItem(group[this.groupOptionsKey])));
            });
        }
        else {
            return this.createOptionItem(this.options);
        }
    }
    renderDropdown() {
        var _a;
        if (!this.isOpen)
            return null;
        return (h("div", { class: "tk-dropdown-panel", ref: el => (this.panelRef = el), "data-tk-dropdown-id": this.uniqueId }, h("div", { class: "tk-dropdown-item-holder" }, ((_a = this.options) === null || _a === void 0 ? void 0 : _a.length) > 0 ? this.createOptions() : this.hasEmptyDataSlot ? h("slot", { name: "empty-data" }) : this.emptyMessage)));
    }
    render() {
        const rootClasses = classNames('tk-dropdown-container', this.size);
        return (h("div", { key: '50792dda7bef333a0a0d45ed6bad04dc26dc7534', class: rootClasses, "data-tk-dropdown-id": this.uniqueId }, h("slot", { key: '49a2bf076febbac475e3c9b23a0518b6db3d6ee0', name: "trigger" }), this.renderDropdown()));
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "disabled": [{
                "disabledChanged": 0
            }]
    }; }
};
TkDropdown.style = tkDropdownCss();

export { TkDropdown as tk_dropdown };
