// @takeoff-ui/core@0.11.9/dist/esm/tk-phone-input.entry.js
import { r as registerInstance, c as createEvent, h, d as getElement } from './index-D4ALLk7q.js';
import { c as classNames } from './index-D7YaJBkB.js';
import { g as getIconElementProps } from './icon-utils-SXopALR1.js';
import { f as floatingElementAutoUpdate, a as applyStyles } from './position-utils-BiHzhysf.js';
import { r as renderHint } from './hint-utils-BJaEKoxW.js';
import { v as v4 } from './v4-CK3_k8jD.js';
import './test-id-utils-BDMnqEGL.js';

/**
 * Default list of countries with their dialing codes and masks
 * Used when no custom country list is provided
 */
const INTERNAL_COUNTRIES = [
    { label: 'Afghanistan', id: 'AF', dialCode: '+93', mask: '999 99 999 9999' },
    { label: 'Åland Islands', id: 'AX', dialCode: '+358', mask: '99 999 99 99' },
    { label: 'Albania', id: 'AL', dialCode: '+355', mask: '999 999 999' },
    { label: 'Algeria', id: 'DZ', dialCode: '+213', mask: '99 999 99 99' },
    { label: 'American Samoa', id: 'AS', dialCode: '+1684', mask: '(999) 999-9999' },
    { label: 'Andorra', id: 'AD', dialCode: '+376', mask: '999 999' },
    { label: 'Angola', id: 'AO', dialCode: '+244', mask: '999 999 999' },
    { label: 'Anguilla', id: 'AI', dialCode: '+1264', mask: '(999) 999-9999' },
    { label: 'Antarctica', id: 'AQ', dialCode: '+672', mask: '99 999999' },
    { label: 'Antigua and Barbuda', id: 'AG', dialCode: '+1268', mask: '(999) 999-9999' },
    { label: 'Argentina', id: 'AR', dialCode: '+54', mask: '(999) 9999-9999' },
    { label: 'Armenia', id: 'AM', dialCode: '+374', mask: '99 999-9999' },
    { label: 'Aruba', id: 'AW', dialCode: '+297', mask: '999 9999' },
    { label: 'Australia', id: 'AU', dialCode: '+61', mask: '(99) 9999 9999' },
    { label: 'Austria', id: 'AT', dialCode: '+43', mask: '9999 99999999' },
    { label: 'Azerbaijan', id: 'AZ', dialCode: '+994', mask: '99 999 99 99' },
    { label: 'Bahamas', id: 'BS', dialCode: '+1242', mask: '(999) 999-9999' },
    { label: 'Bahrain', id: 'BH', dialCode: '+973', mask: '9999 9999' },
    { label: 'Bangladesh', id: 'BD', dialCode: '+880', mask: '9999-999999' },
    { label: 'Barbados', id: 'BB', dialCode: '+1246', mask: '(999) 999-9999' },
    { label: 'Belarus', id: 'BY', dialCode: '+375', mask: '(99) 999-99-99' },
    { label: 'Belgium', id: 'BE', dialCode: '+32', mask: '999 99 99 99' },
    { label: 'Belize', id: 'BZ', dialCode: '+501', mask: '999-9999' },
    { label: 'Benin', id: 'BJ', dialCode: '+229', mask: '99 99 99 99' },
    { label: 'Bermuda', id: 'BM', dialCode: '+1441', mask: '(999) 999-9999' },
    { label: 'Bhutan', id: 'BT', dialCode: '+975', mask: '99 999 999' },
    { label: 'Bolivia', id: 'BO', dialCode: '+591', mask: '99999999' },
    { label: 'Bonaire, Sint Eustatius and Saba', id: 'BQ', dialCode: '+599', mask: '999 9999' },
    { label: 'Bosnia and Herzegovina', id: 'BA', dialCode: '+387', mask: '99 999-999' },
    { label: 'Botswana', id: 'BW', dialCode: '+267', mask: '99 999 999' },
    { label: 'Bouvet Island', id: 'BV', dialCode: '+47', mask: '999 99 999' },
    { label: 'Brazil', id: 'BR', dialCode: '+55', mask: '(99) 99999-9999' },
    { label: 'British Indian Ocean Territory', id: 'IO', dialCode: '+246', mask: '999 9999' },
    { label: 'Brunei Darussalam', id: 'BN', dialCode: '+673', mask: '999-9999' },
    { label: 'Bulgaria', id: 'BG', dialCode: '+359', mask: '99 999 999' },
    { label: 'Burkina Faso', id: 'BF', dialCode: '+226', mask: '99 99 99 99' },
    { label: 'Burundi', id: 'BI', dialCode: '+257', mask: '99 99 9999' },
    { label: 'Cambodia', id: 'KH', dialCode: '+855', mask: '99 999 999' },
    { label: 'Cameroon', id: 'CM', dialCode: '+237', mask: '9999 9999' },
    { label: 'Canada', id: 'CA', dialCode: '+1', mask: '(999) 999-9999' },
    { label: 'Cape Verde', id: 'CV', dialCode: '+238', mask: '999 99 99' },
    { label: 'Cayman Islands', id: 'KY', dialCode: '+1345', mask: '(999) 999-9999' },
    { label: 'Central African Republic', id: 'CF', dialCode: '+236', mask: '99 99 99 99' },
    { label: 'Chad', id: 'TD', dialCode: '+235', mask: '99 99 99 99' },
    { label: 'Chile', id: 'CL', dialCode: '+56', mask: '9 9999 9999' },
    { label: 'China', id: 'CN', dialCode: '+86', mask: '999-99999999' },
    { label: 'Christmas Island', id: 'CX', dialCode: '+61', mask: '(99) 9999 9999' },
    { label: 'Cocos (Keeling) Islands', id: 'CC', dialCode: '+61', mask: '(99) 9999 9999' },
    { label: 'Colombia', id: 'CO', dialCode: '+57', mask: '999 999 9999' },
    { label: 'Comoros', id: 'KM', dialCode: '+269', mask: '99 99999' },
    { label: 'Congo', id: 'CG', dialCode: '+242', mask: '99 999 9999' },
    { label: 'Congo, Democratic Republic of the', id: 'CD', dialCode: '+243', mask: '99 999 9999' },
    { label: 'Cook Islands', id: 'CK', dialCode: '+682', mask: '99 999' },
    { label: 'Costa Rica', id: 'CR', dialCode: '+506', mask: '9999-9999' },
    { label: "Côte d'Ivoire", id: 'CI', dialCode: '+225', mask: '99 99 99 99' },
    { label: 'Croatia', id: 'HR', dialCode: '+385', mask: '99 999 9999' },
    { label: 'Cuba', id: 'CU', dialCode: '+53', mask: '9 9999999' },
    { label: 'Curaçao', id: 'CW', dialCode: '+599', mask: '999 9999' },
    { label: 'Cyprus', id: 'CY', dialCode: '+357', mask: '99 999 999' },
    { label: 'Czech Republic', id: 'CZ', dialCode: '+420', mask: '999 999 999' },
    { label: 'Denmark', id: 'DK', dialCode: '+45', mask: '99 99 99 99' },
    { label: 'Djibouti', id: 'DJ', dialCode: '+253', mask: '99 99 99 99' },
    { label: 'Dominica', id: 'DM', dialCode: '+1767', mask: '(999) 999-9999' },
    { label: 'Dominican Republic', id: 'DO', dialCode: '+1', mask: '(999) 999-9999' },
    { label: 'Ecuador', id: 'EC', dialCode: '+593', mask: '99 999-9999' },
    { label: 'Egypt', id: 'EG', dialCode: '+20', mask: '999 999 9999' },
    { label: 'El Salvador', id: 'SV', dialCode: '+503', mask: '9999-9999' },
    { label: 'Equatorial Guinea', id: 'GQ', dialCode: '+240', mask: '99 999 9999' },
    { label: 'Eritrea', id: 'ER', dialCode: '+291', mask: '9 999 999' },
    { label: 'Estonia', id: 'EE', dialCode: '+372', mask: '9999 9999' },
    { label: 'Eswatini', id: 'SZ', dialCode: '+268', mask: '99 99 9999' },
    { label: 'Ethiopia', id: 'ET', dialCode: '+251', mask: '99 999 9999' },
    { label: 'Falkland Islands (Malvinas)', id: 'FK', dialCode: '+500', mask: '99999' },
    { label: 'Faroe Islands', id: 'FO', dialCode: '+298', mask: '999999' },
    { label: 'Fiji', id: 'FJ', dialCode: '+679', mask: '99 99999' },
    { label: 'Finland', id: 'FI', dialCode: '+358', mask: '99 999 99 99' },
    { label: 'France', id: 'FR', dialCode: '+33', mask: '9 99 99 99 99' },
    { label: 'French Guiana', id: 'GF', dialCode: '+594', mask: '99999-9999' },
    { label: 'French Polynesia', id: 'PF', dialCode: '+689', mask: '99 99 99' },
    { label: 'French Southern Territories', id: 'TF', dialCode: '+262', mask: '99999-9999' },
    { label: 'Gabon', id: 'GA', dialCode: '+241', mask: '9 99 99 99' },
    { label: 'Gambia', id: 'GM', dialCode: '+220', mask: '999-9999' },
    { label: 'Georgia', id: 'GE', dialCode: '+995', mask: '999 99 99 99' },
    { label: 'Germany', id: 'DE', dialCode: '+49', mask: '9999 99999999' },
    { label: 'Ghana', id: 'GH', dialCode: '+233', mask: '99 999 9999' },
    { label: 'Gibraltar', id: 'GI', dialCode: '+350', mask: '999 99999' },
    { label: 'Greece', id: 'GR', dialCode: '+30', mask: '999 999 9999' },
    { label: 'Greenland', id: 'GL', dialCode: '+299', mask: '99 99 99' },
    { label: 'Grenada', id: 'GD', dialCode: '+1473', mask: '(999) 999-9999' },
    { label: 'Guadeloupe', id: 'GP', dialCode: '+590', mask: '999999999' },
    { label: 'Guam', id: 'GU', dialCode: '+1671', mask: '(999) 999-9999' },
    { label: 'Guatemala', id: 'GT', dialCode: '+502', mask: '9999-9999' },
    { label: 'Guernsey', id: 'GG', dialCode: '+44', mask: '99999 999999' },
    { label: 'Guinea', id: 'GN', dialCode: '+224', mask: '99 999 999' },
    { label: 'Guinea-Bissau', id: 'GW', dialCode: '+245', mask: '9 999999' },
    { label: 'Guyana', id: 'GY', dialCode: '+592', mask: '999-9999' },
    { label: 'Haiti', id: 'HT', dialCode: '+509', mask: '99 99-9999' },
    { label: 'Heard Island and McDonald Islands', id: 'HM', dialCode: '+672', mask: '' },
    { label: 'Holy See (Vatican City State)', id: 'VA', dialCode: '+39', mask: '99 9999 9999' },
    { label: 'Honduras', id: 'HN', dialCode: '+504', mask: '9999-9999' },
    { label: 'Hong Kong', id: 'HK', dialCode: '+852', mask: '9999 9999' },
    { label: 'Hungary', id: 'HU', dialCode: '+36', mask: '(99) 999-9999' },
    { label: 'Iceland', id: 'IS', dialCode: '+354', mask: '999-9999' },
    { label: 'India', id: 'IN', dialCode: '+91', mask: '99999-99999' },
    { label: 'Indonesia', id: 'ID', dialCode: '+62', mask: '999-9999-9999' },
    { label: 'Iran, Islamic Republic of', id: 'IR', dialCode: '+98', mask: '999 999 9999' },
    { label: 'Iraq', id: 'IQ', dialCode: '+964', mask: '999 999 9999' },
    { label: 'Ireland', id: 'IE', dialCode: '+353', mask: '99 999 9999' },
    { label: 'Isle of Man', id: 'IM', dialCode: '+44', mask: '99999 999999' },
    { label: 'Israel', id: 'IL', dialCode: '+972', mask: '999 999-9999' },
    { label: 'Italy', id: 'IT', dialCode: '+39', mask: '999 9999 9999' },
    { label: 'Jamaica', id: 'JM', dialCode: '+1876', mask: '(999) 999-9999' },
    { label: 'Japan', id: 'JP', dialCode: '+81', mask: '99-9999-9999' },
    { label: 'Jersey', id: 'JE', dialCode: '+44', mask: '99999 999999' },
    { label: 'Jordan', id: 'JO', dialCode: '+962', mask: '9 9999-9999' },
    { label: 'Kazakhstan', id: 'KZ', dialCode: '+7', mask: '(999) 999-99-99' },
    { label: 'Kenya', id: 'KE', dialCode: '+254', mask: '999 999999' },
    { label: 'Kiribati', id: 'KI', dialCode: '+686', mask: '99 999' },
    { label: "Korea (Democratic People's Republic of)", id: 'KP', dialCode: '+850', mask: '999 999-9999' },
    { label: 'Korea, Republic of', id: 'KR', dialCode: '+82', mask: '999-9999-9999' },
    { label: 'Kosovo', id: 'XK', dialCode: '+383', mask: '999 999 999' },
    { label: 'Kuwait', id: 'KW', dialCode: '+965', mask: '9999 9999' },
    { label: 'Kyrgyzstan', id: 'KG', dialCode: '+996', mask: '999 999-9999' },
    { label: "Lao People's Democratic Republic", id: 'LA', dialCode: '+856', mask: '99 999 999' },
    { label: 'Latvia', id: 'LV', dialCode: '+371', mask: '99 999 999' },
    { label: 'Lebanon', id: 'LB', dialCode: '+961', mask: '99 999 999' },
    { label: 'Lesotho', id: 'LS', dialCode: '+266', mask: '99 999 999' },
    { label: 'Liberia', id: 'LR', dialCode: '+231', mask: '99 999 999' },
    { label: 'Libya', id: 'LY', dialCode: '+218', mask: '99-9999999' },
    { label: 'Liechtenstein', id: 'LI', dialCode: '+423', mask: '999 99 99' },
    { label: 'Lithuania', id: 'LT', dialCode: '+370', mask: '(999) 99999' },
    { label: 'Luxembourg', id: 'LU', dialCode: '+352', mask: '999 99 99 99' },
    { label: 'Macau', id: 'MO', dialCode: '+853', mask: '9999 9999' },
    { label: 'Republic of North Macedonia', id: 'MK', dialCode: '+389', mask: '99 999 999' },
    { label: 'Madagascar', id: 'MG', dialCode: '+261', mask: '99 99 999 99' },
    { label: 'Malawi', id: 'MW', dialCode: '+265', mask: '9 9999 9999' },
    { label: 'Malaysia', id: 'MY', dialCode: '+60', mask: '99-9999-9999' },
    { label: 'Maldives', id: 'MV', dialCode: '+960', mask: '999-9999' },
    { label: 'Mali', id: 'ML', dialCode: '+223', mask: '99 99 99 99' },
    { label: 'Malta', id: 'MT', dialCode: '+356', mask: '9999 9999' },
    { label: 'Marshall Islands', id: 'MH', dialCode: '+692', mask: '999-9999' },
    { label: 'Martinique', id: 'MQ', dialCode: '+596', mask: '999999999' },
    { label: 'Mauritania', id: 'MR', dialCode: '+222', mask: '99 99 99 99' },
    { label: 'Mauritius', id: 'MU', dialCode: '+230', mask: '9999-9999' },
    { label: 'Mayotte', id: 'YT', dialCode: '+262', mask: '99999-9999' },
    { label: 'Mexico', id: 'MX', dialCode: '+52', mask: '999 999 9999' },
    { label: 'Micronesia (Federated States of)', id: 'FM', dialCode: '+691', mask: '999-9999' },
    { label: 'Republic of Moldova', id: 'MD', dialCode: '+373', mask: '9999 99999' },
    { label: 'Monaco', id: 'MC', dialCode: '+377', mask: '99 99 99 99' },
    { label: 'Mongolia', id: 'MN', dialCode: '+976', mask: '9999-9999' },
    { label: 'Montenegro', id: 'ME', dialCode: '+382', mask: '99 999 999' },
    { label: 'Montserrat', id: 'MS', dialCode: '+1664', mask: '(999) 999-9999' },
    { label: 'Morocco', id: 'MA', dialCode: '+212', mask: '99-9999999' },
    { label: 'Mozambique', id: 'MZ', dialCode: '+258', mask: '99 999 9999' },
    { label: 'Myanmar', id: 'MM', dialCode: '+95', mask: '99 999 9999' },
    { label: 'Namibia', id: 'NA', dialCode: '+264', mask: '99 999 9999' },
    { label: 'Nauru', id: 'NR', dialCode: '+674', mask: '999-9999' },
    { label: 'Nepal', id: 'NP', dialCode: '+977', mask: '999-9999999' },
    { label: 'Netherlands', id: 'NL', dialCode: '+31', mask: '9 99999999' },
    { label: 'Netherlands Antilles', id: 'AN', dialCode: '+599', mask: '999 9999' },
    { label: 'New Caledonia', id: 'NC', dialCode: '+687', mask: '999999' },
    { label: 'New Zealand', id: 'NZ', dialCode: '+64', mask: '999 999 9999' },
    { label: 'Nicaragua', id: 'NI', dialCode: '+505', mask: '9999-9999' },
    { label: 'Niger', id: 'NE', dialCode: '+227', mask: '99 99 99 99' },
    { label: 'Nigeria', id: 'NG', dialCode: '+234', mask: '999 999 9999' },
    { label: 'Niue', id: 'NU', dialCode: '+683', mask: '9999' },
    { label: 'Norfolk Island', id: 'NF', dialCode: '+672', mask: '999999' },
    { label: 'Northern Mariana Islands', id: 'MP', dialCode: '+1670', mask: '(999) 999-9999' },
    { label: 'Norway', id: 'NO', dialCode: '+47', mask: '999 99 999' },
    { label: 'Oman', id: 'OM', dialCode: '+968', mask: '9999 9999' },
    { label: 'Pakistan', id: 'PK', dialCode: '+92', mask: '999-9999999' },
    { label: 'Palau', id: 'PW', dialCode: '+680', mask: '999-9999' },
    { label: 'State of Palestine', id: 'PS', dialCode: '+970', mask: '99 999-9999' },
    { label: 'Panama', id: 'PA', dialCode: '+507', mask: '9999-9999' },
    { label: 'Papua New Guinea', id: 'PG', dialCode: '+675', mask: '999-9999' },
    { label: 'Paraguay', id: 'PY', dialCode: '+595', mask: '999 999-9999' },
    { label: 'Peru', id: 'PE', dialCode: '+51', mask: '999 999 999' },
    { label: 'Philippines', id: 'PH', dialCode: '+63', mask: '9999 9999999' },
    { label: 'Pitcairn', id: 'PN', dialCode: '+64', mask: '99999999' },
    { label: 'Poland', id: 'PL', dialCode: '+48', mask: '999-999-999' },
    { label: 'Portugal', id: 'PT', dialCode: '+351', mask: '999 999 999' },
    { label: 'Puerto Rico', id: 'PR', dialCode: '+1', mask: '(999) 999-9999' },
    { label: 'Qatar', id: 'QA', dialCode: '+974', mask: '9999 9999' },
    { label: 'Réunion', id: 'RE', dialCode: '+262', mask: '99999-9999' },
    { label: 'Romania', id: 'RO', dialCode: '+40', mask: '99 999-9999' },
    { label: 'Russian Federation', id: 'RU', dialCode: '+7', mask: '(999) 999-9999' },
    { label: 'Rwanda', id: 'RW', dialCode: '+250', mask: '999 999 999' },
    { label: 'Saint Barthélemy', id: 'BL', dialCode: '+590', mask: '999 99 99 99' },
    { label: 'Saint Helena', id: 'SH', dialCode: '+290', mask: '99999' },
    { label: 'Saint Kitts and Nevis', id: 'KN', dialCode: '+1869', mask: '(999) 999-9999' },
    { label: 'Saint Lucia', id: 'LC', dialCode: '+1758', mask: '(999) 999-9999' },
    { label: 'Sint Maarten', id: 'SX', dialCode: '+1721', mask: '(999) 999-9999' },
    { label: 'Saint Martin', id: 'MF', dialCode: '+590', mask: '999 99 99 99' },
    { label: 'Saint Pierre and Miquelon', id: 'PM', dialCode: '+508', mask: '99-9999' },
    { label: 'Saint Vincent and the Grenadines', id: 'VC', dialCode: '+1784', mask: '(999) 999-9999' },
    { label: 'Samoa', id: 'WS', dialCode: '+685', mask: '99-9999' },
    { label: 'San Marino', id: 'SM', dialCode: '+378', mask: '9999 999999' },
    { label: 'São Tomé and Príncipe', id: 'ST', dialCode: '+239', mask: '9 999999' },
    { label: 'Saudi Arabia', id: 'SA', dialCode: '+966', mask: '99 999 9999' },
    { label: 'Senegal', id: 'SN', dialCode: '+221', mask: '99 999 99 99' },
    { label: 'Serbia', id: 'RS', dialCode: '+381', mask: '99 999-9999' },
    { label: 'Seychelles', id: 'SC', dialCode: '+248', mask: '9 999 999' },
    { label: 'Sierra Leone', id: 'SL', dialCode: '+232', mask: '99 999999' },
    { label: 'Singapore', id: 'SG', dialCode: '+65', mask: '9999 9999' },
    { label: 'Slovakia', id: 'SK', dialCode: '+421', mask: '999 999 999' },
    { label: 'Slovenia', id: 'SI', dialCode: '+386', mask: '99 999 999' },
    { label: 'Solomon Islands', id: 'SB', dialCode: '+677', mask: '99999' },
    { label: 'Somalia', id: 'SO', dialCode: '+252', mask: '99 999-99' },
    { label: 'South Africa', id: 'ZA', dialCode: '+27', mask: '99 999 9999' },
    { label: 'South Georgia and the South Sandwich Islands', id: 'GS', dialCode: '+500', mask: '99999' },
    { label: 'South Sudan', id: 'SS', dialCode: '+211', mask: '99 999 9999' },
    { label: 'Spain', id: 'ES', dialCode: '+34', mask: '999 99 99 99' },
    { label: 'Sri Lanka', id: 'LK', dialCode: '+94', mask: '99 999 9999' },
    { label: 'Sudan', id: 'SD', dialCode: '+249', mask: '99 999-9999' },
    { label: 'Suriname', id: 'SR', dialCode: '+597', mask: '999-9999' },
    { label: 'Svalbard and Jan Mayen', id: 'SJ', dialCode: '+47', mask: '999 99 999' },
    { label: 'Sweden', id: 'SE', dialCode: '+46', mask: '99-9999999' },
    { label: 'Switzerland', id: 'CH', dialCode: '+41', mask: '99 999 99 99' },
    { label: 'Syrian Arab Republic', id: 'SY', dialCode: '+963', mask: '99 9999-9999' },
    { label: 'Taiwan, Province of China', id: 'TW', dialCode: '+886', mask: '9999-9999' },
    { label: 'Tajikistan', id: 'TJ', dialCode: '+992', mask: '99 999-9999' },
    { label: 'Tanzania, United Republic of', id: 'TZ', dialCode: '+255', mask: '99 999 9999' },
    { label: 'Thailand', id: 'TH', dialCode: '+66', mask: '99 999 9999' },
    { label: 'Timor-Leste', id: 'TL', dialCode: '+670', mask: '999-9999' },
    { label: 'Togo', id: 'TG', dialCode: '+228', mask: '99 99 99 99' },
    { label: 'Tokelau', id: 'TK', dialCode: '+690', mask: '9999' },
    { label: 'Tonga', id: 'TO', dialCode: '+676', mask: '99999' },
    { label: 'Trinidad and Tobago', id: 'TT', dialCode: '+1868', mask: '(999) 999-9999' },
    { label: 'Tunisia', id: 'TN', dialCode: '+216', mask: '99 999 999' },
    { label: 'Turkey', id: 'TR', dialCode: '+90', mask: '(999) 999 9999' },
    { label: 'Turkmenistan', id: 'TM', dialCode: '+993', mask: '99 999999' },
    { label: 'Turks and Caicos Islands', id: 'TC', dialCode: '+1649', mask: '(999) 999-9999' },
    { label: 'Tuvalu', id: 'TV', dialCode: '+688', mask: '99999' },
    { label: 'Uganda', id: 'UG', dialCode: '+256', mask: '99 999-9999' },
    { label: 'Ukraine', id: 'UA', dialCode: '+380', mask: '(99) 999 99 99' },
    { label: 'United Arab Emirates', id: 'AE', dialCode: '+971', mask: '99 999 9999' },
    { label: 'United Kingdom', id: 'GB', dialCode: '+44', mask: '9999 999999' },
    { label: 'United States', id: 'US', dialCode: '+1', mask: '(999) 999-9999' },
    { label: 'United States Minor Outlying Islands', id: 'UM', dialCode: '+1', mask: '(999) 999-9999' },
    { label: 'Uruguay', id: 'UY', dialCode: '+598', mask: '9999 9999' },
    { label: 'Uzbekistan', id: 'UZ', dialCode: '+998', mask: '99 999-9999' },
    { label: 'Vanuatu', id: 'VU', dialCode: '+678', mask: '99999' },
    { label: 'Venezuela', id: 'VE', dialCode: '+58', mask: '999-9999999' },
    { label: 'Viet Nam', id: 'VN', dialCode: '+84', mask: '999 99 99-99' },
    { label: 'Virgin Islands, British', id: 'VG', dialCode: '+1284', mask: '(999) 999-9999' },
    { label: 'Virgin Islands, U.S.', id: 'VI', dialCode: '+1340', mask: '(999) 999-9999' },
    { label: 'Wallis and Futuna', id: 'WF', dialCode: '+681', mask: '99-9999' },
    { label: 'Western Sahara', id: 'EH', dialCode: '+212', mask: '99-9999999' },
    { label: 'Yemen', id: 'YE', dialCode: '+967', mask: '999 999 999' },
    { label: 'Zambia', id: 'ZM', dialCode: '+260', mask: '99-9999999' },
    { label: 'Zimbabwe', id: 'ZW', dialCode: '+263', mask: '99 999-9999' },
];

const tkPhoneInputCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block;min-width:0}.tk-phone-input-container{display:flex;flex-direction:column;align-items:flex-start;color:var(--text-darkest);font-family:var(--family-body);gap:var(--input-external-label-small-container-gap);min-width:0}.tk-phone-input-container .tk-phone-input-label{font-weight:var(--weight-regular);gap:var(--input-external-label-small-container-gap)}.tk-phone-input-container .tk-phone-input-wrapper{display:flex;align-items:center;gap:var(--input-external-label-small-container-gap);align-self:stretch;border-radius:var(--button-sizes-large-radius);border:var(--spacing-px) solid var(--border-light);background:var(--static-light);overflow:hidden}.tk-phone-input-container .tk-phone-input-wrapper:focus{border:var(--spacing-px) solid var(--states-info-base)}.tk-phone-input-container .tk-phone-input-wrapper:hover{background-color:var(--background-lightest)}.tk-phone-input-container .tk-phone-input-dropdown{flex-shrink:0}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-button{cursor:pointer;display:flex;align-items:center;border:none;background-color:transparent;padding:0}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-button .tk-phone-input-dropdown-button-selected{display:flex;align-items:center;gap:var(--spacing-xs);border-right:var(--spacing-px) solid var(--border-light)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-button .tk-phone-input-dropdown-button-selected .tk-phone-input-dial-code{padding-right:var(--input-external-label-large-label-h-padding)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-button .tk-phone-input-dropdown-button-selected.tk-phone-input-no-dial-code{padding-right:var(--input-external-label-large-label-h-padding)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu{position:fixed;z-index:800;background:var(--static-light);border:var(--spacing-px) solid var(--border-light);border-radius:var(--button-sizes-large-radius);box-shadow:var(--effect-1-default-sm);display:flex;flex-direction:column;overflow:hidden}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-dropdown-menu-search{padding:var(--spacing-m-base)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu{list-style:none;overflow-y:auto;margin:var(--spacing-m-base)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu::-webkit-scrollbar{width:4px;height:4px}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu .tk-phone-input-menu-item{display:flex;align-items:center;gap:var(--spacing-xs);padding:var(--input-external-label-base-input-v-padding) var(--spacing-4xl);cursor:pointer;transition:background-color 0.2s}.tk-phone-input-container .tk-phone-input-dropdown .tk-phone-input-dropdown-menu .tk-phone-input-menu .tk-phone-input-menu-item:hover{background-color:var(--background-lightest)}.tk-phone-input-container .tk-phone-input-input{flex:1;min-width:0;border:none;background-color:transparent;outline:none}.tk-phone-input-container .tk-phone-input-input::placeholder{color:var(--text-sub-base)}.tk-phone-input-container[aria-invalid] .tk-phone-input-wrapper{border-color:var(--states-danger-base)}.tk-phone-input-container[aria-disabled],.tk-phone-input-container[aria-readonly]{pointer-events:none}.tk-phone-input-container[aria-disabled] .tk-phone-input-wrapper,.tk-phone-input-container[aria-readonly] .tk-phone-input-wrapper{background-color:var(--background-lightest)}.tk-phone-input-container[aria-disabled] .tk-phone-input-label,.tk-phone-input-container[aria-disabled] .tk-phone-input-label-asterisk,.tk-phone-input-container[aria-disabled] .tk-phone-input-dropdown-button,.tk-phone-input-container[aria-readonly] .tk-phone-input-label,.tk-phone-input-container[aria-readonly] .tk-phone-input-label-asterisk,.tk-phone-input-container[aria-readonly] .tk-phone-input-dropdown-button{color:var(--text-sub-base)}.tk-phone-input-container[aria-disabled] .tk-phone-input-input{color:var(--text-sub-base)}.tk-phone-input-container.tk-phone-input-container-large{font-size:var(--desktop-label-l-size);line-height:var(--desktop-label-l-line-height)}.tk-phone-input-container.tk-phone-input-container-large .tk-phone-input-wrapper{height:var(--spacing-12xl);padding:var(--input-external-label-large-input-v-padding) var(--input-external-label-large-input-h-padding)}.tk-phone-input-container.tk-phone-input-container-large .tk-phone-input-wrapper .tk-phone-input-dropdown-button,.tk-phone-input-container.tk-phone-input-container-large .tk-phone-input-wrapper .tk-phone-input-input{font-size:var(--desktop-label-l-size)}.tk-phone-input-container.tk-phone-input-container-large .tk-phone-input-label{padding:var(--input-external-label-large-label-v-padding) var(--input-external-label-large-label-h-padding)}.tk-phone-input-container.tk-phone-input-container-large .tk-phone-input-menu{max-height:288px}.tk-phone-input-container.tk-phone-input-container-base{font-size:var(--desktop-label-m-base-size);line-height:var(--desktop-label-m-base-line-height)}.tk-phone-input-container.tk-phone-input-container-base .tk-phone-input-wrapper{height:var(--spacing-10xl);padding:var(--input-external-label-base-input-v-padding) var(--input-external-label-base-input-h-padding)}.tk-phone-input-container.tk-phone-input-container-base .tk-phone-input-wrapper .tk-phone-input-dropdown-button,.tk-phone-input-container.tk-phone-input-container-base .tk-phone-input-wrapper .tk-phone-input-input{font-size:var(--desktop-label-m-base-size)}.tk-phone-input-container.tk-phone-input-container-base .tk-phone-input-label{padding:var(--input-external-label-base-label-v-padding) var(--input-external-label-base-label-h-padding)}.tk-phone-input-container.tk-phone-input-container-base .tk-phone-input-menu{max-height:276px}.tk-phone-input-container.tk-phone-input-container-small{font-size:var(--desktop-label-s-size);line-height:var(--desktop-label-s-line-height)}.tk-phone-input-container.tk-phone-input-container-small .tk-phone-input-wrapper{height:var(--spacing-8xl);padding:var(--input-external-label-small-input-v-padding) var(--input-external-label-small-input-h-padding)}.tk-phone-input-container.tk-phone-input-container-small .tk-phone-input-wrapper .tk-phone-input-dropdown-button,.tk-phone-input-container.tk-phone-input-container-small .tk-phone-input-wrapper .tk-phone-input-input{font-size:var(--desktop-label-s-size)}.tk-phone-input-container.tk-phone-input-container-small .tk-phone-input-label{padding:var(--input-external-label-small-label-v-padding) var(--input-external-label-small-label-h-padding)}.tk-phone-input-container.tk-phone-input-container-small .tk-phone-input-menu{max-height:264px}.tk-phone-input-container .tk-phone-input-asterisk{color:var(--states-danger-base);font-size:var(--desktop-label-s-size);line-height:var(--desktop-label-s-line-height);font-weight:var(--weight-light)}`;

const TkPhoneInput = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.tkChange = createEvent(this, "tk-change", 5);
        this.tkBlur = createEvent(this, "tk-blur", 7);
        this.tkFocus = createEvent(this, "tk-focus", 7);
        this.uniqueId = v4();
        /**
         * The list of countries to display in the dropdown.
         */
        this.countries = [];
        /**
         * The current input value with mask applied.
         */
        this.inputValue = '';
        /**
         * Whether the input has been touched.
         */
        this.hasFocus = false;
        /**
         * Whether the country dropdown is open.
         */
        this.isDropdownOpen = false;
        /**
         * The current search term for filtering countries.
         */
        this.searchTerm = '';
        /**
         * Whether the input is disabled.
         * * @defaultValue false
         */
        this.disabled = false;
        /**
         * If `true`, the user cannot modify the value.
         * @defaultValue false
         */
        this.readonly = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * The default country to select (ISO country code).
         */
        this.defaultCountry = 'TR';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * Sets size for the component.
         */
        this.size = 'base';
        /**
         * Hides the phone flag icon in the dropdown button and list.
         * @default false
         */
        this.hideFlag = false;
        /**
         * Determines the width of the dropdown. Accepts values like 'match-parent', 'auto', or a specific width in '300px'.
         * @defaultValue match-parent
         */
        this.dropdownWidthMode = 'match-parent';
        /**
         * Toggle the country dropdown.
         */
        this.toggleDropdown = () => {
            var _a;
            if (!this.disabled) {
                this.isDropdownOpen = !this.isDropdownOpen;
                if (this.isDropdownOpen) {
                    (_a = this.searchInput) === null || _a === void 0 ? void 0 : _a.focus();
                }
            }
        };
        /**
         * Handle country selection from the dropdown.
         */
        this.handleCountrySelect = (country) => {
            var _a;
            this.setSelectedCountry(country.id);
            this.value = {
                rawValue: '',
                maskedValue: '',
                country: {
                    id: this.selectedCountry.id,
                    label: this.selectedCountry.label,
                    dialCode: this.selectedCountry.dialCode,
                    mask: this.selectedCountry.mask,
                    placeholder: this.selectedCountry.placeholder,
                },
            };
            this.tkChange.emit(this.value);
            this.inputValue = '';
            this.isDropdownOpen = false;
            this.searchTerm = '';
            (_a = this.inputRef) === null || _a === void 0 ? void 0 : _a.focus();
        };
        /**
         * Event handler for clicks outside the component to close dropdown.
         */
        this.handleClickOutside = (event) => {
            if (!this.el.contains(event.target)) {
                this.isDropdownOpen = false;
            }
        };
        /**
         * Handle changes to the country search input.
         */
        this.handleSearchChange = (event) => {
            this.searchTerm = event.target.value;
        };
        /**
         * Handle changes to the phone number input.
         */
        this.handleInput = (event) => {
            var _a;
            const inputElement = event.target;
            const rawValue = inputElement.value.replace(/\D/g, '');
            if (!((_a = this.selectedCountry.mask) === null || _a === void 0 ? void 0 : _a.length)) {
                this.inputValue = rawValue;
                this.inputRef.value = this.inputValue;
                // reassign the value object and emit the change event before returning. Masked Value consists of raw value
                this.value = {
                    rawValue,
                    maskedValue: rawValue,
                    country: {
                        id: this.selectedCountry.id,
                        label: this.selectedCountry.label,
                        dialCode: this.selectedCountry.dialCode,
                        placeholder: this.selectedCountry.placeholder,
                    },
                };
                this.tkChange.emit(this.value);
                return;
            }
            const currentMask = this.selectedCountry.mask;
            const maxDigits = (currentMask.match(/9/g) || []).length;
            const hasInvalidCharacters = /[^\d() -]/.test(inputElement.value);
            if (rawValue.length > maxDigits || hasInvalidCharacters)
                return (this.inputRef.value = this.inputValue);
            this.hasFocus = false;
            this.inputValue = this.applyMask(rawValue, currentMask);
            this.inputRef.value = this.inputValue;
            this.value = {
                rawValue,
                maskedValue: this.inputValue,
                country: {
                    id: this.selectedCountry.id,
                    label: this.selectedCountry.label,
                    dialCode: this.selectedCountry.dialCode,
                    mask: currentMask,
                    placeholder: this.selectedCountry.placeholder,
                },
            };
            this.tkChange.emit(this.value);
        };
        this.handleInputBlur = () => {
            this.hasFocus = false;
            this.tkBlur.emit();
        };
        this.handleInputFocus = () => {
            this.hasFocus = true;
            this.tkFocus.emit();
        };
        this.handleFormReset = () => {
            var _a;
            this.value = {
                rawValue: '',
                maskedValue: '',
                country: {
                    id: this.selectedCountry.id,
                    label: this.selectedCountry.label,
                    dialCode: this.selectedCountry.dialCode,
                    mask: this.selectedCountry.mask,
                    placeholder: this.selectedCountry.placeholder,
                },
            };
            this.tkChange.emit(this.value);
            this.isDropdownOpen = false;
            this.inputValue = '';
            this.inputRef.value = '';
            this.searchTerm = '';
            (_a = this.inputRef) === null || _a === void 0 ? void 0 : _a.focus();
        };
    }
    valueChanged(newValue) {
        var _a, _b;
        if (!newValue || (typeof newValue === 'object' && Object.keys(newValue).length === 0)) {
            this.handleFormReset();
        }
        if (!newValue.rawValue && !newValue.maskedValue) {
            this.inputValue = '';
        }
        if (newValue) {
            if ((_a = newValue === null || newValue === void 0 ? void 0 : newValue.country) === null || _a === void 0 ? void 0 : _a.id) {
                this.setSelectedCountry((_b = newValue === null || newValue === void 0 ? void 0 : newValue.country) === null || _b === void 0 ? void 0 : _b.id);
            }
            this.inputValue = this.applyMask(newValue.rawValue, this.selectedCountry.mask);
        }
    }
    disabledChanged(newValue) {
        if (newValue)
            this.isDropdownOpen = false;
    }
    /**
     * Add event listeners when the component is connected to the DOM.
     */
    connectedCallback() {
        document.addEventListener('click', this.handleClickOutside);
    }
    /**
     * Remove event listeners when the component is disconnected from the DOM.
     */
    disconnectedCallback() {
        var _a;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        document.removeEventListener('click', this.handleClickOutside);
    }
    /**
     * Initialize the component before it is rendered.
     */
    componentWillLoad() {
        var _a, _b, _c, _d;
        this.initializeCountries();
        if ((this === null || this === void 0 ? void 0 : this.value) && ((_a = Object.keys(this === null || this === void 0 ? void 0 : this.value)) === null || _a === void 0 ? void 0 : _a.length)) {
            if ((_c = (_b = this.value) === null || _b === void 0 ? void 0 : _b.country) === null || _c === void 0 ? void 0 : _c.id) {
                this.setSelectedCountry(this.value.country.id);
            }
            this.inputValue = this.applyMask((_d = this === null || this === void 0 ? void 0 : this.value) === null || _d === void 0 ? void 0 : _d.rawValue, this === null || this === void 0 ? void 0 : this.selectedCountry.mask);
        }
        else {
            this.setSelectedCountry(this.defaultCountry);
        }
    }
    /**
     * Update the component when properties change.
     */
    componentDidUpdate() {
        var _a, _b, _c;
        if (this.isDropdownOpen) {
            // Clean up old floating UI listeners before setting up new ones
            (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
            this.updatePosition();
        }
        else {
            // Remove floating UI listeners when dropdown closes
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
            (_c = this.panelRef) === null || _c === void 0 ? void 0 : _c.remove();
        }
    }
    formResetCallback() {
        this.handleFormReset();
    }
    updatePosition() {
        const dropdownWidthMode = this.dropdownWidthMode;
        const tkInputRootEl = this.el.querySelector('.tk-phone-input-wrapper');
        this.cleanup = floatingElementAutoUpdate(tkInputRootEl, this.panelRef, undefined, {
            placement: 'bottom-start',
            size: {
                apply({ rects, elements }) {
                    if (dropdownWidthMode === 'match-parent') {
                        applyStyles(elements.floating, {
                            width: `${rects.reference.width}px`,
                        });
                    }
                    else if (dropdownWidthMode !== 'auto' && dropdownWidthMode.length > 0) {
                        applyStyles(elements.floating, {
                            width: dropdownWidthMode,
                        });
                    }
                },
            },
        });
    }
    /**
     * Initialize the list of countries from the provided prop or fallback to internal list.
     */
    initializeCountries() {
        if (this.countryList) {
            this.countries = this.countryList;
        }
        else {
            this.countries = INTERNAL_COUNTRIES;
        }
    }
    /**
     * Set the selected country by its ID.
     */
    setSelectedCountry(countryId) {
        const country = this.countries.find(c => c.id.toUpperCase() === countryId.toUpperCase());
        this.selectedCountry = country || this.countries[0];
    }
    /**
     * Apply the mask to a raw phone number.
     */
    applyMask(rawValue, mask) {
        if (!(mask === null || mask === void 0 ? void 0 : mask.length) || !(rawValue === null || rawValue === void 0 ? void 0 : rawValue.length))
            return rawValue;
        let maskedValue = '';
        let digitIndex = 0;
        for (let i = 0; i < mask.length && digitIndex < rawValue.length; i++) {
            if (mask[i] === '9') {
                maskedValue += rawValue[digitIndex++];
            }
            else {
                maskedValue += mask[i];
            }
        }
        return maskedValue;
    }
    /**
     * Get the filtered list of countries based on the search term.
     */
    getFilteredCountries() {
        const term = this.searchTerm.toLowerCase();
        if (!term)
            return this.countries;
        return this.countries.filter(country => { var _a; return country.label.toLowerCase().includes(term) || ((_a = country.dialCode) === null || _a === void 0 ? void 0 : _a.includes(term)); });
    }
    /**
     * Get the flag class based on the country object.
     */
    getFlagClass(country) {
        return classNames('flag', (country === null || country === void 0 ? void 0 : country.dialCode) ? `flag-${country === null || country === void 0 ? void 0 : country.id.toLowerCase()}` : 'flag-none');
    }
    renderLabel() {
        var _a;
        let label;
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            const asterisk = h("span", { class: "tk-phone-input-asterisk" }, "*");
            label = (h("label", { htmlFor: this.uniqueId, class: "tk-phone-input-label" }, this.label, this.showAsterisk ? asterisk : ''));
        }
        return label;
    }
    /**
     * Create the country selector dropdown.
     */
    renderCountrySelector() {
        return (h("div", { class: "tk-phone-input-dropdown" }, this.createDropdownButton(), this.isDropdownOpen && (h("div", { class: "tk-phone-input-dropdown-menu", role: "listbox", ref: el => (this.panelRef = el) }, this.createDropdownSearch(), this.createDropdownList()))));
    }
    /**
     * Create the dropdown button for selecting a country.
     */
    createDropdownButton() {
        const selectedClass = classNames('tk-phone-input-dropdown-button-selected', {
            'tk-phone-input-no-dial-code': !this.selectedCountry.dialCode,
        });
        return (h("button", { class: "tk-phone-input-dropdown-button", onClick: this.toggleDropdown, type: "button" }, h("div", { class: selectedClass }, h("tk-icon", Object.assign({}, getIconElementProps('stat_minus_1', { variant: null, size: 'large' }, 'rounded', 'span'))), !this.hideFlag && this.renderFlag(this.selectedCountry), this.selectedCountry.dialCode && h("span", { class: "tk-phone-input-dial-code" }, this.selectedCountry.dialCode))));
    }
    /**
     * Create the search input for filtering countries in the dropdown.
     */
    createDropdownSearch() {
        return (h("tk-input", { class: "tk-phone-input-dropdown-menu-search", size: this.size, placeholder: "Search", value: this.searchTerm, "onTk-change": this.handleSearchChange, ref: el => (this.searchInput = el), icon: "search", iconPosition: "right", onClick: (e) => e.stopPropagation() }));
    }
    /**
     * Create the dropdown list of countries.
     */
    createDropdownList() {
        return (h("ul", { class: "tk-phone-input-menu" }, this.getFilteredCountries().map(country => (h("li", { class: "tk-phone-input-menu-item", onClick: () => this.handleCountrySelect(country), key: country.id, role: "option", "aria-selected": country.id === this.selectedCountry.id }, !this.hideFlag && this.renderFlag(country), country.label && h("span", { class: "tk-phone-input-menu-country-label" }, country.label), country.dialCode && h("span", { class: "tk-phone-input-menu-dial-id" }, country.dialCode))))));
    }
    /**
     * Create the phone number input field.
     */
    renderPhoneInput() {
        var _a, _b;
        return (h("input", { id: this.uniqueId, type: "tel", class: "tk-phone-input-input", autoComplete: "off", placeholder: ((_a = this.selectedCountry) === null || _a === void 0 ? void 0 : _a.placeholder) || this.placeholder || ((_b = this.selectedCountry.mask) === null || _b === void 0 ? void 0 : _b.replace(/9/g, '9')), value: this.inputValue, onInput: this.handleInput, onBlur: this.handleInputBlur, onFocus: this.handleInputFocus, disabled: this.disabled, ref: el => (this.inputRef = el) }));
    }
    /**
     * Render the flag element for a country. Shows a close icon for countries without a dial code.
     */
    renderFlag(country) {
        const flagClass = this.getFlagClass(country);
        return country.dialCode ? (h("div", { class: flagClass, "aria-label": `${country.label} flag` })) : (h("div", { class: flagClass }, h("tk-icon", Object.assign({}, getIconElementProps('close', { color: 'var(--static-white)' })))));
    }
    /**
     * Render the component.
     */
    render() {
        return (h("div", { key: '7622740fa83a9e25aca50bed39634dfe9f12b578', class: classNames('tk-phone-input-container', `tk-phone-input-container-${this.size}`), "aria-invalid": this.invalid, "aria-disabled": this.disabled, "aria-readonly": this.readonly }, this.renderLabel(), h("div", { key: '86a1c6d319cea43d53824732efb1095244c247bb', class: "tk-phone-input-wrapper" }, this.renderCountrySelector(), this.renderPhoneInput()), renderHint(this.hint, this.error, this.invalid)));
    }
    static get formAssociated() { return true; }
    get el() { return getElement(this); }
    static get watchers() { return {
        "value": [{
                "valueChanged": 0
            }],
        "disabled": [{
                "disabledChanged": 0
            }]
    }; }
};
TkPhoneInput.style = tkPhoneInputCss();

export { TkPhoneInput as tk_phone_input };
