// @takeoff-ui/core@0.11.9/dist/esm/tk-popover.entry.js
import { r as registerInstance, c as createEvent, h, d as getElement } from './index-D4ALLk7q.js';
import { f as floatingElementAutoUpdate } from './position-utils-BiHzhysf.js';
import { C as ClickOutsideMixin } from './clickoutside-mixin-C0U9wSBw.js';
import { g as getDataTestId } from './test-id-utils-BDMnqEGL.js';

const tkPopoverCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-popover-content{display:flex;gap:8px;position:fixed;padding:var(--tooltips-base-v-padding) var(--tooltips-base-h-padding);border-radius:var(--tooltips-base-radius, 8px);width:max-content;color:var(--text-base);font-family:var(--family-body);line-height:20px;z-index:1300;box-shadow:var(--shadow-base)}.tk-popover-content[popover]{margin:0;inset:auto;overflow:visible}.tk-popover-content.tk-popover-dark{border:1px solid var(--border-light);background:var(--background-darkest)}.tk-popover-content.tk-popover-dark .tk-popover-arrow{background:var(--background-darkest)}.tk-popover-content.tk-popover-white{border:1px solid var(--border-light);background:var(--static-light)}.tk-popover-content.tk-popover-white .tk-popover-arrow{background:var(--static-light)}.tk-popover-content.tk-popover-basic{border:1px solid var(--states-light);background:var(--background-lightest)}.tk-popover-content.tk-popover-basic .tk-popover-arrow{background:var(--background-lightest)}`;

const TkPopover = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.tkChange = createEvent(this, "tk-change", 7);
        this.isOpen = false;
        /**
         * Controls if popover has custom content.
         * @defaultValue false
         */
        this.hasContentSlot = false;
        /**
         * Sets the action of the popover.
         * @defaultValue 'click'
         */
        this.trigger = 'click';
        /**
         * Sets the type of the popover.
         * @defaultValue "white"
         */
        this.type = 'basic';
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        /**
         * Click outside handler implementation - called by the mixin
         */
        this.closeHandler = () => {
            this.isOpen = false;
        };
    }
    isOpenChanged() {
        this.tkChange.emit(this.isOpen);
    }
    positionChanged() {
        if (this.popoverElement) {
            this.updatePosition();
        }
    }
    get isHover() {
        return this.trigger === 'hover';
    }
    componentWillLoad() {
        this.hasContentSlot = !!this.el.querySelector('[slot="content"]');
    }
    componentDidLoad() {
        var _a, _b, _c;
        // Initialize click outside mixin
        this.clickOutsideMixin = new ClickOutsideMixin({
            referenceElement: this.el,
            handler: this.closeHandler,
            disabled: this.isHover || !this.isOpen,
        });
        this.triggerElement = this.el.querySelector('[slot="trigger"]');
        if (this.trigger === 'hover') {
            (_a = this.triggerElement) === null || _a === void 0 ? void 0 : _a.addEventListener('mouseenter', () => (this.isOpen = true));
            (_b = this.triggerElement) === null || _b === void 0 ? void 0 : _b.addEventListener('mouseleave', () => (this.isOpen = false));
        }
        else {
            (_c = this.triggerElement) === null || _c === void 0 ? void 0 : _c.addEventListener('click', () => (this.isOpen = !this.isOpen));
        }
    }
    disconnectedCallback() {
        var _a, _b, _c, _d, _e;
        if (this.trigger === 'hover') {
            (_a = this.triggerElement) === null || _a === void 0 ? void 0 : _a.removeEventListener('mouseenter', () => (this.isOpen = true));
            (_b = this.triggerElement) === null || _b === void 0 ? void 0 : _b.removeEventListener('mouseleave', () => (this.isOpen = false));
        }
        else {
            (_c = this.triggerElement) === null || _c === void 0 ? void 0 : _c.removeEventListener('click', () => (this.isOpen = !this.isOpen));
        }
        // Clean up floating UI listeners on component unmount
        (_d = this.cleanup) === null || _d === void 0 ? void 0 : _d.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        // Call mixin's disconnectedCallback for cleanup
        (_e = this.clickOutsideMixin) === null || _e === void 0 ? void 0 : _e.disconnectedCallback();
    }
    componentDidUpdate() {
        var _a, _b;
        // Update click outside disabled state based on trigger type and open state
        this.clickOutsideMixin.updateConfig({ disabled: this.isHover || !this.isOpen });
        if (this.isOpen) {
            // Promote to the top layer before positioning so it cannot be trapped
            // behind a sticky table cell's stacking context.
            this.showInTopLayer();
            // Clean up old floating UI listeners before setting up new ones
            (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
            this.updatePosition();
        }
        else {
            // Remove floating UI listeners when popover closes
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
        }
    }
    /**
     * Closes the popover
     */
    async close() {
        this.isOpen = false;
    }
    updatePosition() {
        this.cleanup = floatingElementAutoUpdate(this.triggerElement, this.popoverElement, this.arrowElement, { placement: this.position });
    }
    /**
     * Promote the panel to the browser top layer via the native Popover API.
     * The top layer escapes every ancestor stacking context (e.g. the per-cell
     * z-index of sticky tk-table columns), so the panel can no longer be hidden
     * behind a neighbouring cell. The element stays inside this component's shadow
     * root, so slotted content, scoped styles and click-outside detection keep
     * working. No-ops when the API is unavailable (older browsers).
     */
    showInTopLayer() {
        const el = this.popoverElement;
        if (typeof (el === null || el === void 0 ? void 0 : el.showPopover) !== 'function' || !el.isConnected)
            return;
        // showPopover() throws if the element is already in the top layer.
        if (el.matches(':popover-open'))
            return;
        try {
            el.showPopover();
        }
        catch (_a) {
            /* element not eligible (e.g. detached during a re-render) — ignore */
        }
    }
    render() {
        return (h("div", { key: 'a8102448b35b6d36483f3fc894cf3e501e65f913', class: "tk-popover", "data-testid": getDataTestId(this.dataTestid, 'container') }, h("slot", { key: 'e990eab12540553215603e8a4048f4d4e6bd0a2f', name: "trigger" }), this.isOpen && (h("div", { key: '170c6b08b4a9abdc133513b1a55c514f6dc3f4fe', ref: el => (this.popoverElement = el),
            // `manual` keeps the browser's light-dismiss from fighting our
            // ClickOutsideMixin, while still rendering the panel in the top layer.
            popover: "manual", class: {
                'tk-popover-content': true,
                [`tk-popover-${this.type}`]: true,
            }, style: Object.assign({}, this.containerStyle), role: "popover", onClick: e => e.stopPropagation(), "data-testid": getDataTestId(this.dataTestid, 'content') }, this.hasContentSlot && h("slot", { key: '217923e970a0c1050fafd595a1977ea3d0789b43', name: "content" }), h("div", { key: '0ab91df3542c594b8185fa73f4fe4c4fb0ee778f', ref: el => (this.arrowElement = el), class: "tk-popover-arrow", "data-testid": getDataTestId(this.dataTestid, 'arrow') })))));
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "isOpen": [{
                "isOpenChanged": 0
            }],
        "position": [{
                "positionChanged": 0
            }]
    }; }
};
TkPopover.style = tkPopoverCss();

export { TkPopover as tk_popover };
