// @takeoff-ui/core@0.11.9/dist/esm/tk-toggle-button_2.entry.js
import { r as registerInstance, c as createEvent, h, d as getElement } from './index-D4ALLk7q.js';
import { c as classNames } from './index-D7YaJBkB.js';
import { r as renderIcons } from './icon-utils-SXopALR1.js';
import { g as getDataTestId } from './test-id-utils-BDMnqEGL.js';
import { i as isEqual } from './isEqual-CxgWR-W5.js';
import './_Uint8Array-B3_jBex_.js';

const tkToggleButtonCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:flex;width:100%}.tk-toggle-button{display:flex;align-items:center;justify-content:center;flex:1 0 auto;border:0px;padding:var(--button-sizes-large-v-padding) var(--button-sizes-base-h-padding);gap:var(--button-sizes-base-gap);cursor:pointer;border-radius:var(--button-sizes-base-radius);background-color:transparent}.tk-toggle-button.selected{box-shadow:0 2px 2px -1px var(--shadow-black-alpha-base, rgba(82, 88, 102, 0.08)), 0 6px 12px -2px var(--shadow-black-alpha-base, rgba(82, 88, 102, 0.08)), 0 8px 20px 0 var(--shadow-black-alpha-base, rgba(82, 88, 102, 0.08))}.tk-toggle-button.neutral.filled.selected{background-color:var(--background-darkest)}.tk-toggle-button.neutral.outlined.selected{background-color:var(--background-lightest);border:1px solid var(--border-sub-base)}.tk-toggle-button.neutral.text.selected{background-color:var(--static-light)}.tk-toggle-button.neutral.raised.selected{background-color:var(--background-darkest);border:1px solid var(--static-black)}.tk-toggle-button.neutral.filled-light.selected{background-color:var(--background-light)}.tk-toggle-button.neutral.filled.selected .tk-toggle-button-label,.tk-toggle-button.neutral.raised.selected .tk-toggle-button-label{color:var(--text-lightest)}.tk-toggle-button.neutral.outlined.selected .tk-toggle-button-label,.tk-toggle-button.neutral.text.selected .tk-toggle-button-label,.tk-toggle-button.neutral.filled-light.selected .tk-toggle-button-label{color:var(--background-darkest)}.tk-toggle-button.primary.filled.selected{background-color:var(--primary-base)}.tk-toggle-button.primary.outlined.selected{background-color:var(--background-lightest);border:1px solid var(--primary-sub-base)}.tk-toggle-button.primary.text.selected{background-color:var(--static-light)}.tk-toggle-button.primary.raised.selected{background-color:var(--primary-base);border:1px solid var(--primary-sub-base)}.tk-toggle-button.primary.filled-light.selected{background-color:var(--primary-lightest)}.tk-toggle-button.primary.filled.selected .tk-toggle-button-label,.tk-toggle-button.primary.raised.selected .tk-toggle-button-label{color:var(--static-white)}.tk-toggle-button.primary.outlined.selected .tk-toggle-button-label,.tk-toggle-button.primary.text.selected .tk-toggle-button-label,.tk-toggle-button.primary.filled-light.selected .tk-toggle-button-label{color:var(--primary-base)}.tk-toggle-button .tk-toggle-button-label{color:var(--text-base);font-family:var(--desktop-body-m-base-font);font-weight:400;font-style:normal}.tk-toggle-button .tk-toggle-button-label.large{line-height:var(--desktop-label-l-line-height);font-size:var(--desktop-label-l-size)}.tk-toggle-button .tk-toggle-button-label.base{line-height:var(--desktop-label-m-base-line-height);font-size:var(--desktop-label-m-base-size)}.tk-toggle-button .tk-toggle-button-label.small{line-height:var(--desktop-label-s-line-height);font-size:var(--desktop-label-s-size)}.tk-toggle-button.small{height:0px;padding:var(--button-sizes-large-v-padding) var(--button-sizes-small-h-padding)}.tk-toggle-button.base{height:8px}.tk-toggle-button.large{height:16px;padding:var(--button-sizes-large-v-padding) var(--button-sizes-large-h-padding);gap:var(--button-sizes-large-gap)}.tk-toggle-button.rounded{border-radius:999px}.tk-toggle-button:hover .tk-toggle-button-label,.tk-toggle-button:hover tk-icon i{color:var(--text-darkest)}.tk-toggle-button.disabled{pointer-events:none;user-select:none}.tk-toggle-button.disabled .tk-toggle-button-label{color:var(--text-sub-base)}`;

const TkToggleButton = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.tkToggle = createEvent(this, "tk-toggle", 7);
        /**
         * Defines the position of the icon.
         * @defaultValue 'left'
         */
        this.iconPosition = 'left';
        /**
         * Label text displayed inside the button.
         */
        this.label = '';
        /**
         * Sets size for the component.
         * @defaultValue 'base'
         */
        this.size = 'base';
        /**
         * Determines the button's variant for different styles.
         */
        this.variant = 'neutral';
        /**
         * The value of the type toggle button.
         */
        this.type = 'filled';
        /**
         * Whether the button is selected.
         */
        this.selected = false;
        this.handleClick = (e) => {
            if (this.disabled) {
                e.preventDefault();
                e.stopPropagation();
                return;
            }
            this.tkToggle.emit({ value: this.value, selected: !this.selected });
        };
    }
    getIconColor() {
        if (this.disabled)
            return 'var(--icon-sub-base)';
        if (this.selected) {
            if (this.type === 'filled' || this.type === 'raised')
                return 'var(--static-white)';
            if (this.variant === 'neutral')
                return 'var(--icon-darkest)';
            if (this.variant === 'primary')
                return 'var(--primary-base)';
        }
        return null;
    }
    createIcons() {
        return renderIcons(this.icon, {
            size: this.size,
            variant: 'neutral',
            additionalProps: { color: this.getIconColor() },
            dataTestid: this.dataTestid,
        }, this.iconPosition);
    }
    renderLabel() {
        var _a;
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            return (h("span", { class: classNames('tk-toggle-button-label', this.size), "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label));
        }
        return null;
    }
    render() {
        const rootClasses = classNames('tk-toggle-button', this.variant, this.type, this.size, {
            rounded: this.rounded,
            selected: this.selected && !this.disabled,
            disabled: this.disabled,
        });
        const icon = this.createIcons();
        return (h("button", { key: 'eb6b89c8ba9d6df6d97625d0142a0e9e73d827ae', class: rootClasses, disabled: this.disabled, onClick: this.handleClick, "data-testid": getDataTestId(this.dataTestid, 'button') }, icon.leftIcon, this.renderLabel(), icon.rightIcon));
    }
    get el() { return getElement(this); }
};
TkToggleButton.style = tkToggleButtonCss();

const tkToggleButtonGroupCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-toggle-button-group{display:flex;align-items:center;gap:8px;padding:4px;align-self:stretch;background:var(--static-light);border-radius:12px}.tk-toggle-button-group.basic{background-color:transparent}.tk-toggle-button-group.light{border:1px solid var(--border-light, #e1e4ea);background:var(--background-lightest)}.tk-toggle-button-group.divided{border:1px solid var(--border-light, #e1e4ea)}.tk-toggle-button-group.horizontal{flex-direction:row}.tk-toggle-button-group.horizontal.rounded{border-radius:999px}.tk-toggle-button-group.vertical{flex-direction:column}.tk-toggle-button-group.vertical.rounded{border-radius:24px}`;

const TkToggleButtonGroup = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.tkChange = createEvent(this, "tk-change", 7);
        /**
         * The value type of the toggle button group.
         */
        this.type = 'basic';
        /**
         * The value of the rounded toggle button group.
         */
        this.rounded = false;
        /**
         * The direction of the toggle button group.
         */
        this.direction = 'horizontal';
    }
    roundedChanged() {
        this.updateSlottedItems();
    }
    valueChanged() {
        this.updateSelected();
        this.tkChange.emit(this.value);
    }
    componentWillLoad() {
        this.updateSlottedItems();
    }
    updateSlottedItems() {
        this.slottedItems = this.el.querySelectorAll('tk-toggle-button');
        this.updateSelected();
        this.slottedItems.forEach(item => {
            item.rounded = this.rounded;
            item.addEventListener('tk-toggle', this.handleToggle.bind(this));
        });
    }
    updateSelected() {
        var _a;
        if (((_a = this.slottedItems) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.slottedItems.forEach(item => {
                if (item.disabled) {
                    if (item.selected) {
                        item.selected = false;
                        item.value = undefined;
                    }
                    return;
                }
                item.selected = isEqual(this.value, item.value);
            });
        }
    }
    handleToggle(e) {
        const { value } = e.detail;
        this.value = value;
        this.updateSelected();
    }
    render() {
        const groupClasses = classNames('tk-toggle-button-group', this.type, {
            rounded: this.rounded,
            vertical: this.direction === 'vertical',
            horizontal: this.direction === 'horizontal',
        });
        return (h("div", { key: '3bb7ea989256d18ee0e94fbcaf825e0c64d4ea69', class: groupClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("slot", { key: '88e50487b5e744a4082df70f626a41259ab41c18', onSlotchange: () => this.updateSlottedItems() })));
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "rounded": [{
                "roundedChanged": 0
            }],
        "value": [{
                "valueChanged": 0
            }]
    }; }
};
TkToggleButtonGroup.style = tkToggleButtonGroupCss();

export { TkToggleButton as tk_toggle_button, TkToggleButtonGroup as tk_toggle_button_group };
