// @takeoff-ui/core@0.11.9/dist/collection/components/tk-timeline-item/tk-timeline-item.js
import { h, Host } from "@stencil/core";
/**
 * The TimelineItem component for custom timeline content.
 * This component is used within the `tk-timeline` component to represent individual items in a timeline.
 * It automatically assigns a slot name based on its index within the parent timeline.
 * * @react `import { TkTimelineItem } from '@takeoff-ui/react'`
 * * @vue `import { TkTimelineItem } from '@takeoff-ui/vue'`
 * * @angular `import { TkTimelineItem } from '@takeoff-ui/angular'`
 */
export class TkTimelineItem {
    componentDidLoad() {
        // Timeline item parent'a slot ismi ata
        const timeline = this.el.closest('tk-timeline');
        if (timeline) {
            const siblings = Array.from(timeline.querySelectorAll('tk-timeline-item'));
            const index = siblings.indexOf(this.el);
            this.el.slot = `item-${index}`;
        }
    }
    render() {
        return (h(Host, { key: 'ef999f4a728a6a7f248b3ab70f5f6ae3c24987ca' }, h("slot", { key: '0093189d42688a21ba6f75451793104e3dbd8b1f' })));
    }
    static get is() { return "tk-timeline-item"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-timeline-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-timeline-item.css"]
        };
    }
    static get elementRef() { return "el"; }
}
