// @takeoff-ui/core@0.11.9/dist/collection/components/tk-gantt-chart/tk-gantt-chart.js
import { h, Fragment } from "@stencil/core";
import classNames from "classnames";
import { DEFAULT_COLUMNS, DEFAULT_ROW_HEIGHT, DEFAULT_PANEL_WIDTH, DEFAULT_HEADER_HEIGHT, DEFAULT_WEEK_START_DAY, VIEW_TYPE_DAY_WIDTH } from "./defaults";
import { autoFitViewType, flattenTasks, getTaskBarPosition, getBarPositionFromDates, getIndicatorPosition, getTimelineRange, generatePrimaryHeaders, generateSecondaryHeaders, isWeekend, isHoliday, addDays, toDateOnly, } from "./helpers";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * TkGanttChart is a display-only Gantt chart component with expandable tasks,
 * configurable view types (weekly, monthly, quarterly, yearly), two-layer headers,
 * indicators, weekend/holiday highlighting, and customisable tooltips / task bars.
 * @slot empty - Content to show when there are no tasks
 * @react `import { TkGanttChart } from '@takeoff-ui/react'`
 * @vue `import { TkGanttChart } from '@takeoff-ui/vue'`
 * @angular `import { TkGanttChart } from '@takeoff-ui/angular'`
 */
export class TkGanttChart {
    constructor() {
        // ---------------------------------------------------------------------------
        // State
        // ---------------------------------------------------------------------------
        this.expandedIds = new Set();
        this.computedViewType = 'monthly';
        // ---------------------------------------------------------------------------
        // Props
        // ---------------------------------------------------------------------------
        /**
         * Array of tasks to display. Each task may contain nested children for sub-tasks.
         */
        this.tasks = [];
        /**
         * Column definitions for the left-side task list panel.
         * Columns are fed from the task data.
         */
        this.columns = DEFAULT_COLUMNS;
        /**
         * Secondary header display mode.
         * - `'days'` shows individual day numbers or day names
         * - `'weeks'` shows week numbers (W1, W2 … / H1, H2 … for Turkish locale)
         * @defaultValue 'days'
         */
        this.secondaryHeaderMode = 'days';
        /**
         * Day the week starts on (0 = Sunday … 6 = Saturday).
         * @defaultValue 1 (Monday)
         */
        this.weekStartDay = DEFAULT_WEEK_START_DAY;
        /**
         * Array of indicator lines rendered on the timeline (e.g. today, startDate, freezeDate).
         */
        this.indicators = [];
        /**
         * Whether to automatically show a today indicator line.
         * @defaultValue true
         */
        this.showTodayIndicator = true;
        /**
         * Whether to highlight weekend columns with a different background.
         * @defaultValue true
         */
        this.highlightWeekends = true;
        /**
         * Array of holiday dates to highlight on the timeline.
         */
        this.holidays = [];
        /**
         * Row height in pixels.
         * @defaultValue 40
         */
        this.rowHeight = DEFAULT_ROW_HEIGHT;
        /**
         * Width of the left-side task list panel in pixels.
         * @defaultValue 320
         */
        this.panelWidth = DEFAULT_PANEL_WIDTH;
        /**
         * Whether to hide the left-side task list panel.
         * @defaultValue false
         */
        this.hidePanel = false;
        /**
         * Locale string for date formatting (e.g. 'tr-TR', 'en-US').
         * Uses native `toLocaleString` under the hood.
         * @defaultValue 'default'
         */
        this.locale = 'default';
    }
    tasksChanged() {
        this.recalculate();
    }
    viewTypeChanged() {
        this.recalculate();
    }
    // ---------------------------------------------------------------------------
    // Lifecycle
    // ---------------------------------------------------------------------------
    componentWillLoad() {
        this.recalculate();
    }
    // ---------------------------------------------------------------------------
    // Private methods
    // ---------------------------------------------------------------------------
    recalculate() {
        var _a;
        this.computedViewType = (_a = this.viewType) !== null && _a !== void 0 ? _a : autoFitViewType(this.tasks);
    }
    // ---------------------------------------------------------------------------
    // Handle event handlers
    // ---------------------------------------------------------------------------
    handleToggleRow(task, e) {
        e.stopPropagation();
        const newSet = new Set(this.expandedIds);
        const isExpanded = newSet.has(task.id);
        if (isExpanded) {
            newSet.delete(task.id);
        }
        else {
            newSet.add(task.id);
        }
        this.expandedIds = newSet;
        this.tkTaskToggle.emit({ task, expanded: !isExpanded });
    }
    handleTaskClick(task) {
        this.tkTaskClick.emit(task);
    }
    handleTimelineScroll(e) {
        const target = e.target;
        if (this.panelRef) {
            this.panelRef.scrollTop = target.scrollTop;
        }
    }
    handlePanelScroll(e) {
        const target = e.target;
        if (this.timelineRef) {
            this.timelineRef.scrollTop = target.scrollTop;
        }
    }
    // ---------------------------------------------------------------------------
    // Create factory methods
    // ---------------------------------------------------------------------------
    createTaskBar(task, barPos, segment, segmentIndex) {
        var _a, _b, _c;
        const taskKey = String(task.id);
        const barKey = segmentIndex != null ? `${taskKey}-segment-${segmentIndex}` : taskKey;
        const progress = (_a = segment === null || segment === void 0 ? void 0 : segment.progress) !== null && _a !== void 0 ? _a : task.progress;
        const barStyle = (_b = segment === null || segment === void 0 ? void 0 : segment.style) !== null && _b !== void 0 ? _b : task.style;
        const barLabel = (_c = segment === null || segment === void 0 ? void 0 : segment.label) !== null && _c !== void 0 ? _c : task.name;
        const commonStyle = Object.assign({ left: `${barPos.left}px`, width: `${barPos.width}px`, height: `${this.rowHeight - 12}px` }, barStyle);
        const tooltipContent = this.createTooltipContent(task, barKey);
        if (this.taskBarHtml) {
            const content = this.taskBarHtml(task);
            const barContent = typeof content === 'string' ? (h("div", { class: "custom-task-bar", innerHTML: content, "data-testid": getDataTestId(this.dataTestid, 'custom-task-bar', barKey) })) : (h("div", { class: "custom-task-bar", ref: el => el && el.replaceChildren(content), "data-testid": getDataTestId(this.dataTestid, 'custom-task-bar', barKey) }));
            return (h("tk-tooltip", { position: "top", variant: "dark", dataTestid: getDataTestId(this.dataTestid, 'custom-task-bar-tooltip', barKey) }, h("div", { slot: "trigger", class: "tk-gantt-chart-task-bar tk-gantt-chart-task-bar-custom", style: commonStyle, onClick: () => this.handleTaskClick(task), "data-testid": getDataTestId(this.dataTestid, 'custom-task-bar-trigger', barKey) }, barContent), h("div", { slot: "content", "data-testid": getDataTestId(this.dataTestid, 'custom-task-bar-tooltip-content', barKey) }, tooltipContent)));
        }
        return (h("tk-tooltip", { position: "top", variant: "dark", dataTestid: getDataTestId(this.dataTestid, 'task-bar-tooltip', barKey) }, h("div", { slot: "trigger", class: classNames('tk-gantt-chart-task-bar', { 'tk-gantt-chart-task-bar-segment': segmentIndex != null }), style: commonStyle, onClick: () => this.handleTaskClick(task), "data-testid": getDataTestId(this.dataTestid, 'task-bar', barKey) }, progress != null && progress > 0 && (h("div", { class: "tk-gantt-chart-task-bar-progress", style: { width: `${Math.min(progress, 100)}%` }, "data-testid": getDataTestId(this.dataTestid, 'task-bar-progress', barKey) })), h("span", { class: "tk-gantt-chart-task-bar-label", "data-testid": getDataTestId(this.dataTestid, 'task-bar-label', barKey) }, barLabel)), h("div", { slot: "content", "data-testid": getDataTestId(this.dataTestid, 'task-bar-tooltip-content', barKey) }, tooltipContent)));
    }
    createIndicatorLine(indicator, minDate, bodyHeight, indicatorKey) {
        const left = getIndicatorPosition(indicator.date, minDate, this.computedViewType);
        return (h("div", { class: classNames('tk-gantt-chart-indicator', {
                'tk-gantt-chart-indicator-dashed': indicator.lineStyle === 'dashed',
            }), style: {
                left: `${left}px`,
                height: `${bodyHeight}px`,
                borderLeftColor: indicator.color || 'var(--primary-base)',
                borderLeftStyle: indicator.lineStyle === 'dashed' ? 'dashed' : 'solid',
            }, "data-testid": getDataTestId(this.dataTestid, 'indicator', indicatorKey) }, indicator.label && (h("span", { class: "tk-gantt-chart-indicator-label", "data-testid": getDataTestId(this.dataTestid, 'indicator-label', indicatorKey) }, indicator.label))));
    }
    createTooltipContent(task, tooltipKey) {
        if (this.tooltipHtml) {
            const content = this.tooltipHtml(task);
            if (typeof content === 'string') {
                return h("div", { innerHTML: content, "data-testid": getDataTestId(this.dataTestid, 'custom-tooltip', tooltipKey) });
            }
            return h("div", { ref: el => el && el.replaceChildren(content), "data-testid": getDataTestId(this.dataTestid, 'custom-tooltip', tooltipKey) });
        }
        const startLabel = toDateOnly(task.startDate).toLocaleDateString(this.locale);
        const endLabel = toDateOnly(task.endDate).toLocaleDateString(this.locale);
        return (h(Fragment, null, h("div", { class: "tk-gantt-chart-tooltip-title", "data-testid": getDataTestId(this.dataTestid, 'tooltip-title', tooltipKey) }, task.name), h("div", { class: "tk-gantt-chart-tooltip-row", "data-testid": getDataTestId(this.dataTestid, 'tooltip-start', tooltipKey) }, h("span", { class: "tk-gantt-chart-tooltip-label", "data-testid": getDataTestId(this.dataTestid, 'tooltip-start-label', tooltipKey) }, "Start:"), startLabel), h("div", { class: "tk-gantt-chart-tooltip-row", "data-testid": getDataTestId(this.dataTestid, 'tooltip-end', tooltipKey) }, h("span", { class: "tk-gantt-chart-tooltip-label", "data-testid": getDataTestId(this.dataTestid, 'tooltip-end-label', tooltipKey) }, "End:"), endLabel), task.progress != null && (h("div", { class: "tk-gantt-chart-tooltip-row", "data-testid": getDataTestId(this.dataTestid, 'tooltip-progress', tooltipKey) }, h("span", { class: "tk-gantt-chart-tooltip-label", "data-testid": getDataTestId(this.dataTestid, 'tooltip-progress-label', tooltipKey) }, "Progress:"), task.progress, "%"))));
    }
    // ---------------------------------------------------------------------------
    // Render helpers
    // ---------------------------------------------------------------------------
    renderPanel(flatRows) {
        if (this.hidePanel)
            return null;
        const headerHeight = DEFAULT_HEADER_HEIGHT * 2; // two header layers
        return (h("div", { class: "tk-gantt-chart-panel", style: { width: `${this.panelWidth}px` }, ref: el => (this.panelRef = el), onScroll: (e) => this.handlePanelScroll(e), "data-testid": getDataTestId(this.dataTestid, 'panel') }, h("div", { class: "tk-gantt-chart-panel-header", style: { height: `${headerHeight}px` }, "data-testid": getDataTestId(this.dataTestid, 'panel-header') }, this.columns.map(col => (h("div", { class: "tk-gantt-chart-panel-header-cell", style: { width: col.width || 'auto', minWidth: col.width || '80px' }, "data-testid": getDataTestId(this.dataTestid, 'panel-header-cell', String(col.field)) }, col.header)))), h("div", { class: "tk-gantt-chart-panel-body", "data-testid": getDataTestId(this.dataTestid, 'panel-body') }, flatRows.map(({ task, depth }) => (h("div", { class: "tk-gantt-chart-panel-row", style: { height: `${this.rowHeight}px` }, "data-testid": getDataTestId(this.dataTestid, 'panel-row', String(task.id)) }, this.columns.map((col, colIndex) => {
            var _a, _b;
            return (h("div", { class: "tk-gantt-chart-panel-cell", style: {
                    width: col.width || 'auto',
                    minWidth: col.width || '80px',
                    paddingLeft: colIndex === 0 ? `${depth * 20 + 8}px` : '8px',
                }, "data-testid": getDataTestId(this.dataTestid, 'panel-cell', String(task.id), String(col.field)) }, colIndex === 0 && ((_b = (_a = task.children) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) > 0 && (h("button", { class: classNames('tk-gantt-chart-expand-btn', {
                    'tk-gantt-chart-expand-btn-open': this.expandedIds.has(task.id),
                }), onClick: (e) => this.handleToggleRow(task, e), "data-testid": getDataTestId(this.dataTestid, 'panel-expand-button', String(task.id)) }, h("tk-icon", { icon: "chevron_right", size: "small", dataTestid: getDataTestId(this.dataTestid, 'panel-expand-icon', String(task.id)) }))), col.html ? (h("span", { innerHTML: col.html(task), "data-testid": getDataTestId(this.dataTestid, 'panel-cell-content', String(task.id), String(col.field)) })) : (h("span", { "data-testid": getDataTestId(this.dataTestid, 'panel-cell-content', String(task.id), String(col.field)) }, task[col.field]))));
        })))))));
    }
    renderTimelineHeader(start, end, totalWidth) {
        const vt = this.computedViewType;
        const dayWidth = VIEW_TYPE_DAY_WIDTH[vt];
        const primaryCells = generatePrimaryHeaders(start, end, vt, this.locale, this.weekStartDay);
        const secondaryCells = generateSecondaryHeaders(start, end, vt, this.locale, this.weekStartDay, this.secondaryHeaderMode);
        return (h("div", { class: "tk-gantt-chart-timeline-headers", style: { width: `${totalWidth}px` }, "data-testid": getDataTestId(this.dataTestid, 'timeline-headers') }, h("div", { class: "tk-gantt-chart-timeline-header tk-gantt-chart-timeline-header-primary", style: { height: `${DEFAULT_HEADER_HEIGHT}px` }, "data-testid": getDataTestId(this.dataTestid, 'timeline-primary') }, primaryCells.map((cell, index) => (h("div", { class: "tk-gantt-chart-timeline-header-cell", style: { width: `${cell.span * dayWidth}px` }, "data-testid": getDataTestId(this.dataTestid, 'timeline-primary-cell', index.toString()) }, cell.label)))), h("div", { class: "tk-gantt-chart-timeline-header tk-gantt-chart-timeline-header-secondary", style: { height: `${DEFAULT_HEADER_HEIGHT}px` }, "data-testid": getDataTestId(this.dataTestid, 'timeline-secondary') }, secondaryCells.map((cell, index) => {
            const cellStart = cell.startDate;
            const isWeekendCell = cell.span === 1 && isWeekend(cellStart);
            const isHolidayCell = cell.span === 1 && isHoliday(cellStart, this.holidays);
            return (h("div", { class: classNames('tk-gantt-chart-timeline-header-cell', {
                    'tk-gantt-chart-timeline-header-cell-weekend': isWeekendCell && this.highlightWeekends,
                    'tk-gantt-chart-timeline-header-cell-holiday': isHolidayCell,
                }), style: { width: `${cell.span * dayWidth}px` }, "data-testid": getDataTestId(this.dataTestid, 'timeline-secondary-cell', index.toString()) }, cell.label));
        }))));
    }
    renderTimelineBody(flatRows, start, totalWidth, totalDays) {
        const dayWidth = VIEW_TYPE_DAY_WIDTH[this.computedViewType];
        const bodyHeight = flatRows.length * this.rowHeight;
        return (h("div", { class: "tk-gantt-chart-timeline-body", style: { width: `${totalWidth}px` }, "data-testid": getDataTestId(this.dataTestid, 'timeline-body') }, (this.highlightWeekends || this.holidays.length > 0) &&
            Array.from({ length: totalDays }).map((_, i) => {
                const d = addDays(start, i);
                const weekend = this.highlightWeekends && isWeekend(d);
                const holiday = isHoliday(d, this.holidays);
                if (!weekend && !holiday)
                    return null;
                return (h("div", { class: classNames('tk-gantt-chart-column-highlight', {
                        'tk-gantt-chart-weekend': weekend,
                        'tk-gantt-chart-holiday': holiday,
                    }), style: {
                        left: `${i * dayWidth}px`,
                        width: `${dayWidth}px`,
                        height: `${bodyHeight}px`,
                    }, "data-testid": getDataTestId(this.dataTestid, 'column-highlight', i.toString()) }));
            }), this.showTodayIndicator &&
            this.createIndicatorLine({ date: new Date().toISOString(), label: 'Today', color: 'var(--primary-base)', lineStyle: 'solid' }, start, bodyHeight, 'today'), this.indicators.map((ind, index) => this.createIndicatorLine(ind, start, bodyHeight, `custom-${index}`)), flatRows.map(({ task }) => {
            const hasSegments = task.segments && task.segments.length > 0;
            return (h("div", { class: "tk-gantt-chart-timeline-row", style: { height: `${this.rowHeight}px` }, "data-testid": getDataTestId(this.dataTestid, 'timeline-row', String(task.id)) }, hasSegments
                ? task.segments.map((seg, idx) => {
                    const segPos = getBarPositionFromDates(seg.startDate, seg.endDate, start, this.computedViewType);
                    return this.createTaskBar(task, segPos, seg, idx);
                })
                : this.createTaskBar(task, getTaskBarPosition(task, start, this.computedViewType))));
        })));
    }
    renderEmptyState() {
        return (h("div", { class: "tk-gantt-chart-empty", "data-testid": getDataTestId(this.dataTestid, 'empty') }, h("slot", { name: "empty" }, h("span", { "data-testid": getDataTestId(this.dataTestid, 'empty-text') }, "No tasks to display"))));
    }
    // ---------------------------------------------------------------------------
    // render
    // ---------------------------------------------------------------------------
    render() {
        if (!this.tasks.length) {
            return (h("div", { class: "tk-gantt-chart", style: this.containerStyle, "data-testid": getDataTestId(this.dataTestid, 'container') }, this.renderEmptyState()));
        }
        const flatRows = flattenTasks(this.tasks, this.expandedIds);
        const { start, end, totalDays } = getTimelineRange(this.tasks, this.computedViewType, this.weekStartDay);
        const dayWidth = VIEW_TYPE_DAY_WIDTH[this.computedViewType];
        const totalWidth = totalDays * dayWidth;
        return (h("div", { class: "tk-gantt-chart", style: this.containerStyle, "data-testid": getDataTestId(this.dataTestid, 'container') }, this.renderPanel(flatRows), h("div", { class: "tk-gantt-chart-timeline", ref: el => (this.timelineRef = el), onScroll: (e) => this.handleTimelineScroll(e), "data-testid": getDataTestId(this.dataTestid, 'timeline') }, this.renderTimelineHeader(start, end, totalWidth), this.renderTimelineBody(flatRows, start, totalWidth, totalDays))));
    }
    static get is() { return "tk-gantt-chart"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-gantt-chart.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-gantt-chart.css"]
        };
    }
    static get properties() {
        return {
            "tasks": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IGanttTask[]",
                    "resolved": "IGanttTask[]",
                    "references": {
                        "IGanttTask": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::IGanttTask",
                            "referenceLocation": "IGanttTask"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Array of tasks to display. Each task may contain nested children for sub-tasks."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "columns": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IGanttColumn[]",
                    "resolved": "IGanttColumn[]",
                    "references": {
                        "IGanttColumn": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::IGanttColumn",
                            "referenceLocation": "IGanttColumn"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Column definitions for the left-side task list panel.\nColumns are fed from the task data."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "DEFAULT_COLUMNS"
            },
            "viewType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "GanttViewType",
                    "resolved": "\"monthly\" | \"quarterly\" | \"weekly\" | \"yearly\"",
                    "references": {
                        "GanttViewType": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::GanttViewType",
                            "referenceLocation": "GanttViewType"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "undefined (auto-fit)"
                        }],
                    "text": "Timeline view type. When not set the component auto-fits based on the task date range."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "view-type"
            },
            "secondaryHeaderMode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "GanttSecondaryHeaderMode",
                    "resolved": "\"days\" | \"weeks\"",
                    "references": {
                        "GanttSecondaryHeaderMode": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::GanttSecondaryHeaderMode",
                            "referenceLocation": "GanttSecondaryHeaderMode"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'days'"
                        }],
                    "text": "Secondary header display mode.\n- `'days'` shows individual day numbers or day names\n- `'weeks'` shows week numbers (W1, W2 \u2026 / H1, H2 \u2026 for Turkish locale)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "secondary-header-mode",
                "defaultValue": "'days'"
            },
            "weekStartDay": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "GanttWeekStartDay",
                    "resolved": "0 | 1 | 2 | 3 | 4 | 5 | 6",
                    "references": {
                        "GanttWeekStartDay": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::GanttWeekStartDay",
                            "referenceLocation": "GanttWeekStartDay"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "1 (Monday)"
                        }],
                    "text": "Day the week starts on (0 = Sunday \u2026 6 = Saturday)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "week-start-day",
                "defaultValue": "DEFAULT_WEEK_START_DAY"
            },
            "indicators": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IGanttIndicator[]",
                    "resolved": "IGanttIndicator[]",
                    "references": {
                        "IGanttIndicator": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::IGanttIndicator",
                            "referenceLocation": "IGanttIndicator"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Array of indicator lines rendered on the timeline (e.g. today, startDate, freezeDate)."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "showTodayIndicator": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Whether to automatically show a today indicator line."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-today-indicator",
                "defaultValue": "true"
            },
            "highlightWeekends": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Whether to highlight weekend columns with a different background."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "highlight-weekends",
                "defaultValue": "true"
            },
            "holidays": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IGanttHoliday[]",
                    "resolved": "IGanttHoliday[]",
                    "references": {
                        "IGanttHoliday": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::IGanttHoliday",
                            "referenceLocation": "IGanttHoliday"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Array of holiday dates to highlight on the timeline."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "tooltipHtml": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "GanttTooltipFunction",
                    "resolved": "(task: IGanttTask) => string | HTMLElement",
                    "references": {
                        "GanttTooltipFunction": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::GanttTooltipFunction",
                            "referenceLocation": "GanttTooltipFunction"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom tooltip render function. Receives the hovered task and should return\nan HTMLElement or an HTML string."
                },
                "getter": false,
                "setter": false
            },
            "taskBarHtml": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "GanttTaskBarFunction",
                    "resolved": "(task: IGanttTask) => string | HTMLElement",
                    "references": {
                        "GanttTaskBarFunction": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::GanttTaskBarFunction",
                            "referenceLocation": "GanttTaskBarFunction"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom task bar render function. Receives the task and should return\nan HTMLElement or an HTML string."
                },
                "getter": false,
                "setter": false
            },
            "rowHeight": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "40"
                        }],
                    "text": "Row height in pixels."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "row-height",
                "defaultValue": "DEFAULT_ROW_HEIGHT"
            },
            "panelWidth": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "320"
                        }],
                    "text": "Width of the left-side task list panel in pixels."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "panel-width",
                "defaultValue": "DEFAULT_PANEL_WIDTH"
            },
            "hidePanel": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether to hide the left-side task list panel."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-panel",
                "defaultValue": "false"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS style applied to the root container element."
                },
                "getter": false,
                "setter": false
            },
            "locale": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'default'"
                        }],
                    "text": "Locale string for date formatting (e.g. 'tr-TR', 'en-US').\nUses native `toLocaleString` under the hood."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "locale",
                "defaultValue": "'default'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "expandedIds": {},
            "computedViewType": {}
        };
    }
    static get events() {
        return [{
                "method": "tkTaskClick",
                "name": "tk-task-click",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a task bar is clicked. Detail contains the clicked task."
                },
                "complexType": {
                    "original": "IGanttTask",
                    "resolved": "IGanttTask",
                    "references": {
                        "IGanttTask": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::IGanttTask",
                            "referenceLocation": "IGanttTask"
                        }
                    }
                }
            }, {
                "method": "tkTaskToggle",
                "name": "tk-task-toggle",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a task row is expanded or collapsed."
                },
                "complexType": {
                    "original": "{ task: IGanttTask; expanded: boolean }",
                    "resolved": "{ task: IGanttTask; expanded: boolean; }",
                    "references": {
                        "IGanttTask": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-gantt-chart/types.ts::IGanttTask",
                            "referenceLocation": "IGanttTask"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "tasks",
                "methodName": "tasksChanged"
            }, {
                "propName": "viewType",
                "methodName": "viewTypeChanged"
            }];
    }
}
