// @takeoff-ui/core@0.11.9/dist/collection/components/tk-breadcrumb/tk-breadcrumb.js
import { h, Fragment } from "@stencil/core";
import classNames from "classnames";
import { getIconElementProps } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkBreadcrumb` provides a navigational aid, allowing users to keep track of their location within the application's hierarchy.
 * @slot default - Default slot to detect TkBreadcrumbItem components.
 * @react `import { TkBreadcrumb, TkBreadcrumbItem } from '@takeoff-ui/react'`
 * @vue `import { TkBreadcrumb, TkBreadcrumbItem } from '@takeoff-ui/vue'`
 * @angular `import { TkBreadcrumb, TkBreadcrumbItem } from '@takeoff-ui/angular'`
 */
export class TkBreadcrumb {
    constructor() {
        this.hasSlottedItems = false;
        /**
         * Type of separator to use between breadcrumb items
         * @default 'icon'
         */
        this.separator = 'icon';
        /**
         * Icon to use as separator when separator is set to 'icon'
         * @default 'chevron_right'
         */
        this.separatorIcon = 'chevron_right';
        /**
         * Defines the visual style of the component, possible values are 'basic' and 'outlined'.
         * @default 'basic'
         */
        this.type = 'basic';
    }
    componentWillLoad() {
        this.hasSlottedItems = !!this.el.querySelector('tk-breadcrumb-item');
    }
    getBreadcrumbItemProps(item, index) {
        var _a, _b;
        const { label, href, isExternal, icon } = item;
        return {
            key: index,
            label,
            href,
            icon,
            isExternal,
            isCurrent: index === ((_b = (_a = this.model) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) - 1,
            dataTestid: getDataTestId(this.dataTestid, 'item', index.toString()),
        };
    }
    renderSeparator(index) {
        const separatorDataTestid = getDataTestId(this.dataTestid, 'separator', index.toString());
        const separatorClasses = classNames('tk-breadcrumb-separator-icon', {
            'tk-breadcrumb-dot-separator': this.separator === 'dot',
            'tk-breadcrumb-slash-separator': this.separator === 'slash',
            'tk-breadcrumb-vertical-separator': this.separator === 'vertical',
        });
        if (this.separator === 'icon') {
            return (h("tk-icon", Object.assign({}, getIconElementProps(this.separatorIcon, { class: separatorClasses, color: 'var(--icon-sub-base)' }, 'rounded', 'span'), { dataTestid: getDataTestId(separatorDataTestid, 'icon') })));
        }
        return h("span", { class: separatorClasses, "data-testid": getDataTestId(separatorDataTestid, 'item') });
    }
    render() {
        var _a;
        const rootClasses = classNames('tk-breadcrumb', `tk-breadcrumb-${this.type}`);
        return (h("nav", { key: '266cb8b62ff2feda77d3bc749f83554f5bc7f8ae', class: rootClasses, "aria-label": "breadcrumb", "data-testid": getDataTestId(this.dataTestid, 'container') }, h("ol", { key: '8496e0dc3f511daa1c659cf74c4e9996fa9f6b7b', class: "tk-breadcrumb-list", "data-testid": getDataTestId(this.dataTestid, 'list') }, this.hasSlottedItems ? (h("slot", null)) : ((_a = this.model) === null || _a === void 0 ? void 0 : _a.map((item, index) => (h(Fragment, null, h("tk-breadcrumb-item", Object.assign({}, this.getBreadcrumbItemProps(item, index))), index < this.model.length - 1 && (h("li", { class: "tk-breadcrumb-separator", "data-testid": getDataTestId(this.dataTestid, 'separator', index.toString()) }, this.renderSeparator(index))))))))));
    }
    static get is() { return "tk-breadcrumb"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-breadcrumb.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-breadcrumb.css"]
        };
    }
    static get properties() {
        return {
            "model": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IBreadcrumbModel[]",
                    "resolved": "IBreadcrumbModel[]",
                    "references": {
                        "IBreadcrumbModel": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-breadcrumb/types.ts::IBreadcrumbModel",
                            "referenceLocation": "IBreadcrumbModel"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Array of breadcrumb items"
                },
                "getter": false,
                "setter": false
            },
            "separator": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'icon' | 'dot' | 'slash' | 'vertical'",
                    "resolved": "\"dot\" | \"icon\" | \"slash\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "'icon'"
                        }],
                    "text": "Type of separator to use between breadcrumb items"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "separator",
                "defaultValue": "'icon'"
            },
            "separatorIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "'chevron_right'"
                        }],
                    "text": "Icon to use as separator when separator is set to 'icon'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "separator-icon",
                "defaultValue": "'chevron_right'"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'outlined'",
                    "resolved": "\"basic\" | \"outlined\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "'basic'"
                        }],
                    "text": "Defines the visual style of the component, possible values are 'basic' and 'outlined'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'basic'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hasSlottedItems": {}
        };
    }
    static get elementRef() { return "el"; }
}
