// @takeoff-ui/core@0.11.9/dist/collection/components/tk-dialog/tk-dialog.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { getIconElementProps } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkDialog` component provides a customizable modal dialog for displaying important information or requesting user input. It supports various configurations including different header types, variants, and customizable content.
 * @slot default - Default slot to detect child to inner content.
 * @slot container - Custom container template.
 * @slot header - Custom header template.
 * @slot content - Custom content template.
 * @slot footer - Custom footer template.
 * @slot footer-actions - Custom actions template to default footer.
 * @react `import { TkDialog } from '@takeoff-ui/react'`
 * @vue `import { TkDialog } from '@takeoff-ui/vue'`
 * @angular `import { TkDialog } from '@takeoff-ui/angular'`
 */
export class TkDialog {
    constructor() {
        /**
         * Controls the visibility of the dialog
         */
        this.visible = false;
        /**
         * Header type
         */
        this.headerType = 'basic';
        /**
         * Controls whether the close button is shown
         */
        this.showCloseButton = true;
        /**
         * Controls whether the header is shown
         */
        this.showHeader = true;
        /**
         * Controls whether the variant sign is shown
         */
        this.showVariantSign = true;
        /**
         * The variant of the dialog
         */
        this.variant = 'info';
        /**
         * Controls whether the backdrop is shown
         * @defaultValue false
         */
        this.hideBackdrop = false;
        /**
         * Appearance of the mask
         * @defaultValue base
         */
        this.maskVariant = 'base';
        /**
         * Controls whether the dialog has a blur background
         * @defaultValue false
         */
        this.isMaskBlur = false;
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        /**
         * Prevents the dialog from being dismissed by clicking outside
         * @defaultValue false
         */
        this.preventDismiss = false;
        this.handleCloseButtonClick = () => {
            this.tkClose.emit();
            this.tkVisibleChange.emit(false);
        };
        this.handleOverlayClick = () => {
            if (!this.preventDismiss) {
                this.tkClose.emit();
                this.tkVisibleChange.emit(false);
            }
        };
    }
    visibleChanged(newValue) {
        if (newValue) {
            this.tkOpen.emit();
            if (!this.hideBackdrop) {
                this.blockBodyScroll();
            }
        }
        else {
            this.tkClose.emit();
            if (!this.hideBackdrop) {
                this.unblockBodyScroll();
            }
        }
    }
    componentWillLoad() {
        this.hasContainerSlot = !!this.el.querySelector(':scope > [slot="container"]');
        this.hasHeaderSlot = !!this.el.querySelector(':scope > [slot="header"]');
        this.hasContentSlot = !!this.el.querySelector(':scope > [slot="content"]');
        this.hasFooterSlot = !!this.el.querySelector(':scope > [slot="footer"]');
        this.hasFooterActionsSlot = !!this.el.querySelector(':scope > [slot="footer-actions"]');
        this.hasDefaultSlotContent = Array.from(this.el.childNodes).some(node => {
            return node.nodeType === Node.ELEMENT_NODE && !node.hasAttribute('slot');
        });
    }
    disconnectedCallback() {
        if (!this.hideBackdrop) {
            this.unblockBodyScroll();
        }
    }
    /**
     * Requests to open the dialog by emitting a tk-open event.
     * Note: This method only emits an event. The dialog will only open if the parent component
     * listens for this event and updates the 'visible' prop to true.
     */
    async open() {
        this.tkOpen.emit();
    }
    /**
     * Requests to close the dialog by emitting a tk-close event.
     * Note: This method only emits an event. The dialog will only close if the parent component
     * listens for this event and updates the 'visible' prop to false.
     */
    async close() {
        this.tkClose.emit();
    }
    blockBodyScroll() {
        if (document.body.style.overflow !== 'hidden') {
            this.bodyOverflow = document.body.style.overflow;
            this.bodyPaddingRight = document.body.style.paddingRight;
            const scrollbarWidth = window.innerWidth - document.documentElement.clientWidth;
            document.body.style.paddingRight = `${scrollbarWidth}px`;
            document.body.style.overflow = 'hidden';
        }
    }
    unblockBodyScroll() {
        document.body.style.overflow = this.bodyOverflow;
        document.body.style.paddingRight = this.bodyPaddingRight;
    }
    getVariantIcon() {
        switch (this.variant) {
            case 'success':
                return 'check_circle';
            case 'warning':
                return 'warning';
            case 'danger':
                return 'error';
            default:
                return 'info';
        }
    }
    createHeader() {
        if (this.showHeader) {
            if (this.hasHeaderSlot) {
                return h("slot", { name: "header" });
            }
            else {
                const headerClasses = classNames('tk-dialog-header', `tk-dialog-header-${this.headerType}`);
                return (h("div", { class: headerClasses, "data-testid": getDataTestId(this.dataTestid, 'header-container') }, h("div", { class: "tk-dialog-header-content", "data-testid": getDataTestId(this.dataTestid, 'header') }, this.showVariantSign && (h("tk-icon", Object.assign({ sign: true, dataTestid: getDataTestId(this.dataTestid, 'sign-icon'), size: "xlarge" }, getIconElementProps(this.getVariantIcon(), { class: classNames('fill tk-dialog-sign-icon'), variant: this.variant }, 'rounded', 'span')))), h("div", { class: "tk-dialog-title-container", "data-testid": getDataTestId(this.dataTestid, 'title-container') }, this.subheader && (h("span", { class: "tk-dialog-subtitle", "data-testid": getDataTestId(this.dataTestid, 'subheader') }, this.subheader)), this.header && (h("span", { class: "tk-dialog-title", "data-testid": getDataTestId(this.dataTestid, 'title') }, this.header)))), this.showCloseButton && (h("tk-button", { variant: "neutral", icon: "close", size: "small", type: "text", "onTk-click": this.handleCloseButtonClick, "aria-label": "Close dialog", dataTestid: getDataTestId(this.dataTestid, 'close-button') }))));
            }
        }
        return null;
    }
    createContent() {
        if (this.hasContentSlot) {
            return (h("div", { class: "tk-dialog-content", "data-testid": getDataTestId(this.dataTestid, 'content') }, h("slot", { name: "content" })));
        }
        else if (this.hasDefaultSlotContent) {
            return (h("div", { class: "tk-dialog-content", "data-testid": getDataTestId(this.dataTestid, 'content') }, h("slot", null)));
        }
        return null;
    }
    createFooter() {
        if (this.hasFooterSlot) {
            return h("slot", { name: "footer" });
        }
        else if (this.hasFooterActionsSlot) {
            return (h("div", { class: "tk-dialog-footer", "data-testid": getDataTestId(this.dataTestid, 'footer') }, h("slot", { name: "footer-actions" })));
        }
        return null;
    }
    createDialog() {
        const header = this.createHeader();
        const content = this.createContent();
        const footer = this.createFooter();
        const dialogClasses = classNames('tk-dialog', `tk-dialog-${this.variant}`);
        const rootProps = {
            'class': dialogClasses,
            'style': Object.assign({ display: 'flex', flexDirection: 'column' }, this.containerStyle),
            'role': 'dialog',
            'aria-modal': true,
            'data-testid': getDataTestId(this.dataTestid, 'root'),
        };
        return (h("div", Object.assign({}, rootProps), header, content, footer));
    }
    renderMask() {
        const dialog = this.hasContainerSlot ? h("slot", { name: "container" }) : this.createDialog();
        const maskClasses = classNames('tk-dialog-mask', `tk-dialog-mask-${this.maskVariant}`, {
            'tk-dialog-visible': this.visible,
            'tk-dialog-mask-hidden': this.hideBackdrop,
            'tk-dialog-mask-blur': this.isMaskBlur,
        });
        return (h("div", { class: maskClasses, "data-testid": getDataTestId(this.dataTestid, 'mask') }, h("div", { class: "tk-dialog-overlay", onClick: () => this.handleOverlayClick(), "data-testid": getDataTestId(this.dataTestid, 'overlay') }), dialog));
    }
    render() {
        return this.renderMask();
    }
    static get is() { return "tk-dialog"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-dialog.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-dialog.css"]
        };
    }
    static get properties() {
        return {
            "visible": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls the visibility of the dialog"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "visible",
                "defaultValue": "false"
            },
            "header": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The header text"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header"
            },
            "headerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light' | 'dark' | 'primary'",
                    "resolved": "\"basic\" | \"dark\" | \"divided\" | \"light\" | \"primary\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Header type"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header-type",
                "defaultValue": "'basic'"
            },
            "showCloseButton": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls whether the close button is shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-close-button",
                "defaultValue": "true"
            },
            "showHeader": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls whether the header is shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-header",
                "defaultValue": "true"
            },
            "showVariantSign": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls whether the variant sign is shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-variant-sign",
                "defaultValue": "true"
            },
            "subheader": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The subheader text"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "subheader"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'success' | 'info' | 'warning' | 'danger'",
                    "resolved": "\"danger\" | \"info\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The variant of the dialog"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'info'"
            },
            "hideBackdrop": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the backdrop is shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-backdrop",
                "defaultValue": "false"
            },
            "maskVariant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'lightest' | 'light' | 'base' | 'dark' | 'darkest'",
                    "resolved": "\"base\" | \"dark\" | \"darkest\" | \"light\" | \"lightest\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "base"
                        }],
                    "text": "Appearance of the mask"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mask-variant",
                "defaultValue": "'base'"
            },
            "isMaskBlur": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the dialog has a blur background"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "is-mask-blur",
                "defaultValue": "false"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of container element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "preventDismiss": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Prevents the dialog from being dismissed by clicking outside"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "prevent-dismiss",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hasContainerSlot": {},
            "hasContentSlot": {},
            "hasDefaultSlotContent": {},
            "hasFooterActionsSlot": {},
            "hasFooterSlot": {},
            "hasHeaderSlot": {}
        };
    }
    static get events() {
        return [{
                "method": "tkOpen",
                "name": "tk-open",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the dialog is opened"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkClose",
                "name": "tk-close",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the dialog is closed"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkVisibleChange",
                "name": "tk-visible-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the dialog visibility changes"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "open": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Requests to open the dialog by emitting a tk-open event.\nNote: This method only emits an event. The dialog will only open if the parent component\nlistens for this event and updates the 'visible' prop to true.",
                    "tags": []
                }
            },
            "close": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Requests to close the dialog by emitting a tk-close event.\nNote: This method only emits an event. The dialog will only close if the parent component\nlistens for this event and updates the 'visible' prop to false.",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "visible",
                "methodName": "visibleChanged"
            }];
    }
}
