// @takeoff-ui/core@0.11.9/dist/collection/components/tk-pagination/tk-pagination.js
import { Fragment, h } from "@stencil/core";
import classNames from "classnames";
import { getIconElementProps } from "../../utils/icon-utils";
import { formatTemplate, getTemplateValues } from "./helpers";
/**
 * TkPagination component description.
 * @react `import { TkPagination } from '@takeoff-ui/react'`
 * @vue `import { TkPagination } from '@takeoff-ui/vue'`
 * @angular `import { TkPagination } from '@takeoff-ui/angular'`
 */
export class TkPagination {
    constructor() {
        this.ellipsis = '...';
        this.internalCurrentPage = 1;
        /**
         * Whether the pagination elements should have rounded corners
         * @defaultValue false
         */
        this.rounded = false;
        /**
         * Number of items per page.
         * @defaultValue 10
         */
        this.rowsPerPage = 10;
        /**
         * Number of total items.
         * @defaultValue 0
         */
        this.totalItems = 0;
        /**
         * Number of items per page options
         * @defaultValue [5, 10, 20, 50]
         */
        this.rowsPerPageOptions = [5, 10, 20, 50];
        /**
         * The type of the pagination
         * @defaultValue 'outlined'
         */
        this.type = 'outlined';
        /**
         * Template string for current page report in pagination.
         * Available placeholders: {currentPage}, {totalPages}
         * @defaultValue 'page: {currentPage} of {totalPages}'
         */
        this.pageReportTemplate = 'page: {currentPage} of {totalPages}';
        /**
         * Template string for items report in pagination.
         * Available placeholders: {startItem}, {endItem}, {totalItems}
         * @defaultValue 'item: {startItem}-{endItem} of {totalItems}'
         */
        this.itemsReportTemplate = 'item: {startItem}-{endItem} of {totalItems}';
        /**
         * The current page of the pagination.
         * @defaultValue 1
         */
        this.currentPage = 1;
        this.handlePageClick = (page) => {
            if (page >= 1 && page <= this.getTotalPages()) {
                this.internalCurrentPage = page;
            }
        };
        this.handlePrevClick = () => {
            if (this.internalCurrentPage > 1) {
                this.internalCurrentPage--;
                this.tkPrevPage.emit({ page: this.internalCurrentPage });
            }
        };
        this.handleNextClick = () => {
            if (this.internalCurrentPage < this.getTotalPages()) {
                this.internalCurrentPage++;
                this.tkNextPage.emit({ page: this.internalCurrentPage });
            }
        };
    }
    internalCurrentPageChanged(newValue) {
        this.currentPage = newValue;
        this.updateInputValue(newValue);
        this.emitPageChangeEvents();
    }
    currentPageChanged(newValue) {
        this.internalCurrentPage = newValue;
        this.updateInputValue(newValue);
    }
    componentWillLoad() {
        this.internalCurrentPage = this.currentPage;
        this.updateInputValue(this.currentPage);
    }
    updateInputValue(value) {
        this.inputValue = value.toString();
    }
    getTotalPages() {
        return Math.ceil(this.totalItems / this.rowsPerPage);
    }
    getPageNumbers() {
        const totalPages = this.getTotalPages();
        const currentPage = this.internalCurrentPage;
        const pageNumbers = [];
        if (totalPages <= 5) {
            for (let i = 1; i <= totalPages; i++) {
                pageNumbers.push(i);
            }
        }
        else if (totalPages <= 7) {
            this.generatePageNumbersForMediumRange(pageNumbers, currentPage, totalPages);
        }
        else {
            this.generatePageNumbersForLargeRange(pageNumbers, currentPage, totalPages);
        }
        return pageNumbers;
    }
    generatePageNumbersForMediumRange(pageNumbers, currentPage, totalPages) {
        pageNumbers.push(1);
        if (currentPage <= 3) {
            for (let i = 2; i <= 5; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis);
        }
        else if (currentPage >= totalPages - 2) {
            pageNumbers.push(this.ellipsis);
            for (let i = totalPages - 4; i < totalPages; i++) {
                pageNumbers.push(i);
            }
        }
        else {
            pageNumbers.push(this.ellipsis);
            for (let i = currentPage - 1; i <= currentPage + 1; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis);
        }
        pageNumbers.push(totalPages);
    }
    generatePageNumbersForLargeRange(pageNumbers, currentPage, totalPages) {
        pageNumbers.push(1);
        if (currentPage <= 4) {
            for (let i = 2; i <= 5; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis, totalPages);
        }
        else if (currentPage >= totalPages - 3) {
            pageNumbers.push(this.ellipsis);
            for (let i = totalPages - 4; i <= totalPages; i++) {
                pageNumbers.push(i);
            }
        }
        else {
            pageNumbers.push(this.ellipsis);
            for (let i = currentPage - 2; i <= currentPage + 2; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis, totalPages);
        }
    }
    validateAndUpdatePage() {
        const pageNumber = parseInt(this.inputValue, 10);
        const totalPages = this.getTotalPages();
        if (!isNaN(pageNumber) && pageNumber >= 1 && pageNumber <= totalPages) {
            this.handlePageClick(pageNumber);
        }
        else {
            this.updateInputValue(this.internalCurrentPage);
        }
    }
    emitPageChangeEvents() {
        const totalPages = this.getTotalPages();
        const startItem = (this.internalCurrentPage - 1) * this.rowsPerPage + 1;
        const endItem = Math.min(this.internalCurrentPage * this.rowsPerPage, this.totalItems);
        this.tkPageChange.emit({
            page: this.internalCurrentPage,
            totalPages,
            startItem,
            endItem,
        });
    }
    handlePageInputChange(value) {
        this.inputValue = value.replace(/[^0-9]/g, '');
    }
    createPageNumbers() {
        return this.getPageNumbers().map(pageNumber => {
            if (pageNumber === this.ellipsis) {
                return h("span", { class: "tk-pagination-ellipsis tk-pagination-cell" }, "...");
            }
            return (h("button", { class: classNames('tk-pagination-cell tk-pagination-page', { 'tk-pagination-page-active': pageNumber === this.internalCurrentPage }), type: "button", onClick: () => this.handlePageClick(pageNumber) }, h("span", { class: "tk-pagination-page-label" }, pageNumber)));
        });
    }
    renderTag(totalPages) {
        if (this.mode !== 'compact') {
            let tagContent;
            const templateValues = getTemplateValues(this.internalCurrentPage, this.rowsPerPage, this.totalItems, totalPages);
            if (this.mode === 'compact-expanded') {
                const itemsText = formatTemplate(this.itemsReportTemplate, templateValues);
                tagContent = h("span", { class: "tk-pagination-tag-label" }, itemsText);
            }
            else {
                const pageText = formatTemplate(this.pageReportTemplate, templateValues);
                const itemsText = formatTemplate(this.itemsReportTemplate, templateValues);
                tagContent = totalPages > 0 && (h(Fragment, null, h("span", { class: "tk-pagination-tag-label" }, pageText), h("span", { class: "tk-pagination-tag-label" }, itemsText)));
            }
            return h("div", { class: "tk-pagination-tag" }, tagContent);
        }
        return null;
    }
    renderContent(totalPages) {
        const ContentWrapperTag = this.mode === 'compact' || this.mode === 'compact-expanded' ? 'div' : 'nav';
        const contentClasses = classNames('tk-pagination', `tk-pagination-${this.mode || this.type}`, {
            'tk-pagination-rounded': this.rounded,
        });
        let content;
        if (this.mode === 'compact') {
            const templateValues = getTemplateValues(this.internalCurrentPage, this.rowsPerPage, this.totalItems, totalPages);
            const compactLabel = formatTemplate(this.pageReportTemplate, templateValues);
            content = (h(Fragment, null, h("button", { class: "tk-pagination-cell tk-pagination-prev", type: "button", onClick: this.handlePrevClick, disabled: this.internalCurrentPage === 1 }, h("tk-icon", Object.assign({}, getIconElementProps('chevron_left', { color: 'var(--icon-base)' })))), h("tk-input", { mode: "text", value: this.inputValue, "onTk-change": (event) => this.handlePageInputChange(event.detail.toString()), "onTk-blur": () => this.validateAndUpdatePage(), min: 1, max: totalPages }), h("span", { class: "tk-pagination-current-label" }, compactLabel), h("button", { class: "tk-pagination-cell tk-pagination-next", type: "button", onClick: this.handleNextClick, disabled: this.internalCurrentPage === totalPages }, h("tk-icon", Object.assign({}, getIconElementProps('chevron_right', { color: 'var(--icon-base)' }))))));
        }
        // else if(this.type === 'compact-expanded') {
        //   @TODO: Add pagination content for compact-expanded after the tk-dropdown component is completed
        //   content =
        // }
        else {
            content = (h(Fragment, null, h("div", { class: "tk-pagination-prev-actions" }, h("button", { class: "tk-pagination-cell tk-pagination-first", type: "button", onClick: () => this.handlePageClick(1), disabled: this.internalCurrentPage === 1 }, h("tk-icon", Object.assign({}, getIconElementProps('keyboard_double_arrow_left', { color: 'var(--icon-base)' })))), this.type === 'grouped' && h("span", { class: "tk-pagination-divider" }), h("button", { class: "tk-pagination-cell tk-pagination-prev", type: "button", onClick: this.handlePrevClick, disabled: this.internalCurrentPage === 1 }, h("tk-icon", Object.assign({}, getIconElementProps('chevron_left', { color: 'var(--icon-base)' }))), ' ')), this.createPageNumbers(), h("div", { class: "tk-pagination-next-actions" }, h("button", { class: "tk-pagination-cell tk-pagination-next", type: "button", onClick: this.handleNextClick, disabled: this.internalCurrentPage === totalPages }, h("tk-icon", Object.assign({}, getIconElementProps('chevron_right', { color: 'var(--icon-base)' })))), this.type === 'grouped' && h("span", { class: "tk-pagination-divider" }), h("button", { class: "tk-pagination-cell tk-pagination-last", type: "button", onClick: () => this.handlePageClick(totalPages), disabled: this.internalCurrentPage === totalPages }, h("tk-icon", Object.assign({}, getIconElementProps('keyboard_double_arrow_right', { color: 'var(--icon-base)' })))))));
        }
        return h(ContentWrapperTag, { class: contentClasses }, content);
    }
    renderSelect() {
        return (h("tk-select", { style: { width: '75px' }, value: this.rowsPerPage, options: this.rowsPerPageOptions, "onTk-change": e => {
                if (e.detail !== this.rowsPerPage) {
                    this.rowsPerPage = e.detail;
                    this.tkRowsPerPageChange.emit(e.detail);
                }
            } }));
    }
    renderInput(totalPages) {
        return (h("tk-input", { style: { width: '70px' }, mode: "text", min: 1, max: totalPages, value: this.inputValue, icon: "chevron_right", iconPosition: "right", "onTk-change": (event) => {
                this.handlePageInputChange(event.detail.toString());
            }, "onTk-blur": () => this.validateAndUpdatePage() }));
    }
    render() {
        const totalPages = this.getTotalPages();
        return (h("div", { key: 'ffc6159bb2ead21b8e793abe3e72abc5c33a62bb', class: "tk-pagination-container" }, h("div", { key: '64c521213434bc4a3164f0a04cddc7237a8b3e98', class: "tk-pagination-start" }, this.renderTag(totalPages)), this.renderContent(totalPages), h("div", { key: 'c26ec2ec063ccbd2c3f84c6c6eb92ea27ee801c0', class: "tk-pagination-end" }, this.renderSelect(), this.renderInput(totalPages))));
    }
    static get is() { return "tk-pagination"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-pagination.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-pagination.css"]
        };
    }
    static get properties() {
        return {
            "rounded": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the pagination elements should have rounded corners"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rounded",
                "defaultValue": "false"
            },
            "rowsPerPage": {
                "type": "number",
                "mutable": true,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "10"
                        }],
                    "text": "Number of items per page."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rows-per-page",
                "defaultValue": "10"
            },
            "totalItems": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "0"
                        }],
                    "text": "Number of total items."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "total-items",
                "defaultValue": "0"
            },
            "rowsPerPageOptions": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "number[]",
                    "resolved": "number[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "[5, 10, 20, 50]"
                        }],
                    "text": "Number of items per page options"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[5, 10, 20, 50]"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'outlined' | 'text' | 'grouped'",
                    "resolved": "\"grouped\" | \"outlined\" | \"text\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'outlined'"
                        }],
                    "text": "The type of the pagination"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'outlined'"
            },
            "mode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'compact' | 'compact-expanded'",
                    "resolved": "\"compact\" | \"compact-expanded\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'outlined'"
                        }],
                    "text": "The mode of the pagination"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mode"
            },
            "pageReportTemplate": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'page: {currentPage} of {totalPages}'"
                        }],
                    "text": "Template string for current page report in pagination.\nAvailable placeholders: {currentPage}, {totalPages}"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "page-report-template",
                "defaultValue": "'page: {currentPage} of {totalPages}'"
            },
            "itemsReportTemplate": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'item: {startItem}-{endItem} of {totalItems}'"
                        }],
                    "text": "Template string for items report in pagination.\nAvailable placeholders: {startItem}, {endItem}, {totalItems}"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "items-report-template",
                "defaultValue": "'item: {startItem}-{endItem} of {totalItems}'"
            },
            "currentPage": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "1"
                        }],
                    "text": "The current page of the pagination."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "current-page",
                "defaultValue": "1"
            }
        };
    }
    static get states() {
        return {
            "inputValue": {},
            "internalCurrentPage": {}
        };
    }
    static get events() {
        return [{
                "method": "tkNextPage",
                "name": "tk-next-page",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Pagination next button click event"
                },
                "complexType": {
                    "original": "{ page: number }",
                    "resolved": "{ page: number; }",
                    "references": {}
                }
            }, {
                "method": "tkPageChange",
                "name": "tk-page-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Pagination page change event"
                },
                "complexType": {
                    "original": "{ page: number; totalPages: number; startItem: number; endItem: number }",
                    "resolved": "{ page: number; totalPages: number; startItem: number; endItem: number; }",
                    "references": {}
                }
            }, {
                "method": "tkPrevPage",
                "name": "tk-prev-page",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Pagination prev button click event"
                },
                "complexType": {
                    "original": "{ page: number }",
                    "resolved": "{ page: number; }",
                    "references": {}
                }
            }, {
                "method": "tkRowsPerPageChange",
                "name": "tk-rows-per-page-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "RowsPerPage change event"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get watchers() {
        return [{
                "propName": "internalCurrentPage",
                "methodName": "internalCurrentPageChanged"
            }, {
                "propName": "currentPage",
                "methodName": "currentPageChanged"
            }];
    }
}
