// @takeoff-ui/core@0.11.9/dist/collection/components/tk-editor/tk-editor.js
import { h } from "@stencil/core";
import { Editor } from "@tiptap/core";
import Placeholder from "@tiptap/extension-placeholder";
import CharacterCount from "@tiptap/extension-character-count";
import StarterKit from "@tiptap/starter-kit";
import TextAlign from "@tiptap/extension-text-align";
import Underline from "@tiptap/extension-underline";
import Link from "@tiptap/extension-link";
import Image from "@tiptap/extension-image";
import { TOOLBAR_ICONS } from "./constants";
import { STARTER_KIT_EXTENSION_NAMES, DEFAULT_TOOLBAR_CONFIG } from "./defaults";
import classNames from "classnames";
import { renderHint } from "../../utils/hint-utils";
/**
 * TkEditor is a WYSIWYG editor component that wraps Tiptap editor.
 */
export class TkEditor {
    constructor() {
        this.isExternalUpdate = false;
        this.isFocused = false;
        this.selectionState = {};
        this.isEmpty = false;
        this.charCount = 0;
        /**
         * The value of the editor
         * @default ''
         */
        this.value = '';
        /**
         * Whether the editor is disabled
         * @default false
         */
        this.disabled = false;
        /**
         * Whether the editor is readonly
         * @default false
         */
        this.readonly = false;
        /**
         * Whether to hide the toolbar
         * @default false
         */
        this.hideToolbar = false;
        /**
         * Limits the number of characters.
         * @default 140
         */
        this.maxLength = 140;
        /**
         * Whether to show the character counter
         * @default false
         */
        this.showCounter = false;
        /**
         * Whether the editor is resizable
         * @default false
         */
        this.resizable = false;
        /**
         * The style attribute of tabs item element
         */
        this.contentStyle = null;
        this.editorHeight = 0;
        this.isResizing = false;
        this.resizeStartY = 0;
        this.resizeStartHeight = 0;
        /**
         * Custom extensions
         */
        this.extensions = [];
        /**
         * Custom toolbar buttons for extensions
         */
        this.customToolbarButtons = [];
        /**
         * Indicates whether the editor is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        this.handleToolbarAction = (button) => {
            if (!this.editor)
                return;
            if (button.command) {
                button.command(this.editor);
                return;
            }
            const commands = {
                undo: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.undo(); },
                redo: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.redo(); },
                heading: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.toggleHeading({ level: button.level }); },
                bold: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.toggleBold(); },
                italic: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.toggleItalic(); },
                underline: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.toggleUnderline(); },
                strike: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.toggleStrike(); },
                clearFormat: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.unsetAllMarks().clearNodes().setParagraph(); },
                bulletList: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.toggleBulletList(); },
                orderedList: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.toggleOrderedList(); },
                alignLeft: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.setTextAlign('left'); },
                alignCenter: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.setTextAlign('center'); },
                alignRight: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.setTextAlign('right'); },
                alignJustify: () => { var _a; return (_a = this.command()) === null || _a === void 0 ? void 0 : _a.setTextAlign('justify'); },
                link: () => {
                    var _a;
                    const url = prompt('Enter URL:');
                    return url ? (_a = this.command()) === null || _a === void 0 ? void 0 : _a.setLink({ href: url }) : undefined;
                },
                image: () => {
                    var _a;
                    const imageUrl = prompt('Enter image URL:');
                    return imageUrl ? (_a = this.command()) === null || _a === void 0 ? void 0 : _a.setImage({ src: imageUrl }) : undefined;
                },
            };
            const command = commands[button.action];
            const chain = command === null || command === void 0 ? void 0 : command();
            if (chain) {
                chain.run();
            }
        };
        this.handleResizeMouseDown = (e) => {
            if (!this.resizable || this.disabled || this.readonly)
                return;
            e.preventDefault(); // Prevent text selection
            this.isResizing = true;
            this.resizeStartY = e.clientY;
            const content = this.el.querySelector('.tk-editor-content');
            this.resizeStartHeight = content.offsetHeight;
            document.addEventListener('mousemove', this.handleResizeMouseMove);
            document.addEventListener('mouseup', this.handleResizeMouseUp);
        };
        this.handleResizeMouseMove = (e) => {
            if (!this.isResizing)
                return;
            const deltaY = e.clientY - this.resizeStartY;
            const newHeight = Math.max(120, this.resizeStartHeight + deltaY); // Min height 120px
            this.editorHeight = newHeight;
        };
        this.handleResizeMouseUp = () => {
            this.isResizing = false;
            document.removeEventListener('mousemove', this.handleResizeMouseMove);
            document.removeEventListener('mouseup', this.handleResizeMouseUp);
        };
    }
    valueChanged(newValue) {
        if (this.editor && newValue !== this.editor.getHTML()) {
            this.isExternalUpdate = true;
            this.editor.commands.setContent(newValue);
        }
    }
    editableChanged() {
        if (this.editor) {
            this.editor.setEditable(!this.disabled && !this.readonly);
        }
    }
    componentDidLoad() {
        this.initEditor();
    }
    disconnectedCallback() {
        if (this.editor) {
            this.editor.destroy();
        }
    }
    /**
     * Gets the current content of the editor
     */
    async getContent(format = 'html') {
        if (this.editor) {
            switch (format) {
                case 'json':
                    return this.editor.getJSON();
                case 'html':
                    return this.editor.getHTML();
                case 'text':
                    return this.editor.getText();
                default:
                    return this.editor.getHTML();
            }
        }
        return '';
    }
    /**
     * Sets the content of the editor
     */
    async setContent(content) {
        if (this.editor) {
            this.value = content;
            this.editor.commands.setContent(content);
        }
    }
    /**
     * Returns the Tiptap Editor instance
     */
    async getEditor() {
        return this.editor;
    }
    get activeExtensions() {
        const userExtensions = this.extensions || [];
        const userExtensionNames = new Set(userExtensions.map(ext => ext.name));
        const starterKitExclusions = STARTER_KIT_EXTENSION_NAMES.filter(name => userExtensionNames.has(name)).reduce((acc, name) => (Object.assign(Object.assign({}, acc), { [name]: false })), {});
        const defaultExtensions = [Placeholder.configure({ placeholder: this.placeholder || '' }), StarterKit.configure(starterKitExclusions)];
        if (!userExtensionNames.has('underline')) {
            defaultExtensions.push(Underline.configure({}));
        }
        if (!userExtensionNames.has('textAlign')) {
            defaultExtensions.push(TextAlign.configure({
                types: ['heading', 'paragraph'],
                alignments: ['left', 'center', 'right', 'justify'],
            }));
        }
        if (!userExtensionNames.has('link')) {
            defaultExtensions.push(Link.configure({
                openOnClick: true,
                HTMLAttributes: {
                    class: 'tk-editor-link',
                },
            }));
        }
        if (!userExtensionNames.has('image')) {
            defaultExtensions.push(Image.configure({
                HTMLAttributes: {
                    class: 'tk-editor-image',
                },
            }));
        }
        if (this.showCounter) {
            defaultExtensions.push(CharacterCount.configure({ limit: this.maxLength }));
        }
        return [...defaultExtensions, ...userExtensions];
    }
    command() {
        var _a;
        return (_a = this.editor) === null || _a === void 0 ? void 0 : _a.chain().focus();
    }
    initEditor() {
        if (this.editorRef) {
            this.editor = new Editor({
                element: this.editorRef,
                extensions: this.activeExtensions,
                content: this.value,
                editable: !this.disabled && !this.readonly,
                onCreate: ({ editor }) => {
                    this.isEmpty = editor.getText().length === 0;
                    if (this.showCounter) {
                        this.charCount = editor.storage.characterCount.characters();
                    }
                },
                onUpdate: ({ editor }) => {
                    if (this.isExternalUpdate) {
                        this.isExternalUpdate = false;
                        this.isEmpty = editor.getText().length === 0;
                        if (this.showCounter) {
                            this.charCount = editor.storage.characterCount.characters();
                        }
                        return;
                    }
                    const html = editor.getHTML();
                    this.value = html;
                    this.tkChange.emit(html);
                    this.updateSelectionState();
                    this.isEmpty = editor.getText().length === 0;
                    if (this.showCounter) {
                        this.charCount = editor.storage.characterCount.characters();
                    }
                },
                onFocus: () => {
                    this.isFocused = true;
                    this.tkFocus.emit();
                },
                onBlur: () => {
                    this.isFocused = false;
                    this.tkBlur.emit();
                },
                onSelectionUpdate: () => {
                    this.updateSelectionState();
                },
            });
        }
    }
    updateSelectionState() {
        if (!this.editor)
            return;
        this.selectionState = {
            heading: this.editor.isActive('heading') ? { level: this.editor.getAttributes('heading').level } : undefined,
            bold: this.editor.isActive('bold'),
            italic: this.editor.isActive('italic'),
            underline: this.editor.isActive('underline'),
            strike: this.editor.isActive('strike'),
            bulletList: this.editor.isActive('bulletList'),
            orderedList: this.editor.isActive('orderedList'),
            textAlign: this.editor.getAttributes('textAlign').textAlign,
            link: this.editor.isActive('link'),
        };
    }
    getDefaultButtonConfig(button) {
        const configs = {
            'h1': { icon: 'H1', label: 'Heading 1', behavior: 'toggle', level: 1 },
            'h2': { icon: 'H2', label: 'Heading 2', behavior: 'toggle', level: 2 },
            'h3': { icon: 'H3', label: 'Heading 3', behavior: 'toggle', level: 3 },
            'h4': { icon: 'H4', label: 'Heading 4', behavior: 'toggle', level: 4 },
            'bold': { icon: 'BOLD', label: 'Bold', behavior: 'toggle' },
            'italic': { icon: 'ITALIC', label: 'Italic', behavior: 'toggle' },
            'underline': { icon: 'UNDERLINED', label: 'Underline', behavior: 'toggle' },
            'strike': { icon: 'STRIKETHROUGH', label: 'Strikethrough', behavior: 'toggle' },
            'clear-format': { icon: 'CLEAR_FORMAT', label: 'Clear Formatting', behavior: 'action' },
            'bullet-list': { icon: 'LIST', label: 'Bullet List', behavior: 'toggle' },
            'ordered-list': { icon: 'LIST_NUMBERED', label: 'Numbered List', behavior: 'toggle' },
            'align-left': { icon: 'ALIGN_LEFT', label: 'Align Left', behavior: 'action' },
            'align-center': { icon: 'ALIGN_CENTER', label: 'Align Center', behavior: 'action' },
            'align-right': { icon: 'ALIGN_RIGHT', label: 'Align Right', behavior: 'action' },
            'align-justify': { icon: 'ALIGN_JUSTIFY', label: 'Align Justify', behavior: 'action' },
            'link': { icon: 'LINK', label: 'Add Link', behavior: 'toggle' },
            'image': { icon: 'PHOTO', label: 'Add Image', behavior: 'action' },
            'undo': { icon: 'UNDO', label: 'Undo', behavior: 'action' },
            'redo': { icon: 'REDO', label: 'Redo', behavior: 'action' },
        };
        return Object.assign({ action: this.getButtonAction(button) }, configs[button]);
    }
    getButtonAction(button) {
        const actionMap = {
            'h1': 'heading',
            'h2': 'heading',
            'h3': 'heading',
            'h4': 'heading',
            'bold': 'bold',
            'italic': 'italic',
            'underline': 'underline',
            'strike': 'strike',
            'clear-format': 'clearFormat',
            'bullet-list': 'bulletList',
            'ordered-list': 'orderedList',
            'align-left': 'alignLeft',
            'align-center': 'alignCenter',
            'align-right': 'alignRight',
            'align-justify': 'alignJustify',
            'link': 'link',
            'image': 'image',
            'undo': 'undo',
            'redo': 'redo',
        };
        return actionMap[button];
    }
    isButtonActive(button) {
        if (!this.editor)
            return false;
        if (button.command) {
            return this.editor.isActive(button.action);
        }
        switch (button.action) {
            case 'heading':
                return this.editor.isActive('heading', { level: button.level });
            case 'bold':
                return this.editor.isActive('bold');
            case 'italic':
                return this.editor.isActive('italic');
            case 'underline':
                return this.editor.isActive('underline');
            case 'strike':
                return this.editor.isActive('strike');
            case 'bulletList':
                return this.editor.isActive('bulletList');
            case 'orderedList':
                return this.editor.isActive('orderedList');
            case 'alignLeft':
                return this.editor.isActive({ textAlign: 'left' });
            case 'alignCenter':
                return this.editor.isActive({ textAlign: 'center' });
            case 'alignRight':
                return this.editor.isActive({ textAlign: 'right' });
            case 'alignJustify':
                return this.editor.isActive({ textAlign: 'justify' });
            case 'link':
                return this.editor.isActive('link');
            case 'undo':
                return this.editor.can().undo();
            case 'redo':
                return this.editor.can().redo();
            default:
                return this.editor.isActive(button.action);
        }
    }
    isHistoryButtonDisabled(action) {
        var _a;
        if (!this.editor)
            return true;
        return !((_a = this.editor.can()) === null || _a === void 0 ? void 0 : _a[action]());
    }
    createToolbarButton(button) {
        const isCustomButton = !TOOLBAR_ICONS[button.icon];
        const iconContent = TOOLBAR_ICONS[button.icon] || button.icon;
        return (h("button", { class: classNames('tk-editor-button', `tk-editor-button-${button.action}`, `tk-editor-button-${button.behavior}`, {
                'tk-editor-button-active': button.behavior === 'toggle' && this.isButtonActive(button),
                'tk-editor-button-disabled': this.disabled,
                'tk-editor-button-readonly': this.readonly,
                [`tk-editor-button-${button.action}-disabled`]: (button.action === 'undo' || button.action === 'redo') && this.isHistoryButtonDisabled(button.action),
            }), type: "button", title: button.label, disabled: this.disabled || this.readonly, innerHTML: iconContent, onClick: () => !this.disabled && this.handleToolbarAction(button), "data-tk-editor-custom-button": isCustomButton }));
    }
    renderToolbar() {
        if (this.hideToolbar)
            return null;
        const toolbarConfig = this.toolbar || DEFAULT_TOOLBAR_CONFIG;
        return (h("div", { class: "tk-editor-toolbar" }, h("div", { class: "tk-editor-controlsgroup-container" }, toolbarConfig.map((group, groupIndex) => (h("div", { class: "tk-editor-controlsgroup", key: groupIndex }, group.map(button => {
            const buttonConfig = typeof button === 'string' ? this.getDefaultButtonConfig(button) : button;
            return this.createToolbarButton(buttonConfig);
        })))))));
    }
    renderFooter() {
        if (this.showCounter || this.resizable) {
            return (h("div", { class: "tk-editor-footer" }, this.showCounter && (h("div", { class: classNames('tk-editor-counter', { 'tk-editor-counter-maxed': this.charCount === this.maxLength }) }, this.charCount, "/", this.maxLength)), this.resizable && (h("div", { class: "tk-editor-resize-icon", onMouseDown: this.handleResizeMouseDown }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "6", height: "6", viewBox: "0 0 6 6", fill: "none" }, h("path", { d: "M2.18564e-05 5.89256H5.89258V0L2.18564e-05 5.89256Z", fill: "var(--icon-light)" }))))));
        }
        return null;
    }
    render() {
        const contentStyle = Object.assign(Object.assign({}, this.contentStyle), (this.editorHeight > 0 ? { height: `${this.editorHeight}px`, minHeight: 'unset' } : {}));
        const labelElement = this.label && (h("label", { key: '6c92769f4717846926319d0cb93704ab94e7f8aa', class: "tk-editor-label" }, this.label, this.showAsterisk && h("span", { key: 'dc65aa93dc78b5bd354a1c1e739c03afb6f359b5', class: "tk-editor-label-asterisk" }, "*")));
        return (h("div", { key: '7e54eb1ac65bd5f11153a2056865bd28c5db452b', class: classNames('tk-editor-container', {
                'tk-editor-focused': this.isFocused,
                'tk-editor-disabled': this.disabled,
                'tk-editor-readonly': this.readonly,
                'tk-editor-invalid': this.invalid,
            }) }, labelElement, h("div", { key: '1bf88c1aa8ca71fdbc8af60a02c6ca178741467a', class: "tk-editor" }, this.renderToolbar(), h("div", { key: 'f865329c0ee325f1abf85e6679628caeb42943fa', class: classNames('tk-editor-content', { 'tk-editor-placeholder-visible': this.isEmpty && !!this.placeholder }), ref: el => (this.editorRef = el), "data-placeholder": this.placeholder, style: contentStyle }), this.renderFooter()), renderHint(this.hint, this.error, this.invalid)));
    }
    static get is() { return "tk-editor"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-editor.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-editor.css"]
        };
    }
    static get properties() {
        return {
            "value": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "''"
                        }],
                    "text": "The value of the editor"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value",
                "defaultValue": "''"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }],
                    "text": "Whether the editor is disabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "readonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }],
                    "text": "Whether the editor is readonly"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "readonly",
                "defaultValue": "false"
            },
            "placeholder": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The placeholder text when editor is empty"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "placeholder"
            },
            "hideToolbar": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }],
                    "text": "Whether to hide the toolbar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-toolbar",
                "defaultValue": "false"
            },
            "maxLength": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "140"
                        }],
                    "text": "Limits the number of characters."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max-length",
                "defaultValue": "140"
            },
            "showCounter": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }],
                    "text": "Whether to show the character counter"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-counter",
                "defaultValue": "false"
            },
            "resizable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }],
                    "text": "Whether the editor is resizable"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "resizable",
                "defaultValue": "false"
            },
            "contentStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of tabs item element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "extensions": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "AnyExtension[]",
                    "resolved": "AnyExtension[]",
                    "references": {
                        "AnyExtension": {
                            "location": "import",
                            "path": "@tiptap/core",
                            "id": "../../node_modules/.pnpm/@tiptap+core@2.27.2_@tiptap+pm@2.27.2/node_modules/@tiptap/core/dist/index.d.ts::AnyExtension",
                            "referenceLocation": "AnyExtension"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom extensions"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "customToolbarButtons": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "TkEditorCustomButton[]",
                    "resolved": "TkEditorCustomButton[]",
                    "references": {
                        "TkEditorCustomButton": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-editor/types.ts::TkEditorCustomButton",
                            "referenceLocation": "TkEditorCustomButton"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom toolbar buttons for extensions"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "toolbar": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "TkEditorToolbarConfig",
                    "resolved": "(TkEditorDefaultButton | TkEditorCustomButton)[][]",
                    "references": {
                        "TkEditorToolbarConfig": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-editor/types.ts::TkEditorToolbarConfig",
                            "referenceLocation": "TkEditorToolbarConfig"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Toolbar configuration"
                },
                "getter": false,
                "setter": false
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The label for the toggle"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Provided a hint or additional information about the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the editor is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This is the error message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "editor": {},
            "isFocused": {},
            "selectionState": {},
            "isEmpty": {},
            "charCount": {},
            "editorHeight": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when editor content changes"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }, {
                "method": "tkFocus",
                "name": "tkFocus",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when editor gets focus"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkBlur",
                "name": "tkBlur",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when editor loses focus"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "getContent": {
                "complexType": {
                    "signature": "(format?: \"json\" | \"html\" | \"text\") => Promise<JSONContent | string>",
                    "parameters": [{
                            "name": "format",
                            "type": "\"text\" | \"html\" | \"json\"",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "JSONContent": {
                            "location": "import",
                            "path": "@tiptap/core",
                            "id": "../../node_modules/.pnpm/@tiptap+core@2.27.2_@tiptap+pm@2.27.2/node_modules/@tiptap/core/dist/index.d.ts::JSONContent",
                            "referenceLocation": "JSONContent"
                        }
                    },
                    "return": "Promise<string | JSONContent>"
                },
                "docs": {
                    "text": "Gets the current content of the editor",
                    "tags": []
                }
            },
            "setContent": {
                "complexType": {
                    "signature": "(content: string) => Promise<void>",
                    "parameters": [{
                            "name": "content",
                            "type": "string",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Sets the content of the editor",
                    "tags": []
                }
            },
            "getEditor": {
                "complexType": {
                    "signature": "() => Promise<Editor>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "Editor": {
                            "location": "import",
                            "path": "@tiptap/core",
                            "id": "../../node_modules/.pnpm/@tiptap+core@2.27.2_@tiptap+pm@2.27.2/node_modules/@tiptap/core/dist/index.d.ts::Editor",
                            "referenceLocation": "Editor"
                        }
                    },
                    "return": "Promise<Editor>"
                },
                "docs": {
                    "text": "Returns the Tiptap Editor instance",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "valueChanged"
            }, {
                "propName": "disabled",
                "methodName": "editableChanged"
            }, {
                "propName": "readonly",
                "methodName": "editableChanged"
            }];
    }
}
