// @takeoff-ui/core@0.11.9/dist/collection/components/tk-divider/tk-divider.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * TkDivider is a component that allows you to separate different sections, elements and content.
 * @react `import { TkDivider } from '@takeoff-ui/react'`
 * @vue `import { TkDivider } from '@takeoff-ui/vue'`
 * @angular `import { TkDivider } from '@takeoff-ui/angular'`
 */
export class TkDivider {
    constructor() {
        /**
         * Controls the orientation of the divider component.
         * @defaultValue horizontal
         */
        this.orientation = 'horizontal';
    }
    componentWillLoad() {
        this.hasDefaultSlot = Array.from(this.el.childNodes).some(node => { var _a; return (node.nodeType === Node.ELEMENT_NODE && !node.hasAttribute('slot')) || (node.nodeType === Node.TEXT_NODE && ((_a = node.textContent) === null || _a === void 0 ? void 0 : _a.trim().length) > 0); });
    }
    getRootStyles() {
        const styles = {};
        if (this.my) {
            const space = typeof this.my === 'number' ? `${this.my}px` : this.my;
            styles.marginTop = space;
            styles.marginBottom = space;
        }
        if (this.mx) {
            const space = typeof this.mx === 'number' ? `${this.mx}px` : this.mx;
            styles.marginLeft = space;
            styles.marginRight = space;
        }
        return styles;
    }
    render() {
        return (h("div", { key: '1f2c3c461d6b7a9483bd112c4c0689248009469c', class: classNames('tk-divider', `tk-divider-${this.orientation}`), role: "separator", "aria-orientation": this.orientation, style: this.getRootStyles(), "data-testid": getDataTestId(this.dataTestid, 'container') }, this.hasDefaultSlot && (h("div", { key: 'b0f6cc048d7a580c92dd311ec7021ec90c0830d7', class: "tk-divider-content", "data-testid": getDataTestId(this.dataTestid, 'content') }, h("slot", { key: 'dde0629cd11c49f689f983241215c8212bae7ded' })))));
    }
    static get is() { return "tk-divider"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-divider.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-divider.css"]
        };
    }
    static get properties() {
        return {
            "orientation": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'horizontal' | 'vertical'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "horizontal"
                        }],
                    "text": "Controls the orientation of the divider component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "orientation",
                "defaultValue": "'horizontal'"
            },
            "mx": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "string | number",
                    "resolved": "number | string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Controls vertical spacing"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mx"
            },
            "my": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "string | number",
                    "resolved": "number | string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Controls horizontal spacing"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "my"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hasDefaultSlot": {}
        };
    }
    static get elementRef() { return "el"; }
}
