// @takeoff-ui/core@0.11.9/dist/collection/components/tk-stepper/tk-step.js
import { h, Host } from "@stencil/core";
/**
 * TkStep sub-component for individual steps in TkStepper.
 */
export class TkStep {
    constructor() {
        /**
         * Indicates if the step has been completed.
         */
        this.complete = false;
        /**
         * Indicates if the step has encountered an error.
         */
        this.error = false;
        /**
         * Indicates if the step is currently active.
         */
        this.isActive = false;
        /**
         * Whether the step is clickable.
         */
        this.isClickable = true;
        /**
         * Controls the label position of the step.
         * @defaultValue 'non-flip'
         */
        this.labelPosition = 'non-flip';
        /**
         * Controls the step mode of the stepper component.
         * @defaultValue 'basic'
         */
        this.stepMode = 'basic';
        /**
         * Indicates if the step is disabled (cannot be selected or interacted with).
         */
        this.disabled = false;
    }
    render() {
        return (h(Host, { key: '45beb75797f35160514c95035cdade44b36d2099' }, h("slot", { key: '689ac301bf415c2c13ff23e27a867475b5177e7c' })));
    }
    static get is() { return "tk-step"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-step.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-step.css"]
        };
    }
    static get properties() {
        return {
            "header": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The header text to be displayed for the step."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header"
            },
            "subheader": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Optional subheader text to provide additional context for the step."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "subheader"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon to be displayed for the step. Can be either a string (icon name) or an IIconOptions object."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "completeIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon or icon options for completed steps."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "complete-icon"
            },
            "activeIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon or icon options for active steps."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "active-icon"
            },
            "inactiveIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon or icon options for inactive steps."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "inactive-icon"
            },
            "errorIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon or icon options for error steps."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error-icon"
            },
            "complete": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the step has been completed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "complete",
                "defaultValue": "false"
            },
            "error": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the step has encountered an error."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error",
                "defaultValue": "false"
            },
            "isActive": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the step is currently active."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "is-active",
                "defaultValue": "false"
            },
            "index": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The index of the step in the stepper."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "index"
            },
            "isClickable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the step is clickable."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "is-clickable",
                "defaultValue": "true"
            },
            "labelPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'flip' | 'non-flip'",
                    "resolved": "\"flip\" | \"non-flip\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'non-flip'"
                        }],
                    "text": "Controls the label position of the step."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label-position",
                "defaultValue": "'non-flip'"
            },
            "stepMode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'number'",
                    "resolved": "\"basic\" | \"number\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'basic'"
                        }],
                    "text": "Controls the step mode of the stepper component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "step-mode",
                "defaultValue": "'basic'"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the step is disabled (cannot be selected or interacted with)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "el"; }
}
