// @takeoff-ui/core@0.11.9/dist/collection/components/tk-select/tk-select.js
import { Fragment, h } from "@stencil/core";
import classNames from "classnames";
import { v4 as uuidv4 } from "uuid";
import { isEqual, some, remove } from "lodash-es";
import { getNestedValue } from "../../utils/object-utils";
import { applyStyles } from "../../utils/style-utils";
import { ClickOutsideMixin } from "../../utils/clickoutside-mixin";
import { floatingElementAutoUpdate } from "../../utils/position-utils";
/**
 * TkSelect component description.
 * @slot empty-data - Set how the dropdown will appear when there is no data
 * @react `import { TkSelect } from '@takeoff-ui/react'`
 * @vue `import { TkSelect } from '@takeoff-ui/vue'`
 * @angular `import { TkSelect } from '@takeoff-ui/angular'`
 */
export class TkSelect {
    constructor() {
        this.hasEmptyDataSlot = false;
        this.uniqueId = uuidv4();
        this.isItemClickFlag = false;
        this.flatOptions = [];
        this.isOpen = false;
        /**
         * The key to use for option group names.
         * Required if grouped options are used.
         */
        this.groupNameKey = 'label';
        /**
         * The key to use for accessing grouped options array.
         * Required if grouped options are used.
         */
        this.groupOptionsKey = 'options';
        /**
         * Represents whether the options are fethecd from service or not.
         * If true renders spinner in options dropdown.
         */
        this.loading = false;
        /**
         * Enables users to enter custom values that are not part of the predefined options.
         * @defaultValue false
         */
        this.allowCustomValue = false;
        /**
         * Indicates whether the input can be cleared
         * @defaultValue false
         */
        this.clearable = false;
        /**
         * If `true`, the user cannot interact with the input.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * Determines the width of the dropdown. Accepts values like 'match-parent', 'auto', or a specific width in '300px'.
         * @defaultValue match-parent
         */
        this.dropdownWidthMode = 'match-parent';
        /**
         * This property determines whether the input field within the select box is editable.
         * @defaultValue false
         */
        this.editable = false;
        /**
         * The message to display when there is no data available.
         */
        this.emptyMessage = 'No options available';
        /**
         *  Function used to filter current options based on the input value. Comes with a default filter function, but can be overridden with a custom function.
         */
        this.filter = this.defaultFilter;
        /**
         * Sets the delay (in ms) before triggering the filter function.
         */
        this.filterDebounceDelay = 0;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * If `true`, the user cannot modify the value.
         * @defaultValue false
         */
        this.readonly = false;
        /**
         * Sets size for the component.
         * @defaultValue base
         */
        this.size = 'base';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * The key to use for option labels
         * @defaultValue label
         */
        this.optionLabelKey = 'label';
        /**
         * If true enables selectAll option
         *  @defaultValue false
         */
        this.selectAll = false;
        /**
         * Sets the label of the selectAll option
         *  @defaultValue 'All'
         */
        this.selectAllLabel = 'All';
        /**
         * Click outside handler implementation - called by the mixin
         */
        this.closeHandler = () => {
            this.isOpen = false;
        };
        this.boundRunFilterForMultiple = this.runFilterForMultiple.bind(this);
    }
    isOpenChanged(newValue) {
        if (newValue)
            this.tkOpen.emit();
        else
            this.tkClose.emit();
    }
    disabledChanged(newValue) {
        if (newValue)
            this.isOpen = false;
    }
    optionsChanged(newValue, oldValue) {
        var _a;
        if (isEqual(newValue, oldValue))
            return;
        this.renderOptions = ((_a = this.options) === null || _a === void 0 ? void 0 : _a.length) > 0 ? [...this.options] : [];
        this.setValue();
    }
    /**
     * Update the native input element when the value changes
     */
    valueChanged(newValue, oldValue) {
        if (isEqual(newValue, oldValue))
            return;
        this.setValue();
        if (this.multiple && this.selectAll) {
            const newValues = Array.isArray(newValue) ? newValue : [];
            this.tkSelectAll.emit(this.isAllSelected(newValues));
        }
    }
    componentWillLoad() {
        var _a;
        this.hasEmptyDataSlot = !!this.el.querySelector('[slot="empty-data"]');
        this.setFlatOptions();
        this.renderOptions = ((_a = this.options) === null || _a === void 0 ? void 0 : _a.length) > 0 ? [...this.options] : [];
    }
    componentDidLoad() {
        var _a, _b;
        (_b = (_a = this.internals) === null || _a === void 0 ? void 0 : _a.form) === null || _b === void 0 ? void 0 : _b.addEventListener('reset', () => {
            this.handleFormReset();
        });
        this.nativeInputRef = this.inputRef.querySelector('input');
        const tkInputArea = this.inputRef.querySelector('.tk-input');
        this.clickOutsideMixin = new ClickOutsideMixin({
            referenceElement: tkInputArea !== null && tkInputArea !== void 0 ? tkInputArea : this.el,
            handler: this.closeHandler,
            disabled: this.disabled || this.readonly || !this.isOpen,
        });
        if (this.allowCustomValue) {
            this.editable = true;
        }
        if (this.value !== undefined && this.value !== null) {
            this.setValue();
        }
    }
    componentDidRender() {
        var _a, _b, _c;
        // multiple durumda chips li input çalıştığı için ve tk-input value olarak chips leri geri döndürdüğü için
        // tk-input'un içindeki inputa yazılan değerlerin filtering için çalışabilmesini sağlamak için yapılmıştır.
        if (this.readonly) {
            const nativeInput = (_a = this.inputRef) === null || _a === void 0 ? void 0 : _a.querySelector('input');
            if (nativeInput) {
                nativeInput.setAttribute('readonly', 'true');
            }
        }
        if (this.multiple && this.editable) {
            (_b = this.nativeInputRef) === null || _b === void 0 ? void 0 : _b.removeEventListener('input', this.boundRunFilterForMultiple);
            (_c = this.nativeInputRef) === null || _c === void 0 ? void 0 : _c.addEventListener('input', this.boundRunFilterForMultiple);
        }
    }
    componentDidUpdate() {
        var _a, _b, _c, _d;
        this.nativeInputRef = this.inputRef.querySelector('input');
        // Update click outside mixin configuration based on current state
        (_a = this.clickOutsideMixin) === null || _a === void 0 ? void 0 : _a.updateConfig({
            disabled: this.disabled || this.readonly || !this.isOpen,
            ignoredElements: this.panelRef ? [this.panelRef] : [],
        });
        if (this.isOpen) {
            if (this.inputRef && this.panelRef) {
                // Clean up old floating UI listeners before setting up new ones
                (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
                this.updatePosition();
                this.setFlatOptions();
                // Panel açıldığında ilk itemin active olmasını sağlamak için
                const activeItem = this.el.querySelector('.dropdown-item[data-active]');
                if (!activeItem && !this.allowCustomValue) {
                    const firstEnabledIndex = this.getNextEnabledItemIndex(-1, 'down');
                    if (firstEnabledIndex !== null) {
                        const firstItem = this.el.querySelector(`.dropdown-item[data-option-index="${firstEnabledIndex}"]`);
                        firstItem === null || firstItem === void 0 ? void 0 : firstItem.setAttribute('data-active', 'true');
                    }
                }
            }
        }
        else {
            // Remove floating UI listeners when select closes
            (_c = this.cleanup) === null || _c === void 0 ? void 0 : _c.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
            (_d = this.panelRef) === null || _d === void 0 ? void 0 : _d.remove();
        }
    }
    disconnectedCallback() {
        var _a, _b, _c, _d;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        (_c = (_b = this.internals) === null || _b === void 0 ? void 0 : _b.form) === null || _c === void 0 ? void 0 : _c.removeEventListener('reset', this.handleFormReset.bind(this));
        // Call mixin's disconnectedCallback for cleanup
        (_d = this.clickOutsideMixin) === null || _d === void 0 ? void 0 : _d.disconnectedCallback();
    }
    formResetCallback() {
        this.handleFormReset();
    }
    /**
     * Closes the select dropdown;
     */
    async close() {
        this.isOpen = false;
    }
    updatePosition() {
        const dropdownWidthMode = this.dropdownWidthMode;
        const tkInputRootEl = this.inputRef.querySelector('.tk-input');
        this.cleanup = floatingElementAutoUpdate(tkInputRootEl, this.panelRef, undefined, {
            placement: 'bottom-start',
            size: {
                apply({ rects, elements }) {
                    if (dropdownWidthMode === 'match-parent') {
                        applyStyles(elements.floating, {
                            width: `${rects.reference.width}px`,
                        });
                    }
                    else if (dropdownWidthMode !== 'auto' && dropdownWidthMode.length > 0) {
                        applyStyles(elements.floating, {
                            width: dropdownWidthMode,
                        });
                    }
                },
            },
        });
    }
    isGrouped() {
        var _a, _b;
        return ((_a = this.options) === null || _a === void 0 ? void 0 : _a.length) > 0 && ((_b = this.options[0]) === null || _b === void 0 ? void 0 : _b[this.groupOptionsKey]);
    }
    setFlatOptions() {
        if (this.isGrouped()) {
            this.flatOptions = this.options.flatMap(group => group[this.groupOptionsKey]);
        }
        else {
            this.flatOptions = this.options;
        }
    }
    async runFilterForMultiple() {
        this.renderOptions = await this.filter(this.nativeInputRef.value, this.options);
    }
    async defaultFilter(text, options) {
        if (!text) {
            return [...this.options];
        }
        if (this.isGrouped()) {
            return options
                .map(group => (Object.assign(Object.assign({}, group), { [this.groupOptionsKey]: group[this.groupOptionsKey].filter(option => this.getOptionLabel(option).toLowerCase().includes(text.toLowerCase())) })))
                .filter(group => group[this.groupOptionsKey].length > 0);
        }
        else {
            return options.filter(item => this.getOptionLabel(item).toLowerCase().indexOf(text.toLowerCase()) > -1);
        }
    }
    isOptionSelected(valueArr, optionValue) {
        if (typeof optionValue === 'object' && !Array.isArray(optionValue) && optionValue !== null) {
            return valueArr.some(v => isEqual(v, optionValue));
        }
        else {
            return valueArr.includes(optionValue);
        }
    }
    //edited to omit disabled options from select all check
    isAllSelected(valueArr) {
        var _a, _b;
        let arr;
        if (Array.isArray(valueArr)) {
            arr = valueArr;
        }
        else if (Array.isArray((_b = (_a = this.value) === null || _a === void 0 ? void 0 : _a.filter) === null || _b === void 0 ? void 0 : _b.call(_a, item => { var _a; return !((_a = this.optionDisabled) === null || _a === void 0 ? void 0 : _a.call(this, item)); }))) {
            arr = this.value;
        }
        else {
            arr = [];
        }
        const optionValues = this.flatOptions.filter(item => { var _a; return !((_a = this.optionDisabled) === null || _a === void 0 ? void 0 : _a.call(this, item)); }).map(opt => this.getOptionValue(opt));
        return optionValues.length > 0 && optionValues.every(val => this.isOptionSelected(arr, val));
    }
    getOptionLabel(item) {
        if (typeof item === 'object' && item !== null) {
            const label = getNestedValue(item, this.optionLabelKey);
            return label != null ? String(label) : '';
        }
        return item != null ? String(item) : '';
    }
    getOptionValue(item) {
        var _a;
        if (typeof item === 'object' && item !== null) {
            if (((_a = this.optionValueKey) === null || _a === void 0 ? void 0 : _a.length) > 0) {
                return getNestedValue(item, this.optionValueKey);
            }
            else {
                return item;
            }
        }
        else {
            return item;
        }
    }
    getDisplayValueForMultiple(selectedItems) {
        if (!this.visibleItemCount || selectedItems.length <= this.visibleItemCount) {
            return selectedItems;
        }
        // When allowCustomValue is true, prioritize predefined options over custom values
        if (this.allowCustomValue) {
            // Separate predefined options from custom values
            const predefinedItems = selectedItems.filter(item => {
                return this.flatOptions.some(opt => {
                    if (this.optionValueKey) {
                        return this.getOptionValue(opt) === this.getOptionValue(item);
                    }
                    else {
                        return isEqual(opt, item);
                    }
                });
            });
            const customItems = selectedItems.filter(item => {
                return !this.flatOptions.some(opt => {
                    if (this.optionValueKey) {
                        return this.getOptionValue(opt) === this.getOptionValue(item);
                    }
                    else {
                        return isEqual(opt, item);
                    }
                });
            });
            // Take predefined items first, then custom items if there's room
            const visibleItems = [...predefinedItems.slice(0, this.visibleItemCount), ...customItems.slice(0, Math.max(0, this.visibleItemCount - predefinedItems.length))];
            if (selectedItems.length > this.visibleItemCount) {
                const remainingCount = selectedItems.length - this.visibleItemCount;
                const othersIndicator = {
                    __isOthersIndicator: true,
                    label: `+${remainingCount}`,
                    removable: false,
                };
                return [...visibleItems, othersIndicator];
            }
            return visibleItems;
        }
        // Original logic for when allowCustomValue is false
        const visibleItems = selectedItems.slice(0, this.visibleItemCount);
        const remainingCount = selectedItems.length - this.visibleItemCount;
        const othersIndicator = {
            __isOthersIndicator: true,
            label: `+${remainingCount}`,
            removable: false,
        };
        return [...visibleItems, othersIndicator];
    }
    async setRenderOptions(value) {
        this.renderOptions = await this.filter(value, this.options);
    }
    getSelectedItem() {
        var _a, _b;
        if (((_a = this.renderOptions) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            let searchOptions = this.renderOptions;
            if (this.isGrouped()) {
                searchOptions = this.renderOptions.flatMap(group => group[this.groupOptionsKey] || []);
            }
            if (typeof this.value != 'object' && (searchOptions === null || searchOptions === void 0 ? void 0 : searchOptions.every(item => typeof item != 'object'))) {
                // value ve her bir option object değilse. Yani bu primitive tiplerle çalışan bir selectbox ise
                return searchOptions.find(item => item == this.value);
            }
            else if (searchOptions === null || searchOptions === void 0 ? void 0 : searchOptions.every(item => typeof item === 'object')) {
                if (((_b = this.optionValueKey) === null || _b === void 0 ? void 0 : _b.length) > 0) {
                    return searchOptions.find(item => this.getOptionValue(item) == this.value);
                }
                else {
                    return searchOptions.find(item => isEqual(item, this.value));
                }
            }
            else {
                return null;
            }
        }
        else {
            return null;
        }
    }
    setValue() {
        var _a, _b, _c, _d, _e;
        if (!this.inputRef)
            return;
        this.setFlatOptions();
        // Handle multiple selection case
        if (this.multiple) {
            // Ensure value is always an array
            const currentValue = Array.isArray(this.value) ? this.value : [];
            this.selectedItem = currentValue
                .map(val => {
                let found;
                if (this.optionValueKey) {
                    found = this.flatOptions.find(opt => this.getOptionValue(opt) === val);
                }
                else {
                    found = this.flatOptions.find(opt => isEqual(opt, val));
                }
                if (found !== undefined)
                    return found;
                if (this.allowCustomValue)
                    return val;
                return null;
            })
                .filter(val => val !== null && val !== undefined);
            // Apply visible item count logic for display
            const displayValue = this.getDisplayValueForMultiple(this.selectedItem);
            this.inputRef.value = displayValue;
            return;
        }
        // Handle single selection case
        if (this.editable && this.allowCustomValue) {
            // For editable with custom values, show the value directly
            if (this.value !== undefined && this.value !== null) {
                this.inputRef.value = this.getOptionLabel(this.value);
            }
            else {
                this.inputRef.value = null;
            }
            return;
        }
        // Find the selected item based on value type
        if (typeof this.value !== 'object' && ((_a = this.flatOptions) === null || _a === void 0 ? void 0 : _a.every(item => typeof item !== 'object'))) {
            // Handle primitive values
            this.selectedItem = (_b = this.flatOptions) === null || _b === void 0 ? void 0 : _b.find(item => item === this.value);
        }
        else if (((_c = this.optionValueKey) === null || _c === void 0 ? void 0 : _c.length) > 0) {
            // Handle object values with optionValueKey
            this.selectedItem = (_d = this.flatOptions) === null || _d === void 0 ? void 0 : _d.find(item => this.getOptionValue(item) === this.value);
        }
        else {
            // Handle object values without optionValueKey
            this.selectedItem = (_e = this.flatOptions) === null || _e === void 0 ? void 0 : _e.find(item => isEqual(item, this.value));
        }
        // Set input value based on selection state
        if (this.selectedItem !== null && this.selectedItem !== undefined) {
            if (this.multiple) {
                this.inputRef.value = this.selectedItem;
            }
            else {
                const label = this.getOptionLabel(this.selectedItem);
                this.inputRef.value = label;
            }
        }
        else {
            if (this.filter) {
                this.selectedItem = this.getSelectedItem();
                if (this.selectedItem) {
                    if (this.multiple) {
                        this.inputRef.value = this.selectedItem;
                    }
                    else {
                        this.inputRef.value = this.getOptionLabel(this.selectedItem);
                    }
                }
                else {
                    this.inputRef.value = null;
                }
            }
            else {
                this.inputRef.value = null;
            }
        }
    }
    scrollItem(item) {
        item.scrollIntoView({
            behavior: 'smooth',
            block: 'nearest',
        });
    }
    getNextEnabledItemIndex(currentIndex, direction, visited = 0) {
        const totalItems = this.flatOptions.length;
        const maxSteps = this.selectAll && this.multiple ? totalItems + 1 : totalItems;
        if (visited >= maxSteps)
            return null;
        const startIndex = this.selectAll && this.multiple ? -1 : 0;
        const endIndex = totalItems - 1;
        let nextIndex = direction === 'down' ? currentIndex + 1 : currentIndex - 1;
        if (nextIndex > endIndex) {
            nextIndex = startIndex;
        }
        else if (nextIndex < startIndex) {
            nextIndex = endIndex;
        }
        if (nextIndex === -1 && this.selectAll && this.multiple) {
            return -1;
        }
        const item = this.flatOptions[nextIndex];
        const isDisabled = this.optionDisabled ? this.optionDisabled(item) : false;
        if (isDisabled)
            return this.getNextEnabledItemIndex(nextIndex, direction, visited + 1);
        else
            return nextIndex;
    }
    navigateToItem(direction) {
        const activeItem = this.el.querySelector('.dropdown-item[data-active="true"]');
        const activeIndex = activeItem ? Number(activeItem.getAttribute('data-option-index')) : null;
        // activeindex varsa o indexten başlayarak yönüne göre sonraki enabled itemi bul, yoksa aşağı yönünde ilk enabled itemi bul
        const nextIndex = activeIndex !== null ? this.getNextEnabledItemIndex(activeIndex, direction) : this.getNextEnabledItemIndex(-1, 'down');
        if (nextIndex !== null) {
            const newActiveItem = this.el.querySelector(`.dropdown-item[data-option-index='${nextIndex}']`);
            if (newActiveItem) {
                activeItem === null || activeItem === void 0 ? void 0 : activeItem.setAttribute('data-active', 'false');
                newActiveItem.setAttribute('data-active', 'true');
                this.scrollItem(newActiveItem);
            }
        }
    }
    handleFormReset() {
        if (this.multiple && this.optionDisabled && Array.isArray(this.value)) {
            this.value = this.value.filter(item => this.optionDisabled(item));
        }
        else {
            this.value = null;
        }
        this.tkChange.emit(this.value);
    }
    async handleSelectAllClick() {
        var _a;
        if (this.readonly)
            return;
        if (this.multiple) {
            let tmpValue;
            const checking = this.isAllSelected();
            if (checking) {
                // Deselect all
                tmpValue = this.value.filter(item => { var _a; return (_a = this.optionDisabled) === null || _a === void 0 ? void 0 : _a.call(this, item); });
                this.tkSelectAll.emit(false);
            }
            else {
                //optionsdaki değerleri almak için
                const optionValues = this.flatOptions.filter(item => { var _a; return !((_a = this.optionDisabled) === null || _a === void 0 ? void 0 : _a.call(this, item)); }).map(opt => this.getOptionValue(opt));
                // allowcustom trueyken optionsda olmayan valueların eklenmesi için
                const customValues = Array.isArray(this.value) ? (_a = this.value) === null || _a === void 0 ? void 0 : _a.filter(val => !this.isOptionSelected(optionValues, val)) : [];
                // Select all (optionValue + custom values)
                tmpValue = [...optionValues, ...customValues];
                this.tkSelectAll.emit(true);
            }
            // filtreleme ardında yapılan seçimden sonra filtrelem için kullandığımız tk-input içerisindeki native inputu temizleme işlemi
            if (this.multiple && this.editable) {
                this.nativeInputRef.value = null;
                this.renderOptions = await this.filter(null, this.options);
            }
            // Prevent adding 'all' item to value
            this.inputRef.value = [...tmpValue];
            this.value = [...tmpValue];
            this.tkChange.emit([...tmpValue]);
        }
        this.isItemClickFlag = false;
    }
    async handleItemClick(item) {
        var _a;
        if (this.readonly || ((_a = this.optionDisabled) === null || _a === void 0 ? void 0 : _a.call(this, item)))
            return;
        if (this.multiple) {
            const tmpValue = Array.isArray(this.value) ? [...this.value] : [];
            const tmpItem = this.getOptionValue(item);
            if (some(tmpValue, itemValue => isEqual(itemValue, this.getOptionValue(tmpItem)))) {
                // tıklanan item zaten seçili ise seçimi kaldırır
                remove(tmpValue, itemValue => isEqual(itemValue, tmpItem));
            }
            else {
                // tıklanan item seçili değilse ekler
                tmpValue.push(tmpItem);
            }
            // filtreleme ardında yapılan seçimden sonra filtrelem için kullandığımız tk-input içerisindeki native inputu temizleme işlemi
            if (this.multiple && this.editable) {
                this.nativeInputRef.value = null;
            }
            this.inputRef.value = [...tmpValue];
            this.value = [...tmpValue];
            this.tkChange.emit([...tmpValue]);
        }
        else {
            this.value = this.getOptionValue(item);
            this.tkChange.emit(this.getOptionValue(item));
            this.isOpen = false;
        }
        // seçim yapıldıktan sonra eğer filtreleme yapılarak bir seçim yapıldıysa eski filtreleme sonuçlarının tutulmaması ve tüm listesinin optionlarda render edilebilmesi için yapılmıştır.
        this.renderOptions = await this.filter(null, this.options);
        this.isItemClickFlag = false;
    }
    async handleInputChange(value) {
        if (this.multiple) {
            if (value == null) {
                this.value = [];
            }
            else {
                const incomingChips = Array.isArray(value) ? value : [value];
                // Filter out the "others" indicator
                const validChips = incomingChips.filter(val => !(typeof val === 'object' && val !== null && val.__isOthersIndicator));
                // When visibleItemCount is active and we have reordered display, we need to handle removals carefully
                if (this.visibleItemCount && this.selectedItem && this.selectedItem.length > this.visibleItemCount) {
                    // Get the current display value (what the user sees)
                    const currentDisplayValue = this.getDisplayValueForMultiple(this.selectedItem);
                    const currentValidDisplayChips = currentDisplayValue.filter(val => !(typeof val === 'object' && val !== null && val.__isOthersIndicator));
                    // Find which chip was removed by comparing the arrays
                    const removedChip = currentValidDisplayChips.find(displayChip => !validChips.some(validChip => {
                        if (this.optionValueKey && typeof displayChip === 'object' && typeof validChip === 'object') {
                            return this.getOptionValue(displayChip) === this.getOptionValue(validChip);
                        }
                        return isEqual(displayChip, validChip);
                    }));
                    if (removedChip) {
                        // Remove the chip from the actual value array (not the display array)
                        const currentValue = Array.isArray(this.value) ? [...this.value] : [];
                        const removedValue = this.optionValueKey && typeof removedChip === 'object' ? this.getOptionValue(removedChip) : removedChip;
                        const updatedValue = currentValue.filter(val => {
                            if (this.optionValueKey && typeof removedChip === 'object') {
                                return val !== removedValue;
                            }
                            return !isEqual(val, removedValue);
                        });
                        this.value = updatedValue;
                        this.tkChange.emit(this.value);
                        return;
                    }
                }
                // Normal case: no visibleItemCount or no reordering needed
                const resolvedValues = validChips.map(val => {
                    if (typeof val === 'object' && val !== null && this.optionValueKey) {
                        return this.getOptionValue(val);
                    }
                    return val;
                });
                this.value = resolvedValues;
            }
            this.tkChange.emit(this.value);
        }
        else {
            if (this.editable && this.allowCustomValue) {
                this.value = value;
                this.tkChange.emit(value);
            }
            // hızlıca inputda arama yapılırsa verilen filterDebounceDelay süresi içinde
            // tekrar inputda değişiklik yapılırsa sadece bir defa filtre fonksiyonu çalıştırma için yapıldı.
            if (this.filterDebounceDelay > 0) {
                clearTimeout(this.filterDebounceTimeout);
                this.filterDebounceTimeout = setTimeout(async () => {
                    await this.setRenderOptions(value);
                }, this.filterDebounceDelay);
            }
            else {
                await this.setRenderOptions(value);
            }
            if (value === '') {
                this.value = null;
                this.tkChange.emit(null);
            }
        }
    }
    handleInputClick(e) {
        if (this.disabled || this.readonly)
            return;
        const path = e.composedPath();
        const isClearButton = path.some(el => { var _a; return (_a = el.classList) === null || _a === void 0 ? void 0 : _a.contains('tk-input-clear-button'); });
        const isChevron = path.some((el) => el.tagName === 'TK-ICON' && (el.icon === 'keyboard_arrow_up' || el.icon === 'keyboard_arrow_down'));
        const isChipsClearButton = path.some((el) => { var _a; return (_a = el.classList) === null || _a === void 0 ? void 0 : _a.contains('tk-chips-clear-button'); });
        const isInputElement = path.some((el) => { var _a; return (_a = el.classList) === null || _a === void 0 ? void 0 : _a.contains('tk-input'); });
        if (isClearButton || isChipsClearButton)
            return;
        if (isChevron) {
            this.isOpen = !this.isOpen;
            return;
        }
        if (isInputElement && !this.isOpen) {
            this.isOpen = true;
        }
    }
    async handleInputBlur() {
        // item click'den geldiğinde blur çalıştırma
        if (this.isItemClickFlag) {
            this.isItemClickFlag = false;
            return;
        }
        if (!this.editable)
            return;
        // filtreleme ardında yapılan seçimden sonra filtrelem için kullandığımız tk-input içerisindeki native inputu temizleme işlemi
        if (this.multiple) {
            this.nativeInputRef.value = null;
        }
        if (!this.multiple && !this.allowCustomValue) {
            const selectedItem = this.getSelectedItem();
            const inputValue = this.nativeInputRef.value;
            if (!inputValue)
                return;
            // custom value'ya izin verilmiyor ise inputu boşalt
            if (
            // seçili item yok ise ama inutda bir değer var ise
            (!selectedItem && inputValue) ||
                // seçili item var ise ama inputta yazar değer seçili item ile uyuşmuyor ise
                (selectedItem && this.getOptionLabel(selectedItem) != inputValue)) {
                this.value = null;
                this.inputRef.value = null;
                this.tkChange.emit(null);
            }
        }
        this.renderOptions = await this.filter(null, this.options);
    }
    async handleInputKeydown(e) {
        // Prevent default behavior for accessibility keys
        if (['ArrowDown', 'ArrowUp', 'Enter', 'Escape'].includes(e.key)) {
            e.preventDefault();
        }
        // Handle keyboard shortcuts when dropdown is closed
        if (!this.isOpen) {
            if (e.key === 'Enter' || e.key === 'ArrowDown' || e.key === 'ArrowUp') {
                // Enter, Arrow Up/Down: Open dropdown
                if (!this.disabled && !this.readonly) {
                    this.isOpen = true;
                }
                return;
            }
        }
        // Handle keyboard shortcuts when dropdown is open
        if (this.isOpen) {
            if (e.key === 'Escape') {
                // Escape: Close dropdown without selecting
                this.isOpen = false;
                return;
            }
            if (e.key === 'ArrowDown') {
                this.navigateToItem('down');
                return;
            }
            if (e.key === 'ArrowUp') {
                this.navigateToItem('up');
                return;
            }
            if (e.key === 'Enter') {
                const activeItem = this.el.querySelector('.dropdown-item[data-active="true"]');
                if (this.multiple && this.editable && this.allowCustomValue) {
                    this.nativeInputRef.dispatchEvent(new InputEvent('input', { bubbles: true }));
                }
                if (activeItem) {
                    activeItem.click();
                }
                return;
            }
        }
        // Handle Tab key
        if (e.key === 'Tab') {
            this.isOpen = false;
        }
    }
    handleInputClearClick() {
        if (this.multiple && this.optionDisabled && Array.isArray(this.value)) {
            this.value = this.value.filter(item => this.optionDisabled(item));
        }
        else {
            this.value = null;
        }
        this.tkChange.emit(this.value);
        this.selectAll && this.multiple && this.isAllSelected() && this.tkSelectAll.emit(false);
    }
    createOptionItem(options, startIndex = 0) {
        return options === null || options === void 0 ? void 0 : options.map((item, index) => {
            var _a;
            const isDisabled = this.optionDisabled ? (_a = this.optionDisabled) === null || _a === void 0 ? void 0 : _a.call(this, item) : false;
            let itemProps = {};
            let children;
            const checking = some(this.value, itemValue => isEqual(itemValue, this.getOptionValue(item)));
            if (this.multiple) {
                if (this.optionHtml != undefined) {
                    children = (h(Fragment, null, h("tk-checkbox", { value: checking, disabled: isDisabled, "onTk-change": e => e.stopPropagation(), onClick: e => e.preventDefault() }), h("div", { class: "multiple-option-content", innerHTML: this.optionHtml(item) })));
                }
                else {
                    children = (h(Fragment, null, h("tk-checkbox", { value: checking, disabled: isDisabled, "onTk-change": e => e.stopPropagation(), onClick: e => e.preventDefault() }), h("div", null, this.getOptionLabel(item))));
                }
            }
            else {
                if (this.optionHtml != undefined) {
                    itemProps = { innerHTML: this.optionHtml(item) };
                }
                else {
                    itemProps = { innerHTML: this.getOptionLabel(item) };
                }
            }
            return (h("div", Object.assign({ class: classNames('dropdown-item', { multiple: this.multiple }, { disabled: isDisabled }), "data-option-index": startIndex + index, "data-selected": this.multiple && checking ? 'true' : this.value === this.getOptionValue(item) ? 'true' : 'false', onPointerDown: () => (this.isItemClickFlag = true), onClick: () => this.handleItemClick(item) }, itemProps), children));
        });
    }
    createSelectAllOption() {
        var _a;
        if (this.selectAll && this.multiple) {
            const isIndeterminate = !this.isAllSelected() && ((_a = this.value) === null || _a === void 0 ? void 0 : _a.length) > 0;
            const checking = this.isAllSelected();
            return (h("div", null, h("div", { class: classNames('dropdown-item', { multiple: this.multiple }), "data-selected": this.multiple && checking ? 'true' : 'false', onPointerDown: () => (this.isItemClickFlag = true), onClick: () => this.handleSelectAllClick(), "data-option-index": "-1" }, h("tk-checkbox", { indeterminate: isIndeterminate, value: isIndeterminate ? null : checking, "onTk-change": e => e.stopPropagation(), onClick: e => e.preventDefault() }), h("div", null, this.selectAllLabel)), h("tk-divider", { my: 1, style: { margin: '6px 4px 0px' } })));
        }
    }
    createOptions() {
        if (this.isGrouped()) {
            let currentIndex = 0;
            return this.renderOptions.map(group => {
                var _a;
                const groupItems = this.createOptionItem(group[this.groupOptionsKey], currentIndex);
                currentIndex += ((_a = group[this.groupOptionsKey]) === null || _a === void 0 ? void 0 : _a.length) || 0;
                return (h("div", { class: "dropdown-group" }, h("div", { class: "dropdown-group-label" }, h("label", null, group[this.groupNameKey]), h("div", { class: "line" })), groupItems));
            });
        }
        else {
            return this.createOptionItem(this.renderOptions);
        }
    }
    renderInput() {
        var _a;
        return (h("tk-input", { ref: el => (this.inputRef = el), class: classNames('tk-select-input', {
                'editable-select': this.editable,
                'readonly-select': this.readonly,
                'tk-table-input': this.el.classList.contains('tk-table-select'),
                'multiple-select': this.multiple,
                'allow-custom-value-select': this.allowCustomValue,
            }), label: this.label, size: this.size, showAsterisk: this.showAsterisk, hint: this.hint, placeholder: ((_a = this.value) === null || _a === void 0 ? void 0 : _a.length) > 0 ? '' : this.placeholder, invalid: this.invalid, error: this.error, icon: { left: this.icon, right: this.isOpen ? 'keyboard_arrow_up' : 'keyboard_arrow_down' }, mode: this.multiple ? 'chips' : 'text', chipLabelKey: this.optionLabelKey, readonly: this.readonly, disabled: this.disabled, loading: this.loading, clearable: this.clearable, chipOptions: this.chipOptions, chipDisabled: this.optionDisabled, "aria-describedby": "dropdown", "aria-expanded": !!this.isOpen, onClick: e => this.handleInputClick(e), "onTk-change": e => {
                e.stopPropagation();
                this.handleInputChange(e.detail);
            }, "onTk-blur": () => this.handleInputBlur(), "onTk-clear-click": () => this.handleInputClearClick(), onKeyDown: e => this.handleInputKeydown(e) }));
    }
    renderDropdown() {
        var _a;
        if (!this.isOpen)
            return null;
        return (h("div", { class: "tk-select-panel", ref: el => (this.panelRef = el), "data-tk-select-id": this.uniqueId }, h("div", { class: "dropdown-item-holder" }, this.loading ? (h("tk-spinner", { size: this.size })) : (h(Fragment, null, this.panelTopHtml && (h("div", { class: "dropdown-item-top", ref: el => {
                if (el) {
                    const htmlContent = this.panelTopHtml();
                    if (htmlContent instanceof HTMLElement) {
                        el.innerHTML = '';
                        el.appendChild(htmlContent);
                    }
                    else {
                        el.innerHTML = htmlContent;
                    }
                }
            } })), ((_a = this.renderOptions) === null || _a === void 0 ? void 0 : _a.length) > 0 ? (h(Fragment, null, this.createSelectAllOption(), this.createOptions())) : this.hasEmptyDataSlot ? (h("slot", { name: "empty-data" })) : (this.emptyMessage))))));
    }
    render() {
        const rootClasses = classNames('tk-select-container', this.size);
        return (h("div", { key: '64f931ec47171d320e918857784947bb496f8016', "aria-readonly": this.readonly, "aria-disabled": this.disabled, "aria-invalid": this.invalid, class: rootClasses }, this.renderInput(), this.renderDropdown()));
    }
    static get is() { return "tk-select"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-select.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-select.css"]
        };
    }
    static get properties() {
        return {
            "groupNameKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The key to use for option group names.\nRequired if grouped options are used."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "group-name-key",
                "defaultValue": "'label'"
            },
            "groupOptionsKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The key to use for accessing grouped options array.\nRequired if grouped options are used."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "group-options-key",
                "defaultValue": "'options'"
            },
            "loading": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Represents whether the options are fethecd from service or not.\nIf true renders spinner in options dropdown."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "loading",
                "defaultValue": "false"
            },
            "allowCustomValue": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Enables users to enter custom values that are not part of the predefined options."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "allow-custom-value",
                "defaultValue": "false"
            },
            "clearable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input can be cleared"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "clearable",
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The icon displayed in the select box."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If `true`, the user cannot interact with the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "dropdownWidthMode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'match-parent' | 'auto' | string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "match-parent"
                        }],
                    "text": "Determines the width of the dropdown. Accepts values like 'match-parent', 'auto', or a specific width in '300px'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "dropdown-width-mode",
                "defaultValue": "'match-parent'"
            },
            "editable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "This property determines whether the input field within the select box is editable."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "editable",
                "defaultValue": "false"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This is the error message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "emptyMessage": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The message to display when there is no data available."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "empty-message",
                "defaultValue": "'No options available'"
            },
            "filter": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Function",
                    "resolved": "Function",
                    "references": {
                        "Function": {
                            "location": "global",
                            "id": "global::Function"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Function used to filter current options based on the input value. Comes with a default filter function, but can be overridden with a custom function."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "this.defaultFilter"
            },
            "filterDebounceDelay": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets the delay (in ms) before triggering the filter function."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "filter-debounce-delay",
                "defaultValue": "0"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provided a hint or additional information about the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the label for the element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "multiple": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If `true` the user can make multiple selections."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "multiple"
            },
            "visibleItemCount": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The number of items to show in the collapsed select before listing `+N others`."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "visible-item-count"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The name of the control, which is submitted with the form data."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name"
            },
            "placeholder": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Placeholder text displayed when the input is empty."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "placeholder"
            },
            "readonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If `true`, the user cannot modify the value."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "readonly",
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "base"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "chipOptions": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IChipOptions",
                    "resolved": "IChipOptions",
                    "references": {
                        "IChipOptions": {
                            "location": "import",
                            "path": "../tk-chips/types",
                            "id": "src/components/tk-chips/types.ts::IChipOptions",
                            "referenceLocation": "IChipOptions"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets options for all chips rendered in multiple selection mode."
                },
                "getter": false,
                "setter": false
            },
            "panelTopHtml": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Function",
                    "resolved": "Function",
                    "references": {
                        "Function": {
                            "location": "global",
                            "id": "global::Function"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provides a function to customize the panel top content."
                },
                "getter": false,
                "setter": false
            },
            "options": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "any[]",
                    "resolved": "any[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The list of options to be displayed in the select box."
                },
                "getter": false,
                "setter": false
            },
            "optionHtml": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Function",
                    "resolved": "Function",
                    "references": {
                        "Function": {
                            "location": "global",
                            "id": "global::Function"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provides a function to customize the options."
                },
                "getter": false,
                "setter": false
            },
            "optionLabelKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "label"
                        }],
                    "text": "The key to use for option labels"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "option-label-key",
                "defaultValue": "'label'"
            },
            "optionValueKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "value"
                        }],
                    "text": "The key to use for option values"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "option-value-key"
            },
            "selectAll": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If true enables selectAll option"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "select-all",
                "defaultValue": "false"
            },
            "selectAllLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'All'"
                        }],
                    "text": "Sets the label of the selectAll option"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "select-all-label",
                "defaultValue": "'All'"
            },
            "optionDisabled": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Function",
                    "resolved": "Function",
                    "references": {
                        "Function": {
                            "location": "global",
                            "id": "global::Function"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "A function to determine whether an option should be disabled."
                },
                "getter": false,
                "setter": false
            },
            "value": {
                "type": "any",
                "mutable": true,
                "complexType": {
                    "original": "any | any[]",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            }
        };
    }
    static get states() {
        return {
            "renderOptions": {},
            "isOpen": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the value has changed."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "tkSelectAll",
                "name": "tk-select-all",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the selectAll option is changed"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "tkOpen",
                "name": "tk-open",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the select is opened"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkClose",
                "name": "tk-close",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the select is closed"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "close": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Closes the select dropdown;",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "isOpen",
                "methodName": "isOpenChanged"
            }, {
                "propName": "disabled",
                "methodName": "disabledChanged"
            }, {
                "propName": "options",
                "methodName": "optionsChanged"
            }, {
                "propName": "value",
                "methodName": "valueChanged"
            }];
    }
    static get attachInternalsMemberName() { return "internals"; }
}
