// @takeoff-ui/core@0.11.9/dist/collection/components/tk-dropdown/tk-dropdown.js
import { h } from "@stencil/core";
import { v4 as uuidv4 } from "uuid";
import classNames from "classnames";
import { ClickOutsideMixin } from "../../utils/clickoutside-mixin";
import { floatingElementAutoUpdate } from "../../utils/position-utils";
/**
 * TkDropdown creates a dropdown with a trigger element. Items in the options prop can be listed and templated.
 * @slot trigger - The trigger slot defines the element that will trigger the dropdown
 * @slot empty-data - Set how the dropdown will appear when there is no data
 * @react `import { TkDropdown } from '@takeoff-ui/react'`
 * @vue `import { TkDropdown } from '@takeoff-ui/vue'`
 * @angular `import { TkDropdown } from '@takeoff-ui/angular'`
 */
export class TkDropdown {
    constructor() {
        this.hasEmptyDataSlot = false;
        this.uniqueId = uuidv4();
        this.isOpen = false;
        /**
         * The disabled status.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * The message to display when there is no data available.
         */
        this.emptyMessage = 'No options available';
        /**
         * The key to use for option labels
         * @defaultValue label
         */
        this.optionLabelKey = 'label';
        /**
         * The key to use for option values
         * @defaultValue value
         */
        this.optionValueKey = 'value';
        /**
         * The key to use for group names when options are grouped
         * @defaultValue groupName
         */
        this.groupNameKey = 'groupName';
        /**
         * The key to use for grouped options array
         * @defaultValue options
         */
        this.groupOptionsKey = 'options';
        /**
         * Indicates the alignment of options.
         * @defaultValue left
         */
        this.optionsAlign = 'left';
        /**
         * Sets the position of the tooltip.
         * @defaultValue 'right'
         */
        this.position = 'bottom';
        /**
         * Sets size for the dropdown panel.
         * @defaultValue base
         */
        this.size = 'base';
        /**
         * Click outside handler implementation - called by the mixin
         */
        this.closeHandler = () => {
            this.isOpen = false;
        };
    }
    disabledChanged(newValue) {
        if (newValue)
            this.isOpen = false;
    }
    componentWillLoad() {
        this.hasEmptyDataSlot = !!this.el.querySelector('[slot="empty-data"]');
        this.triggerRef = this.el.querySelector('[slot="trigger"]');
        if (this.triggerRef && !this.disabled) {
            this.triggerRef.style.cursor = 'pointer';
            this.triggerRef.addEventListener('click', () => {
                this.isOpen = !this.isOpen;
            });
        }
    }
    componentDidLoad() {
        // Initialize click outside mixin
        this.clickOutsideMixin = new ClickOutsideMixin({
            referenceElement: this.el,
            handler: this.closeHandler,
            disabled: this.disabled || !this.isOpen,
        });
    }
    componentDidUpdate() {
        var _a, _b;
        // Update click outside disabled state based on disabled prop and open state
        this.clickOutsideMixin.updateConfig({ disabled: this.disabled || !this.isOpen });
        if (this.isOpen) {
            // Clean up old floating UI listeners before setting up new ones
            (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
            this.updatePosition();
        }
        else {
            // Remove floating UI listeners when dropdown closes
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
        }
    }
    disconnectedCallback() {
        var _a, _b;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        // Call mixin's disconnectedCallback for cleanup
        (_b = this.clickOutsideMixin) === null || _b === void 0 ? void 0 : _b.disconnectedCallback();
    }
    updatePosition() {
        this.cleanup = floatingElementAutoUpdate(this.triggerRef, this.panelRef, undefined, {
            placement: this.position,
        });
    }
    isGrouped() {
        var _a;
        return ((_a = this.options) === null || _a === void 0 ? void 0 : _a.length) > 0 && this.options[0].hasOwnProperty(this.groupNameKey);
    }
    getOptionLabel(item) {
        return typeof item === 'object' ? item[this.optionLabelKey] : item;
    }
    handleItemClick(item) {
        this.isOpen = false;
        this.tkItemClick.emit(item);
    }
    createOptionItem(options) {
        return options === null || options === void 0 ? void 0 : options.map((item, index) => {
            let optionItem;
            if (this.optionHtml != undefined) {
                optionItem = this.optionHtml(item);
            }
            else {
                optionItem = this.getOptionLabel(item);
            }
            return (h("div", { class: classNames('tk-dropdown-item', this.optionsAlign, { disabled: item === null || item === void 0 ? void 0 : item.disabled }), "data-option-index": index, "data-active": index == 0 ? 'true' : 'false', onClick: () => this.handleItemClick(item), innerHTML: optionItem }));
        });
    }
    createOptions() {
        var _a;
        if (this.isGrouped()) {
            return (_a = this.options) === null || _a === void 0 ? void 0 : _a.map(group => {
                return (h("div", { class: "tk-dropdown-group" }, h("div", { class: "tk-dropdown-group-label" }, h("label", null, group[this.groupNameKey]), h("div", { class: "line" })), this.createOptionItem(group[this.groupOptionsKey])));
            });
        }
        else {
            return this.createOptionItem(this.options);
        }
    }
    renderDropdown() {
        var _a;
        if (!this.isOpen)
            return null;
        return (h("div", { class: "tk-dropdown-panel", ref: el => (this.panelRef = el), "data-tk-dropdown-id": this.uniqueId }, h("div", { class: "tk-dropdown-item-holder" }, ((_a = this.options) === null || _a === void 0 ? void 0 : _a.length) > 0 ? this.createOptions() : this.hasEmptyDataSlot ? h("slot", { name: "empty-data" }) : this.emptyMessage)));
    }
    render() {
        const rootClasses = classNames('tk-dropdown-container', this.size);
        return (h("div", { key: '50792dda7bef333a0a0d45ed6bad04dc26dc7534', class: rootClasses, "data-tk-dropdown-id": this.uniqueId }, h("slot", { key: '49a2bf076febbac475e3c9b23a0518b6db3d6ee0', name: "trigger" }), this.renderDropdown()));
    }
    static get is() { return "tk-dropdown"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-dropdown.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-dropdown.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "The disabled status."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "emptyMessage": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The message to display when there is no data available."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "empty-message",
                "defaultValue": "'No options available'"
            },
            "optionHtml": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Function",
                    "resolved": "Function",
                    "references": {
                        "Function": {
                            "location": "global",
                            "id": "global::Function"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provides a function to customize the options."
                },
                "getter": false,
                "setter": false
            },
            "optionLabelKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "label"
                        }],
                    "text": "The key to use for option labels"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "option-label-key",
                "defaultValue": "'label'"
            },
            "optionValueKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "value"
                        }],
                    "text": "The key to use for option values"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "option-value-key",
                "defaultValue": "'value'"
            },
            "groupNameKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "groupName"
                        }],
                    "text": "The key to use for group names when options are grouped"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "group-name-key",
                "defaultValue": "'groupName'"
            },
            "groupOptionsKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "options"
                        }],
                    "text": "The key to use for grouped options array"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "group-options-key",
                "defaultValue": "'options'"
            },
            "options": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "any[]",
                    "resolved": "any[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The list of options to be displayed in the select box."
                },
                "getter": false,
                "setter": false
            },
            "optionsAlign": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'center' | 'right'",
                    "resolved": "\"center\" | \"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "left"
                        }],
                    "text": "Indicates the alignment of options."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "options-align",
                "defaultValue": "'left'"
            },
            "position": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'top' | 'top-start' | 'top-end' | 'bottom' | 'bottom-start' | 'bottom-end' | 'left' | 'left-start' | 'left-end' | 'right' | 'right-start' | 'right-end'",
                    "resolved": "\"bottom\" | \"bottom-end\" | \"bottom-start\" | \"left\" | \"left-end\" | \"left-start\" | \"right\" | \"right-end\" | \"right-start\" | \"top\" | \"top-end\" | \"top-start\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'right'"
                        }],
                    "text": "Sets the position of the tooltip."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "position",
                "defaultValue": "'bottom'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "base"
                        }],
                    "text": "Sets size for the dropdown panel."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            }
        };
    }
    static get states() {
        return {
            "isOpen": {}
        };
    }
    static get events() {
        return [{
                "method": "tkItemClick",
                "name": "tk-item-click",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the value has changed."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "disabled",
                "methodName": "disabledChanged"
            }];
    }
}
