// @takeoff-ui/core@0.11.9/dist/collection/components/tk-textarea/tk-textarea.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { v4 as uuidv4 } from "uuid";
import { renderHint } from "../../utils/hint-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The TkTextarea component enables multi-line text input with customizable size, validation, and styling options.
 * @react `import { TkTextarea } from '@takeoff-ui/react'`
 * @vue `import { TkTextarea } from '@takeoff-ui/vue'`
 * @angular `import { TkTextarea } from '@takeoff-ui/angular'`
 */
export class TkTextarea {
    constructor() {
        this.uniqueId = uuidv4();
        this.hasFocus = false;
        this.passwordStrength = 0;
        this.charCount = 0;
        /**
         * If `true`, the user cannot interact with the input.
         */
        this.disabled = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Represents the rows value of the component
         * @defaultValue 3
         */
        this.rows = 3;
        /**
         * If `true`, the user cannot modify the value.
         */
        this.readonly = false;
        /**
         * Sets size for the component.
         */
        this.size = 'base';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * The value of the input.
         */
        this.value = '';
        this.updateCharCount = () => {
            if (this.maxLength) {
                this.charCount = this.value.toString().trim().length;
            }
        };
        this.handleInput = (ev) => {
            const input = ev.target;
            this.value = (input === null || input === void 0 ? void 0 : input.value) || '';
            this.tkInput.emit(ev);
        };
        this.handleBlur = () => {
            this.hasFocus = false;
            this.tkBlur.emit();
        };
        this.handleFocus = () => {
            this.hasFocus = true;
            this.tkFocus.emit();
        };
    }
    /**
     * Update the native input element when the value changes
     */
    valueChanged() {
        this.tkChange.emit(this.value);
        this.updateCharCount();
    }
    componentWillLoad() {
        // If the tk-input has a tabindex attribute we get the value
        // and pass it down to the native input, then remove it from the
        // tk-input to avoid causing tabbing twice on the same element
        if (this.el.hasAttribute('tabindex')) {
            const tabindex = this.el.getAttribute('tabindex');
            this.tabindex = tabindex !== null ? tabindex : undefined;
            this.el.removeAttribute('tabindex');
        }
    }
    componentDidLoad() {
        this.nativeInput = this.el.shadowRoot.querySelector('textarea');
    }
    formResetCallback() {
        this.handleFormReset();
    }
    /**
     * Sets focus on the specified `tk-input`. Use this method instead of the global
     * `input.focus()`.
     */
    async setFocus() {
        var _a;
        (_a = this.nativeInput) === null || _a === void 0 ? void 0 : _a.focus();
    }
    handleFormReset() {
        this.value = null;
        this.nativeInput.value = null;
        this.tkChange.emit(this.value);
    }
    renderLabel() {
        var _a;
        let label;
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            const asterisk = (h("span", { class: "asterisk", "data-testid": getDataTestId(this.dataTestid, 'label-asterisk') }, "*"));
            label = (h("label", { htmlFor: this.uniqueId, class: "label", "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label, this.showAsterisk ? asterisk : ''));
        }
        return label;
    }
    render() {
        const rootClasses = classNames('tk-textarea-container', this.size, { focus: this.hasFocus });
        let counter;
        const counterClasses = classNames('counter', this.charCount == this.maxLength && 'maxed');
        if (this.maxLength) {
            counter = (h("span", { key: '16acb14e4ef37bae5e34d28408cc6a697cc48a88', class: counterClasses, "data-testid": getDataTestId(this.dataTestid, 'counter') }, this.charCount, "/", this.maxLength));
        }
        return (h("div", { key: 'f5c34cc6cb7b59b835b54b8b79830b21471ef6b5', "aria-readonly": this.readonly, "aria-disabled": this.disabled, "aria-invalid": this.invalid, class: rootClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, this.renderLabel(), h("div", { key: 'be29df5ed44ca9f4bc9b52903ddc267833557a24', class: "tk-textarea", "data-testid": getDataTestId(this.dataTestid, 'control') }, h("textarea", { key: '5d3fc8d5d32271eb726d0cb103b99951b3f0053c', id: this.uniqueId, ref: el => (this.nativeInput = el), disabled: this.disabled, autoComplete: "off", name: this.name, rows: this.rows, maxLength: this.maxLength, placeholder: this.placeholder || '', readOnly: this.readonly, tabindex: this.tabindex, onInput: this.handleInput, onBlur: this.handleBlur, onFocus: this.handleFocus, value: this.value, "data-testid": getDataTestId(this.dataTestid, 'native-textarea') }), counter), renderHint(this.hint, this.error, this.invalid, this.dataTestid)));
    }
    static get is() { return "tk-textarea"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-textarea.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-textarea.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If `true`, the user cannot interact with the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This is the error message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provided a hint or additional information about the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the label for the element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The name of the control, which is submitted with the form data."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name"
            },
            "placeholder": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Placeholder text displayed when the input is empty."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "placeholder"
            },
            "rows": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "3"
                        }],
                    "text": "Represents the rows value of the component"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rows",
                "defaultValue": "3"
            },
            "readonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If `true`, the user cannot modify the value."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "readonly",
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "maxLength": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Limits the number of characters."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max-length"
            },
            "value": {
                "type": "any",
                "mutable": true,
                "complexType": {
                    "original": "string | number | null",
                    "resolved": "number | string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value",
                "defaultValue": "''"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hasFocus": {},
            "passwordStrength": {},
            "charCount": {}
        };
    }
    static get events() {
        return [{
                "method": "tkInput",
                "name": "tk-input",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a keyboard input occurred."
                },
                "complexType": {
                    "original": "KeyboardEvent",
                    "resolved": "KeyboardEvent",
                    "references": {
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the value has changed."
                },
                "complexType": {
                    "original": "string | number | undefined | null",
                    "resolved": "number | string",
                    "references": {}
                }
            }, {
                "method": "tkBlur",
                "name": "tk-blur",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the input loses focus."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkFocus",
                "name": "tk-focus",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the input has focus."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "setFocus": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Sets focus on the specified `tk-input`. Use this method instead of the global\n`input.focus()`.",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "valueChanged"
            }];
    }
    static get attachInternalsMemberName() { return "internals"; }
}
