// @takeoff-ui/core@0.11.9/dist/collection/components/tk-color-picker/tk-color-picker.js
import { h, Fragment } from "@stencil/core";
import classNames from "classnames";
import { v4 as uuidv4 } from "uuid";
import { parseColorToHsva, hsvaToCss, hsvaToHex, hsvaToRgb, rgbToHsva } from "./color-utils";
import { ClickOutsideMixin } from "../../utils/clickoutside-mixin";
import { floatingElementAutoUpdate } from "../../utils/position-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkColorPicker` component provides a color selection interface with various input formats.
 * It supports HEX and RGB color formats with an optional alpha channel.
 * @slot header - Custom header template (replaces default header completely).
 * @slot header-actions - Custom actions template for header (replaces close button).
 * @slot footer - Custom footer template (replaces default footer container).
 * @slot footer-actions - Custom actions template for footer layout with styled container.
 * @react `import { TkColorPicker } from '@takeoff-ui/react'`
 * @vue `import { TkColorPicker } from '@takeoff-ui/vue'`
 * @angular `import { TkColorPicker } from '@takeoff-ui/angular'`
 */
export class TkColorPicker {
    constructor() {
        this.uniqueId = uuidv4();
        this.isDragging = false;
        this.tempInputVal = null;
        this.pendingValue = null;
        this.isOpen = false;
        this.internalHSVA = { h: 0, s: 0, v: 0, a: 1 };
        this.currentFormat = 'hex';
        this.hasHeaderSlot = false;
        this.hasHeaderActionsSlot = false;
        this.hasFooterSlot = false;
        this.hasFooterActionsSlot = false;
        this.triggerInputValue = '';
        this.isTriggerInputFocused = false;
        /**
         * The current color value (supports HEX and RGB formats)
         */
        this.value = '';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         * @defaultValue false
         */
        this.showAsterisk = false;
        /**
         * If `true`, the user cannot modify the value.
         * @defaultValue false
         */
        this.readonly = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Disables the color picker
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * Sets size for the component.
         * @defaultValue base
         */
        this.size = 'base';
        /**
         * Displays the picker inline without trigger
         * @defaultValue false
         */
        this.inline = false;
        /**
         * Panel layout orientation
         * @defaultValue 'vertical'
         */
        this.orientation = 'vertical';
        /**
         * Color format for display and output
         * @defaultValue 'hex'
         */
        this.format = 'hex';
        /**
         * Array of preset color values
         */
        this.presets = ['#326FD1', '#C79807', '#A45E3C', '#119C8D', '#EDBBA3', '#ABC9FB', '#D0E1FD', '#FF6259', '#717784', '#85B2F9', '#EAD6FD'];
        /**
         * Controls whether the header is shown
         * @defaultValue false
         */
        this.showHeader = false;
        /**
         * Shows/hides the alpha slider
         * @defaultValue true
         */
        this.showAlphaSlider = true;
        /**
         * Shows/hides the preset colors section
         * @defaultValue true
         */
        this.showPresets = true;
        /**
         * Shows/hides the format selector
         * @defaultValue true
         */
        this.showFormatSelector = true;
        /**
         * The type of the header styling
         * @defaultValue 'basic'
         */
        this.headerType = 'basic';
        /**
         * The type of the footer styling
         * @defaultValue 'basic'
         */
        this.footerType = 'basic';
        /**
         * Prevents the panel from being dismissed by clicking outside.
         * Use with footer actions (Apply/Cancel buttons) for controlled closing.
         * @defaultValue false
         */
        this.preventDismiss = false;
        /**
         * Controls whether the close button is shown in the header.
         * Set to false when using footer actions for controlled closing.
         * @defaultValue false
         */
        this.showCloseButton = false;
        this.handleDocumentMouseMove = (e) => {
            if (this.isDragging && this.currentDragHandler) {
                e.preventDefault();
                this.currentDragHandler(e);
            }
        };
        this.handleDocumentMouseUp = () => {
            this.stopDragging();
        };
        this.handleClickOutside = (_) => {
            this.handleApply();
        };
        this.handleTriggerClick = (e) => {
            e.stopPropagation();
            if (this.disabled || this.readonly || this.inline)
                return;
            if (!this.isOpen) {
                this.pendingValue = Object.assign({}, this.internalHSVA);
                this.isOpen = true;
            }
        };
        this.handleSaturationPointer = (e) => {
            if (!this.saturationAreaRef)
                return;
            const rect = this.saturationAreaRef.getBoundingClientRect();
            const x = Math.max(0, Math.min(1, (e.clientX - rect.left) / rect.width));
            const y = Math.max(0, Math.min(1, (e.clientY - rect.top) / rect.height));
            this.updateColor({
                s: Math.round(x * 100),
                v: Math.round((1 - y) * 100),
            });
        };
        this.handleSaturationMouseDown = (e) => {
            this.saturationAreaRef = e.currentTarget;
            this.startDragging(ev => {
                this.handleSaturationPointer(ev);
            }, e);
        };
        this.handleHueChange = (e) => {
            if (!this.hueSliderRef)
                return;
            const rect = this.hueSliderRef.getBoundingClientRect();
            const isVerticalSlider = this.hueSliderRef.classList.contains('tk-color-picker-slider-vertical');
            let percent;
            if (isVerticalSlider) {
                percent = Math.max(0, Math.min(1, (e.clientY - rect.top) / rect.height));
            }
            else {
                percent = Math.max(0, Math.min(1, (e.clientX - rect.left) / rect.width));
            }
            this.updateColor({ h: percent * 360 });
        };
        this.handleHueMouseDown = (e) => {
            this.hueSliderRef = e.currentTarget;
            this.startDragging(this.handleHueChange, e);
        };
        this.handleAlphaChange = (e) => {
            if (!this.alphaSliderRef)
                return;
            const rect = this.alphaSliderRef.getBoundingClientRect();
            const isVerticalSlider = this.alphaSliderRef.classList.contains('tk-color-picker-slider-vertical');
            let percent;
            if (isVerticalSlider) {
                percent = Math.max(0, Math.min(1, (e.clientY - rect.top) / rect.height));
            }
            else {
                percent = Math.max(0, Math.min(1, (e.clientX - rect.left) / rect.width));
            }
            this.updateColor({ a: percent });
        };
        this.handleAlphaMouseDown = (e) => {
            this.alphaSliderRef = e.currentTarget;
            this.startDragging(this.handleAlphaChange, e);
        };
        this.handleRgbChange = (component, value) => {
            const val = Math.max(0, Math.min(255, parseInt(value) || 0));
            const { r, g, b } = hsvaToRgb(this.internalHSVA.h, this.internalHSVA.s, this.internalHSVA.v);
            let newR = r;
            let newG = g;
            let newB = b;
            if (component === 'r')
                newR = val;
            if (component === 'g')
                newG = val;
            if (component === 'b')
                newB = val;
            const newHsva = rgbToHsva(newR, newG, newB, this.internalHSVA.a);
            // Preserve hue when color becomes achromatic (gray/black/white)
            if (newHsva.s === 0 || newHsva.v === 0) {
                newHsva.h = this.internalHSVA.h;
            }
            this.internalHSVA = newHsva;
        };
        this.handleInputBlur = (type) => {
            if (this.tempInputVal === null)
                return;
            const val = String(this.tempInputVal);
            this.tempInputVal = null;
            if (type === 'hex') {
                const newHsva = parseColorToHsva('#' + val);
                // Preserve hue when color becomes achromatic
                if (newHsva.s === 0 || newHsva.v === 0) {
                    newHsva.h = this.internalHSVA.h;
                }
                this.internalHSVA = newHsva;
            }
            else if (type === 'a') {
                const a = Math.max(0, Math.min(100, Number(val))) / 100;
                this.updateColor({ a });
            }
            else {
                this.handleRgbChange(type, val);
            }
        };
        this.handleEyeDropper = async () => {
            if (!window.EyeDropper)
                return;
            try {
                const eyeDropper = new window.EyeDropper();
                const result = await eyeDropper.open();
                this.internalHSVA = parseColorToHsva(result.sRGBHex);
            }
            catch (_a) {
                // User canceled or browser does not support
            }
        };
        this.handleFormatChange = (e) => {
            this.currentFormat = e.detail;
            e.stopPropagation();
        };
        this.handleApply = () => {
            const colorValue = hsvaToCss(this.internalHSVA, this.currentFormat);
            this.value = colorValue;
            this.pendingValue = null;
            this.tkApply.emit(colorValue);
            this.tkChange.emit(colorValue);
            this.isOpen = false;
        };
        this.handleCancel = () => {
            if (this.pendingValue) {
                this.internalHSVA = this.pendingValue;
                this.pendingValue = null;
            }
            this.tkCancel.emit();
            this.isOpen = false;
        };
        this.handleCloseClick = () => {
            this.handleCancel();
        };
        this.handleTriggerInputFocus = () => {
            this.isTriggerInputFocused = true;
            // Only set triggerInputValue if there's an actual value
            // If value is empty, keep triggerInputValue empty so user can type from scratch
            this.triggerInputValue = this.value || '';
        };
        this.handleTriggerInputBlur = () => {
            this.isTriggerInputFocused = false;
            // Apply the typed value on blur
            this.applyTriggerInputValue();
        };
        this.handleTriggerInputChange = (e) => {
            const input = e.detail;
            this.triggerInputValue = input.value;
        };
        this.handleTriggerInputKeyDown = (e) => {
            if (e.key === 'Enter') {
                e.preventDefault();
                this.applyTriggerInputValue();
            }
            else if (e.key === 'Escape') {
                e.preventDefault();
                // Reset to current color value
                const colorCss = hsvaToCss(this.internalHSVA, this.currentFormat);
                this.triggerInputValue = colorCss;
            }
        };
    }
    isOpenChanged(newVal) {
        if (newVal) {
            this.tkOpen.emit();
        }
        else {
            this.tkClose.emit();
        }
    }
    valueChanged(newVal) {
        if (this.isDragging)
            return;
        this.internalHSVA = parseColorToHsva(newVal);
    }
    disabledChanged(newValue) {
        if (newValue)
            this.isOpen = false;
    }
    componentWillLoad() {
        this.internalHSVA = parseColorToHsva(this.value || '#000000');
        this.currentFormat = this.format === 'hex' || this.format === 'rgba' ? this.format : 'hex';
        this.hasHeaderSlot = !!this.el.querySelector(':scope > [slot="header"]');
        this.hasHeaderActionsSlot = !!this.el.querySelector(':scope > [slot="header-actions"]');
        this.hasFooterSlot = !!this.el.querySelector(':scope > [slot="footer"]');
        this.hasFooterActionsSlot = !!this.el.querySelector(':scope > [slot="footer-actions"]');
    }
    componentDidLoad() {
        var _a;
        const tkInputArea = (_a = this.inputRef) === null || _a === void 0 ? void 0 : _a.querySelector('.tk-input');
        this.clickOutsideMixin = new ClickOutsideMixin({
            referenceElement: tkInputArea !== null && tkInputArea !== void 0 ? tkInputArea : this.el,
            handler: this.handleClickOutside,
            disabled: !this.isOpen || this.disabled || this.readonly || this.inline || this.preventDismiss,
        });
    }
    componentDidUpdate() {
        var _a, _b, _c;
        (_a = this.clickOutsideMixin) === null || _a === void 0 ? void 0 : _a.updateConfig({
            disabled: !this.isOpen || this.disabled || this.readonly || this.inline || this.preventDismiss,
            ignoredElements: this.panelRef ? [this.panelRef] : [],
        });
        if (this.isOpen && this.inputRef && this.panelRef) {
            // Clean up old floating UI listeners before setting up new ones
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            this.updatePosition();
        }
        else {
            // Remove floating UI listeners when dropdown closes
            (_c = this.cleanup) === null || _c === void 0 ? void 0 : _c.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
        }
    }
    disconnectedCallback() {
        var _a, _b;
        (_a = this.clickOutsideMixin) === null || _a === void 0 ? void 0 : _a.disconnectedCallback();
        // Clean up floating UI listeners on component unmount
        (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        this.stopDragging();
    }
    /**
     * Opens the color picker panel
     */
    async open() {
        if (!this.disabled && !this.readonly && !this.inline) {
            this.pendingValue = Object.assign({}, this.internalHSVA);
            this.isOpen = true;
        }
    }
    /**
     * Closes the color picker panel
     */
    async close() {
        this.isOpen = false;
    }
    /**
     * Gets the current color value in the specified format
     * @param format - Output format (defaults to current format)
     */
    async getValue(format) {
        return hsvaToCss(this.internalHSVA, format || this.currentFormat);
    }
    /**
     * Applies the current color selection and closes the panel
     * Use this method in custom footer actions
     */
    async apply() {
        this.handleApply();
    }
    /**
     * Cancels the color selection, reverts to previous value, and closes the panel
     * Use this method in custom footer actions
     */
    async cancel() {
        this.handleCancel();
    }
    startDragging(handler, initialEvent) {
        this.isDragging = true;
        this.currentDragHandler = handler;
        handler(initialEvent);
        document.addEventListener('mousemove', this.handleDocumentMouseMove);
        document.addEventListener('mouseup', this.handleDocumentMouseUp);
    }
    stopDragging() {
        this.isDragging = false;
        this.currentDragHandler = undefined;
        document.removeEventListener('mousemove', this.handleDocumentMouseMove);
        document.removeEventListener('mouseup', this.handleDocumentMouseUp);
    }
    updatePosition() {
        const inputEl = this.inputRef.querySelector('.tk-input');
        this.cleanup = floatingElementAutoUpdate(inputEl, this.panelRef, undefined, {
            placement: 'bottom-end',
        });
    }
    /**
     * Updates internal HSVA state with partial values
     */
    updateColor(partial) {
        this.internalHSVA = Object.assign(Object.assign({}, this.internalHSVA), partial);
    }
    applyTriggerInputValue() {
        if (!this.triggerInputValue)
            return;
        const parsed = parseColorToHsva(this.triggerInputValue);
        // Check if parsed successfully (not all zeros unless input was black)
        if (parsed.h !== 0 || parsed.s !== 0 || parsed.v !== 0 || this.triggerInputValue.toLowerCase().includes('000')) {
            this.internalHSVA = parsed;
            const newValue = hsvaToCss(this.internalHSVA, this.currentFormat);
            this.value = newValue;
            this.tkChange.emit(newValue);
        }
    }
    handlePresetSelect(color) {
        this.internalHSVA = parseColorToHsva(color);
    }
    createEyedropperButton() {
        return (h("tk-button", { variant: "neutral", type: "outlined", size: "small", icon: "colorize", "onTk-click": this.handleEyeDropper, disabled: this.disabled, dataTestid: getDataTestId(this.dataTestid, 'eyedropper-button') }));
    }
    createColorPreview() {
        return (h("div", { class: "tk-color-picker-preview-box", style: { backgroundColor: hsvaToCss(this.internalHSVA, this.currentFormat) }, "data-testid": getDataTestId(this.dataTestid, 'preview') }));
    }
    createSaturation() {
        const { h: hue, s, v } = this.internalHSVA;
        return (h("div", { class: "tk-color-picker-saturation-container", style: { backgroundColor: `hsl(${hue}, 100%, 50%)` }, onMouseDown: this.handleSaturationMouseDown, "data-testid": getDataTestId(this.dataTestid, 'saturation') }, h("div", { class: "tk-color-picker-saturation-white", "data-testid": getDataTestId(this.dataTestid, 'white') }), h("div", { class: "tk-color-picker-saturation-black", "data-testid": getDataTestId(this.dataTestid, 'black') }), h("div", { class: "tk-color-picker-saturation-pointer", style: {
                left: `${s}%`,
                top: `${100 - v}%`,
            }, "data-testid": getDataTestId(this.dataTestid, 'pointer') })));
    }
    createSliders() {
        const isHorizontalLayout = this.orientation === 'horizontal';
        const sliderClass = classNames('tk-color-picker-slider', {
            'tk-color-picker-slider-vertical': isHorizontalLayout,
        });
        const huePercent = (this.internalHSVA.h / 360) * 100;
        return (h("div", { class: classNames('tk-color-picker-sliders', { 'tk-color-picker-sliders-vertical-layout': isHorizontalLayout }), "data-testid": getDataTestId(this.dataTestid, 'sliders') }, h("div", { class: `${sliderClass} tk-color-picker-hue`, ref: el => (this.hueSliderRef = el), onMouseDown: this.handleHueMouseDown, "data-testid": getDataTestId(this.dataTestid, 'hue') }, h("div", { class: "tk-color-picker-slider-thumb", style: isHorizontalLayout ? { top: `${huePercent}%`, left: '50%' } : { left: `${huePercent}%`, top: '50%' }, "data-testid": getDataTestId(this.dataTestid, 'hue-thumb') }, h("div", { class: "tk-color-picker-slider-thumb-inner", "data-testid": getDataTestId(this.dataTestid, 'hue-inner-thumb') }))), this.showAlphaSlider && (h("div", { class: sliderClass, ref: el => (this.alphaSliderRef = el), onMouseDown: this.handleAlphaMouseDown, "data-testid": getDataTestId(this.dataTestid, 'alpha') }, h("div", { class: "tk-color-picker-alpha-bg", "data-testid": getDataTestId(this.dataTestid, 'slider') }), h("div", { class: "tk-color-picker-alpha-overlay", style: { background: `linear-gradient(${isHorizontalLayout ? 'to bottom' : 'to right'}, transparent, ${hsvaToHex(this.internalHSVA)})` }, "data-testid": getDataTestId(this.dataTestid, 'overlay') }), h("div", { class: "tk-color-picker-slider-thumb", style: isHorizontalLayout ? { top: `${this.internalHSVA.a * 100}%`, left: '50%' } : { left: `${this.internalHSVA.a * 100}%`, top: '50%' }, "data-testid": getDataTestId(this.dataTestid, 'alpha-thumb') }, h("div", { class: "tk-color-picker-slider-thumb-inner", "data-testid": getDataTestId(this.dataTestid, 'alpha-inner-thumb') }))))));
    }
    createAlphaInput() {
        if (!this.showAlphaSlider)
            return null;
        return (h("tk-input", { class: "tk-color-picker-alpha-input", mode: "number", size: "small", min: "0", max: "100", value: (this.internalHSVA.a * 100).toFixed(0), disabled: this.disabled, "onTk-change": (e) => (this.tempInputVal = e.detail), "onTk-blur": () => this.handleInputBlur('a'), dataTestid: getDataTestId(this.dataTestid, 'alpha-input') }));
    }
    createInputs() {
        const { r, g, b } = hsvaToRgb(this.internalHSVA.h, this.internalHSVA.s, this.internalHSVA.v);
        return (h("div", { class: "tk-color-picker-inputs", "data-testid": getDataTestId(this.dataTestid, 'inputs') }, this.showFormatSelector && (h("tk-select", { class: "tk-color-picker-format-select", size: "small", value: this.currentFormat, optionValueKey: "value", disabled: this.disabled, options: [
                { value: 'hex', label: 'HEX' },
                { value: 'rgba', label: 'RGB' },
            ], "onTk-change": this.handleFormatChange, "data-testid": getDataTestId(this.dataTestid, 'format-select') })), h("div", { class: "tk-color-picker-input-fields", "data-testid": getDataTestId(this.dataTestid, 'input-fields') }, this.currentFormat === 'hex' && (h(Fragment, null, h("tk-input", { class: "tk-color-picker-hex-input", size: "small", icon: "tag", iconPosition: "left", value: hsvaToHex(this.internalHSVA).replace('#', ''), disabled: this.disabled, "onTk-change": (e) => (this.tempInputVal = e.detail), "onTk-blur": () => this.handleInputBlur('hex'), dataTestid: getDataTestId(this.dataTestid, 'hex-input') }), this.createAlphaInput())), this.currentFormat === 'rgba' && (h("div", { class: "tk-color-picker-rgb-inputs", "data-testid": getDataTestId(this.dataTestid, 'rgb') }, h("tk-input", { class: "tk-color-picker-rgb-input", size: "small", value: r.toString(), disabled: this.disabled, "onTk-change": (e) => (this.tempInputVal = e.detail), "onTk-blur": () => this.handleInputBlur('r'), dataTestid: getDataTestId(this.dataTestid, 'red-input') }), h("tk-input", { class: "tk-color-picker-rgb-input", size: "small", value: g.toString(), disabled: this.disabled, "onTk-change": (e) => (this.tempInputVal = e.detail), "onTk-blur": () => this.handleInputBlur('g'), dataTestid: getDataTestId(this.dataTestid, 'green-input') }), h("tk-input", { class: "tk-color-picker-rgb-input", size: "small", value: b.toString(), disabled: this.disabled, "onTk-change": (e) => (this.tempInputVal = e.detail), "onTk-blur": () => this.handleInputBlur('b'), dataTestid: getDataTestId(this.dataTestid, 'blue-input') }), this.createAlphaInput())))));
    }
    createPresets() {
        var _a;
        if (!this.showPresets || !((_a = this.presets) === null || _a === void 0 ? void 0 : _a.length))
            return null;
        return (h("div", { class: "tk-color-picker-presets", "data-testid": getDataTestId(this.dataTestid, 'presets') }, this.presets.map((color, index) => (h("div", { class: "tk-color-picker-preset", style: { backgroundColor: color }, onClick: () => this.handlePresetSelect(color), "data-testid": getDataTestId(this.dataTestid, 'preset', index.toString()) })))));
    }
    createFooter() {
        if (this.hasFooterSlot) {
            return h("slot", { name: "footer" });
        }
        if (this.hasFooterActionsSlot) {
            const footerClasses = classNames('tk-color-picker-panel-footer', `tk-color-picker-panel-footer-${this.footerType}`);
            return (h("div", { class: footerClasses, "data-testid": getDataTestId(this.dataTestid, 'footer-actions') }, h("slot", { name: "footer-actions" })));
        }
        // No footer by default - footer is entirely slot-based
        return null;
    }
    createPanelHeader() {
        if (!this.showHeader)
            return null;
        if (this.hasHeaderSlot) {
            return h("slot", { name: "header" });
        }
        else {
            const headerClasses = classNames('tk-color-picker-panel-header', `tk-color-picker-panel-header-${this.headerType}`);
            return (h("div", { class: headerClasses, "data-testid": getDataTestId(this.dataTestid, 'header') }, h("span", { class: "tk-color-picker-title", "data-testid": getDataTestId(this.dataTestid, 'title') }, this.header), this.hasHeaderActionsSlot ? (h("slot", { name: "header-actions" })) : (!this.inline &&
                this.showCloseButton && (h("tk-button", { variant: "neutral", type: "text", size: "small", icon: "close", onClick: this.handleCloseClick, dataTestid: getDataTestId(this.dataTestid, 'close-button') })))));
        }
    }
    createPanelBody() {
        if (this.orientation === 'horizontal') {
            return (h(Fragment, null, h("div", { class: "tk-color-picker-body-horizontal", "data-testid": getDataTestId(this.dataTestid, 'horizontal') }, h("div", { class: "tk-color-picker-controls-side", "data-testid": getDataTestId(this.dataTestid, 'controls') }, this.createColorPreview(), this.createSliders(), this.createEyedropperButton()), h("div", { class: "tk-color-picker-saturation-wrapper", "data-testid": getDataTestId(this.dataTestid, 'saturation-wrapper') }, this.createSaturation())), this.createInputs(), this.createPresets()));
        }
        return (h(Fragment, null, h("div", { class: "tk-color-picker-body-vertical", "data-testid": getDataTestId(this.dataTestid, 'vertical') }, this.createSaturation(), h("div", { class: "tk-color-picker-controls-bottom", "data-testid": getDataTestId(this.dataTestid, 'controls') }, h("div", { class: "tk-color-picker-eyedropper-row", "data-testid": getDataTestId(this.dataTestid, 'eyedropper-row') }, this.createEyedropperButton(), this.createSliders(), this.createColorPreview()), this.createInputs())), this.createPresets()));
    }
    renderInput() {
        if (this.inline)
            return null;
        const displayValue = this.isTriggerInputFocused ? this.triggerInputValue : this.value;
        const iconColor = this.value || '#000000';
        return (h("tk-input", { ref: el => (this.inputRef = el), label: this.label, readonly: this.readonly, disabled: this.disabled, invalid: this.invalid, error: this.error, hint: this.hint, showAsterisk: this.showAsterisk, name: this.name, size: this.size, placeholder: this.placeholder, icon: {
                left: {
                    name: 'circle',
                    color: iconColor,
                    style: 'outlined',
                    fill: true,
                },
                right: this.isOpen ? 'keyboard_arrow_up' : 'keyboard_arrow_down',
            }, value: displayValue, "aria-describedby": "dropdown", "aria-expanded": this.isOpen, onClick: this.handleTriggerClick, "onTk-focus": this.handleTriggerInputFocus, "onTk-blur": this.handleTriggerInputBlur, "onTk-change": this.handleTriggerInputChange, onKeyDown: this.handleTriggerInputKeyDown, dataTestid: getDataTestId(this.dataTestid, 'trigger-input') }));
    }
    renderPanel() {
        const panelCls = classNames('tk-color-picker-panel', {
            'tk-color-picker-panel-inline': this.inline,
            'tk-color-picker-panel-horizontal': this.orientation === 'horizontal',
        });
        return (h("div", { class: panelCls, ref: el => (this.panelRef = el), "data-tk-id": this.uniqueId, "data-testid": getDataTestId(this.dataTestid, 'panel') }, this.createPanelHeader(), h("div", { class: "tk-color-picker-panel-body", "data-testid": getDataTestId(this.dataTestid, 'body') }, this.createPanelBody()), this.createFooter()));
    }
    render() {
        const rootClasses = classNames('tk-color-picker', {
            'tk-color-picker-inline': this.inline,
            'tk-color-picker-horizontal': this.orientation === 'horizontal',
        });
        return (h("div", { key: '78165c5fbbf530e7837fb46c6341b4ffbf77f0c1', class: rootClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, this.renderInput(), (this.isOpen || this.inline) && this.renderPanel()));
    }
    static get is() { return "tk-color-picker"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-color-picker.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-color-picker.css"]
        };
    }
    static get properties() {
        return {
            "value": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The current color value (supports HEX and RGB formats)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value",
                "defaultValue": "''"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label text displayed above the trigger input"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "header": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'Color Picker'"
                        }],
                    "text": "Title displayed in the panel header"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "readonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If `true`, the user cannot modify the value."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "readonly",
                "defaultValue": "false"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The name of the control, which is submitted with the form data."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This is the error message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provided a hint or additional information about the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "placeholder": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Placeholder text displayed when the input is empty."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "placeholder"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Disables the color picker"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "base"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "inline": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Displays the picker inline without trigger"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "inline",
                "defaultValue": "false"
            },
            "orientation": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'vertical' | 'horizontal'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'vertical'"
                        }],
                    "text": "Panel layout orientation"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "orientation",
                "defaultValue": "'vertical'"
            },
            "format": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'hex' | 'rgba'",
                    "resolved": "\"hex\" | \"rgba\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'hex'"
                        }],
                    "text": "Color format for display and output"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "format",
                "defaultValue": "'hex'"
            },
            "presets": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Array of preset color values"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "['#326FD1', '#C79807', '#A45E3C', '#119C8D', '#EDBBA3', '#ABC9FB', '#D0E1FD', '#FF6259', '#717784', '#85B2F9', '#EAD6FD']"
            },
            "showHeader": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the header is shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-header",
                "defaultValue": "false"
            },
            "showAlphaSlider": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Shows/hides the alpha slider"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-alpha-slider",
                "defaultValue": "true"
            },
            "showPresets": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Shows/hides the preset colors section"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-presets",
                "defaultValue": "true"
            },
            "showFormatSelector": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Shows/hides the format selector"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-format-selector",
                "defaultValue": "true"
            },
            "headerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light' | 'dark' | 'primary'",
                    "resolved": "\"basic\" | \"dark\" | \"divided\" | \"light\" | \"primary\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'basic'"
                        }],
                    "text": "The type of the header styling"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header-type",
                "defaultValue": "'basic'"
            },
            "footerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light'",
                    "resolved": "\"basic\" | \"divided\" | \"light\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'basic'"
                        }],
                    "text": "The type of the footer styling"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "footer-type",
                "defaultValue": "'basic'"
            },
            "preventDismiss": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Prevents the panel from being dismissed by clicking outside.\nUse with footer actions (Apply/Cancel buttons) for controlled closing."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "prevent-dismiss",
                "defaultValue": "false"
            },
            "showCloseButton": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the close button is shown in the header.\nSet to false when using footer actions for controlled closing."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-close-button",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "isOpen": {},
            "internalHSVA": {},
            "currentFormat": {},
            "hasHeaderSlot": {},
            "hasHeaderActionsSlot": {},
            "hasFooterSlot": {},
            "hasFooterActionsSlot": {},
            "triggerInputValue": {},
            "isTriggerInputFocused": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the color is applied/confirmed"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }, {
                "method": "tkOpen",
                "name": "tk-open",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the panel opens"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkClose",
                "name": "tk-close",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the panel closes"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkApply",
                "name": "tk-apply",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the apply button is clicked"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }, {
                "method": "tkCancel",
                "name": "tk-cancel",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the cancel button is clicked"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "open": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Opens the color picker panel",
                    "tags": []
                }
            },
            "close": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Closes the color picker panel",
                    "tags": []
                }
            },
            "getValue": {
                "complexType": {
                    "signature": "(format?: \"hex\" | \"rgba\") => Promise<string>",
                    "parameters": [{
                            "name": "format",
                            "type": "\"hex\" | \"rgba\"",
                            "docs": "- Output format (defaults to current format)"
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<string>"
                },
                "docs": {
                    "text": "Gets the current color value in the specified format",
                    "tags": [{
                            "name": "param",
                            "text": "format - Output format (defaults to current format)"
                        }]
                }
            },
            "apply": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Applies the current color selection and closes the panel\nUse this method in custom footer actions",
                    "tags": []
                }
            },
            "cancel": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Cancels the color selection, reverts to previous value, and closes the panel\nUse this method in custom footer actions",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "isOpen",
                "methodName": "isOpenChanged"
            }, {
                "propName": "value",
                "methodName": "valueChanged"
            }, {
                "propName": "disabled",
                "methodName": "disabledChanged"
            }];
    }
}
