// @takeoff-ui/core@0.11.9/dist/cjs/clickoutside-mixin-BGIuHTh6.js
'use strict';

/**
 * Mixin that provides click outside functionality for components.
 * Handles event listener management, binding, cleanup, and click detection.
 */
class ClickOutsideMixin {
    checkConfig() {
        if (!this.config.referenceElement) {
            throw new Error('Reference element must be provided');
        }
        if (!this.config.handler) {
            throw new Error('Handler must be provided');
        }
    }
    constructor(initialConfig = {}) {
        /**
         * Configuration object containing all click outside settings.
         */
        this.config = {
            useCapture: true,
            ignoredElements: [],
            disabled: false,
        };
        this._isListenerActive = false;
        /**
         * Handles the window click event and determines if it's an outside click.
         */
        this._handleWindowClick = (event) => {
            var _a;
            const composedPath = event.composedPath();
            // Check if click is inside the reference element
            const isInsideReference = composedPath.some(item => item === this.config.referenceElement);
            // Check if click is on any ignored elements
            const isOnIgnoredElement = (_a = this.config.ignoredElements) === null || _a === void 0 ? void 0 : _a.some(ignoredEl => composedPath.some(item => item === ignoredEl));
            // If click is outside reference and not on ignored elements, trigger handler
            if (!isInsideReference && !isOnIgnoredElement) {
                this.config.handler(event);
            }
        };
        this.updateConfig(initialConfig);
        this.checkConfig();
    }
    evaluateListenerState() {
        if (!this.config.disabled && !this._isListenerActive) {
            this.bindListener();
        }
        else if (this.config.disabled && this._isListenerActive) {
            this.unbindListener();
        }
    }
    /**
     * Updates the configuration with partial config object.
     * @param partialConfig - Partial configuration to merge with current config
     */
    updateConfig(partialConfig) {
        this.config = Object.assign(Object.assign({}, this.config), partialConfig);
        // Re-evaluate listener state based on new config
        this.evaluateListenerState();
    }
    /**
     * Binds the window click listener.
     * Should be called when the component needs to start listening for outside clicks.
     */
    bindListener() {
        if (this.config.disabled || this._isListenerActive)
            return;
        window.addEventListener('click', this._handleWindowClick, this.config.useCapture);
        this._isListenerActive = true;
    }
    /**
     * Unbinds the window click listener.
     * Should be called when the component no longer needs to listen for outside clicks.
     */
    unbindListener() {
        if (!this._isListenerActive)
            return;
        window.removeEventListener('click', this._handleWindowClick, this.config.useCapture);
        this._isListenerActive = false;
    }
    /**
     * Cleanup in disconnectedCallback.
     */
    disconnectedCallback() {
        this.unbindListener();
    }
}

exports.ClickOutsideMixin = ClickOutsideMixin;
