// @takeoff-ui/core@0.11.9/dist/cjs/index-CUpr0td9.js
'use strict';

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

const NAMESPACE = 'core';
const BUILD = /* core */ { hotModuleReplacement: false, hydratedSelectorName: "hydrated", lazyLoad: true, propChangeCallback: true, shadowDom: true, slotRelocation: true, state: true, updatable: true};

const globalScripts = () => {};
const globalStyles = "@import url(\"https://fonts.googleapis.com/css2?family=Geologica:wght@100..900&display=swap\");@font-face{font-family:\"Material Symbols Outlined\";src:url(\"./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2\") format(\"woff2\");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:\"Material Symbols Outlined\";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:\"liga\";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:\"FILL\" 0}.material-symbols-outlined.fill{font-variation-settings:\"FILL\" 1}@font-face{font-family:\"Material Symbols Rounded\";src:url(\"./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2\") format(\"woff2\");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:\"Material Symbols Rounded\";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:\"liga\";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:\"FILL\" 0}.material-symbols-rounded.fill{font-variation-settings:\"FILL\" 1}@font-face{font-family:\"Material Symbols Sharp\";src:url(\"./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2\") format(\"woff2\");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:\"Material Symbols Sharp\";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:\"liga\";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:\"FILL\" 0}.material-symbols-sharp.fill{font-variation-settings:\"FILL\" 1}:root{--primary-50:#fae6e8;--primary-100:#eeb0b8;--primary-200:#e68a95;--primary-300:#db5465;--primary-400:#d43347;--primary-500:#c90019;--primary-600:#b70017;--primary-700:#8f0012;--primary-800:#6f000e;--primary-900:#54000b;--static-black:#000000;--static-white:#ffffff;--secondary-50:#f2f3f5;--secondary-100:#dadee3;--secondary-200:#c1c8d1;--secondary-300:#a9b3be;--secondary-400:#919dac;--secondary-500:#79889a;--secondary-600:#607083;--secondary-700:#49586a;--secondary-800:#34404f;--secondary-900:#1f2833;--neutral-0:#ffffff;--neutral-50:#f9fafc;--neutral-100:#f2f5f8;--neutral-200:#e1e4ea;--neutral-300:#cacfd8;--neutral-400:#99a0ae;--neutral-500:#717784;--neutral-600:#525866;--neutral-700:#2b303b;--neutral-800:#222530;--neutral-900:#181b25;--neutral-1000:#111216;--red-50:#fff5f5;--red-100:#ffd0ce;--red-200:#ffaca7;--red-300:#ff8780;--red-400:#ff6259;--red-500:#ff3d32;--red-600:#d9342b;--red-700:#b32b23;--red-800:#8c221c;--red-900:#661814;--blue-50:#f5f9ff;--blue-100:#d0e1fd;--blue-200:#abc9fb;--blue-300:#85b2f9;--blue-400:#609af8;--blue-500:#3b82f6;--blue-600:#326fd1;--blue-700:#295bac;--blue-800:#204887;--blue-900:#183462;--green-50:#f4fcf7;--green-100:#caf1d8;--green-200:#a0e6ba;--green-300:#76db9b;--green-400:#4cd07d;--green-500:#22c55e;--green-600:#1da750;--green-700:#188a42;--green-800:#136c34;--green-900:#0e4f26;--yellow-50:#fefbf3;--yellow-100:#faedc4;--yellow-200:#f6de95;--yellow-300:#f2d066;--yellow-400:#eec137;--yellow-500:#eab308;--yellow-600:#c79807;--yellow-700:#a47d06;--yellow-800:#816204;--yellow-900:#5e4803;--purple-50:#fbf7ff;--purple-100:#ead6fd;--purple-200:#dab6fc;--purple-300:#c996fa;--purple-400:#b975f9;--purple-500:#a855f7;--purple-600:#8f48d2;--purple-700:#763cad;--purple-800:#5c2f88;--purple-900:#432263;--cyan-50:#f3fbfd;--cyan-100:#c3edf5;--cyan-200:#94e0ed;--cyan-300:#65d2e4;--cyan-400:#35c4dc;--cyan-500:#06b6d4;--cyan-600:#059bb4;--cyan-700:#047f94;--cyan-800:#036475;--cyan-900:#024955;--business-50:#fdf0e9;--business-100:#f7d5c4;--business-200:#edbba3;--business-300:#e0a385;--business-400:#cf8c6b;--business-500:#ba7655;--business-600:#a45e3c;--business-700:#894727;--business-800:#6a3216;--business-900:#4e3224;--teal-50:#f3fbfb;--teal-100:#c7eeea;--teal-200:#9ae0d9;--teal-300:#6dd3c8;--teal-400:#41c5b7;--teal-500:#14b8a6;--teal-600:#119c8d;--teal-700:#0e8174;--teal-800:#0b655b;--teal-900:#084a42;--alpha-brand-primary-500-48:#c900197a;--alpha-brand-primary-500-24:#c900193d;--alpha-brand-primary-500-16:#c9001928;--alpha-brand-primary-500-8:#c9001914;--alpha-brand-primary-500-4:#c900190a;--alpha-brand-primary-500-2:#c9001905;--alpha-brand-secondary-500-24:#79889a3d;--alpha-brand-secondary-500-16:#79889a28;--alpha-brand-secondary-500-8:#79889a14;--alpha-base-black-72:#000000b7;--alpha-base-black-64:#000000a3;--alpha-base-black-48:#0000007a;--alpha-base-black-32:#00000051;--alpha-base-black-24:#0000003d;--alpha-base-black-16:#00000028;--alpha-base-black-12:#0000001e;--alpha-base-black-8:#00000014;--alpha-base-black-4:#0000000a;--alpha-base-black-2:#00000005;--alpha-base-black-none:#00000000;--alpha-base-white-72:#ffffffb7;--alpha-base-white-64:#ffffffa3;--alpha-base-white-48:#ffffff7a;--alpha-base-white-32:#ffffff51;--alpha-base-white-24:#ffffff3d;--alpha-base-white-16:#ffffff28;--alpha-base-white-12:#ffffff1e;--alpha-base-white-8:#ffffff14;--alpha-base-white-4:#ffffff0a;--alpha-base-white-2:#ffffff05;--alpha-base-white-none:#ffffff00;--alpha-blue-blue-500-72:#3b82f6b7;--alpha-blue-blue-500-64:#3b82f6a3;--alpha-blue-blue-500-48:#3b82f67a;--alpha-blue-blue-500-32:#3b82f651;--alpha-blue-blue-500-24:#3b82f63d;--alpha-blue-blue-500-16:#3b82f628;--alpha-blue-blue-500-12:#3b82f61e;--alpha-blue-blue-500-8:#3b82f614;--alpha-blue-blue-500-4:#3b82f60a;--alpha-blue-blue-500-2:#3b82f605;--alpha-neutral-neutral-500-72:#525866b7;--alpha-neutral-neutral-500-64:#525866a3;--alpha-neutral-neutral-500-48:#5258667a;--alpha-neutral-neutral-500-32:#52586651;--alpha-neutral-neutral-500-24:#5258663d;--alpha-neutral-neutral-500-16:#52586628;--alpha-neutral-neutral-500-12:#5258661e;--alpha-neutral-neutral-500-8:#52586614;--alpha-neutral-neutral-500-4:#5258660a;--alpha-neutral-neutral-500-2:#52586605;--aviation-amber-50:#fff2e5;--aviation-amber-100:#ffd9b2;--aviation-amber-200:#ffcc99;--aviation-amber-300:#ffb366;--aviation-amber-400:#ff9933;--aviation-amber-500:#ff8000;--aviation-amber-600:#cc6600;--aviation-amber-700:#994d00;--aviation-amber-800:#663300;--aviation-amber-900:#4d2600;--family-title:'Geologica';--family-label:'Geologica';--family-body:'Geologica';--family-subheading:'Geologica';--size-xxs:11px;--size-xs:12px;--size-sm:14px;--size-base:16px;--size-lg:18px;--size-xl:20px;--size-2xl:24px;--size-3xl:30px;--size-4xl:36px;--size-5xl:48px;--size-6xl:60px;--size-7xl:72px;--size-8xl:96px;--size-9xl:128px;--weight-thin:100;--weight-extralight:200;--weight-light:300;--weight-regular:400;--weight-medium:500;--weight-semibold:600;--weight-bold:700;--weight-extrabold:800;--weight-black:900;--line-height-xxs-normal:16px;--line-height-xs-normal:18px;--line-height-sm-normal:20px;--line-height-base-normal:24px;--line-height-lg-normal:27px;--line-height-xl-normal:30px;--line-height-2xl-normal:36px;--line-height-3xl-normal:45px;--line-height-4xl-normal:54px;--line-height-5xl-normal:72px;--line-height-6xl-normal:90px;--line-height-7xl-normal:108px;--line-height-8xl-normal:144px;--line-height-9xl-normal:128px;--line-height-xxs-tight:14px;--line-height-xs-tight:15px;--line-height-sm-tight:17.5px;--line-height-base-tight:20px;--line-height-lg-tight:22.5px;--line-height-xl-tight:25px;--line-height-2xl-tight:30px;--line-height-3xl-tight:37.5px;--line-height-4xl-tight:45px;--line-height-5xl-tight:60px;--line-height-6xl-tight:75px;--line-height-7xl-tight:90px;--line-height-8xl-tight:120px;--line-height-9xl-tight:128px;--line-height-xxs-none:11px;--line-height-xs-none:12px;--line-height-sm-none:14px;--line-height-base-none:16px;--line-height-lg-none:18px;--line-height-xl-none:20px;--line-height-2xl-none:24px;--line-height-3xl-none:30px;--line-height-4xl-none:36px;--line-height-5xl-none:48px;--line-height-6xl-none:60px;--line-height-7xl-none:72px;--line-height-8xl-none:96px;--line-height-9xl-none:128px;--desktop-title-display1-font:var(--family-title);--desktop-title-display1-size:var(--size-8xl);--desktop-title-display1-line-weight:var(--weight-medium);--desktop-title-display1-line-height:var(--line-height-8xl-none);--desktop-title-display2-font:var(--family-title);--desktop-title-display2-size:var(--size-7xl);--desktop-title-display2-line-weight:var(--weight-medium);--desktop-title-display2-line-height:var(--line-height-7xl-none);--desktop-title-h1-font:var(--family-title);--desktop-title-h1-size:var(--size-6xl);--desktop-title-h1-line-weight:var(--weight-medium);--desktop-title-h1-line-height:var(--line-height-6xl-none);--desktop-title-h2-font:var(--family-title);--desktop-title-h2-size:var(--size-5xl);--desktop-title-h2-line-weight:var(--weight-medium);--desktop-title-h2-line-height:var(--line-height-5xl-none);--desktop-title-h3-font:var(--family-title);--desktop-title-h3-size:var(--size-4xl);--desktop-title-h3-line-weight:var(--weight-bold);--desktop-title-h3-line-height:var(--line-height-4xl-none);--desktop-title-h4-font:var(--family-title);--desktop-title-h4-size:var(--size-3xl);--desktop-title-h4-line-weight:var(--weight-bold);--desktop-title-h4-line-height:var(--line-height-3xl-tight);--desktop-title-h5-font:var(--family-title);--desktop-title-h5-size:var(--size-2xl);--desktop-title-h5-line-weight:var(--weight-bold);--desktop-title-h5-line-height:var(--line-height-2xl-tight);--desktop-body-2xl-font:var(--family-body);--desktop-body-2xl-size:var(--size-2xl);--desktop-body-2xl-line-weight:var(--weight-medium);--desktop-body-2xl-line-height:var(--line-height-2xl-normal);--desktop-body-xl-font:var(--family-body);--desktop-body-xl-size:var(--size-xl);--desktop-body-xl-line-weight:var(--weight-medium);--desktop-body-xl-line-height:var(--line-height-xl-normal);--desktop-body-l-font:var(--family-body);--desktop-body-l-size:var(--size-lg);--desktop-body-l-line-weight:var(--weight-medium);--desktop-body-l-line-height:var(--line-height-lg-normal);--desktop-body-m-base-font:var(--family-body);--desktop-body-m-base-size:var(--size-base);--desktop-body-m-base-line-weight:var(--weight-medium);--desktop-body-m-base-line-height:var(--line-height-base-normal);--desktop-subheading-m-base-font:var(--family-body);--desktop-subheading-m-base-size:var(--size-base);--desktop-subheading-m-base-line-weight:var(--weight-medium);--desktop-subheading-m-base-line-height:var(--line-height-base-normal);--desktop-subheading-s-font:var(--family-body);--desktop-subheading-s-size:var(--size-sm);--desktop-subheading-s-line-weight:var(--weight-medium);--desktop-subheading-s-line-height:var(--line-height-sm-normal);--desktop-subheading-xs-font:var(--family-body);--desktop-subheading-xs-size:var(--size-xs);--desktop-subheading-xs-line-weight:var(--weight-medium);--desktop-subheading-xs-line-height:var(--line-height-xs-normal);--desktop-subheading-2xs-font:var(--family-body);--desktop-subheading-2xs-size:var(--size-xxs);--desktop-subheading-2xs-line-weight:var(--weight-medium);--desktop-subheading-2xs-line-height:var(--line-height-xxs-normal);--desktop-label-l-font:var(--family-body);--desktop-label-m-base-font:var(--family-body);--desktop-label-m-base-size:var(--size-sm);--desktop-label-m-base-line-weight:var(--weight-medium);--desktop-label-m-base-line-height:var(--line-height-sm-normal);--desktop-label-l-size:var(--size-base);--desktop-label-l-line-weight:var(--weight-medium);--desktop-label-l-line-height:var(--line-height-base-normal);--desktop-label-s-font:var(--family-body);--desktop-label-s-size:var(--size-xs);--desktop-label-s-line-weight:var(--weight-medium);--desktop-label-s-line-height:var(--line-height-xs-normal);--desktop-label-underline-s-font:var(--family-body);--desktop-label-underline-s-size:var(--size-xs);--desktop-label-underline-s-line-weight:var(--weight-medium);--desktop-label-underline-s-line-height:var(--line-height-xs-normal);--desktop-body-s-font:var(--family-body);--desktop-body-s-size:var(--size-sm);--desktop-body-s-line-weight:var(--weight-medium);--desktop-body-s-line-height:var(--line-height-sm-normal);--desktop-body-xs-font:var(--family-body);--desktop-body-xs-size:var(--size-xs);--desktop-body-xs-line-weight:var(--weight-medium);--desktop-body-xs-line-height:var(--line-height-xs-normal);--desktop-title-h6-font:var(--family-title);--desktop-title-h6-size:var(--size-xl);--desktop-title-h6-line-weight:var(--weight-bold);--desktop-title-h6-line-height:var(--line-height-xl-tight);--desktop-label-underline-m-base-font:var(--family-body);--desktop-label-underline-m-base-size:var(--size-sm);--desktop-label-underline-m-base-line-weight:var(--weight-medium);--desktop-label-underline-m-base-line-height:var(--line-height-sm-normal);--desktop-label-underline-l-font:var(--family-body);--desktop-label-underline-l-size:var(--size-base);--desktop-label-underline-l-line-weight:var(--weight-medium);--desktop-label-underline-l-line-height:var(--line-height-base-normal);--desktop-body-2xs-font:var(--family-body);--desktop-body-2xs-size:var(--size-xxs);--desktop-body-2xs-line-weight:var(--weight-medium);--desktop-body-2xs-line-height:var(--line-height-xxs-normal);--tablet-title-display1-font:var(--family-title);--tablet-title-display1-size:var(--size-7xl);--tablet-title-display1-line-weight:var(--weight-medium);--tablet-title-display1-line-height:var(--line-height-7xl-none);--tablet-title-display2-font:var(--family-title);--tablet-title-display2-size:var(--size-6xl);--tablet-title-display2-line-weight:var(--weight-medium);--tablet-title-display2-line-height:var(--line-height-7xl-none);--tablet-title-h1-font:var(--family-title);--tablet-title-h1-size:var(--size-5xl);--tablet-title-h1-line-weight:var(--weight-medium);--tablet-title-h1-line-height:var(--line-height-6xl-none);--tablet-title-h2-font:var(--family-title);--tablet-title-h2-size:var(--size-4xl);--tablet-title-h2-line-weight:var(--weight-medium);--tablet-title-h2-line-height:var(--line-height-5xl-none);--tablet-title-h3-font:var(--family-title);--tablet-title-h3-size:var(--size-3xl);--tablet-title-h3-line-weight:var(--weight-bold);--tablet-title-h3-line-height:var(--line-height-4xl-none);--tablet-title-h4-font:var(--family-title);--tablet-title-h4-size:var(--size-2xl);--tablet-title-h4-line-weight:var(--weight-bold);--tablet-title-h4-line-height:var(--line-height-3xl-tight);--tablet-title-h5-font:var(--family-title);--tablet-title-h5-size:var(--size-xl);--tablet-title-h5-line-weight:var(--weight-bold);--tablet-title-h5-line-height:var(--line-height-2xl-tight);--tablet-body-2xl-font:var(--family-body);--tablet-body-2xl-size:var(--size-xl);--tablet-body-2xl-line-weight:var(--weight-medium);--tablet-body-2xl-line-height:var(--line-height-2xl-normal);--tablet-body-xl-font:var(--family-body);--tablet-body-xl-size:var(--size-lg);--tablet-body-xl-line-weight:var(--weight-medium);--tablet-body-xl-line-height:var(--line-height-xl-normal);--tablet-body-l-font:var(--family-body);--tablet-body-l-size:var(--size-base);--tablet-body-l-line-weight:var(--weight-medium);--tablet-body-l-line-height:var(--line-height-lg-normal);--tablet-body-m-base-font:var(--family-body);--tablet-body-m-base-size:var(--size-sm);--tablet-body-m-base-line-weight:var(--weight-medium);--tablet-body-m-base-line-height:var(--line-height-base-normal);--tablet-subheading-m-base-font:var(--family-body);--tablet-subheading-m-base-size:var(--size-base);--tablet-subheading-m-base-line-weight:var(--weight-medium);--tablet-subheading-m-base-line-height:var(--line-height-base-normal);--tablet-subheading-s-font:var(--family-body);--tablet-subheading-s-size:var(--size-sm);--tablet-subheading-s-line-weight:var(--weight-medium);--tablet-subheading-s-line-height:var(--line-height-sm-normal);--tablet-subheading-xs-font:var(--family-body);--tablet-subheading-xs-size:var(--size-xs);--tablet-subheading-xs-line-weight:var(--weight-medium);--tablet-subheading-xs-line-height:var(--line-height-xs-normal);--tablet-subheading-2xs-font:var(--family-body);--tablet-subheading-2xs-size:var(--size-xxs);--tablet-subheading-2xs-line-weight:var(--weight-medium);--tablet-subheading-2xs-line-height:var(--line-height-xxs-normal);--tablet-label-l-font:var(--family-body);--tablet-label-m-base-font:var(--family-body);--tablet-label-m-base-size:var(--size-sm);--tablet-label-m-base-line-weight:var(--weight-medium);--tablet-label-m-base-line-height:var(--line-height-sm-normal);--tablet-label-l-size:var(--size-sm);--tablet-label-l-line-weight:var(--weight-medium);--tablet-label-l-line-height:var(--line-height-sm-normal);--tablet-label-s-font:var(--family-body);--tablet-label-s-size:var(--size-xs);--tablet-label-s-line-weight:var(--weight-medium);--tablet-label-s-line-height:var(--line-height-xs-normal);--tablet-label-underline-s-font:var(--family-body);--tablet-label-underline-s-size:var(--size-xs);--tablet-label-underline-s-line-weight:var(--weight-medium);--tablet-label-underline-s-line-height:var(--line-height-xs-normal);--tablet-body-s-font:var(--family-body);--tablet-body-s-size:var(--size-xs);--tablet-body-s-line-weight:var(--weight-medium);--tablet-body-s-line-height:var(--line-height-sm-normal);--tablet-body-xs-font:var(--family-body);--tablet-body-xs-size:var(--size-xs);--tablet-body-xs-line-weight:var(--weight-medium);--tablet-body-xs-line-height:var(--line-height-xs-normal);--tablet-title-h6-font:var(--family-title);--tablet-title-h6-size:var(--size-lg);--tablet-title-h6-line-weight:var(--weight-bold);--tablet-title-h6-line-height:var(--line-height-xl-tight);--tablet-label-underline-m-base-font:var(--family-body);--tablet-label-underline-m-base-size:var(--size-xs);--tablet-label-underline-m-base-line-weight:var(--weight-medium);--tablet-label-underline-m-base-line-height:var(--line-height-xs-normal);--tablet-label-underline-l-font:var(--family-body);--tablet-label-underline-l-size:var(--size-sm);--tablet-label-underline-l-line-weight:var(--weight-medium);--tablet-label-underline-l-line-height:var(--line-height-sm-normal);--tablet-body-2xs-font:var(--family-body);--tablet-body-2xs-size:var(--size-xxs);--tablet-body-2xs-line-weight:var(--weight-medium);--tablet-body-2xs-line-height:var(--line-height-xxs-normal);--mobile-title-display1-font:var(--family-title);--mobile-title-display1-size:var(--size-6xl);--mobile-title-display1-line-weight:var(--weight-medium);--mobile-title-display1-line-height:var(--line-height-8xl-none);--mobile-title-display2-font:var(--family-title);--mobile-title-display2-size:var(--size-5xl);--mobile-title-display2-line-weight:var(--weight-medium);--mobile-title-display2-line-height:var(--line-height-7xl-none);--mobile-title-h1-font:var(--family-title);--mobile-title-h1-size:var(--size-4xl);--mobile-title-h1-line-weight:var(--weight-medium);--mobile-title-h1-line-height:var(--line-height-6xl-none);--mobile-title-h2-font:var(--family-title);--mobile-title-h2-size:var(--size-3xl);--mobile-title-h2-line-weight:var(--weight-medium);--mobile-title-h2-line-height:var(--line-height-5xl-none);--mobile-title-h3-font:var(--family-title);--mobile-title-h3-size:var(--size-2xl);--mobile-title-h3-line-weight:var(--weight-bold);--mobile-title-h3-line-height:var(--line-height-4xl-none);--mobile-title-h4-font:var(--family-title);--mobile-title-h4-size:var(--size-xl);--mobile-title-h4-line-weight:var(--weight-bold);--mobile-title-h4-line-height:var(--line-height-3xl-tight);--mobile-title-h5-font:var(--family-title);--mobile-title-h5-size:var(--size-lg);--mobile-title-h5-line-weight:var(--weight-bold);--mobile-title-h5-line-height:var(--line-height-2xl-tight);--mobile-body-2xl-font:var(--family-body);--mobile-body-2xl-size:var(--size-lg);--mobile-body-2xl-line-weight:var(--weight-medium);--mobile-body-2xl-line-height:var(--line-height-2xl-normal);--mobile-body-xl-font:var(--family-body);--mobile-body-xl-size:var(--size-base);--mobile-body-xl-line-weight:var(--weight-medium);--mobile-body-xl-line-height:var(--line-height-base-normal);--mobile-body-l-font:var(--family-body);--mobile-body-l-size:var(--size-sm);--mobile-body-l-line-weight:var(--weight-medium);--mobile-body-l-line-height:var(--line-height-lg-normal);--mobile-body-m-base-font:var(--family-body);--mobile-body-m-base-size:var(--size-xs);--mobile-body-m-base-line-weight:var(--weight-medium);--mobile-body-m-base-line-height:var(--line-height-base-normal);--mobile-subheading-m-base-font:var(--family-body);--mobile-subheading-m-base-size:var(--size-xs);--mobile-subheading-m-base-line-weight:var(--weight-medium);--mobile-subheading-m-base-line-height:var(--line-height-xs-normal);--mobile-subheading-s-font:var(--family-body);--mobile-subheading-s-size:var(--size-xs);--mobile-subheading-s-line-weight:var(--weight-medium);--mobile-subheading-s-line-height:var(--line-height-xs-normal);--mobile-subheading-xs-font:var(--family-body);--mobile-subheading-xs-size:var(--size-xs);--mobile-subheading-xs-line-weight:var(--weight-medium);--mobile-subheading-xs-line-height:var(--line-height-xs-normal);--mobile-subheading-2xs-font:var(--family-body);--mobile-subheading-2xs-size:var(--size-xxs);--mobile-subheading-2xs-line-weight:var(--weight-medium);--mobile-subheading-2xs-line-height:var(--line-height-xxs-normal);--mobile-label-l-font:var(--family-body);--mobile-label-m-base-font:var(--family-body);--mobile-label-m-base-size:var(--size-xs);--mobile-label-m-base-line-weight:var(--weight-medium);--mobile-label-m-base-line-height:var(--line-height-xs-normal);--mobile-label-l-size:var(--size-xs);--mobile-label-l-line-weight:var(--weight-medium);--mobile-label-l-line-height:var(--line-height-xs-normal);--mobile-label-s-font:var(--family-body);--mobile-label-s-size:var(--size-xxs);--mobile-label-s-line-weight:var(--weight-medium);--mobile-label-s-line-height:var(--line-height-xxs-normal);--mobile-label-underline-s-font:var(--family-body);--mobile-label-underline-s-size:var(--size-xxs);--mobile-label-underline-s-line-weight:var(--weight-medium);--mobile-label-underline-s-line-height:var(--line-height-xxs-normal);--mobile-body-s-font:var(--family-body);--mobile-body-s-size:var(--size-xs);--mobile-body-s-line-weight:var(--weight-medium);--mobile-body-s-line-height:var(--line-height-sm-normal);--mobile-body-xs-font:var(--family-body);--mobile-body-xs-size:var(--size-xs);--mobile-body-xs-line-weight:var(--weight-medium);--mobile-body-xs-line-height:var(--line-height-xs-normal);--mobile-title-h6-font:var(--family-title);--mobile-title-h6-size:var(--size-base);--mobile-title-h6-line-weight:var(--weight-bold);--mobile-title-h6-line-height:var(--line-height-xl-tight);--mobile-label-underline-m-base-font:var(--family-body);--mobile-label-underline-m-base-size:var(--size-xxs);--mobile-label-underline-m-base-line-weight:var(--weight-medium);--mobile-label-underline-m-base-line-height:var(--line-height-xxs-normal);--mobile-label-underline-l-font:var(--family-body);--mobile-label-underline-l-size:var(--size-xs);--mobile-label-underline-l-line-weight:var(--weight-medium);--mobile-label-underline-l-line-height:var(--line-height-xs-normal);--mobile-body-2xs-font:var(--family-body);--mobile-body-2xs-size:var(--size-xxs);--mobile-body-2xs-line-weight:var(--weight-medium);--mobile-body-2xs-line-height:var(--line-height-xxs-normal);--line-height-normal-xxs:16;--line-height-normal-xs:18;--line-height-normal-sm:20;--line-height-normal-base:24;--line-height-normal-lg:27;--line-height-normal-xl:30;--line-height-normal-2xl:36;--line-height-normal-3xl:45;--line-height-normal-4xl:54;--line-height-normal-5xl:72;--line-height-normal-6xl:90;--line-height-normal-7xl:108;--line-height-normal-8xl:128;--line-height-normal-9xl:144;--line-height-tight-xxs:14;--line-height-tight-xs:15;--line-height-tight-sm:17.5;--line-height-tight-base:20;--line-height-tight-lg:22.5;--line-height-tight-xl:25;--line-height-tight-2xl:30;--line-height-tight-3xl:37.5;--line-height-tight-4xl:45;--line-height-tight-5xl:60;--line-height-tight-6xl:75;--line-height-tight-7xl:90;--line-height-tight-8xl:120;--line-height-tight-9xl:128;--line-height-none-xxs:11;--line-height-none-xs:12;--line-height-none-sm:14;--line-height-none-base:16;--line-height-none-lg:18;--line-height-none-xl:20;--line-height-none-2xl:24;--line-height-none-3xl:30;--line-height-none-4xl:36;--line-height-none-5xl:48;--line-height-none-6xl:60;--line-height-none-7xl:72;--line-height-none-8xl:96;--line-height-none-9xl:128;--effect-1-default-xl:0px 2px 2px -1px var(--shadow-black-alpha-darkest), 0px 6px 12px -2px var(--shadow-black-alpha-darkest), 0px 8px 20px 0px var(--shadow-black-alpha-darkest);--effect-1-default-lg:0px 2px 2px -1px var(--shadow-black-alpha-dark), 0px 6px 12px -2px var(--shadow-black-alpha-dark), 0px 8px 20px 0px var(--shadow-black-alpha-dark);--effect-1-default-base:0px 2px 2px -1px var(--shadow-black-alpha-base), 0px 6px 12px -2px var(--shadow-black-alpha-base), 0px 8px 20px 0px var(--shadow-black-alpha-base);--effect-1-default-sm:0px 2px 2px -1px var(--shadow-black-alpha-light), 0px 6px 12px -2px var(--shadow-black-alpha-light), 0px 8px 20px 0px var(--shadow-black-alpha-light);--effect-1-default-xs:0px 2px 2px -1px var(--shadow-black-alpha-lightest), 0px 6px 12px -2px var(--shadow-black-alpha-lightest), 0px 8px 20px 0px var(--shadow-black-alpha-lightest);--effect-1-hover-xl:0px 2px 2px -1px var(--shadow-black-alpha-darkest), 0px 6px 12px -2px var(--shadow-black-alpha-darkest), 0px 8px 20px 0px var(--shadow-black-alpha-darkest);--effect-1-hover-lg:0px 2px 2px -1px var(--shadow-black-alpha-dark), 0px 6px 12px 2px var(--shadow-black-alpha-dark), 0px 8px 20px 4px var(--shadow-black-alpha-dark);--effect-1-hover-base:0px 2px 2px -1px var(--shadow-black-alpha-base), 0px 6px 12px 2px var(--shadow-black-alpha-base), 0px 8px 20px 4px var(--shadow-black-alpha-base);--effect-1-hover-sm:0px 2px 2px -1px var(--shadow-black-alpha-light), 0px 6px 12px 2px var(--shadow-black-alpha-light), 0px 8px 20px 4px var(--shadow-black-alpha-light);--effect-1-hover-xs:0px 2px 2px -1px var(--shadow-black-alpha-lightest), 0px 6px 12px 2px var(--shadow-black-alpha-lightest), 0px 8px 20px 4px var(--shadow-black-alpha-lightest);--effect-2-default-xl:0px 0px 20px 3px var(--shadow-black-alpha-darkest);--effect-2-default-lg:0px 0px 20px 3px var(--shadow-black-alpha-dark);--effect-2-default-base:0px 0px 20px 3px var(--shadow-black-alpha-base);--effect-2-default-sm:0px 0px 20px 3px var(--shadow-black-alpha-light);--effect-2-default-xs:0px 0px 20px 6px var(--shadow-black-alpha-lightest);--effect-2-hover-xl:0px 0px 20px 8px var(--shadow-black-alpha-darkest);--effect-2-hover-lg:0px 0px 20px 8px var(--shadow-black-alpha-dark);--effect-2-hover-base:0px 0px 20px 8px var(--shadow-black-alpha-base);--effect-2-hover-sm:0px 0px 20px 8px var(--shadow-black-alpha-light);--effect-2-hover-xs:0px 0px 20px 8px var(--shadow-black-alpha-lightest);--primary-focus:0px 0px 0px 2px var(--static-light), 0px 0px 0px 4px var(--primary-light);--secondary-focus:0px 0px 0px 2px var(--static-light), 0px 0px 0px 4px var(--states-info-base);--info-focus:0px 0px 0px 2px var(--static-light), 0px 0px 0px 4px var(--states-info-sub-base);--secondary-light-focus:0px 0px 0px 1px var(--static-light), 0px 0px 0px 2px var(--states-info-base);--neutral-focus:0px 0px 0px 2px var(--static-light), 0px 0px 0px 4px var(--background-light);--radius-none:0px;--radius-xxs:2px;--radius-xs:4px;--radius-s:6px;--radius-m-base:8px;--radius-l:12px;--radius-xl:16px;--radius-2xl:20px;--radius-3xl:24px;--radius-4xl:32px;--radius-full:999px;--spacing-none:0px;--spacing-px:1px;--spacing-xxs:2px;--spacing-xs:4px;--spacing-s:6px;--spacing-m-base:8px;--spacing-l:10px;--spacing-xl:12px;--spacing-2xl:14px;--spacing-4xl:16px;--spacing-5xl:20px;--spacing-6xl:24px;--spacing-7xl:28px;--spacing-8xl:32px;--spacing-9xl:36px;--spacing-10xl:40px;--spacing-11xl:44px;--spacing-12xl:48px;--spacing-14xl:56px;--spacing-15xl:60px;--spacing-16xl:64px;--spacing-20xl:80px;--spacing-24xl:96px;--spacing-28xl:112px;--spacing-32xl:128px;--spacing-36xl:144px;--spacing-40xl:160px;--spacing-44xl:176px;--spacing-48xl:192px;--spacing-52xl:208px;--spacing-56xl:224px;--spacing-60xl:240px;--spacing-64xl:256px;--spacing-72xl:288px;--spacing-80xl:320px;--spacing-96xl:384px;--primary-lightest:var(--primary-50);--secondary-lightest:var(--secondary-50);--static-dark:var(--static-black);--static-light:var(--static-white);--secondary-light:var(--secondary-100);--secondary-sub-base:var(--secondary-300);--secondary-base:var(--secondary-500);--secondary-dark:var(--secondary-600);--secondary-darkest:var(--secondary-800);--text-lightest:var(--neutral-50);--text-light:var(--neutral-300);--text-sub-base:var(--neutral-400);--text-base:var(--neutral-500);--text-dark:var(--neutral-600);--text-darkest:var(--neutral-800);--primary-light:var(--primary-100);--primary-sub-base:var(--primary-300);--primary-base:var(--primary-500);--primary-dark:var(--primary-700);--primary-darkest:var(--primary-800);--background-lightest:var(--neutral-50);--background-light:var(--neutral-200);--background-sub-base:var(--neutral-300);--background-base:var(--neutral-500);--background-dark:var(--neutral-600);--background-darkest:var(--neutral-800);--frames-lightest:var(--neutral-0);--border-lightest:var(--neutral-50);--icon-lightest:var(--neutral-50);--icon-light:var(--neutral-200);--icon-sub-base:var(--neutral-300);--icon-base:var(--neutral-500);--icon-dark:var(--neutral-600);--icon-darkest:var(--neutral-800);--border-light:var(--neutral-200);--border-sub-base:var(--neutral-300);--border-base:var(--neutral-500);--border-dark:var(--neutral-600);--border-darkest:var(--neutral-800);--frames-light:var(--neutral-50);--frames-sub-base:var(--neutral-400);--frames-base:var(--neutral-600);--frames-dark:var(--neutral-900);--frames-darkest:var(--neutral-1000);--shadow-black-alpha-lightest:var(--alpha-neutral-neutral-500-2);--shadow-black-alpha-light:var(--alpha-neutral-neutral-500-4);--shadow-black-alpha-base:var(--alpha-neutral-neutral-500-8);--shadow-black-alpha-dark:var(--alpha-neutral-neutral-500-12);--shadow-black-alpha-darkest:var(--alpha-neutral-neutral-500-16);--overlay-black-none:var(--alpha-base-black-none);--overlay-black-xxlightest:var(--alpha-base-black-2);--overlay-black-xlightest:var(--alpha-base-black-8);--overlay-black-lightest:var(--alpha-base-black-16);--overlay-black-light:var(--alpha-base-black-32);--overlay-black-base:var(--alpha-base-black-48);--overlay-black-dark:var(--alpha-base-black-64);--overlay-black-darkest:var(--alpha-base-black-72);--states-info-lightest:var(--blue-50);--overlay-white-none:var(--alpha-base-white-none);--overlay-white-lightest:var(--alpha-base-white-16);--overlay-white-light:var(--alpha-base-white-32);--overlay-white-base:var(--alpha-base-white-48);--overlay-white-dark:var(--alpha-base-white-64);--overlay-white-darkest:var(--alpha-base-white-72);--states-lightest:var(--neutral-50);--states-info-light:var(--blue-100);--states-info-sub-base:var(--blue-200);--states-light:var(--neutral-300);--states-info-base:var(--blue-500);--states-base:var(--secondary-500);--states-info-dark:var(--blue-700);--states-dark:var(--neutral-600);--states-darkest:var(--neutral-800);--states-info-darkest:var(--blue-800);--states-success-lightest:var(--green-50);--states-success-light:var(--green-100);--states-success-sub-base:var(--green-200);--states-success-base:var(--green-500);--states-success-dark:var(--green-700);--states-danger-lightest:var(--red-50);--states-success-darkest:var(--green-800);--states-danger-light:var(--red-100);--states-danger-sub-base:var(--red-200);--states-danger-base:var(--red-500);--states-danger-dark:var(--red-700);--states-danger-darkest:var(--red-800);--states-warning-lightest:var(--yellow-50);--states-warning-light:var(--yellow-100);--states-warning-sub-base:var(--yellow-200);--states-warning-base:var(--yellow-500);--states-warning-dark:var(--yellow-700);--states-verified-lightest:var(--blue-50);--states-warning-darkest:var(--yellow-800);--states-verified-light:var(--blue-100);--states-verified-sub-base:var(--blue-200);--states-verified-base:var(--blue-500);--states-verified-dark:var(--blue-700);--states-verified-darkest:var(--blue-800);--shadow-blue-alpha-lightest:var(--alpha-blue-blue-500-2);--shadow-blue-alpha-light:var(--alpha-blue-blue-500-4);--shadow-blue-alpha-base:var(--alpha-blue-blue-500-8);--shadow-blue-alpha-dark:var(--alpha-blue-blue-500-12);--shadow-blue-alpha-darkest:var(--alpha-blue-blue-500-16);--color:#ffffff;--skeleton-skeleton-gradient-none:#f9fafc00;--skeleton-skeleton-gradient:#f9fafc;--shadow-primary-alpha-lightest:var(--alpha-brand-primary-500-2);--shadow-primary-alpha-light:var(--alpha-brand-primary-500-4);--shadow-primary-alpha-base:var(--alpha-brand-primary-500-8);--shadow-primary-alpha-dark:var(--alpha-brand-primary-500-16);--shadow-primary-alpha-darkest:var(--alpha-brand-primary-500-24);--button-sizes-large-radius:var(--radius-m-base);--button-sizes-large-gap:var(--spacing-m-base);--button-sizes-large-h-padding:var(--spacing-2xl);--button-sizes-large-v-padding:var(--spacing-xl);--button-sizes-large-full-padding:var(--spacing-xl);--button-sizes-large-icon-size:var(--spacing-6xl);--button-sizes-base-radius:var(--radius-m-base);--button-sizes-base-gap:var(--spacing-s);--button-sizes-base-h-padding:var(--spacing-l);--button-sizes-base-v-padding:var(--spacing-m-base);--button-sizes-base-full-padding:var(--spacing-m-base);--button-sizes-base-icon-size:var(--spacing-6xl);--button-sizes-small-radius:var(--radius-m-base);--button-sizes-small-gap:var(--spacing-xs);--button-sizes-small-h-padding:var(--spacing-m-base);--button-sizes-small-v-padding:var(--spacing-s);--button-sizes-small-full-padding:var(--spacing-s);--button-sizes-small-icon-size:var(--spacing-5xl);--link-button-sizes-large-radius:var(--radius-none);--link-button-sizes-large-gap:var(--spacing-xs);--link-button-sizes-large-h-padding:var(--spacing-none);--link-button-sizes-large-v-padding:var(--spacing-none);--link-button-sizes-large-full-padding:var(--spacing-s);--link-button-sizes-large-icon-size:var(--spacing-6xl);--link-button-sizes-base-radius:var(--radius-none);--link-button-sizes-base-gap:var(--spacing-xs);--link-button-sizes-base-h-padding:var(--spacing-none);--link-button-sizes-base-v-padding:var(--spacing-none);--link-button-sizes-base-full-padding:var(--spacing-s);--link-button-sizes-base-icon-size:var(--spacing-5xl);--link-button-sizes-small-radius:var(--radius-none);--link-button-sizes-small-gap:var(--spacing-xs);--link-button-sizes-small-h-padding:var(--spacing-none);--link-button-sizes-small-v-padding:var(--spacing-none);--link-button-sizes-small-full-padding:var(--spacing-s);--link-button-sizes-small-icon-size:var(--spacing-4xl);--fab-button-sizes-large-radius:var(--radius-s);--fab-button-sizes-large-gap:var(--spacing-xs);--fab-button-sizes-large-h-padding:var(--spacing-2xl);--fab-button-sizes-large-v-padding:var(--spacing-2xl);--fab-button-sizes-large-full-padding:var(--spacing-xl);--fab-button-sizes-large-icon-size:var(--spacing-7xl);--fab-button-sizes-base-radius:var(--radius-s);--fab-button-sizes-base-gap:var(--spacing-xs);--fab-button-sizes-base-h-padding:var(--spacing-xl);--fab-button-sizes-base-v-padding:var(--spacing-xl);--fab-button-sizes-base-full-padding:var(--spacing-xl);--fab-button-sizes-base-icon-size:var(--spacing-6xl);--fab-button-sizes-small-radius:var(--radius-s);--fab-button-sizes-small-gap:var(--spacing-xs);--fab-button-sizes-small-h-padding:var(--spacing-l);--fab-button-sizes-small-v-padding:var(--spacing-l);--fab-button-sizes-small-full-padding:var(--spacing-xl);--fab-button-sizes-small-icon-size:var(--spacing-5xl);--segmented-button-sizes-large-radius:var(--radius-none);--segmented-button-sizes-large-gap:var(--spacing-xs);--segmented-button-sizes-large-h-padding:var(--spacing-4xl);--segmented-button-sizes-large-v-padding:var(--spacing-m-base);--segmented-button-sizes-large-full-padding:var(--spacing-m-base);--segmented-button-sizes-large-icon-size:var(--spacing-6xl);--segmented-button-sizes-base-radius:var(--radius-none);--segmented-button-sizes-base-gap:var(--spacing-xs);--segmented-button-sizes-base-h-padding:var(--spacing-xl);--segmented-button-sizes-base-v-padding:var(--spacing-s);--segmented-button-sizes-base-full-padding:var(--spacing-s);--segmented-button-sizes-base-icon-size:var(--spacing-6xl);--segmented-button-sizes-small-radius:var(--radius-none);--segmented-button-sizes-small-gap:var(--spacing-xs);--segmented-button-sizes-small-h-padding:var(--spacing-m-base);--segmented-button-sizes-small-v-padding:var(--spacing-xs);--segmented-button-sizes-small-full-padding:var(--spacing-xs);--segmented-button-sizes-small-icon-size:var(--spacing-6xl);--card-body-relaxed-h-padding:var(--spacing-4xl);--card-body-relaxed-v-padding:var(--spacing-4xl);--card-body-relaxed-gap:var(--spacing-m-base);--card-header-base-h-padding:var(--spacing-4xl);--card-header-base-v-padding:var(--spacing-xl);--card-header-base-more-icon-size:var(--spacing-10xl);--card-header-base-gap:var(--spacing-m-base);--card-header-small-h-padding:var(--spacing-4xl);--card-header-small-v-padding:var(--spacing-l);--card-header-small-more-icon-size:var(--spacing-10xl);--card-header-small-gap:var(--spacing-s);--card-body-base-h-padding:var(--spacing-4xl);--card-body-base-v-padding:var(--spacing-xl);--card-body-base-gap:var(--spacing-s);--card-body-tight-h-padding:var(--spacing-4xl);--card-body-tight-v-padding:var(--spacing-m-base);--card-body-tight-gap:var(--spacing-xs);--card-image-full-screen-h-padding:var(--spacing-none);--card-image-full-screen-v-padding:var(--spacing-none);--card-image-windowed-h-padding:var(--spacing-4xl);--card-image-windowed-v-padding:var(--spacing-4xl);--card-footer-base-h-padding:var(--spacing-4xl);--card-footer-base-v-padding:var(--spacing-4xl);--card-footer-base-gap:var(--spacing-m-base);--chips-large-radius:var(--radius-s);--chips-large-h-padding:var(--spacing-m-base);--chips-large-v-padding:var(--spacing-xs);--chips-large-gap:var(--spacing-s);--chips-large-icon-size:var(--spacing-5xl);--chips-base-radius:var(--radius-s);--chips-base-h-padding:var(--spacing-s);--chips-base-v-padding:var(--spacing-xxs);--chips-base-gap:var(--spacing-xs);--chips-base-icon-size:var(--spacing-4xl);--chips-small-radius:var(--radius-s);--chips-small-h-padding:var(--spacing-xs);--chips-small-v-padding:var(--spacing-px);--chips-small-gap:var(--spacing-xxs);--chips-small-icon-size:var(--spacing-2xl);--input-external-label-large-text-area-radius:var(--radius-m-base);--input-external-label-large-icon-size:var(--spacing-6xl);--input-external-label-large-container-gap:var(--spacing-xs);--input-external-label-large-text-field-gap:var(--spacing-xs);--input-external-label-large-label-h-padding:var(--spacing-m-base);--input-external-label-large-label-v-padding:var(--spacing-none);--input-external-label-large-input-h-padding:var(--spacing-xl);--input-external-label-large-input-v-padding:var(--spacing-xl);--input-external-label-large-input-gap:var(--spacing-xs);--input-external-label-base-text-area-radius:var(--radius-m-base);--input-external-label-base-icon-size:var(--spacing-5xl);--input-external-label-base-div-gap:var(--spacing-xs);--input-external-label-base-text-field-gap:var(--spacing-xs);--input-external-label-base-label-h-padding:var(--spacing-m-base);--input-external-label-base-label-v-padding:var(--spacing-none);--input-external-label-base-input-h-padding:var(--spacing-l);--input-external-label-base-input-v-padding:var(--spacing-l);--input-external-label-base-input-gap:var(--spacing-xs);--input-external-label-small-text-area-radius:var(--radius-m-base);--input-external-label-small-icon-size:var(--spacing-5xl);--input-external-label-small-container-gap:var(--spacing-xs);--input-external-label-small-text-field-gap:var(--spacing-xxs);--input-external-label-small-label-h-padding:var(--spacing-s);--input-external-label-small-label-v-padding:var(--spacing-none);--input-external-label-small-input-h-padding:var(--spacing-m-base);--input-external-label-small-input-v-padding:var(--spacing-s);--input-external-label-small-input-gap:var(--spacing-xs);--basic-switcher-container-radius:var(--radius-none);--basic-switcher-container-h-padding:var(--spacing-none);--basic-switcher-container-v-padding:var(--spacing-none);--basic-switcher-container-gap:var(--spacing-s);--basic-switcher-content-gap:var(--spacing-m-base);--basic-switcher-text-gap:var(--spacing-none);--basic-switcher-button-size:var(--spacing-6xl);--card-switcher-container-radius:var(--radius-m-base);--card-switcher-container-h-padding:var(--spacing-xl);--card-switcher-container-v-padding:var(--spacing-xl);--card-switcher-container-gap:var(--spacing-m-base);--card-switcher-content-gap:var(--spacing-m-base);--card-switcher-text-gap:var(--spacing-none);--card-switcher-button-size:var(--spacing-6xl);--avatar-switcher-container-radius:var(--radius-m-base);--avatar-switcher-container-h-padding:var(--spacing-xl);--avatar-switcher-container-v-padding:var(--spacing-xl);--avatar-switcher-container-gap:var(--spacing-xl);--avatar-switcher-v-content-gap:var(--spacing-m-base);--avatar-switcher-h-content-gap:var(--spacing-m-base);--avatar-switcher-text-gap:var(--spacing-none);--avatar-switcher-button-size:var(--spacing-6xl);--avatar-switcher-avatar-size:var(--spacing-12xl);--breadcrumb-symbol-size:var(--spacing-5xl);--breadcrumb-container-gap:var(--spacing-xs);--basic-item-h-padding:var(--spacing-none);--backgrounded-item-h-padding:var(--spacing-s);--backgrounded-item-v-padding:var(--spacing-xxs);--backgrounded-item-gap:var(--spacing-xxs);--backgrounded-item-icon-size:var(--spacing-4xl);--basic-item-v-padding:var(--spacing-none);--basic-item-gap:var(--spacing-xxs);--basic-item-icon-size:var(--spacing-4xl);--pagination-cell-small-cell-radius:var(--radius-m-base);--pagination-cell-small-cell-size:var(--spacing-6xl);--pagination-cell-base-cell-radius:var(--radius-m-base);--pagination-cell-base-cell-size:var(--spacing-8xl);--pagination-cell-large-cell-radius:var(--radius-m-base);--pagination-cell-large-cell-size:var(--spacing-10xl);--pagination-base-tag-gap:var(--spacing-none);--pagination-base-right-gap:var(--spacing-xl);--pagination-base-container-gap:var(--spacing-xl);--pagination-base-container-v-padding:var(--spacing-xxs);--pagination-base-arrow-gap:var(--spacing-xs);--pagination-base-gap:var(--spacing-s);--slider-h-padding:var(--spacing-none);--slider-v-padding:var(--spacing-none);--slider-gap:var(--spacing-xxs);--slider-content-gap:var(--spacing-none);--slider-slider-h-padding:var(--spacing-m-base);--slider-slider-v-padding:var(--spacing-s);--badge-small-basic-radius:var(--radius-xs);--badge-small-basic-gap:0;--badge-small-basic-left-padding:var(--spacing-s);--badge-small-basic-right-padding:var(--spacing-s);--badge-small-basic-v-padding:var(--spacing-none);--badge-small-basic-icon-size:0;--badge-small-left-icon-radius:var(--radius-xs);--badge-small-left-icon-gap:var(--spacing-none);--badge-small-left-icon-left-padding:var(--spacing-xs);--badge-small-left-icon-right-padding:var(--spacing-s);--badge-small-left-icon-v-padding:var(--spacing-none);--badge-small-left-icon-icon-size:var(--spacing-2xl);--badge-small-right-icon-radius:var(--radius-xs);--badge-small-right-icon-gap:var(--spacing-none);--badge-small-right-icon-left-padding:var(--spacing-s);--badge-small-right-icon-right-padding:var(--spacing-xs);--badge-small-right-icon-v-padding:var(--spacing-none);--badge-small-right-icon-icon-size:var(--spacing-2xl);--badge-base-basic-radius:var(--radius-s);--badge-base-basic-gap:0;--badge-base-basic-left-padding:var(--spacing-m-base);--badge-base-basic-right-padding:var(--spacing-m-base);--badge-base-basic-v-padding:var(--spacing-px);--badge-base-basic-icon-size:0;--badge-base-left-icon-radius:var(--radius-s);--badge-base-left-icon-gap:var(--spacing-xxs);--badge-base-left-icon-left-padding:var(--spacing-xs);--badge-base-left-icon-right-padding:var(--spacing-m-base);--badge-base-left-icon-v-padding:var(--spacing-px);--badge-base-left-icon-icon-size:var(--spacing-4xl);--badge-base-right-icon-radius:var(--radius-s);--badge-base-right-icon-gap:var(--spacing-xxs);--badge-base-right-icon-left-padding:var(--spacing-m-base);--badge-base-right-icon-right-padding:var(--spacing-xs);--badge-base-right-icon-v-padding:var(--spacing-px);--badge-base-right-icon-icon-size:var(--spacing-4xl);--badge-large-basic-radius:var(--radius-m-base);--badge-large-basic-gap:0;--badge-large-basic-left-padding:var(--spacing-l);--badge-large-basic-right-padding:var(--spacing-l);--badge-large-basic-v-padding:var(--spacing-xxs);--badge-large-basic-icon-size:0;--badge-large-left-icon-radius:var(--radius-m-base);--badge-large-left-icon-gap:var(--spacing-xs);--badge-large-left-icon-left-padding:var(--spacing-s);--badge-large-left-icon-right-padding:var(--spacing-l);--badge-large-left-icon-v-padding:var(--spacing-xxs);--badge-large-left-icon-icon-size:var(--spacing-4xl);--badge-large-right-icon-radius:var(--radius-m-base);--badge-large-right-icon-gap:var(--spacing-xs);--badge-large-right-icon-left-padding:var(--spacing-l);--badge-large-right-icon-right-padding:var(--spacing-s);--badge-large-right-icon-v-padding:var(--spacing-xxs);--badge-large-right-icon-icon-size:var(--spacing-4xl);--datepicker-items-base-radius:var(--radius-m-base);--datepicker-items-base-size:var(--spacing-10xl);--datepicker-items-small-radius:var(--radius-m-base);--datepicker-items-small-size:var(--spacing-8xl);--datepicker-footer-base-h-padding:var(--spacing-4xl);--datepicker-footer-base-v-padding:var(--spacing-4xl);--datepicker-footer-base-gap:var(--spacing-m-base);--datepicker-header-base-h-padding:var(--spacing-4xl);--datepicker-header-base-top-padding:var(--spacing-m-base);--datepicker-header-base-bottom-padding:var(--spacing-xs);--datepicker-header-base-navigation-gap:var(--spacing-none);--datepicker-header-base-date-text-gap:var(--spacing-xs);--datepicker-header-base-date-text-h-padding:var(--spacing-m-base);--datepicker-body-base-day-h-padding:var(--spacing-4xl);--datepicker-body-base-day-v-padding:var(--spacing-xl);--datepicker-body-base-day-row-group-gap:var(--spacing-xs);--datepicker-body-base-day-row-gap:var(--spacing-xs);--datepicker-body-base-month-year-h-padding:var(--spacing-4xl);--datepicker-body-base-month-year-v-padding:var(--spacing-7xl);--datepicker-body-base-month-year-row-gap:var(--spacing-xs);--dialog-header-base-h-padding:var(--spacing-4xl);--dialog-header-base-v-padding:var(--spacing-xl);--dialog-header-base-sign-size:var(--spacing-12xl);--dialog-header-base-close-icon-size:var(--spacing-6xl);--dialog-header-base-gap:var(--spacing-m-base);--dialog-header-small-h-padding:var(--spacing-4xl);--dialog-header-small-v-padding:var(--spacing-l);--dialog-header-small-sign-size:var(--spacing-11xl);--dialog-header-small-close-icon-size:var(--spacing-6xl);--dialog-header-small-gap:var(--spacing-s);--dialog-body-base-h-padding:var(--spacing-4xl);--dialog-body-base-v-padding:var(--spacing-4xl);--dialog-body-base-gap:var(--spacing-m-base);--dialog-footer-base-h-padding:var(--spacing-4xl);--dialog-footer-base-v-padding:var(--spacing-4xl);--dialog-footer-base-gap:var(--spacing-m-base);--dialog-extra-base-h-padding:var(--spacing-4xl);--dialog-extra-base-v-padding:var(--spacing-xl);--dialog-extra-base-gap:var(--spacing-m-base);--dropdown-items-basic-base-radius:var(--radius-m-base);--dropdown-items-basic-base-h-padding:var(--spacing-m-base);--dropdown-items-basic-base-v-padding:var(--spacing-m-base);--dropdown-items-basic-base-gap:var(--spacing-xs);--dropdown-items-basic-base-icon-size:var(--spacing-5xl);--dropdown-items-basic-large-radius:var(--radius-m-base);--dropdown-items-basic-large-h-padding:var(--spacing-m-base);--dropdown-items-basic-large-v-padding:var(--spacing-m-base);--dropdown-items-basic-large-gap:var(--spacing-xs);--dropdown-items-basic-large-icon-size:var(--spacing-5xl);--dropdown-items-basic-small-radius:var(--radius-m-base);--dropdown-items-basic-small-h-padding:var(--spacing-m-base);--dropdown-items-basic-small-v-padding:var(--spacing-s);--dropdown-items-basic-small-gap:var(--spacing-xs);--dropdown-items-basic-small-icon-size:var(--spacing-5xl);--dropdown-items-avatar-base-radius:var(--radius-m-base);--dropdown-items-avatar-base-h-padding:var(--spacing-m-base);--dropdown-items-avatar-base-v-padding:var(--spacing-m-base);--dropdown-items-avatar-base-gap:var(--spacing-xs);--dropdown-items-avatar-base-icon-size:var(--spacing-5xl);--dropdown-items-avatar-large-radius:var(--radius-m-base);--dropdown-items-avatar-large-h-padding:var(--spacing-m-base);--dropdown-items-avatar-large-v-padding:var(--spacing-m-base);--dropdown-items-avatar-large-gap:var(--spacing-xs);--dropdown-items-avatar-large-icon-size:var(--spacing-5xl);--dropdown-items-avatar-small-radius:var(--radius-m-base);--dropdown-items-avatar-small-h-padding:var(--spacing-m-base);--dropdown-items-avatar-small-v-padding:var(--spacing-s);--dropdown-items-avatar-small-gap:var(--spacing-xs);--dropdown-items-avatar-small-icon-size:var(--spacing-5xl);--dropdown-items-radio-checkbox-base-radius:var(--radius-m-base);--dropdown-items-radio-checkbox-base-h-padding:var(--spacing-m-base);--dropdown-items-radio-checkbox-base-v-padding:var(--spacing-s);--dropdown-items-radio-checkbox-base-container-gap:var(--spacing-m-base);--dropdown-items-radio-checkbox-large-radius:var(--radius-m-base);--dropdown-items-radio-checkbox-large-h-padding:var(--spacing-m-base);--dropdown-items-radio-checkbox-large-v-padding:var(--spacing-m-base);--dropdown-items-radio-checkbox-large-container-gap:var(--spacing-l);--dropdown-items-radio-checkbox-small-radius:var(--radius-m-base);--dropdown-items-radio-checkbox-small-h-padding:var(--spacing-m-base);--dropdown-items-radio-checkbox-small-v-padding:var(--spacing-xs);--dropdown-items-radio-checkbox-small-container-gap:var(--spacing-s);--editor-gap:var(--spacing-m-base);--editor-header-compact-h-padding:var(--spacing-4xl);--editor-body-h-padding:var(--spacing-4xl);--editor-body-v-padding:var(--spacing-4xl);--editor-body-gap:var(--spacing-xs);--editor-body-content-gap:var(--spacing-m-base);--editor-header-normal-h-padding:var(--spacing-4xl);--editor-header-normal-v-padding:var(--spacing-xl);--editor-header-normal-gap:var(--spacing-xl);--editor-header-normal-operation-gap:var(--spacing-m-base);--editor-header-compact-v-padding:var(--spacing-xl);--editor-header-compact-gap:var(--spacing-xl);--editor-header-compact-operation-gap:var(--spacing-m-base);--table-items-head-h-padding:0;--table-items-head-v-padding:0;--table-items-body-base-text-h-padding:var(--spacing-6xl);--table-items-body-base-text-text-h-padding:var(--spacing-6xl);--table-items-body-base-nontext-h-padding:var(--spacing-4xl);--table-items-body-base-nontext-v-padding:var(--spacing-4xl);--table-items-body-base-nontext-gap:var(--spacing-m-base);--table-items-body-base-text-text-v-padding:var(--spacing-s);--table-items-body-base-text-text-content-gap:var(--spacing-xxs);--table-items-body-base-text-text-gap:var(--spacing-m-base);--table-items-body-base-text-v-padding:var(--spacing-4xl);--table-items-body-base-text-gap:var(--spacing-m-base);--table-items-body-small-text-h-padding:var(--spacing-6xl);--table-items-body-small-text-text-h-padding:var(--spacing-6xl);--table-items-body-small-nontext-h-padding:var(--spacing-xl);--table-items-body-small-nontext-v-padding:var(--spacing-m-base);--table-items-body-small-nontext-gap:var(--spacing-m-base);--table-items-body-small-text-text-v-padding:var(--spacing-xxs);--table-items-body-small-text-text-content-gap:var(--spacing-xxs);--table-items-body-small-text-text-gap:var(--spacing-m-base);--table-items-body-small-text-v-padding:var(--spacing-m-base);--table-items-body-small-text-gap:var(--spacing-m-base);--table-items-head-base-text-h-padding:var(--spacing-xl);--table-items-head-base-text-text-h-padding:var(--spacing-xl);--table-items-head-base-nontext-h-padding:var(--spacing-4xl);--table-items-head-base-nontext-v-padding:var(--spacing-4xl);--table-items-head-base-nontext-gap:var(--spacing-m-base);--table-items-head-base-text-text-v-padding:var(--spacing-s);--table-items-head-base-text-text-content-h-padding:var(--spacing-xl);--table-items-head-base-text-text-content-gap:var(--spacing-xxs);--table-items-head-base-text-text-gap:var(--spacing-m-base);--table-items-head-base-text-v-padding:var(--spacing-4xl);--table-items-head-base-text-content-h-padding:var(--spacing-xl);--table-items-head-base-text-gap:var(--spacing-m-base);--table-items-head-small-text-h-padding:var(--spacing-xl);--table-items-head-small-text-text-h-padding:var(--spacing-xl);--table-items-head-small-nontext-h-padding:var(--spacing-xl);--table-items-head-small-nontext-v-padding:var(--spacing-xl);--table-items-head-small-nontext-gap:var(--spacing-m-base);--table-items-head-small-text-text-v-padding:var(--spacing-xs);--table-items-head-small-text-text-content-h-padding:var(--spacing-xl);--table-items-head-small-text-text-content-gap:var(--spacing-none);--table-items-head-small-text-text-gap:var(--spacing-m-base);--table-items-head-small-text-v-padding:var(--spacing-m-base);--table-items-head-small-text-gap:var(--spacing-m-base);--table-items-head-xsmall-text-h-padding:var(--spacing-xs);--table-items-head-xsmall-text-v-padding:var(--spacing-xs);--table-items-body-xsmall-text-v-padding:var(--spacing-xs);--table-items-body-xsmall-text-h-padding:var(--spacing-xs);--tabs-items-large-h-padding:var(--spacing-4xl);--tabs-items-large-v-padding:var(--spacing-2xl);--tabs-items-large-basic-active-gap:var(--spacing-xl);--tabs-items-large-basic-vertical-active-gap:var(--spacing-2xl);--tabs-items-large-tab-content-gap:var(--spacing-m-base);--tabs-items-large-tab-label-gap:var(--spacing-xs);--tabs-items-large-icon-size:var(--spacing-5xl);--tabs-items-base-h-padding:var(--spacing-2xl);--tabs-items-base-v-padding:var(--spacing-l);--tabs-items-base-basic-active-gap:var(--spacing-m-base);--tabs-items-base-basic-vertical-active-gap:var(--spacing-xl);--tabs-items-base-tab-content-gap:var(--spacing-m-base);--tabs-items-base-tab-label-gap:var(--spacing-xs);--tabs-items-base-icon-size:var(--spacing-5xl);--tabs-items-small-h-padding:var(--spacing-xl);--tabs-items-small-v-padding:var(--spacing-s);--tabs-items-small-basic-active-gap:var(--spacing-xs);--tabs-items-small-basic-vertical-active-gap:var(--spacing-l);--tabs-items-small-tab-content-gap:var(--spacing-s);--tabs-items-small-tab-label-gap:var(--spacing-xs);--tabs-items-small-icon-size:var(--spacing-5xl);--tabs-items-xsmall-h-padding:var(--spacing-l);--tabs-items-xsmall-v-padding:var(--spacing-xs);--tabs-items-xsmall-basic-active-gap:var(--spacing-xxs);--tabs-items-xsmall-basic-vertical-active-gap:var(--spacing-m-base);--tabs-items-xsmall-tab-content-gap:var(--spacing-xs);--tabs-items-xsmall-tab-label-gap:var(--spacing-xxs);--tabs-items-xsmall-icon-size:var(--spacing-4xl);--tabs-items-xxsmall-h-padding:var(--spacing-m-base);--tabs-items-xxsmall-v-padding:var(--spacing-xxs);--tabs-items-xxsmall-basic-active-gap:var(--spacing-px);--tabs-items-xxsmall-basic-vertical-active-gap:var(--spacing-s);--tabs-items-xxsmall-tab-content-gap:var(--spacing-xxs);--tabs-items-xxsmall-tab-label-gap:var(--spacing-px);--tabs-items-xxsmall-icon-size:var(--spacing-3xl);--timeline-h-padding:var(--spacing-m-base);--timeline-v-padding:var(--spacing-m-base);--timeline-gap:var(--spacing-m-base);--timeline-content-gap:var(--spacing-m-base);--timeline-text-gap:var(--spacing-m-base);--upload-basic-h-padding:var(--spacing-6xl);--upload-basic-v-padding:var(--spacing-8xl);--upload-basic-gap:var(--spacing-6xl);--upload-basic-content-gap:var(--spacing-4xl);--upload-basic-text-gap:var(--spacing-xs);--upload-card-h-padding:var(--spacing-xl);--upload-card-v-padding:var(--spacing-xl);--upload-card-gap:var(--spacing-xl);--upload-card-content-gap:var(--spacing-xxs);--upload-compact-h-padding:var(--spacing-xl);--upload-compact-v-padding:var(--spacing-xl);--upload-compact-gap:var(--spacing-m-base);--upload-compact-content-gap:var(--spacing-4xl);--upload-compact-text-gap:var(--spacing-none);--calendar-header-h-padding:var(--spacing-4xl);--calendar-header-v-padding:var(--spacing-4xl);--calendar-header-selector-gap:var(--spacing-xl);--calendar-header-actions-gap:var(--spacing-xl);--calendar-activity-h-padding:var(--spacing-xxs);--calendar-activity-v-padding:var(--spacing-xxs);--calendar-activity-gap:var(--spacing-xs);--calendar-month-cell-h-padding:var(--spacing-m-base);--calendar-month-cell-v-padding:var(--spacing-m-base);--calendar-week-cell-h-padding:var(--spacing-m-base);--calendar-week-cell-v-padding:var(--spacing-m-base);--calendar-week-cell-gap:var(--spacing-xs);--toast-base-radius:var(--radius-m-base);--toast-base-h-padding:var(--spacing-xl);--toast-base-v-padding:var(--spacing-xl);--toast-base-gap:var(--spacing-m-base);--toast-base-content-gap:var(--spacing-m-base);--toast-sign-base-size:var(--spacing-5xl);--toast-sign-base-full-padding:var(--spacing-s);--toast-sign-large-size:var(--spacing-6xl);--toast-sign-large-full-padding:var(--spacing-m-base);--toast-sign-small-size:var(--spacing-4xl);--toast-sign-small-full-padding:var(--spacing-xs);--gradient-info-gradient-tozero:linear-gradient(270deg, var(--background-lightest) 89%, var(--states-info-sub-base) 100%);--gradient-success-gradient-tozero:linear-gradient(270deg, var(--background-lightest) 89%, var(--states-success-sub-base) 100%);--gradient-danger-gradient-tozero:linear-gradient(270deg, var(--background-lightest) 89%, var(--states-danger-sub-base) 100%);--gradient-warning-gradient-tozero:linear-gradient(270deg, var(--background-lightest) 89%, var(--states-warning-sub-base) 100%);--gradient-gradient-gray:linear-gradient(270deg, var(--background-lightest) 89%, var(--background-sub-base) 100%);--sign-2xlarge-radius:var(--radius-xl);--sign-2xlarge-icon-size:var(--spacing-8xl);--sign-2xlarge-h-padding:var(--spacing-4xl);--sign-2xlarge-v-padding:var(--spacing-4xl);--sign-xlarge-radius:var(--radius-l);--sign-xlarge-icon-size:var(--spacing-6xl);--sign-xlarge-h-padding:var(--spacing-xl);--sign-xlarge-v-padding:var(--spacing-xl);--sign-large-radius:var(--radius-l);--sign-large-icon-size:var(--spacing-6xl);--sign-large-h-padding:var(--spacing-m-base);--sign-large-v-padding:var(--spacing-m-base);--sign-base-radius:var(--radius-m-base);--sign-base-icon-size:var(--spacing-5xl);--sign-base-h-padding:var(--spacing-s);--sign-base-v-padding:var(--spacing-s);--sign-small-radius:var(--radius-s);--sign-small-icon-size:var(--spacing-4xl);--sign-small-h-padding:var(--spacing-xs);--sign-small-v-padding:var(--spacing-xs);--sign-xsmall-radius:var(--radius-s);--sign-xsmall-icon-size:var(--spacing-xl);--sign-xsmall-h-padding:var(--spacing-xs);--sign-xsmall-v-padding:var(--spacing-xs);--sign-xxsmall-radius:var(--radius-s);--sign-xxsmall-icon-size:var(--spacing-l);--sign-xxsmall-h-padding:var(--spacing-xxs);--sign-xxsmall-v-padding:var(--spacing-xxs);--notification-header-h-padding:var(--spacing-4xl);--notification-header-v-padding:var(--spacing-4xl);--notification-header-gap:var(--spacing-m-base);--notification-header-button-gap:var(--spacing-xs);--notification-body-h-padding:var(--spacing-4xl);--notification-footer-h-padding:var(--spacing-4xl);--notification-footer-v-padding:var(--spacing-xl);--notification-footer-gap:var(--spacing-m-base);--notification-body-v-padding:var(--spacing-xl);--notification-body-gap:var(--spacing-m-base);--notification-body-content-gap:var(--spacing-m-base);--notification-body-text-gap:var(--spacing-xs);--notification-body-header-gap:var(--spacing-m-base);--notification-body-head-gap:var(--spacing-none);--tooltips-base-max-width:var(--spacing-52xl);--tooltips-base-radius:var(--radius-m-base);--tooltips-base-h-padding:var(--spacing-xl);--tooltips-base-v-padding:var(--spacing-m-base);--tooltips-base-gap:var(--spacing-m-base);--tooltips-base-content-gap:var(--spacing-none);--accordion-base-radius:var(--radius-m-base);--accordion-base-h-padding:var(--spacing-xl);--accordion-base-v-padding:var(--spacing-xl);--accordion-base-gap:var(--spacing-s);--accordion-base-content-gap:var(--spacing-s);--accordion-large-radius:var(--radius-m-base);--accordion-large-h-padding:var(--spacing-xl);--accordion-large-v-padding:var(--spacing-xl);--accordion-large-gap:var(--spacing-m-base);--accordion-large-content-gap:var(--spacing-m-base);--spinner-size-xlarge:var(--spacing-12xl);--spinner-size-large:var(--spacing-10xl);--spinner-size-base:var(--spacing-8xl);--spinner-size-small:var(--spacing-6xl);--spinner-default-gap:var(--spacing-m-base);--spinner-centered-gap:var(--spacing-xs);--stepper-items-content-gap:var(--spacing-xs);--stepper-items-rail-gap:var(--spacing-xs);--stepper-segment-gap:var(--spacing-xs);--rating-basic-rate-gap:var(--spacing-none);--rating-basic-rate-content-gap:var(--spacing-xs);--rating-basic-rate-content-h-padding:var(--spacing-xxs);--rating-basic-rate-content-v-padding:var(--spacing-xxs);--rating-emoji-rate-gap:var(--spacing-m-base);--rating-bar-rate-compact-gap:var(--spacing-none);--rating-bar-rate-default-gap:var(--spacing-xs);--rating-number-rate-cells-h:var(--spacing-8xl);--rating-number-rate-cells-w:var(--spacing-8xl);--rating-number-rate-compact-gap:var(--spacing-none);--rating-number-rate-default-gap:var(--spacing-xs);--breakpoint-xl:1440px;--breakpoint-lg:1248px;--breakpoint-md:905px;--breakpoint-sm:600px;--breakpoint-xs:0px}[data-theme=dark]{--primary-lightest:var(--primary-900);--secondary-lightest:var(--secondary-900);--static-dark:var(--static-white);--static-light:var(--static-black);--secondary-light:var(--secondary-800);--secondary-sub-base:var(--secondary-600);--secondary-base:var(--secondary-400);--secondary-dark:var(--secondary-300);--secondary-darkest:var(--secondary-100);--text-lightest:var(--neutral-900);--text-light:var(--neutral-600);--text-sub-base:var(--neutral-500);--text-base:var(--neutral-400);--text-dark:var(--neutral-300);--text-darkest:var(--neutral-100);--primary-light:var(--primary-800);--primary-sub-base:var(--primary-700);--primary-base:var(--primary-500);--primary-dark:var(--primary-200);--primary-darkest:var(--primary-100);--background-lightest:var(--neutral-900);--background-light:var(--neutral-700);--background-sub-base:var(--neutral-600);--background-base:var(--neutral-400);--background-dark:var(--neutral-300);--background-darkest:var(--neutral-100);--frames-lightest:var(--neutral-1000);--border-lightest:var(--neutral-900);--icon-lightest:var(--neutral-900);--icon-light:var(--neutral-700);--icon-sub-base:var(--neutral-600);--icon-base:var(--neutral-400);--icon-dark:var(--neutral-300);--icon-darkest:var(--neutral-100);--border-light:var(--neutral-700);--border-sub-base:var(--neutral-600);--border-base:var(--neutral-400);--border-dark:var(--neutral-300);--border-darkest:var(--neutral-100);--frames-light:var(--neutral-900);--frames-sub-base:var(--neutral-600);--frames-base:var(--neutral-400);--frames-dark:var(--neutral-50);--frames-darkest:var(--neutral-0);--shadow-black-alpha-lightest:var(--alpha-neutral-neutral-500-2);--shadow-black-alpha-light:var(--alpha-neutral-neutral-500-4);--shadow-black-alpha-base:var(--alpha-neutral-neutral-500-8);--shadow-black-alpha-dark:var(--alpha-neutral-neutral-500-12);--shadow-black-alpha-darkest:var(--alpha-neutral-neutral-500-16);--overlay-black-none:var(--alpha-base-white-none);--overlay-black-xxlightest:var(--alpha-base-white-8);--overlay-black-xlightest:var(--alpha-base-white-12);--overlay-black-lightest:var(--alpha-base-white-16);--overlay-black-light:var(--alpha-base-white-32);--overlay-black-base:var(--alpha-base-white-48);--overlay-black-dark:var(--alpha-base-white-64);--overlay-black-darkest:var(--alpha-base-white-72);--states-info-lightest:var(--blue-900);--overlay-white-none:var(--alpha-base-black-none);--overlay-white-lightest:var(--alpha-base-black-16);--overlay-white-light:var(--alpha-base-black-32);--overlay-white-base:var(--alpha-base-black-48);--overlay-white-dark:var(--alpha-base-black-64);--overlay-white-darkest:var(--alpha-base-black-72);--states-lightest:var(--neutral-50);--states-info-light:var(--blue-800);--states-info-sub-base:var(--blue-700);--states-light:var(--neutral-300);--states-info-base:var(--blue-500);--states-base:var(--secondary-500);--states-info-dark:var(--blue-200);--states-dark:var(--neutral-600);--states-darkest:var(--neutral-800);--states-info-darkest:var(--blue-100);--states-success-lightest:var(--green-900);--states-success-light:var(--green-800);--states-success-sub-base:var(--green-700);--states-success-base:var(--green-500);--states-success-dark:var(--green-200);--states-danger-lightest:var(--red-900);--states-success-darkest:var(--green-100);--states-danger-light:var(--red-800);--states-danger-sub-base:var(--red-700);--states-danger-base:var(--red-400);--states-danger-dark:var(--red-300);--states-danger-darkest:var(--red-100);--states-warning-lightest:var(--yellow-900);--states-warning-light:var(--yellow-800);--states-warning-sub-base:var(--yellow-700);--states-warning-base:var(--yellow-500);--states-warning-dark:var(--yellow-400);--states-verified-lightest:var(--blue-900);--states-warning-darkest:var(--yellow-200);--states-verified-light:var(--blue-800);--states-verified-sub-base:var(--blue-700);--states-verified-base:var(--blue-400);--states-verified-dark:var(--blue-300);--states-verified-darkest:var(--blue-100);--shadow-blue-alpha-lightest:var(--alpha-blue-blue-500-2);--shadow-blue-alpha-light:var(--alpha-blue-blue-500-4);--shadow-blue-alpha-base:var(--alpha-blue-blue-500-8);--shadow-blue-alpha-dark:var(--alpha-blue-blue-500-12);--shadow-blue-alpha-darkest:var(--alpha-blue-blue-500-16);--color:#ffffff;--skeleton-skeleton-gradient-none:#52586600;--skeleton-skeleton-gradient:#525866;--shadow-primary-alpha-lightest:var(--alpha-brand-primary-500-2);--shadow-primary-alpha-light:var(--alpha-brand-primary-500-4);--shadow-primary-alpha-base:var(--alpha-brand-primary-500-8);--shadow-primary-alpha-dark:var(--alpha-brand-primary-500-16);--shadow-primary-alpha-darkest:var(--alpha-brand-primary-500-24);}@media (min-width: var(--breakpoint-md)){:root{--primary-213:32px}}@media (min-width: var(--breakpoint-md)){:root{--primary-213:50px}}.container{width:100%;margin:0 auto;padding:0 1rem}@media (min-width: 768px){.container{max-width:750px}}@media (min-width: 992px){.container{max-width:970px}}@media (min-width: 1200px){.container{max-width:1170px}}.row{display:flex;flex-wrap:wrap;margin-right:var(--spacing-m-base);margin-left:var(--spacing-m-base)}.col{flex:1;padding-right:var(--spacing-m-base);padding-left:var(--spacing-m-base)}@media (min-width: 576px){.col-6{flex:0 0 50%;max-width:50%}.col-12{flex:0 0 100%;max-width:100%}}@media (min-width: 768px){.col-4{flex:0 0 33.33%;max-width:33.33%}.col-8{flex:0 0 66.67%;max-width:66.67%}}@media (min-width: 992px){.col-3{flex:0 0 25%;max-width:25%}.col-9{flex:0 0 75%;max-width:75%}}.frame{border:1px solid lightgrey;border-radius:10px;padding:20px;margin:40px auto}.geologica{font-family:\"Geologica\", sans-serif}@font-face{font-family:\"tk-text\";src:url(\"./assets/fonts/tk-text/tk-text.woff2\") format(\"woff2\"), url(\"./assets/fonts/tk-text/tk-text.ttf\") format(\"truetype\");font-weight:400;font-style:normal;font-display:swap}@font-face{font-family:\"tk-display\";src:url(\"./assets/fonts/tk-display/tk-display.woff2\") format(\"woff2\"), url(\"./assets/fonts/tk-display/tk-display.ttf\") format(\"truetype\");font-weight:400;font-style:normal;font-display:swap}.tk-text,.tk-font{font-family:\"tk-text\", sans-serif}.tk-display{font-family:\"tk-display\", sans-serif}:root.tk-font,:root[data-tk-font=true],:root[data-tk-font],.tk-font,[data-tk-font=true],[data-tk-font]{--family-title:'tk-display', sans-serif;--family-label:'tk-text', sans-serif;--family-body:'tk-text', sans-serif;--family-subheading:'tk-text', sans-serif}.tk-toaster{position:fixed;display:flex;flex-direction:column;gap:16px;min-width:360px;max-width:480px;z-index:2000}.tk-toaster.top-right{top:20px;right:20px;bottom:unset;left:unset}.tk-toaster.top-left{top:20px;right:unset;bottom:unset;left:20px}.tk-toaster.bottom-right{top:unset;right:20px;bottom:20px;left:unset}.tk-toaster.bottom-left{top:unset;right:unset;bottom:20px;left:20px}.tk-toaster .tk-toast{opacity:0;transition:opacity 0.3s ease-out}.tk-toaster .tk-toast.open{opacity:1}tk-accordion *::-webkit-scrollbar,tk-accordion-item *::-webkit-scrollbar,tk-alert *::-webkit-scrollbar,tk-badge *::-webkit-scrollbar,tk-breadcrumb *::-webkit-scrollbar,tk-card *::-webkit-scrollbar,tk-carousel *::-webkit-scrollbar,tk-checkbox *::-webkit-scrollbar,tk-color-picker *::-webkit-scrollbar,tk-currency-input *::-webkit-scrollbar,tk-datepicker *::-webkit-scrollbar,tk-dialog *::-webkit-scrollbar,tk-drawer *::-webkit-scrollbar,tk-popover *::-webkit-scrollbar,tk-radio *::-webkit-scrollbar,tk-select *::-webkit-scrollbar,tk-stepper *::-webkit-scrollbar,tk-table *::-webkit-scrollbar,tk-timeline-item *::-webkit-scrollbar,tk-tooltip *::-webkit-scrollbar{width:4px;height:4px}tk-accordion *::-webkit-scrollbar-track,tk-accordion-item *::-webkit-scrollbar-track,tk-alert *::-webkit-scrollbar-track,tk-badge *::-webkit-scrollbar-track,tk-breadcrumb *::-webkit-scrollbar-track,tk-card *::-webkit-scrollbar-track,tk-carousel *::-webkit-scrollbar-track,tk-checkbox *::-webkit-scrollbar-track,tk-color-picker *::-webkit-scrollbar-track,tk-currency-input *::-webkit-scrollbar-track,tk-datepicker *::-webkit-scrollbar-track,tk-dialog *::-webkit-scrollbar-track,tk-drawer *::-webkit-scrollbar-track,tk-popover *::-webkit-scrollbar-track,tk-radio *::-webkit-scrollbar-track,tk-select *::-webkit-scrollbar-track,tk-stepper *::-webkit-scrollbar-track,tk-table *::-webkit-scrollbar-track,tk-timeline-item *::-webkit-scrollbar-track,tk-tooltip *::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}tk-accordion *::-webkit-scrollbar-thumb,tk-accordion-item *::-webkit-scrollbar-thumb,tk-alert *::-webkit-scrollbar-thumb,tk-badge *::-webkit-scrollbar-thumb,tk-breadcrumb *::-webkit-scrollbar-thumb,tk-card *::-webkit-scrollbar-thumb,tk-carousel *::-webkit-scrollbar-thumb,tk-checkbox *::-webkit-scrollbar-thumb,tk-color-picker *::-webkit-scrollbar-thumb,tk-currency-input *::-webkit-scrollbar-thumb,tk-datepicker *::-webkit-scrollbar-thumb,tk-dialog *::-webkit-scrollbar-thumb,tk-drawer *::-webkit-scrollbar-thumb,tk-popover *::-webkit-scrollbar-thumb,tk-radio *::-webkit-scrollbar-thumb,tk-select *::-webkit-scrollbar-thumb,tk-stepper *::-webkit-scrollbar-thumb,tk-table *::-webkit-scrollbar-thumb,tk-timeline-item *::-webkit-scrollbar-thumb,tk-tooltip *::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}tk-accordion *::-webkit-scrollbar-thumb:hover,tk-accordion-item *::-webkit-scrollbar-thumb:hover,tk-alert *::-webkit-scrollbar-thumb:hover,tk-badge *::-webkit-scrollbar-thumb:hover,tk-breadcrumb *::-webkit-scrollbar-thumb:hover,tk-card *::-webkit-scrollbar-thumb:hover,tk-carousel *::-webkit-scrollbar-thumb:hover,tk-checkbox *::-webkit-scrollbar-thumb:hover,tk-color-picker *::-webkit-scrollbar-thumb:hover,tk-currency-input *::-webkit-scrollbar-thumb:hover,tk-datepicker *::-webkit-scrollbar-thumb:hover,tk-dialog *::-webkit-scrollbar-thumb:hover,tk-drawer *::-webkit-scrollbar-thumb:hover,tk-popover *::-webkit-scrollbar-thumb:hover,tk-radio *::-webkit-scrollbar-thumb:hover,tk-select *::-webkit-scrollbar-thumb:hover,tk-stepper *::-webkit-scrollbar-thumb:hover,tk-table *::-webkit-scrollbar-thumb:hover,tk-timeline-item *::-webkit-scrollbar-thumb:hover,tk-tooltip *::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}tk-accordion *::-webkit-scrollbar-corner,tk-accordion-item *::-webkit-scrollbar-corner,tk-alert *::-webkit-scrollbar-corner,tk-badge *::-webkit-scrollbar-corner,tk-breadcrumb *::-webkit-scrollbar-corner,tk-card *::-webkit-scrollbar-corner,tk-carousel *::-webkit-scrollbar-corner,tk-checkbox *::-webkit-scrollbar-corner,tk-color-picker *::-webkit-scrollbar-corner,tk-currency-input *::-webkit-scrollbar-corner,tk-datepicker *::-webkit-scrollbar-corner,tk-dialog *::-webkit-scrollbar-corner,tk-drawer *::-webkit-scrollbar-corner,tk-popover *::-webkit-scrollbar-corner,tk-radio *::-webkit-scrollbar-corner,tk-select *::-webkit-scrollbar-corner,tk-stepper *::-webkit-scrollbar-corner,tk-table *::-webkit-scrollbar-corner,tk-timeline-item *::-webkit-scrollbar-corner,tk-tooltip *::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-table-filter-panel{width:220px;position:fixed;border-radius:16px;border:1px solid var(--border-light);background-color:var(--static-light);z-index:800;box-shadow:var(--effect-1-default-base);display:flex;flex-direction:column;padding:8px;gap:8px}.tk-table-filter-panel .tk-table-filter-checkbox-container,.tk-table-filter-panel .tk-table-filter-radio-container,.tk-table-filter-panel .tk-table-filter-treeview-container{display:flex;flex-direction:column;gap:6px;max-height:200px;overflow-y:auto}.tk-table-filter-panel .tk-table-filter-checkbox-container::-webkit-scrollbar,.tk-table-filter-panel .tk-table-filter-radio-container::-webkit-scrollbar,.tk-table-filter-panel .tk-table-filter-treeview-container::-webkit-scrollbar{width:4px;height:4px}.tk-table-filter-panel .tk-table-filter-checkbox-container::-webkit-scrollbar-track,.tk-table-filter-panel .tk-table-filter-radio-container::-webkit-scrollbar-track,.tk-table-filter-panel .tk-table-filter-treeview-container::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-table-filter-panel .tk-table-filter-checkbox-container::-webkit-scrollbar-thumb,.tk-table-filter-panel .tk-table-filter-radio-container::-webkit-scrollbar-thumb,.tk-table-filter-panel .tk-table-filter-treeview-container::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-table-filter-panel .tk-table-filter-checkbox-container::-webkit-scrollbar-thumb:hover,.tk-table-filter-panel .tk-table-filter-radio-container::-webkit-scrollbar-thumb:hover,.tk-table-filter-panel .tk-table-filter-treeview-container::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-table-filter-panel .tk-table-filter-checkbox-container::-webkit-scrollbar-corner,.tk-table-filter-panel .tk-table-filter-radio-container::-webkit-scrollbar-corner,.tk-table-filter-panel .tk-table-filter-treeview-container::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-table-filter-panel .tk-table-filter-checkbox-container .tk-table-filter-checkbox-item,.tk-table-filter-panel .tk-table-filter-checkbox-container .tk-table-filter-radio-item,.tk-table-filter-panel .tk-table-filter-radio-container .tk-table-filter-checkbox-item,.tk-table-filter-panel .tk-table-filter-radio-container .tk-table-filter-radio-item,.tk-table-filter-panel .tk-table-filter-treeview-container .tk-table-filter-checkbox-item,.tk-table-filter-panel .tk-table-filter-treeview-container .tk-table-filter-radio-item{display:flex;align-items:center;padding:4px}.tk-table-filter-panel .tk-table-filter-panel-buttons{display:flex;gap:8px}.tk-table-filter-panel .tk-table-filter-panel-buttons tk-button{flex:1}.tk-hint-wrapper{color:var(--text-sub-base);font-size:var(--desktop-body-xs-size);line-height:var(--desktop-body-xs-line-height);font-weight:var(--weight-light);display:flex;align-items:center;gap:var(--spacing-xs);padding-left:var(--spacing-xs);cursor:default;font-family:var(--family-body)}.tk-hint-wrapper.invalid,.tk-hint-wrapper.error{color:var(--states-danger-base)}.floating-hidden,.floating-hidden *{visibility:hidden !important}tk-currency-input .tk-currency-input-container-small .flag,tk-currency-input .tk-phone-input-container-small .flag,tk-phone-input .tk-currency-input-container-small .flag,tk-phone-input .tk-phone-input-container-small .flag{width:20px;height:20px}tk-currency-input .tk-currency-input-container-base .flag,tk-currency-input .tk-phone-input-container-base .flag,tk-phone-input .tk-currency-input-container-base .flag,tk-phone-input .tk-phone-input-container-base .flag{width:22px;height:22px}tk-currency-input .tk-currency-input-container-large .flag,tk-currency-input .tk-phone-input-container-large .flag,tk-phone-input .tk-currency-input-container-large .flag,tk-phone-input .tk-phone-input-container-large .flag{width:24px;height:24px}tk-currency-input .flag,tk-phone-input .flag{background:url(https://cdn.turkishairlines.com/m/544f39bc4b8f8e7e) no-repeat;background-size:150%;vertical-align:middle;border-radius:100%;background-position-x:50%;align-self:center}tk-currency-input .flag.flag-ad,tk-phone-input .flag.flag-ad{background-position-y:0%}tk-currency-input .flag.flag-ae,tk-phone-input .flag.flag-ae{background-position-y:0.395222%}tk-currency-input .flag.flag-af,tk-phone-input .flag.flag-af{background-position-y:0.790444%}tk-currency-input .flag.flag-ag,tk-phone-input .flag.flag-ag{background-position-y:1.185667%}tk-currency-input .flag.flag-ai,tk-phone-input .flag.flag-ai{background-position-y:1.580889%}tk-currency-input .flag.flag-al,tk-phone-input .flag.flag-al{background-position-y:1.976111%}tk-currency-input .flag.flag-am,tk-phone-input .flag.flag-am{background-position-y:2.371333%}tk-currency-input .flag.flag-an,tk-phone-input .flag.flag-an{background-position-y:2.766555%}tk-currency-input .flag.flag-ao,tk-phone-input .flag.flag-ao{background-position-y:3.161778%}tk-currency-input .flag.flag-aq,tk-phone-input .flag.flag-aq{background-position-y:3.557%}tk-currency-input .flag.flag-ar,tk-phone-input .flag.flag-ar{background-position-y:3.952222%}tk-currency-input .flag.flag-as,tk-phone-input .flag.flag-as{background-position-y:4.347444%}tk-currency-input .flag.flag-at,tk-phone-input .flag.flag-at{background-position-y:4.742666%}tk-currency-input .flag.flag-au,tk-phone-input .flag.flag-au{background-position-y:5.137889%}tk-currency-input .flag.flag-aw,tk-phone-input .flag.flag-aw{background-position-y:5.533111%}tk-currency-input .flag.flag-ax,tk-phone-input .flag.flag-ax{background-position-y:5.928333%}tk-currency-input .flag.flag-az,tk-phone-input .flag.flag-az{background-position-y:6.323555%}tk-currency-input .flag.flag-ba,tk-phone-input .flag.flag-ba{background-position-y:6.718777%}tk-currency-input .flag.flag-bb,tk-phone-input .flag.flag-bb{background-position-y:7.114%}tk-currency-input .flag.flag-bd,tk-phone-input .flag.flag-bd{background-position-y:7.509222%}tk-currency-input .flag.flag-be,tk-phone-input .flag.flag-be{background-position-y:7.904444%}tk-currency-input .flag.flag-bf,tk-phone-input .flag.flag-bf{background-position-y:8.299666%}tk-currency-input .flag.flag-bg,tk-phone-input .flag.flag-bg{background-position-y:8.694888%}tk-currency-input .flag.flag-bh,tk-phone-input .flag.flag-bh{background-position-y:9.090111%}tk-currency-input .flag.flag-bi,tk-phone-input .flag.flag-bi{background-position-y:9.485333%}tk-currency-input .flag.flag-bj,tk-phone-input .flag.flag-bj{background-position-y:9.880555%}tk-currency-input .flag.flag-bl,tk-phone-input .flag.flag-bl{background-position-y:10.275777%}tk-currency-input .flag.flag-bm,tk-phone-input .flag.flag-bm{background-position-y:10.670999%}tk-currency-input .flag.flag-bn,tk-phone-input .flag.flag-bn{background-position-y:11.066222%}tk-currency-input .flag.flag-bo,tk-phone-input .flag.flag-bo{background-position-y:11.461444%}tk-currency-input .flag.flag-bq,tk-phone-input .flag.flag-bq{background-position-y:11.856666%}tk-currency-input .flag.flag-br,tk-phone-input .flag.flag-br{background-position-y:12.251888%}tk-currency-input .flag.flag-bs,tk-phone-input .flag.flag-bs{background-position-y:12.64711%}tk-currency-input .flag.flag-bt,tk-phone-input .flag.flag-bt{background-position-y:13.042333%}tk-currency-input .flag.flag-bv,tk-phone-input .flag.flag-bv{background-position-y:13.437555%}tk-currency-input .flag.flag-bw,tk-phone-input .flag.flag-bw{background-position-y:13.832777%}tk-currency-input .flag.flag-by,tk-phone-input .flag.flag-by{background-position-y:14.227999%}tk-currency-input .flag.flag-bz,tk-phone-input .flag.flag-bz{background-position-y:14.623222%}tk-currency-input .flag.flag-ca,tk-phone-input .flag.flag-ca{background-position-y:15.018444%}tk-currency-input .flag.flag-cc,tk-phone-input .flag.flag-cc{background-position-y:15.413666%}tk-currency-input .flag.flag-cd,tk-phone-input .flag.flag-cd{background-position-y:15.808888%}tk-currency-input .flag.flag-cf,tk-phone-input .flag.flag-cf{background-position-y:16.20411%}tk-currency-input .flag.flag-cg,tk-phone-input .flag.flag-cg{background-position-y:16.599333%}tk-currency-input .flag.flag-ch,tk-phone-input .flag.flag-ch{background-position-y:16.994555%}tk-currency-input .flag.flag-ci,tk-phone-input .flag.flag-ci{background-position-y:17.389777%}tk-currency-input .flag.flag-ck,tk-phone-input .flag.flag-ck{background-position-y:17.784999%}tk-currency-input .flag.flag-cl,tk-phone-input .flag.flag-cl{background-position-y:18.180221%}tk-currency-input .flag.flag-cm,tk-phone-input .flag.flag-cm{background-position-y:18.575444%}tk-currency-input .flag.flag-cn,tk-phone-input .flag.flag-cn{background-position-y:18.970666%}tk-currency-input .flag.flag-co,tk-phone-input .flag.flag-co{background-position-y:19.365888%}tk-currency-input .flag.flag-cr,tk-phone-input .flag.flag-cr{background-position-y:19.76111%}tk-currency-input .flag.flag-cu,tk-phone-input .flag.flag-cu{background-position-y:20.156332%}tk-currency-input .flag.flag-cv,tk-phone-input .flag.flag-cv{background-position-y:20.551555%}tk-currency-input .flag.flag-cw,tk-phone-input .flag.flag-cw{background-position-y:20.946777%}tk-currency-input .flag.flag-cx,tk-phone-input .flag.flag-cx{background-position-y:21.341999%}tk-currency-input .flag.flag-cy,tk-phone-input .flag.flag-cy{background-position-y:21.737221%}tk-currency-input .flag.flag-cz,tk-phone-input .flag.flag-cz{background-position-y:22.132443%}tk-currency-input .flag.flag-de,tk-phone-input .flag.flag-de{background-position-y:22.527666%}tk-currency-input .flag.flag-dj,tk-phone-input .flag.flag-dj{background-position-y:22.922888%}tk-currency-input .flag.flag-dk,tk-phone-input .flag.flag-dk{background-position-y:23.31811%}tk-currency-input .flag.flag-dm,tk-phone-input .flag.flag-dm{background-position-y:23.713332%}tk-currency-input .flag.flag-do,tk-phone-input .flag.flag-do{background-position-y:24.108554%}tk-currency-input .flag.flag-dz,tk-phone-input .flag.flag-dz{background-position-y:24.503777%}tk-currency-input .flag.flag-ec,tk-phone-input .flag.flag-ec{background-position-y:24.898999%}tk-currency-input .flag.flag-ee,tk-phone-input .flag.flag-ee{background-position-y:25.294221%}tk-currency-input .flag.flag-eg,tk-phone-input .flag.flag-eg{background-position-y:25.689443%}tk-currency-input .flag.flag-eh,tk-phone-input .flag.flag-eh{background-position-y:26.084665%}tk-currency-input .flag.flag-er,tk-phone-input .flag.flag-er{background-position-y:26.479888%}tk-currency-input .flag.flag-es,tk-phone-input .flag.flag-es{background-position-y:26.87511%}tk-currency-input .flag.flag-et,tk-phone-input .flag.flag-et{background-position-y:27.270332%}tk-currency-input .flag.flag-fi,tk-phone-input .flag.flag-fi{background-position-y:27.665554%}tk-currency-input .flag.flag-fj,tk-phone-input .flag.flag-fj{background-position-y:28.060776%}tk-currency-input .flag.flag-fk,tk-phone-input .flag.flag-fk{background-position-y:28.455999%}tk-currency-input .flag.flag-fm,tk-phone-input .flag.flag-fm{background-position-y:28.851221%}tk-currency-input .flag.flag-fo,tk-phone-input .flag.flag-fo{background-position-y:29.246443%}tk-currency-input .flag.flag-fr,tk-phone-input .flag.flag-fr{background-position-y:29.641665%}tk-currency-input .flag.flag-fx,tk-phone-input .flag.flag-fx{background-position-y:30.036887%}tk-currency-input .flag.flag-ga,tk-phone-input .flag.flag-ga{background-position-y:30.43211%}tk-currency-input .flag.flag-gb,tk-currency-input .flag.flag-uk,tk-phone-input .flag.flag-gb,tk-phone-input .flag.flag-uk{background-position-y:30.827332%}tk-currency-input .flag.flag-gd,tk-phone-input .flag.flag-gd{background-position-y:31.222554%}tk-currency-input .flag.flag-ge,tk-phone-input .flag.flag-ge{background-position-y:31.617776%}tk-currency-input .flag.flag-gf,tk-phone-input .flag.flag-gf{background-position-y:32.012998%}tk-currency-input .flag.flag-gg,tk-phone-input .flag.flag-gg{background-position-y:32.408221%}tk-currency-input .flag.flag-gh,tk-phone-input .flag.flag-gh{background-position-y:32.803443%}tk-currency-input .flag.flag-gi,tk-phone-input .flag.flag-gi{background-position-y:33.198665%}tk-currency-input .flag.flag-gl,tk-phone-input .flag.flag-gl{background-position-y:33.593887%}tk-currency-input .flag.flag-gm,tk-phone-input .flag.flag-gm{background-position-y:33.989109%}tk-currency-input .flag.flag-gn,tk-phone-input .flag.flag-gn{background-position-y:34.384332%}tk-currency-input .flag.flag-gp,tk-phone-input .flag.flag-gp{background-position-y:34.779554%}tk-currency-input .flag.flag-gq,tk-phone-input .flag.flag-gq{background-position-y:35.174776%}tk-currency-input .flag.flag-gr,tk-phone-input .flag.flag-gr{background-position-y:35.569998%}tk-currency-input .flag.flag-gs,tk-phone-input .flag.flag-gs{background-position-y:35.96522%}tk-currency-input .flag.flag-gt,tk-phone-input .flag.flag-gt{background-position-y:36.360443%}tk-currency-input .flag.flag-gu,tk-phone-input .flag.flag-gu{background-position-y:36.755665%}tk-currency-input .flag.flag-gw,tk-phone-input .flag.flag-gw{background-position-y:37.150887%}tk-currency-input .flag.flag-gy,tk-phone-input .flag.flag-gy{background-position-y:37.546109%}tk-currency-input .flag.flag-hk,tk-phone-input .flag.flag-hk{background-position-y:37.941331%}tk-currency-input .flag.flag-hm,tk-phone-input .flag.flag-hm{background-position-y:38.336554%}tk-currency-input .flag.flag-hn,tk-phone-input .flag.flag-hn{background-position-y:38.731776%}tk-currency-input .flag.flag-hr,tk-phone-input .flag.flag-hr{background-position-y:39.126998%}tk-currency-input .flag.flag-ht,tk-phone-input .flag.flag-ht{background-position-y:39.52222%}tk-currency-input .flag.flag-hu,tk-phone-input .flag.flag-hu{background-position-y:39.917442%}tk-currency-input .flag.flag-id,tk-phone-input .flag.flag-id{background-position-y:40.312665%}tk-currency-input .flag.flag-ie,tk-phone-input .flag.flag-ie{background-position-y:40.707887%}tk-currency-input .flag.flag-il,tk-phone-input .flag.flag-il{background-position-y:41.103109%}tk-currency-input .flag.flag-im,tk-phone-input .flag.flag-im{background-position-y:41.498331%}tk-currency-input .flag.flag-in,tk-phone-input .flag.flag-in{background-position-y:41.893553%}tk-currency-input .flag.flag-io,tk-phone-input .flag.flag-io{background-position-y:42.288776%}tk-currency-input .flag.flag-iq,tk-phone-input .flag.flag-iq{background-position-y:42.683998%}tk-currency-input .flag.flag-ir,tk-phone-input .flag.flag-ir{background-position-y:43.07922%}tk-currency-input .flag.flag-is,tk-phone-input .flag.flag-is{background-position-y:43.474442%}tk-currency-input .flag.flag-it,tk-phone-input .flag.flag-it{background-position-y:43.869665%}tk-currency-input .flag.flag-je,tk-phone-input .flag.flag-je{background-position-y:44.264887%}tk-currency-input .flag.flag-jm,tk-phone-input .flag.flag-jm{background-position-y:44.660109%}tk-currency-input .flag.flag-jo,tk-phone-input .flag.flag-jo{background-position-y:45.055331%}tk-currency-input .flag.flag-jp,tk-phone-input .flag.flag-jp{background-position-y:45.450553%}tk-currency-input .flag.flag-ke,tk-phone-input .flag.flag-ke{background-position-y:45.845776%}tk-currency-input .flag.flag-kg,tk-phone-input .flag.flag-kg{background-position-y:46.240998%}tk-currency-input .flag.flag-kh,tk-phone-input .flag.flag-kh{background-position-y:46.63622%}tk-currency-input .flag.flag-ki,tk-phone-input .flag.flag-ki{background-position-y:47.031442%}tk-currency-input .flag.flag-km,tk-phone-input .flag.flag-km{background-position-y:47.426664%}tk-currency-input .flag.flag-kn,tk-phone-input .flag.flag-kn{background-position-y:47.821887%}tk-currency-input .flag.flag-kp,tk-phone-input .flag.flag-kp{background-position-y:48.217109%}tk-currency-input .flag.flag-kr,tk-phone-input .flag.flag-kr{background-position-y:48.612331%}tk-currency-input .flag.flag-kw,tk-phone-input .flag.flag-kw{background-position-y:49.007553%}tk-currency-input .flag.flag-ky,tk-phone-input .flag.flag-ky{background-position-y:49.402775%}tk-currency-input .flag.flag-kz,tk-phone-input .flag.flag-kz{background-position-y:49.797998%}tk-currency-input .flag.flag-la,tk-phone-input .flag.flag-la{background-position-y:50.19322%}tk-currency-input .flag.flag-lb,tk-phone-input .flag.flag-lb{background-position-y:50.588442%}tk-currency-input .flag.flag-lc,tk-phone-input .flag.flag-lc{background-position-y:50.983664%}tk-currency-input .flag.flag-li,tk-phone-input .flag.flag-li{background-position-y:51.378886%}tk-currency-input .flag.flag-lk,tk-phone-input .flag.flag-lk{background-position-y:51.774109%}tk-currency-input .flag.flag-lr,tk-phone-input .flag.flag-lr{background-position-y:52.169331%}tk-currency-input .flag.flag-ls,tk-phone-input .flag.flag-ls{background-position-y:52.564553%}tk-currency-input .flag.flag-lt,tk-phone-input .flag.flag-lt{background-position-y:52.959775%}tk-currency-input .flag.flag-lu,tk-phone-input .flag.flag-lu{background-position-y:53.354997%}tk-currency-input .flag.flag-lv,tk-phone-input .flag.flag-lv{background-position-y:53.75022%}tk-currency-input .flag.flag-ly,tk-phone-input .flag.flag-ly{background-position-y:54.145442%}tk-currency-input .flag.flag-ma,tk-phone-input .flag.flag-ma{background-position-y:54.540664%}tk-currency-input .flag.flag-mc,tk-phone-input .flag.flag-mc{background-position-y:54.935886%}tk-currency-input .flag.flag-md,tk-phone-input .flag.flag-md{background-position-y:55.331108%}tk-currency-input .flag.flag-me,tk-phone-input .flag.flag-me{background-position-y:55.726331%}tk-currency-input .flag.flag-mf,tk-phone-input .flag.flag-mf{background-position-y:56.121553%}tk-currency-input .flag.flag-mg,tk-phone-input .flag.flag-mg{background-position-y:56.516775%}tk-currency-input .flag.flag-mh,tk-phone-input .flag.flag-mh{background-position-y:56.911997%}tk-currency-input .flag.flag-mk,tk-phone-input .flag.flag-mk{background-position-y:57.307219%}tk-currency-input .flag.flag-ml,tk-phone-input .flag.flag-ml{background-position-y:57.702442%}tk-currency-input .flag.flag-mm,tk-phone-input .flag.flag-mm{background-position-y:58.097664%}tk-currency-input .flag.flag-mn,tk-phone-input .flag.flag-mn{background-position-y:58.492886%}tk-currency-input .flag.flag-mo,tk-phone-input .flag.flag-mo{background-position-y:58.888108%}tk-currency-input .flag.flag-mp,tk-phone-input .flag.flag-mp{background-position-y:59.28333%}tk-currency-input .flag.flag-mq,tk-phone-input .flag.flag-mq{background-position-y:59.678553%}tk-currency-input .flag.flag-mr,tk-phone-input .flag.flag-mr{background-position-y:60.073775%}tk-currency-input .flag.flag-ms,tk-phone-input .flag.flag-ms{background-position-y:60.468997%}tk-currency-input .flag.flag-mt,tk-phone-input .flag.flag-mt{background-position-y:60.864219%}tk-currency-input .flag.flag-mu,tk-phone-input .flag.flag-mu{background-position-y:61.259441%}tk-currency-input .flag.flag-mv,tk-phone-input .flag.flag-mv{background-position-y:61.654664%}tk-currency-input .flag.flag-mw,tk-phone-input .flag.flag-mw{background-position-y:62.049886%}tk-currency-input .flag.flag-mx,tk-phone-input .flag.flag-mx{background-position-y:62.445108%}tk-currency-input .flag.flag-my,tk-phone-input .flag.flag-my{background-position-y:62.84033%}tk-currency-input .flag.flag-mz,tk-phone-input .flag.flag-mz{background-position-y:63.235552%}tk-currency-input .flag.flag-na,tk-phone-input .flag.flag-na{background-position-y:63.630775%}tk-currency-input .flag.flag-nc,tk-phone-input .flag.flag-nc{background-position-y:64.025997%}tk-currency-input .flag.flag-ne,tk-phone-input .flag.flag-ne{background-position-y:64.421219%}tk-currency-input .flag.flag-nf,tk-phone-input .flag.flag-nf{background-position-y:64.816441%}tk-currency-input .flag.flag-ng,tk-phone-input .flag.flag-ng{background-position-y:65.211663%}tk-currency-input .flag.flag-ni,tk-phone-input .flag.flag-ni{background-position-y:65.606886%}tk-currency-input .flag.flag-nl,tk-phone-input .flag.flag-nl{background-position-y:66.002108%}tk-currency-input .flag.flag-no,tk-phone-input .flag.flag-no{background-position-y:66.39733%}tk-currency-input .flag.flag-np,tk-phone-input .flag.flag-np{background-position-y:66.792552%}tk-currency-input .flag.flag-nr,tk-phone-input .flag.flag-nr{background-position-y:67.187774%}tk-currency-input .flag.flag-nu,tk-phone-input .flag.flag-nu{background-position-y:67.582997%}tk-currency-input .flag.flag-nz,tk-phone-input .flag.flag-nz{background-position-y:67.978219%}tk-currency-input .flag.flag-om,tk-phone-input .flag.flag-om{background-position-y:68.373441%}tk-currency-input .flag.flag-pa,tk-phone-input .flag.flag-pa{background-position-y:68.768663%}tk-currency-input .flag.flag-pe,tk-phone-input .flag.flag-pe{background-position-y:69.163885%}tk-currency-input .flag.flag-pf,tk-phone-input .flag.flag-pf{background-position-y:69.559108%}tk-currency-input .flag.flag-pg,tk-phone-input .flag.flag-pg{background-position-y:69.95433%}tk-currency-input .flag.flag-ph,tk-phone-input .flag.flag-ph{background-position-y:70.349552%}tk-currency-input .flag.flag-pk,tk-phone-input .flag.flag-pk{background-position-y:70.744774%}tk-currency-input .flag.flag-pl,tk-phone-input .flag.flag-pl{background-position-y:71.139996%}tk-currency-input .flag.flag-pm,tk-phone-input .flag.flag-pm{background-position-y:71.535219%}tk-currency-input .flag.flag-pn,tk-phone-input .flag.flag-pn{background-position-y:71.930441%}tk-currency-input .flag.flag-pr,tk-phone-input .flag.flag-pr{background-position-y:72.325663%}tk-currency-input .flag.flag-ps,tk-phone-input .flag.flag-ps{background-position-y:72.720885%}tk-currency-input .flag.flag-pt,tk-phone-input .flag.flag-pt{background-position-y:73.116108%}tk-currency-input .flag.flag-pw,tk-phone-input .flag.flag-pw{background-position-y:73.51133%}tk-currency-input .flag.flag-py,tk-phone-input .flag.flag-py{background-position-y:73.906552%}tk-currency-input .flag.flag-qa,tk-phone-input .flag.flag-qa{background-position-y:74.301774%}tk-currency-input .flag.flag-re,tk-phone-input .flag.flag-re{background-position-y:74.696996%}tk-currency-input .flag.flag-ro,tk-phone-input .flag.flag-ro{background-position-y:75.092219%}tk-currency-input .flag.flag-rs,tk-phone-input .flag.flag-rs{background-position-y:75.487441%}tk-currency-input .flag.flag-ru,tk-phone-input .flag.flag-ru{background-position-y:75.882663%}tk-currency-input .flag.flag-rw,tk-phone-input .flag.flag-rw{background-position-y:76.277885%}tk-currency-input .flag.flag-sa,tk-phone-input .flag.flag-sa{background-position-y:76.673107%}tk-currency-input .flag.flag-sb,tk-phone-input .flag.flag-sb{background-position-y:77.06833%}tk-currency-input .flag.flag-sc,tk-phone-input .flag.flag-sc{background-position-y:77.463552%}tk-currency-input .flag.flag-sd,tk-phone-input .flag.flag-sd{background-position-y:77.858774%}tk-currency-input .flag.flag-se,tk-phone-input .flag.flag-se{background-position-y:78.253996%}tk-currency-input .flag.flag-sg,tk-phone-input .flag.flag-sg{background-position-y:78.649218%}tk-currency-input .flag.flag-sh,tk-phone-input .flag.flag-sh{background-position-y:79.044441%}tk-currency-input .flag.flag-si,tk-phone-input .flag.flag-si{background-position-y:79.439663%}tk-currency-input .flag.flag-sj,tk-phone-input .flag.flag-sj{background-position-y:79.834885%}tk-currency-input .flag.flag-sk,tk-phone-input .flag.flag-sk{background-position-y:80.230107%}tk-currency-input .flag.flag-sl,tk-phone-input .flag.flag-sl{background-position-y:80.625329%}tk-currency-input .flag.flag-sm,tk-phone-input .flag.flag-sm{background-position-y:81.020552%}tk-currency-input .flag.flag-sn,tk-phone-input .flag.flag-sn{background-position-y:81.415774%}tk-currency-input .flag.flag-so,tk-phone-input .flag.flag-so{background-position-y:81.810996%}tk-currency-input .flag.flag-sr,tk-phone-input .flag.flag-sr{background-position-y:82.206218%}tk-currency-input .flag.flag-ss,tk-phone-input .flag.flag-ss{background-position-y:82.60144%}tk-currency-input .flag.flag-st,tk-phone-input .flag.flag-st{background-position-y:82.996663%}tk-currency-input .flag.flag-sv,tk-phone-input .flag.flag-sv{background-position-y:83.391885%}tk-currency-input .flag.flag-sx,tk-phone-input .flag.flag-sx{background-position-y:83.787107%}tk-currency-input .flag.flag-sy,tk-phone-input .flag.flag-sy{background-position-y:84.182329%}tk-currency-input .flag.flag-sz,tk-phone-input .flag.flag-sz{background-position-y:84.577551%}tk-currency-input .flag.flag-tc,tk-phone-input .flag.flag-tc{background-position-y:84.972774%}tk-currency-input .flag.flag-td,tk-phone-input .flag.flag-td{background-position-y:85.367996%}tk-currency-input .flag.flag-tf,tk-phone-input .flag.flag-tf{background-position-y:85.763218%}tk-currency-input .flag.flag-tg,tk-phone-input .flag.flag-tg{background-position-y:86.15844%}tk-currency-input .flag.flag-th,tk-phone-input .flag.flag-th{background-position-y:86.553662%}tk-currency-input .flag.flag-tj,tk-phone-input .flag.flag-tj{background-position-y:86.948885%}tk-currency-input .flag.flag-tk,tk-phone-input .flag.flag-tk{background-position-y:87.344107%}tk-currency-input .flag.flag-tl,tk-phone-input .flag.flag-tl{background-position-y:87.739329%}tk-currency-input .flag.flag-tm,tk-phone-input .flag.flag-tm{background-position-y:88.134551%}tk-currency-input .flag.flag-tn,tk-phone-input .flag.flag-tn{background-position-y:88.529773%}tk-currency-input .flag.flag-to,tk-phone-input .flag.flag-to{background-position-y:88.924996%}tk-currency-input .flag.flag-tp,tk-phone-input .flag.flag-tp{background-position-y:89.320218%}tk-currency-input .flag.flag-tr,tk-phone-input .flag.flag-tr{background-position-y:89.71544%}tk-currency-input .flag.flag-tt,tk-phone-input .flag.flag-tt{background-position-y:90.110662%}tk-currency-input .flag.flag-tv,tk-phone-input .flag.flag-tv{background-position-y:90.505884%}tk-currency-input .flag.flag-tw,tk-phone-input .flag.flag-tw{background-position-y:90.901107%}tk-currency-input .flag.flag-tz,tk-phone-input .flag.flag-tz{background-position-y:91.296329%}tk-currency-input .flag.flag-ua,tk-phone-input .flag.flag-ua{background-position-y:91.691551%}tk-currency-input .flag.flag-ug,tk-phone-input .flag.flag-ug{background-position-y:92.086773%}tk-currency-input .flag.flag-um,tk-phone-input .flag.flag-um{background-position-y:92.481995%}tk-currency-input .flag.flag-us,tk-phone-input .flag.flag-us{background-position-y:92.877218%}tk-currency-input .flag.flag-uy,tk-phone-input .flag.flag-uy{background-position-y:93.27244%}tk-currency-input .flag.flag-uz,tk-phone-input .flag.flag-uz{background-position-y:93.667662%}tk-currency-input .flag.flag-va,tk-phone-input .flag.flag-va{background-position-y:94.062884%}tk-currency-input .flag.flag-vc,tk-phone-input .flag.flag-vc{background-position-y:94.458106%}tk-currency-input .flag.flag-ve,tk-phone-input .flag.flag-ve{background-position-y:94.853329%}tk-currency-input .flag.flag-vg,tk-phone-input .flag.flag-vg{background-position-y:95.248551%}tk-currency-input .flag.flag-vi,tk-phone-input .flag.flag-vi{background-position-y:95.643773%}tk-currency-input .flag.flag-vn,tk-phone-input .flag.flag-vn{background-position-y:96.038995%}tk-currency-input .flag.flag-vu,tk-phone-input .flag.flag-vu{background-position-y:96.434217%}tk-currency-input .flag.flag-wf,tk-phone-input .flag.flag-wf{background-position-y:96.82944%}tk-currency-input .flag.flag-ws,tk-phone-input .flag.flag-ws{background-position-y:97.224662%}tk-currency-input .flag.flag-xk,tk-phone-input .flag.flag-xk{background-position-y:97.619884%}tk-currency-input .flag.flag-ye,tk-phone-input .flag.flag-ye{background-position-y:98.015106%}tk-currency-input .flag.flag-yt,tk-phone-input .flag.flag-yt{background-position-y:98.410328%}tk-currency-input .flag.flag-za,tk-phone-input .flag.flag-za{background-position-y:98.805551%}tk-currency-input .flag.flag-zm,tk-phone-input .flag.flag-zm{background-position-y:99.200773%}tk-currency-input .flag.flag-zr,tk-phone-input .flag.flag-zr{background-position-y:99.595995%}tk-currency-input .flag.flag-zw,tk-phone-input .flag.flag-zw{background-position-y:99.991217%}tk-currency-input .flag.flag-none,tk-phone-input .flag.flag-none{background:var(--background-sub-base);display:flex;align-items:center;justify-content:center}";

/*
 Stencil Client Platform v4.41.2 | MIT Licensed | https://stenciljs.com
 */


// src/utils/constants.ts
var SVG_NS = "http://www.w3.org/2000/svg";
var HTML_NS = "http://www.w3.org/1999/xhtml";
var reWireGetterSetter = (instance, hostRef) => {
  var _a;
  const cmpMeta = hostRef.$cmpMeta$;
  const members = Object.entries((_a = cmpMeta.$members$) != null ? _a : {});
  members.map(([memberName, [memberFlags]]) => {
    if ((memberFlags & 31 /* Prop */ || memberFlags & 32 /* State */)) {
      const ogValue = instance[memberName];
      const ogDescriptor = getPropertyDescriptor(Object.getPrototypeOf(instance), memberName) || Object.getOwnPropertyDescriptor(instance, memberName);
      if (ogDescriptor) {
        Object.defineProperty(instance, memberName, {
          get() {
            return ogDescriptor.get.call(this);
          },
          set(newValue) {
            ogDescriptor.set.call(this, newValue);
          },
          configurable: true,
          enumerable: true
        });
      }
      instance[memberName] = hostRef.$instanceValues$.has(memberName) ? hostRef.$instanceValues$.get(memberName) : ogValue;
    }
  });
};
function getPropertyDescriptor(obj, memberName) {
  while (obj) {
    const desc = Object.getOwnPropertyDescriptor(obj, memberName);
    if (desc == null ? void 0 : desc.get) return desc;
    obj = Object.getPrototypeOf(obj);
  }
  return void 0;
}

// src/client/client-host-ref.ts
var getHostRef = (ref) => {
  if (ref.__stencil__getHostRef) {
    return ref.__stencil__getHostRef();
  }
  return void 0;
};
var registerInstance = (lazyInstance, hostRef) => {
  if (!hostRef) return;
  lazyInstance.__stencil__getHostRef = () => hostRef;
  hostRef.$lazyInstance$ = lazyInstance;
  if (hostRef.$cmpMeta$.$flags$ & 512 /* hasModernPropertyDecls */ && (BUILD.state)) {
    reWireGetterSetter(lazyInstance, hostRef);
  }
};
var registerHost = (hostElement, cmpMeta) => {
  const hostRef = {
    $flags$: 0,
    $hostElement$: hostElement,
    $cmpMeta$: cmpMeta,
    $instanceValues$: /* @__PURE__ */ new Map(),
    $serializerValues$: /* @__PURE__ */ new Map()
  };
  {
    hostRef.$onInstancePromise$ = new Promise((r) => hostRef.$onInstanceResolve$ = r);
  }
  {
    hostRef.$onReadyPromise$ = new Promise((r) => hostRef.$onReadyResolve$ = r);
    hostElement["s-p"] = [];
    hostElement["s-rc"] = [];
  }
  {
    hostRef.$fetchedCbList$ = [];
  }
  const ref = hostRef;
  hostElement.__stencil__getHostRef = () => ref;
  return ref;
};
var isMemberInElement = (elm, memberName) => memberName in elm;
var consoleError = (e, el) => (0, console.error)(e, el);

// src/client/client-load-module.ts
var cmpModules = /* @__PURE__ */ new Map();
var loadModule = (cmpMeta, hostRef, hmrVersionId) => {
  const exportName = cmpMeta.$tagName$.replace(/-/g, "_");
  const bundleId = cmpMeta.$lazyBundleId$;
  if (!bundleId) {
    return void 0;
  }
  const module = cmpModules.get(bundleId) ;
  if (module) {
    return module[exportName];
  }
  
        if (!hmrVersionId || !BUILD.hotModuleReplacement) {
          const processMod = importedModule => {
              cmpModules.set(bundleId, importedModule);
              return importedModule[exportName];
          }
          switch(bundleId) {
              
                case 'tk-button_9.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-button_9.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-accordion.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-accordion.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-accordion-item.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-accordion-item.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-alert.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-alert.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-avatar-group.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-avatar-group.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-breadcrumb.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-breadcrumb.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-card.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-card.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-carousel.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-carousel.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-chart.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-chart.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-currency-input.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-currency-input.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-dialog.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-dialog.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-drawer.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-drawer.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-dropdown.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-dropdown.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-editor.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-editor.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-gantt-chart.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-gantt-chart.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-org-chart.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-org-chart.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-phone-input.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-phone-input.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-popover.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-popover.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-radio-group.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-radio-group.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-rating.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-rating.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-slider.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-slider.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-step.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-step.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-stepper.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-stepper.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-tabs.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-tabs.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-textarea.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-textarea.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-timeline.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-timeline.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-timeline-item.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-timeline-item.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-toggle.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-toggle.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-upload.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-upload.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-avatar.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-avatar.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-breadcrumb-item.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-breadcrumb-item.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-datepicker_4.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-datepicker_4.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-tabs-item.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-tabs-item.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-toggle-button_2.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-toggle-button_2.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-tooltip.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-tooltip.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-badge.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-badge.cjs.entry.js')); }).then(processMod, consoleError);
                case 'tk-table.cjs':
                    return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(
                        /* webpackMode: "lazy" */
                        './tk-table.cjs.entry.js')); }).then(processMod, consoleError);
          }
      }
  return (function (t) { return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(t)); }); })(
    /* @vite-ignore */
    /* webpackInclude: /\.entry\.js$/ */
    /* webpackExclude: /\.system\.entry\.js$/ */
    /* webpackMode: "lazy" */
    `./${bundleId}.entry.js${""}`
  ).then(
    (importedModule) => {
      {
        cmpModules.set(bundleId, importedModule);
      }
      return importedModule[exportName];
    },
    (e) => {
      consoleError(e, hostRef.$hostElement$);
    }
  );
};

// src/client/client-style.ts
var styles = /* @__PURE__ */ new Map();
var HYDRATED_CSS = "{visibility:hidden}.hydrated{visibility:inherit}";
var SLOT_FB_CSS = "slot-fb{display:contents}slot-fb[hidden]{display:none}";
var XLINK_NS = "http://www.w3.org/1999/xlink";
var FORM_ASSOCIATED_CUSTOM_ELEMENT_CALLBACKS = [
  "formAssociatedCallback",
  "formResetCallback",
  "formDisabledCallback",
  "formStateRestoreCallback"
];
var win = typeof window !== "undefined" ? window : {};
var H = win.HTMLElement || class {
};
var plt = {
  $flags$: 0,
  $resourcesUrl$: "",
  jmp: (h2) => h2(),
  raf: (h2) => requestAnimationFrame(h2),
  ael: (el, eventName, listener, opts) => el.addEventListener(eventName, listener, opts),
  rel: (el, eventName, listener, opts) => el.removeEventListener(eventName, listener, opts),
  ce: (eventName, opts) => new CustomEvent(eventName, opts)
};
var supportsShadow = BUILD.shadowDom;
var supportsListenerOptions = /* @__PURE__ */ (() => {
  var _a;
  let supportsListenerOptions2 = false;
  try {
    (_a = win.document) == null ? void 0 : _a.addEventListener(
      "e",
      null,
      Object.defineProperty({}, "passive", {
        get() {
          supportsListenerOptions2 = true;
        }
      })
    );
  } catch (e) {
  }
  return supportsListenerOptions2;
})();
var promiseResolve = (v) => Promise.resolve(v);
var supportsConstructableStylesheets = /* @__PURE__ */ (() => {
  try {
    if (!win.document.adoptedStyleSheets) {
      return false;
    }
    new CSSStyleSheet();
    return typeof new CSSStyleSheet().replaceSync === "function";
  } catch (e) {
  }
  return false;
})() ;
var supportsMutableAdoptedStyleSheets = supportsConstructableStylesheets ? /* @__PURE__ */ (() => !!win.document && Object.getOwnPropertyDescriptor(win.document.adoptedStyleSheets, "length").writable)() : false;
var queuePending = false;
var queueDomReads = [];
var queueDomWrites = [];
var queueTask = (queue, write) => (cb) => {
  queue.push(cb);
  if (!queuePending) {
    queuePending = true;
    if (write && plt.$flags$ & 4 /* queueSync */) {
      nextTick(flush);
    } else {
      plt.raf(flush);
    }
  }
};
var consume = (queue) => {
  for (let i2 = 0; i2 < queue.length; i2++) {
    try {
      queue[i2](performance.now());
    } catch (e) {
      consoleError(e);
    }
  }
  queue.length = 0;
};
var flush = () => {
  consume(queueDomReads);
  {
    consume(queueDomWrites);
    if (queuePending = queueDomReads.length > 0) {
      plt.raf(flush);
    }
  }
};
var nextTick = (cb) => promiseResolve().then(cb);
var writeTask = /* @__PURE__ */ queueTask(queueDomWrites, true);

// src/utils/style.ts
function createStyleSheetIfNeededAndSupported(styles2) {
  if (!supportsConstructableStylesheets) return void 0;
  const sheet = new CSSStyleSheet();
  sheet.replaceSync(styles2);
  return sheet;
}

// src/utils/shadow-root.ts
var globalStyleSheet;
function createShadowRoot(cmpMeta) {
  var _a;
  const opts = { mode: "open" };
  const shadowRoot = this.attachShadow(opts);
  if (globalStyleSheet === void 0) globalStyleSheet = (_a = createStyleSheetIfNeededAndSupported(globalStyles)) != null ? _a : null;
  if (globalStyleSheet) {
    if (supportsMutableAdoptedStyleSheets) {
      shadowRoot.adoptedStyleSheets.push(globalStyleSheet);
    } else {
      shadowRoot.adoptedStyleSheets = [...shadowRoot.adoptedStyleSheets, globalStyleSheet];
    }
  } else if (!supportsConstructableStylesheets) {
    const styleElm = document.createElement("style");
    styleElm.innerHTML = globalStyles;
    shadowRoot.prepend(styleElm);
  }
}
var updateFallbackSlotVisibility = (elm) => {
  const childNodes = internalCall(elm, "childNodes");
  if (elm.tagName && elm.tagName.includes("-") && elm["s-cr"] && elm.tagName !== "SLOT-FB") {
    getHostSlotNodes(childNodes, elm.tagName).forEach((slotNode) => {
      if (slotNode.nodeType === 1 /* ElementNode */ && slotNode.tagName === "SLOT-FB") {
        if (getSlotChildSiblings(slotNode, getSlotName(slotNode), false).length) {
          slotNode.hidden = true;
        } else {
          slotNode.hidden = false;
        }
      }
    });
  }
  let i2 = 0;
  for (i2 = 0; i2 < childNodes.length; i2++) {
    const childNode = childNodes[i2];
    if (childNode.nodeType === 1 /* ElementNode */ && internalCall(childNode, "childNodes").length) {
      updateFallbackSlotVisibility(childNode);
    }
  }
};
var getSlottedChildNodes = (childNodes) => {
  const result = [];
  for (let i2 = 0; i2 < childNodes.length; i2++) {
    const slottedNode = childNodes[i2]["s-nr"] || void 0;
    if (slottedNode && slottedNode.isConnected) {
      result.push(slottedNode);
    }
  }
  return result;
};
function getHostSlotNodes(childNodes, hostName, slotName) {
  let i2 = 0;
  let slottedNodes = [];
  let childNode;
  for (; i2 < childNodes.length; i2++) {
    childNode = childNodes[i2];
    if (childNode["s-sr"] && (!hostName || childNode["s-hn"] === hostName) && (slotName === void 0 || getSlotName(childNode) === slotName)) {
      slottedNodes.push(childNode);
      if (typeof slotName !== "undefined") return slottedNodes;
    }
    slottedNodes = [...slottedNodes, ...getHostSlotNodes(childNode.childNodes, hostName, slotName)];
  }
  return slottedNodes;
}
var getSlotChildSiblings = (slot, slotName, includeSlot = true) => {
  const childNodes = [];
  if (includeSlot && slot["s-sr"] || !slot["s-sr"]) childNodes.push(slot);
  let node = slot;
  while (node = node.nextSibling) {
    if (getSlotName(node) === slotName && (includeSlot || !node["s-sr"])) childNodes.push(node);
  }
  return childNodes;
};
var isNodeLocatedInSlot = (nodeToRelocate, slotName) => {
  if (nodeToRelocate.nodeType === 1 /* ElementNode */) {
    if (nodeToRelocate.getAttribute("slot") === null && slotName === "") {
      return true;
    }
    if (nodeToRelocate.getAttribute("slot") === slotName) {
      return true;
    }
    return false;
  }
  if (nodeToRelocate["s-sn"] === slotName) {
    return true;
  }
  return slotName === "";
};
var getSlotName = (node) => typeof node["s-sn"] === "string" ? node["s-sn"] : node.nodeType === 1 && node.getAttribute("slot") || void 0;
function patchSlotNode(node) {
  if (node.assignedElements || node.assignedNodes || !node["s-sr"]) return;
  const assignedFactory = (elementsOnly) => (function(opts) {
    const toReturn = [];
    const slotName = this["s-sn"];
    if (opts == null ? void 0 : opts.flatten) {
      console.error(`
          Flattening is not supported for Stencil non-shadow slots.
          You can use \`.childNodes\` to nested slot fallback content.
          If you have a particular use case, please open an issue on the Stencil repo.
        `);
    }
    const parent = this["s-cr"].parentElement;
    const slottedNodes = parent.__childNodes ? parent.childNodes : getSlottedChildNodes(parent.childNodes);
    slottedNodes.forEach((n) => {
      if (slotName === getSlotName(n)) {
        toReturn.push(n);
      }
    });
    if (elementsOnly) {
      return toReturn.filter((n) => n.nodeType === 1 /* ElementNode */);
    }
    return toReturn;
  }).bind(node);
  node.assignedElements = assignedFactory(true);
  node.assignedNodes = assignedFactory(false);
}
function dispatchSlotChangeEvent(elm) {
  elm.dispatchEvent(new CustomEvent("slotchange", { bubbles: false, cancelable: false, composed: false }));
}
function findSlotFromSlottedNode(slottedNode, parentHost) {
  var _a;
  parentHost = parentHost || ((_a = slottedNode["s-ol"]) == null ? void 0 : _a.parentElement);
  if (!parentHost) return { slotNode: null, slotName: "" };
  const slotName = slottedNode["s-sn"] = getSlotName(slottedNode) || "";
  const childNodes = internalCall(parentHost, "childNodes");
  const slotNode = getHostSlotNodes(childNodes, parentHost.tagName, slotName)[0];
  return { slotNode, slotName };
}
var patchCloneNode = (HostElementPrototype) => {
  if (HostElementPrototype.__cloneNode) return;
  const orgCloneNode = HostElementPrototype.__cloneNode = HostElementPrototype.cloneNode;
  HostElementPrototype.cloneNode = function(deep) {
    const srcNode = this;
    const isShadowDom = srcNode.shadowRoot && supportsShadow ;
    const clonedNode = orgCloneNode.call(srcNode, isShadowDom ? deep : false);
    if (!isShadowDom && deep) {
      let i2 = 0;
      let slotted, nonStencilNode;
      const stencilPrivates = [
        "s-id",
        "s-cr",
        "s-lr",
        "s-rc",
        "s-sc",
        "s-p",
        "s-cn",
        "s-sr",
        "s-sn",
        "s-hn",
        "s-ol",
        "s-nr",
        "s-si",
        "s-rf",
        "s-scs"
      ];
      const childNodes = this.__childNodes || this.childNodes;
      for (; i2 < childNodes.length; i2++) {
        slotted = childNodes[i2]["s-nr"];
        nonStencilNode = stencilPrivates.every((privateField) => !childNodes[i2][privateField]);
        if (slotted) {
          {
            clonedNode.appendChild(slotted.cloneNode(true));
          }
        }
        if (nonStencilNode) {
          clonedNode.appendChild(childNodes[i2].cloneNode(true));
        }
      }
    }
    return clonedNode;
  };
};
function internalCall(node, method) {
  if ("__" + method in node) {
    const toReturn = node["__" + method];
    if (typeof toReturn !== "function") return toReturn;
    return toReturn.bind(node);
  } else {
    if (typeof node[method] !== "function") return node[method];
    return node[method].bind(node);
  }
}
var createTime = (fnName, tagName = "") => {
  {
    return () => {
      return;
    };
  }
};
var uniqueTime = (key, measureText) => {
  {
    return () => {
      return;
    };
  }
};

// src/utils/query-nonce-meta-tag-content.ts
function queryNonceMetaTagContent(doc) {
  var _a, _b, _c;
  return (_c = (_b = (_a = doc.head) == null ? void 0 : _a.querySelector('meta[name="csp-nonce"]')) == null ? void 0 : _b.getAttribute("content")) != null ? _c : void 0;
}

// src/runtime/styles.ts
var rootAppliedStyles = /* @__PURE__ */ new WeakMap();
var registerStyle = (scopeId2, cssText, allowCS) => {
  let style = styles.get(scopeId2);
  if (supportsConstructableStylesheets && allowCS) {
    style = style || new CSSStyleSheet();
    if (typeof style === "string") {
      style = cssText;
    } else {
      style.replaceSync(cssText);
    }
  } else {
    style = cssText;
  }
  styles.set(scopeId2, style);
};
var addStyle = (styleContainerNode, cmpMeta, mode) => {
  var _a, _b, _c;
  const scopeId2 = getScopeId(cmpMeta);
  const style = styles.get(scopeId2);
  if (!win.document) {
    return scopeId2;
  }
  styleContainerNode = styleContainerNode.nodeType === 11 /* DocumentFragment */ ? styleContainerNode : win.document;
  if (style) {
    if (typeof style === "string") {
      styleContainerNode = styleContainerNode.head || styleContainerNode;
      let appliedStyles = rootAppliedStyles.get(styleContainerNode);
      let styleElm;
      if (!appliedStyles) {
        rootAppliedStyles.set(styleContainerNode, appliedStyles = /* @__PURE__ */ new Set());
      }
      if (!appliedStyles.has(scopeId2)) {
        styleElm = win.document.createElement("style");
        styleElm.textContent = style;
        const nonce = (_a = plt.$nonce$) != null ? _a : queryNonceMetaTagContent(win.document);
        if (nonce != null) {
          styleElm.setAttribute("nonce", nonce);
        }
        if (!(cmpMeta.$flags$ & 1 /* shadowDomEncapsulation */)) {
          if (styleContainerNode.nodeName === "HEAD") {
            const preconnectLinks = styleContainerNode.querySelectorAll("link[rel=preconnect]");
            const referenceNode2 = preconnectLinks.length > 0 ? preconnectLinks[preconnectLinks.length - 1].nextSibling : styleContainerNode.querySelector("style");
            styleContainerNode.insertBefore(
              styleElm,
              (referenceNode2 == null ? void 0 : referenceNode2.parentNode) === styleContainerNode ? referenceNode2 : null
            );
          } else if ("host" in styleContainerNode) {
            if (supportsConstructableStylesheets) {
              const currentWindow = (_b = styleContainerNode.defaultView) != null ? _b : styleContainerNode.ownerDocument.defaultView;
              const stylesheet = new currentWindow.CSSStyleSheet();
              stylesheet.replaceSync(style);
              if (supportsMutableAdoptedStyleSheets) {
                styleContainerNode.adoptedStyleSheets.unshift(stylesheet);
              } else {
                styleContainerNode.adoptedStyleSheets = [stylesheet, ...styleContainerNode.adoptedStyleSheets];
              }
            } else {
              const existingStyleContainer = styleContainerNode.querySelector("style");
              if (existingStyleContainer) {
                existingStyleContainer.textContent = style + existingStyleContainer.textContent;
              } else {
                styleContainerNode.prepend(styleElm);
              }
            }
          } else {
            styleContainerNode.append(styleElm);
          }
        }
        if (cmpMeta.$flags$ & 1 /* shadowDomEncapsulation */) {
          styleContainerNode.insertBefore(styleElm, null);
        }
        if (cmpMeta.$flags$ & 4 /* hasSlotRelocation */) {
          styleElm.textContent += SLOT_FB_CSS;
        }
        if (appliedStyles) {
          appliedStyles.add(scopeId2);
        }
      }
    } else {
      let appliedStyles = rootAppliedStyles.get(styleContainerNode);
      if (!appliedStyles) {
        rootAppliedStyles.set(styleContainerNode, appliedStyles = /* @__PURE__ */ new Set());
      }
      if (!appliedStyles.has(scopeId2)) {
        const currentWindow = (_c = styleContainerNode.defaultView) != null ? _c : styleContainerNode.ownerDocument.defaultView;
        let stylesheet;
        if (style.constructor === currentWindow.CSSStyleSheet) {
          stylesheet = style;
        } else {
          stylesheet = new currentWindow.CSSStyleSheet();
          for (let i2 = 0; i2 < style.cssRules.length; i2++) {
            stylesheet.insertRule(style.cssRules[i2].cssText, i2);
          }
        }
        if (supportsMutableAdoptedStyleSheets) {
          styleContainerNode.adoptedStyleSheets.push(stylesheet);
        } else {
          styleContainerNode.adoptedStyleSheets = [...styleContainerNode.adoptedStyleSheets, stylesheet];
        }
        appliedStyles.add(scopeId2);
      }
    }
  }
  return scopeId2;
};
var attachStyles = (hostRef) => {
  const cmpMeta = hostRef.$cmpMeta$;
  const elm = hostRef.$hostElement$;
  const flags = cmpMeta.$flags$;
  const endAttachStyles = createTime("attachStyles", cmpMeta.$tagName$);
  const scopeId2 = addStyle(
    elm.shadowRoot ? elm.shadowRoot : elm.getRootNode(),
    cmpMeta);
  if (flags & 10 /* needsScopedEncapsulation */) {
    elm["s-sc"] = scopeId2;
    elm.classList.add(scopeId2 + "-h");
  }
  endAttachStyles();
};
var getScopeId = (cmp, mode) => "sc-" + (cmp.$tagName$);
var isComplexType = (o) => {
  o = typeof o;
  return o === "object" || o === "function";
};

// src/runtime/vdom/h.ts
var h = (nodeName, vnodeData, ...children) => {
  if (typeof nodeName === "string") {
    nodeName = transformTag(nodeName);
  }
  let child = null;
  let key = null;
  let slotName = null;
  let simple = false;
  let lastSimple = false;
  const vNodeChildren = [];
  const walk = (c) => {
    for (let i2 = 0; i2 < c.length; i2++) {
      child = c[i2];
      if (Array.isArray(child)) {
        walk(child);
      } else if (child != null && typeof child !== "boolean") {
        if (simple = typeof nodeName !== "function" && !isComplexType(child)) {
          child = String(child);
        }
        if (simple && lastSimple) {
          vNodeChildren[vNodeChildren.length - 1].$text$ += child;
        } else {
          vNodeChildren.push(simple ? newVNode(null, child) : child);
        }
        lastSimple = simple;
      }
    }
  };
  walk(children);
  if (vnodeData) {
    if (vnodeData.key) {
      key = vnodeData.key;
    }
    if (vnodeData.name) {
      slotName = vnodeData.name;
    }
    {
      const classData = vnodeData.className || vnodeData.class;
      if (classData) {
        vnodeData.class = typeof classData !== "object" ? classData : Object.keys(classData).filter((k) => classData[k]).join(" ");
      }
    }
  }
  if (typeof nodeName === "function") {
    return nodeName(
      vnodeData === null ? {} : vnodeData,
      vNodeChildren,
      vdomFnUtils
    );
  }
  const vnode = newVNode(nodeName, null);
  vnode.$attrs$ = vnodeData;
  if (vNodeChildren.length > 0) {
    vnode.$children$ = vNodeChildren;
  }
  {
    vnode.$key$ = key;
  }
  {
    vnode.$name$ = slotName;
  }
  return vnode;
};
var newVNode = (tag, text) => {
  const vnode = {
    $flags$: 0,
    $tag$: tag,
    // Normalize undefined to null to prevent rendering "undefined" as text
    $text$: text != null ? text : null,
    $elm$: null,
    $children$: null
  };
  {
    vnode.$attrs$ = null;
  }
  {
    vnode.$key$ = null;
  }
  {
    vnode.$name$ = null;
  }
  return vnode;
};
var Host = {};
var isHost = (node) => node && node.$tag$ === Host;
var vdomFnUtils = {
  forEach: (children, cb) => children.map(convertToPublic).forEach(cb),
  map: (children, cb) => children.map(convertToPublic).map(cb).map(convertToPrivate)
};
var convertToPublic = (node) => ({
  vattrs: node.$attrs$,
  vchildren: node.$children$,
  vkey: node.$key$,
  vname: node.$name$,
  vtag: node.$tag$,
  vtext: node.$text$
});
var convertToPrivate = (node) => {
  if (typeof node.vtag === "function") {
    const vnodeData = { ...node.vattrs };
    if (node.vkey) {
      vnodeData.key = node.vkey;
    }
    if (node.vname) {
      vnodeData.name = node.vname;
    }
    return h(node.vtag, vnodeData, ...node.vchildren || []);
  }
  const vnode = newVNode(node.vtag, node.vtext);
  vnode.$attrs$ = node.vattrs;
  vnode.$children$ = node.vchildren;
  vnode.$key$ = node.vkey;
  vnode.$name$ = node.vname;
  return vnode;
};

// src/runtime/parse-property-value.ts
var parsePropertyValue = (propValue, propType, isFormAssociated) => {
  if (propValue != null && !isComplexType(propValue)) {
    if (propType & 4 /* Boolean */) {
      if (isFormAssociated && typeof propValue === "string") {
        return propValue === "" || !!propValue;
      } else {
        return propValue === "false" ? false : propValue === "" || !!propValue;
      }
    }
    if (propType & 2 /* Number */) {
      return typeof propValue === "string" ? parseFloat(propValue) : typeof propValue === "number" ? propValue : NaN;
    }
    if (propType & 1 /* String */) {
      return String(propValue);
    }
    return propValue;
  }
  return propValue;
};
var getElement = (ref) => {
  var _a;
  return (_a = getHostRef(ref)) == null ? void 0 : _a.$hostElement$ ;
};

// src/runtime/event-emitter.ts
var createEvent = (ref, name, flags) => {
  const elm = getElement(ref);
  return {
    emit: (detail) => {
      return emitEvent(elm, name, {
        bubbles: !!(flags & 4 /* Bubbles */),
        composed: !!(flags & 2 /* Composed */),
        cancelable: !!(flags & 1 /* Cancellable */),
        detail
      });
    }
  };
};
var emitEvent = (elm, name, opts) => {
  const ev = plt.ce(name, opts);
  elm.dispatchEvent(ev);
  return ev;
};
var setAccessor = (elm, memberName, oldValue, newValue, isSvg, flags, initialRender) => {
  if (oldValue === newValue) {
    return;
  }
  let isProp = isMemberInElement(elm, memberName);
  let ln = memberName.toLowerCase();
  if (memberName === "class") {
    const classList = elm.classList;
    const oldClasses = parseClassList(oldValue);
    let newClasses = parseClassList(newValue);
    {
      classList.remove(...oldClasses.filter((c) => c && !newClasses.includes(c)));
      classList.add(...newClasses.filter((c) => c && !oldClasses.includes(c)));
    }
  } else if (memberName === "style") {
    {
      for (const prop in oldValue) {
        if (!newValue || newValue[prop] == null) {
          if (prop.includes("-")) {
            elm.style.removeProperty(prop);
          } else {
            elm.style[prop] = "";
          }
        }
      }
    }
    for (const prop in newValue) {
      if (!oldValue || newValue[prop] !== oldValue[prop]) {
        if (prop.includes("-")) {
          elm.style.setProperty(prop, newValue[prop]);
        } else {
          elm.style[prop] = newValue[prop];
        }
      }
    }
  } else if (memberName === "key") ; else if (memberName === "ref") {
    if (newValue) {
      newValue(elm);
    }
  } else if ((!isProp ) && memberName[0] === "o" && memberName[1] === "n") {
    if (memberName[2] === "-") {
      memberName = memberName.slice(3);
    } else if (isMemberInElement(win, ln)) {
      memberName = ln.slice(2);
    } else {
      memberName = ln[2] + memberName.slice(3);
    }
    if (oldValue || newValue) {
      const capture = memberName.endsWith(CAPTURE_EVENT_SUFFIX);
      memberName = memberName.replace(CAPTURE_EVENT_REGEX, "");
      if (oldValue) {
        plt.rel(elm, memberName, oldValue, capture);
      }
      if (newValue) {
        plt.ael(elm, memberName, newValue, capture);
      }
    }
  } else {
    const isComplex = isComplexType(newValue);
    if ((isProp || isComplex && newValue !== null) && !isSvg) {
      try {
        if (!elm.tagName.includes("-")) {
          const n = newValue == null ? "" : newValue;
          if (memberName === "list") {
            isProp = false;
          } else if (oldValue == null || elm[memberName] != n) {
            if (typeof elm.__lookupSetter__(memberName) === "function") {
              elm[memberName] = n;
            } else {
              elm.setAttribute(memberName, n);
            }
          }
        } else if (elm[memberName] !== newValue) {
          elm[memberName] = newValue;
        }
      } catch (e) {
      }
    }
    let xlink = false;
    {
      if (ln !== (ln = ln.replace(/^xlink\:?/, ""))) {
        memberName = ln;
        xlink = true;
      }
    }
    if (newValue == null || newValue === false) {
      if (newValue !== false || elm.getAttribute(memberName) === "") {
        if (xlink) {
          elm.removeAttributeNS(XLINK_NS, memberName);
        } else {
          elm.removeAttribute(memberName);
        }
      }
    } else if ((!isProp || flags & 4 /* isHost */ || isSvg) && !isComplex && elm.nodeType === 1 /* ElementNode */) {
      newValue = newValue === true ? "" : newValue;
      if (xlink) {
        elm.setAttributeNS(XLINK_NS, memberName, newValue);
      } else {
        elm.setAttribute(memberName, newValue);
      }
    }
  }
};
var parseClassListRegex = /\s/;
var parseClassList = (value) => {
  if (typeof value === "object" && value && "baseVal" in value) {
    value = value.baseVal;
  }
  if (!value || typeof value !== "string") {
    return [];
  }
  return value.split(parseClassListRegex);
};
var CAPTURE_EVENT_SUFFIX = "Capture";
var CAPTURE_EVENT_REGEX = new RegExp(CAPTURE_EVENT_SUFFIX + "$");

// src/runtime/vdom/update-element.ts
var updateElement = (oldVnode, newVnode, isSvgMode2, isInitialRender) => {
  const elm = newVnode.$elm$.nodeType === 11 /* DocumentFragment */ && newVnode.$elm$.host ? newVnode.$elm$.host : newVnode.$elm$;
  const oldVnodeAttrs = oldVnode && oldVnode.$attrs$ || {};
  const newVnodeAttrs = newVnode.$attrs$ || {};
  {
    for (const memberName of sortedAttrNames(Object.keys(oldVnodeAttrs))) {
      if (!(memberName in newVnodeAttrs)) {
        setAccessor(
          elm,
          memberName,
          oldVnodeAttrs[memberName],
          void 0,
          isSvgMode2,
          newVnode.$flags$);
      }
    }
  }
  for (const memberName of sortedAttrNames(Object.keys(newVnodeAttrs))) {
    setAccessor(
      elm,
      memberName,
      oldVnodeAttrs[memberName],
      newVnodeAttrs[memberName],
      isSvgMode2,
      newVnode.$flags$);
  }
};
function sortedAttrNames(attrNames) {
  return attrNames.includes("ref") ? (
    // we need to sort these to ensure that `'ref'` is the last attr
    [...attrNames.filter((attr) => attr !== "ref"), "ref"]
  ) : (
    // no need to sort, return the original array
    attrNames
  );
}
var contentRef;
var hostTagName;
var useNativeShadowDom = false;
var checkSlotFallbackVisibility = false;
var checkSlotRelocate = false;
var isSvgMode = false;
var createElm = (oldParentVNode, newParentVNode, childIndex) => {
  var _a;
  const newVNode2 = newParentVNode.$children$[childIndex];
  let i2 = 0;
  let elm;
  let childNode;
  let oldVNode;
  if (!useNativeShadowDom) {
    checkSlotRelocate = true;
    if (newVNode2.$tag$ === "slot") {
      newVNode2.$flags$ |= newVNode2.$children$ ? (
        // slot element has fallback content
        // still create an element that "mocks" the slot element
        2 /* isSlotFallback */
      ) : (
        // slot element does not have fallback content
        // create an html comment we'll use to always reference
        // where actual slot content should sit next to
        1 /* isSlotReference */
      );
    }
  }
  if (newVNode2.$text$ != null) {
    elm = newVNode2.$elm$ = win.document.createTextNode(newVNode2.$text$);
  } else if (newVNode2.$flags$ & 1 /* isSlotReference */) {
    elm = newVNode2.$elm$ = win.document.createTextNode("");
    {
      updateElement(null, newVNode2, isSvgMode);
    }
  } else {
    if (!isSvgMode) {
      isSvgMode = newVNode2.$tag$ === "svg";
    }
    if (!win.document) {
      throw new Error("You are trying to render a Stencil component in an environment that doesn't support the DOM.");
    }
    elm = newVNode2.$elm$ = win.document.createElementNS(
      isSvgMode ? SVG_NS : HTML_NS,
      !useNativeShadowDom && BUILD.slotRelocation && newVNode2.$flags$ & 2 /* isSlotFallback */ ? "slot-fb" : newVNode2.$tag$
    ) ;
    if (isSvgMode && newVNode2.$tag$ === "foreignObject") {
      isSvgMode = false;
    }
    {
      updateElement(null, newVNode2, isSvgMode);
    }
    if (newVNode2.$children$) {
      const appendTarget = newVNode2.$tag$ === "template" ? elm.content : elm;
      for (i2 = 0; i2 < newVNode2.$children$.length; ++i2) {
        childNode = createElm(oldParentVNode, newVNode2, i2);
        if (childNode) {
          appendTarget.appendChild(childNode);
        }
      }
    }
    {
      if (newVNode2.$tag$ === "svg") {
        isSvgMode = false;
      } else if (elm.tagName === "foreignObject") {
        isSvgMode = true;
      }
    }
  }
  elm["s-hn"] = hostTagName;
  {
    if (newVNode2.$flags$ & (2 /* isSlotFallback */ | 1 /* isSlotReference */)) {
      elm["s-sr"] = true;
      elm["s-cr"] = contentRef;
      elm["s-sn"] = newVNode2.$name$ || "";
      elm["s-rf"] = (_a = newVNode2.$attrs$) == null ? void 0 : _a.ref;
      patchSlotNode(elm);
      oldVNode = oldParentVNode && oldParentVNode.$children$ && oldParentVNode.$children$[childIndex];
      if (oldVNode && oldVNode.$tag$ === newVNode2.$tag$ && oldParentVNode.$elm$) {
        relocateToHostRoot(oldParentVNode.$elm$);
      }
    }
  }
  return elm;
};
var relocateToHostRoot = (parentElm) => {
  plt.$flags$ |= 1 /* isTmpDisconnected */;
  const host = parentElm.closest(hostTagName.toLowerCase());
  if (host != null) {
    const contentRefNode = Array.from(host.__childNodes || host.childNodes).find(
      (ref) => ref["s-cr"]
    );
    const childNodeArray = Array.from(
      parentElm.__childNodes || parentElm.childNodes
    );
    for (const childNode of contentRefNode ? childNodeArray.reverse() : childNodeArray) {
      if (childNode["s-sh"] != null) {
        insertBefore(host, childNode, contentRefNode != null ? contentRefNode : null);
        childNode["s-sh"] = void 0;
        checkSlotRelocate = true;
      }
    }
  }
  plt.$flags$ &= -2 /* isTmpDisconnected */;
};
var putBackInOriginalLocation = (parentElm, recursive) => {
  plt.$flags$ |= 1 /* isTmpDisconnected */;
  const oldSlotChildNodes = Array.from(parentElm.__childNodes || parentElm.childNodes);
  if (parentElm["s-sr"]) {
    let node = parentElm;
    while (node = node.nextSibling) {
      if (node && node["s-sn"] === parentElm["s-sn"] && node["s-sh"] === hostTagName) {
        oldSlotChildNodes.push(node);
      }
    }
  }
  for (let i2 = oldSlotChildNodes.length - 1; i2 >= 0; i2--) {
    const childNode = oldSlotChildNodes[i2];
    if (childNode["s-hn"] !== hostTagName && childNode["s-ol"]) {
      insertBefore(referenceNode(childNode).parentNode, childNode, referenceNode(childNode));
      childNode["s-ol"].remove();
      childNode["s-ol"] = void 0;
      childNode["s-sh"] = void 0;
      checkSlotRelocate = true;
    }
    if (recursive) {
      putBackInOriginalLocation(childNode, recursive);
    }
  }
  plt.$flags$ &= -2 /* isTmpDisconnected */;
};
var addVnodes = (parentElm, before, parentVNode, vnodes, startIdx, endIdx) => {
  let containerElm = parentElm["s-cr"] && parentElm["s-cr"].parentNode || parentElm;
  let childNode;
  if (containerElm.shadowRoot && containerElm.tagName === hostTagName) {
    containerElm = containerElm.shadowRoot;
  }
  if (parentVNode.$tag$ === "template") {
    containerElm = containerElm.content;
  }
  for (; startIdx <= endIdx; ++startIdx) {
    if (vnodes[startIdx]) {
      childNode = createElm(null, parentVNode, startIdx);
      if (childNode) {
        vnodes[startIdx].$elm$ = childNode;
        insertBefore(containerElm, childNode, referenceNode(before) );
      }
    }
  }
};
var removeVnodes = (vnodes, startIdx, endIdx) => {
  for (let index = startIdx; index <= endIdx; ++index) {
    const vnode = vnodes[index];
    if (vnode) {
      const elm = vnode.$elm$;
      nullifyVNodeRefs(vnode);
      if (elm) {
        {
          checkSlotFallbackVisibility = true;
          if (elm["s-ol"]) {
            elm["s-ol"].remove();
          } else {
            putBackInOriginalLocation(elm, true);
          }
        }
        elm.remove();
      }
    }
  }
};
var updateChildren = (parentElm, oldCh, newVNode2, newCh, isInitialRender = false) => {
  let oldStartIdx = 0;
  let newStartIdx = 0;
  let idxInOld = 0;
  let i2 = 0;
  let oldEndIdx = oldCh.length - 1;
  let oldStartVnode = oldCh[0];
  let oldEndVnode = oldCh[oldEndIdx];
  let newEndIdx = newCh.length - 1;
  let newStartVnode = newCh[0];
  let newEndVnode = newCh[newEndIdx];
  let node;
  let elmToMove;
  const containerElm = newVNode2.$tag$ === "template" ? parentElm.content : parentElm;
  while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
    if (oldStartVnode == null) {
      oldStartVnode = oldCh[++oldStartIdx];
    } else if (oldEndVnode == null) {
      oldEndVnode = oldCh[--oldEndIdx];
    } else if (newStartVnode == null) {
      newStartVnode = newCh[++newStartIdx];
    } else if (newEndVnode == null) {
      newEndVnode = newCh[--newEndIdx];
    } else if (isSameVnode(oldStartVnode, newStartVnode, isInitialRender)) {
      patch(oldStartVnode, newStartVnode, isInitialRender);
      oldStartVnode = oldCh[++oldStartIdx];
      newStartVnode = newCh[++newStartIdx];
    } else if (isSameVnode(oldEndVnode, newEndVnode, isInitialRender)) {
      patch(oldEndVnode, newEndVnode, isInitialRender);
      oldEndVnode = oldCh[--oldEndIdx];
      newEndVnode = newCh[--newEndIdx];
    } else if (isSameVnode(oldStartVnode, newEndVnode, isInitialRender)) {
      if ((oldStartVnode.$tag$ === "slot" || newEndVnode.$tag$ === "slot")) {
        putBackInOriginalLocation(oldStartVnode.$elm$.parentNode, false);
      }
      patch(oldStartVnode, newEndVnode, isInitialRender);
      insertBefore(containerElm, oldStartVnode.$elm$, oldEndVnode.$elm$.nextSibling);
      oldStartVnode = oldCh[++oldStartIdx];
      newEndVnode = newCh[--newEndIdx];
    } else if (isSameVnode(oldEndVnode, newStartVnode, isInitialRender)) {
      if ((oldStartVnode.$tag$ === "slot" || newEndVnode.$tag$ === "slot")) {
        putBackInOriginalLocation(oldEndVnode.$elm$.parentNode, false);
      }
      patch(oldEndVnode, newStartVnode, isInitialRender);
      insertBefore(containerElm, oldEndVnode.$elm$, oldStartVnode.$elm$);
      oldEndVnode = oldCh[--oldEndIdx];
      newStartVnode = newCh[++newStartIdx];
    } else {
      idxInOld = -1;
      {
        for (i2 = oldStartIdx; i2 <= oldEndIdx; ++i2) {
          if (oldCh[i2] && oldCh[i2].$key$ !== null && oldCh[i2].$key$ === newStartVnode.$key$) {
            idxInOld = i2;
            break;
          }
        }
      }
      if (idxInOld >= 0) {
        elmToMove = oldCh[idxInOld];
        if (elmToMove.$tag$ !== newStartVnode.$tag$) {
          node = createElm(oldCh && oldCh[newStartIdx], newVNode2, idxInOld);
        } else {
          patch(elmToMove, newStartVnode, isInitialRender);
          oldCh[idxInOld] = void 0;
          node = elmToMove.$elm$;
        }
        newStartVnode = newCh[++newStartIdx];
      } else {
        node = createElm(oldCh && oldCh[newStartIdx], newVNode2, newStartIdx);
        newStartVnode = newCh[++newStartIdx];
      }
      if (node) {
        {
          insertBefore(
            referenceNode(oldStartVnode.$elm$).parentNode,
            node,
            referenceNode(oldStartVnode.$elm$)
          );
        }
      }
    }
  }
  if (oldStartIdx > oldEndIdx) {
    addVnodes(
      parentElm,
      newCh[newEndIdx + 1] == null ? null : newCh[newEndIdx + 1].$elm$,
      newVNode2,
      newCh,
      newStartIdx,
      newEndIdx
    );
  } else if (newStartIdx > newEndIdx) {
    removeVnodes(oldCh, oldStartIdx, oldEndIdx);
  }
};
var isSameVnode = (leftVNode, rightVNode, isInitialRender = false) => {
  if (leftVNode.$tag$ === rightVNode.$tag$) {
    if (leftVNode.$tag$ === "slot") {
      return leftVNode.$name$ === rightVNode.$name$;
    }
    if (!isInitialRender) {
      return leftVNode.$key$ === rightVNode.$key$;
    }
    if (isInitialRender && !leftVNode.$key$ && rightVNode.$key$) {
      leftVNode.$key$ = rightVNode.$key$;
    }
    return true;
  }
  return false;
};
var referenceNode = (node) => node && node["s-ol"] || node;
var patch = (oldVNode, newVNode2, isInitialRender = false) => {
  const elm = newVNode2.$elm$ = oldVNode.$elm$;
  const oldChildren = oldVNode.$children$;
  const newChildren = newVNode2.$children$;
  const tag = newVNode2.$tag$;
  const text = newVNode2.$text$;
  let defaultHolder;
  if (text == null) {
    {
      isSvgMode = tag === "svg" ? true : tag === "foreignObject" ? false : isSvgMode;
    }
    {
      if (tag === "slot" && !useNativeShadowDom) {
        if (oldVNode.$name$ !== newVNode2.$name$) {
          newVNode2.$elm$["s-sn"] = newVNode2.$name$ || "";
          relocateToHostRoot(newVNode2.$elm$.parentElement);
        }
      }
      updateElement(oldVNode, newVNode2, isSvgMode);
    }
    if (oldChildren !== null && newChildren !== null) {
      updateChildren(elm, oldChildren, newVNode2, newChildren, isInitialRender);
    } else if (newChildren !== null) {
      if (oldVNode.$text$ !== null) {
        elm.textContent = "";
      }
      addVnodes(elm, null, newVNode2, newChildren, 0, newChildren.length - 1);
    } else if (
      // don't do this on initial render as it can cause non-hydrated content to be removed
      !isInitialRender && BUILD.updatable && oldChildren !== null
    ) {
      removeVnodes(oldChildren, 0, oldChildren.length - 1);
    } else ;
    if (isSvgMode && tag === "svg") {
      isSvgMode = false;
    }
  } else if ((defaultHolder = elm["s-cr"])) {
    defaultHolder.parentNode.textContent = text;
  } else if (oldVNode.$text$ !== text) {
    elm.data = text;
  }
};
var relocateNodes = [];
var markSlotContentForRelocation = (elm) => {
  let node;
  let hostContentNodes;
  let j;
  const children = elm.__childNodes || elm.childNodes;
  for (const childNode of children) {
    if (childNode["s-sr"] && (node = childNode["s-cr"]) && node.parentNode) {
      hostContentNodes = node.parentNode.__childNodes || node.parentNode.childNodes;
      const slotName = childNode["s-sn"];
      for (j = hostContentNodes.length - 1; j >= 0; j--) {
        node = hostContentNodes[j];
        if (!node["s-cn"] && !node["s-nr"] && node["s-hn"] !== childNode["s-hn"] && (!node["s-sh"] || node["s-sh"] !== childNode["s-hn"])) {
          if (isNodeLocatedInSlot(node, slotName)) {
            let relocateNodeData = relocateNodes.find((r) => r.$nodeToRelocate$ === node);
            checkSlotFallbackVisibility = true;
            node["s-sn"] = node["s-sn"] || slotName;
            if (relocateNodeData) {
              relocateNodeData.$nodeToRelocate$["s-sh"] = childNode["s-hn"];
              relocateNodeData.$slotRefNode$ = childNode;
            } else {
              node["s-sh"] = childNode["s-hn"];
              relocateNodes.push({
                $slotRefNode$: childNode,
                $nodeToRelocate$: node
              });
            }
            if (node["s-sr"]) {
              relocateNodes.map((relocateNode) => {
                if (isNodeLocatedInSlot(relocateNode.$nodeToRelocate$, node["s-sn"])) {
                  relocateNodeData = relocateNodes.find((r) => r.$nodeToRelocate$ === node);
                  if (relocateNodeData && !relocateNode.$slotRefNode$) {
                    relocateNode.$slotRefNode$ = relocateNodeData.$slotRefNode$;
                  }
                }
              });
            }
          } else if (!relocateNodes.some((r) => r.$nodeToRelocate$ === node)) {
            relocateNodes.push({
              $nodeToRelocate$: node
            });
          }
        }
      }
    }
    if (childNode.nodeType === 1 /* ElementNode */) {
      markSlotContentForRelocation(childNode);
    }
  }
};
var nullifyVNodeRefs = (vNode) => {
  {
    vNode.$attrs$ && vNode.$attrs$.ref && vNode.$attrs$.ref(null);
    vNode.$children$ && vNode.$children$.map(nullifyVNodeRefs);
  }
};
var insertBefore = (parent, newNode, reference, isInitialLoad) => {
  {
    if (typeof newNode["s-sn"] === "string") {
      parent.insertBefore(newNode, reference);
      const { slotNode } = findSlotFromSlottedNode(newNode);
      if (slotNode && !isInitialLoad) dispatchSlotChangeEvent(slotNode);
      return newNode;
    }
  }
  if (parent.__insertBefore) {
    return parent.__insertBefore(newNode, reference);
  } else {
    return parent == null ? void 0 : parent.insertBefore(newNode, reference);
  }
};
var renderVdom = (hostRef, renderFnResults, isInitialLoad = false) => {
  var _a, _b, _c, _d, _e;
  const hostElm = hostRef.$hostElement$;
  const cmpMeta = hostRef.$cmpMeta$;
  const oldVNode = hostRef.$vnode$ || newVNode(null, null);
  const isHostElement = isHost(renderFnResults);
  const rootVnode = isHostElement ? renderFnResults : h(null, null, renderFnResults);
  hostTagName = hostElm.tagName;
  if (cmpMeta.$attrsToReflect$) {
    rootVnode.$attrs$ = rootVnode.$attrs$ || {};
    cmpMeta.$attrsToReflect$.forEach(([propName, attribute]) => {
      {
        rootVnode.$attrs$[attribute] = hostElm[propName];
      }
    });
  }
  if (isInitialLoad && rootVnode.$attrs$) {
    for (const key of Object.keys(rootVnode.$attrs$)) {
      if (hostElm.hasAttribute(key) && !["key", "ref", "style", "class"].includes(key)) {
        rootVnode.$attrs$[key] = hostElm[key];
      }
    }
  }
  rootVnode.$tag$ = null;
  rootVnode.$flags$ |= 4 /* isHost */;
  hostRef.$vnode$ = rootVnode;
  rootVnode.$elm$ = oldVNode.$elm$ = hostElm.shadowRoot || hostElm ;
  useNativeShadowDom = !!(cmpMeta.$flags$ & 1 /* shadowDomEncapsulation */) && !(cmpMeta.$flags$ & 128 /* shadowNeedsScopedCss */);
  {
    contentRef = hostElm["s-cr"];
    checkSlotFallbackVisibility = false;
  }
  patch(oldVNode, rootVnode, isInitialLoad);
  {
    plt.$flags$ |= 1 /* isTmpDisconnected */;
    if (checkSlotRelocate) {
      markSlotContentForRelocation(rootVnode.$elm$);
      for (const relocateData of relocateNodes) {
        const nodeToRelocate = relocateData.$nodeToRelocate$;
        if (!nodeToRelocate["s-ol"] && win.document) {
          const orgLocationNode = win.document.createTextNode("");
          orgLocationNode["s-nr"] = nodeToRelocate;
          insertBefore(
            nodeToRelocate.parentNode,
            nodeToRelocate["s-ol"] = orgLocationNode,
            nodeToRelocate,
            isInitialLoad
          );
        }
      }
      for (const relocateData of relocateNodes) {
        const nodeToRelocate = relocateData.$nodeToRelocate$;
        const slotRefNode = relocateData.$slotRefNode$;
        if (nodeToRelocate.nodeType === 1 /* ElementNode */ && isInitialLoad) {
          nodeToRelocate["s-ih"] = (_a = nodeToRelocate.hidden) != null ? _a : false;
        }
        if (slotRefNode) {
          const parentNodeRef = slotRefNode.parentNode;
          let insertBeforeNode = slotRefNode.nextSibling;
          if (insertBeforeNode && insertBeforeNode.nodeType === 1 /* ElementNode */) {
            let orgLocationNode = (_b = nodeToRelocate["s-ol"]) == null ? void 0 : _b.previousSibling;
            while (orgLocationNode) {
              let refNode = (_c = orgLocationNode["s-nr"]) != null ? _c : null;
              if (refNode && refNode["s-sn"] === nodeToRelocate["s-sn"] && parentNodeRef === (refNode.__parentNode || refNode.parentNode)) {
                refNode = refNode.nextSibling;
                while (refNode === nodeToRelocate || (refNode == null ? void 0 : refNode["s-sr"])) {
                  refNode = refNode == null ? void 0 : refNode.nextSibling;
                }
                if (!refNode || !refNode["s-nr"]) {
                  insertBeforeNode = refNode;
                  break;
                }
              }
              orgLocationNode = orgLocationNode.previousSibling;
            }
          }
          const parent = nodeToRelocate.__parentNode || nodeToRelocate.parentNode;
          const nextSibling = nodeToRelocate.__nextSibling || nodeToRelocate.nextSibling;
          if (!insertBeforeNode && parentNodeRef !== parent || nextSibling !== insertBeforeNode) {
            if (nodeToRelocate !== insertBeforeNode) {
              insertBefore(parentNodeRef, nodeToRelocate, insertBeforeNode, isInitialLoad);
              if (nodeToRelocate.nodeType === 8 /* CommentNode */ && nodeToRelocate.nodeValue.startsWith("s-nt-")) {
                const textNode = win.document.createTextNode(nodeToRelocate.nodeValue.replace(/^s-nt-/, ""));
                textNode["s-hn"] = nodeToRelocate["s-hn"];
                textNode["s-sn"] = nodeToRelocate["s-sn"];
                textNode["s-sh"] = nodeToRelocate["s-sh"];
                textNode["s-sr"] = nodeToRelocate["s-sr"];
                textNode["s-ol"] = nodeToRelocate["s-ol"];
                textNode["s-ol"]["s-nr"] = textNode;
                insertBefore(nodeToRelocate.parentNode, textNode, nodeToRelocate, isInitialLoad);
                nodeToRelocate.parentNode.removeChild(nodeToRelocate);
              }
              if (nodeToRelocate.nodeType === 1 /* ElementNode */ && nodeToRelocate.tagName !== "SLOT-FB") {
                nodeToRelocate.hidden = (_d = nodeToRelocate["s-ih"]) != null ? _d : false;
              }
            }
          }
          nodeToRelocate && typeof slotRefNode["s-rf"] === "function" && slotRefNode["s-rf"](slotRefNode);
        } else if (nodeToRelocate.nodeType === 1 /* ElementNode */) {
          nodeToRelocate.hidden = true;
        }
      }
    }
    if (checkSlotFallbackVisibility) {
      updateFallbackSlotVisibility(rootVnode.$elm$);
    }
    plt.$flags$ &= -2 /* isTmpDisconnected */;
    relocateNodes.length = 0;
  }
  if (!useNativeShadowDom && !(cmpMeta.$flags$ & 1 /* shadowDomEncapsulation */) && hostElm["s-cr"]) {
    const children = rootVnode.$elm$.__childNodes || rootVnode.$elm$.childNodes;
    for (const childNode of children) {
      if (childNode["s-hn"] !== hostTagName && !childNode["s-sh"]) {
        if (isInitialLoad && childNode["s-ih"] == null) {
          childNode["s-ih"] = (_e = childNode.hidden) != null ? _e : false;
        }
        if (childNode.nodeType === 1 /* ElementNode */) {
          childNode.hidden = true;
        } else if (childNode.nodeType === 3 /* TextNode */ && !!childNode.nodeValue.trim()) {
          const textCommentNode = win.document.createComment("s-nt-" + childNode.nodeValue);
          textCommentNode["s-sn"] = childNode["s-sn"];
          insertBefore(childNode.parentNode, textCommentNode, childNode, isInitialLoad);
          childNode.parentNode.removeChild(childNode);
        }
      }
    }
  }
  contentRef = void 0;
};

// src/runtime/update-component.ts
var attachToAncestor = (hostRef, ancestorComponent) => {
  if (ancestorComponent && !hostRef.$onRenderResolve$ && ancestorComponent["s-p"]) {
    const index = ancestorComponent["s-p"].push(
      new Promise(
        (r) => hostRef.$onRenderResolve$ = () => {
          ancestorComponent["s-p"].splice(index - 1, 1);
          r();
        }
      )
    );
  }
};
var scheduleUpdate = (hostRef, isInitialLoad) => {
  {
    hostRef.$flags$ |= 16 /* isQueuedForUpdate */;
  }
  if (hostRef.$flags$ & 4 /* isWaitingForChildren */) {
    hostRef.$flags$ |= 512 /* needsRerender */;
    return;
  }
  attachToAncestor(hostRef, hostRef.$ancestorComponent$);
  const dispatch = () => dispatchHooks(hostRef, isInitialLoad);
  if (isInitialLoad) {
    queueMicrotask(() => {
      dispatch();
    });
    return;
  }
  return writeTask(dispatch) ;
};
var dispatchHooks = (hostRef, isInitialLoad) => {
  const elm = hostRef.$hostElement$;
  const endSchedule = createTime("scheduleUpdate", hostRef.$cmpMeta$.$tagName$);
  const instance = hostRef.$lazyInstance$ ;
  if (!instance) {
    throw new Error(
      `Can't render component <${elm.tagName.toLowerCase()} /> with invalid Stencil runtime! Make sure this imported component is compiled with a \`externalRuntime: true\` flag. For more information, please refer to https://stenciljs.com/docs/custom-elements#externalruntime`
    );
  }
  let maybePromise;
  if (isInitialLoad) {
    {
      if (hostRef.$deferredConnectedCallback$) {
        hostRef.$deferredConnectedCallback$ = false;
        safeCall(instance, "connectedCallback", void 0, elm);
      }
      {
        hostRef.$flags$ |= 256 /* isListenReady */;
        if (hostRef.$queuedListeners$) {
          hostRef.$queuedListeners$.map(([methodName, event]) => safeCall(instance, methodName, event, elm));
          hostRef.$queuedListeners$ = void 0;
        }
      }
      if (hostRef.$fetchedCbList$.length) {
        hostRef.$fetchedCbList$.forEach((cb) => cb(elm));
      }
    }
    maybePromise = safeCall(instance, "componentWillLoad", void 0, elm);
  } else {
    maybePromise = safeCall(instance, "componentWillUpdate", void 0, elm);
  }
  maybePromise = enqueue(maybePromise, () => safeCall(instance, "componentWillRender", void 0, elm));
  endSchedule();
  return enqueue(maybePromise, () => updateComponent(hostRef, instance, isInitialLoad));
};
var enqueue = (maybePromise, fn) => isPromisey(maybePromise) ? maybePromise.then(fn).catch((err) => {
  console.error(err);
  fn();
}) : fn();
var isPromisey = (maybePromise) => maybePromise instanceof Promise || maybePromise && maybePromise.then && typeof maybePromise.then === "function";
var updateComponent = async (hostRef, instance, isInitialLoad) => {
  var _a;
  const elm = hostRef.$hostElement$;
  const endUpdate = createTime("update", hostRef.$cmpMeta$.$tagName$);
  const rc = elm["s-rc"];
  if (isInitialLoad) {
    attachStyles(hostRef);
  }
  const endRender = createTime("render", hostRef.$cmpMeta$.$tagName$);
  {
    callRender(hostRef, instance, elm, isInitialLoad);
  }
  if (rc) {
    rc.map((cb) => cb());
    elm["s-rc"] = void 0;
  }
  endRender();
  endUpdate();
  {
    const childrenPromises = (_a = elm["s-p"]) != null ? _a : [];
    const postUpdate = () => postUpdateComponent(hostRef);
    if (childrenPromises.length === 0) {
      postUpdate();
    } else {
      Promise.all(childrenPromises).then(postUpdate);
      hostRef.$flags$ |= 4 /* isWaitingForChildren */;
      childrenPromises.length = 0;
    }
  }
};
var callRender = (hostRef, instance, elm, isInitialLoad) => {
  try {
    instance = instance.render() ;
    {
      hostRef.$flags$ &= -17 /* isQueuedForUpdate */;
    }
    {
      hostRef.$flags$ |= 2 /* hasRendered */;
    }
    {
      {
        {
          renderVdom(hostRef, instance, isInitialLoad);
        }
      }
    }
  } catch (e) {
    consoleError(e, hostRef.$hostElement$);
  }
  return null;
};
var postUpdateComponent = (hostRef) => {
  const tagName = hostRef.$cmpMeta$.$tagName$;
  const elm = hostRef.$hostElement$;
  const endPostUpdate = createTime("postUpdate", tagName);
  const instance = hostRef.$lazyInstance$ ;
  const ancestorComponent = hostRef.$ancestorComponent$;
  safeCall(instance, "componentDidRender", void 0, elm);
  if (!(hostRef.$flags$ & 64 /* hasLoadedComponent */)) {
    hostRef.$flags$ |= 64 /* hasLoadedComponent */;
    {
      addHydratedFlag(elm);
    }
    safeCall(instance, "componentDidLoad", void 0, elm);
    endPostUpdate();
    {
      hostRef.$onReadyResolve$(elm);
      if (!ancestorComponent) {
        appDidLoad();
      }
    }
  } else {
    safeCall(instance, "componentDidUpdate", void 0, elm);
    endPostUpdate();
  }
  {
    hostRef.$onInstanceResolve$(elm);
  }
  {
    if (hostRef.$onRenderResolve$) {
      hostRef.$onRenderResolve$();
      hostRef.$onRenderResolve$ = void 0;
    }
    if (hostRef.$flags$ & 512 /* needsRerender */) {
      nextTick(() => scheduleUpdate(hostRef, false));
    }
    hostRef.$flags$ &= -517;
  }
};
var appDidLoad = (who) => {
  nextTick(() => emitEvent(win, "appload", { detail: { namespace: NAMESPACE } }));
};
var safeCall = (instance, method, arg, elm) => {
  if (instance && instance[method]) {
    try {
      return instance[method](arg);
    } catch (e) {
      consoleError(e, elm);
    }
  }
  return void 0;
};
var addHydratedFlag = (elm) => {
  var _a;
  return elm.classList.add((_a = BUILD.hydratedSelectorName) != null ? _a : "hydrated") ;
};

// src/runtime/set-value.ts
var getValue = (ref, propName) => getHostRef(ref).$instanceValues$.get(propName);
var setValue = (ref, propName, newVal, cmpMeta) => {
  const hostRef = getHostRef(ref);
  if (!hostRef) {
    return;
  }
  if (!hostRef) {
    throw new Error(
      `Couldn't find host element for "${cmpMeta.$tagName$}" as it is unknown to this Stencil runtime. This usually happens when integrating a 3rd party Stencil component with another Stencil component or application. Please reach out to the maintainers of the 3rd party Stencil component or report this on the Stencil Discord server (https://chat.stenciljs.com) or comment on this similar [GitHub issue](https://github.com/stenciljs/core/issues/5457).`
    );
  }
  const elm = hostRef.$hostElement$ ;
  const oldVal = hostRef.$instanceValues$.get(propName);
  const flags = hostRef.$flags$;
  const instance = hostRef.$lazyInstance$ ;
  newVal = parsePropertyValue(
    newVal,
    cmpMeta.$members$[propName][0],
    !!(cmpMeta.$flags$ & 64 /* formAssociated */)
  );
  const areBothNaN = Number.isNaN(oldVal) && Number.isNaN(newVal);
  const didValueChange = newVal !== oldVal && !areBothNaN;
  if ((!(flags & 8 /* isConstructingInstance */) || oldVal === void 0) && didValueChange) {
    hostRef.$instanceValues$.set(propName, newVal);
    if (cmpMeta.$watchers$) {
      const watchMethods = cmpMeta.$watchers$[propName];
      if (watchMethods) {
        watchMethods.map((watcher) => {
          try {
            const [[watchMethodName, watcherFlags]] = Object.entries(watcher);
            if (flags & 128 /* isWatchReady */ || watcherFlags & 1 /* Immediate */) {
              if (!instance) {
                hostRef.$fetchedCbList$.push(() => {
                  hostRef.$lazyInstance$[watchMethodName](newVal, oldVal, propName);
                });
              } else {
                instance[watchMethodName](newVal, oldVal, propName);
              }
            }
          } catch (e) {
            consoleError(e, elm);
          }
        });
      }
    }
    if ((flags & (2 /* hasRendered */ | 16 /* isQueuedForUpdate */)) === 2 /* hasRendered */) {
      if (instance.componentShouldUpdate) {
        if (instance.componentShouldUpdate(newVal, oldVal, propName) === false) {
          return;
        }
      }
      scheduleUpdate(hostRef, false);
    }
  }
};

// src/runtime/proxy-component.ts
var proxyComponent = (Cstr, cmpMeta, flags) => {
  var _a, _b;
  const prototype = Cstr.prototype;
  if (cmpMeta.$flags$ & 64 /* formAssociated */ && flags & 1 /* isElementConstructor */) {
    FORM_ASSOCIATED_CUSTOM_ELEMENT_CALLBACKS.forEach((cbName) => {
      Object.defineProperty(prototype, cbName, {
        value(...args) {
          var _a2;
          const hostRef = getHostRef(this);
          const instance = hostRef == null ? void 0 : hostRef.$lazyInstance$ ;
          if (!instance) {
            (_a2 = hostRef == null ? void 0 : hostRef.$onReadyPromise$) == null ? void 0 : _a2.then((asyncInstance) => {
              const cb = asyncInstance[cbName];
              typeof cb === "function" && cb.call(asyncInstance, ...args);
            });
          } else {
            const cb = instance[cbName] ;
            typeof cb === "function" && cb.call(instance, ...args);
          }
        }
      });
    });
  }
  if (cmpMeta.$members$ || BUILD.propChangeCallback) {
    {
      if (Cstr.watchers && !cmpMeta.$watchers$) {
        cmpMeta.$watchers$ = Cstr.watchers;
      }
      if (Cstr.deserializers && !cmpMeta.$deserializers$) {
        cmpMeta.$deserializers$ = Cstr.deserializers;
      }
      if (Cstr.serializers && !cmpMeta.$serializers$) {
        cmpMeta.$serializers$ = Cstr.serializers;
      }
    }
    const members = Object.entries((_a = cmpMeta.$members$) != null ? _a : {});
    members.map(([memberName, [memberFlags]]) => {
      if ((memberFlags & 31 /* Prop */ || (flags & 2 /* proxyState */) && memberFlags & 32 /* State */)) {
        const { get: origGetter, set: origSetter } = Object.getOwnPropertyDescriptor(prototype, memberName) || {};
        if (origGetter) cmpMeta.$members$[memberName][0] |= 2048 /* Getter */;
        if (origSetter) cmpMeta.$members$[memberName][0] |= 4096 /* Setter */;
        if (flags & 1 /* isElementConstructor */ || !origGetter) {
          Object.defineProperty(prototype, memberName, {
            get() {
              {
                if ((cmpMeta.$members$[memberName][0] & 2048 /* Getter */) === 0) {
                  return getValue(this, memberName);
                }
                const ref = getHostRef(this);
                const instance = ref ? ref.$lazyInstance$ : prototype;
                if (!instance) return;
                return instance[memberName];
              }
            },
            configurable: true,
            enumerable: true
          });
        }
        Object.defineProperty(prototype, memberName, {
          set(newValue) {
            const ref = getHostRef(this);
            if (!ref) {
              return;
            }
            if (origSetter) {
              const currentValue = memberFlags & 32 /* State */ ? this[memberName] : ref.$hostElement$[memberName];
              if (typeof currentValue === "undefined" && ref.$instanceValues$.get(memberName)) {
                newValue = ref.$instanceValues$.get(memberName);
              }
              origSetter.apply(this, [
                parsePropertyValue(
                  newValue,
                  memberFlags,
                  !!(cmpMeta.$flags$ & 64 /* formAssociated */)
                )
              ]);
              newValue = memberFlags & 32 /* State */ ? this[memberName] : ref.$hostElement$[memberName];
              setValue(this, memberName, newValue, cmpMeta);
              return;
            }
            {
              if ((flags & 1 /* isElementConstructor */) === 0 || (cmpMeta.$members$[memberName][0] & 4096 /* Setter */) === 0) {
                setValue(this, memberName, newValue, cmpMeta);
                if (flags & 1 /* isElementConstructor */ && !ref.$lazyInstance$) {
                  ref.$fetchedCbList$.push(() => {
                    if (cmpMeta.$members$[memberName][0] & 4096 /* Setter */ && ref.$lazyInstance$[memberName] !== ref.$instanceValues$.get(memberName)) {
                      ref.$lazyInstance$[memberName] = newValue;
                    }
                  });
                }
                return;
              }
              const setterSetVal = () => {
                const currentValue = ref.$lazyInstance$[memberName];
                if (!ref.$instanceValues$.get(memberName) && currentValue) {
                  ref.$instanceValues$.set(memberName, currentValue);
                }
                ref.$lazyInstance$[memberName] = parsePropertyValue(
                  newValue,
                  memberFlags,
                  !!(cmpMeta.$flags$ & 64 /* formAssociated */)
                );
                setValue(this, memberName, ref.$lazyInstance$[memberName], cmpMeta);
              };
              if (ref.$lazyInstance$) {
                setterSetVal();
              } else {
                ref.$fetchedCbList$.push(() => {
                  setterSetVal();
                });
              }
            }
          }
        });
      } else if (flags & 1 /* isElementConstructor */ && memberFlags & 64 /* Method */) {
        Object.defineProperty(prototype, memberName, {
          value(...args) {
            var _a2;
            const ref = getHostRef(this);
            return (_a2 = ref == null ? void 0 : ref.$onInstancePromise$) == null ? void 0 : _a2.then(() => {
              var _a3;
              return (_a3 = ref.$lazyInstance$) == null ? void 0 : _a3[memberName](...args);
            });
          }
        });
      }
    });
    if ((flags & 1 /* isElementConstructor */)) {
      const attrNameToPropName = /* @__PURE__ */ new Map();
      prototype.attributeChangedCallback = function(attrName, oldValue, newValue) {
        plt.jmp(() => {
          var _a2;
          const propName = attrNameToPropName.get(attrName);
          const hostRef = getHostRef(this);
          if (this.hasOwnProperty(propName) && BUILD.lazyLoad) {
            newValue = this[propName];
            delete this[propName];
          }
          if (prototype.hasOwnProperty(propName) && typeof this[propName] === "number" && // cast type to number to avoid TS compiler issues
          this[propName] == newValue) {
            return;
          } else if (propName == null) {
            const flags2 = hostRef == null ? void 0 : hostRef.$flags$;
            if (hostRef && flags2 && !(flags2 & 8 /* isConstructingInstance */) && newValue !== oldValue) {
              const instance = hostRef.$lazyInstance$ ;
              const entry = (_a2 = cmpMeta.$watchers$) == null ? void 0 : _a2[attrName];
              entry == null ? void 0 : entry.forEach((watcher) => {
                const [[watchMethodName, watcherFlags]] = Object.entries(watcher);
                if (instance[watchMethodName] != null && (flags2 & 128 /* isWatchReady */ || watcherFlags & 1 /* Immediate */)) {
                  instance[watchMethodName].call(instance, newValue, oldValue, attrName);
                }
              });
            }
            return;
          }
          const propFlags = members.find(([m]) => m === propName);
          if (propFlags && propFlags[1][0] & 4 /* Boolean */) {
            newValue = newValue === null || newValue === "false" ? false : true;
          }
          const propDesc = Object.getOwnPropertyDescriptor(prototype, propName);
          if (newValue != this[propName] && (!propDesc.get || !!propDesc.set)) {
            this[propName] = newValue;
          }
        });
      };
      Cstr.observedAttributes = Array.from(
        /* @__PURE__ */ new Set([
          ...Object.keys((_b = cmpMeta.$watchers$) != null ? _b : {}),
          ...members.filter(([_, m]) => m[0] & 31 /* HasAttribute */).map(([propName, m]) => {
            var _a2;
            const attrName = m[1] || propName;
            attrNameToPropName.set(attrName, propName);
            if (m[0] & 512 /* ReflectAttr */) {
              (_a2 = cmpMeta.$attrsToReflect$) == null ? void 0 : _a2.push([propName, attrName]);
            }
            return attrName;
          })
        ])
      );
    }
  }
  return Cstr;
};

// src/runtime/initialize-component.ts
var initializeComponent = async (elm, hostRef, cmpMeta, hmrVersionId) => {
  let Cstr;
  if ((hostRef.$flags$ & 32 /* hasInitializedComponent */) === 0) {
    hostRef.$flags$ |= 32 /* hasInitializedComponent */;
    const bundleId = cmpMeta.$lazyBundleId$;
    if (bundleId) {
      const CstrImport = loadModule(cmpMeta, hostRef);
      if (CstrImport && "then" in CstrImport) {
        const endLoad = uniqueTime();
        Cstr = await CstrImport;
        endLoad();
      } else {
        Cstr = CstrImport;
      }
      if (!Cstr) {
        throw new Error(`Constructor for "${cmpMeta.$tagName$}#${hostRef.$modeName$}" was not found`);
      }
      if (!Cstr.isProxied) {
        {
          cmpMeta.$watchers$ = Cstr.watchers;
          cmpMeta.$serializers$ = Cstr.serializers;
          cmpMeta.$deserializers$ = Cstr.deserializers;
        }
        proxyComponent(Cstr, cmpMeta, 2 /* proxyState */);
        Cstr.isProxied = true;
      }
      const endNewInstance = createTime("createInstance", cmpMeta.$tagName$);
      {
        hostRef.$flags$ |= 8 /* isConstructingInstance */;
      }
      try {
        new Cstr(hostRef);
      } catch (e) {
        consoleError(e, elm);
      }
      {
        hostRef.$flags$ &= -9 /* isConstructingInstance */;
      }
      {
        hostRef.$flags$ |= 128 /* isWatchReady */;
      }
      endNewInstance();
      const needsDeferredCallback = cmpMeta.$flags$ & 4 /* hasSlotRelocation */;
      if (!needsDeferredCallback) {
        fireConnectedCallback(hostRef.$lazyInstance$, elm);
      } else {
        hostRef.$deferredConnectedCallback$ = true;
      }
    } else {
      Cstr = elm.constructor;
      const cmpTag = elm.localName;
      customElements.whenDefined(cmpTag).then(() => hostRef.$flags$ |= 128 /* isWatchReady */);
    }
    if (Cstr && Cstr.style) {
      let style;
      if (typeof Cstr.style === "string") {
        style = Cstr.style;
      }
      const scopeId2 = getScopeId(cmpMeta);
      if (!styles.has(scopeId2) || BUILD.hotModuleReplacement) {
        const endRegisterStyles = createTime("registerStyles", cmpMeta.$tagName$);
        registerStyle(scopeId2, style, !!(cmpMeta.$flags$ & 1 /* shadowDomEncapsulation */));
        endRegisterStyles();
      }
    }
  }
  const ancestorComponent = hostRef.$ancestorComponent$;
  const schedule = () => scheduleUpdate(hostRef, true);
  if (ancestorComponent && ancestorComponent["s-rc"]) {
    ancestorComponent["s-rc"].push(schedule);
  } else {
    schedule();
  }
};
var fireConnectedCallback = (instance, elm) => {
  {
    safeCall(instance, "connectedCallback", void 0, elm);
  }
};

// src/runtime/connected-callback.ts
var connectedCallback = (elm) => {
  if ((plt.$flags$ & 1 /* isTmpDisconnected */) === 0) {
    const hostRef = getHostRef(elm);
    if (!hostRef) {
      return;
    }
    const cmpMeta = hostRef.$cmpMeta$;
    const endConnected = createTime("connectedCallback", cmpMeta.$tagName$);
    if (!(hostRef.$flags$ & 1 /* hasConnected */)) {
      hostRef.$flags$ |= 1 /* hasConnected */;
      {
        if (// TODO(STENCIL-854): Remove code related to legacy shadowDomShim field
        cmpMeta.$flags$ & (4 /* hasSlotRelocation */ | 8 /* needsShadowDomShim */)) {
          setContentReference(elm);
        }
      }
      {
        let ancestorComponent = elm;
        while (ancestorComponent = ancestorComponent.parentNode || ancestorComponent.host) {
          if (ancestorComponent["s-p"]) {
            attachToAncestor(hostRef, hostRef.$ancestorComponent$ = ancestorComponent);
            break;
          }
        }
      }
      if (cmpMeta.$members$) {
        Object.entries(cmpMeta.$members$).map(([memberName, [memberFlags]]) => {
          if (memberFlags & 31 /* Prop */ && memberName in elm && elm[memberName] !== Object.prototype[memberName]) {
            const value = elm[memberName];
            delete elm[memberName];
            elm[memberName] = value;
          }
        });
      }
      {
        initializeComponent(elm, hostRef, cmpMeta);
      }
    } else {
      addHostEventListeners(elm, hostRef, cmpMeta.$listeners$);
      if (hostRef == null ? void 0 : hostRef.$lazyInstance$) {
        fireConnectedCallback(hostRef.$lazyInstance$, elm);
      } else if (hostRef == null ? void 0 : hostRef.$onReadyPromise$) {
        hostRef.$onReadyPromise$.then(() => fireConnectedCallback(hostRef.$lazyInstance$, elm));
      }
    }
    endConnected();
  }
};
var setContentReference = (elm) => {
  if (!win.document) {
    return;
  }
  const contentRefElm = elm["s-cr"] = win.document.createComment(
    ""
  );
  contentRefElm["s-cn"] = true;
  insertBefore(elm, contentRefElm, elm.firstChild);
};
var disconnectInstance = (instance, elm) => {
  {
    safeCall(instance, "disconnectedCallback", void 0, elm || instance);
  }
};
var disconnectedCallback = async (elm) => {
  if ((plt.$flags$ & 1 /* isTmpDisconnected */) === 0) {
    const hostRef = getHostRef(elm);
    {
      if (hostRef == null ? void 0 : hostRef.$rmListeners$) {
        hostRef.$rmListeners$.map((rmListener) => rmListener());
        hostRef.$rmListeners$ = void 0;
      }
    }
    if (hostRef == null ? void 0 : hostRef.$lazyInstance$) {
      disconnectInstance(hostRef.$lazyInstance$, elm);
    } else if (hostRef == null ? void 0 : hostRef.$onReadyPromise$) {
      hostRef.$onReadyPromise$.then(() => disconnectInstance(hostRef.$lazyInstance$, elm));
    }
  }
  if (rootAppliedStyles.has(elm)) {
    rootAppliedStyles.delete(elm);
  }
  if (elm.shadowRoot && rootAppliedStyles.has(elm.shadowRoot)) {
    rootAppliedStyles.delete(elm.shadowRoot);
  }
};

// src/runtime/bootstrap-lazy.ts
var bootstrapLazy = (lazyBundles, options = {}) => {
  var _a;
  if (!win.document) {
    console.warn("Stencil: No document found. Skipping bootstrapping lazy components.");
    return;
  }
  const endBootstrap = createTime();
  const cmpTags = [];
  const exclude = options.exclude || [];
  const customElements2 = win.customElements;
  const head = win.document.head;
  const metaCharset = /* @__PURE__ */ head.querySelector("meta[charset]");
  const dataStyles = /* @__PURE__ */ win.document.createElement("style");
  const deferredConnectedCallbacks = [];
  let appLoadFallback;
  let isBootstrapping = true;
  Object.assign(plt, options);
  plt.$resourcesUrl$ = new URL(options.resourcesUrl || "./", win.document.baseURI).href;
  let hasSlotRelocation = false;
  lazyBundles.map((lazyBundle) => {
    lazyBundle[1].map((compactMeta) => {
      var _a2, _b, _c;
      const cmpMeta = {
        $flags$: compactMeta[0],
        $tagName$: compactMeta[1],
        $members$: compactMeta[2],
        $listeners$: compactMeta[3]
      };
      if (cmpMeta.$flags$ & 4 /* hasSlotRelocation */) {
        hasSlotRelocation = true;
      }
      {
        cmpMeta.$members$ = compactMeta[2];
      }
      {
        cmpMeta.$listeners$ = compactMeta[3];
      }
      {
        cmpMeta.$attrsToReflect$ = [];
      }
      {
        cmpMeta.$watchers$ = (_a2 = compactMeta[4]) != null ? _a2 : {};
        cmpMeta.$serializers$ = (_b = compactMeta[5]) != null ? _b : {};
        cmpMeta.$deserializers$ = (_c = compactMeta[6]) != null ? _c : {};
      }
      const tagName = transformTag(cmpMeta.$tagName$);
      const HostElement = class extends HTMLElement {
        ["s-p"];
        ["s-rc"];
        hasRegisteredEventListeners = false;
        // StencilLazyHost
        constructor(self) {
          super(self);
          self = this;
          registerHost(self, cmpMeta);
          if (cmpMeta.$flags$ & 1 /* shadowDomEncapsulation */) {
            {
              if (!self.shadowRoot) {
                createShadowRoot.call(self, cmpMeta);
              } else {
                if (self.shadowRoot.mode !== "open") {
                  throw new Error(
                    `Unable to re-use existing shadow root for ${cmpMeta.$tagName$}! Mode is set to ${self.shadowRoot.mode} but Stencil only supports open shadow roots.`
                  );
                }
              }
            }
          }
        }
        connectedCallback() {
          const hostRef = getHostRef(this);
          if (!hostRef) {
            return;
          }
          if (!this.hasRegisteredEventListeners) {
            this.hasRegisteredEventListeners = true;
            addHostEventListeners(this, hostRef, cmpMeta.$listeners$);
          }
          if (appLoadFallback) {
            clearTimeout(appLoadFallback);
            appLoadFallback = null;
          }
          if (isBootstrapping) {
            deferredConnectedCallbacks.push(this);
          } else {
            plt.jmp(() => connectedCallback(this));
          }
        }
        disconnectedCallback() {
          plt.jmp(() => disconnectedCallback(this));
          plt.raf(() => {
            var _a3;
            const hostRef = getHostRef(this);
            if (!hostRef) {
              return;
            }
            const i2 = deferredConnectedCallbacks.findIndex((host) => host === this);
            if (i2 > -1) {
              deferredConnectedCallbacks.splice(i2, 1);
            }
            if (((_a3 = hostRef == null ? void 0 : hostRef.$vnode$) == null ? void 0 : _a3.$elm$) instanceof Node && !hostRef.$vnode$.$elm$.isConnected) {
              delete hostRef.$vnode$.$elm$;
            }
          });
        }
        componentOnReady() {
          var _a3;
          return (_a3 = getHostRef(this)) == null ? void 0 : _a3.$onReadyPromise$;
        }
      };
      if (!(cmpMeta.$flags$ & 1 /* shadowDomEncapsulation */) && cmpMeta.$flags$ & 256 /* hasSlot */) {
        {
          {
            patchCloneNode(HostElement.prototype);
          }
        }
      } else {
        patchCloneNode(HostElement.prototype);
      }
      if (cmpMeta.$flags$ & 64 /* formAssociated */) {
        HostElement.formAssociated = true;
      }
      cmpMeta.$lazyBundleId$ = lazyBundle[0];
      if (!exclude.includes(tagName) && !customElements2.get(tagName)) {
        cmpTags.push(tagName);
        customElements2.define(
          tagName,
          proxyComponent(HostElement, cmpMeta, 1 /* isElementConstructor */)
        );
      }
    });
  });
  if (cmpTags.length > 0) {
    if (hasSlotRelocation) {
      dataStyles.textContent += SLOT_FB_CSS;
    }
    {
      dataStyles.textContent += cmpTags.sort() + HYDRATED_CSS;
    }
    if (dataStyles.innerHTML.length) {
      dataStyles.setAttribute("data-styles", "");
      const nonce = (_a = plt.$nonce$) != null ? _a : queryNonceMetaTagContent(win.document);
      if (nonce != null) {
        dataStyles.setAttribute("nonce", nonce);
      }
      head.insertBefore(dataStyles, metaCharset ? metaCharset.nextSibling : head.firstChild);
    }
  }
  isBootstrapping = false;
  if (deferredConnectedCallbacks.length) {
    deferredConnectedCallbacks.map((host) => host.connectedCallback());
  } else {
    {
      plt.jmp(() => appLoadFallback = setTimeout(appDidLoad, 30));
    }
  }
  endBootstrap();
};

// src/runtime/fragment.ts
var Fragment = (_, children) => children;
var addHostEventListeners = (elm, hostRef, listeners, attachParentListeners) => {
  if (listeners && win.document) {
    listeners.map(([flags, name, method]) => {
      const target = getHostListenerTarget(win.document, elm, flags) ;
      const handler = hostListenerProxy(hostRef, method);
      const opts = hostListenerOpts(flags);
      plt.ael(target, name, handler, opts);
      (hostRef.$rmListeners$ = hostRef.$rmListeners$ || []).push(() => plt.rel(target, name, handler, opts));
    });
  }
};
var hostListenerProxy = (hostRef, methodName) => (ev) => {
  var _a;
  try {
    {
      if (hostRef.$flags$ & 256 /* isListenReady */) {
        (_a = hostRef.$lazyInstance$) == null ? void 0 : _a[methodName](ev);
      } else {
        (hostRef.$queuedListeners$ = hostRef.$queuedListeners$ || []).push([methodName, ev]);
      }
    }
  } catch (e) {
    consoleError(e, hostRef.$hostElement$);
  }
};
var getHostListenerTarget = (doc, elm, flags) => {
  if (flags & 8 /* TargetWindow */) {
    return win;
  }
  return elm;
};
var hostListenerOpts = (flags) => supportsListenerOptions ? {
  passive: (flags & 1 /* Passive */) !== 0,
  capture: (flags & 2 /* Capture */) !== 0
} : (flags & 2 /* Capture */) !== 0;

// src/runtime/nonce.ts
var setNonce = (nonce) => plt.$nonce$ = nonce;
function transformTag(tag) {
  return tag;
}

exports.Fragment = Fragment;
exports.H = H;
exports.Host = Host;
exports.bootstrapLazy = bootstrapLazy;
exports.createEvent = createEvent;
exports.getElement = getElement;
exports.globalScripts = globalScripts;
exports.h = h;
exports.promiseResolve = promiseResolve;
exports.registerInstance = registerInstance;
exports.setNonce = setNonce;
