// @takeoff-ui/core@0.11.9/dist/cjs/tk-datepicker_4.cjs.entry.js
'use strict';

var index = require('./index-CUpr0td9.js');
var index$1 = require('./index-CgG3VNXc.js');
var index$2 = require('./index-DmcCMEYj.js');
var clickoutsideMixin = require('./clickoutside-mixin-BGIuHTh6.js');
var positionUtils = require('./position-utils-DGFX5w3y.js');
var testIdUtils = require('./test-id-utils-CAIe6ggJ.js');
var v4 = require('./v4-_2BfiRUa.js');
var iconUtils = require('./icon-utils-CoP4H8B1.js');

const tkDatepickerCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-datepicker-container{display:block}.tk-datepicker-panel{display:inline-flex;flex-direction:column;align-items:center;border-radius:var(--radius-xl);box-shadow:var(--effect-1-default-base);background:var(--static-light);font-family:var(--family-body)}.tk-datepicker-panel:not(.tk-datepicker-panel-inline){position:fixed;z-index:800}.tk-datepicker-panel.tk-datepicker-panel-inline{position:static;box-shadow:none;border:1px solid var(--border-light)}.tk-datepicker-panel .tk-datepicker-panel-inner{display:inline-flex}.tk-datepicker-panel .tk-datepicker-panel-inner .tk-datepicker-calendar-container{display:inline-flex;flex-direction:column}.tk-datepicker-panel .tk-datepicker-panel-inner .tk-datepicker-timepicker-panel{display:flex;flex-direction:column;border-left:1px solid var(--border-light)}.tk-datepicker-panel .tk-datepicker-panel-inner .tk-datepicker-timepicker-panel.tk-datepicker-timepicker-panel-only{border-left:0}.tk-datepicker-header{display:flex;width:336px;padding:var(--datepicker-header-base-top-padding) var(--datepicker-header-base-h-padding) var(--datepicker-header-base-bottom-padding) var(--datepicker-header-base-h-padding);flex-direction:column;align-items:flex-start}.tk-datepicker-header.tk-datepicker-header-basic{border-bottom:var(--spacing-px) solid var(--border-light)}.tk-datepicker-header.tk-datepicker-header-divided .tk-datepicker-header-content{border:1px solid var(--border-light)}.tk-datepicker-header.tk-datepicker-header-light .tk-datepicker-header-content{background:var(--background-lightest)}.tk-datepicker-header.tk-datepicker-header-primary .tk-datepicker-header-content{background:var(--primary-base)}.tk-datepicker-header.tk-datepicker-header-dark .tk-datepicker-header-content{background:var(--background-darkest)}.tk-datepicker-header .tk-datepicker-header-content{display:flex;justify-content:space-between;align-items:center;align-self:stretch;width:100%;border-radius:8px}.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-select-container{display:flex;padding:0px var(--datepicker-header-base-date-text-h-padding);align-items:center;gap:var(--datepicker-header-base-date-text-gap)}.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-month{color:var(--text-darkest)}.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-year{color:var(--text-base)}.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-month,.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-year{text-align:center;font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400;line-height:var(--desktop-body-s-line-height);cursor:pointer}.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-month.disabled,.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-year.disabled{pointer-events:none;user-select:none;color:var(--text-sub-base)}.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-header-content-start,.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-header-content-end{display:inline-flex;align-items:center;gap:var(--datepicker-header-base-navigation-gap)}.tk-datepicker-header .tk-datepicker-header-content .tk-datepicker-divider{width:1px;height:24px;background-color:var(--border-light)}.tk-datepicker-header.tk-datepicker-header-primary .tk-datepicker-header-content .tk-datepicker-divider{background-color:var(--primary-lightest)}.tk-datepicker-header.tk-datepicker-header-primary .tk-datepicker-header-content .tk-datepicker-divider.disabled{background-color:var(--primary-sub-base)}.tk-datepicker-header.tk-datepicker-header-primary .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-month,.tk-datepicker-header.tk-datepicker-header-primary .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-year{color:var(--primary-lightest)}.tk-datepicker-header.tk-datepicker-header-primary .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-month.disabled,.tk-datepicker-header.tk-datepicker-header-primary .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-year.disabled{color:var(--primary-light)}.tk-datepicker-header.tk-datepicker-header-dark .tk-datepicker-header-content .tk-datepicker-divider{background-color:var(--background-base)}.tk-datepicker-header.tk-datepicker-header-dark .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-month,.tk-datepicker-header.tk-datepicker-header-dark .tk-datepicker-header-content .tk-datepicker-select-container .tk-datepicker-select-year{color:var(--text-lightest)}.tk-datepicker-body{display:flex;width:336px;padding:var(--datepicker-body-base-day-v-padding) var(--datepicker-body-base-day-h-padding);flex-direction:column;align-items:center;gap:8px;flex-shrink:0}.tk-datepicker-body .tk-datepicker-table{width:100%;height:100%;table-layout:fixed;border-collapse:collapse;border-spacing:8px}.tk-datepicker-body.tk-datepicker-months-view,.tk-datepicker-body.tk-datepicker-years-view{height:260px}.tk-datepicker-week-day-cell{width:var(--datepicker-items-base-size);height:var(--datepicker-items-base-size);min-width:var(--datepicker-items-base-size);min-height:var(--datepicker-items-base-size);text-align:center;vertical-align:middle;padding:0}.tk-datepicker-week-day-cell .tk-datepicker-week-day{color:var(--text-base, #717784);text-align:center;font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400;line-height:var(--desktop-body-s-line-height)}.tk-datepicker-day-cell{position:relative;min-width:var(--datepicker-items-base-size);height:var(--datepicker-items-base-size);text-align:center;vertical-align:middle;padding-top:2px;padding-bottom:2px;z-index:100;cursor:pointer}.tk-datepicker-day-cell::before{content:"";position:absolute;top:50%;transform:translateY(-50%);background-color:transparent;height:40px}.tk-datepicker-day-cell .tk-datepicker-day{display:inline-block;position:relative;min-width:var(--datepicker-items-base-size);height:var(--datepicker-items-base-size);line-height:var(--datepicker-items-base-size);z-index:100;text-align:center;color:var(--text-darkest);font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400}.tk-datepicker-day-cell.disabled,.tk-datepicker-day-cell.readonly,.tk-datepicker-day-cell.adjacent-month{pointer-events:none;user-select:none}.tk-datepicker-day-cell.disabled .tk-datepicker-day,.tk-datepicker-day-cell.readonly .tk-datepicker-day,.tk-datepicker-day-cell.adjacent-month .tk-datepicker-day{color:var(--text-sub-base)}.tk-datepicker-day-cell.selected .tk-datepicker-day{color:var(--text-lightest);background:var(--primary-base);border-radius:var(--datepicker-items-base-radius)}.tk-datepicker-day-cell.selected.disabled .tk-datepicker-day,.tk-datepicker-day-cell.selected.readonly .tk-datepicker-day{color:var(--primary-light);background:var(--primary-lightest)}.tk-datepicker-day-cell.today .tk-datepicker-day{border-radius:var(--datepicker-items-base-radius);border:var(--spacing-px) solid var(--states-info-sub-base);background:var(--background-lightest);box-sizing:border-box}.tk-datepicker-day-cell.in-range:first-child::before{border-top-left-radius:var(--datepicker-items-base-radius);border-bottom-left-radius:var(--datepicker-items-base-radius)}.tk-datepicker-day-cell.in-range:last-child::before{border-top-right-radius:var(--datepicker-items-base-radius);border-bottom-right-radius:var(--datepicker-items-base-radius)}.tk-datepicker-day-cell.in-range::before{background:var(--primary-lightest);width:100%}.tk-datepicker-day-cell.range-start:has(+.in-range)::before,.tk-datepicker-day-cell.range-start:has(+.range-end)::before{background:var(--primary-lightest);width:100%;border-top-left-radius:var(--datepicker-items-base-radius);border-bottom-left-radius:var(--datepicker-items-base-radius)}.tk-datepicker-day-cell.range-start:not(.range-end) .tk-datepicker-day{border-top-right-radius:0;border-bottom-right-radius:0}.tk-datepicker-day-cell.range-end:not(.range-start) .tk-datepicker-day{border-top-left-radius:0;border-bottom-left-radius:0}.tk-datepicker-day-cell:not(.selected):not(.disabled):not(.in-range):hover .tk-datepicker-day{background:var(--background-lightest);border-radius:var(--datepicker-items-base-radius)}.tk-datepicker-table{border-collapse:separate;border-spacing:0}.tk-datepicker-months{display:flex;flex-direction:column;justify-content:space-between;height:100%;width:100%}.tk-datepicker-month-row{display:flex;justify-content:space-between}.tk-datepicker-month,.tk-datepicker-year{display:flex;width:var(--datepicker-items-base-size);height:var(--datepicker-items-base-size);justify-content:center;align-items:center;border-radius:var(--datepicker-items-base-radius);background:var(--static-light);color:var(--text-darkest);text-align:center;font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);font-style:normal;font-weight:400;line-height:var(--desktop-body-s-line-height);cursor:pointer}.tk-datepicker-month:hover,.tk-datepicker-year:hover{background:var(--background-lightest)}.tk-datepicker-month.selected,.tk-datepicker-year.selected{background:var(--primary-base);color:var(--text-lightest)}.tk-datepicker-month.disabled,.tk-datepicker-year.disabled{color:var(--text-sub-base);pointer-events:none}.tk-datepicker-years{display:flex;flex-direction:column;justify-content:space-between;height:100%;width:100%}.tk-datepicker-year-row{display:flex;justify-content:space-between}.tk-datepicker-footer{display:flex;padding:var(--datepicker-footer-base-v-padding) var(--datepicker-footer-base-h-padding);align-items:center;align-self:stretch}.tk-datepicker-footer.tk-datepicker-footer-divided,.tk-datepicker-footer.tk-datepicker-footer-light{border-top:1px solid var(--border-light)}.tk-datepicker-footer.tk-datepicker-footer-light{background:var(--background-lightest);border-bottom-left-radius:var(--radius-xl);border-bottom-right-radius:var(--radius-xl)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-header{width:88px;height:52px;padding:0 var(--datepicker-header-base-h-padding) 0 var(--datepicker-header-base-h-padding);border-bottom:1px solid var(--border-light);border-top-right-radius:var(--radius-xl);display:flex;justify-content:center;align-items:center}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-header.tk-datepicker-timepicker-header-only{border-left:0;border-top-left-radius:var(--radius-xl)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-header.tk-datepicker-timepicker-header-light{background:var(--background-lightest)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-header.tk-datepicker-timepicker-header-primary{background:var(--primary-base);border-bottom:1px solid var(--primary-sub-base)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-header.tk-datepicker-timepicker-header-dark{background:var(--background-darkest);border-bottom:1px solid var(--border-dark)}.tk-datepicker-timepicker-header-colon{margin:0 var(--spacing-1)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body{display:flex;gap:4px;padding:var(--datepicker-body-base-day-v-padding) var(--datepicker-body-base-day-h-padding);flex:none;height:260px;overflow-y:auto;position:relative}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body::-webkit-scrollbar{width:4px;height:4px}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-only{border-left:0;border-bottom-left-radius:var(--radius-xl);border-bottom-right-radius:var(--radius-xl)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-light{background:var(--background-lightest)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-light::before{background:var(--static-light)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-primary{background:var(--primary-base)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-primary::before{background:var(--primary-300);color:var(--primary-50);border-color:var(--primary-100)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-dark{background:var(--background-darkest)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-dark::before{background:var(--static-dark);color:var(--text-lightest);border-color:var(--border-dark)}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body.tk-datepicker-timepicker-body-4-weeks::before{top:100px}.tk-datepicker-timepicker-panel .tk-datepicker-timepicker-body::before{place-content:center;display:grid;color:var(--text-darkest);content:":";position:absolute;top:144px;left:16px;right:16px;height:40px;box-sizing:border-box;border-radius:8px;border:1px solid var(--border-light);background:var(--background-lightest);pointer-events:none;z-index:0}.tk-datepicker-timepicker-col{display:flex;flex-direction:column;align-items:center;gap:4px;flex:1}.tk-datepicker-timepicker-arrow{background:none;border:none;color:var(--icon-base);cursor:pointer;padding:8px;font-size:12px;line-height:1}.tk-datepicker-timepicker-value{font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);color:var(--text-darkest);text-align:center;width:40px;height:40px;line-height:40px;position:relative;z-index:100;cursor:pointer;font-weight:400}.tk-datepicker-timepicker-value.tk-datepicker-timepicker-value-dark{color:var(--text-lightest)}.tk-datepicker-timepicker-value.tk-datepicker-timepicker-value-primary{color:var(--primary-50)}.tk-datepicker-timepicker-value.tk-datepicker-timepicker-value-empty{visibility:hidden;height:40px;min-height:40px;line-height:40px;flex:0 0 40px}.tk-datepicker-timepicker-value.disabled{pointer-events:none;user-select:none;color:var(--text-sub-base)}.tk-datepicker-timepicker-value-selected{background:transparent;border:none;padding:0}.tk-datepicker-timepicker-separator{width:100%;height:1px;background-color:var(--border-light)}.tk-datepicker-timepicker-separator.tk-datepicker-timepicker-separator-dark{background-color:var(--border-dark)}.tk-datepicker-timepicker-separator.tk-datepicker-timepicker-separator-primary{background-color:var(--primary-300)}.tk-datepicker-loading{display:flex;align-items:center;justify-content:center;width:100%;height:260px}`;

const TkDatePicker = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.tkInputChange = index.createEvent(this, "tk-input-change", 7);
        this.tkChange = index.createEvent(this, "tk-change", 7);
        this.tkInvalid = index.createEvent(this, "tk-invalid", 7);
        if (hostRef.$hostElement$["s-ei"]) {
            this.internals = hostRef.$hostElement$["s-ei"];
        }
        else {
            this.internals = hostRef.$hostElement$.attachInternals();
            hostRef.$hostElement$["s-ei"] = this.internals;
        }
        this.today = new Date();
        this.uniqueId = v4.v4();
        this.isUpdatingAmPm = false;
        this.weeksLength = 0;
        this.currentMonth = new Date();
        this.internalSelectedDates = { start: null, end: null };
        this.inputValue = '';
        this.internalStartTime = null;
        this.internalEndTime = null;
        this.internalAmPm = 'AM';
        this.hoverDate = null;
        this.currentView = 'days';
        this.isInvalid = false;
        this.isOpen = false;
        this.concealUntilMeasured = false;
        /**
         * Defines the size for the label
         * @defaultValue base
         */
        this.size = 'base';
        /**
         * Whether the datepicker is disabled
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * Whether the datepicker is in an invalid state
         */
        this.invalid = false;
        /**
         * Whether the datepicker is read-only
         */
        this.readonly = false;
        /**
         * Indicates whether the input of datepicker can be cleared
         * @defaultValue false
         */
        this.clearable = false;
        /**
         * Represents whether the datepicker is in a loading state.
         * If true, renders a spinner inside the input and panel.
         * @defaultValue false
         */
        this.loading = false;
        /**
         * Specifies a material icon name to be displayed.
         */
        this.icon = 'calendar_month';
        /**
         * Defines the position of the icon.
         * @defaultValue left
         */
        this.iconPosition = 'left';
        /**
         * Minimum selectable date
         */
        this.minDate = '';
        /**
         * Maximum selectable date
         */
        this.maxDate = '';
        /**
         * Array of dates that are allowed to be selected. All other dates will be disabled.
         *
         * Note: Format should match dateFormat prop
         */
        this.allowedDates = [];
        /**
         * Array of dates that are disabled for selection.
         *
         * Format should match dateFormat prop
         */
        this.disabledDates = [];
        /**
         * Whether to display inline panel
         * @defaultValue false
         */
        this.inline = false;
        /**
         * Whether to require manual confirmation (Apply button) before committing changes.
         * If true, changes are only applied when apply() is called.
         * @defaultValue false
         */
        this.allowApplyButton = false;
        /**
         * The selection mode of the date picker: 'single' for single date selection, 'range' for date range selection.
         * @defaultValue single
         */
        this.mode = 'single';
        /**
         * Locale for date formatting
         * @defaultValue en
         */
        this.locale = 'en';
        /**
         * The visual variant of the footer: 'basic', 'divided', or 'light'.
         * @defaultValue basic
         */
        this.footerType = 'basic';
        /**
         * Date format pattern
         * @defaultValue yyyy-MM-dd
         */
        this.dateFormat = 'yyyy-MM-dd';
        /**
         * Whether to disable input mask
         * @defaultValue false
         */
        this.disableMask = false;
        /**
         * Header visual variant
         * @defaultValue basic
         */
        this.headerType = 'basic';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * Disabled week days (0-6, where 0 is Sunday and 6 is Saturday)
         * Example: [0,6] will disable Sunday and Saturday
         */
        this.disabledWeekDays = [];
        /**
         * Whether to show the timepicker panel next to the calendar.
         * @defaultValue false
         */
        this.showTimePicker = false;
        /**
         * Enables time-only mode. In this mode, no date selection is required and the input shows a time mask.
         * When enabled, the panel renders only the time picker and `tk-change` emits a time string (e.g. HH:mm or hh:mm a).
         * @defaultValue false
         */
        this.timeOnly = false;
        /**
         * Hour increment step.
         * @defaultValue 1
         */
        this.hourStep = 1;
        /**
         * Minute increment step.
         * @defaultValue 1
         */
        this.minuteStep = 1;
        /**
         * Time format: '12' or '24'.
         * @defaultValue '24'
         */
        this.timeFormat = '24';
        this.remeasureCalendarOnNextFrame = () => {
            if (!(this.showTimePicker && (this.inline || this.isOpen)))
                return;
            requestAnimationFrame(() => {
                const h = this.measureCalendarTableHeight();
                if (h > 0)
                    this.calendarTableHeightPx = h;
            });
        };
        /**
         * Click outside handler implementation - called by the mixin
         */
        this.closeHandler = () => {
            if (this.inline)
                return;
            this.isOpen = false;
        };
        this.handleApply = () => {
            const { start, end } = this.internalSelectedDates;
            let emitValue = null;
            if (this.timeOnly) {
                if (this.internalStartTime) {
                    const tempDate = new Date();
                    tempDate.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
                    emitValue = index$1.dateFnsExports.format(tempDate, this.getOnlyTimeFormat());
                }
            }
            else {
                const formattedStart = start ? this.formatDateOrDateTime(start, 'start') : null;
                if (this.mode === 'range') {
                    emitValue = {
                        start: formattedStart || '',
                        end: end ? this.formatDateOrDateTime(end, 'end') : undefined,
                    };
                }
                else {
                    emitValue = formattedStart;
                }
            }
            this.inputValue = this.formatInputValue();
            this.value = emitValue;
            this.tkChange.emit(emitValue);
            const isEndSelected = this.mode === 'range' ? !!this.internalSelectedDates.end : true;
            if (!this.inline && this.isOpen && !this.showTimePicker && isEndSelected) {
                this.isOpen = false;
            }
            this.isInvalid = false;
            this.currentView = 'days';
        };
        this.handleIncreaseHour = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (this.timeFormat === '12') {
                const hoursList = Array.from({ length: Math.ceil(12 / this.hourStep) }, (_, i) => i * this.hourStep + 1);
                let displayHour = targetTimeState.time.hour % 12;
                displayHour = displayHour === 0 ? 12 : displayHour;
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - displayHour) < Math.abs(prev - displayHour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const nextIdx = Math.min(idx + 1, hoursList.length - 1);
                const newDisplayHour = hoursList[nextIdx];
                // Convert back to 24-hour format
                let newHour24 = newDisplayHour === 12 ? 0 : newDisplayHour;
                if (this.internalAmPm === 'PM') {
                    newHour24 += 12;
                }
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
            }
            else {
                // 24h mode - move to next hour in step list
                const hoursList = Array.from({ length: Math.ceil(24 / this.hourStep) }, (_, i) => i * this.hourStep);
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.hour) < Math.abs(prev - targetTimeState.time.hour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const nextIdx = Math.min(idx + 1, hoursList.length - 1);
                const hour = hoursList[nextIdx];
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
            }
            this.emitTimeChange();
        };
        this.handleDecreaseHour = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (this.timeFormat === '12') {
                const hoursList = Array.from({ length: Math.ceil(12 / this.hourStep) }, (_, i) => i * this.hourStep + 1);
                let displayHour = targetTimeState.time.hour % 12;
                displayHour = displayHour === 0 ? 12 : displayHour;
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - displayHour) < Math.abs(prev - displayHour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const prevIdx = Math.max(idx - 1, 0);
                const newDisplayHour = hoursList[prevIdx];
                // Convert back to 24-hour format
                let newHour24 = newDisplayHour === 12 ? 0 : newDisplayHour;
                if (this.internalAmPm === 'PM') {
                    newHour24 += 12;
                }
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
            }
            else {
                // 24h mode - move to previous hour in step list
                const hoursList = Array.from({ length: Math.ceil(24 / this.hourStep) }, (_, i) => i * this.hourStep);
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.hour) < Math.abs(prev - targetTimeState.time.hour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const prevIdx = Math.max(idx - 1, 0);
                const hour = hoursList[prevIdx];
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
            }
            this.emitTimeChange();
        };
        this.handleHourClick = (hour) => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
            }
            this.emitTimeChange();
        };
        this.handleIncreaseMinute = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            const minutesList = Array.from({ length: Math.ceil(60 / this.minuteStep) }, (_, i) => i * this.minuteStep);
            // Find closest minute in step list
            const closestMinute = minutesList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.minute) < Math.abs(prev - targetTimeState.time.minute) ? curr : prev));
            const idx = minutesList.indexOf(closestMinute);
            const nextIdx = Math.min(idx + 1, minutesList.length - 1);
            const minute = minutesList[nextIdx];
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            this.emitTimeChange();
        };
        this.handleDecreaseMinute = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            const minutesList = Array.from({ length: Math.ceil(60 / this.minuteStep) }, (_, i) => i * this.minuteStep);
            // Find closest minute in step list
            const closestMinute = minutesList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.minute) < Math.abs(prev - targetTimeState.time.minute) ? curr : prev));
            const idx = minutesList.indexOf(closestMinute);
            const prevIdx = Math.max(idx - 1, 0);
            const minute = minutesList[prevIdx];
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            this.emitTimeChange();
        };
        this.handleMinuteClick = (min) => {
            if (this.disabled || this.readonly)
                return;
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: min });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: min });
            }
            this.emitTimeChange();
        };
        this.handleAmPmToggle = (event) => {
            // Prevent inner toggle group's tk-change from leaking outside and overwriting consumer value
            if (this.disabled || this.readonly)
                return;
            event.stopPropagation();
            if (this.timeFormat !== '12')
                return;
            const newAmPm = event.detail;
            // Just update the state - the watcher will handle the time conversion
            this.internalAmPm = newAmPm;
        };
        this.handleInputBlur = () => {
            if ((this.timeOnly || this.showTimePicker) && this.timeFormat === '12' && this.internalStartTime) {
                this.internalAmPm = this.internalStartTime.hour >= 12 ? 'PM' : 'AM';
                this.inputValue = this.formatInputValue();
            }
        };
        this.handleInputClick = (e) => {
            if (this.disabled || this.readonly) {
                e.preventDefault();
                return;
            }
            this.isOpen = !this.isOpen;
            this.remeasureCalendarOnNextFrame();
        };
        this.handleDateClick = (date) => {
            if (this.disabled || this.readonly || this.isDateDisabled(date))
                return;
            const normalizedDate = this.normalizeDate(date);
            const defaultTime = this.getDefaultTime();
            if (this.mode === 'single') {
                this.internalSelectedDates = { start: normalizedDate, end: null };
                if (this.showTimePicker) {
                    if (!this.internalStartTime) {
                        this.internalStartTime = defaultTime;
                        if (this.timeFormat === '12')
                            this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
                    }
                }
                else {
                    this.internalStartTime = null;
                }
                this.internalEndTime = this.internalStartTime;
                if (!this.inline && !this.showTimePicker && !this.allowApplyButton) {
                    this.isOpen = false;
                }
            }
            else if (this.mode === 'range') {
                const { start, end } = this.internalSelectedDates;
                if (!start || (start && end)) {
                    this.internalSelectedDates = { start: normalizedDate, end: null };
                    if (this.showTimePicker) {
                        if (!this.internalStartTime) {
                            this.internalStartTime = defaultTime;
                            if (this.timeFormat === '12')
                                this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
                        }
                        this.internalEndTime = null;
                    }
                    else {
                        this.internalStartTime = null;
                        this.internalEndTime = null;
                    }
                    this.hoverDate = null;
                }
                else {
                    let newStart;
                    let newEnd;
                    if (normalizedDate < start) {
                        newStart = normalizedDate;
                        newEnd = start;
                        if (this.showTimePicker) {
                            const tempTime = this.internalStartTime;
                            this.internalStartTime = this.internalEndTime || this.internalStartTime || defaultTime;
                            this.internalEndTime = tempTime || defaultTime;
                        }
                    }
                    else {
                        newStart = start;
                        newEnd = normalizedDate;
                        if (this.showTimePicker && !this.internalEndTime) {
                            this.internalEndTime = this.internalStartTime || defaultTime;
                        }
                    }
                    this.internalSelectedDates = { start: newStart, end: newEnd };
                    this.hoverDate = null;
                    if (!this.inline && !this.showTimePicker && !this.allowApplyButton) {
                        this.isOpen = false;
                    }
                }
            }
            // if swap occurred, bail
            if (!this.ensureRangeOrder()) {
                return;
            }
            this.remeasureCalendarOnNextFrame();
            if (!this.allowApplyButton) {
                this.handleApply();
            }
            this.isInvalid = false;
        };
        this.handleInputKeyDown = (event) => {
            if (this.timeOnly)
                return; // let time mask/type
            if (this.disableMask || this.mode === 'range') {
                event.preventDefault();
            }
        };
        this.handleInputChange = (event) => {
            if (this.disableMask || this.mode === 'range') {
                event.preventDefault();
                return;
            }
            this.remeasureCalendarOnNextFrame();
            this.inputValue = event.detail;
            this.tkInputChange.emit(this.inputValue);
        };
        this.handleInputClearClick = () => {
            if (this.clearable) {
                this.inputValue = '';
                this.internalSelectedDates = { start: null, end: null };
                this.internalStartTime = null;
                this.internalEndTime = null;
                this.tkChange.emit(null);
            }
        };
        this.handleInputInput = () => {
            if (this.disableMask || this.mode === 'range')
                return;
            clearTimeout(this.debounceTimer);
            this.debounceTimer = window.setTimeout(() => {
                if (this.inputValue) {
                    if (this.timeOnly) {
                        const parsedTime = this.parseTimeString(this.inputValue);
                        if (parsedTime) {
                            // min/maxTime validation
                            if (this.isTimeDisabled(parsedTime.getHours(), parsedTime.getMinutes())) {
                                this.isInvalid = true;
                                this.tkChange.emit(undefined);
                                return;
                            }
                            this.internalStartTime = { hour: parsedTime.getHours(), minute: parsedTime.getMinutes() };
                            this.internalEndTime = this.internalStartTime;
                            this.isInvalid = false;
                            if (!this.allowApplyButton) {
                                this.tkChange.emit(index$1.dateFnsExports.format(parsedTime, this.getOnlyTimeFormat()));
                            }
                        }
                        else if (!this.allowApplyButton) {
                            this.isInvalid = true;
                            this.tkInvalid.emit({ message: 'Invalid time format', value: this.inputValue });
                            this.tkChange.emit(undefined);
                        }
                    }
                    else {
                        const parser = this.showTimePicker ? this.parseFullDateTime.bind(this) : this.parseInputDate.bind(this);
                        const parsedDate = parser(this.inputValue);
                        if (parsedDate && !this.isDateDisabled(parsedDate)) {
                            // min/maxTime validation for timepicker mode
                            if (this.showTimePicker && this.isTimeDisabled(parsedDate.getHours(), parsedDate.getMinutes())) {
                                this.isInvalid = true;
                                this.tkChange.emit(undefined);
                                return;
                            }
                            const normalized = this.normalizeDate(parsedDate);
                            this.internalSelectedDates = {
                                start: normalized,
                                end: null,
                            };
                            let formattedValue;
                            if (this.showTimePicker) {
                                const time = { hour: parsedDate.getHours(), minute: parsedDate.getMinutes() };
                                this.internalStartTime = time;
                                this.internalEndTime = time;
                                formattedValue = index$1.dateFnsExports.format(parsedDate, this.getFullDateTimeFormat());
                            }
                            else {
                                this.internalStartTime = null;
                                this.internalEndTime = null;
                                formattedValue = this.formatDateOrDateTime(parsedDate, 'start');
                            }
                            this.isInvalid = false;
                            this.tkChange.emit(formattedValue);
                        }
                        else if (!this.allowApplyButton) {
                            this.isInvalid = true;
                            this.tkInvalid.emit({ message: 'Invalid date format', value: this.inputValue });
                        }
                    }
                }
                else {
                    this.isInvalid = false;
                    this.internalSelectedDates = { start: null, end: null };
                    this.internalStartTime = null;
                    this.internalEndTime = null;
                    if (!this.allowApplyButton) {
                        this.tkChange.emit(undefined);
                    }
                }
            }, 300);
        };
        this.handleDateHover = (date) => {
            if (this.mode === 'range' && this.internalSelectedDates.start && !this.internalSelectedDates.end) {
                this.hoverDate = date;
            }
            this.remeasureCalendarOnNextFrame();
        };
        this.handleMonthChange = (increment) => {
            this.currentMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() + increment);
            this.remeasureCalendarOnNextFrame();
        };
        this.handleYearChange = (increment) => {
            if (this.currentView === 'years') {
                this.currentMonth = new Date(this.currentMonth.getFullYear() + increment * 12, this.currentMonth.getMonth());
                return;
            }
            this.currentMonth = new Date(this.currentMonth.getFullYear() + increment, this.currentMonth.getMonth());
            this.remeasureCalendarOnNextFrame();
        };
        this.handleViewChange = (e, view) => {
            e.stopPropagation();
            this.currentView = view;
            this.remeasureCalendarOnNextFrame();
        };
    }
    internalAmPmChanged(newValue) {
        this.updateTimeBasedOnAmPm(newValue);
    }
    isOpenChanged(newValue) {
        if (!this.inline) {
            if (newValue) {
                if (this.internalSelectedDates.start) {
                    this.currentMonth = new Date(this.internalSelectedDates.start.getFullYear(), this.internalSelectedDates.start.getMonth());
                }
                if (this.showTimePicker) {
                    this.concealUntilMeasured = true;
                    requestAnimationFrame(() => {
                        const h = this.measureCalendarTableHeight();
                        if (h > 0)
                            this.calendarTableHeightPx = h;
                        this.concealUntilMeasured = false;
                    });
                }
            }
            else {
                // Kullanıcı değişikliklerini uygulamadan paneli kapatırsa, geçici seçimleri temizle
                if (this.allowApplyButton) {
                    this.processDateValue(this.value, true);
                }
                this.currentView = 'days';
                this.calendarTableHeightPx = undefined;
                this.concealUntilMeasured = false;
            }
        }
    }
    valueChanged(newValue, oldValue) {
        if (JSON.stringify(newValue) === JSON.stringify(oldValue)) {
            return;
        }
        this.processDateValue(newValue, true);
        this.remeasureCalendarOnNextFrame();
    }
    dateFormatChanged() {
        if (this.timeOnly)
            return;
        this.updateMaskOptions();
    }
    showTimePickerChanged() {
        this.updateMaskOptions();
    }
    timeOnlyChanged() {
        // Update mask according to the active mode
        this.updateMaskOptions();
    }
    timeFormatChanged() {
        if (this.timeOnly || this.showTimePicker) {
            this.updateMaskOptions();
            // Sync AM/PM with the current hour when switching to 12h
            if (this.timeFormat === '12' && this.internalStartTime) {
                this.internalAmPm = this.internalStartTime.hour >= 12 ? 'PM' : 'AM';
            }
        }
    }
    componentWillLoad() {
        this.updateMaskOptions();
        if (this.allowedDates) {
            this.allowedDates = this.allowedDates.filter(date => {
                const parsed = this.parseInputDate(date);
                return parsed !== null;
            });
        }
        if (this.disabledDates) {
            this.disabledDates = this.disabledDates.filter(date => {
                const parsed = this.parseInputDate(date);
                return parsed !== null;
            });
        }
        this.initializeDates();
        this.hasFooterSlot = !!this.el.querySelector('[slot="footer"]');
        this.hasFooterActionsSlot = !!this.el.querySelector('[slot="footer-actions"]');
    }
    componentDidLoad() {
        var _a, _b, _c;
        (_b = (_a = this.internals) === null || _a === void 0 ? void 0 : _a.form) === null || _b === void 0 ? void 0 : _b.addEventListener('reset', () => {
            this.handleFormReset();
        });
        // Initialize click outside mixin only if not inline mode
        if (!this.inline) {
            const tkInputArea = (_c = this.inputRef) === null || _c === void 0 ? void 0 : _c.querySelector('.tk-input');
            this.clickOutsideMixin = new clickoutsideMixin.ClickOutsideMixin({
                referenceElement: tkInputArea !== null && tkInputArea !== void 0 ? tkInputArea : this.el,
                handler: this.closeHandler,
                disabled: this.disabled || this.readonly || !this.isOpen,
            });
        }
        if (this.inline && this.showTimePicker) {
            requestAnimationFrame(() => {
                const h = this.measureCalendarTableHeight();
                if (h > 0)
                    this.calendarTableHeightPx = h;
            });
        }
    }
    componentDidUpdate() {
        var _a, _b, _c;
        // Update click outside mixin configuration based on current state
        (_a = this.clickOutsideMixin) === null || _a === void 0 ? void 0 : _a.updateConfig({
            disabled: this.disabled || this.readonly || this.inline || !this.isOpen,
            ignoredElements: this.panelRef ? [this.panelRef] : [],
        });
        if (this.isOpen) {
            if (this.inputRef && this.panelRef) {
                // Clean up old floating UI listeners before setting up new ones
                (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
                this.updatePosition();
            }
        }
        else {
            // Remove floating UI listeners when datepicker closes
            (_c = this.cleanup) === null || _c === void 0 ? void 0 : _c.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
        }
    }
    disconnectedCallback() {
        var _a, _b, _c, _d;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        (_c = (_b = this.internals) === null || _b === void 0 ? void 0 : _b.form) === null || _c === void 0 ? void 0 : _c.removeEventListener('reset', this.handleFormReset);
        // Call mixin's disconnectedCallback for cleanup
        (_d = this.clickOutsideMixin) === null || _d === void 0 ? void 0 : _d.disconnectedCallback();
    }
    formResetCallback() {
        this.handleFormReset();
    }
    /**
     * Sets the date to today
     */
    async setToday() {
        const today = this.normalizeDate(new Date());
        this.currentMonth = today;
        this.internalSelectedDates = { start: today, end: null };
        if (this.showTimePicker) {
            const defaultTime = this.getDefaultTime();
            this.internalStartTime = defaultTime;
            this.internalEndTime = null;
            if (this.timeFormat === '12')
                this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
        }
        else {
            this.internalStartTime = null;
            this.internalEndTime = null;
        }
    }
    /**
     * Applies the current internal selection and emits tk-change
     */
    async apply() {
        this.handleApply();
    }
    /**
     * Closes the datepicker panel if it is open.
     */
    async closePanel() {
        if (this.isOpen) {
            this.isOpen = false;
        }
    }
    updatePosition() {
        const inputEl = this.inputRef.querySelector('.tk-input');
        this.cleanup = positionUtils.floatingElementAutoUpdate(inputEl, this.panelRef, undefined, {
            placement: 'bottom-start',
        });
    }
    isTimeDisabled(hour, minute) {
        const toTotalMinutes = (h, m) => h * 60 + m;
        let testHour = hour;
        if (this.timeFormat === '12') {
            // 12h format - convert to 24h for comparison
            testHour = (hour === 12 ? 0 : hour) + (this.internalAmPm === 'PM' ? 12 : 0);
        }
        const total = toTotalMinutes(testHour, minute);
        const parseBound = (val) => {
            if (!val)
                return null;
            // Try parsing with current timeFormat and AM/PM context
            let parsed = null;
            if (this.timeFormat === '12' && /am|pm|AM|PM/i.test(val)) {
                // If AM/PM present, parse as 12h
                parsed = this.parseTimeString(val);
            }
            else if (this.timeFormat === '12') {
                // If not present, append current AM/PM
                parsed = this.parseTimeString(val + ' ' + this.internalAmPm);
            }
            else {
                // 24h mode
                parsed = this.parseTimeString(val);
            }
            if (!parsed)
                return null;
            return toTotalMinutes(parsed.getHours(), parsed.getMinutes());
        };
        const minBound = parseBound(this.minTime);
        const maxBound = parseBound(this.maxTime);
        if (minBound !== null && total < minBound)
            return true;
        if (maxBound !== null && total > maxBound)
            return true;
        return false;
    }
    updateTimeBasedOnAmPm(newAmPm) {
        if (this.timeFormat !== '12' || this.isUpdatingAmPm) {
            return;
        }
        this.isUpdatingAmPm = true;
        const targetTimeState = this.getTimeStateToModify();
        if (!targetTimeState) {
            this.isUpdatingAmPm = false;
            return;
        }
        // Convert current hour based on AM/PM change
        let currentHour = targetTimeState.time.hour;
        let needsUpdate = false;
        if (newAmPm === 'PM' && currentHour < 12) {
            currentHour += 12;
            needsUpdate = true;
        }
        else if (newAmPm === 'AM' && currentHour >= 12) {
            currentHour -= 12;
            needsUpdate = true;
        }
        // Only update if the hour actually changed
        if (needsUpdate) {
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: currentHour });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: currentHour });
            }
            this.emitTimeChange();
        }
        this.isUpdatingAmPm = false;
    }
    getResolvedFirstDayIndex() {
        if (this.firstDayOfWeekIndex !== undefined && this.firstDayOfWeekIndex !== null) {
            if (this.firstDayOfWeekIndex >= 0 && this.firstDayOfWeekIndex <= 6) {
                return this.firstDayOfWeekIndex;
            }
            else {
                console.warn(`Invalid firstDayOfWeekIndex value: ${this.firstDayOfWeekIndex}. Defaulting to 0 (Monday).`);
                return 0;
            }
        }
        try {
            const resolvedLocale = this.locale || 'en';
            if (typeof Intl !== 'undefined' && Intl.Locale && typeof Intl.Locale === 'function' && Intl.Locale.prototype.hasOwnProperty('weekInfo')) {
                // @ts-ignore: Accessing weekInfo from potentially unknown Intl.Locale type
                const localeInfo = new Intl.Locale(resolvedLocale).getWeekInfo();
                if (localeInfo && localeInfo.firstDay !== undefined) {
                    // Formula: (IntlDay + 6) % 7 maps 1(Mon) to 0, 7(Sun) to 6
                    return (localeInfo.firstDay + 6) % 7;
                }
            }
        }
        catch (e) {
            console.warn(`Could not determine first day of week from locale '${this.locale}'. Error or unsupported: ${e}`);
        }
        return 0;
    }
    getFullDateTimeFormat() {
        const timePattern = this.timeFormat === '12' ? 'hh:mm a' : 'HH:mm';
        return /[hH]/.test(this.dateFormat) ? this.dateFormat : `${this.dateFormat} ${timePattern}`;
    }
    getOnlyTimeFormat() {
        return this.timeFormat === '12' ? 'hh:mm a' : 'HH:mm';
    }
    getDefaultTime() {
        const now = new Date();
        let hour = now.getHours();
        let minute = now.getMinutes();
        // Align hour to the configured hourStep (floor to nearest step)
        const hourStep = Math.max(1, this.hourStep || 1);
        hour = Math.floor(hour / hourStep) * hourStep;
        // Align minutes to the configured step (floor to nearest step)
        const minuteStep = Math.max(1, this.minuteStep || 1);
        minute = Math.floor(minute / minuteStep) * minuteStep;
        if (!this.isTimeDisabled(hour, minute)) {
            return { hour, minute };
        }
        const hours = this.timeFormat === '12'
            ? Array.from({ length: Math.ceil(12 / hourStep) }, (_, i) => Math.min(i * hourStep + 1, 12))
            : Array.from({ length: Math.ceil(24 / hourStep) }, (_, i) => i * hourStep);
        const minutes = Array.from({ length: Math.ceil(60 / minuteStep) }, (_, i) => i * minuteStep);
        for (const h of hours) {
            for (const m of minutes) {
                if (!this.isTimeDisabled(h, m)) {
                    return { hour: h, minute: m };
                }
            }
        }
        // No valid time found, log warning
        console.warn('TkDatepicker: No valid time slots available for the current configuration.');
        return { hour, minute };
    }
    getDateWithTime(date, type) {
        if (!date)
            return null;
        const newDate = new Date(date);
        const timeToApply = type === 'start' ? this.internalStartTime : this.internalEndTime;
        if (this.showTimePicker && timeToApply) {
            newDate.setHours(timeToApply.hour, timeToApply.minute);
        }
        return newDate;
    }
    getMaskOptionsFromDateFormat(format) {
        var _a;
        const delimiter = ((_a = format.match(/[^a-zA-Z]/)) === null || _a === void 0 ? void 0 : _a[0]) || '';
        const datePattern = [];
        const blockSizes = [];
        const parts = format.split(/[^a-zA-Z]/);
        parts.forEach(part => {
            switch (part.toLowerCase()) {
                case 'yyyy':
                    datePattern.push('Y');
                    blockSizes.push(4);
                    break;
                case 'yy':
                    datePattern.push('Y');
                    blockSizes.push(2);
                    break;
                case 'mm':
                case 'm':
                    datePattern.push('m');
                    blockSizes.push(2);
                    break;
                case 'dd':
                case 'd':
                    datePattern.push('d');
                    blockSizes.push(2);
                    break;
            }
        });
        const dateDelimiters = Array(blockSizes.length - 1).fill(delimiter);
        if (this.showTimePicker) {
            this.maskOptions = {
                blocks: [...blockSizes, 2, 2], // date blocks + HH:MM
                delimiters: [...dateDelimiters, ' ', ':'],
                numericOnly: true,
            };
        }
        else {
            this.maskOptions = {
                date: true,
                delimiter,
                datePattern,
            };
        }
    }
    processDateValue(value, updateCurrentMonth = false) {
        if (this.timeOnly) {
            // In time-only mode, value is expected to be a time string (e.g., HH:mm or hh:mm a)
            let startTime = null;
            if (typeof value === 'string' && value) {
                const parsed = this.parseTimeString(value);
                if (parsed) {
                    startTime = { hour: parsed.getHours(), minute: parsed.getMinutes() };
                }
            }
            // For initial load, set AM/PM based on actual time, but respect user changes after that
            if (this.timeFormat === '12' && startTime) {
                // Only set if this is the initial default state, otherwise respect user choice
                if (this.internalAmPm === 'AM' && startTime.hour >= 12) {
                    this.internalAmPm = 'PM';
                }
            }
            this.internalStartTime = startTime;
            this.internalEndTime = startTime;
            // Do not set any dates in time-only mode
            this.internalSelectedDates = { start: null, end: null };
            this.inputValue = this.formatInputValue();
            return;
        }
        let startDate = null;
        let endDate = null;
        let startTime = null;
        let endTime = null;
        const now = new Date();
        const defaultTime = this.getDefaultTime();
        if (value) {
            let startString = null;
            let endString = null;
            if (typeof value === 'string') {
                startString = value;
            }
            else if (typeof value === 'object' && value.start) {
                startString = value.start;
                endString = value.end || null;
            }
            if (startString) {
                const parsedStartDateTime = this.showTimePicker ? this.parseFullDateTime(startString) : this.parseInputDate(startString);
                if (parsedStartDateTime && !this.isDateDisabled(parsedStartDateTime)) {
                    startDate = this.normalizeDate(parsedStartDateTime);
                    if (this.showTimePicker) {
                        const expectedFullFormat = this.formatFullDateTime(parsedStartDateTime);
                        if (startString === expectedFullFormat || startString.includes(' ')) {
                            startTime = { hour: parsedStartDateTime.getHours(), minute: parsedStartDateTime.getMinutes() };
                        }
                        else {
                            startTime = defaultTime;
                        }
                    }
                }
                else {
                    startDate = null;
                    startTime = null;
                }
            }
            if (this.mode === 'range' && endString && startDate) {
                const parsedEndDateTime = this.showTimePicker ? this.parseFullDateTime(endString) : this.parseInputDate(endString);
                if (parsedEndDateTime && !this.isDateDisabled(parsedEndDateTime)) {
                    endDate = this.normalizeDate(parsedEndDateTime);
                    if (this.showTimePicker) {
                        const expectedFullFormat = this.formatFullDateTime(parsedEndDateTime);
                        if (endString === expectedFullFormat || endString.includes(' ')) {
                            endTime = { hour: parsedEndDateTime.getHours(), minute: parsedEndDateTime.getMinutes() };
                        }
                        else {
                            endTime = startTime || defaultTime;
                        }
                    }
                }
                else {
                    endDate = null;
                    endTime = null;
                }
            }
            else if (this.mode === 'range') {
                endDate = null;
                endTime = null;
            }
            else {
                endDate = null;
                endTime = null;
            }
        }
        else {
            if (this.showTimePicker) {
                startTime = defaultTime;
                endTime = this.mode === 'range' ? null : defaultTime;
            }
        }
        this.internalSelectedDates = { start: startDate, end: endDate };
        this.internalStartTime = startTime;
        this.internalEndTime = this.mode === 'range' ? endTime : startTime;
        if (updateCurrentMonth && startDate) {
            this.currentMonth = new Date(startDate.getFullYear(), startDate.getMonth());
        }
        else if (!startDate) {
            this.currentMonth = new Date(this.currentMonth || now);
        }
        this.inputValue = this.formatInputValue();
    }
    initializeDates() {
        var _a;
        // Initialize currentMonth with priority: defaultDate > first allowedDate month > first non-disabled month > now
        const now = new Date();
        const parsedDefaultDate = this.parseDefaultDate(this.defaultDate);
        if (parsedDefaultDate) {
            this.currentMonth = parsedDefaultDate;
        }
        else if (this.allowedDates && this.allowedDates.length > 0) {
            this.currentMonth = (_a = this.getFirstAvailableAllowedMonth()) !== null && _a !== void 0 ? _a : now;
        }
        else {
            let candidateMonth = now;
            if (this.disabledDates && this.disabledDates.length > 0) {
                let limit = 12; // Prevent infinite loop by limiting to 12 months of searching
                while (this.isMonthFullyDisabled(candidateMonth.getFullYear(), candidateMonth.getMonth()) && limit > 0) {
                    candidateMonth = new Date(candidateMonth.getFullYear(), candidateMonth.getMonth() + 1, 1);
                    limit--;
                }
            }
            this.currentMonth = candidateMonth;
        }
        this.processDateValue(this.value, true);
        if (this.showTimePicker && !this.internalStartTime && this.internalSelectedDates.start) {
            const defaultTime = this.getDefaultTime();
            this.internalStartTime = defaultTime;
            if (this.timeFormat === '12')
                this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
            if (this.mode === 'range' && !this.internalEndTime && this.internalSelectedDates.end) {
                this.internalEndTime = this.internalStartTime;
            }
            else if (this.mode === 'single') {
                this.internalEndTime = this.internalStartTime;
            }
        }
        this.inputValue = this.formatInputValue();
    }
    measureCalendarTableHeight() {
        var _a;
        const table = (_a = this.panelRef) === null || _a === void 0 ? void 0 : _a.querySelector('.tk-datepicker-table');
        return table ? Math.ceil(table.getBoundingClientRect().height) : 0;
    }
    normalizeDate(date) {
        return new Date(date.getFullYear(), date.getMonth(), date.getDate());
    }
    formatDate(date) {
        return index$1.dateFnsExports.format(date, this.dateFormat);
    }
    formatFullDateTime(date) {
        if (!date)
            return '';
        return index$1.dateFnsExports.format(date, this.getFullDateTimeFormat());
    }
    formatDateOrDateTime(date, type) {
        if (this.timeOnly) {
            if (!this.internalStartTime)
                return '';
            const temp = new Date();
            temp.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
            return index$1.dateFnsExports.format(temp, this.getOnlyTimeFormat());
        }
        if (this.showTimePicker && date && type) {
            const dateWithCorrectTime = this.getDateWithTime(date, type);
            return dateWithCorrectTime ? this.formatFullDateTime(dateWithCorrectTime) : '';
        }
        else if (date) {
            return this.formatDate(date);
        }
        return '';
    }
    parseInputDate(dateString) {
        const parsedDate = index$1.dateFnsExports.parse(dateString, this.dateFormat, new Date());
        if (index$1.dateFnsExports.isValid(parsedDate) && this.formatDate(parsedDate) === dateString) {
            return parsedDate;
        }
        return null;
    }
    parseFullDateTime(dateTimeString) {
        // Try full date+time format first
        const formatString = this.getFullDateTimeFormat();
        const parsedDate = index$1.dateFnsExports.parse(dateTimeString, formatString, new Date());
        if (index$1.dateFnsExports.isValid(parsedDate) && index$1.dateFnsExports.format(parsedDate, formatString) === dateTimeString) {
            return parsedDate;
        }
        // Fallback: allow typing HH:mm without AM/PM for 12h mode
        if (this.timeFormat === '12') {
            const altFormat = `${this.dateFormat} HH:mm`;
            const parsedAlt = index$1.dateFnsExports.parse(dateTimeString, altFormat, new Date());
            if (index$1.dateFnsExports.isValid(parsedAlt) && index$1.dateFnsExports.format(parsedAlt, altFormat) === dateTimeString) {
                return parsedAlt;
            }
        }
        return null;
    }
    parseTimeString(timeString) {
        const base = new Date();
        const primaryFmt = this.getOnlyTimeFormat();
        let parsed = index$1.dateFnsExports.parse(timeString, primaryFmt, base);
        if (index$1.dateFnsExports.isValid(parsed) && index$1.dateFnsExports.format(parsed, primaryFmt) === timeString) {
            return parsed;
        }
        // Fallbacks
        // 1) In 12h mode, allow typing time without AM/PM (e.g., "11:00").
        //    We interpret it using the current AM/PM selection.
        if (this.timeFormat === '12') {
            const simpleMatch = timeString.match(/^(\d{1,2}):(\d{2})$/);
            if (simpleMatch) {
                let hour = parseInt(simpleMatch[1], 10);
                const minute = parseInt(simpleMatch[2], 10);
                if (Number.isFinite(hour) && Number.isFinite(minute) && minute >= 0 && minute <= 59) {
                    if (this.internalAmPm === 'PM' && hour < 12)
                        hour += 12;
                    if (this.internalAmPm === 'AM' && hour === 12)
                        hour = 0;
                    const t = new Date(base);
                    t.setHours(Math.max(0, Math.min(23, hour)), minute, 0, 0);
                    return t;
                }
            }
            // Also accept explicit 24h input (HH:mm) and convert to a Date directly
            const alt24 = index$1.dateFnsExports.parse(timeString, 'HH:mm', base);
            if (index$1.dateFnsExports.isValid(alt24) && index$1.dateFnsExports.format(alt24, 'HH:mm') === timeString) {
                return alt24;
            }
        }
        return null;
    }
    isDateDisabled(date) {
        const normalizedDate = this.normalizeDate(date);
        const dateStr = this.formatDate(normalizedDate);
        const minDate = this.minDate ? this.parseInputDate(this.minDate) : null;
        const maxDate = this.maxDate ? this.parseInputDate(this.maxDate) : null;
        if (minDate) {
            minDate.setHours(0, 0, 0, 0);
        }
        if (maxDate) {
            maxDate.setHours(23, 59, 59, 999);
        }
        const isOutOfRange = (minDate && date < minDate) || (maxDate && date > maxDate);
        if (isOutOfRange)
            return true;
        if (this.allowedDates && this.allowedDates.length > 0) {
            const isAllowed = this.allowedDates.some(allowedDate => {
                const parsedAllowedDate = this.parseInputDate(allowedDate);
                return parsedAllowedDate && this.formatDate(this.normalizeDate(parsedAllowedDate)) === dateStr;
            });
            return !isAllowed;
        }
        if (this.disabledDates && this.disabledDates.length > 0) {
            return this.disabledDates.some(disabledDate => {
                const parsedDisabledDate = this.parseInputDate(disabledDate);
                return parsedDisabledDate && this.formatDate(this.normalizeDate(parsedDisabledDate)) === dateStr;
            });
        }
        if (this.disabledWeekDays && this.disabledWeekDays.length > 0) {
            const dayOfWeek = date.getDay();
            return this.disabledWeekDays.includes(dayOfWeek);
        }
        return false;
    }
    isMonthFullyDisabled(year, month) {
        // Check if the entire month has all days disabled
        const daysInMonth = new Date(year, month + 1, 0).getDate();
        for (let day = 1; day <= daysInMonth; day++) {
            const date = new Date(year, month, day);
            if (!this.isDateDisabled(date)) {
                return false; // Found at least one enabled date
            }
        }
        return true; // All days in the month are disabled
    }
    parseDefaultDate(value) {
        if (!value)
            return null;
        const now = new Date();
        const delimiters = ['.', '/', '_', '-'];
        const delimiterRegex = new RegExp(`[${delimiters.join('')}]`);
        const parts = value.trim().split(delimiterRegex);
        if (parts.length === 2) {
            const [first, second] = parts;
            if (/^\d{4}$/.test(first) && /^\d{2}$/.test(second)) {
                // yyyy-MM
                const year = Number.parseInt(first, 10);
                const month = Number.parseInt(second, 10);
                if (month >= 1 && month <= 12)
                    return new Date(year, month - 1, 1);
            }
            else if (/^\d{2}$/.test(first) && /^\d{4}$/.test(second)) {
                // MM-yyyy
                const month = Number.parseInt(first, 10);
                const year = Number.parseInt(second, 10);
                if (month >= 1 && month <= 12)
                    return new Date(year, month - 1, 1);
            }
            return null;
        }
        // (yyyy)
        if (parts.length === 1 && /^\d{4}$/.test(parts[0])) {
            const year = Number.parseInt(parts[0], 10);
            return new Date(year, now.getMonth(), 1);
        }
        //  (MM)
        if (parts.length === 1 && /^\d{2}$/.test(parts[0])) {
            const month = Number.parseInt(parts[0], 10);
            if (month >= 1 && month <= 12)
                return new Date(now.getFullYear(), month - 1, 1);
        }
        return null;
    }
    getFirstAvailableAllowedMonth() {
        var _a;
        if (!((_a = this.allowedDates) === null || _a === void 0 ? void 0 : _a.length))
            return null;
        const parsedAllowedDates = this.allowedDates.map(date => this.parseInputDate(date)).filter((date) => date !== null);
        if (!parsedAllowedDates.length)
            return null;
        const minDate = parsedAllowedDates.reduce((min, curr) => (curr.getTime() < min.getTime() ? curr : min), parsedAllowedDates[0]);
        return new Date(minDate.getFullYear(), minDate.getMonth(), 1);
    }
    formatInputValue() {
        if (this.timeOnly) {
            if (this.internalStartTime) {
                const temp = new Date();
                temp.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
                return index$1.dateFnsExports.format(temp, this.getOnlyTimeFormat());
            }
            return '';
        }
        const { start, end } = this.internalSelectedDates;
        if (start) {
            const formattedStart = this.formatDateOrDateTime(start, 'start');
            if (this.mode === 'range' && end) {
                const formattedEnd = this.formatDateOrDateTime(end, 'end');
                return `${formattedStart} - ${formattedEnd}`;
            }
            return formattedStart;
        }
        return '';
    }
    isToday(date) {
        return date.getDate() === this.today.getDate() && date.getMonth() === this.today.getMonth() && date.getFullYear() === this.today.getFullYear();
    }
    ensureDateTimeInitialized(type) {
        if (this.timeOnly) {
            // In time-only mode, prefer parsing the current input value instead of resetting to default
            if (!this.internalStartTime) {
                let parsedFromInput = null;
                if (this.inputValue) {
                    parsedFromInput = this.parseTimeString(this.inputValue);
                }
                const resolvedTime = parsedFromInput ? { hour: parsedFromInput.getHours(), minute: parsedFromInput.getMinutes() } : this.getDefaultTime();
                this.internalStartTime = resolvedTime;
                this.internalEndTime = resolvedTime;
                if (this.timeFormat === '12')
                    this.internalAmPm = resolvedTime.hour >= 12 ? 'PM' : 'AM';
            }
            return;
        }
        if (!this.internalSelectedDates.start && type === 'start') {
            const today = this.normalizeDate(new Date());
            this.internalSelectedDates = Object.assign(Object.assign({}, this.internalSelectedDates), { start: today });
            this.currentMonth = new Date(today);
        }
        if (!this.internalSelectedDates.end && type === 'end' && this.mode === 'range') {
            return; // Don't initialize end date if not set
        }
        if (this.showTimePicker) {
            const defaultTime = this.getDefaultTime();
            if (type === 'start' && !this.internalStartTime) {
                this.internalStartTime = defaultTime;
                if (this.timeFormat === '12')
                    this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
            }
            if (type === 'end' && this.mode === 'range' && this.internalSelectedDates.end && !this.internalEndTime) {
                this.internalEndTime = this.internalStartTime || defaultTime;
            }
        }
    }
    ensureRangeOrder() {
        var _a, _b;
        if (this.mode !== 'range' || !this.showTimePicker)
            return true;
        const { start, end } = this.internalSelectedDates;
        if (!start || !end)
            return true;
        const startTs = (_a = this.getDateWithTime(start, 'start')) === null || _a === void 0 ? void 0 : _a.getTime();
        const endTs = (_b = this.getDateWithTime(end, 'end')) === null || _b === void 0 ? void 0 : _b.getTime();
        if (startTs !== undefined && endTs !== undefined && endTs < startTs) {
            this.internalSelectedDates = { start: end, end: start };
            [this.internalStartTime, this.internalEndTime] = [this.internalEndTime, this.internalStartTime];
            this.inputValue = this.formatInputValue();
            this.tkChange.emit({
                start: this.formatDateOrDateTime(end, 'start'),
                end: this.formatDateOrDateTime(start, 'end'),
            });
            return false;
        }
        return true;
    }
    emitTimeChange() {
        // Apply buttona basılmadıkça anlık değişiklikleri emit etmesin diye
        if (this.allowApplyButton)
            return;
        if (this.timeOnly) {
            if (!this.internalStartTime)
                return;
            const temp = new Date();
            temp.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
            const value = index$1.dateFnsExports.format(temp, this.getOnlyTimeFormat());
            this.tkChange.emit(value);
            this.inputValue = this.formatInputValue();
            return;
        }
        if (!this.showTimePicker || !this.internalSelectedDates.start)
            return;
        // if swap occurred, bail
        if (!this.ensureRangeOrder()) {
            return;
        }
        const { start, end } = this.internalSelectedDates;
        let emitValue;
        const formattedStart = this.formatDateOrDateTime(start, 'start');
        if (this.mode === 'range') {
            const formattedEnd = end ? this.formatDateOrDateTime(end, 'end') : null;
            emitValue = {
                start: formattedStart,
                end: formattedEnd,
            };
        }
        else {
            emitValue = formattedStart;
        }
        this.tkChange.emit(emitValue);
        this.inputValue = this.formatInputValue();
    }
    updateMaskOptions() {
        if (this.timeOnly) {
            // timeformat'a göre mask options'ları belirle
            this.maskOptions = {
                time: true,
                timePattern: ['h', 'm'],
                timeFormat: this.timeFormat,
            };
        }
        else {
            this.getMaskOptionsFromDateFormat(this.dateFormat);
        }
    }
    getTimeStateToModify() {
        // Allow time changes if any time UI is active: showTimePicker or timeOnly mode
        if (!(this.showTimePicker || this.timeOnly))
            return null;
        let targetType = 'start';
        if (this.mode === 'range' && this.internalSelectedDates.end) {
            targetType = 'end';
        }
        this.ensureDateTimeInitialized(targetType);
        const timeState = targetType === 'start' ? this.internalStartTime : this.internalEndTime;
        if (!timeState)
            return null;
        return { time: timeState, type: targetType };
    }
    handleYearSelect(e, year) {
        e.stopPropagation();
        this.currentMonth = new Date(year, this.currentMonth.getMonth());
        this.currentView = 'months';
        this.remeasureCalendarOnNextFrame();
    }
    handleFormReset() {
        const initialValueAttr = this.el.getAttribute('value');
        let initialValue = null;
        if (initialValueAttr) {
            try {
                initialValue = JSON.parse(initialValueAttr);
            }
            catch (e) {
                initialValue = initialValueAttr;
            }
        }
        this.processDateValue(initialValue, true);
        this.tkChange.emit(this.value);
    }
    createDayCell(date, isAdjacentMonth) {
        const { start = null, end = null } = this.internalSelectedDates;
        const dateTime = date.getTime();
        const isSelectedStart = start && dateTime === start.getTime();
        const isSelectedEnd = start && end && dateTime === end.getTime();
        const isSelected = isSelectedStart || isSelectedEnd;
        let isInRange = false;
        const rangeEnd = end || this.hoverDate;
        let visualRangeStart = null;
        let visualRangeEnd = null;
        if (start && rangeEnd) {
            if (start.getTime() <= (rangeEnd === null || rangeEnd === void 0 ? void 0 : rangeEnd.getTime())) {
                visualRangeStart = start;
                visualRangeEnd = rangeEnd;
            }
            else {
                // Reverse selection: user selected end date first, then hovering/selecting earlier date
                visualRangeStart = rangeEnd;
                visualRangeEnd = start;
            }
            isInRange = dateTime > (visualRangeStart === null || visualRangeStart === void 0 ? void 0 : visualRangeStart.getTime()) && dateTime < (visualRangeEnd === null || visualRangeEnd === void 0 ? void 0 : visualRangeEnd.getTime());
        }
        const isVisualRangeStart = visualRangeStart && dateTime === (visualRangeStart === null || visualRangeStart === void 0 ? void 0 : visualRangeStart.getTime());
        const isVisualRangeEnd = visualRangeEnd && dateTime === (visualRangeEnd === null || visualRangeEnd === void 0 ? void 0 : visualRangeEnd.getTime());
        const isDisabled = this.isDateDisabled(date);
        const isToday = this.isToday(date);
        const dateKey = `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`;
        return (index.h("td", { class: index$2.classNames('tk-datepicker-day-cell', {
                'selected': isSelected,
                'in-range': isInRange,
                'range-start': isVisualRangeStart && this.mode === 'range',
                'range-end': isVisualRangeEnd && this.mode === 'range',
                'today': isToday && !isSelected && !isInRange,
                'disabled': isDisabled || this.disabled,
                'readonly': this.readonly,
                'adjacent-month': isAdjacentMonth,
            }), onClick: () => !isDisabled && this.handleDateClick(date), onMouseEnter: () => this.handleDateHover(date), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'day-cell', dateKey) }, index.h("span", { class: "tk-datepicker-day", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'day', dateKey) }, date.getDate())));
    }
    createWeekDays() {
        const firstDayOfMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth(), 1);
        const lastDayOfMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() + 1, 0);
        const daysInPreviousMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth(), 0).getDate();
        const daysInCurrentMonth = lastDayOfMonth.getDate();
        const weeks = [];
        let days = [];
        const resolvedStartOfWeekIndex = this.getResolvedFirstDayIndex();
        const startOfWeekForGetDay = (resolvedStartOfWeekIndex + 1) % 7;
        const firstDayOfWeek = firstDayOfMonth.getDay();
        let emptyCells = (firstDayOfWeek - startOfWeekForGetDay + 7) % 7;
        // Previous month's days
        for (let i = emptyCells - 1; i >= 0; i--) {
            const date = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() - 1, daysInPreviousMonth - i);
            days.push(this.createDayCell(date, true));
        }
        // Current month's days
        for (let i = 1; i <= daysInCurrentMonth; i++) {
            const date = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth(), i);
            days.push(this.createDayCell(date, false));
            if (days.length === 7) {
                weeks.push(index.h("tr", { "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'week', weeks.length.toString()) }, days));
                days = [];
            }
        }
        // Next month's days
        if (days.length > 0) {
            let nextMonthDay = 1;
            while (days.length < 7) {
                const date = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() + 1, nextMonthDay);
                days.push(this.createDayCell(date, true));
                nextMonthDay++;
            }
            weeks.push(index.h("tr", { "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'week', weeks.length.toString()) }, days));
        }
        this.weeksLength = weeks.length;
        return (index.h("tbody", { class: "tk-datepicker-days", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'days') }, weeks));
    }
    createWeekDayNames() {
        const startOfWeekIndex = this.getResolvedFirstDayIndex();
        // Use a known Monday as a base to generate other days reliably
        const baseMonday = new Date(2023, 0, 2);
        const weekdays = [...Array(7)].map((_, i) => {
            const dayOffset = (i + startOfWeekIndex) % 7;
            const dateForDay = new Date(baseMonday);
            dateForDay.setDate(baseMonday.getDate() + dayOffset);
            return dateForDay.toLocaleString(this.locale, { weekday: 'short' });
        });
        return (index.h("thead", { "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'week-days-header') }, index.h("tr", { class: "tk-datepicker-week-days", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'week-days-row') }, weekdays.map((day, index$1) => (index.h("th", { class: "tk-datepicker-week-day-cell", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'week-cell', index$1.toString()) }, index.h("span", { class: "tk-datepicker-week-day", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'week-day', index$1.toString()) }, day)))))));
    }
    createMonths() {
        const months = [...Array(12)].map((_, i) => new Date(2023, i).toLocaleString(this.locale, { month: 'short' }));
        const rows = [];
        for (let i = 0; i < 12; i += 4) {
            const monthGroup = months.slice(i, i + 4);
            rows.push(index.h("tr", { class: "tk-datepicker-month-row", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'month', (i / 4).toString()) }, monthGroup.map((month, index$1) => {
                const monthIndex = i + index$1;
                const isSelected = this.currentMonth.getMonth() === monthIndex;
                return (index.h("td", { class: index$2.classNames('tk-datepicker-month', {
                        selected: isSelected,
                    }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'month-option', monthIndex.toString()), onClick: (e) => {
                        e.stopPropagation();
                        this.currentMonth = new Date(this.currentMonth.getFullYear(), monthIndex);
                        this.currentView = 'days';
                    } }, month));
            })));
        }
        return (index.h("tbody", { class: "tk-datepicker-months", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'months') }, rows));
    }
    createYears() {
        const currentYear = this.currentMonth.getFullYear();
        const startYear = Math.floor(currentYear / 12) * 12;
        const years = Array.from({ length: 12 }, (_, i) => startYear + i);
        const rows = [];
        for (let i = 0; i < years.length; i += 4) {
            const yearGroup = years.slice(i, i + 4);
            rows.push(index.h("tr", { class: "tk-datepicker-year-row", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'year', (i / 4).toString()) }, yearGroup.map(year => (index.h("td", { class: index$2.classNames('tk-datepicker-year', {
                    selected: year === currentYear,
                }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'year', year.toString()), onClick: e => this.handleYearSelect(e, year) }, year)))));
        }
        return (index.h("tbody", { class: "tk-datepicker-years", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'years') }, rows));
    }
    createHeader() {
        const monthName = this.currentMonth.toLocaleString(this.locale, {
            month: 'long',
        });
        const year = this.currentMonth.getFullYear().toString();
        const headerClasses = index$2.classNames('tk-datepicker-header', `tk-datepicker-header-${this.headerType}`);
        return (index.h("div", { class: headerClasses, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header') }, index.h("div", { class: "tk-datepicker-header-content", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-content') }, index.h("div", { class: "tk-datepicker-header-content-start", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-start') }, index.h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "keyboard_double_arrow_left", "onTk-click": () => this.handleYearChange(-1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'prev-year-button') }), index.h("span", { class: "tk-datepicker-divider", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-start-divider') }), index.h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "chevron_left", "onTk-click": () => this.handleMonthChange(-1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'prev-month-button') })), index.h("div", { class: "tk-datepicker-select-container", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-select') }, index.h("div", { class: index$2.classNames('tk-datepicker-select-month', { disabled: this.readonly || this.disabled || this.loading }), onClick: e => !this.loading && this.handleViewChange(e, 'months'), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-month') }, monthName), index.h("div", { class: index$2.classNames('tk-datepicker-select-year', { disabled: this.readonly || this.disabled || this.loading }), onClick: e => !this.loading && this.handleViewChange(e, 'years'), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-year') }, year)), index.h("div", { class: "tk-datepicker-header-content-end", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-end') }, index.h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "chevron_right", "onTk-click": () => this.handleMonthChange(1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'next-month-button') }), index.h("span", { class: "tk-datepicker-divider", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'header-end-divider') }), index.h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "keyboard_double_arrow_right", "onTk-click": () => this.handleYearChange(1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'next-year-button') })))));
    }
    createFooter() {
        if (this.hasFooterSlot) {
            return index.h("slot", { name: "footer" });
        }
        else if (this.hasFooterActionsSlot) {
            const footerClass = index$2.classNames('tk-datepicker-footer', `tk-datepicker-footer-${this.footerType}`);
            return (index.h("div", { class: footerClass, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'footer') }, index.h("slot", { name: "footer-actions" })));
        }
        return null;
    }
    createTimePicker() {
        var _a, _b;
        let timeToDisplay = null;
        if (this.mode === 'single') {
            timeToDisplay = this.internalStartTime;
        }
        else {
            if (this.internalSelectedDates.end) {
                timeToDisplay = this.internalEndTime;
            }
            else if (this.internalSelectedDates.start) {
                timeToDisplay = this.internalStartTime;
            }
        }
        const defaultTime = this.getDefaultTime();
        const currentTime = timeToDisplay || defaultTime;
        let displayHour = currentTime.hour;
        if (this.timeFormat === '12') {
            displayHour = displayHour % 12 === 0 ? 12 : displayHour % 12;
        }
        const displayMinute = currentTime.minute;
        const hours = this.timeFormat === '12'
            ? Array.from({ length: Math.ceil(12 / this.hourStep) }, (_, i) => i * this.hourStep + 1)
            : Array.from({ length: Math.ceil(24 / this.hourStep) }, (_, i) => i * this.hourStep);
        const minutes = Array.from({ length: Math.ceil(60 / this.minuteStep) }, (_, i) => i * this.minuteStep);
        const isHourDisabled = (hour) => this.isTimeDisabled(hour, displayMinute);
        const isMinuteDisabled = (minute) => this.isTimeDisabled(displayHour, minute);
        // Find closest hour in the hours array
        const findClosestInArray = (value, arr) => {
            if (arr.includes(value))
                return value;
            // Find the closest value in the array
            return arr.reduce((prev, curr) => (Math.abs(curr - value) < Math.abs(prev - value) ? curr : prev));
        };
        const currentHour = findClosestInArray(displayHour, hours);
        const currentMinute = findClosestInArray(displayMinute, minutes);
        const sliceRange = (options, selected) => {
            const idx = options.indexOf(selected);
            return Array.from({ length: this.weeksLength === 6 ? 5 : this.weeksLength === 4 ? 3 : 4 }, (_, i) => {
                const optIndex = idx - (this.weeksLength === 4 ? 1 : 2) + i;
                return optIndex >= 0 && optIndex < options.length ? options[optIndex] : null;
            });
        };
        const isDisabled = this.disabled || this.readonly;
        const visibleHours = sliceRange(hours, currentHour);
        const visibleMinutes = sliceRange(minutes, currentMinute);
        const isMinHour = currentHour === hours[0];
        const isMaxHour = currentHour === hours[hours.length - 1];
        const isMinMinute = currentMinute === minutes[0];
        const isMaxMinute = currentMinute === minutes[minutes.length - 1];
        const nextHourDisabled = isHourDisabled(hours[hours.indexOf(currentHour) + 1]);
        const prevHourDisabled = isHourDisabled(hours[hours.indexOf(currentHour) - 1]);
        const nextMinuteDisabled = isMinuteDisabled(minutes[minutes.indexOf(currentMinute) + 1]);
        const prevMinuteDisabled = isMinuteDisabled(minutes[minutes.indexOf(currentMinute) - 1]);
        let AMDisabled = false;
        let PMDisabled = false;
        if (this.timeFormat === '12') {
            if ((_a = this.minTime) === null || _a === void 0 ? void 0 : _a.toLowerCase().includes('pm')) {
                this.internalAmPm = 'PM';
                AMDisabled = true;
            }
            if ((_b = this.maxTime) === null || _b === void 0 ? void 0 : _b.toLowerCase().includes('am')) {
                this.internalAmPm = 'AM';
                PMDisabled = true;
            }
        }
        return (index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-panel', this.timeOnly && 'tk-datepicker-timepicker-panel-only'), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker') }, index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-header', `tk-datepicker-timepicker-header-${this.headerType}`, this.timeOnly && 'tk-datepicker-timepicker-header-only'), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-header') }, this.timeFormat === '12' && (index.h("tk-toggle-button-group", { value: this.internalAmPm, type: this.headerType === 'basic' ? 'basic' : this.headerType === 'light' ? 'light' : 'divided', "onTk-change": e => this.handleAmPmToggle(e), class: "tk-datepicker-ampm-toggle", dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'timepicker-toggle') }, index.h("tk-toggle-button", { key: "AM", type: this.headerType === 'primary' || this.headerType === 'dark' ? 'filled' : 'text', variant: this.headerType === 'dark' ? 'neutral' : 'primary', value: "AM", label: "AM", size: "small", disabled: isDisabled || AMDisabled, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'timepicker-am') }), index.h("tk-toggle-button", { key: "PM", type: this.headerType === 'primary' || this.headerType === 'dark' ? 'filled' : 'text', variant: this.headerType === 'dark' ? 'neutral' : 'primary', value: "PM", label: "PM", size: "small", disabled: isDisabled || PMDisabled, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'timepicker-pm') })))), index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-body', ['primary', 'light', 'dark'].includes(this.headerType) && `tk-datepicker-timepicker-body-${this.headerType}`, this.weeksLength === 4 && 'tk-datepicker-timepicker-body-4-weeks', this.timeOnly && 'tk-datepicker-timepicker-body-only'), style: {
                height: this.calendarTableHeightPx ? `${this.calendarTableHeightPx}px` : undefined,
            }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-body') }, index.h("div", { class: "tk-datepicker-timepicker-col", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-hours') }, index.h("div", null, index.h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_less", "onTk-click": this.handleDecreaseHour, disabled: isMinHour || prevHourDisabled || isDisabled, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'timepicker-hours-decrease') }), index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-hours-separator-top') })), visibleHours.map((hour, hourIndex) => hour === null ? (index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-value tk-datepicker-timepicker-value-empty', { disabled: isDisabled }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-hours-value-empty', hourIndex.toString()) })) : (index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-value', {
                'selected': hour === currentHour,
                'tk-datepicker-timepicker-value-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-value-primary': this.headerType === 'primary',
                'disabled': isDisabled || isHourDisabled(hour),
            }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-hour', hour.toString()), onClick: () => {
                if (isDisabled || isHourDisabled(hour)) {
                    return;
                }
                else if (this.timeFormat === '12') {
                    const hour24 = (hour === 12 ? 0 : hour) + (this.internalAmPm === 'PM' ? 12 : 0);
                    this.handleHourClick(hour24);
                }
                else {
                    this.handleHourClick(hour);
                }
            } }, hour.toString().padStart(2, '0')))), index.h("div", null, index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-hours-separator-bottom') }), index.h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_more", "onTk-click": this.handleIncreaseHour, disabled: isMaxHour || nextHourDisabled || isDisabled, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'timepicker-hours-increase') }))), index.h("div", { class: "tk-datepicker-timepicker-col", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-minutes') }, index.h("div", null, index.h("tk-button", { variant: this.headerType === 'dark' || this.headerType === 'primary' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_less", "onTk-click": this.handleDecreaseMinute, disabled: isMinMinute || prevMinuteDisabled || isDisabled, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'timepicker-minutes-decrease') }), index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-minutes-separator-top') })), visibleMinutes.map((m, minuteIndex) => m === null ? (index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-value tk-datepicker-timepicker-value-empty', { disabled: isDisabled }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-minutes-value-empty', minuteIndex.toString()) })) : (index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-value', {
                'selected': m === currentMinute,
                'tk-datepicker-timepicker-value-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-value-primary': this.headerType === 'primary',
                'disabled': isDisabled || isMinuteDisabled(m),
            }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-minute', m.toString()), onClick: () => {
                if (isDisabled || isMinuteDisabled(m))
                    return;
                this.handleMinuteClick(m);
            } }, String(m).padStart(2, '0')))), index.h("div", null, index.h("div", { class: index$2.classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timepicker-minutes-separator-bottom') }), index.h("tk-button", { variant: this.headerType === 'dark' || this.headerType === 'primary' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_more", "onTk-click": this.handleIncreaseMinute, disabled: isMaxMinute || nextMinuteDisabled || isDisabled, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'timepicker-minutes-increase') }))))));
    }
    createLoading() {
        return (index.h("div", { class: "tk-datepicker-loading", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'loading') }, index.h("tk-spinner", { size: this.size, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'loading-spinner') })));
    }
    renderInput() {
        if (this.inline)
            return null;
        const shouldUseMask = !this.disableMask && this.mode !== 'range';
        const maskOptionsToPass = shouldUseMask ? this.maskOptions : undefined;
        return (index.h("tk-input", { ref: el => (this.inputRef = el), label: this.label, size: this.size, mode: "text", icon: this.icon, iconPosition: this.iconPosition, class: index$2.classNames('tk-datepicker-input', { 'tk-table-input': this.el.classList.contains('tk-table-datepicker') }), name: this.name, hint: this.hint, clearable: this.clearable, loading: this.loading, disabled: this.disabled, invalid: this.invalid || this.isInvalid, readonly: this.readonly, error: this.error, placeholder: this.placeholder || (this.timeOnly ? this.getOnlyTimeFormat() : this.showTimePicker ? this.getFullDateTimeFormat() : this.dateFormat).toUpperCase(), value: this.inputValue, maskOptions: maskOptionsToPass, "onTk-change": this.handleInputChange, "onTk-clear-click": this.handleInputClearClick, onInput: this.handleInputInput, onKeyDown: this.handleInputKeyDown, onClick: this.handleInputClick, "onTk-blur": this.handleInputBlur, "aria-expanded": !!this.isOpen, "aria-haspopup": "true", "data-tk-datepicker-id": this.uniqueId, showAsterisk: this.showAsterisk, dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'input') }));
    }
    renderPanel() {
        if (!this.isOpen && !this.inline)
            return null;
        const panelClasses = index$2.classNames('tk-datepicker-panel', {
            'tk-datepicker-panel-inline': this.inline,
        });
        const bodyClasses = index$2.classNames('tk-datepicker-body', {
            'tk-datepicker-months-view': this.currentView === 'months',
            'tk-datepicker-years-view': this.currentView === 'years',
        });
        // Time-only mode: render only time picker
        if (this.timeOnly) {
            return (index.h("div", { class: panelClasses, ref: el => (this.panelRef = el), role: !this.inline ? 'dialog' : undefined, "aria-modal": "true", "data-tk-datepicker-id": this.uniqueId, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel') }, index.h("div", { class: "tk-datepicker-panel-inner", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-inner') }, this.createTimePicker()), this.createFooter()));
        }
        // Default: date (calendar), optionally alongside time picker
        return (index.h("div", { class: panelClasses, ref: el => (this.panelRef = el), role: !this.inline ? 'dialog' : undefined, "aria-modal": "true", "data-tk-datepicker-id": this.uniqueId, style: { visibility: this.concealUntilMeasured ? 'hidden' : undefined }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel') }, index.h("div", { class: "tk-datepicker-panel-inner", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-inner') }, index.h("div", { class: "tk-datepicker-calendar-container", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'calendar') }, this.createHeader(), index.h("div", { class: bodyClasses, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'body') }, this.loading ? (this.createLoading()) : (index.h("table", { class: "tk-datepicker-table", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'table') }, this.currentView === 'days' && (index.h(index.Fragment, null, this.createWeekDayNames(), this.createWeekDays())), this.currentView === 'months' && this.createMonths(), this.currentView === 'years' && this.createYears())))), this.showTimePicker && this.createTimePicker()), this.createFooter()));
    }
    render() {
        return (index.h("div", { key: '80b177ebfad0d28bc21c37aaa1991158c83cfc26', class: "tk-datepicker-container", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'container') }, this.renderInput(), this.renderPanel()));
    }
    static get formAssociated() { return true; }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "internalAmPm": [{
                "internalAmPmChanged": 0
            }],
        "isOpen": [{
                "isOpenChanged": 0
            }],
        "value": [{
                "valueChanged": 0
            }],
        "dateFormat": [{
                "dateFormatChanged": 0
            }],
        "showTimePicker": [{
                "showTimePickerChanged": 0
            }],
        "timeOnly": [{
                "timeOnlyChanged": 0
            }],
        "timeFormat": [{
                "timeFormatChanged": 0
            }]
    }; }
};
TkDatePicker.style = tkDatepickerCss();

/**
 * Formats a template string by replacing placeholders with corresponding values.
 *
 * @param template - The template string containing placeholders in the format {key}.
 * @param values - An object containing key-value pairs for replacement.
 * @returns The formatted string with placeholders replaced by their corresponding values.
 * @example
 * const template = "Page {currentPage} of {totalPages}";
 * const values = { currentPage: 2, totalPages: 5 };
 * const result = formatTemplate(template, values);
 * // result: "Page 2 of 5"
 */
function formatTemplate(template, values) {
    return template.replace(/\{(\w+)\}/g, (match, key) => {
        var _a;
        return ((_a = values[key]) === null || _a === void 0 ? void 0 : _a.toString()) || match;
    });
}
function getTemplateValues(internalCurrentPage, rowsPerPage, totalItems, totalPages) {
    const startItem = (internalCurrentPage - 1) * rowsPerPage + 1;
    const endItem = Math.min(internalCurrentPage * rowsPerPage, totalItems);
    return {
        currentPage: internalCurrentPage,
        totalPages: totalPages,
        startItem,
        endItem,
        totalItems: totalItems,
    };
}

const tkPaginationCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-pagination{display:block}.tk-pagination-container{display:flex;justify-content:space-between;align-items:center;flex-wrap:wrap;font-family:var(--family-body)}.tk-pagination-container .tk-pagination-start{margin-right:auto}.tk-pagination-container .tk-pagination-end{margin-left:auto;display:flex;gap:var(--pagination-base-right-gap, 12px)}.tk-pagination-container .tk-pagination-tag{display:flex;flex-direction:column;justify-content:center;align-items:flex-start;gap:var(--pagination-base-tag-gap, 0px)}.tk-pagination-container .tk-pagination-tag .tk-pagination-tag-label{color:var(--text-base, #717784);font-family:var(--desktop-body-xs-font);font-size:var(--desktop-body-xs-size, 12px);font-style:normal;font-weight:400;line-height:var(--desktop-body-xs-line-height, 18px);}.tk-pagination-container .tk-pagination{display:flex;align-items:center;flex-wrap:wrap;gap:var(--pagination-base-gap, 6px)}.tk-pagination-container .tk-pagination .tk-pagination-prev-actions,.tk-pagination-container .tk-pagination .tk-pagination-next-actions{display:flex;align-items:center;gap:var(--pagination-base-arrow-gap, 4px)}.tk-pagination-container .tk-pagination .tk-pagination-cell{display:flex;min-width:32px;height:32px;padding:6px 11px;justify-content:center;align-items:center;border-radius:8px;background:none;border:none;cursor:pointer;color:var(--text-darkest, #222530);font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size, 14px);font-style:normal;font-weight:400;line-height:var(--desktop-body-s-line-height, 20px);}.tk-pagination-container .tk-pagination .tk-pagination-cell .tk-pagination-page-label{color:var(--text-darkest)}.tk-pagination-container .tk-pagination .tk-pagination-cell:hover:not(:disabled):not(.tk-pagination-ellipsis){background:var(--background-lightest, #f9fafc)}.tk-pagination-container .tk-pagination .tk-pagination-cell:focus-visible{border:var(--spacing-px, 1px) solid var(--states-info-base, #3b82f6)}.tk-pagination-container .tk-pagination .tk-pagination-cell:disabled{pointer-events:none}.tk-pagination-container .tk-pagination .tk-pagination-cell:disabled .tk-pagination-page-label{color:var(--text-sub-base, #99a0ae)}.tk-pagination-container .tk-pagination.tk-pagination-rounded:not(.tk-pagination-compact-expanded) .tk-pagination-page{border-radius:var(--radius-full)}.tk-pagination-container .tk-pagination.tk-pagination-outlined .tk-pagination-page{border:var(--spacing-px, 1px) solid var(--border-light, #e1e4ea)}.tk-pagination-container .tk-pagination.tk-pagination-grouped{display:inline-flex;align-items:center;padding:var(--pagination-base-container-v-padding, 2px) 0px;border-radius:8px;border:var(--spacing-px, 1px) solid var(--border-light, #e1e4ea);background:var(--static-light, #fff);overflow:hidden}.tk-pagination-container .tk-pagination.tk-pagination-grouped .tk-pagination-divider{width:1px;height:16px;background-color:var(--border-light, #e1e4ea)}.tk-pagination-container .tk-pagination.tk-pagination-compact .tk-pagination-cell{width:36px;height:36px}.tk-pagination-container .tk-pagination.tk-pagination-compact .tk-pagination-current-label{color:var(--text-base, #717784);font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size, 14px);font-style:normal;font-weight:400;line-height:var(--desktop-body-s-line-height, 20px);}.tk-pagination-container .tk-pagination.tk-pagination-compact .tk-pagination-first,.tk-pagination-container .tk-pagination.tk-pagination-compact .tk-pagination-prev,.tk-pagination-container .tk-pagination.tk-pagination-compact .tk-pagination-next,.tk-pagination-container .tk-pagination.tk-pagination-compact .tk-pagination-last{border:var(--spacing-px, 1px) solid var(--border-light, #e1e4ea)}.tk-pagination-container .tk-pagination.tk-pagination-compact.tk-pagination-rounded .tk-pagination-first,.tk-pagination-container .tk-pagination.tk-pagination-compact.tk-pagination-rounded .tk-pagination-prev,.tk-pagination-container .tk-pagination.tk-pagination-compact.tk-pagination-rounded .tk-pagination-next,.tk-pagination-container .tk-pagination.tk-pagination-compact.tk-pagination-rounded .tk-pagination-last{border-radius:var(--radius-full)}.tk-pagination-container .tk-pagination.tk-pagination-compact tk-input{width:48px;height:40px}.tk-pagination-container .tk-pagination .tk-pagination-page.tk-pagination-page-active{border:var(--spacing-px, 1px) solid var(--states-info-sub-base);background:var(--background-lightest, #f9fafc)}.tk-pagination-container .tk-pagination .tk-pagination-last:focus-visible,.tk-pagination-container .tk-pagination .tk-pagination-prev:focus-visible,.tk-pagination-container .tk-pagination .tk-pagination-page:focus-visible,.tk-pagination-container .tk-pagination .tk-pagination-next:focus-visible,.tk-pagination-container .tk-pagination .tk-pagination-first:focus-visible{border:var(--spacing-px, 1px) solid var(--states-info-base, #3b82f6);outline:none}@media (max-width: 600px){.tk-pagination-container{display:grid;gap:8px;grid-template-columns:1fr}.tk-pagination-container .tk-pagination-start{display:none}.tk-pagination-container .tk-pagination,.tk-pagination-container .tk-pagination-end{width:100%;justify-content:center}}`;

const TkPagination = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.tkNextPage = index.createEvent(this, "tk-next-page", 7);
        this.tkPageChange = index.createEvent(this, "tk-page-change", 7);
        this.tkPrevPage = index.createEvent(this, "tk-prev-page", 7);
        this.tkRowsPerPageChange = index.createEvent(this, "tk-rows-per-page-change", 7);
        this.ellipsis = '...';
        this.internalCurrentPage = 1;
        /**
         * Whether the pagination elements should have rounded corners
         * @defaultValue false
         */
        this.rounded = false;
        /**
         * Number of items per page.
         * @defaultValue 10
         */
        this.rowsPerPage = 10;
        /**
         * Number of total items.
         * @defaultValue 0
         */
        this.totalItems = 0;
        /**
         * Number of items per page options
         * @defaultValue [5, 10, 20, 50]
         */
        this.rowsPerPageOptions = [5, 10, 20, 50];
        /**
         * The type of the pagination
         * @defaultValue 'outlined'
         */
        this.type = 'outlined';
        /**
         * Template string for current page report in pagination.
         * Available placeholders: {currentPage}, {totalPages}
         * @defaultValue 'page: {currentPage} of {totalPages}'
         */
        this.pageReportTemplate = 'page: {currentPage} of {totalPages}';
        /**
         * Template string for items report in pagination.
         * Available placeholders: {startItem}, {endItem}, {totalItems}
         * @defaultValue 'item: {startItem}-{endItem} of {totalItems}'
         */
        this.itemsReportTemplate = 'item: {startItem}-{endItem} of {totalItems}';
        /**
         * The current page of the pagination.
         * @defaultValue 1
         */
        this.currentPage = 1;
        this.handlePageClick = (page) => {
            if (page >= 1 && page <= this.getTotalPages()) {
                this.internalCurrentPage = page;
            }
        };
        this.handlePrevClick = () => {
            if (this.internalCurrentPage > 1) {
                this.internalCurrentPage--;
                this.tkPrevPage.emit({ page: this.internalCurrentPage });
            }
        };
        this.handleNextClick = () => {
            if (this.internalCurrentPage < this.getTotalPages()) {
                this.internalCurrentPage++;
                this.tkNextPage.emit({ page: this.internalCurrentPage });
            }
        };
    }
    internalCurrentPageChanged(newValue) {
        this.currentPage = newValue;
        this.updateInputValue(newValue);
        this.emitPageChangeEvents();
    }
    currentPageChanged(newValue) {
        this.internalCurrentPage = newValue;
        this.updateInputValue(newValue);
    }
    componentWillLoad() {
        this.internalCurrentPage = this.currentPage;
        this.updateInputValue(this.currentPage);
    }
    updateInputValue(value) {
        this.inputValue = value.toString();
    }
    getTotalPages() {
        return Math.ceil(this.totalItems / this.rowsPerPage);
    }
    getPageNumbers() {
        const totalPages = this.getTotalPages();
        const currentPage = this.internalCurrentPage;
        const pageNumbers = [];
        if (totalPages <= 5) {
            for (let i = 1; i <= totalPages; i++) {
                pageNumbers.push(i);
            }
        }
        else if (totalPages <= 7) {
            this.generatePageNumbersForMediumRange(pageNumbers, currentPage, totalPages);
        }
        else {
            this.generatePageNumbersForLargeRange(pageNumbers, currentPage, totalPages);
        }
        return pageNumbers;
    }
    generatePageNumbersForMediumRange(pageNumbers, currentPage, totalPages) {
        pageNumbers.push(1);
        if (currentPage <= 3) {
            for (let i = 2; i <= 5; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis);
        }
        else if (currentPage >= totalPages - 2) {
            pageNumbers.push(this.ellipsis);
            for (let i = totalPages - 4; i < totalPages; i++) {
                pageNumbers.push(i);
            }
        }
        else {
            pageNumbers.push(this.ellipsis);
            for (let i = currentPage - 1; i <= currentPage + 1; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis);
        }
        pageNumbers.push(totalPages);
    }
    generatePageNumbersForLargeRange(pageNumbers, currentPage, totalPages) {
        pageNumbers.push(1);
        if (currentPage <= 4) {
            for (let i = 2; i <= 5; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis, totalPages);
        }
        else if (currentPage >= totalPages - 3) {
            pageNumbers.push(this.ellipsis);
            for (let i = totalPages - 4; i <= totalPages; i++) {
                pageNumbers.push(i);
            }
        }
        else {
            pageNumbers.push(this.ellipsis);
            for (let i = currentPage - 2; i <= currentPage + 2; i++) {
                pageNumbers.push(i);
            }
            pageNumbers.push(this.ellipsis, totalPages);
        }
    }
    validateAndUpdatePage() {
        const pageNumber = parseInt(this.inputValue, 10);
        const totalPages = this.getTotalPages();
        if (!isNaN(pageNumber) && pageNumber >= 1 && pageNumber <= totalPages) {
            this.handlePageClick(pageNumber);
        }
        else {
            this.updateInputValue(this.internalCurrentPage);
        }
    }
    emitPageChangeEvents() {
        const totalPages = this.getTotalPages();
        const startItem = (this.internalCurrentPage - 1) * this.rowsPerPage + 1;
        const endItem = Math.min(this.internalCurrentPage * this.rowsPerPage, this.totalItems);
        this.tkPageChange.emit({
            page: this.internalCurrentPage,
            totalPages,
            startItem,
            endItem,
        });
    }
    handlePageInputChange(value) {
        this.inputValue = value.replace(/[^0-9]/g, '');
    }
    createPageNumbers() {
        return this.getPageNumbers().map(pageNumber => {
            if (pageNumber === this.ellipsis) {
                return index.h("span", { class: "tk-pagination-ellipsis tk-pagination-cell" }, "...");
            }
            return (index.h("button", { class: index$2.classNames('tk-pagination-cell tk-pagination-page', { 'tk-pagination-page-active': pageNumber === this.internalCurrentPage }), type: "button", onClick: () => this.handlePageClick(pageNumber) }, index.h("span", { class: "tk-pagination-page-label" }, pageNumber)));
        });
    }
    renderTag(totalPages) {
        if (this.mode !== 'compact') {
            let tagContent;
            const templateValues = getTemplateValues(this.internalCurrentPage, this.rowsPerPage, this.totalItems, totalPages);
            if (this.mode === 'compact-expanded') {
                const itemsText = formatTemplate(this.itemsReportTemplate, templateValues);
                tagContent = index.h("span", { class: "tk-pagination-tag-label" }, itemsText);
            }
            else {
                const pageText = formatTemplate(this.pageReportTemplate, templateValues);
                const itemsText = formatTemplate(this.itemsReportTemplate, templateValues);
                tagContent = totalPages > 0 && (index.h(index.Fragment, null, index.h("span", { class: "tk-pagination-tag-label" }, pageText), index.h("span", { class: "tk-pagination-tag-label" }, itemsText)));
            }
            return index.h("div", { class: "tk-pagination-tag" }, tagContent);
        }
        return null;
    }
    renderContent(totalPages) {
        const ContentWrapperTag = this.mode === 'compact' || this.mode === 'compact-expanded' ? 'div' : 'nav';
        const contentClasses = index$2.classNames('tk-pagination', `tk-pagination-${this.mode || this.type}`, {
            'tk-pagination-rounded': this.rounded,
        });
        let content;
        if (this.mode === 'compact') {
            const templateValues = getTemplateValues(this.internalCurrentPage, this.rowsPerPage, this.totalItems, totalPages);
            const compactLabel = formatTemplate(this.pageReportTemplate, templateValues);
            content = (index.h(index.Fragment, null, index.h("button", { class: "tk-pagination-cell tk-pagination-prev", type: "button", onClick: this.handlePrevClick, disabled: this.internalCurrentPage === 1 }, index.h("tk-icon", Object.assign({}, iconUtils.getIconElementProps('chevron_left', { color: 'var(--icon-base)' })))), index.h("tk-input", { mode: "text", value: this.inputValue, "onTk-change": (event) => this.handlePageInputChange(event.detail.toString()), "onTk-blur": () => this.validateAndUpdatePage(), min: 1, max: totalPages }), index.h("span", { class: "tk-pagination-current-label" }, compactLabel), index.h("button", { class: "tk-pagination-cell tk-pagination-next", type: "button", onClick: this.handleNextClick, disabled: this.internalCurrentPage === totalPages }, index.h("tk-icon", Object.assign({}, iconUtils.getIconElementProps('chevron_right', { color: 'var(--icon-base)' }))))));
        }
        // else if(this.type === 'compact-expanded') {
        //   @TODO: Add pagination content for compact-expanded after the tk-dropdown component is completed
        //   content =
        // }
        else {
            content = (index.h(index.Fragment, null, index.h("div", { class: "tk-pagination-prev-actions" }, index.h("button", { class: "tk-pagination-cell tk-pagination-first", type: "button", onClick: () => this.handlePageClick(1), disabled: this.internalCurrentPage === 1 }, index.h("tk-icon", Object.assign({}, iconUtils.getIconElementProps('keyboard_double_arrow_left', { color: 'var(--icon-base)' })))), this.type === 'grouped' && index.h("span", { class: "tk-pagination-divider" }), index.h("button", { class: "tk-pagination-cell tk-pagination-prev", type: "button", onClick: this.handlePrevClick, disabled: this.internalCurrentPage === 1 }, index.h("tk-icon", Object.assign({}, iconUtils.getIconElementProps('chevron_left', { color: 'var(--icon-base)' }))), ' ')), this.createPageNumbers(), index.h("div", { class: "tk-pagination-next-actions" }, index.h("button", { class: "tk-pagination-cell tk-pagination-next", type: "button", onClick: this.handleNextClick, disabled: this.internalCurrentPage === totalPages }, index.h("tk-icon", Object.assign({}, iconUtils.getIconElementProps('chevron_right', { color: 'var(--icon-base)' })))), this.type === 'grouped' && index.h("span", { class: "tk-pagination-divider" }), index.h("button", { class: "tk-pagination-cell tk-pagination-last", type: "button", onClick: () => this.handlePageClick(totalPages), disabled: this.internalCurrentPage === totalPages }, index.h("tk-icon", Object.assign({}, iconUtils.getIconElementProps('keyboard_double_arrow_right', { color: 'var(--icon-base)' })))))));
        }
        return index.h(ContentWrapperTag, { class: contentClasses }, content);
    }
    renderSelect() {
        return (index.h("tk-select", { style: { width: '75px' }, value: this.rowsPerPage, options: this.rowsPerPageOptions, "onTk-change": e => {
                if (e.detail !== this.rowsPerPage) {
                    this.rowsPerPage = e.detail;
                    this.tkRowsPerPageChange.emit(e.detail);
                }
            } }));
    }
    renderInput(totalPages) {
        return (index.h("tk-input", { style: { width: '70px' }, mode: "text", min: 1, max: totalPages, value: this.inputValue, icon: "chevron_right", iconPosition: "right", "onTk-change": (event) => {
                this.handlePageInputChange(event.detail.toString());
            }, "onTk-blur": () => this.validateAndUpdatePage() }));
    }
    render() {
        const totalPages = this.getTotalPages();
        return (index.h("div", { key: 'ffc6159bb2ead21b8e793abe3e72abc5c33a62bb', class: "tk-pagination-container" }, index.h("div", { key: '64c521213434bc4a3164f0a04cddc7237a8b3e98', class: "tk-pagination-start" }, this.renderTag(totalPages)), this.renderContent(totalPages), index.h("div", { key: 'c26ec2ec063ccbd2c3f84c6c6eb92ea27ee801c0', class: "tk-pagination-end" }, this.renderSelect(), this.renderInput(totalPages))));
    }
    static get watchers() { return {
        "internalCurrentPage": [{
                "internalCurrentPageChanged": 0
            }],
        "currentPage": [{
                "currentPageChanged": 0
            }]
    }; }
};
TkPagination.style = tkPaginationCss();

const tkRadioCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-radio{display:inline-block}.tk-radio-container{display:flex;flex-direction:row;align-items:center;justify-content:flex-start;gap:var(--spacing-s);color:var(--text-darkest);font-family:var(--family-body);font-size:var(--size-sm);font-weight:var(--weight-medium);cursor:pointer;user-select:none}.tk-radio-container.right{flex-direction:row-reverse}.tk-radio-container.width-description{align-items:start}.tk-radio-container input{margin:0;appearance:none;outline:none;display:none}.tk-radio-container .mask{width:20px;min-width:20px;height:20px;background-color:var(--static-light);border:2px solid var(--border-light);border-radius:var(--radius-full);box-sizing:border-box;-moz-box-sizing:border-box;-webkit-box-sizing:border-box;padding:2px}.tk-radio-container .mask div{display:none;width:12px;height:12px;color:var(--static-light);border-radius:var(--radius-full);background-color:var(--states-info-base)}.tk-radio-container .tk-radio-text-holder .tk-radio-description{color:var(--text-base);font-size:var(--size-xs);font-weight:var(--weight-regular)}.tk-radio-container input:checked+.mask{border-color:var(--states-info-lightest);background-color:var(--static-light)}.tk-radio-container input:checked+.mask div{display:block}tk-radio:not([disabled]) .tk-radio-container:hover .mask{background-color:var(--background-lightest)}tk-radio:not([disabled]) .tk-radio-container input:focus-visible+.mask{border-color:var(--states-info-base)}tk-radio:not([disabled]) .tk-radio-container:hover input:checked+.mask{border-color:var(--states-info-light);background-color:var(--states-info-lightest)}tk-radio[disabled] .tk-radio-container{cursor:not-allowed;color:var(--text-sub-base);pointer-events:none}tk-radio[disabled] .tk-radio-container input+.mask{border:2px solid var(--border-light);background-color:var(--background-light)}tk-radio[disabled] .tk-radio-container input+.mask div{background-color:var(--background-sub-base)}tk-radio[invalid]:not([disabled]) .tk-radio-container .mask{border:2px solid var(--states-danger-base);background:var(--static-light)}tk-radio[invalid]:not([disabled]) .tk-radio-container input:checked+.mask{border-color:var(--states-danger-light)}tk-radio[invalid]:not([disabled]) .tk-radio-container input:checked+.mask div{background-color:var(--states-danger-base)}tk-radio[data-type=card]{border-radius:var(--radius-m-base);border:1px solid var(--border-light)}tk-radio[data-type=card] .tk-radio-container{padding:var(--card-switcher-container-v-padding) var(--card-switcher-container-h-padding)}tk-radio[data-type=card]:not([disabled]):hover{background-color:var(--background-lightest)}`;

const TkRadio = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.tkChange = index.createEvent(this, "tk-change", 7);
        if (hostRef.$hostElement$["s-ei"]) {
            this.internals = hostRef.$hostElement$["s-ei"];
        }
        else {
            this.internals = hostRef.$hostElement$.attachInternals();
            hostRef.$hostElement$["s-ei"] = this.internals;
        }
        this.uniqueId = v4.v4();
        /**
         * Controls if radio has custom content.
         * @defaultValue false
         */
        this.hasContentSlot = false;
        /**
         * Disables the radio button if true.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Marks the radio button as checked or unchecked.
         * @defaultValue false
         */
        this.checked = false;
        // radio group ile birlikte kullanımlarda diğer radio ya tıklandığı durumda kendini false yapmaya ihtiyaç olmadığı için bu işlemi radio group tarafında yapıldığı için outside click eventi atanmasına gerek yoktur.
        if (!this.el.closest('tk-radio-group')) {
            this.windowClickHandler = this.handleWindowClick.bind(this);
        }
    }
    componentWillLoad() {
        this.hasContentSlot = !!this.el.querySelector('[slot="content"]');
        this.parentEl = this.el.closest('tk-radio-group');
        if (this.parentEl && !this.position) {
            this.position = this.parentEl.position;
        }
    }
    componentDidRender() {
        this.bindWindowClickListener();
    }
    disconnectedCallback() {
        this.unbindWindowClickListener();
    }
    bindWindowClickListener() {
        window.addEventListener('click', this.windowClickHandler);
    }
    unbindWindowClickListener() {
        window.removeEventListener('click', this.windowClickHandler);
    }
    handleWindowClick(event) {
        var _a, _b;
        let clickedElement = event.target;
        if (clickedElement.tagName !== 'TK-RADIO') {
            clickedElement = clickedElement.closest('tk-radio');
        }
        if (((_a = clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('name')) === null || _a === void 0 ? void 0 : _a.length) > 0 &&
            (clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('name')) == this.name &&
            ((_b = clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('data-tk-radio-id')) === null || _b === void 0 ? void 0 : _b.length) > 0 &&
            (clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('data-tk-radio-id')) !== this.uniqueId &&
            !(clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('disabled'))) {
            this.checked = false;
        }
    }
    handleInputChange() {
        if (!this.disabled) {
            this.checked = true;
            this.tkChange.emit(this.value);
        }
    }
    render() {
        const labelClass = index$2.classNames('tk-radio-container', this.position, {
            'disabled': this.disabled,
            'width-description': !!this.description,
        });
        return (index.h(index.Host, { key: '92262fceb575a73daaccdaec498a70dd9073ede3', "data-tk-radio-id": this.uniqueId, invalid: this.invalid, disabled: this.disabled }, index.h("label", { key: 'e2f683494e14e42f333b449037471d6f0f24e865', htmlFor: this.uniqueId, class: labelClass, "aria-disabled": this.disabled, "aria-invalid": this.invalid }, index.h("input", { key: '7a15a9f05c0ee1c08dc95803404caa1ee6008d59', id: this.uniqueId, type: "radio", name: this.name, value: this.value, checked: this.checked, disabled: this.disabled, onChange: () => this.handleInputChange() }), index.h("div", { key: 'bc1d16c1d7007e67848a67ed652931ca4b218db7', class: "mask" }, index.h("div", { key: '282040082c5e42618c7bf7a28f7d43d088ef8dbf' })), this.hasContentSlot ? (index.h("slot", { name: "content" })) : (index.h("div", { class: "tk-radio-text-holder" }, this.label && index.h("div", { class: "tk-radio-label" }, this.label), this.description && index.h("div", { class: "tk-radio-description" }, this.description))))));
    }
    static get formAssociated() { return true; }
    get el() { return index.getElement(this); }
};
TkRadio.style = tkRadioCss();

const tkTreeViewCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-tree-view{display:inline-flex;flex-direction:column;align-items:flex-start;background:transparent}.tk-tree-view::-webkit-scrollbar{width:4px;height:4px}.tk-tree-view::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-tree-view::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-tree-view::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-tree-view::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-tree-view.stepper{display:flex;flex-direction:row;align-items:flex-start;gap:var(--tree-gap, 12px);overflow-x:auto}.tk-tree-view.stepper .select-all{width:auto}.tk-tree-view.stepper .column .node .tk-tree-view.pointer{height:100%}.tk-tree-view.divided{padding:var(--tree-v-padding, 8px) var(--tree-h-padding, 10px);border-radius:8px;border:1px solid var(--border-light, #e1e4ea)}.tk-tree-view.light{padding:var(--tree-v-padding, 8px) var(--tree-h-padding, 10px);border-radius:8px;border:1px solid var(--border-light, #e1e4ea);background:var(--background-lightest, #f9fafc)}.tk-tree-view.content{width:100%;display:flex;flex-direction:column;gap:var(--tree-gap, 2px);font-family:var(--family-body)}.tk-tree-view .select-all{display:flex;flex-direction:row;align-items:center;justify-content:space-between;gap:4px;cursor:pointer;color:var(--text-dark);user-select:none;width:100%}.tk-tree-view .select-all.large{padding:var(--items-tree-large-v-padding, 10px) var(--items-tree-large-h-padding, 12px);font-size:var(--desktop-body-m-base-size, 16px);line-height:var(--desktop-body-m-base-line-height, 24px)}.tk-tree-view .select-all.base{padding:var(--items-tree-base-v-padding, 6px) var(--items-tree-base-h-padding, 8px);font-size:var(--desktop-body-s-size, 14px);line-height:var(--desktop-body-s-line-height, 20px)}.tk-tree-view .select-all.small{padding:var(--items-tree-small-v-padding, 4px) var(--items-tree-small-h-padding, 6px);font-size:var(--desktop-body-2xs-size, 11px);line-height:var(--line-height-xxs-normal, 16px)}.tk-tree-view .select-all.disabled{color:var(--text-sub-base);cursor:not-allowed}.tk-tree-view .select-all:not(.disabled):hover{color:var(--text-darkest)}.tk-tree-view .select-all .select-all-content{display:flex;flex-direction:row;align-items:center;gap:4px}.tk-tree-view.node{display:flex;align-items:flex-start;border-radius:4px;flex-direction:column;position:relative;width:100%}.tk-tree-view.node.expanded{font-weight:500;gap:var(--tree-gap, 4px)}.tk-tree-view.node.directory{font-weight:500}.tk-tree-view.node.file{font-weight:400}.tk-tree-view.node .tk-tree-view.pointer{background-color:var(--primary-base);width:4px;border-radius:0 4px 4px 0;position:absolute;left:0;top:0;z-index:1}.tk-tree-view.node .tk-tree-view.pointer.base{height:32px}.tk-tree-view.node .tk-tree-view.pointer.small{height:24px}.tk-tree-view.node .tk-tree-view.pointer.large{height:44px}.tk-tree-view.text-container{display:flex;flex-direction:row;width:100%}.tk-tree-view.text-container.large{gap:var(--items-tree-large-gap, 10px)}.tk-tree-view.text-container.base{gap:var(--items-tree-base-gap, 8px)}.tk-tree-view.text-container.small{gap:var(--items-tree-small-gap, 4px)}.tk-tree-view.label{display:flex;flex-direction:row;align-items:center;gap:4px;cursor:pointer;color:var(--text-dark, #525866);user-select:none;border-radius:4px;width:100%;justify-content:space-between}.tk-tree-view.label.large{padding:var(--items-tree-large-v-padding, 10px) var(--items-tree-large-h-padding, 12px)}.tk-tree-view.label.base{padding:var(--items-tree-base-v-padding, 6px) var(--items-tree-base-h-padding, 8px)}.tk-tree-view.label.small{padding:var(--items-tree-small-v-padding, 4px) var(--items-tree-small-h-padding, 6px)}.tk-tree-view.label.selected{background:var(--primary-lightest, #fae6e8);color:var(--text-darkest, #222530)}.tk-tree-view.label.disabled{color:var(--text-sub-base, #99a0ae);cursor:not-allowed}.tk-tree-view.label:not(.disabled,.selected):hover{color:var(--text-darkest, #222530);background:var(--background-lightest, #f9fafc)}.tk-tree-view.text{font-family:var(--desktop-body-s-font);font-style:normal;font-weight:400}.tk-tree-view.text.base{font-size:var(--body-s-size, 14px);line-height:var(--body-s-line-height, 20px)}.tk-tree-view.text.small{font-size:var(--body-2xs-size, 11px);line-height:var(--line-height-xxs-normal, 16px)}.tk-tree-view.text.large{font-size:var(--body-m-base-size, 16px);line-height:var(--body-m-base-line-height, 24px)}.tk-tree-view.children{margin-left:18px;border-left:1px solid var(--border-light, #e1e4ea);padding-left:10px;gap:2px;width:calc(100% - 18px);box-sizing:border-box;}`;

const TkTreeView = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.tkItemClick = index.createEvent(this, "tk-item-click", 7);
        this.tkChange = index.createEvent(this, "tk-change", 7);
        this.tkExpandChange = index.createEvent(this, "tk-expand-change", 7);
        this.isAllSelected = false;
        this.expandedPaths = new Set();
        this.selectedPath = null;
        /**
         * başlangıçta expanded keys lerin seçili gibi görünmesini ve herhangi bir item'a tıklandığında
         * bu state'in false'a çekilerek expandedKeys'lerin sürekli seçili görünmesini engellemek için kurgulanmıştır.s
         */
        this.isInitialLoad = true;
        /**
         * Array of tree items data. This is the primary way to provide data to the tree view.
         */
        this.items = [];
        /**
         * Tree view mode: 'basic' or 'stepper'.
         */
        this.mode = 'basic';
        /**
         * Tree view type: 'basic', 'divided', or 'light'.
         */
        this.type = 'basic';
        /**
         * Tree view size: 'large', 'base' or 'small'.
         */
        this.size = 'base';
        /**
         * If true, disables all interaction with the tree view.
         */
        this.disabled = false;
        /**
         * Icon for branch items (items with children). When empty, no icon is shown.
         */
        this.branchIcon = '';
        /**
         * Icon for leaf items (items without children). When empty, no icon is shown.
         */
        this.leafIcon = '';
        /**
         * Show/hide the badge for children count on directories.
         */
        this.showBadge = true;
        /**
         * Show/hide badges with zero count. Default is true.
         * When false, badges with 0 count will be hidden (works for both selected count and children count).
         */
        this.showZeroCountBadges = true;
        /**
         * If true, enables checkbox selection for tree items.
         */
        this.selectable = false;
        /**
         * Show/hide the pointer icon for selected items.
         */
        this.showPointer = true;
        /**
         * Selection strategy for checkboxes:
         * <br />
         * **all:** selecting a node selects the node itself and all descendants
         * <br />
         * **leaf:** selecting a node selects only leaf descendants (and leaf itself if it is a leaf)
         */
        this.selectionStrategy = 'all';
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        /**
         * The style attribute of column element for stepper mode
         */
        this.stepStyle = null;
        /**
         * If true, shows a "Select All" checkbox above the tree items. Only effective when `selectable` is true.
         */
        this.selectAll = false;
        /**
         * Label for the "Select All" checkbox row.
         * @defaultValue 'Select All'
         */
        this.selectAllLabel = 'Select All';
        /**
         * If true, expands all nodes in basic mode.
         * <br />
         * **Note:** This prop is ignored when expandedKeys is provided.
         *
         */
        this.expandAll = false;
        this.handleToggleUnified = (pathStr, item) => {
            const descendants = this.collectDescendantPaths(pathStr, item);
            if (this.mode === 'stepper') {
                // Generate ancestor paths inline: "0-1-2" -> ["0", "0-1", "0-1-2"]
                const pathParts = pathStr.split('-');
                const ancestors = [];
                for (let i = 1; i <= pathParts.length; i++) {
                    ancestors.push(pathParts.slice(0, i).join('-'));
                }
                const isExpanded = this.expandedPaths.has(pathStr);
                if (isExpanded) {
                    // Stepper: Close self and descendants
                    if (!this.isControlled()) {
                        this.expandedPaths.delete(pathStr);
                        descendants.forEach(descPath => this.expandedPaths.delete(descPath));
                    }
                    if (pathStr === this.selectedPath || descendants.includes(this.selectedPath)) {
                        this.selectedPath = null;
                    }
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const newPaths = new Set(this.expandedPaths);
                        newPaths.delete(pathStr);
                        descendants.forEach(descPath => newPaths.delete(descPath));
                        const deepestPaths = this.getDeepestPaths(newPaths);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                }
                else {
                    // Stepper: Open path to this item
                    if (!this.isControlled()) {
                        this.expandedPaths = new Set(ancestors);
                    }
                    this.handleSelect(pathStr, item);
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const ancestorSet = new Set(ancestors);
                        const deepestPaths = this.getDeepestPaths(ancestorSet);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                    return;
                }
            }
            else {
                if (this.expandedPaths.has(pathStr)) {
                    // Basic: Close self and descendants
                    if (!this.isControlled()) {
                        this.expandedPaths.delete(pathStr);
                        descendants.forEach(descPath => this.expandedPaths.delete(descPath));
                    }
                    if (this.selectedPath === pathStr || descendants.includes(this.selectedPath)) {
                        this.selectedPath = null;
                    }
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const newPaths = new Set(this.expandedPaths);
                        newPaths.delete(pathStr);
                        descendants.forEach(descPath => newPaths.delete(descPath));
                        const deepestPaths = this.getDeepestPaths(newPaths);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                }
                else {
                    // Basic: Open only self
                    if (!this.isControlled()) {
                        this.expandedPaths.add(pathStr);
                    }
                    this.handleSelect(pathStr, item);
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const newPaths = new Set(this.expandedPaths);
                        newPaths.add(pathStr);
                        const deepestPaths = this.getDeepestPaths(newPaths);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                }
            }
        };
        this.handleSelect = (pathStr, item) => {
            this.selectedPath = pathStr;
            this.isInitialLoad = false;
            this.tkItemClick.emit(item);
        };
        /**
         * Get all keys (including branches and leaves) from a list of items recursively.
         */
        this.getAllKeys = (items) => {
            const keys = [];
            for (const item of items) {
                if (item.key)
                    keys.push(item.key);
                if (item.children && item.children.length > 0) {
                    keys.push(...this.getAllKeys(item.children));
                }
            }
            return keys;
        };
        /**
         * Get all child keys recursively
         */
        this.getAllChildKeys = (item) => {
            const keys = [];
            if (item.children && item.children.length > 0) {
                item.children.forEach(child => {
                    keys.push(child.key);
                    keys.push(...this.getAllChildKeys(child));
                });
            }
            return keys;
        };
        /**
         * Get all leaf descendant keys including the item's own key if it is a leaf
         */
        this.getLeafKeys = (item) => {
            const keys = [];
            const isLeaf = !item.children || item.children.length === 0;
            if (isLeaf) {
                if (item.key)
                    keys.push(item.key);
                return keys;
            }
            for (const child of item.children) {
                keys.push(...this.getLeafKeys(child));
            }
            return keys;
        };
        /**
         * Handle checkbox change events.
         */
        this.handleCheckboxChange = (checked, item) => {
            var _a;
            if (this.selectionStrategy === 'all') {
                const childKeys = this.getAllChildKeys(item);
                if (checked) {
                    // Add current item and all children
                    this.value = [...(this.value || []), item.key, ...childKeys];
                }
                else {
                    // Remove current item and all children
                    this.value = (_a = this.value) === null || _a === void 0 ? void 0 : _a.filter(key => key !== item.key && !childKeys.includes(key));
                }
            }
            else {
                // 'leaf' strategy
                const leafKeys = this.getLeafKeys(item);
                if (checked) {
                    this.value = [...(this.value || []), ...leafKeys];
                }
                else {
                    this.value = (this.value || []).filter(key => !leafKeys.includes(key));
                }
                // Ensure we don't keep non-leaf keys selected under leaf strategy
                const allLeafsUnderRoot = this.items.flatMap(root => this.getLeafKeys(root));
                this.value = (this.value || []).filter(key => allLeafsUnderRoot.includes(key));
            }
            // Deduplicate and clean undefined/null
            this.value = Array.from(new Set((this.value || []).filter((k) => !!k)));
            this.tkChange.emit(this.value);
        };
        /**
         * Check if all child nodes are selected and return checkbox state
         */
        this.getCheckboxState = (item) => {
            var _a;
            const isLeaf = !item.children || item.children.length === 0;
            const isDirectlySelected = ((_a = this.value) === null || _a === void 0 ? void 0 : _a.includes(item.key)) || false;
            // Leaf nodes: direct selection matters in both strategies
            if (isLeaf) {
                return { checked: isDirectlySelected, indeterminate: false };
            }
            // Directory nodes: compute based on children
            const childStates = item.children.map(child => this.getCheckboxState(child));
            const checkedChildren = childStates.filter(state => state.checked || state.indeterminate);
            const allChecked = childStates.every(state => state.checked);
            if (this.selectionStrategy === 'all') {
                if (allChecked && checkedChildren.length === item.children.length) {
                    return { checked: true, indeterminate: false };
                }
                else if (checkedChildren.length > 0) {
                    return { checked: false, indeterminate: true };
                }
                else {
                    return { checked: isDirectlySelected, indeterminate: false };
                }
            }
            else {
                // 'leaf' strategy: ignore isDirectlySelected for directories
                if (allChecked && checkedChildren.length === item.children.length) {
                    return { checked: true, indeterminate: false };
                }
                else if (checkedChildren.length > 0) {
                    return { checked: false, indeterminate: true };
                }
                else {
                    return { checked: false, indeterminate: false };
                }
            }
        };
        /**
         * Count selected items within the subtree rooted at the given item
         */
        this.getSelectedCount = (item) => {
            var _a;
            let count = 0;
            if (item.key && ((_a = this.value) === null || _a === void 0 ? void 0 : _a.includes(item.key))) {
                count += 1;
            }
            if (item.children && item.children.length > 0) {
                for (const child of item.children) {
                    count += this.getSelectedCount(child);
                }
            }
            return count;
        };
        /**
         * Handle item click events.
         */
        this.handleItemClick = (pathStr, item, isDisabled, isDirectory) => {
            if (this.disabled || isDisabled)
                return;
            if (isDirectory) {
                this.handleToggleUnified(pathStr, item);
            }
            else {
                // In stepper mode, when a file (leaf) is clicked, collapse any expanded
                // directory at the same level by keeping only the ancestors of the file's parent
                if (this.mode === 'stepper') {
                    const parentPath = pathStr.includes('-') ? pathStr.split('-').slice(0, -1).join('-') : '';
                    const ancestors = [];
                    if (parentPath) {
                        const parentParts = parentPath.split('-');
                        for (let i = 1; i <= parentParts.length; i++) {
                            ancestors.push(parentParts.slice(0, i).join('-'));
                        }
                    }
                    if (this.isControlled()) {
                        const ancestorSet = new Set(ancestors);
                        const deepestPaths = this.getDeepestPaths(ancestorSet);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                    else {
                        this.expandedPaths = new Set(ancestors);
                    }
                }
                this.handleSelect(pathStr, item);
            }
        };
        this.handleSelectAllClick = () => {
            if (this.disabled)
                return;
            if (this.isAllSelected) {
                this.value = [];
            }
            else {
                this.value = this.selectionStrategy === 'leaf' ? this.items.flatMap(item => this.getLeafKeys(item)) : this.getAllKeys(this.items);
            }
            this.tkChange.emit(this.value);
        };
        this.renderItem = (item, basePath = '', index$1, depth = 0) => {
            var _a;
            const pathStr = basePath ? `${basePath}-${index$1}` : `${index$1}`;
            const isDirectory = !!(item.children && item.children.length > 0);
            const isExpanded = this.expandedPaths.has(pathStr);
            const isSelected = this.selectedPath === pathStr || (this.isInitialLoad && ((_a = this.expandedKeys) === null || _a === void 0 ? void 0 : _a.includes(item.key)));
            const isDisabled = this.disabled || item.disabled;
            const nodeClass = index$2.classNames('tk-tree-view', 'node', {
                directory: isDirectory,
                file: !isDirectory,
                expanded: isExpanded,
                selected: isSelected,
                disabled: isDisabled,
            });
            const selectedCount = this.getSelectedCount(item);
            return (index.h("div", { class: nodeClass }, this.showPointer && (isExpanded || isSelected) && index.h("span", { class: index$2.classNames('tk-tree-view', 'pointer', this.size) }), index.h("div", { class: index$2.classNames('tk-tree-view', 'label', {
                    selected: isSelected,
                    disabled: isDisabled,
                }, this.size), onClick: () => {
                    this.handleItemClick(pathStr, item, isDisabled, isDirectory);
                } }, isDirectory && this.mode === 'basic' && index.h("tk-icon", { variant: isSelected ? 'primary' : 'neutral', icon: isExpanded ? 'arrow_drop_down' : 'arrow_right', size: this.size }), this.selectable && (index.h("tk-checkbox", { onClick: e => {
                    e.stopPropagation();
                }, value: this.getCheckboxState(item).checked, indeterminate: this.getCheckboxState(item).indeterminate, disabled: isDisabled, "onTk-change": e => {
                    e.stopPropagation();
                    this.handleCheckboxChange(e.detail, item);
                } })), isDirectory && this.branchIcon && index.h("tk-icon", { icon: this.branchIcon, variant: isSelected ? 'primary' : 'neutral', size: this.size }), !isDirectory && this.leafIcon && index.h("tk-icon", { icon: this.leafIcon, variant: isSelected ? 'primary' : 'neutral', size: this.size }), index.h("div", { class: index$2.classNames('tk-tree-view', 'text-container', this.size) }, index.h("span", { class: index$2.classNames('tk-tree-view', 'text', this.size) }, item.label), (() => {
                var _a, _b, _c, _d, _e, _f, _g, _h, _j;
                const badgeCount = this.selectable ? selectedCount : (_a = item.children) === null || _a === void 0 ? void 0 : _a.length;
                const shouldShowBadge = isDirectory && this.showBadge && (this.showZeroCountBadges || badgeCount > 0);
                return (shouldShowBadge && (index.h("tk-badge", { count: badgeCount, size: this.size, type: (_c = (_b = this.badgeOptions) === null || _b === void 0 ? void 0 : _b.type) !== null && _c !== void 0 ? _c : 'filledlight', variant: (_e = (_d = this.badgeOptions) === null || _d === void 0 ? void 0 : _d.variant) !== null && _e !== void 0 ? _e : 'neutral', rounded: (_g = (_f = this.badgeOptions) === null || _f === void 0 ? void 0 : _f.rounded) !== null && _g !== void 0 ? _g : true, icon: (_h = this.badgeOptions) === null || _h === void 0 ? void 0 : _h.icon, iconPosition: (_j = this.badgeOptions) === null || _j === void 0 ? void 0 : _j.iconPosition })));
            })()), this.mode === 'stepper' && isDirectory && item.children && item.children.length > 0 && (index.h("tk-icon", { variant: isSelected ? 'primary' : 'neutral', icon: !isExpanded ? 'keyboard_arrow_down' : 'keyboard_arrow_right', size: this.size }))), this.mode === 'basic' && isDirectory && isExpanded && item.children && item.children.length > 0 && (index.h("div", { class: index$2.classNames('tk-tree-view', 'children') }, item.children.map((child, childIndex) => this.renderItem(child, pathStr, childIndex, depth + 1))))));
        };
    }
    itemsChanged() {
        this.initializeExpandedPaths();
    }
    modeChanged() {
        this.initializeExpandedPaths();
    }
    expandAllChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            this.initializeExpandedPaths();
        }
    }
    expandedKeysChanged(newValue) {
        if (Array.isArray(newValue)) {
            // For each value, find its index path in the tree
            const indexPaths = [];
            const invalidKeys = [];
            newValue.forEach((key, index) => {
                // Find the index path for this key
                if (this.mode === 'stepper' && index !== 0) {
                    return;
                }
                const indexPath = this.findIndexPath(key);
                if (indexPath) {
                    indexPaths.push(indexPath);
                }
                else {
                    invalidKeys.push(key);
                }
            });
            // Log invalid keys
            if (invalidKeys.length > 0) {
                console.error('Invalid keys given to expandedKeys prop:', invalidKeys);
            }
            this.expandedPaths = this.expandKeysWithAncestors(indexPaths);
            // In controlled mode, clear selection if the selected node is no longer expanded
            if (this.selectedPath !== null) {
                // Check if the selected path or any of its ancestors is collapsed
                const selectedPathParts = this.selectedPath.split('-');
                let isSelectedPathVisible = true;
                // Check each ancestor up to and including the selected path
                for (let i = 1; i <= selectedPathParts.length; i++) {
                    const ancestorPath = selectedPathParts.slice(0, i).join('-');
                    // For directories, they need to be expanded to show their children
                    if (i < selectedPathParts.length && !this.expandedPaths.has(ancestorPath)) {
                        isSelectedPathVisible = false;
                        break;
                    }
                }
                // Clear selection if not visible
                if (!isSelectedPathVisible) {
                    this.selectedPath = null;
                }
            }
        }
    }
    componentWillLoad() {
        this.initializeExpandedPaths();
    }
    componentWillRender() {
        this.isAllSelected = this.computeIsAllSelected();
    }
    /**
     * Check if the component is in controlled mode.
     * Controlled mode is when the parent component manages expansion state via expandedKeys prop.
     * We use Array.isArray() instead of checking undefined to handle edge cases like null.
     */
    isControlled() {
        return Array.isArray(this.expandedKeys);
    }
    /**
     * Find the full index-based path to a given key by searching the tree recursively
     * Example: key "4" -> "0-1-0" (index path from root to target)
     * Returns null if the key is not found
     */
    findIndexPath(targetKey, items = this.items, currentPath = []) {
        for (let i = 0; i < items.length; i++) {
            const item = items[i];
            const newPath = [...currentPath, i];
            if (item.key === targetKey) {
                return newPath.join('-');
            }
            if (item.children && item.children.length > 0) {
                const found = this.findIndexPath(targetKey, item.children, newPath);
                if (found) {
                    return found;
                }
            }
        }
        return null;
    }
    /**
     * Convert an index-based path to its corresponding key (last item in path)
     * Example: "0-1-2" -> "4" (returns only the key of the last item, not the full path)
     * Returns null if the path is invalid
     */
    indexPathToKey(indexPath) {
        const indices = indexPath.split('-').map(Number);
        let currentLevel = this.items;
        // Navigate to the target item
        for (let i = 0; i < indices.length; i++) {
            const index = indices[i];
            if (!currentLevel || index < 0 || index >= currentLevel.length) {
                return null;
            }
            const item = currentLevel[index];
            // If this is the last index, return its key
            if (i === indices.length - 1) {
                return item.key || null;
            }
            // Otherwise, navigate to children
            currentLevel = item.children;
        }
        return null;
    }
    /**
     * Get the deepest (longest) paths from a set of expanded paths
     * Example: Set(["0", "0-1", "0-1-2", "3"]) -> ["0-1-2", "3"]
     * Returns only the paths with maximum depth in each branch
     */
    getDeepestPaths(paths) {
        const allPaths = Array.from(paths);
        // Find paths that don't have any child path in the set
        return allPaths.filter(path => {
            // A path is deepest if no other path starts with it followed by a hyphen
            return !allPaths.some(otherPath => otherPath.startsWith(path + '-'));
        });
    }
    /**
     * Expand keys and automatically include all ancestor paths
     * Example: ["0-1-2"] -> Set(["0", "0-1", "0-1-2"])
     * Note: Expects index-based paths. Validation should be done before calling this method.
     */
    expandKeysWithAncestors(keys) {
        const expandedSet = new Set();
        keys.forEach(key => {
            // Add the key itself
            expandedSet.add(key);
            // Add all ancestor paths
            const parts = key.split('-');
            for (let i = 1; i < parts.length; i++) {
                const ancestorPath = parts.slice(0, i).join('-');
                expandedSet.add(ancestorPath);
            }
        });
        return expandedSet;
    }
    /**
     * Collect all descendant paths recursively
     */
    collectDescendantPaths(pathStr, item) {
        const paths = [];
        if (item.children && item.children.length > 0) {
            item.children.forEach((child, index) => {
                const childPath = `${pathStr}-${index}`;
                paths.push(childPath);
                paths.push(...this.collectDescendantPaths(childPath, child));
            });
        }
        return paths;
    }
    /**
     * Initialize expanded paths based on mode and expandAll configuration
     */
    initializeExpandedPaths() {
        if (!this.items || this.items.length === 0) {
            this.expandedPaths = new Set();
            return;
        }
        // Skip initialization if expandedKeys is being used for control
        if (this.isControlled()) {
            const indexPaths = [];
            const invalidKeys = [];
            this.expandedKeys.forEach((key, index) => {
                // Find the index path for this key
                if (this.mode === 'stepper' && index !== 0) {
                    return;
                }
                const indexPath = this.findIndexPath(key);
                if (indexPath) {
                    indexPaths.push(indexPath);
                }
                else {
                    invalidKeys.push(key);
                }
            });
            if (invalidKeys.length > 0) {
                console.error('Invalid keys given to expandedKeys prop:', invalidKeys);
            }
            this.expandedPaths = this.expandKeysWithAncestors(indexPaths);
            return;
        }
        if (!this.expandAll)
            return;
        const expanded = new Set();
        if (this.mode === 'basic') {
            // Expand all directory nodes
            const traverse = (nodes, base = '') => {
                nodes.forEach((node, idx) => {
                    const path = base ? `${base}-${idx}` : `${idx}`;
                    if (node.children && node.children.length > 0) {
                        expanded.add(path);
                        traverse(node.children, path);
                    }
                });
            };
            traverse(this.items);
        }
        else if (this.mode === 'stepper') {
            // Expand the first directory all the way down
            let level = this.items;
            let base = '';
            while (level && level.length > 0) {
                const idx = level.findIndex(n => n.children && n.children.length > 0);
                if (idx === -1)
                    break;
                const path = base ? `${base}-${idx}` : `${idx}`;
                expanded.add(path);
                base = path;
                level = level[idx].children;
            }
        }
        this.expandedPaths = expanded;
    }
    /**
     * Get stepper columns for stepper mode display.
     */
    getStepperColumns() {
        const columns = [];
        let currentLevel = this.items;
        let currentBasePath = '';
        while (currentLevel && currentLevel.length > 0) {
            columns.push({ items: currentLevel, basePath: currentBasePath });
            const expandedItem = currentLevel.find((_, index) => {
                const itemPath = currentBasePath ? `${currentBasePath}-${index}` : `${index}`;
                return this.expandedPaths.has(itemPath);
            });
            if (expandedItem && expandedItem.children && expandedItem.children.length > 0) {
                const expandedIndex = currentLevel.indexOf(expandedItem);
                currentBasePath = currentBasePath ? `${currentBasePath}-${expandedIndex}` : `${expandedIndex}`;
                currentLevel = expandedItem.children;
            }
            else {
                break;
            }
        }
        return columns;
    }
    computeIsAllSelected() {
        const allKeys = this.selectionStrategy === 'leaf' ? this.items.flatMap(item => this.getLeafKeys(item)) : this.getAllKeys(this.items);
        return allKeys.length > 0 && allKeys.every(k => { var _a; return (_a = this.value) === null || _a === void 0 ? void 0 : _a.includes(k); });
    }
    renderSelectAll() {
        var _a, _b, _c, _d;
        if (!this.selectAll || !this.selectable)
            return null;
        const isIndeterminate = !this.isAllSelected && ((_b = (_a = this.value) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) > 0;
        const selectedCount = (_d = (_c = this.value) === null || _c === void 0 ? void 0 : _c.length) !== null && _d !== void 0 ? _d : 0;
        return (index.h("div", { class: index$2.classNames('select-all', this.size, { disabled: this.disabled }), onClick: this.handleSelectAllClick }, index.h("div", { class: index$2.classNames('select-all-content') }, index.h("tk-checkbox", { value: this.isAllSelected, indeterminate: isIndeterminate, disabled: this.disabled, onClick: e => e.stopPropagation(), "onTk-change": e => {
                e.stopPropagation();
                this.handleSelectAllClick();
            } }), index.h("div", null, this.selectAllLabel)), selectedCount > 0 && index.h("tk-badge", { count: selectedCount, size: this.size, type: "filledlight", variant: "neutral", rounded: true })));
    }
    render() {
        if (!this.items || !this.items.length)
            return null;
        if (this.mode === 'stepper') {
            const columns = this.getStepperColumns();
            return (index.h("div", { class: index$2.classNames('tk-tree-view', 'stepper'), style: this.containerStyle }, this.renderSelectAll(), columns.map((column, idx) => (index.h("div", { class: index$2.classNames('tk-tree-view', 'column', 'divided'), key: idx, style: this.stepStyle }, column.items.map((item, itemIndex) => this.renderItem(item, column.basePath, itemIndex)))))));
        }
        else {
            const rootClasses = index$2.classNames('tk-tree-view', this.type, { disabled: this.disabled });
            return (index.h("div", { class: rootClasses, style: this.containerStyle }, this.renderSelectAll(), index.h("div", { class: index$2.classNames('tk-tree-view', 'content') }, this.items.map((item, index) => this.renderItem(item, '', index)))));
        }
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "items": [{
                "itemsChanged": 0
            }],
        "mode": [{
                "modeChanged": 0
            }],
        "expandAll": [{
                "expandAllChanged": 0
            }],
        "expandedKeys": [{
                "expandedKeysChanged": 0
            }]
    }; }
};
TkTreeView.style = tkTreeViewCss();

exports.tk_datepicker = TkDatePicker;
exports.tk_pagination = TkPagination;
exports.tk_radio = TkRadio;
exports.tk_tree_view = TkTreeView;
