// @takeoff-ui/core@0.11.9/dist/cjs/tk-gantt-chart.cjs.entry.js
'use strict';

var index = require('./index-CUpr0td9.js');
var index$1 = require('./index-DmcCMEYj.js');
var testIdUtils = require('./test-id-utils-CAIe6ggJ.js');

/** Default row height in pixels. */
const DEFAULT_ROW_HEIGHT = 40;
/** Default width of the left-side panel in pixels. */
const DEFAULT_PANEL_WIDTH = 320;
/** Default header row height in pixels (each of the two header layers). */
const DEFAULT_HEADER_HEIGHT = 36;
/** Default week start day (1 = Monday). */
const DEFAULT_WEEK_START_DAY = 1;
/** Base unit width in pixels per day for each view type. */
const VIEW_TYPE_DAY_WIDTH = {
    weekly: 40,
    monthly: 24,
    quarterly: 6,
    yearly: 2,
};
/** Default columns shown in the left panel. */
const DEFAULT_COLUMNS = [
    { field: 'name', header: 'Task Name', width: '200px' },
    { field: 'startDate', header: 'Start', width: '100px' },
    { field: 'endDate', header: 'End', width: '100px' },
];

// ---------------------------------------------------------------------------
// Date helpers
// ---------------------------------------------------------------------------
/** Strip time portion and return a pure-date Date at midnight UTC. */
function toDateOnly(input) {
    const d = typeof input === 'string' ? new Date(input) : new Date(input.getTime());
    d.setHours(0, 0, 0, 0);
    return d;
}
/** Number of calendar days between two dates (inclusive of start, exclusive of end). */
function daysBetween(a, b) {
    const msPerDay = 86400000;
    return Math.max(Math.round((b.getTime() - a.getTime()) / msPerDay), 0);
}
/** Add N days to a date (immutable). */
function addDays(date, days) {
    const d = new Date(date.getTime());
    d.setDate(d.getDate() + days);
    return d;
}
/** Is the given date a Saturday (6) or Sunday (0)? */
function isWeekend(date) {
    const day = date.getDay();
    return day === 0 || day === 6;
}
/** Get the ISO week number for a given date. */
function getWeekNumber(date) {
    const d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
    d.setUTCDate(d.getUTCDate() + 4 - (d.getUTCDay() || 7));
    const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
    return Math.ceil(((d.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);
}
/** Get the start of the week containing `date` based on the configured start day. */
function getWeekStart(date, weekStartDay) {
    const d = new Date(date.getTime());
    d.setHours(0, 0, 0, 0);
    const diff = (d.getDay() - weekStartDay + 7) % 7;
    d.setDate(d.getDate() - diff);
    return d;
}
/** Get the quarter (1-4) for a date. */
function getQuarter(date) {
    return Math.floor(date.getMonth() / 3) + 1;
}
// ---------------------------------------------------------------------------
// Date range helpers
// ---------------------------------------------------------------------------
/** Compute the overall min/max date range across all tasks (recursively). */
function getDateRange(tasks) {
    let min = Infinity;
    let max = -Infinity;
    const walk = (items) => {
        var _a, _b;
        for (const task of items) {
            const start = toDateOnly(task.startDate).getTime();
            const end = toDateOnly(task.endDate).getTime();
            if (start < min)
                min = start;
            if (end > max)
                max = end;
            // Also consider segment date ranges
            if ((_a = task.segments) === null || _a === void 0 ? void 0 : _a.length) {
                for (const seg of task.segments) {
                    const segStart = toDateOnly(seg.startDate).getTime();
                    const segEnd = toDateOnly(seg.endDate).getTime();
                    if (segStart < min)
                        min = segStart;
                    if (segEnd > max)
                        max = segEnd;
                }
            }
            if ((_b = task.children) === null || _b === void 0 ? void 0 : _b.length)
                walk(task.children);
        }
    };
    walk(tasks);
    return { min: new Date(min), max: addDays(new Date(max), 1) };
}
// ---------------------------------------------------------------------------
// Auto-fit view type
// ---------------------------------------------------------------------------
/** Determine the best view type based on the task date range span. */
function autoFitViewType(tasks) {
    if (!tasks.length)
        return 'monthly';
    const { min, max } = getDateRange(tasks);
    const totalDays = daysBetween(min, max);
    if (totalDays <= 14)
        return 'weekly';
    if (totalDays <= 90)
        return 'monthly';
    if (totalDays <= 365)
        return 'quarterly';
    return 'yearly';
}
// ---------------------------------------------------------------------------
// Flatten tree
// ---------------------------------------------------------------------------
/** Flatten the task tree into rows, respecting expanded state. */
function flattenTasks(tasks, expandedIds, depth = 0) {
    var _a;
    const rows = [];
    for (const task of tasks) {
        rows.push({ task, depth });
        if (((_a = task.children) === null || _a === void 0 ? void 0 : _a.length) && expandedIds.has(task.id)) {
            rows.push(...flattenTasks(task.children, expandedIds, depth + 1));
        }
    }
    return rows;
}
// ---------------------------------------------------------------------------
// Task bar positioning
// ---------------------------------------------------------------------------
/** Returns left offset (px) and width (px) for a task bar relative to minDate. */
function getTaskBarPosition(task, minDate, viewType) {
    const dayWidth = VIEW_TYPE_DAY_WIDTH[viewType];
    const start = toDateOnly(task.startDate);
    const end = toDateOnly(task.endDate);
    const left = daysBetween(minDate, start) * dayWidth;
    const width = Math.max((daysBetween(start, end) + 1) * dayWidth, dayWidth);
    return { left, width };
}
/** Returns the pixel left offset and width for a date range relative to the timeline start. */
function getBarPositionFromDates(startDate, endDate, minDate, viewType) {
    const dayWidth = VIEW_TYPE_DAY_WIDTH[viewType];
    const start = toDateOnly(startDate);
    const end = toDateOnly(endDate);
    const left = daysBetween(minDate, start) * dayWidth;
    const width = Math.max((daysBetween(start, end) + 1) * dayWidth, dayWidth);
    return { left, width };
}
/** Returns the pixel left offset for an indicator date. */
function getIndicatorPosition(dateStr, minDate, viewType) {
    const dayWidth = VIEW_TYPE_DAY_WIDTH[viewType];
    const d = toDateOnly(dateStr);
    return daysBetween(minDate, d) * dayWidth;
}
// ---------------------------------------------------------------------------
// Timeline range with padding
// ---------------------------------------------------------------------------
/** Compute padded timeline start/end based on view type. */
function getTimelineRange(tasks, viewType, weekStartDay) {
    if (!tasks.length) {
        const today = toDateOnly(new Date());
        return { start: today, end: addDays(today, 30), totalDays: 30 };
    }
    const { min, max } = getDateRange(tasks);
    let start;
    let end;
    switch (viewType) {
        case 'weekly': {
            start = getWeekStart(min, weekStartDay);
            end = addDays(getWeekStart(max, weekStartDay), 7);
            break;
        }
        case 'monthly': {
            start = new Date(min.getFullYear(), min.getMonth(), 1);
            end = new Date(max.getFullYear(), max.getMonth() + 1, 1);
            break;
        }
        case 'quarterly': {
            const qStart = Math.floor(min.getMonth() / 3) * 3;
            start = new Date(min.getFullYear(), qStart, 1);
            const qEnd = Math.floor(max.getMonth() / 3) * 3 + 3;
            end = new Date(max.getFullYear(), qEnd, 1);
            break;
        }
        case 'yearly': {
            start = new Date(min.getFullYear(), 0, 1);
            end = new Date(max.getFullYear() + 1, 0, 1);
            break;
        }
        default: {
            start = min;
            end = max;
        }
    }
    const totalDays = daysBetween(start, end);
    return { start, end, totalDays };
}
// ---------------------------------------------------------------------------
// Header generation
// ---------------------------------------------------------------------------
/** Generate primary (upper) header cells. */
function generatePrimaryHeaders(start, end, viewType, locale, _weekStartDay) {
    const cells = [];
    switch (viewType) {
        case 'weekly': {
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const weekEnd = addDays(cursor, 6);
                const label = formatWeekRange(cursor, weekEnd, locale);
                cells.push({ label, span: 7, startDate: new Date(cursor.getTime()) });
                cursor = addDays(cursor, 7);
            }
            break;
        }
        case 'monthly': {
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const year = cursor.getFullYear();
                const month = cursor.getMonth();
                const monthEnd = new Date(year, month + 1, 1);
                const effectiveEnd = monthEnd < end ? monthEnd : end;
                const span = daysBetween(cursor, effectiveEnd);
                const label = cursor.toLocaleString(locale, { month: 'long', year: 'numeric' });
                cells.push({ label, span, startDate: new Date(cursor.getTime()) });
                cursor = monthEnd;
            }
            break;
        }
        case 'quarterly': {
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const q = getQuarter(cursor);
                const year = cursor.getFullYear();
                const qEnd = new Date(year, q * 3, 1);
                const effectiveEnd = qEnd < end ? qEnd : end;
                const span = daysBetween(cursor, effectiveEnd);
                const label = `Q${q} ${year}`;
                cells.push({ label, span, startDate: new Date(cursor.getTime()) });
                cursor = qEnd;
            }
            break;
        }
        case 'yearly': {
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const year = cursor.getFullYear();
                const yearEnd = new Date(year + 1, 0, 1);
                const effectiveEnd = yearEnd < end ? yearEnd : end;
                const span = daysBetween(cursor, effectiveEnd);
                const label = `${year}`;
                cells.push({ label, span, startDate: new Date(cursor.getTime()) });
                cursor = yearEnd;
            }
            break;
        }
    }
    return cells;
}
/** Generate secondary (lower) header cells. */
function generateSecondaryHeaders(start, end, viewType, locale, weekStartDay, secondaryMode = 'days') {
    const cells = [];
    if (secondaryMode === 'weeks') {
        return generateWeekHeaders(start, end, locale, weekStartDay);
    }
    switch (viewType) {
        case 'weekly': {
            // Days of the week
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const label = cursor.toLocaleString(locale, { weekday: 'short' });
                cells.push({ label, span: 1, startDate: new Date(cursor.getTime()) });
                cursor = addDays(cursor, 1);
            }
            break;
        }
        case 'monthly': {
            // Day numbers
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const label = `${cursor.getDate()}`;
                cells.push({ label, span: 1, startDate: new Date(cursor.getTime()) });
                cursor = addDays(cursor, 1);
            }
            break;
        }
        case 'quarterly': {
            // Months
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const year = cursor.getFullYear();
                const month = cursor.getMonth();
                const monthEnd = new Date(year, month + 1, 1);
                const effectiveEnd = monthEnd < end ? monthEnd : end;
                const span = daysBetween(cursor, effectiveEnd);
                const label = cursor.toLocaleString(locale, { month: 'short' });
                cells.push({ label, span, startDate: new Date(cursor.getTime()) });
                cursor = monthEnd;
            }
            break;
        }
        case 'yearly': {
            // Months
            let cursor = new Date(start.getTime());
            while (cursor < end) {
                const year = cursor.getFullYear();
                const month = cursor.getMonth();
                const monthEnd = new Date(year, month + 1, 1);
                const effectiveEnd = monthEnd < end ? monthEnd : end;
                const span = daysBetween(cursor, effectiveEnd);
                const label = cursor.toLocaleString(locale, { month: 'short' });
                cells.push({ label, span, startDate: new Date(cursor.getTime()) });
                cursor = monthEnd;
            }
            break;
        }
    }
    return cells;
}
/** Generate week number header cells (W1, W2, etc.). */
function generateWeekHeaders(start, end, locale, weekStartDay) {
    const cells = [];
    let cursor = getWeekStart(start, weekStartDay);
    // If week start is before range start, adjust the first cell span
    if (cursor < start) {
        cursor = new Date(start.getTime());
    }
    while (cursor < end) {
        const weekNum = getWeekNumber(cursor);
        const weekPrefix = getWeekPrefix(locale);
        const nextWeekStart = addDays(getWeekStart(cursor, weekStartDay), 7);
        const effectiveEnd = nextWeekStart < end ? nextWeekStart : end;
        const span = daysBetween(cursor, effectiveEnd);
        if (span > 0) {
            cells.push({ label: `${weekPrefix}${weekNum}`, span, startDate: new Date(cursor.getTime()) });
        }
        cursor = nextWeekStart;
    }
    return cells;
}
/** Get the locale-aware week prefix (W for English, H for Turkish, etc.). */
function getWeekPrefix(locale) {
    const lower = locale.toLowerCase();
    if (lower.startsWith('tr'))
        return 'H';
    return 'W';
}
/** Format a week range label such as "Jan 13–19". */
function formatWeekRange(start, end, locale) {
    const startStr = start.toLocaleString(locale, { month: 'short', day: 'numeric' });
    const endDay = end.getDate();
    return `${startStr}–${endDay}`;
}
// ---------------------------------------------------------------------------
// Holiday helpers
// ---------------------------------------------------------------------------
/** Check whether a date is in the holidays list. */
function isHoliday(date, holidays) {
    const ts = toDateOnly(date).getTime();
    return holidays.some(h => toDateOnly(h.date).getTime() === ts);
}

const tkGanttChartCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}.tk-gantt-chart{display:flex;position:relative;overflow:hidden;border:1px solid var(--border-light);border-radius:var(--radius-m-base);background:var(--static-white);font-size:var(--desktop-body-s-size);color:var(--text-darkest);font-family:var(--family-body)}.tk-gantt-chart-panel{flex-shrink:0;border-right:1px solid var(--border-light);overflow-y:auto;overflow-x:auto}.tk-gantt-chart-panel::-webkit-scrollbar{width:4px;height:4px}.tk-gantt-chart-panel::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-gantt-chart-panel::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-gantt-chart-panel::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-gantt-chart-panel::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-gantt-chart-panel-header{display:flex;align-items:center;position:sticky;top:0;z-index:2}.tk-gantt-chart-panel-header-cell{display:flex;align-items:center;padding:0 8px;font-weight:600;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;height:100%;border-right:1px solid var(--border-light);background:var(--background-lightest);border-bottom:1px solid var(--border-light)}.tk-gantt-chart-panel-header-cell:last-child{border-right:none}.tk-gantt-chart-panel-body{position:relative}.tk-gantt-chart-panel-row{display:flex;align-items:center}.tk-gantt-chart-panel-row:last-child .tk-gantt-chart-panel-cell{border-bottom:none}.tk-gantt-chart-panel-row:hover{background:var(--background-lightest)}.tk-gantt-chart-panel-cell{display:flex;align-items:center;gap:4px;padding:0 8px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;height:100%;border-right:1px solid var(--border-light);border-bottom:1px solid var(--border-light)}.tk-gantt-chart-panel-cell:last-child{border-right:none}.tk-gantt-chart-expand-btn{display:inline-flex;align-items:center;justify-content:center;background:none;border:none;cursor:pointer;padding:0;margin-right:4px;transition:transform 0.2s ease;color:var(--icon-base)}.tk-gantt-chart-expand-btn:hover{color:var(--icon-darkest)}.tk-gantt-chart-expand-btn-open{transform:rotate(90deg)}.tk-gantt-chart-timeline{flex:1;overflow:auto;position:relative}.tk-gantt-chart-timeline::-webkit-scrollbar{width:4px;height:4px}.tk-gantt-chart-timeline::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-gantt-chart-timeline::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-gantt-chart-timeline::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-gantt-chart-timeline::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-gantt-chart-timeline-headers{position:sticky;top:0;z-index:2;background:var(--background-lightest);border-bottom:1px solid var(--border-light)}.tk-gantt-chart-timeline-header{display:flex}.tk-gantt-chart-timeline-header-primary{border-bottom:1px solid var(--border-light)}.tk-gantt-chart-timeline-header-cell{display:flex;align-items:center;justify-content:center;border-right:1px solid var(--border-light);font-weight:500;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;font-size:var(--desktop-body-xs-size, 11px)}.tk-gantt-chart-timeline-header-cell:last-child{border-right:none}.tk-gantt-chart-timeline-header-cell-weekend{background:var(--neutral-100);color:var(--text-light)}.tk-gantt-chart-timeline-header-cell-holiday{background:var(--primary-100, #fff0f0);color:var(--primary-base)}.tk-gantt-chart-timeline-body{position:relative}.tk-gantt-chart-timeline-row{position:relative;display:flex;align-items:center;border-bottom:1px solid var(--border-light)}.tk-gantt-chart-timeline-row .custom-task-bar{width:100%;height:100%}.tk-gantt-chart-timeline-row:last-child{border-bottom:none}.tk-gantt-chart-column-highlight{position:absolute;top:0;pointer-events:none}.tk-gantt-chart-weekend{background:rgba(0, 0, 0, 0.03)}.tk-gantt-chart-holiday{background:rgba(201, 0, 25, 0.06)}.tk-gantt-chart-task-bar{position:absolute;top:50%;transform:translateY(-50%);border-radius:var(--radius-s-base, 4px);background:var(--primary-base);cursor:pointer;display:flex;align-items:center;overflow:hidden;transition:opacity 0.15s ease}.tk-gantt-chart-task-bar:hover{opacity:0.85}.tk-gantt-chart-task-bar-custom{background:none}.tk-gantt-chart-task-bar-segment{border-radius:var(--radius-s-base, 4px)}.tk-gantt-chart-task-bar-progress{position:absolute;left:0;top:0;height:100%;background:rgba(255, 255, 255, 0.35);border-radius:var(--radius-s-base, 4px) 0 0 var(--radius-s-base, 4px);pointer-events:none}.tk-gantt-chart-task-bar-label{position:relative;z-index:1;padding:0 6px;color:var(--static-white);font-size:var(--desktop-body-xs-size, 11px);white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.tk-gantt-chart-indicator{position:absolute;top:0;width:0;border-left-width:2px;z-index:1;pointer-events:none}.tk-gantt-chart-indicator-label{position:absolute;top:-18px;left:4px;font-size:var(--desktop-body-xs-size, 10px);white-space:nowrap;color:inherit;font-weight:500}.tk-gantt-chart-tooltip{position:absolute;z-index:10;padding:8px 12px;border-radius:var(--radius-s-base, 4px);background:var(--frames-darkest);color:var(--static-white);font-size:var(--desktop-body-xs-size, 11px);pointer-events:none;box-shadow:0 2px 8px rgba(0, 0, 0, 0.15);min-width:140px}.tk-gantt-chart-tooltip-title{font-weight:600;margin-bottom:4px}.tk-gantt-chart-tooltip-row{line-height:1.6}.tk-gantt-chart-tooltip-label{font-weight:500;margin-right:4px}.tk-gantt-chart-empty{display:flex;align-items:center;justify-content:center;width:100%;min-height:200px;color:var(--text-light)}`;

const TkGanttChart = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.tkTaskClick = index.createEvent(this, "tk-task-click", 7);
        this.tkTaskToggle = index.createEvent(this, "tk-task-toggle", 7);
        // ---------------------------------------------------------------------------
        // State
        // ---------------------------------------------------------------------------
        this.expandedIds = new Set();
        this.computedViewType = 'monthly';
        // ---------------------------------------------------------------------------
        // Props
        // ---------------------------------------------------------------------------
        /**
         * Array of tasks to display. Each task may contain nested children for sub-tasks.
         */
        this.tasks = [];
        /**
         * Column definitions for the left-side task list panel.
         * Columns are fed from the task data.
         */
        this.columns = DEFAULT_COLUMNS;
        /**
         * Secondary header display mode.
         * - `'days'` shows individual day numbers or day names
         * - `'weeks'` shows week numbers (W1, W2 … / H1, H2 … for Turkish locale)
         * @defaultValue 'days'
         */
        this.secondaryHeaderMode = 'days';
        /**
         * Day the week starts on (0 = Sunday … 6 = Saturday).
         * @defaultValue 1 (Monday)
         */
        this.weekStartDay = DEFAULT_WEEK_START_DAY;
        /**
         * Array of indicator lines rendered on the timeline (e.g. today, startDate, freezeDate).
         */
        this.indicators = [];
        /**
         * Whether to automatically show a today indicator line.
         * @defaultValue true
         */
        this.showTodayIndicator = true;
        /**
         * Whether to highlight weekend columns with a different background.
         * @defaultValue true
         */
        this.highlightWeekends = true;
        /**
         * Array of holiday dates to highlight on the timeline.
         */
        this.holidays = [];
        /**
         * Row height in pixels.
         * @defaultValue 40
         */
        this.rowHeight = DEFAULT_ROW_HEIGHT;
        /**
         * Width of the left-side task list panel in pixels.
         * @defaultValue 320
         */
        this.panelWidth = DEFAULT_PANEL_WIDTH;
        /**
         * Whether to hide the left-side task list panel.
         * @defaultValue false
         */
        this.hidePanel = false;
        /**
         * Locale string for date formatting (e.g. 'tr-TR', 'en-US').
         * Uses native `toLocaleString` under the hood.
         * @defaultValue 'default'
         */
        this.locale = 'default';
    }
    tasksChanged() {
        this.recalculate();
    }
    viewTypeChanged() {
        this.recalculate();
    }
    // ---------------------------------------------------------------------------
    // Lifecycle
    // ---------------------------------------------------------------------------
    componentWillLoad() {
        this.recalculate();
    }
    // ---------------------------------------------------------------------------
    // Private methods
    // ---------------------------------------------------------------------------
    recalculate() {
        var _a;
        this.computedViewType = (_a = this.viewType) !== null && _a !== void 0 ? _a : autoFitViewType(this.tasks);
    }
    // ---------------------------------------------------------------------------
    // Handle event handlers
    // ---------------------------------------------------------------------------
    handleToggleRow(task, e) {
        e.stopPropagation();
        const newSet = new Set(this.expandedIds);
        const isExpanded = newSet.has(task.id);
        if (isExpanded) {
            newSet.delete(task.id);
        }
        else {
            newSet.add(task.id);
        }
        this.expandedIds = newSet;
        this.tkTaskToggle.emit({ task, expanded: !isExpanded });
    }
    handleTaskClick(task) {
        this.tkTaskClick.emit(task);
    }
    handleTimelineScroll(e) {
        const target = e.target;
        if (this.panelRef) {
            this.panelRef.scrollTop = target.scrollTop;
        }
    }
    handlePanelScroll(e) {
        const target = e.target;
        if (this.timelineRef) {
            this.timelineRef.scrollTop = target.scrollTop;
        }
    }
    // ---------------------------------------------------------------------------
    // Create factory methods
    // ---------------------------------------------------------------------------
    createTaskBar(task, barPos, segment, segmentIndex) {
        var _a, _b, _c;
        const taskKey = String(task.id);
        const barKey = segmentIndex != null ? `${taskKey}-segment-${segmentIndex}` : taskKey;
        const progress = (_a = segment === null || segment === void 0 ? void 0 : segment.progress) !== null && _a !== void 0 ? _a : task.progress;
        const barStyle = (_b = segment === null || segment === void 0 ? void 0 : segment.style) !== null && _b !== void 0 ? _b : task.style;
        const barLabel = (_c = segment === null || segment === void 0 ? void 0 : segment.label) !== null && _c !== void 0 ? _c : task.name;
        const commonStyle = Object.assign({ left: `${barPos.left}px`, width: `${barPos.width}px`, height: `${this.rowHeight - 12}px` }, barStyle);
        const tooltipContent = this.createTooltipContent(task, barKey);
        if (this.taskBarHtml) {
            const content = this.taskBarHtml(task);
            const barContent = typeof content === 'string' ? (index.h("div", { class: "custom-task-bar", innerHTML: content, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'custom-task-bar', barKey) })) : (index.h("div", { class: "custom-task-bar", ref: el => el && el.replaceChildren(content), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'custom-task-bar', barKey) }));
            return (index.h("tk-tooltip", { position: "top", variant: "dark", dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'custom-task-bar-tooltip', barKey) }, index.h("div", { slot: "trigger", class: "tk-gantt-chart-task-bar tk-gantt-chart-task-bar-custom", style: commonStyle, onClick: () => this.handleTaskClick(task), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'custom-task-bar-trigger', barKey) }, barContent), index.h("div", { slot: "content", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'custom-task-bar-tooltip-content', barKey) }, tooltipContent)));
        }
        return (index.h("tk-tooltip", { position: "top", variant: "dark", dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'task-bar-tooltip', barKey) }, index.h("div", { slot: "trigger", class: index$1.classNames('tk-gantt-chart-task-bar', { 'tk-gantt-chart-task-bar-segment': segmentIndex != null }), style: commonStyle, onClick: () => this.handleTaskClick(task), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'task-bar', barKey) }, progress != null && progress > 0 && (index.h("div", { class: "tk-gantt-chart-task-bar-progress", style: { width: `${Math.min(progress, 100)}%` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'task-bar-progress', barKey) })), index.h("span", { class: "tk-gantt-chart-task-bar-label", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'task-bar-label', barKey) }, barLabel)), index.h("div", { slot: "content", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'task-bar-tooltip-content', barKey) }, tooltipContent)));
    }
    createIndicatorLine(indicator, minDate, bodyHeight, indicatorKey) {
        const left = getIndicatorPosition(indicator.date, minDate, this.computedViewType);
        return (index.h("div", { class: index$1.classNames('tk-gantt-chart-indicator', {
                'tk-gantt-chart-indicator-dashed': indicator.lineStyle === 'dashed',
            }), style: {
                left: `${left}px`,
                height: `${bodyHeight}px`,
                borderLeftColor: indicator.color || 'var(--primary-base)',
                borderLeftStyle: indicator.lineStyle === 'dashed' ? 'dashed' : 'solid',
            }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'indicator', indicatorKey) }, indicator.label && (index.h("span", { class: "tk-gantt-chart-indicator-label", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'indicator-label', indicatorKey) }, indicator.label))));
    }
    createTooltipContent(task, tooltipKey) {
        if (this.tooltipHtml) {
            const content = this.tooltipHtml(task);
            if (typeof content === 'string') {
                return index.h("div", { innerHTML: content, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'custom-tooltip', tooltipKey) });
            }
            return index.h("div", { ref: el => el && el.replaceChildren(content), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'custom-tooltip', tooltipKey) });
        }
        const startLabel = toDateOnly(task.startDate).toLocaleDateString(this.locale);
        const endLabel = toDateOnly(task.endDate).toLocaleDateString(this.locale);
        return (index.h(index.Fragment, null, index.h("div", { class: "tk-gantt-chart-tooltip-title", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'tooltip-title', tooltipKey) }, task.name), index.h("div", { class: "tk-gantt-chart-tooltip-row", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'tooltip-start', tooltipKey) }, index.h("span", { class: "tk-gantt-chart-tooltip-label", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'tooltip-start-label', tooltipKey) }, "Start:"), startLabel), index.h("div", { class: "tk-gantt-chart-tooltip-row", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'tooltip-end', tooltipKey) }, index.h("span", { class: "tk-gantt-chart-tooltip-label", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'tooltip-end-label', tooltipKey) }, "End:"), endLabel), task.progress != null && (index.h("div", { class: "tk-gantt-chart-tooltip-row", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'tooltip-progress', tooltipKey) }, index.h("span", { class: "tk-gantt-chart-tooltip-label", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'tooltip-progress-label', tooltipKey) }, "Progress:"), task.progress, "%"))));
    }
    // ---------------------------------------------------------------------------
    // Render helpers
    // ---------------------------------------------------------------------------
    renderPanel(flatRows) {
        if (this.hidePanel)
            return null;
        const headerHeight = DEFAULT_HEADER_HEIGHT * 2; // two header layers
        return (index.h("div", { class: "tk-gantt-chart-panel", style: { width: `${this.panelWidth}px` }, ref: el => (this.panelRef = el), onScroll: (e) => this.handlePanelScroll(e), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel') }, index.h("div", { class: "tk-gantt-chart-panel-header", style: { height: `${headerHeight}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-header') }, this.columns.map(col => (index.h("div", { class: "tk-gantt-chart-panel-header-cell", style: { width: col.width || 'auto', minWidth: col.width || '80px' }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-header-cell', String(col.field)) }, col.header)))), index.h("div", { class: "tk-gantt-chart-panel-body", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-body') }, flatRows.map(({ task, depth }) => (index.h("div", { class: "tk-gantt-chart-panel-row", style: { height: `${this.rowHeight}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-row', String(task.id)) }, this.columns.map((col, colIndex) => {
            var _a, _b;
            return (index.h("div", { class: "tk-gantt-chart-panel-cell", style: {
                    width: col.width || 'auto',
                    minWidth: col.width || '80px',
                    paddingLeft: colIndex === 0 ? `${depth * 20 + 8}px` : '8px',
                }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-cell', String(task.id), String(col.field)) }, colIndex === 0 && ((_b = (_a = task.children) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) > 0 && (index.h("button", { class: index$1.classNames('tk-gantt-chart-expand-btn', {
                    'tk-gantt-chart-expand-btn-open': this.expandedIds.has(task.id),
                }), onClick: (e) => this.handleToggleRow(task, e), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-expand-button', String(task.id)) }, index.h("tk-icon", { icon: "chevron_right", size: "small", dataTestid: testIdUtils.getDataTestId(this.dataTestid, 'panel-expand-icon', String(task.id)) }))), col.html ? (index.h("span", { innerHTML: col.html(task), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-cell-content', String(task.id), String(col.field)) })) : (index.h("span", { "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'panel-cell-content', String(task.id), String(col.field)) }, task[col.field]))));
        })))))));
    }
    renderTimelineHeader(start, end, totalWidth) {
        const vt = this.computedViewType;
        const dayWidth = VIEW_TYPE_DAY_WIDTH[vt];
        const primaryCells = generatePrimaryHeaders(start, end, vt, this.locale);
        const secondaryCells = generateSecondaryHeaders(start, end, vt, this.locale, this.weekStartDay, this.secondaryHeaderMode);
        return (index.h("div", { class: "tk-gantt-chart-timeline-headers", style: { width: `${totalWidth}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline-headers') }, index.h("div", { class: "tk-gantt-chart-timeline-header tk-gantt-chart-timeline-header-primary", style: { height: `${DEFAULT_HEADER_HEIGHT}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline-primary') }, primaryCells.map((cell, index$1) => (index.h("div", { class: "tk-gantt-chart-timeline-header-cell", style: { width: `${cell.span * dayWidth}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline-primary-cell', index$1.toString()) }, cell.label)))), index.h("div", { class: "tk-gantt-chart-timeline-header tk-gantt-chart-timeline-header-secondary", style: { height: `${DEFAULT_HEADER_HEIGHT}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline-secondary') }, secondaryCells.map((cell, index$2) => {
            const cellStart = cell.startDate;
            const isWeekendCell = cell.span === 1 && isWeekend(cellStart);
            const isHolidayCell = cell.span === 1 && isHoliday(cellStart, this.holidays);
            return (index.h("div", { class: index$1.classNames('tk-gantt-chart-timeline-header-cell', {
                    'tk-gantt-chart-timeline-header-cell-weekend': isWeekendCell && this.highlightWeekends,
                    'tk-gantt-chart-timeline-header-cell-holiday': isHolidayCell,
                }), style: { width: `${cell.span * dayWidth}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline-secondary-cell', index$2.toString()) }, cell.label));
        }))));
    }
    renderTimelineBody(flatRows, start, totalWidth, totalDays) {
        const dayWidth = VIEW_TYPE_DAY_WIDTH[this.computedViewType];
        const bodyHeight = flatRows.length * this.rowHeight;
        return (index.h("div", { class: "tk-gantt-chart-timeline-body", style: { width: `${totalWidth}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline-body') }, (this.highlightWeekends || this.holidays.length > 0) &&
            Array.from({ length: totalDays }).map((_, i) => {
                const d = addDays(start, i);
                const weekend = this.highlightWeekends && isWeekend(d);
                const holiday = isHoliday(d, this.holidays);
                if (!weekend && !holiday)
                    return null;
                return (index.h("div", { class: index$1.classNames('tk-gantt-chart-column-highlight', {
                        'tk-gantt-chart-weekend': weekend,
                        'tk-gantt-chart-holiday': holiday,
                    }), style: {
                        left: `${i * dayWidth}px`,
                        width: `${dayWidth}px`,
                        height: `${bodyHeight}px`,
                    }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'column-highlight', i.toString()) }));
            }), this.showTodayIndicator &&
            this.createIndicatorLine({ date: new Date().toISOString(), label: 'Today', color: 'var(--primary-base)', lineStyle: 'solid' }, start, bodyHeight, 'today'), this.indicators.map((ind, index) => this.createIndicatorLine(ind, start, bodyHeight, `custom-${index}`)), flatRows.map(({ task }) => {
            const hasSegments = task.segments && task.segments.length > 0;
            return (index.h("div", { class: "tk-gantt-chart-timeline-row", style: { height: `${this.rowHeight}px` }, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline-row', String(task.id)) }, hasSegments
                ? task.segments.map((seg, idx) => {
                    const segPos = getBarPositionFromDates(seg.startDate, seg.endDate, start, this.computedViewType);
                    return this.createTaskBar(task, segPos, seg, idx);
                })
                : this.createTaskBar(task, getTaskBarPosition(task, start, this.computedViewType))));
        })));
    }
    renderEmptyState() {
        return (index.h("div", { class: "tk-gantt-chart-empty", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'empty') }, index.h("slot", { name: "empty" }, index.h("span", { "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'empty-text') }, "No tasks to display"))));
    }
    // ---------------------------------------------------------------------------
    // render
    // ---------------------------------------------------------------------------
    render() {
        if (!this.tasks.length) {
            return (index.h("div", { class: "tk-gantt-chart", style: this.containerStyle, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'container') }, this.renderEmptyState()));
        }
        const flatRows = flattenTasks(this.tasks, this.expandedIds);
        const { start, end, totalDays } = getTimelineRange(this.tasks, this.computedViewType, this.weekStartDay);
        const dayWidth = VIEW_TYPE_DAY_WIDTH[this.computedViewType];
        const totalWidth = totalDays * dayWidth;
        return (index.h("div", { class: "tk-gantt-chart", style: this.containerStyle, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'container') }, this.renderPanel(flatRows), index.h("div", { class: "tk-gantt-chart-timeline", ref: el => (this.timelineRef = el), onScroll: (e) => this.handleTimelineScroll(e), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'timeline') }, this.renderTimelineHeader(start, end, totalWidth), this.renderTimelineBody(flatRows, start, totalWidth, totalDays))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "tasks": [{
                "tasksChanged": 0
            }],
        "viewType": [{
                "viewTypeChanged": 0
            }]
    }; }
};
TkGanttChart.style = tkGanttChartCss();

exports.tk_gantt_chart = TkGanttChart;
