// @takeoff-ui/core@0.11.9/dist/cjs/tk-radio-group.cjs.entry.js
'use strict';

var index = require('./index-CUpr0td9.js');
var index$1 = require('./index-DmcCMEYj.js');
var hintUtils = require('./hint-utils-D59BxVBS.js');
var testIdUtils = require('./test-id-utils-CAIe6ggJ.js');
var isEqual = require('./isEqual-BbHXdG2o.js');
require('./icon-utils-CoP4H8B1.js');
require('./_Uint8Array-C9PPQ5_G.js');

const tkRadioGroupCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}tk-radio-group{display:block}.tk-radio-group-container{display:flex;align-items:flex-start;flex-direction:column;gap:4px;font-family:var(--family-body)}.tk-radio-group-container label{font-size:var(--size-sm);font-weight:400;color:var(--text-darkest);line-height:var(--desktop-body-s-line-height);padding:var(--input-external-label-base-label-v-padding) var(--input-external-label-base-label-h-padding)}.tk-radio-group-container label .asterisk{color:var(--states-danger-base);font-size:var(--desktop-label-s-size);font-weight:300;line-height:var(--desktop-label-s-line-height)}.tk-radio-group-container .tk-radio-holder{display:flex;flex-direction:row;gap:8px}.tk-radio-group-container .tk-radio-holder.spread{width:100%}.tk-radio-group-container.vertical .tk-radio-holder{flex-direction:column}.tk-radio-group-container[aria-invalid] .tk-radio-holder.card tk-radio{border-color:var(--states-danger-base)}`;

const TkRadioGroup = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.tkChange = index.createEvent(this, "tk-change", 7);
        if (hostRef.$hostElement$["s-ei"]) {
            this.internals = hostRef.$hostElement$["s-ei"];
        }
        else {
            this.internals = hostRef.$hostElement$.attachInternals();
            hostRef.$hostElement$["s-ei"] = this.internals;
        }
        /**
         * The direction of the radio buttons.
         */
        this.direction = 'horizontal';
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Determines the position of the radio group and label.
         */
        this.position = 'left';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * Determines whether the radios will spread evenly across the horizontal space.
         */
        this.spread = false;
        /**
         * Determines the appearance types of radios.
         */
        this.type = 'default';
    }
    invalidChanged() {
        var _a;
        if (((_a = this.slottedItems) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.slottedItems.forEach(item => {
                item.invalid = this.invalid;
            });
        }
    }
    /**
     * Watches for changes in the selected value and emits a custom event when the value changes.
     */
    valueChanged() {
        this.updateTkRadio();
    }
    formResetCallback() {
        this.handleFormReset();
    }
    updateTkRadio() {
        if (this.slottedItems.length > 0) {
            this.slottedItems.forEach(item => {
                item.checked = isEqual.isEqual(this.value, item.value);
            });
        }
    }
    handleFormReset() {
        this.value = null;
        this.tkChange.emit(this.value);
        this.updateTkRadio();
    }
    handleChange(e) {
        this.value = e.detail;
        this.tkChange.emit(this.value);
        this.updateTkRadio();
    }
    handleSlotChange() {
        this.slottedItems = this.el.querySelectorAll('tk-radio');
        if (this.slottedItems.length > 0) {
            this.slottedItems.forEach(item => {
                item.addEventListener('tk-change', e => {
                    e.stopPropagation();
                    this.handleChange(e);
                });
                item.checked = isEqual.isEqual(this.value, item.value);
                item.invalid = this.invalid;
                if (this.spread)
                    item.style.flex = '1';
                item.setAttribute('data-type', this.type);
                // radio componentlerinde name yoksa ve radi group componentinde name var ise
                // group'a verilen name radio componentlerine setlenmesi için yazılmıştır.
                if (!item.name && this.name)
                    item.name = this.name;
            });
        }
    }
    render() {
        var _a;
        let _label;
        const rootClasses = index$1.classNames('tk-radio-group-container', {
            vertical: this.direction === 'vertical',
        });
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            const asterisk = (index.h("span", { key: '43ac6ce9cd0f2cc9f50dc619f011a1ca0ff27d67', class: "asterisk", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'label-asterisk') }, "*"));
            _label = (index.h("label", { key: 'bcbff37d9ee0a14424bae6b954d3f19f27b42ed6', class: "label", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'label') }, this.label, this.showAsterisk && asterisk));
        }
        return (index.h("div", { key: 'bed985255a6a99c6e9490f6b880778489240761d', class: rootClasses, "aria-invalid": this.invalid, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'container') }, _label, index.h("div", { key: 'daba64884f66d2217297a06f517f9eca2cef4ada', class: index$1.classNames('tk-radio-holder', this.type, { spread: this.spread }), "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'holder') }, index.h("slot", { key: '8f21682f7b8c2be6566bd5b1a612dc4feb4e3ec3', onSlotchange: this.handleSlotChange.bind(this) })), hintUtils.renderHint(this.hint, this.error, this.invalid, this.dataTestid)));
    }
    static get formAssociated() { return true; }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "invalid": [{
                "invalidChanged": 0
            }],
        "value": [{
                "valueChanged": 0
            }]
    }; }
};
TkRadioGroup.style = tkRadioGroupCss();

exports.tk_radio_group = TkRadioGroup;
