// @takeoff-ui/core@0.11.9/dist/cjs/tk-step.cjs.entry.js
'use strict';

var index = require('./index-CUpr0td9.js');

const tkStepCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}`;

const TkStep = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Indicates if the step has been completed.
         */
        this.complete = false;
        /**
         * Indicates if the step has encountered an error.
         */
        this.error = false;
        /**
         * Indicates if the step is currently active.
         */
        this.isActive = false;
        /**
         * Whether the step is clickable.
         */
        this.isClickable = true;
        /**
         * Controls the label position of the step.
         * @defaultValue 'non-flip'
         */
        this.labelPosition = 'non-flip';
        /**
         * Controls the step mode of the stepper component.
         * @defaultValue 'basic'
         */
        this.stepMode = 'basic';
        /**
         * Indicates if the step is disabled (cannot be selected or interacted with).
         */
        this.disabled = false;
    }
    render() {
        return (index.h(index.Host, { key: '45beb75797f35160514c95035cdade44b36d2099' }, index.h("slot", { key: '689ac301bf415c2c13ff23e27a867475b5177e7c' })));
    }
    get el() { return index.getElement(this); }
};
TkStep.style = tkStepCss();

exports.tk_step = TkStep;
