// @takeoff-ui/core@0.11.9/dist/cjs/tk-textarea.cjs.entry.js
'use strict';

var index = require('./index-CUpr0td9.js');
var index$1 = require('./index-DmcCMEYj.js');
var hintUtils = require('./hint-utils-D59BxVBS.js');
var testIdUtils = require('./test-id-utils-CAIe6ggJ.js');
var v4 = require('./v4-_2BfiRUa.js');
require('./icon-utils-CoP4H8B1.js');

const tkTextareaCss = () => `@font-face{font-family:"Material Symbols Outlined";src:url("./assets/fonts/material-symbols/Material-Symbols-Outlined.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-outlined{font-family:"Material Symbols Outlined";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-outlined{font-variation-settings:"FILL" 0}.material-symbols-outlined.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Rounded";src:url("./assets/fonts/material-symbols/Material-Symbols-Rounded.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-rounded{font-family:"Material Symbols Rounded";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-rounded{font-variation-settings:"FILL" 0}.material-symbols-rounded.fill{font-variation-settings:"FILL" 1}@font-face{font-family:"Material Symbols Sharp";src:url("./assets/fonts/material-symbols/Material-Symbols-Sharp.woff2") format("woff2");font-weight:400;font-style:normal;font-display:swap}.material-symbols-sharp{font-family:"Material Symbols Sharp";font-weight:normal;font-style:normal;font-size:24px;line-height:1;letter-spacing:normal;text-transform:none;display:inline-block;white-space:nowrap;word-wrap:normal;direction:ltr;-webkit-font-feature-settings:"liga";-webkit-font-smoothing:antialiased}.material-symbols-sharp{font-variation-settings:"FILL" 0}.material-symbols-sharp.fill{font-variation-settings:"FILL" 1}:host{display:block}.tk-textarea-container{display:flex;align-items:flex-start;flex-direction:column;gap:var(--input-external-label-large-container-gap);width:100%}.tk-textarea-container .label{color:var(--text-darkest);font-family:var(--family-body);font-weight:400}.tk-textarea-container .label .asterisk{color:var(--states-danger-base);font-size:var(--desktop-label-s-size);font-weight:300;line-height:var(--desktop-label-s-line-height)}.tk-textarea-container .tk-textarea{display:flex;gap:var(--spacing-xs);position:relative;align-items:center;border:1px solid var(--border-light);border-radius:var(--spacing-m-base);background-color:var(--static-light);width:-webkit-fill-available}.tk-textarea-container .tk-textarea textarea{outline:none;border:none;background-color:transparent;color:var(--text-darkest);padding:0px;width:100%;line-height:var(--desktop-body-m-base-line-height);resize:vertical}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar{width:4px;height:4px}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-track{background:var(--background-lightest);border-radius:2px}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-thumb{border-radius:3px;background:var(--secondary-sub-base)}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-thumb:hover{background:var(--secondary-dark)}.tk-textarea-container .tk-textarea textarea::-webkit-scrollbar-corner{background:var(--background-lightest)}.tk-textarea-container .tk-textarea textarea::placeholder{color:var(--text-sub-base)}.tk-textarea-container .tk-textarea:hover:not(.tk-textarea-container[aria-disabled] .tk-textarea){background-color:var(--background-lightest)}.tk-textarea-container .tk-textarea .counter{position:absolute;right:28px;bottom:6px;color:var(--text-base);line-height:var(--desktop-label-s-line-height);font-size:var(--desktop-label-s-size);font-weight:400}.tk-textarea-container .tk-textarea .counter.maxed{color:var(--states-danger-base)}.tk-textarea-container.large .label{font-size:var(--desktop-label-l-size);line-height:var(--desktop-label-l-line-height);padding:var(--input-external-label-large-label-v-padding) var(--input-external-label-large-label-h-padding)}.tk-textarea-container.large .tk-textarea{padding:var(--input-external-label-large-input-v-padding) var(--input-external-label-large-input-h-padding)}.tk-textarea-container.large .tk-textarea textarea{font-family:var(--desktop-body-m-base-font);font-size:var(--desktop-body-m-base-size);line-height:var(--desktop-body-m-base-line-height)}.tk-textarea-container.base .label{font-size:var(--desktop-label-m-base-size);line-height:var(--desktop-label-m-base-line-height);padding:var(--input-external-label-base-label-v-padding) var(--input-external-label-base-label-h-padding)}.tk-textarea-container.base .tk-textarea{padding:var(--input-external-label-base-input-v-padding) var(--input-external-label-base-input-h-padding)}.tk-textarea-container.base .tk-textarea textarea{font-family:var(--desktop-body-s-font);font-size:var(--desktop-body-s-size);line-height:var(--desktop-body-s-line-height)}.tk-textarea-container.small .label{font-size:var(--desktop-label-s-size);line-height:var(--desktop-label-s-line-height);padding:var(--input-external-label-small-label-v-padding) var(--input-external-label-small-label-h-padding)}.tk-textarea-container.small .tk-textarea{padding:var(--input-external-label-small-input-v-padding) var(--input-external-label-small-input-h-padding)}.tk-textarea-container.small .tk-textarea textarea{font-family:var(--desktop-body-xs-font);font-size:var(--desktop-body-xs-size);line-height:var(--desktop-body-xs-line-height)}.tk-textarea-container.focus .tk-textarea{border:var(--spacing-px) solid var(--states-info-base)}.tk-textarea-container[aria-disabled]{pointer-events:none}.tk-textarea-container[aria-disabled] .label,.tk-textarea-container[aria-disabled] .tk-textarea textarea{color:var(--text-sub-base);pointer-events:none}.tk-textarea-container[aria-disabled] .tk-textarea{background-color:var(--background-lightest)}.tk-textarea-container[aria-readonly] .label{color:var(--text-sub-base)}.tk-textarea-container[aria-readonly] .tk-textarea{background-color:var(--background-lightest)}.tk-textarea-container[aria-invalid] .tk-textarea{border-color:var(--states-danger-base);background-color:var(--background-lightest)}`;

const TkTextarea = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.tkInput = index.createEvent(this, "tk-input", 7);
        this.tkChange = index.createEvent(this, "tk-change", 7);
        this.tkBlur = index.createEvent(this, "tk-blur", 7);
        this.tkFocus = index.createEvent(this, "tk-focus", 7);
        if (hostRef.$hostElement$["s-ei"]) {
            this.internals = hostRef.$hostElement$["s-ei"];
        }
        else {
            this.internals = hostRef.$hostElement$.attachInternals();
            hostRef.$hostElement$["s-ei"] = this.internals;
        }
        this.uniqueId = v4.v4();
        this.hasFocus = false;
        this.passwordStrength = 0;
        this.charCount = 0;
        /**
         * If `true`, the user cannot interact with the input.
         */
        this.disabled = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Represents the rows value of the component
         * @defaultValue 3
         */
        this.rows = 3;
        /**
         * If `true`, the user cannot modify the value.
         */
        this.readonly = false;
        /**
         * Sets size for the component.
         */
        this.size = 'base';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * The value of the input.
         */
        this.value = '';
        this.updateCharCount = () => {
            if (this.maxLength) {
                this.charCount = this.value.toString().trim().length;
            }
        };
        this.handleInput = (ev) => {
            const input = ev.target;
            this.value = (input === null || input === void 0 ? void 0 : input.value) || '';
            this.tkInput.emit(ev);
        };
        this.handleBlur = () => {
            this.hasFocus = false;
            this.tkBlur.emit();
        };
        this.handleFocus = () => {
            this.hasFocus = true;
            this.tkFocus.emit();
        };
    }
    /**
     * Update the native input element when the value changes
     */
    valueChanged() {
        this.tkChange.emit(this.value);
        this.updateCharCount();
    }
    componentWillLoad() {
        // If the tk-input has a tabindex attribute we get the value
        // and pass it down to the native input, then remove it from the
        // tk-input to avoid causing tabbing twice on the same element
        if (this.el.hasAttribute('tabindex')) {
            const tabindex = this.el.getAttribute('tabindex');
            this.tabindex = tabindex !== null ? tabindex : undefined;
            this.el.removeAttribute('tabindex');
        }
    }
    componentDidLoad() {
        this.nativeInput = this.el.shadowRoot.querySelector('textarea');
    }
    formResetCallback() {
        this.handleFormReset();
    }
    /**
     * Sets focus on the specified `tk-input`. Use this method instead of the global
     * `input.focus()`.
     */
    async setFocus() {
        var _a;
        (_a = this.nativeInput) === null || _a === void 0 ? void 0 : _a.focus();
    }
    handleFormReset() {
        this.value = null;
        this.nativeInput.value = null;
        this.tkChange.emit(this.value);
    }
    renderLabel() {
        var _a;
        let label;
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            const asterisk = (index.h("span", { class: "asterisk", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'label-asterisk') }, "*"));
            label = (index.h("label", { htmlFor: this.uniqueId, class: "label", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'label') }, this.label, this.showAsterisk ? asterisk : ''));
        }
        return label;
    }
    render() {
        const rootClasses = index$1.classNames('tk-textarea-container', this.size, { focus: this.hasFocus });
        let counter;
        const counterClasses = index$1.classNames('counter', this.charCount == this.maxLength && 'maxed');
        if (this.maxLength) {
            counter = (index.h("span", { key: '16acb14e4ef37bae5e34d28408cc6a697cc48a88', class: counterClasses, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'counter') }, this.charCount, "/", this.maxLength));
        }
        return (index.h("div", { key: 'f5c34cc6cb7b59b835b54b8b79830b21471ef6b5', "aria-readonly": this.readonly, "aria-disabled": this.disabled, "aria-invalid": this.invalid, class: rootClasses, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'container') }, this.renderLabel(), index.h("div", { key: 'be29df5ed44ca9f4bc9b52903ddc267833557a24', class: "tk-textarea", "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'control') }, index.h("textarea", { key: '5d3fc8d5d32271eb726d0cb103b99951b3f0053c', id: this.uniqueId, ref: el => (this.nativeInput = el), disabled: this.disabled, autoComplete: "off", name: this.name, rows: this.rows, maxLength: this.maxLength, placeholder: this.placeholder || '', readOnly: this.readonly, tabindex: this.tabindex, onInput: this.handleInput, onBlur: this.handleBlur, onFocus: this.handleFocus, value: this.value, "data-testid": testIdUtils.getDataTestId(this.dataTestid, 'native-textarea') }), counter), hintUtils.renderHint(this.hint, this.error, this.invalid, this.dataTestid)));
    }
    static get formAssociated() { return true; }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "value": [{
                "valueChanged": 0
            }]
    }; }
};
TkTextarea.style = tkTextareaCss();

exports.tk_textarea = TkTextarea;
