// @takeoff-ui/core@0.11.9/dist/collection/components/tk-radio/tk-radio.js
import { h, Host } from "@stencil/core";
import classNames from "classnames";
import { v4 as uuidv4 } from "uuid";
/**
 * The TkRadio component is another basic element for user input. You can use this to supply a way for the user to pick an option from multiple choices.
 * @slot content - Custom content template.
 * @react `import { TkRadio } from '@takeoff-ui/react'`
 * @vue `import { TkRadio } from '@takeoff-ui/vue'`
 * @angular `import { TkRadio } from '@takeoff-ui/angular'`
 */
export class TkRadio {
    constructor() {
        this.uniqueId = uuidv4();
        /**
         * Controls if radio has custom content.
         * @defaultValue false
         */
        this.hasContentSlot = false;
        /**
         * Disables the radio button if true.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Marks the radio button as checked or unchecked.
         * @defaultValue false
         */
        this.checked = false;
        // radio group ile birlikte kullanımlarda diğer radio ya tıklandığı durumda kendini false yapmaya ihtiyaç olmadığı için bu işlemi radio group tarafında yapıldığı için outside click eventi atanmasına gerek yoktur.
        if (!this.el.closest('tk-radio-group')) {
            this.windowClickHandler = this.handleWindowClick.bind(this);
        }
    }
    componentWillLoad() {
        this.hasContentSlot = !!this.el.querySelector('[slot="content"]');
        this.parentEl = this.el.closest('tk-radio-group');
        if (this.parentEl && !this.position) {
            this.position = this.parentEl.position;
        }
    }
    componentDidRender() {
        this.bindWindowClickListener();
    }
    disconnectedCallback() {
        this.unbindWindowClickListener();
    }
    bindWindowClickListener() {
        window.addEventListener('click', this.windowClickHandler);
    }
    unbindWindowClickListener() {
        window.removeEventListener('click', this.windowClickHandler);
    }
    handleWindowClick(event) {
        var _a, _b;
        let clickedElement = event.target;
        if (clickedElement.tagName !== 'TK-RADIO') {
            clickedElement = clickedElement.closest('tk-radio');
        }
        if (((_a = clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('name')) === null || _a === void 0 ? void 0 : _a.length) > 0 &&
            (clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('name')) == this.name &&
            ((_b = clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('data-tk-radio-id')) === null || _b === void 0 ? void 0 : _b.length) > 0 &&
            (clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('data-tk-radio-id')) !== this.uniqueId &&
            !(clickedElement === null || clickedElement === void 0 ? void 0 : clickedElement.getAttribute('disabled'))) {
            this.checked = false;
        }
    }
    handleInputChange() {
        if (!this.disabled) {
            this.checked = true;
            this.tkChange.emit(this.value);
        }
    }
    render() {
        const labelClass = classNames('tk-radio-container', this.position, {
            'disabled': this.disabled,
            'width-description': !!this.description,
        });
        return (h(Host, { key: '92262fceb575a73daaccdaec498a70dd9073ede3', "data-tk-radio-id": this.uniqueId, invalid: this.invalid, disabled: this.disabled }, h("label", { key: 'e2f683494e14e42f333b449037471d6f0f24e865', htmlFor: this.uniqueId, class: labelClass, "aria-disabled": this.disabled, "aria-invalid": this.invalid }, h("input", { key: '7a15a9f05c0ee1c08dc95803404caa1ee6008d59', id: this.uniqueId, type: "radio", name: this.name, value: this.value, checked: this.checked, disabled: this.disabled, onChange: () => this.handleInputChange() }), h("div", { key: 'bc1d16c1d7007e67848a67ed652931ca4b218db7', class: "mask" }, h("div", { key: '282040082c5e42618c7bf7a28f7d43d088ef8dbf' })), this.hasContentSlot ? (h("slot", { name: "content" })) : (h("div", { class: "tk-radio-text-holder" }, this.label && h("div", { class: "tk-radio-label" }, this.label), this.description && h("div", { class: "tk-radio-description" }, this.description))))));
    }
    static get is() { return "tk-radio"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-radio.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-radio.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Disables the radio button if true."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "description": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The description sub text displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "description"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the label for the element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "position": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Determines the position of the radio and label."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "position"
            },
            "checked": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Marks the radio button as checked or unchecked."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "checked",
                "defaultValue": "false"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The name of the radio group, used to group radio buttons together."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "name"
            },
            "value": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The value of the radio button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            }
        };
    }
    static get states() {
        return {
            "hasContentSlot": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the radio button's checked state changes."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get attachInternalsMemberName() { return "internals"; }
}
