// @takeoff-ui/core@0.11.9/dist/collection/components/tk-accordion/tk-accordion.js
import { h } from "@stencil/core";
import { isEqual } from "lodash-es";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The TkAccordion component is a user interface element that organizes content under headers, allowing users to expand and collapse sections by clicking on each header. It is particularly useful for improving layout and readability on pages with extensive information.
 * @slot default - Default slot to detect TkAccordionItem components.
 * @react `import { TkAccordion, TkAccordionItem } from '@takeoff-ui/react'`
 * @vue `import { TkAccordion, TkAccordionItem } from '@takeoff-ui/vue'`
 * @angular `import { TkAccordion, TkAccordionItem } from '@takeoff-ui/angular'`
 */
export class TkAccordion {
    constructor() {
        this.internalActiveIndex = [];
        /**
         * Allows multiple accordion items to be expanded simultaneously.
         * @defaultValue false
         */
        this.allowMultiple = false;
        /**
         * Sets the position of opening and closing chevrons.
         * @defaultValue 'right'
         */
        this.arrowPosition = 'right';
        /**
         * Sets the expand icon
         * @defaultValue 'keyboard_arrow_down'
         */
        this.expandIcon = 'keyboard_arrow_down';
        /**
         * Sets the collapse icon
         * @defaultValue 'keyboard_arrow_up'
         */
        this.collapseIcon = 'keyboard_arrow_up';
        /**
         * Whether to hide the arrow icons.
         * @defaultValue false
         */
        this.hideArrows = false;
        /**
         * Sets accordion style for the component.
         * @defaultValue 'grouped'
         */
        this.type = 'grouped';
        /**
         * Controls the display mode of the accordion component.
         * @defaultValue 'default'
         */
        this.mode = 'default';
    }
    internalActiveIndexChanged(newValue, oldValue) {
        // Exit early if the active index hasn't changed
        if (isEqual(newValue, oldValue))
            return;
        this.updateActiveIndex();
        this.getAccordionItems().forEach((item, index) => {
            const itemKey = this.getItemKey(item, index);
            const isActive = newValue.some(activeIndex => activeIndex === itemKey);
            item.active = isActive;
        });
    }
    activeIndexChanged() {
        const normalized = this.normalizeActiveIndex();
        if (isEqual(normalized, this.internalActiveIndex))
            return;
        this.internalActiveIndex = normalized;
    }
    componentDidLoad() {
        this.validateItemKeylessActiveIndex();
        this.validateActiveProps();
        this.validateControlType();
        this.validateKeyType();
        this.initInternalActiveIndex();
        this.initEventListeners();
    }
    initInternalActiveIndex() {
        // if prop activeIndex is set, use it
        if (this.hasActiveIndex())
            return (this.internalActiveIndex = this.normalizeActiveIndex());
        // else if accordion items have active prop, use them
        this.getAccordionItems().forEach((item, index) => {
            if (item.active)
                this.internalActiveIndex = [...this.internalActiveIndex, this.getItemKey(item, index)];
        });
    }
    initEventListeners() {
        this.getAccordionItems().forEach((item, index) => {
            const itemKey = this.getItemKey(item, index);
            // listen to active change
            item.addEventListener('tk-active-change', (e) => {
                this.handleItemActiveChange(itemKey, e.detail);
                item.active = e.detail;
                if (!this.allowMultiple && e.detail) {
                    this.getAccordionItems()
                        .filter(child => child !== item)
                        .forEach(otherItem => {
                        otherItem.active = false;
                    });
                }
            });
        });
    }
    // Validation Logic
    validateActiveProps() {
        const activeCount = this.getAccordionItems().reduce((count, item) => (item.active ? count + 1 : count), 0);
        if (activeCount > 1 && !this.allowMultiple) {
            console.error('TkAccordion: Multiple accordion items are set to active while allowMultiple is false. This may lead to unexpected behavior.');
        }
    }
    validateControlType() {
        const hasActiveAccordionItems = this.getAccordionItems().some(item => item.active !== undefined);
        if (this.hasActiveIndex() && hasActiveAccordionItems)
            console.error('TkAccordion: Accordion cannot have both activeIndex and active accordion items');
    }
    validateKeyType() {
        var _a, _b;
        const type = typeof ((_b = (_a = this.getAccordionItems().find(item => item.itemKey !== undefined)) === null || _a === void 0 ? void 0 : _a.itemKey) !== null && _b !== void 0 ? _b : 0);
        const allItemsHaveSameType = this.getAccordionItems().every(item => item.itemKey === undefined || typeof item.itemKey === type);
        let activeIndexHasSameType = true;
        if (this.hasActiveIndex())
            activeIndexHasSameType = Array.isArray(this.activeIndex) ? this.activeIndex.every(item => typeof item === type) : typeof this.activeIndex === type;
        if (!allItemsHaveSameType || !activeIndexHasSameType)
            console.error('TkAccordion: Accordion item keys must be of the same type');
    }
    validateItemKeylessActiveIndex() {
        if (!this.hasActiveIndex())
            return;
        const hasKeylessItems = this.getAccordionItems().every(item => item.itemKey === undefined);
        const activeIndexType = Array.isArray(this.activeIndex) ? this.activeIndex.map(item => typeof item) : [typeof this.activeIndex];
        if (hasKeylessItems && activeIndexType.some(type => type !== 'number')) {
            console.error('TkAccordion: When using keyless accordion items, activeIndex must be of type number or number array.');
        }
    }
    hasActiveIndex() {
        return this.activeIndex || this.activeIndex === 0;
    }
    normalizeActiveIndex() {
        if (!this.activeIndex && this.activeIndex !== 0)
            return [];
        if (!Array.isArray(this.activeIndex))
            return [this.activeIndex];
        if (this.allowMultiple)
            return this.activeIndex;
        const lastItem = this.activeIndex.at(-1);
        return lastItem ? [lastItem] : [];
    }
    getActiveIndex() {
        return this.allowMultiple ? this.internalActiveIndex : this.internalActiveIndex[this.internalActiveIndex.length - 1];
    }
    updateActiveIndex() {
        const activeIndex = this.getActiveIndex();
        if (!isEqual(activeIndex, this.activeIndex))
            this.tkActiveIndexChange.emit(activeIndex);
    }
    handleItemActiveChange(itemKey, active) {
        if (active && !this.internalActiveIndex.includes(itemKey)) {
            this.internalActiveIndex = [...this.internalActiveIndex, itemKey];
        }
        if (!active && this.internalActiveIndex.includes(itemKey)) {
            this.internalActiveIndex = this.internalActiveIndex.filter(activeIndex => activeIndex !== itemKey);
        }
        this.tkAccordionItemSelected.emit({
            index: itemKey,
            active,
        });
    }
    getAccordionItems() {
        return Array.from(this.el.querySelectorAll('tk-accordion-item')).filter(child => child.parentElement === this.el);
    }
    getItemKey(accordionItem, index) {
        var _a;
        return (_a = accordionItem.itemKey) !== null && _a !== void 0 ? _a : index;
    }
    render() {
        return (h("div", { key: '8a88bf333637906c2bd4218f3015901c5cdbd7aa', class: "tk-accordion", "data-testid": getDataTestId(this.dataTestid, 'container') }, h("slot", { key: '6bbb5335b02b07138985ed6b883a03a5cd1c4254' })));
    }
    static get is() { return "tk-accordion"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-accordion.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-accordion.css"]
        };
    }
    static get properties() {
        return {
            "activeIndex": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "string | number | (string | number)[]",
                    "resolved": "(string | number)[] | number | string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Currently active panel indexes. Can be a single value or an array.\nWhen allowMultiple is false, only the last value in the array will be used.\nHas priority over AccordionItem's active prop. To prevent conflicts, avoid using both simultaneously."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "active-index"
            },
            "allowMultiple": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Allows multiple accordion items to be expanded simultaneously."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "allow-multiple",
                "defaultValue": "false"
            },
            "arrowPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'right'"
                        }],
                    "text": "Sets the position of opening and closing chevrons."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "arrow-position",
                "defaultValue": "'right'"
            },
            "expandIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'keyboard_arrow_down'"
                        }],
                    "text": "Sets the expand icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "expand-icon",
                "defaultValue": "'keyboard_arrow_down'"
            },
            "collapseIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'keyboard_arrow_up'"
                        }],
                    "text": "Sets the collapse icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "collapse-icon",
                "defaultValue": "'keyboard_arrow_up'"
            },
            "hideArrows": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether to hide the arrow icons."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-arrows",
                "defaultValue": "false"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'grouped' | 'divided'",
                    "resolved": "\"divided\" | \"grouped\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'grouped'"
                        }],
                    "text": "Sets accordion style for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'grouped'"
            },
            "mode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'default' | 'compact'",
                    "resolved": "\"compact\" | \"default\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'default'"
                        }],
                    "text": "Controls the display mode of the accordion component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mode",
                "defaultValue": "'default'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "internalActiveIndex": {}
        };
    }
    static get events() {
        return [{
                "method": "tkActiveIndexChange",
                "name": "tk-active-index-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when an active index is changed"
                },
                "complexType": {
                    "original": "string | number | (string | number)[]",
                    "resolved": "(string | number)[] | number | string",
                    "references": {}
                }
            }, {
                "method": "tkAccordionItemSelected",
                "name": "tk-accordion-item-selected",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "deprecated",
                            "text": undefined
                        }],
                    "text": "Emitted when an accordion item is selected"
                },
                "complexType": {
                    "original": "IAccordionItemSelect",
                    "resolved": "IAccordionItemSelect",
                    "references": {
                        "IAccordionItemSelect": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-accordion/types.ts::IAccordionItemSelect",
                            "referenceLocation": "IAccordionItemSelect"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "internalActiveIndex",
                "methodName": "internalActiveIndexChanged"
            }, {
                "propName": "activeIndex",
                "methodName": "activeIndexChanged"
            }];
    }
}
