// @takeoff-ui/core@0.11.9/dist/collection/components/tk-accordion/tk-accordion-item.js
import { h, Host } from "@stencil/core";
import classNames from "classnames";
import { getIconElementProps, renderIcons } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * @slot header - Custom header template that overrides the header prop if provided.
 * @slot content - Custom content template.
 */
export class TkAccordionItem {
    constructor() {
        this.type = 'grouped';
        this.arrowPosition = 'right';
        this.hideArrows = false;
        this.hasHeaderSlot = false;
        this.mode = 'default';
        /**
         * Sets size for the component.
         * @defaultValue 'base'
         */
        this.size = 'base';
    }
    activeChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            this.tkActiveChange.emit(newValue);
        }
    }
    componentWillLoad() {
        this.parentEl = this.el.closest('tk-accordion');
        if (this.parentEl) {
            this.type = this.parentEl.type;
            this.arrowPosition = this.parentEl.arrowPosition;
            this.expandIcon = this.parentEl.expandIcon;
            this.collapseIcon = this.parentEl.collapseIcon;
            this.hideArrows = this.parentEl.hideArrows;
            this.mode = this.parentEl.mode;
        }
        this.hasHeaderSlot = !!this.el.querySelector(':scope > [slot="header"]');
    }
    renderCollapseIcon() {
        if (this.hideArrows)
            return null;
        let _renderIcon;
        if (this.active) {
            _renderIcon = this.collapseIcon;
        }
        else {
            _renderIcon = this.expandIcon;
        }
        return (h("tk-icon", Object.assign({}, getIconElementProps(_renderIcon, { size: 'large', variant: this.active ? 'primary' : 'neutral' }), { dataTestid: getDataTestId(this.dataTestid, this.active ? 'collapse-icon' : 'expand-icon') })));
    }
    createHeader() {
        if (this.hasHeaderSlot) {
            return h("slot", { name: "header" });
        }
        return this.header || '';
    }
    render() {
        const rootClasses = classNames('tk-accordion-item', this.size, this.type, this.mode, {
            open: this.active,
        });
        const icon = renderIcons(this.icon, {
            sign: true,
            variant: 'neutral',
            additionalProps: { class: 'tk-accordion-item-icon' },
            dataTestid: this.dataTestid,
        });
        return (h(Host, { key: '2fc3c3020a58ce6d98746361c01a7f292c76733c' }, h("div", { key: '4b3210c211ac98bc7f22e15bccc53f80d02a9c19', class: rootClasses, "data-testid": getDataTestId(this.dataTestid, 'item') }, h("div", { key: 'f7a15c080331b7738b27b506432b76e998c06996', class: "header", onClick: () => this.tkActiveChange.emit(!this.active), "data-testid": getDataTestId(this.dataTestid, 'header') }, this.arrowPosition === 'left' && this.renderCollapseIcon(), icon.leftIcon, h("span", { key: 'de3161a01072ebd9c7160580a01c75ea4ef51358', class: "title", "data-testid": getDataTestId(this.dataTestid, 'title') }, this.createHeader()), icon.rightIcon, this.arrowPosition === 'right' && this.renderCollapseIcon()), h("div", { key: '895dcbb16edd696ee3fb5560bc260564f1703c5e', class: `content ${this.active ? 'open' : ''}`, "data-testid": getDataTestId(this.dataTestid, 'content') }, h("slot", { key: '74559f5925eaa3232bd5f1d7b9233c53b3bcc6e3', name: "content" })))));
    }
    static get is() { return "tk-accordion-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-accordion-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-accordion-item.css"]
        };
    }
    static get properties() {
        return {
            "active": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Sets if the accordion is active."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "active"
            },
            "itemKey": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "string | number",
                    "resolved": "number | string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Optional key for the accordion item."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "item-key"
            },
            "header": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Header text to display."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'base' | 'large'",
                    "resolved": "\"base\" | \"large\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon for accordion component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "type": {},
            "arrowPosition": {},
            "expandIcon": {},
            "collapseIcon": {},
            "hideArrows": {},
            "hasHeaderSlot": {},
            "mode": {}
        };
    }
    static get events() {
        return [{
                "method": "tkActiveChange",
                "name": "tk-active-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when an active index is changed"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "active",
                "methodName": "activeChanged"
            }];
    }
}
