// @takeoff-ui/core@0.11.9/dist/collection/components/tk-rating/tk-rating.js
import { h } from "@stencil/core";
import getIcon from "./rating-icons";
import classNames from "classnames";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkRating` component is a customizable rating input element that allows users to select a value from a series of icons (such as stars, hearts, or dot).
 * @react `import { TkRating } from '@takeoff-ui/react'`
 * @vue `import { TkRating } from '@takeoff-ui/vue'`
 * @angular `import { TkRating } from '@takeoff-ui/angular'`
 */
export class TkRating {
    constructor() {
        this.hoverRating = 0; // Hover rating for temporary display
        /**
         * The maximum rating value. Possible options are 5 or 10.
         */
        this.maxRating = 5;
        /**
         * The type of icon to display for each rating element. Options include 'star', 'heart', 'dot' and 'number'.
         */
        this.type = 'star';
        /**
         * The currently selected rating value.
         */
        this.value = 0;
        /**
         * Determines whether to show the numerical rating value under to the icon.
         * @defaultValue false
         */
        this.showRatingValue = false;
        /**
         * The user cannot interact with the input.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * If `true`, the user cannot modify the value.
         * @defaultValue false
         */
        this.readonly = false;
    }
    // Set the rating on click, with half steps
    handleRatingClick(rating) {
        if (this.disabled || this.readonly)
            return;
        this.tkChange.emit(rating);
    }
    // Set hover rating with half steps
    handleMouseMove(rating) {
        if (this.disabled || this.readonly)
            return;
        if (rating !== this.hoverRating) {
            this.hoverRating = rating;
        }
    }
    // Reset hover rating when mouse leaves
    handleMouseLeave() {
        if (this.disabled || this.readonly)
            return;
        this.hoverRating = 0;
    }
    renderIcon(ratingValue) {
        const displayRating = this.hoverRating || this.value;
        let state = 'default';
        if (ratingValue <= displayRating) {
            state = 'fill';
        }
        else if (ratingValue - 0.5 === displayRating) {
            state = 'semi';
        }
        if (this.disabled) {
            return (h("div", { class: classNames('tk-rating', this.type, state, { disabled: this.disabled }), "data-testid": getDataTestId(this.dataTestid, 'item', ratingValue.toString()) }, getIcon(this.type, 'disabled'), this.showRatingValue && `0${ratingValue}`));
        }
        return (h("div", { class: classNames('tk-rating', this.type, state, this.readonly, { disabled: this.disabled }), onMouseMove: () => this.handleMouseMove(ratingValue), onMouseLeave: () => this.handleMouseLeave(), onClick: () => this.handleRatingClick(ratingValue), "data-testid": getDataTestId(this.dataTestid, 'item', ratingValue.toString()) }, getIcon(this.type, state), " ", this.showRatingValue && `0${ratingValue}`));
    }
    renderNumberRating() {
        return [1, 2, 3, 4, 5].map(ratingValue => (h("div", { class: classNames('tk-rating', 'number', `rating-number-${ratingValue}`, {
                readonly: this.readonly,
                disabled: this.disabled,
                selected: ratingValue === this.value,
            }), onMouseMove: () => this.handleMouseMove(ratingValue), onMouseLeave: () => this.handleMouseLeave(), onClick: () => this.handleRatingClick(ratingValue), "data-testid": getDataTestId(this.dataTestid, 'item', ratingValue.toString()) }, ratingValue.toString().padStart(2, '0'))));
    }
    render() {
        const isNumberType = this.type === 'number';
        return (h("div", { key: '8ebe4c78e2663354644331d88c278e4c54ba09ad', class: classNames('tk-rating-container', { number: isNumberType }), "data-testid": getDataTestId(this.dataTestid, 'container') }, isNumberType ? this.renderNumberRating() : Array.from({ length: this.maxRating }, (_, index) => this.renderIcon(index + 1))));
    }
    static get is() { return "tk-rating"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-rating.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-rating.css"]
        };
    }
    static get properties() {
        return {
            "maxRating": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The maximum rating value. Possible options are 5 or 10."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max-rating",
                "defaultValue": "5"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'star' | 'heart' | 'dot' | 'number'",
                    "resolved": "\"dot\" | \"heart\" | \"number\" | \"star\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The type of icon to display for each rating element. Options include 'star', 'heart', 'dot' and 'number'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'star'"
            },
            "value": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The currently selected rating value."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value",
                "defaultValue": "0"
            },
            "showRatingValue": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Determines whether to show the numerical rating value under to the icon."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-rating-value",
                "defaultValue": "false"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "The user cannot interact with the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "readonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If `true`, the user cannot modify the value."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "readonly",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hoverRating": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the value has changed."
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
}
