// @takeoff-ui/core@0.11.9/dist/collection/components/tk-checkbox/tk-checkbox.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { v4 as uuidv4 } from "uuid";
import { getIconElementProps } from "../../utils/icon-utils";
/**
 * The TkCheckbox component is another basic element for user input. You can use this to supply a way for the user to toggle an option.
 * @slot content - Custom content template.
 * @react `import { TkCheckbox } from '@takeoff-ui/react'`
 * @vue `import { TkCheckbox } from '@takeoff-ui/vue'`
 * @angular `import { TkCheckbox } from '@takeoff-ui/angular'`
 */
export class TkCheckbox {
    constructor() {
        this.uniqueId = uuidv4();
        /**
         * Controls if checkbox has custom content.
         * @defaultValue false
         */
        this.hasContentSlot = false;
        /**
         * If true, the user cannot interact with the checkbox.
         */
        this.disabled = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Determines the appearance types of the checkbox.
         */
        this.type = 'default';
        /**
         * Determines the size of the checkbox.
         */
        this.size = 'base';
        /**
         * If true, the checkbox will be indeterminate.
         */
        this.indeterminate = false;
        /**
         * Sets the checkbox value
         */
        this.value = false;
    }
    indeterminateChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            this.inputElement.indeterminate = newValue; // update the checkbox's indeterminate property
            if (newValue)
                this.value = null;
        }
    }
    valueChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            if (newValue !== null) {
                this.indeterminate = false;
            }
        }
    }
    componentWillLoad() {
        if (this.indeterminate)
            this.value = null;
        this.hasContentSlot = !!this.el.querySelector('[slot="content"]');
    }
    formResetCallback() {
        this.handleFormReset();
    }
    handleFormReset() {
        this.value = false;
        this.indeterminate = false;
        this.inputElement.checked = false;
        this.tkChange.emit(this.value);
    }
    handleChange() {
        if (!this.disabled) {
            if (this.inputElement.indeterminate) {
                this.value = null;
                this.indeterminate = false;
            }
            else {
                this.value = this.inputElement.checked;
            }
            this.tkChange.emit(this.value);
        }
    }
    renderInput() {
        return (h("input", { id: this.uniqueId, type: "checkbox", ref: el => (this.inputElement = el), checked: this.value, indeterminate: this.indeterminate, disabled: this.disabled, onChange: this.handleChange.bind(this), name: this.name }));
    }
    render() {
        return (h("div", { key: '36149a307c16ce4734f583e625bcfa544e6f9fec', class: classNames('tk-checkbox-container', this.type, this.size), "aria-disabled": this.disabled, "aria-invalid": this.invalid }, h("label", { key: '141ecdcaf29c7fe34b92c8469dc42ca667e78f0c', htmlFor: this.uniqueId }, this.renderInput(), h("div", { key: 'd80febfa0909b18a66f06de79219d4874eb712be', class: "mask" }, h("tk-icon", Object.assign({ key: '9c00972114617557ac5e0878f3ad4dfcd7d206fd', class: "inner-mask" }, getIconElementProps(this.indeterminate ? 'remove' : this.value ? 'check' : null, {
            color: 'var(--static-light)',
            size: this.size === 'small' ? 'xxsmall' : 'xsmall',
        })))), this.hasContentSlot ? (h("slot", { name: "content" })) : (h("div", { class: "tk-checkbox-text-holder" }, h("div", { class: "tk-checkbox-label" }, this.label), h("div", { class: "tk-checkbox-description" }, this.description))))));
    }
    static get is() { return "tk-checkbox"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-checkbox.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-checkbox.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, the user cannot interact with the checkbox."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "description": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The description sub text displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "description"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the label for the checkbox."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Name of the checkbox"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'default' | 'card'",
                    "resolved": "\"card\" | \"default\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines the appearance types of the checkbox."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'default'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'small' | 'base'",
                    "resolved": "\"base\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines the size of the checkbox."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "indeterminate": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, the checkbox will be indeterminate."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "indeterminate",
                "defaultValue": "false"
            },
            "value": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets the checkbox value"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value",
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "hasContentSlot": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the checkbox checked state changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "indeterminate",
                "methodName": "indeterminateChanged"
            }, {
                "propName": "value",
                "methodName": "valueChanged"
            }];
    }
    static get attachInternalsMemberName() { return "internals"; }
}
