// @takeoff-ui/core@0.11.9/dist/collection/components/tk-tabs/tk-tabs.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { getIconElementProps, renderIcons } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * TkTabs is a navigation component that displays menu items as tab headers.
 * @react `import { TkTabs, TkTabsItem } from '@takeoff-ui/react'`
 * @vue `import { TkTabs, TkTabsItem } from '@takeoff-ui/vue'`
 * @angular `import { TkTabs, TkTabsItem } from '@takeoff-ui/angular'`
 */
export class TkTabs {
    constructor() {
        this.internalActiveIndex = 0;
        this.internalTabItems = [];
        /**
         * Sets the alignment of the header.
         */
        this.alignHeaders = 'start';
        /**
         * Controls if the tabs component is controlled.
         * @defaultValue false
         */
        this.controlled = false;
        /**
         * Default Active Index for tabs component.
         * @defaultValue 0
         */
        this.defaultActiveIndex = 0;
        /**
         * Controls the currently active tab index.
         * Can be controlled programmatically from outside.
         * @defaultValue 0
         */
        this.activeIndex = 0;
        /**
         * Controls if tabs are closable.
         * @defaultValue false
         */
        this.isClosable = false;
        /**
         * Controls if new tabs can be added or not.
         * @defaultValue false
         */
        this.isExtendable = false;
        /**
         * Controls the orientation of the tabs component.
         * @defaultValue 'horizontal'
         */
        this.orientation = 'horizontal';
        /**
         * Determines whether the tab headers will spread evenly across the horizontal space.
         * @defaultValue false
         */
        this.spreadHeaders = false;
        /**
         * Controls the size of the tabs component.
         * @defaultValue 'base'
         */
        this.size = 'base';
        /**
         * Controls the tab style of the tabs component.
         * @defaultValue 'basic'
         */
        this.type = 'basic';
        /**
         * Controls the color variant of the tabs component.
         * @defaultValue 'primary'
         */
        this.variant = 'primary';
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        /**
         * The style attribute of headers container element
         */
        this.headerContainerStyle = null;
        /**
         * The style attribute of tabs item element
         */
        this.contentStyle = null;
        this.getIconVariant = (index) => {
            if (index === this.activeIndex && (this.type === 'divided' || this.type === 'compact')) {
                return this.variant;
            }
            else {
                return 'neutral';
            }
        };
        this.createTabIcons = (tab, index, itemDataTestid) => renderIcons(tab.icon, {
            variant: this.getIconVariant(index),
            size: this.size === 'xxsmall' || this.size === 'xsmall' ? 'base' : 'medium',
            dataTestid: itemDataTestid,
        });
    }
    activeIndexChanged(newValue) {
        if (newValue !== undefined && newValue !== this.internalActiveIndex) {
            if (newValue >= 0 && newValue < this.internalTabItems.length) {
                this.internalActiveIndex = newValue;
                this.activeIndex = newValue;
                this.tkTabChange.emit(this.internalActiveIndex);
            }
            else {
                this.activeIndex = this.internalActiveIndex;
                console.warn('Invalid tab index provided');
            }
        }
    }
    componentWillLoad() {
        var _a;
        this.internalTabItems = this.getDirectTabItems();
        // slot ismini kullanıcının vermesine gerek kalmadan içeride setlenmesi sağlandı
        this.internalTabItems.forEach((tab, index) => {
            tab.setAttribute('slot', `tab-content-${index}`);
        });
        this.internalActiveIndex = (_a = this.activeIndex) !== null && _a !== void 0 ? _a : this.defaultActiveIndex;
        this.el.addEventListener('tk-update', this.handleTabUpdate.bind(this));
    }
    disconnectedCallback() {
        this.el.removeEventListener('tk-update', this.handleTabUpdate.bind(this));
    }
    selectTab(index) {
        if (index >= 0 && index < this.internalTabItems.length && !this.internalTabItems[index].disabled) {
            this.internalActiveIndex = index;
            this.activeIndex = index;
            this.tkTabChange.emit(this.internalActiveIndex);
        }
    }
    closeTab(index) {
        this.internalTabItems.splice(index, 1);
        if (this.internalActiveIndex >= index) {
            const newIndex = Math.max(0, this.internalActiveIndex - 1);
            this.selectTab(newIndex);
        }
        this.internalTabItems = [...this.internalTabItems];
    }
    addTab() {
        const newTabItem = document.createElement('tk-tabs-item');
        const newTabIndex = this.internalTabItems.length;
        newTabItem.setAttribute('label', `Tab label`);
        newTabItem.setAttribute('is-closable', 'true');
        const contentDiv = document.createElement('div');
        newTabItem.slot = `tab-content-${newTabIndex}`;
        contentDiv.innerHTML = `TK TAB CONTENT ${newTabIndex + 1}`;
        newTabItem.appendChild(contentDiv);
        this.el.appendChild(newTabItem);
        this.internalTabItems = [...this.internalTabItems, newTabItem];
        this.internalActiveIndex = newTabIndex;
        this.activeIndex = newTabIndex;
        this.tkTabChange.emit(this.internalActiveIndex);
    }
    /**
     * Returns only direct child tab items, filtering out nested ones.
     * This prevents nested tabs from interfering with each other.
     */
    getDirectTabItems() {
        return Array.from(this.el.querySelectorAll('tk-tabs-item')).filter(item => item.closest('tk-tabs') === this.el);
    }
    /**
     * Handles badge update events from tab items
     */
    handleTabUpdate(event) {
        const tab = event.composedPath().find(el => el instanceof HTMLElement && el.tagName.toLowerCase() === 'tk-tabs-item');
        if (!tab)
            return;
        if (tab.closest('tk-tabs') !== this.el)
            return;
        event.stopPropagation();
        const directItems = this.getDirectTabItems();
        const index = directItems.indexOf(tab);
        if (index !== -1 && this.internalTabItems[index]) {
            this.internalTabItems[index].label = event.detail.label;
            this.internalTabItems[index].icon = event.detail.icon;
            this.internalTabItems[index].disabled = event.detail.disabled;
            this.internalTabItems[index].badged = event.detail.badged;
            this.internalTabItems[index].badgeCount = event.detail.badgeCount;
            this.internalTabItems[index].badgeLabel = event.detail.badgeLabel;
            this.internalTabItems[index].tooltipOptions = event.detail.tooltipOptions;
            this.internalTabItems = [...this.internalTabItems];
        }
    }
    handleTabClick(index) {
        if (!this.controlled) {
            this.selectTab(index);
        }
        else {
            this.tkTabClick.emit(index);
        }
    }
    renderTabBadge(tab, index, itemDataTestid) {
        var _a, _b, _c, _d, _e, _f;
        if (tab.badged) {
            const badgeSize = this.size === 'xsmall' || this.size === 'xxsmall' ? 'small' : this.size;
            return (h("div", { class: "tk-tabs-item-badge-container", "data-testid": getDataTestId(itemDataTestid, 'badge-container') }, h("tk-badge", { label: tab.badgeLabel, count: tab.badgeCount, variant: (_b = (_a = tab.badgeOptions) === null || _a === void 0 ? void 0 : _a.variant) !== null && _b !== void 0 ? _b : (this.internalActiveIndex === index ? this.variant : 'neutral'), type: (_d = (_c = tab.badgeOptions) === null || _c === void 0 ? void 0 : _c.type) !== null && _d !== void 0 ? _d : 'filledlight', rounded: (_f = (_e = tab.badgeOptions) === null || _e === void 0 ? void 0 : _e.rounded) !== null && _f !== void 0 ? _f : true, size: badgeSize, dataTestid: getDataTestId(itemDataTestid, 'badge') })));
        }
        return null;
    }
    renderTabTooltip(tab, itemDataTestid) {
        if (tab === null || tab === void 0 ? void 0 : tab.tooltipOptions) {
            if (tab.tooltipOptions.icon) {
                return (h("tk-tooltip", { header: tab.tooltipOptions.header, description: tab.tooltipOptions.description, position: tab.tooltipOptions.position || 'bottom', variant: tab.tooltipOptions.variant || 'dark', dataTestid: getDataTestId(itemDataTestid, 'tooltip') }, h("tk-icon", Object.assign({ slot: "trigger" }, (typeof tab.tooltipOptions.icon === 'string' ? { icon: tab.tooltipOptions.icon } : getIconElementProps(tab.tooltipOptions.icon)), { size: this.size === 'xxsmall' ? 'xsmall' : this.size, dataTestid: getDataTestId(itemDataTestid, 'tooltip-icon') }))));
            }
            return null;
        }
    }
    render() {
        const rootClasses = classNames('tk-tabs', [`tk-tabs-${this.orientation}`], [`tk-tabs-${this.size}`], [`tk-tabs-${this.type}`], [`tk-tabs-${this.variant}`]);
        const headersClasses = classNames('tab-headers', { spread: this.spreadHeaders });
        const rootProps = {
            'class': rootClasses,
            'style': this.containerStyle,
            'data-testid': getDataTestId(this.dataTestid, 'container'),
        };
        const headersProps = {
            'class': headersClasses,
            'style': Object.assign({ justifyContent: this.alignHeaders }, (this.headerContainerStyle && Object.assign({}, this.headerContainerStyle))),
            'data-testid': getDataTestId(this.dataTestid, 'headers'),
        };
        const contentProps = {
            'class': 'tab-content',
            'style': this.contentStyle,
            'data-testid': getDataTestId(this.dataTestid, 'content'),
        };
        return (h("div", Object.assign({ key: '29fe16afc52d6b622b4490d80b594e06aba9eca2' }, rootProps), h("div", Object.assign({ key: '9ff12ed80336dfb4edf179c343110abfbe317fb8' }, headersProps), this.internalTabItems.map((tab, index) => {
            const itemDataTestid = getDataTestId(this.dataTestid, 'item', index.toString());
            const headerClasses = classNames('tab-header', { 'active': this.internalActiveIndex === index, 'tk-tab-header-disabled': tab.disabled });
            const tabIcons = this.createTabIcons(tab, index, itemDataTestid);
            return (h("div", { class: headerClasses, onClick: () => this.handleTabClick(index), "data-testid": getDataTestId(itemDataTestid) }, tabIcons.leftIcon, h("div", { class: "tk-tabs-item-label-container", "data-testid": getDataTestId(itemDataTestid, 'label-container') }, h("span", { class: "tk-tabs-item-label", "data-testid": getDataTestId(itemDataTestid, 'label') }, tab.label), this.renderTabBadge(tab, index, itemDataTestid), this.renderTabTooltip(tab, itemDataTestid)), tabIcons.rightIcon, this.isClosable && (h("tk-icon", Object.assign({}, getIconElementProps('close', { variant: this.getIconVariant(index) }), { dataTestid: getDataTestId(itemDataTestid, 'close-icon'), onClick: e => {
                    e.stopPropagation();
                    this.closeTab(index);
                } })))));
        }), this.isExtendable && (h("tk-icon", Object.assign({ key: '624a829c409b212be618ef318db77af45affa596' }, getIconElementProps('add', { variant: 'neutral', class: classNames('tk-tabs-item-add-icon'), onclick: () => this.addTab() }), { dataTestid: getDataTestId(this.dataTestid, 'add-icon') })))), h("div", Object.assign({ key: '5cbff32a273e53050afd9ffeed2dca7e1b1b6268' }, contentProps), this.internalTabItems.map((_tab, index) => {
            const panelDataTestid = getDataTestId(this.dataTestid, 'panel', index.toString());
            return (h("div", { class: `tab-panel ${this.internalActiveIndex === index ? 'active' : 'hidden'}`, key: index, "data-testid": getDataTestId(panelDataTestid) }, this.internalActiveIndex === index && (h("div", { "data-testid": getDataTestId(panelDataTestid, 'content') }, h("slot", { name: `tab-content-${index}` })))));
        }))));
    }
    static get is() { return "tk-tabs"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-tabs.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-tabs.css"]
        };
    }
    static get properties() {
        return {
            "alignHeaders": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'start' | 'center' | 'end'",
                    "resolved": "\"center\" | \"end\" | \"start\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets the alignment of the header."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "align-headers",
                "defaultValue": "'start'"
            },
            "controlled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls if the tabs component is controlled."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "controlled",
                "defaultValue": "false"
            },
            "defaultActiveIndex": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "0"
                        }],
                    "text": "Default Active Index for tabs component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "default-active-index",
                "defaultValue": "0"
            },
            "activeIndex": {
                "type": "number",
                "mutable": true,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "0"
                        }],
                    "text": "Controls the currently active tab index.\nCan be controlled programmatically from outside."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "active-index",
                "defaultValue": "0"
            },
            "isClosable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls if tabs are closable."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "is-closable",
                "defaultValue": "false"
            },
            "isExtendable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls if new tabs can be added or not."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "is-extendable",
                "defaultValue": "false"
            },
            "orientation": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'horizontal' | 'vertical'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'horizontal'"
                        }],
                    "text": "Controls the orientation of the tabs component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "orientation",
                "defaultValue": "'horizontal'"
            },
            "spreadHeaders": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Determines whether the tab headers will spread evenly across the horizontal space."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "spread-headers",
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small' | 'xsmall' | 'xxsmall'",
                    "resolved": "\"base\" | \"large\" | \"small\" | \"xsmall\" | \"xxsmall\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Controls the size of the tabs component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'compact' | 'expanded'",
                    "resolved": "\"basic\" | \"compact\" | \"divided\" | \"expanded\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'basic'"
                        }],
                    "text": "Controls the tab style of the tabs component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'basic'"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'info' | 'neutral'",
                    "resolved": "\"info\" | \"neutral\" | \"primary\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'primary'"
                        }],
                    "text": "Controls the color variant of the tabs component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'primary'"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of container element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "headerContainerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of headers container element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "contentStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of tabs item element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "internalActiveIndex": {},
            "internalTabItems": {}
        };
    }
    static get events() {
        return [{
                "method": "tkTabClick",
                "name": "tk-tab-click",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Triggered when a tab is clicked. Returns the clicked tab index."
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }, {
                "method": "tkTabChange",
                "name": "tk-tab-change",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Triggered when the currently open tab changes. Returns the active index."
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "activeIndex",
                "methodName": "activeIndexChanged"
            }];
    }
}
