// @takeoff-ui/core@0.11.9/dist/collection/components/tk-tabs/tk-tabs-item.js
import { Host, h } from "@stencil/core";
export class TkTabsItem {
    constructor() {
        /**
         * Checks if tab item has badge component or not.
         * @defaultValue false
         */
        this.badged = false;
    }
    iconChanged() {
        this.emitUpdate();
    }
    labelChanged() {
        this.emitUpdate();
    }
    disabledChanged() {
        this.emitUpdate();
    }
    badgedChanged() {
        this.emitUpdate();
    }
    badgeLabelChanged() {
        this.emitUpdate();
    }
    badgeCountChanged() {
        this.emitUpdate();
    }
    badgeOptionsChanged() {
        this.emitUpdate();
    }
    tooltipOptionsChanged() {
        this.emitUpdate();
    }
    // Dispatch a custom event that the parent component can listen for
    emitUpdate() {
        this.el.dispatchEvent(new CustomEvent('tk-update', {
            bubbles: true,
            composed: true,
            detail: {
                label: this.label,
                icon: this.icon,
                disabled: this.disabled,
                badged: this.badged,
                badgeCount: this.badgeCount,
                badgeLabel: this.badgeLabel,
                badgeOptions: this.badgeOptions,
                tooltipOptions: this.tooltipOptions,
            },
        }));
    }
    render() {
        return (h(Host, { key: '633f113bba8aa0ab19534159b6654cd6b60db20f' }, h("slot", { key: '1a861b6242ada6cdc15071f1187d00480de7037b' })));
    }
    static get is() { return "tk-tabs-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-tabs-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-tabs-item.css"]
        };
    }
    static get properties() {
        return {
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon for tabs item component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label for the tab item."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the tab item is disabled."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled"
            },
            "badged": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Checks if tab item has badge component or not."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "badged",
                "defaultValue": "false"
            },
            "badgeLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets badge component's label."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "badge-label"
            },
            "badgeCount": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "number | string",
                    "resolved": "number | string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets badge component's count."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "badge-count"
            },
            "badgeOptions": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IBadgeOptions",
                    "resolved": "IBadgeOptions",
                    "references": {
                        "IBadgeOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IBadgeOptions",
                            "id": "src/global/interfaces/IBadgeOptions.ts::IBadgeOptions",
                            "referenceLocation": "IBadgeOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets badge component's options."
                },
                "getter": false,
                "setter": false
            },
            "tooltipOptions": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ITooltipOptions",
                    "resolved": "ITooltipOptions",
                    "references": {
                        "ITooltipOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/ITooltipOptions",
                            "id": "src/global/interfaces/ITooltipOptions.ts::ITooltipOptions",
                            "referenceLocation": "ITooltipOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets tooltip options for the tab item."
                },
                "getter": false,
                "setter": false
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "icon",
                "methodName": "iconChanged"
            }, {
                "propName": "label",
                "methodName": "labelChanged"
            }, {
                "propName": "disabled",
                "methodName": "disabledChanged"
            }, {
                "propName": "badged",
                "methodName": "badgedChanged"
            }, {
                "propName": "badgeLabel",
                "methodName": "badgeLabelChanged"
            }, {
                "propName": "badgeCount",
                "methodName": "badgeCountChanged"
            }, {
                "propName": "badgeOptions",
                "methodName": "badgeOptionsChanged"
            }, {
                "propName": "tooltipOptions",
                "methodName": "tooltipOptionsChanged"
            }];
    }
}
