// @takeoff-ui/core@0.11.9/dist/collection/components/tk-table/tk-table.js
import { h, Fragment } from "@stencil/core";
import classNames from "classnames";
import { filterAndSort, handleInputKeydown, calculateColumnStartWidth, calculateNewColumnWidth } from "./helpers";
import { isEqual, some } from "lodash-es";
import jsPDF from "jspdf";
import autoTable from "jspdf-autotable";
import ExcelJs from "exceljs";
import { getIconElementProps } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
import "../../global/sass/fonts/tk-font/tk-text-regular";
import "../../global/sass/fonts/tk-font/tk-text-bold";
import "../../global/sass/fonts/geologica/geologica-regular";
import "../../global/sass/fonts/geologica/geologica-bold";
import { getNestedValue } from "../../utils/object-utils";
import { showElement, hideElement } from "../../utils/style-utils";
import { floatingElementAutoUpdate } from "../../utils/position-utils";
/**
 * TkTable is a component that allows you to display data in a tabular manner. It's generally called a datatable.
 * @react `import { TkTable } from '@takeoff-ui/react'`
 * @vue `import { TkTable } from '@takeoff-ui/vue'`
 * @angular `import { TkTable } from '@takeoff-ui/angular'`
 * @slot empty-data - Set how the table will appear when there is no data
 * @slot body-header - Custom independent rows at the top of tbody (e.g., summary, totals, or custom data rows)
 * @slot body-footer - Custom independent rows at the bottom of tbody (e.g., totals, summary, or additional data rows)
 */
export class TkTable {
    constructor() {
        this.isResizing = false;
        this.resizeColumnIndex = -1;
        this.startX = 0;
        this.startWidth = 0;
        this.customCellCache = new Map();
        this.isSelectionUpdating = false;
        this.filters = [];
        this.currentPage = 1;
        this.renderData = [];
        this.isFilterOpen = false;
        this.columnWidths = {};
        this.stickyOffsets = { left: {}, right: {} };
        this.sorts = [];
        this.groupedData = [];
        this.groupByColumnField = null;
        this.isControlledGrouping = false;
        this.expandedGroups = [];
        /**
         * The column definitions (Array of Objects)
         */
        this.columns = [];
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        /**
         * List of the selected
         */
        this.selection = [];
        /**
         * Style to apply to header of table
         */
        this.headerType = 'basic';
        /**
         * Sets size for the component.
         */
        this.size = 'base';
        /**
         * Rows of data to display
         */
        this.data = [];
        /**
         *
         */
        this.cardTitle = '';
        /**
         * Enables or disables alternating row background colors for easier readability.
         */
        this.striped = false;
        /**
         * Number of items per page.
         */
        this.rowsPerPage = 6;
        /**
         * The type of the pagination
         * @defaultValue 'outlined'
         */
        this.paginationType = 'outlined';
        /**
         * Template string for current page report in pagination.
         * Available placeholders: {currentPage}, {totalPages}
         * @defaultValue 'page: {currentPage} of {totalPages}'
         */
        this.pageReportTemplate = 'page: {currentPage} of {totalPages}';
        /**
         * Template string for items report in pagination.
         * Available placeholders: {startItem}, {endItem}, {totalItems}
         * @defaultValue 'item: {startItem}-{endItem} of {totalItems}'
         */
        this.itemsReportTemplate = 'item: {startItem}-{endItem} of {totalItems}';
        /**
         * Specifies which rows are expanded to show additional content.
         */
        this.expandedRows = [];
        /**
         * Enables multi-column sorting.
         */
        this.multiSort = false;
        /**
         * If true, group headers will have an expand/collapse button to show/hide the rows in that group.
         * @defaultValue false
         */
        this.collapsibleGroups = false;
        this.handleRowClick = (e, row) => {
            const path = e.composedPath();
            const clickableElement = path.some(element => element instanceof HTMLElement && ['tk-popover', 'tk-dropdown'].includes(element.tagName.toLowerCase()));
            if (clickableElement) {
                e.stopPropagation();
            }
            else {
                this.tkRowClick.emit(row);
            }
        };
        this.handleMouseDown = (e, columnIndex) => {
            e.preventDefault();
            this.isResizing = true;
            this.resizeColumnIndex = columnIndex;
            this.startX = e.clientX;
            const th = e.target.closest('th');
            if (!th)
                return;
            const column = this.columns[columnIndex];
            const currentStateWidth = this.columnWidths[column === null || column === void 0 ? void 0 : column.field];
            // Use helper function to calculate starting width
            this.startWidth = calculateColumnStartWidth(th, currentStateWidth);
            this.el.style.cursor = 'col-resize';
            this.el.style.userSelect = 'none';
        };
        this.handleMouseMove = (e) => {
            if (!this.isResizing || this.resizeColumnIndex === -1)
                return;
            e.preventDefault();
            // Use helper function to calculate new width
            const newWidth = calculateNewColumnWidth(this.startX, e.clientX, this.startWidth);
            const column = this.columns[this.resizeColumnIndex];
            if (column) {
                this.columnWidths = Object.assign(Object.assign({}, this.columnWidths), { [column.field]: `${newWidth}px` });
                // Force re-render
                this.columnWidths = Object.assign({}, this.columnWidths);
            }
        };
        this.handleMouseUp = () => {
            if (this.isResizing) {
                this.isResizing = false;
                this.resizeColumnIndex = -1;
                // Reset cursor styles on the table container
                this.el.style.cursor = '';
                this.el.style.userSelect = '';
                this.updateStickyOffsets();
            }
        };
        this.handleScroll = (e) => {
            var _a, _b, _c;
            const target = e.target;
            const scrollLeft = target.scrollLeft;
            const scrollWidth = target.scrollWidth;
            const clientWidth = target.clientWidth;
            const maxScrollLeft = Math.max(0, scrollWidth - clientWidth);
            const EPS = 1; // tolerance for float rounding
            const hasOverflow = scrollWidth > clientWidth + EPS;
            const atLeft = scrollLeft <= EPS;
            const atRight = scrollLeft >= maxScrollLeft - EPS;
            // Update shadow visibility based on scroll position
            const leftStickyElements = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelectorAll('.tk-table-sticky-shadow-right');
            const rightStickyElements = (_b = this.el.shadowRoot) === null || _b === void 0 ? void 0 : _b.querySelectorAll('.tk-table-sticky-shadow-left');
            // Left shadow visibility
            if (leftStickyElements) {
                leftStickyElements.forEach((el) => {
                    el.style.setProperty('--shadow-opacity', hasOverflow && !atLeft ? '1' : '0');
                });
            }
            // Right shadow visibility
            if (rightStickyElements) {
                rightStickyElements.forEach((el) => {
                    el.style.setProperty('--shadow-opacity', hasOverflow && !atRight ? '1' : '0');
                });
            }
            // Header bottom shadow visibility on vertical scroll
            const scrollTop = target.scrollTop;
            const atTop = scrollTop <= EPS;
            const theadElement = (_c = this.el.shadowRoot) === null || _c === void 0 ? void 0 : _c.querySelector('thead');
            if (theadElement) {
                theadElement.style.setProperty('--header-shadow-opacity', !atTop ? '1' : '0');
            }
        };
    }
    // columns prop'u değiştiğinde sticky column'lar varsa offset'leri güncellemek ve scroll listener'ı kurmak için watch eklenmiştir.
    columnsChanged() {
        const stickyColumns = this.columns.filter(col => col.fixed === 'left' || col.fixed === 'right');
        if (stickyColumns.length > 0) {
            setTimeout(() => {
                this.updateStickyOffsets();
                this.setupScrollListener();
            }, 0);
        }
    }
    dataChanged(newValue, oldValue) {
        var _a;
        if (!isEqual(oldValue, newValue)) {
            if (this.paginationMethod == 'client') {
                const tmpData = filterAndSort(newValue, this.columns, this.filters, this.sortField, this.sortOrder, this.sorts);
                this.currentPage = 1;
                const startIndex = (this.currentPage - 1) * this.internalRowsPerPage;
                const endIndex = startIndex + this.internalRowsPerPage;
                this.renderData = ((_a = [...tmpData]) === null || _a === void 0 ? void 0 : _a.slice(startIndex, endIndex)) || [];
            }
            else {
                this.renderData = (newValue === null || newValue === void 0 ? void 0 : newValue.length) > 0 ? [...newValue] : [];
            }
            // Re-apply grouping if it was previously set
            if (this.groupBy || this.groupByColumnField) {
                const groupField = this.groupBy || this.groupByColumnField;
                this.applyGrouping(groupField);
            }
            if (this.refSelectAll)
                this.refSelectAll.value = false;
            this.handleSelectAll(false);
            this.expandedRows = [];
            // Invalidate cached custom cell elements on data change
            this.customCellCache.clear();
        }
    }
    rowsPerPageChanged(newValue, oldValue) {
        if (oldValue !== newValue) {
            this.internalRowsPerPage = newValue;
        }
    }
    groupByChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            if (newValue) {
                this.applyGrouping(newValue);
            }
            else {
                this.clearGroupingInternal();
            }
            this.tkGroupByChange.emit(newValue || null);
        }
    }
    // outside click of search tk-table-filter-panel for close
    checkForClickOutside(ev) {
        if (this.isFilterOpen) {
            let isInside = false;
            ev.composedPath().forEach((item) => {
                if (this.elFilterPanelElement == item || this.elActiveSearchIcon == item) {
                    isInside = true;
                    return;
                }
            });
            if (!isInside) {
                this.closeFilterPanel();
            }
        }
    }
    componentWillLoad() {
        var _a;
        this.hasHeaderRightSlot = !!this.el.querySelector('[slot="header-right"]');
        this.hasEmptyDataSlot = !!this.el.querySelector('[slot="empty-data"]');
        this.hasEmptyFilterSlot = !!this.el.querySelector('[slot="empty-filter"]');
        // Determine if this is a controlled component based on initial groupBy prop
        this.isControlledGrouping = this.groupBy !== undefined;
        this.internalRowsPerPage = this.rowsPerPage;
        if (((_a = this.data) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.generateRenderData(this.data, this.currentPage, true);
        }
        // Apply grouping if groupBy prop is set
        if (this.groupBy) {
            this.applyGrouping(this.groupBy);
        }
        // Initialize column widths from column definitions
        this.columns.forEach(col => {
            if (col.width) {
                this.columnWidths[col.field] = col.width;
            }
        });
        // Add mouse event listeners for column resizing
        document.addEventListener('mousemove', this.handleMouseMove.bind(this));
        document.addEventListener('mouseup', this.handleMouseUp.bind(this));
    }
    componentDidLoad() {
        var _a, _b;
        const stickyColumns = this.columns.filter(col => col.fixed === 'left' || col.fixed === 'right');
        const hasSelectionMode = !!this.selectionMode;
        const hasScrollableContainer = !!(((_a = this.containerStyle) === null || _a === void 0 ? void 0 : _a.height) || ((_b = this.containerStyle) === null || _b === void 0 ? void 0 : _b.maxHeight));
        if (stickyColumns.length > 0 || hasSelectionMode || hasScrollableContainer) {
            this.updateStickyOffsets();
            this.setupScrollListener();
        }
    }
    componentDidRender() {
        // Reset transient selection flag after a render cycle
        if (this.isSelectionUpdating) {
            this.isSelectionUpdating = false;
        }
        this.refreshStickyShadows();
    }
    componentWillUpdate() {
        var _a, _b, _c;
        // empty-data slot'unun data her değiştiğinde görünürlüğünü ayarlamak için yapılmıştır.
        const slotEmptyFilter = this.el.querySelector("[slot='empty-filter']");
        const slotEmptyData = this.el.querySelector("[slot='empty-data']");
        if (slotEmptyFilter) {
            if (this.hasEmptyFilterSlot && ((_a = this.filters) === null || _a === void 0 ? void 0 : _a.length) > 0 && !this.loading && ((_b = this.renderData) === null || _b === void 0 ? void 0 : _b.length) === 0) {
                showElement(slotEmptyFilter);
            }
            else {
                hideElement(slotEmptyFilter);
            }
        }
        if (slotEmptyData) {
            if (this.loading || ((_c = this.renderData) === null || _c === void 0 ? void 0 : _c.length) > 0) {
                hideElement(slotEmptyData);
            }
            else {
                showElement(slotEmptyData);
            }
        }
    }
    componentDidUpdate() {
        var _a, _b;
        if (this.isFilterOpen) {
            if (this.elActiveSearchIcon && this.elFilterPanelElement) {
                // Clean up old floating UI listeners before setting up new ones
                (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
                this.updatePanelPosition();
            }
        }
        else {
            // Remove floating UI listeners when filter closes
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
            this.closeFilterPanel();
        }
    }
    disconnectedCallback() {
        var _a, _b;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        this.elActiveSearchIcon = null;
        this.elFilterPanelElement = null;
        this.isFilterOpen = false;
        const existingFilterPanel = document.querySelector('.tk-table-filter-panel');
        if (existingFilterPanel) {
            existingFilterPanel.remove();
        }
        // Clean up resize event listeners
        document.removeEventListener('mousemove', this.handleMouseMove.bind(this));
        document.removeEventListener('mouseup', this.handleMouseUp.bind(this));
        // Clean up scroll event listener
        const tableHolder = (_b = this.el.shadowRoot) === null || _b === void 0 ? void 0 : _b.querySelector('.table-holder');
        if (tableHolder) {
            tableHolder.removeEventListener('scroll', this.handleScroll.bind(this));
        }
    }
    /**
     * Allows tk-request event to be triggered manually
     */
    async serverRequest() {
        const requestData = {
            currentPage: this.currentPage,
            rowsPerPage: this.internalRowsPerPage,
            sortField: this.sortField,
            sortOrder: this.sortOrder,
            sorts: this.sorts,
            filters: this.filters,
        };
        if (this.paginationMethod === 'client') {
            requestData.totalItems = this.totalItems;
            requestData.data = this.data;
        }
        this.tkRequest.emit(requestData);
    }
    /**
     * Exports the table data to a file
     * @param options
     */
    async exportFile(options) {
        var _a, _b, _c, _d;
        let _data;
        let _columns;
        if (((_a = options === null || options === void 0 ? void 0 : options.externalData) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            _data = options.externalData;
        }
        else if (options.scope == 'all') {
            _data = filterAndSort(this.data, this.columns, this.filters, this.sortField, this.sortOrder, this.sorts);
        }
        else if (options.scope == 'selected') {
            _data = this.selection;
        }
        else if (!options.scope || options.scope == 'current-page') {
            _data = this.renderData;
        }
        if (options.type == 'pdf') {
            const doc = new jsPDF(options.orientation === 'horizontal' ? 'l' : 'p');
            const fontFamily = getComputedStyle(this.el).getPropertyValue('--family-body').trim().replace(/['"]/g, '');
            const pdfFont = fontFamily.startsWith('tk-text') ? 'tk-text' : 'Geologica';
            autoTable(doc, {
                head: [this.columns.map(col => col.header)], // Başlıkları dinamik olarak ekler
                body: _data.map(row => this.columns.map(col => {
                    var _a;
                    if (col.exportFormat) {
                        return col.exportFormat(row);
                    }
                    else {
                        return (_a = getNestedValue(row, col.field)) !== null && _a !== void 0 ? _a : '';
                    }
                })),
                theme: 'striped',
                // styles: { halign: 'center', fontSize: 10 },
                headStyles: { fillColor: [201, 0, 25], fontStyle: 'bold' },
                bodyStyles: { fontStyle: 'normal' },
                styles: { font: pdfFont },
            });
            doc.save(`${(_b = options.fileName) !== null && _b !== void 0 ? _b : 'tk-table'}.pdf`);
        }
        else if (options.type == 'excel') {
            const workbook = new ExcelJs.Workbook();
            const worksheet = workbook.addWorksheet('Sheet 1');
            if (options.columns && options.columns.length > 0) {
                _columns = options.columns;
            }
            else {
                _columns = this.columns;
            }
            worksheet.columns = _columns
                .filter(item => { var _a; return !((_a = options.ignoreColumnsFields) === null || _a === void 0 ? void 0 : _a.includes(item.field)); })
                .map(item => {
                var _a;
                return { header: item.header, key: item.field, width: Number((_a = item.width) === null || _a === void 0 ? void 0 : _a.toString().replace('px', '')) / 7 || 20 }; //Add this 7 because of unit conversion (In excel inches are used and 7px is approximately 1inch)
            });
            worksheet.addRows(_data === null || _data === void 0 ? void 0 : _data.map(item => {
                const rowData = {};
                _columns
                    .filter(col => { var _a; return !((_a = options.ignoreColumnsFields) === null || _a === void 0 ? void 0 : _a.includes(col.field)); })
                    .forEach(col => {
                    var _a;
                    if (col.exportFormat) {
                        rowData[col.field] = col.exportFormat(item);
                    }
                    else {
                        rowData[col.field] = (_a = getNestedValue(item, col.field)) !== null && _a !== void 0 ? _a : '';
                    }
                });
                return rowData;
            }));
            const buffer = await workbook.xlsx.writeBuffer();
            const blob = new Blob([buffer], { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
            const link = document.createElement('a');
            link.href = URL.createObjectURL(blob);
            link.download = `${(_c = options.fileName) !== null && _c !== void 0 ? _c : 'tk-table'}.xlsx`;
            link.click();
        }
        else if (options.type == 'csv') {
            const headers = this.columns.map(col => col.header).join(',');
            const rows = _data
                .map(row => this.columns
                .map(col => {
                var _a;
                if (col.exportFormat) {
                    return col.exportFormat(row);
                }
                else {
                    return (_a = getNestedValue(row, col.field)) !== null && _a !== void 0 ? _a : '';
                }
            })
                .join(','))
                .join('\n');
            const csvContent = headers + '\n' + rows;
            const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' });
            const link = document.createElement('a');
            if (link.download !== undefined) {
                link.setAttribute('href', URL.createObjectURL(blob));
                link.setAttribute('download', `${(_d = options.fileName) !== null && _d !== void 0 ? _d : 'tk-table'}.csv`);
                link.style.visibility = 'hidden';
                document.body.appendChild(link);
                link.click();
                document.body.removeChild(link);
            }
        }
    }
    /**
     * Clears all filters or specific column filters
     * @param columns Optional array of column field names to clear filters for. If not provided, all filters are cleared.
     */
    async clearFilters(columns) {
        var _a;
        if (!((_a = this.filters) === null || _a === void 0 ? void 0 : _a.length))
            return;
        if (columns === null || columns === void 0 ? void 0 : columns.length) {
            // Clear filters for specific columns
            this.filters = this.filters.filter(filter => !columns.includes(filter.field));
        }
        else {
            // Clear all filters
            this.filters = [];
        }
        this.currentPage = 1;
        if (this.paginationMethod !== 'server') {
            const tmpData = filterAndSort(this.data, this.columns, this.filters, this.sortField, this.sortOrder, this.sorts);
            this.generateRenderData(tmpData, 1, true);
        }
    }
    /**
     * Clears all sorting for server side pagination
     */
    async clearSorting() {
        if (this.sortField || this.sorts.length > 0) {
            this.sortField = null;
            this.sortOrder = null;
            this.sorts = [];
            this.currentPage = 1;
            // all sort icons are reset to their default state
            this.el.shadowRoot.querySelectorAll('thead th .tk-table-head-cell .sort-icon').forEach((icon) => (icon.icon = 'swap_vert'));
            if (this.paginationMethod !== 'server') {
                const tmpData = filterAndSort(this.data, this.columns, this.filters, this.sortField, this.sortOrder, this.sorts);
                this.generateRenderData(tmpData, 1, true);
            }
        }
    }
    /**
     * Returns the current filters
     * @returns {Object} The current filters
     */
    async getFilters() {
        return this.filters;
    }
    /**
     * Returns the current sorting settings
     * @returns {Object} The current sorting settings
     */
    async getSorting() {
        if (this.multiSort) {
            return this.sorts;
        }
        else {
            return {
                field: this.sortField,
                order: this.sortOrder,
            };
        }
    }
    /**
     * Sets the current page for pagination
     * @param page The page number to set (1-based index)
     */
    async setCurrentPage(page) {
        this.currentPage = page;
    }
    /**
     * Sets the current filter settings
     */
    async setFilters(filters) {
        this.filters = filters;
    }
    /**
     * Sets the current sorting settings
     */
    async setSorting(sorts) {
        var _a;
        if (this.multiSort && Array.isArray(sorts)) {
            this.sorts = sorts;
            (_a = this.sorts) === null || _a === void 0 ? void 0 : _a.forEach(sort => {
                const sortIcon = this.el.shadowRoot.querySelector(`thead tr th[data-field="${sort.field}"] tk-icon`);
                if (sortIcon) {
                    if (sort.order == 'asc') {
                        sortIcon.icon = 'arrow_drop_up';
                    }
                    else if (sort.order == 'desc') {
                        sortIcon.icon = 'arrow_drop_down';
                    }
                }
            });
        }
        else if (!this.multiSort && !Array.isArray(sorts)) {
            this.sortField = sorts.field;
            this.sortOrder = sorts.order;
            const sortIcon = this.el.shadowRoot.querySelector(`thead tr th[data-field="${sorts.field}"] tk-icon`);
            if (sortIcon) {
                if (sorts.order == 'asc') {
                    sortIcon.icon = 'arrow_drop_up';
                }
                else if (sorts.order == 'desc') {
                    sortIcon.icon = 'arrow_drop_down';
                }
            }
        }
    }
    /**
     * Applies the current filters to the data for client side pagination
     */
    async runFilters() {
        if (this.paginationMethod !== 'server') {
            this.currentPage = 1;
            const tmpData = filterAndSort(this.data, this.columns, this.filters, this.sortField, this.sortOrder, this.sorts);
            this.generateRenderData(tmpData, 1, true);
        }
    }
    /**
     * Groups table data by the specified column field
     * Creates group header rows that display the unique value and count of items in that group.
     * For example, if you have a 'status' column with values 'Open' and 'Closed',
     * this will create group headers like "Open (5)" and "Closed (3)".
     *
     * Always emits tkGroupByChange event. For uncontrolled components, also updates internal state.
     *
     * @param columnField The field name to group by (e.g., 'status', 'category', 'department')
     *
     * @example
     * // Controlled usage (with groupBy prop)
     * await tableRef.groupByColumn('status'); // Emits tkGroupByChange event
     *
     * // Uncontrolled usage (no groupBy prop)
     * await tableRef.groupByColumn('status'); // Updates internal state and emits tkGroupByChange event
     */
    async groupByColumn(columnField) {
        if (!this.isControlledGrouping) {
            // Uncontrolled component - update internal state directly
            if (!columnField) {
                this.clearGroupingInternal();
            }
            else {
                this.applyGrouping(columnField);
            }
        }
        this.tkGroupByChange.emit(columnField || null);
    }
    /**
     * Clears the current grouping and returns to normal table view
     *
     * Always emits tkGroupByChange event with null value. For uncontrolled components, also clears internal state.
     */
    async clearGrouping() {
        if (!this.isControlledGrouping) {
            // Uncontrolled component - clear internal state directly
            this.clearGroupingInternal();
        }
        this.tkGroupByChange.emit(null);
    }
    applyGrouping(columnField) {
        if (!columnField) {
            this.clearGroupingInternal();
            return;
        }
        const filteredData = this.getTableViewData();
        // Group data by the specified column
        const groups = new Map();
        filteredData.forEach(row => {
            const groupValue = getNestedValue(row, columnField);
            if (!groups.has(groupValue)) {
                groups.set(groupValue, []);
            }
            groups.get(groupValue).push(row);
        });
        // Convert to array format for rendering
        this.groupedData = Array.from(groups.entries()).map(([groupValue, rows]) => ({
            groupValue,
            groupCount: rows.length,
            rows,
        }));
        this.groupByColumnField = columnField;
        // Başlangıçta tüm gruplar genişletilmiş olarak gösterilir
        this.expandedGroups = this.groupedData.map(group => group.groupValue);
        // Generate render data from grouped data
        const flatRenderData = [];
        this.groupedData.forEach(group => {
            flatRenderData.push(...group.rows);
        });
        if (this.paginationMethod === 'client') {
            const startIndex = (this.currentPage - 1) * this.internalRowsPerPage;
            const endIndex = startIndex + this.internalRowsPerPage;
            this.renderData = flatRenderData.slice(startIndex, endIndex);
        }
        else {
            this.renderData = flatRenderData;
        }
        this.totalItems = flatRenderData.length;
    }
    updatePanelPosition() {
        this.cleanup = floatingElementAutoUpdate(this.elActiveSearchIcon, this.elFilterPanelElement, undefined, {
            placement: 'bottom',
            offset: 4,
        });
    }
    clearGroupingInternal() {
        this.groupByColumnField = null;
        this.groupedData = [];
        this.expandedGroups = [];
        this.generateRenderData(this.data, this.currentPage, true);
    }
    generateRenderData(data, currentPage, isWillLoad = false) {
        const _data = [...(data !== null && data !== void 0 ? data : [])];
        this.currentPage = currentPage;
        // Clear grouping when generating render data with new dataset
        if (!this.groupByColumnField) {
            this.groupedData = [];
        }
        // component will load dan geldiğinde tkRequest tetiklenmesin diye bu kontrol eklendi.
        if (!isWillLoad) {
            const requestData = {
                currentPage: this.currentPage,
                rowsPerPage: this.internalRowsPerPage,
                sortField: this.sortField,
                sortOrder: this.sortOrder,
                sorts: this.sorts,
                filters: this.filters,
            };
            if (this.paginationMethod === 'client') {
                requestData.totalItems = _data === null || _data === void 0 ? void 0 : _data.length;
                requestData.data = _data;
            }
            this.tkRequest.emit(requestData);
        }
        if (this.paginationMethod == 'client') {
            const startIndex = (this.currentPage - 1) * this.internalRowsPerPage;
            const endIndex = startIndex + this.internalRowsPerPage;
            this.renderData = _data.slice(startIndex, endIndex);
            this.totalItems = _data === null || _data === void 0 ? void 0 : _data.length;
        }
        else {
            this.renderData = _data;
        }
        this.expandedRows = [];
        // Slice/view changed -> clear cached custom cells to avoid stale nodes
        this.customCellCache.clear();
    }
    /**
     * Tabloda gösterilecek satırlar. Server pagination'da filtre/sıralama sunucuda uygulanır;
     * bu yüzden client tarafında filterAndSort tekrar çalıştırılmaz (çift filtreleme ve boş tablo riski).
     */
    getTableViewData() {
        const source = Array.isArray(this.data) ? this.data : [];
        if (this.paginationMethod === 'server') {
            return [...source];
        }
        return filterAndSort(source, this.columns, this.filters, this.sortField, this.sortOrder, this.sorts);
    }
    toggleExpandRow(row, tdExpanderButtonRef) {
        const newExpandedRows = this.expandedRows;
        const findIndex = newExpandedRows.findIndex(item => item[this.dataKey] == row[this.dataKey]);
        if (findIndex > -1) {
            tdExpanderButtonRef.icon = 'keyboard_arrow_down';
            newExpandedRows.splice(findIndex, 1); // Eğer zaten genişlemişse kapat
        }
        else {
            tdExpanderButtonRef.icon = 'keyboard_arrow_up';
            newExpandedRows.push(row); // Değilse genişlet
        }
        this.expandedRows = newExpandedRows;
        this.tkExpandedRowsChange.emit(this.expandedRows);
        // After state change, recalc shadows on next frame
        requestAnimationFrame(() => this.refreshStickyShadows());
    }
    toggleExpandGroup(groupValue) {
        if (!this.collapsibleGroups)
            return;
        const existingIndex = this.expandedGroups.indexOf(groupValue);
        if (existingIndex > -1) {
            // collapse the group
            this.expandedGroups = [...this.expandedGroups.slice(0, existingIndex), ...this.expandedGroups.slice(existingIndex + 1)];
        }
        else {
            // expand the group
            this.expandedGroups = [...this.expandedGroups, groupValue];
        }
        // After state change, recalc shadows on next frame
        requestAnimationFrame(() => this.refreshStickyShadows());
    }
    // Add a new method to safely close the filter panel
    closeFilterPanel() {
        var _a;
        // Clean up floating UI listeners before removing element
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        // Then remove the element from DOM
        if (this.elFilterPanelElement) {
            this.elFilterPanelElement.remove();
            this.elFilterPanelElement = null;
        }
        // Finally update the state
        this.isFilterOpen = false;
    }
    // Checks if all selectable rows are selected
    isAllRowsSelected() {
        if (!Array.isArray(this.selection))
            return false;
        const selectableRows = this.renderData.filter(row => (this.selectionRowDisabled ? !this.selectionRowDisabled(row) : true));
        if (selectableRows.length === 0)
            return false;
        if (this.selection.length < selectableRows.length)
            return false;
        return selectableRows.every(row => this.selection.some(sel => (sel === null || sel === void 0 ? void 0 : sel[this.dataKey]) === (row === null || row === void 0 ? void 0 : row[this.dataKey])));
    }
    setDataTestidAttribute(el, ...suffixes) {
        const testId = getDataTestId(this.dataTestid, ...suffixes);
        if (testId) {
            el.setAttribute('data-testid', testId);
        }
    }
    async handleSearchIconClick(refSearchIcon, field) {
        if (!this.isFilterOpen) {
            this.isFilterOpen = true;
            this.createFilterPanel(refSearchIcon, field);
        }
        else if (this.elActiveSearchIcon != refSearchIcon) {
            this.isFilterOpen = false;
            this.createFilterPanel(refSearchIcon, field);
        }
    }
    handleInputFilterApply(columnField) {
        var _a;
        // if (refSearchInput.value.toString().length > 0) {
        const searchInput = document.querySelector('body > .tk-table-filter-panel > tk-input');
        const value = (_a = searchInput.value) !== null && _a !== void 0 ? _a : '';
        // Bu field da mevcutta bir filtre uygulanmış ise mevcutu değiştirmek için yazıldı.
        // filtre yoksa yeni filtre olarak filters'a eklenmesi sağlandı
        const filter = this.filters.find(filter => filter.field == columnField);
        if (filter)
            filter.value = value.toString();
        else
            this.filters.push({ field: columnField, value });
        // current page değiştiğinde pagination componenti 'handlePageChange' eventini tetiklediğinden 2 defa emit edilmesin diye buraya bu kontrol eklendi
        if (this.currentPage == 1) {
            const tmpData = this.getTableViewData();
            this.generateRenderData(tmpData, 1);
        }
        else {
            this.currentPage = 1;
        }
        // Close the filter panel
        this.closeFilterPanel();
    }
    handleSearchButtonClick(columnField) {
        var _a, _b, _c, _d;
        if (((_a = this.columns.find(col => col.field === columnField)) === null || _a === void 0 ? void 0 : _a.filterType) === 'checkbox') {
            this.handleCheckboxFilterApply(columnField);
        }
        else if (((_b = this.columns.find(col => col.field === columnField)) === null || _b === void 0 ? void 0 : _b.filterType) === 'radio') {
            this.handleRadioFilterApply(columnField);
        }
        else if (((_c = this.columns.find(col => col.field === columnField)) === null || _c === void 0 ? void 0 : _c.filterType) === 'datepicker') {
            this.handleDatepickerFilterApply(columnField);
        }
        else if (((_d = this.columns.find(col => col.field === columnField)) === null || _d === void 0 ? void 0 : _d.filterType) === 'treeview') {
            this.handleTreeviewFilterApply(columnField);
        }
        else {
            this.handleInputFilterApply(columnField);
        }
    }
    handleSearchCancelButtonClick(columnField) {
        const removeFilterIndex = this.filters.findIndex(filter => filter.field == columnField);
        if (removeFilterIndex > -1) {
            this.filters.splice(removeFilterIndex, 1);
            // Reset input value if it's a text filter
            const searchInput = document.querySelector('body > .tk-table-filter-panel > tk-input');
            if (searchInput)
                searchInput.value = '';
            // current page değiştiğinde pagination componenti 'handlePageChange' eventini tetiklediğinden 2 defa emit edilmesin diye buraya bu kontrol eklendi
            if (this.currentPage == 1) {
                const tmpData = this.getTableViewData();
                this.generateRenderData(tmpData, 1);
            }
            else {
                this.currentPage = 1;
            }
        }
        // Close the filter panel
        this.closeFilterPanel();
    }
    handleInputBlur(row, index, colField, editableInputRef) {
        const cellEdit = {
            rowId: row[this.dataKey],
            rowIndex: index,
            field: colField,
            value: editableInputRef.value,
        };
        this.tkCellEdit.emit(cellEdit);
    }
    handleSelectAll(value) {
        this.isSelectionUpdating = true;
        if (value) {
            this.selection = [...this.renderData.filter(row => (this.selectionRowDisabled ? !this.selectionRowDisabled(row) : true))];
        }
        else {
            this.selection = [];
        }
        this.tkSelectionChange.emit(this.selection);
    }
    handleCheckboxSelectChange(isSelect, row) {
        this.isSelectionUpdating = true;
        let tmpSelection = Array.isArray(this.selection) ? [...this.selection] : [];
        const hasSelect = some(tmpSelection, item => isEqual(item, row));
        if (isSelect == false && hasSelect) {
            // seçili ise ve silinmek isteniyor ise
            tmpSelection = tmpSelection.filter(item => item[this.dataKey] !== row[this.dataKey]);
            this.selection = [...tmpSelection];
            this.tkSelectionChange.emit(this.selection);
            this.refSelectAll.indeterminate = true;
        }
        else if (isSelect == true && !hasSelect) {
            // seçili değilse ve eklenmek isteniyor ise
            tmpSelection.push(row);
            this.selection = [...tmpSelection];
            this.tkSelectionChange.emit(this.selection);
            this.refSelectAll.indeterminate = true;
        }
    }
    handleRadioSelectChange(row) {
        this.isSelectionUpdating = true;
        this.selection = row;
        this.tkSelectionChange.emit(this.selection);
    }
    handlePageChange(e) {
        const tmpData = this.getTableViewData();
        this.generateRenderData(tmpData, Number(e.detail.page));
        // sayfa değişikliğinde seçilen değerler sıfırlanır
        if (this.refSelectAll)
            this.refSelectAll.value = false;
        this.handleSelectAll(false);
    }
    handleSortIconClick(refSortIcon, col) {
        if (!col.sortable)
            return;
        if (this.multiSort) {
            this.handleMultiSort(refSortIcon, col);
        }
        else {
            this.handleSingleSort(refSortIcon, col);
        }
    }
    handleSingleSort(refSortIcon, col) {
        this.sortField = col.field;
        const icon = refSortIcon.icon;
        this.sorts = [];
        // tüm sort iconlar default duruma getirilir.
        this.el.shadowRoot.querySelectorAll('thead th .tk-table-head-cell .sort-icon').forEach((icon) => (icon.icon = 'swap_vert'));
        if (icon == 'swap_vert') {
            this.sortOrder = 'asc';
            refSortIcon.icon = 'arrow_drop_up';
        }
        else if (icon == 'arrow_drop_up') {
            this.sortOrder = 'desc';
            refSortIcon.icon = 'arrow_drop_down';
        }
        else if (icon == 'arrow_drop_down') {
            this.sortField = null;
            this.sortOrder = null;
            refSortIcon.icon = 'swap_vert';
        }
        this.applySorting();
    }
    handleMultiSort(refSortIcon, col) {
        const existingIndex = this.sorts.findIndex(s => s.field === col.field);
        const icon = refSortIcon.icon;
        if (existingIndex > -1) {
            const currentSort = this.sorts[existingIndex];
            if (icon === 'arrow_drop_up') {
                currentSort.order = 'desc';
            }
            else if (icon === 'arrow_drop_down') {
                this.sorts.splice(existingIndex, 1);
            }
        }
        else {
            this.sorts.push({
                field: col.field,
                order: 'asc',
            });
        }
        this.applySorting();
    }
    applySorting() {
        if (this.currentPage === 1) {
            const tmpData = this.getTableViewData();
            this.generateRenderData(tmpData, 1);
        }
        else {
            this.currentPage = 1;
        }
    }
    createFilterPanel(refSearchIcon, field) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7;
        // First close any existing filter panel
        this.closeFilterPanel();
        this.elActiveSearchIcon = refSearchIcon;
        this.elFilterPanelElement = document.createElement('div');
        this.elFilterPanelElement.classList.add('tk-table-filter-panel');
        this.elFilterPanelElement.classList.add(`${field}-filter-panel`);
        this.setDataTestidAttribute(this.elFilterPanelElement, 'filter-panel', field);
        // Find the column configuration for this field
        const column = this.columns.find(col => col.field === field);
        // Check if the column has a filterType property and it's set to 'checkbox'
        if ((column === null || column === void 0 ? void 0 : column.filterType) === 'checkbox' && ((_a = column.filterOptions) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            // Create checkbox filter
            const filterContainer = document.createElement('div');
            filterContainer.classList.add('tk-table-filter-checkbox-container');
            this.setDataTestidAttribute(filterContainer, 'filter-checkbox-container', field);
            // Get current filter values for this field
            const currentFilter = this.filters.find(filter => filter.field === field);
            const selectedValues = (currentFilter === null || currentFilter === void 0 ? void 0 : currentFilter.value) || [];
            const divider = document.createElement('tk-divider');
            divider.my = 1;
            const checkboxWrapper = document.createElement('div');
            checkboxWrapper.classList.add('tk-table-filter-checkbox-item');
            if ((_c = (_b = column === null || column === void 0 ? void 0 : column.filterElements) === null || _b === void 0 ? void 0 : _b.optionsSearchInput) === null || _c === void 0 ? void 0 : _c.show) {
                const optionsSearchInput = document.createElement('tk-input');
                optionsSearchInput.placeholder = column.filterElements.optionsSearchInput.placeholder || 'Search';
                optionsSearchInput.dataTestid = getDataTestId(this.dataTestid, 'filter-options-search-input', field);
                optionsSearchInput.addEventListener('tk-change', (e) => {
                    var _a, _b;
                    const searchText = e.detail.toLowerCase();
                    const checkboxWrappers = filterContainer.querySelectorAll('.tk-table-filter-checkbox-item');
                    checkboxWrappers.forEach((wrapper) => {
                        const checkbox = wrapper.querySelector('tk-checkbox:not(.select-all)');
                        if (checkbox) {
                            const label = checkbox.label.toLowerCase();
                            wrapper.style.display = label.includes(searchText) ? 'block' : 'none';
                        }
                    });
                    const visibleCheckboxes = Array.from(checkboxWrappers).filter(wrapper => {
                        const checkbox = wrapper.querySelector('tk-checkbox:not(.select-all)');
                        return checkbox && wrapper.style.display !== 'none';
                    });
                    if (visibleCheckboxes.length === 0) {
                        checkboxWrapper.style.display = 'none';
                        divider.style.display = 'none';
                        optionsSearchInput.hint = ((_b = (_a = column === null || column === void 0 ? void 0 : column.filterElements) === null || _a === void 0 ? void 0 : _a.optionsSearchInput) === null || _b === void 0 ? void 0 : _b.emptyMessage) || 'No results found';
                    }
                    else {
                        checkboxWrapper.style.display = 'block';
                        divider.style.display = 'block';
                        optionsSearchInput.hint = null;
                    }
                });
                filterContainer.appendChild(optionsSearchInput);
            }
            let allCheckbox;
            if (!((_e = (_d = column === null || column === void 0 ? void 0 : column.filterElements) === null || _d === void 0 ? void 0 : _d.selectAllCheckbox) === null || _e === void 0 ? void 0 : _e.hide)) {
                allCheckbox = document.createElement('tk-checkbox');
                allCheckbox.classList.add('select-all');
                this.setDataTestidAttribute(allCheckbox, 'filter-select-all-checkbox', field);
                allCheckbox.label = ((_g = (_f = column === null || column === void 0 ? void 0 : column.filterElements) === null || _f === void 0 ? void 0 : _f.selectAllCheckbox) === null || _g === void 0 ? void 0 : _g.label) || 'Select All';
                allCheckbox.value = selectedValues.length === column.filterOptions.length;
                if (selectedValues.length > 0 && selectedValues.length < column.filterOptions.length) {
                    allCheckbox.indeterminate = true;
                }
                checkboxWrapper.appendChild(allCheckbox);
                allCheckbox.addEventListener('tk-change', (e) => {
                    const allCheckboxes = filterContainer.querySelectorAll('tk-checkbox:not(.select-all)');
                    allCheckboxes.forEach((cb) => {
                        cb.value = e.detail;
                    });
                    if (e.detail) {
                        selectedValues.length = 0;
                        selectedValues.push(...column.filterOptions.map(option => option.value));
                    }
                    else {
                        selectedValues.length = 0;
                    }
                    allCheckbox.indeterminate = false;
                });
                filterContainer.appendChild(checkboxWrapper);
                filterContainer.appendChild(divider);
            }
            // Create checkboxes for each option
            column.filterOptions.forEach((option, optionIndex) => {
                var _a, _b;
                const checkboxWrapper = document.createElement('div');
                checkboxWrapper.classList.add('tk-table-filter-checkbox-item');
                this.setDataTestidAttribute(checkboxWrapper, 'filter-checkbox-item', field, optionIndex.toString());
                const checkbox = document.createElement('tk-checkbox');
                checkbox.value = selectedValues.includes(option.value);
                checkbox.label = option.label || option.value;
                this.setDataTestidAttribute(checkbox, 'filter-option-checkbox', field, optionIndex.toString());
                if (!((_b = (_a = column === null || column === void 0 ? void 0 : column.filterElements) === null || _a === void 0 ? void 0 : _a.selectAllCheckbox) === null || _b === void 0 ? void 0 : _b.hide)) {
                    checkbox.addEventListener('tk-change', () => {
                        const allItemCheckboxes = Array.from(filterContainer.querySelectorAll('tk-checkbox:not(.select-all)'));
                        const visibleCheckboxes = allItemCheckboxes.filter(cb => { var _a; return cb.style.display !== 'none' && ((_a = cb.parentElement) === null || _a === void 0 ? void 0 : _a.style.display) !== 'none'; });
                        const allSelected = visibleCheckboxes.length > 0 && visibleCheckboxes.every(cb => cb.value);
                        const anySelected = visibleCheckboxes.some(cb => cb.value);
                        allCheckbox.value = allSelected;
                        if (!allSelected && anySelected) {
                            allCheckbox.indeterminate = true;
                        }
                        else {
                            allCheckbox.indeterminate = false;
                        }
                    });
                }
                checkboxWrapper.appendChild(checkbox);
                filterContainer.appendChild(checkboxWrapper);
            });
            this.elFilterPanelElement.appendChild(filterContainer);
        }
        // Check if the column has a filterType property and it's set to 'radio'
        else if ((column === null || column === void 0 ? void 0 : column.filterType) === 'radio' && ((_h = column.filterOptions) === null || _h === void 0 ? void 0 : _h.length) > 0) {
            // Create radio filter
            const filterContainer = document.createElement('div');
            filterContainer.classList.add('tk-table-filter-radio-container');
            this.setDataTestidAttribute(filterContainer, 'filter-radio-container', field);
            // Get current filter value for this field
            const currentFilter = this.filters.find(filter => filter.field === field);
            const selectedValue = currentFilter === null || currentFilter === void 0 ? void 0 : currentFilter.value;
            // Create radio group name unique to this column
            const radioGroupName = `radio-filter-${field}`;
            if ((_j = column === null || column === void 0 ? void 0 : column.filterElements) === null || _j === void 0 ? void 0 : _j.optionsSearchInput) {
                const optionsSearchInput = document.createElement('tk-input');
                optionsSearchInput.placeholder = column.filterElements.optionsSearchInput.placeholder || 'Search';
                optionsSearchInput.dataTestid = getDataTestId(this.dataTestid, 'filter-options-search-input', field);
                optionsSearchInput.addEventListener('tk-change', (e) => {
                    const searchText = e.detail.toLowerCase();
                    const radioWrappers = filterContainer.querySelectorAll('.tk-table-filter-radio-item');
                    radioWrappers.forEach((wrapper) => {
                        const radio = wrapper.querySelector('tk-radio');
                        const label = radio.label.toLowerCase();
                        wrapper.style.display = label.includes(searchText) ? 'block' : 'none';
                    });
                });
                filterContainer.appendChild(optionsSearchInput);
            }
            // Create radio buttons for each option
            column.filterOptions.forEach((option, optionIndex) => {
                const radioWrapper = document.createElement('div');
                radioWrapper.classList.add('tk-table-filter-radio-item');
                this.setDataTestidAttribute(radioWrapper, 'filter-radio-item', field, optionIndex.toString());
                const radio = document.createElement('tk-radio');
                radio.value = option.value;
                radio.name = radioGroupName;
                radio.label = option.label || option.value;
                this.setDataTestidAttribute(radio, 'filter-option-radio', field, optionIndex.toString());
                // Check if this option is currently selected
                if (selectedValue === option.value) {
                    radio.checked = true;
                }
                radioWrapper.appendChild(radio);
                filterContainer.appendChild(radioWrapper);
            });
            this.elFilterPanelElement.appendChild(filterContainer);
        }
        else if ((column === null || column === void 0 ? void 0 : column.filterType) === 'datepicker') {
            const filterContainer = document.createElement('div');
            filterContainer.classList.add('tk-table-filter-datepicker-container');
            this.setDataTestidAttribute(filterContainer, 'filter-datepicker-container', field);
            const datepicker = document.createElement('tk-datepicker');
            datepicker.dataTestid = getDataTestId(this.dataTestid, 'filter-datepicker', field);
            const defaultDatepickerProps = {
                label: 'Select a date',
                placeholder: 'Choose a date',
                mode: 'single',
                dateFormat: 'yyyy-MM-dd',
                timeFormat: '24',
                minDate: '',
                maxDate: '',
                hourStep: 1,
                minuteStep: 1,
                locale: 'en',
                showTimePicker: false,
                size: 'base',
            };
            const currentFilter = this.filters.find(filter => filter.field === field);
            const currentValue = (currentFilter === null || currentFilter === void 0 ? void 0 : currentFilter.value) || null;
            Object.assign(datepicker, Object.assign(Object.assign(Object.assign({}, defaultDatepickerProps), (_k = column === null || column === void 0 ? void 0 : column.filterElements) === null || _k === void 0 ? void 0 : _k.optionsSearchDatepicker), { value: currentValue }));
            datepicker.addEventListener('tk-change', (e) => {
                datepicker.value = e.detail;
            });
            filterContainer.appendChild(datepicker);
            this.elFilterPanelElement.appendChild(filterContainer);
        }
        else if ((column === null || column === void 0 ? void 0 : column.filterType) === 'treeview' && ((_l = column.filterOptions) === null || _l === void 0 ? void 0 : _l.length) > 0) {
            // Create treeview filter
            const filterContainer = document.createElement('div');
            filterContainer.classList.add('tk-table-filter-treeview-container');
            this.setDataTestidAttribute(filterContainer, 'filter-treeview-container', field);
            // Get current filter value for this field
            const currentFilter = this.filters.find(filter => filter.field === field);
            const selectedValue = (currentFilter === null || currentFilter === void 0 ? void 0 : currentFilter.value) || [];
            // Create treeview component first
            const treeview = document.createElement('tk-tree-view');
            this.setDataTestidAttribute(treeview, 'filter-treeview', field);
            const treeviewConfig = (_o = (_m = column === null || column === void 0 ? void 0 : column.filterElements) === null || _m === void 0 ? void 0 : _m.treeViewOptions) !== null && _o !== void 0 ? _o : {};
            treeview.selectable = true;
            treeview.size = (_p = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.size) !== null && _p !== void 0 ? _p : 'small';
            treeview.branchIcon = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.branchIcon;
            treeview.leafIcon = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.leafIcon;
            treeview.showBadge = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.showBadge;
            treeview.showZeroCountBadges = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.showZeroCountBadges;
            treeview.badgeOptions = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.badgeOptions;
            treeview.showPointer = (_q = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.showPointer) !== null && _q !== void 0 ? _q : false;
            treeview.selectionStrategy = (_r = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.selectionStrategy) !== null && _r !== void 0 ? _r : 'leaf';
            treeview.containerStyle = (_s = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.containerStyle) !== null && _s !== void 0 ? _s : { width: '100%' };
            treeview.stepStyle = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.stepStyle;
            treeview.expandAll = (_t = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.expandAll) !== null && _t !== void 0 ? _t : true;
            treeview.expandedKeys = treeviewConfig === null || treeviewConfig === void 0 ? void 0 : treeviewConfig.expandedKeys;
            treeview.items = column.filterOptions;
            treeview.value = selectedValue;
            treeview.addEventListener('tk-change', (e) => {
                treeview.value = e.detail;
            });
            if ((_v = (_u = column === null || column === void 0 ? void 0 : column.filterElements) === null || _u === void 0 ? void 0 : _u.optionsSearchInput) === null || _v === void 0 ? void 0 : _v.show) {
                const optionsSearchInput = document.createElement('tk-input');
                optionsSearchInput.placeholder = column.filterElements.optionsSearchInput.placeholder || 'Search';
                optionsSearchInput.dataTestid = getDataTestId(this.dataTestid, 'filter-options-search-input', field);
                optionsSearchInput.addEventListener('tk-change', (e) => {
                    const searchText = e.detail.toLowerCase();
                    // Recursive function to filter tree items including nested children
                    const filterTreeItems = (items) => {
                        return items
                            .map(item => {
                            // Check if current item matches, at start its the parent node
                            const itemMatches = item.label.toLowerCase().includes(searchText);
                            // Recursively filter children
                            const filteredChildren = item.children ? filterTreeItems(item.children) : [];
                            // Include both parent and children if children match, or if parent matches
                            if (itemMatches || filteredChildren.length > 0) {
                                return Object.assign(Object.assign({}, item), { children: filteredChildren.length > 0 ? filteredChildren : item.children });
                            }
                            return null;
                        })
                            .filter(item => item !== null);
                    };
                    treeview.items = filterTreeItems(column.filterOptions);
                });
                filterContainer.appendChild(optionsSearchInput);
            }
            filterContainer.appendChild(treeview);
            this.elFilterPanelElement.appendChild(filterContainer);
        }
        else {
            // Default text input filter
            const input = document.createElement('tk-input');
            const searchInputConfig = (_x = (_w = column === null || column === void 0 ? void 0 : column.filterElements) === null || _w === void 0 ? void 0 : _w.searchInput) !== null && _x !== void 0 ? _x : {};
            input.dataTestid = getDataTestId(this.dataTestid, 'filter-search-input', field);
            input.placeholder = (searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.placeholder) || 'Search';
            input.label = searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.label;
            input.maskOptions = searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.maskOptions;
            input.disabled = !!(searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.disabled);
            input.invalid = !!(searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.invalid);
            input.clearable = !!(searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.clearable);
            input.error = searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.error;
            input.hint = searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.hint;
            input.icon = searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.icon;
            input.iconPosition = searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.iconPosition;
            input.size = (searchInputConfig === null || searchInputConfig === void 0 ? void 0 : searchInputConfig.size) || 'base';
            input.setFocus();
            input.value = ((_z = (_y = this.filters) === null || _y === void 0 ? void 0 : _y.find(item => item.field == field)) === null || _z === void 0 ? void 0 : _z.value) || '';
            input.addEventListener('keydown', (e) => {
                if (e.key === 'Enter') {
                    this.handleSearchButtonClick(field);
                }
            });
            this.elFilterPanelElement.appendChild(input);
        }
        // Create buttons container
        const buttons = document.createElement('div');
        buttons.classList.add('tk-table-filter-panel-buttons');
        this.setDataTestidAttribute(buttons, 'filter-panel-buttons', field);
        // Create cancel button
        const cancelButton = document.createElement('tk-button');
        cancelButton.label = ((_1 = (_0 = column === null || column === void 0 ? void 0 : column.filterElements) === null || _0 === void 0 ? void 0 : _0.cancelButton) === null || _1 === void 0 ? void 0 : _1.label) || ((_3 = (_2 = column === null || column === void 0 ? void 0 : column.filterButtons) === null || _2 === void 0 ? void 0 : _2.cancelButton) === null || _3 === void 0 ? void 0 : _3.label) || 'Remove';
        cancelButton.type = 'outlined';
        cancelButton.fullWidth = true;
        cancelButton.dataTestid = getDataTestId(this.dataTestid, 'filter-cancel-button', field);
        cancelButton.addEventListener('tk-click', () => {
            this.handleSearchCancelButtonClick(field);
            this.isFilterOpen = false;
        });
        // Create search/apply button
        const searchButton = document.createElement('tk-button');
        searchButton.label = ((_5 = (_4 = column === null || column === void 0 ? void 0 : column.filterElements) === null || _4 === void 0 ? void 0 : _4.searchButton) === null || _5 === void 0 ? void 0 : _5.label) || ((_7 = (_6 = column === null || column === void 0 ? void 0 : column.filterButtons) === null || _6 === void 0 ? void 0 : _6.searchButton) === null || _7 === void 0 ? void 0 : _7.label) || 'Apply';
        searchButton.fullWidth = true;
        searchButton.dataTestid = getDataTestId(this.dataTestid, 'filter-apply-button', field);
        searchButton.addEventListener('tk-click', () => {
            this.handleSearchButtonClick(field);
            // We don't need to close the filter panel here anymore since it's closed in the apply methods
        });
        buttons.appendChild(cancelButton);
        buttons.appendChild(searchButton);
        this.elFilterPanelElement.appendChild(buttons);
        this.elFilterPanelElement.style.zIndex = '1700';
        document.body.appendChild(this.elFilterPanelElement);
        this.isFilterOpen = true;
    }
    handleCheckboxFilterApply(columnField) {
        const checkboxes = Array.from(document.querySelectorAll('body > .tk-table-filter-panel .tk-table-filter-checkbox-item tk-checkbox:not(.select-all)'));
        const column = this.columns.find(col => col.field === columnField);
        if (!column || !column.filterOptions || checkboxes.length === 0)
            return;
        // Get selected values
        const selectedValues = [];
        const filterOptions = column.filterOptions;
        checkboxes.forEach((checkbox, index) => {
            if (checkbox.value && filterOptions[index] && checkbox.style.display !== 'none') {
                selectedValues.push(filterOptions[index].value);
            }
        });
        // Update filter
        const filterIndex = this.filters.findIndex(filter => filter.field === columnField);
        if (selectedValues.length > 0) {
            if (filterIndex > -1) {
                this.filters[filterIndex].value = selectedValues;
                this.filters[filterIndex].type = 'checkbox';
            }
            else {
                this.filters.push({
                    field: columnField,
                    value: selectedValues,
                    type: 'checkbox',
                });
            }
        }
        else if (filterIndex > -1) {
            // Remove filter if no values selected
            this.filters.splice(filterIndex, 1);
        }
        // Apply filter
        if (this.currentPage === 1) {
            const tmpData = this.getTableViewData();
            this.generateRenderData(tmpData, 1);
        }
        else {
            this.currentPage = 1;
        }
        // Close the filter panel
        this.closeFilterPanel();
    }
    handleRadioFilterApply(columnField) {
        const radioButtons = Array.from(document.querySelectorAll(`body > .tk-table-filter-panel .tk-table-filter-radio-item tk-radio`));
        const column = this.columns.find(col => col.field === columnField);
        if (!column || !column.filterOptions || radioButtons.length === 0)
            return;
        // Find selected radio button
        const selectedRadio = radioButtons.find((radio) => radio.checked);
        // Update filter
        const filterIndex = this.filters.findIndex(filter => filter.field === columnField);
        if (selectedRadio) {
            const selectedValue = selectedRadio.value;
            if (filterIndex > -1) {
                this.filters[filterIndex].value = selectedValue;
                this.filters[filterIndex].type = 'radio';
            }
            else {
                this.filters.push({
                    field: columnField,
                    value: selectedValue,
                    type: 'radio',
                });
            }
        }
        else if (filterIndex > -1) {
            // Remove filter if no value selected
            this.filters.splice(filterIndex, 1);
        }
        // Apply filter
        if (this.currentPage === 1) {
            const tmpData = this.getTableViewData();
            this.generateRenderData(tmpData, 1);
        }
        else {
            this.currentPage = 1;
        }
        // Close the filter panel
        this.closeFilterPanel();
    }
    handleDatepickerFilterApply(columnField) {
        const datepickerEl = document.querySelector('body > .tk-table-filter-panel .tk-table-filter-datepicker-container tk-datepicker');
        const selectedDate = datepickerEl === null || datepickerEl === void 0 ? void 0 : datepickerEl.value;
        const filterIndex = this.filters.findIndex(filter => filter.field == columnField);
        if (selectedDate) {
            if (filterIndex > -1) {
                this.filters[filterIndex].value = selectedDate;
                this.filters[filterIndex].type = 'datepicker';
            }
            else {
                this.filters.push({ field: columnField, value: selectedDate, type: 'datepicker' });
            }
        }
        else if (filterIndex > -1) {
            // Remove filter if date is cleared
            this.filters.splice(filterIndex, 1);
        }
        // Update table data
        if (this.currentPage === 1) {
            const tmpData = this.getTableViewData();
            this.generateRenderData(tmpData, 1);
        }
        else {
            this.currentPage = 1;
        }
        // Close the filter panel
        this.closeFilterPanel();
    }
    handleTreeviewFilterApply(columnField) {
        const treeview = document.querySelector('.tk-table-filter-treeview-container tk-tree-view');
        if (!treeview)
            return;
        const selectedValues = treeview.value;
        const filterIndex = this.filters.findIndex(filter => filter.field === columnField);
        if (selectedValues.length > 0) {
            if (filterIndex > -1) {
                this.filters[filterIndex].value = selectedValues;
                this.filters[filterIndex].type = 'treeview';
            }
            else {
                this.filters.push({
                    field: columnField,
                    value: selectedValues,
                    type: 'treeview',
                });
            }
        }
        else if (filterIndex > -1) {
            // Remove filter if no values selected
            this.filters.splice(filterIndex, 1);
        }
        // Apply filter
        if (this.currentPage === 1) {
            const tmpData = this.getTableViewData();
            this.generateRenderData(tmpData, 1);
        }
        else {
            this.currentPage = 1;
        }
        // Close the filter panel
        this.closeFilterPanel();
    }
    updateStickyOffsets() {
        var _a;
        const leftColumns = this.columns.filter(col => col.fixed === 'left');
        const rightColumns = this.columns.filter(col => col.fixed === 'right');
        // Reset offsets
        this.stickyOffsets = { left: {}, right: {} };
        const measuredWidths = {};
        let shouldUseMeasuredWidths = false;
        // Try to measure from DOM if available
        const tableElement = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('table');
        if (tableElement) {
            const headerRow = tableElement.querySelector('thead tr');
            if (headerRow) {
                const thElements = Array.from(headerRow.querySelectorAll('th'));
                // Measure actual widths from DOM
                thElements.forEach((th, index) => {
                    const actualWidth = Math.round(th.getBoundingClientRect().width);
                    if (index === 0 && this.selectionMode) {
                        // Selection column
                        measuredWidths['selection'] = actualWidth;
                    }
                    else {
                        // Data column
                        const dataColumnIndex = this.selectionMode ? index - 1 : index;
                        if (dataColumnIndex >= 0 && dataColumnIndex < this.columns.length) {
                            const column = this.columns[dataColumnIndex];
                            if (column) {
                                measuredWidths[column.field] = actualWidth;
                            }
                        }
                    }
                });
                shouldUseMeasuredWidths = true;
            }
        }
        // Calculate left sticky offsets (cumulative from left)
        let leftOffset = 0;
        // Add selection column width if it exists
        if (this.selectionMode) {
            if (shouldUseMeasuredWidths && measuredWidths['selection']) {
                // Use measured width if available
                leftOffset += measuredWidths['selection'];
            }
            else {
                // Fallback to estimated width
                leftOffset += 52;
            }
        }
        leftColumns.forEach(col => {
            this.stickyOffsets.left[col.field] = leftOffset;
            let numericWidth = 120;
            if (shouldUseMeasuredWidths && measuredWidths[col.field]) {
                // Use measured width (priority)
                numericWidth = measuredWidths[col.field];
            }
            else {
                // Use state or column definition width
                const width = this.columnWidths[col.field] || col.width;
                if (width) {
                    if (typeof width === 'string') {
                        numericWidth = parseInt(width.replace(/px|%|em|rem/, '')) || 120;
                    }
                    else if (typeof width === 'number') {
                        numericWidth = width;
                    }
                }
            }
            leftOffset += numericWidth;
        });
        // Calculate right sticky offsets (cumulative from right)
        let rightOffset = 0;
        // Process right columns in reverse order (rightmost first)
        [...rightColumns].reverse().forEach(col => {
            this.stickyOffsets.right[col.field] = rightOffset;
            let numericWidth = 120;
            if (shouldUseMeasuredWidths && measuredWidths[col.field]) {
                // Use measured width (priority)
                numericWidth = measuredWidths[col.field];
            }
            else {
                // Use state or column definition width
                const width = this.columnWidths[col.field] || col.width;
                if (width) {
                    if (typeof width === 'string') {
                        numericWidth = parseInt(width.replace(/px|%|em|rem/, '')) || 120;
                    }
                    else if (typeof width === 'number') {
                        numericWidth = width;
                    }
                }
            }
            rightOffset += numericWidth;
        });
        // Force re-render if using measured widths
        if (shouldUseMeasuredWidths) {
            this.columnWidths = Object.assign({}, this.columnWidths);
        }
    }
    setupScrollListener() {
        var _a;
        const tableHolder = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.table-holder');
        if (tableHolder) {
            tableHolder.addEventListener('scroll', this.handleScroll.bind(this));
            // Initialize shadow state on mount
            this.handleScroll({ target: tableHolder });
        }
    }
    refreshStickyShadows() {
        var _a;
        const tableHolder = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.table-holder');
        if (!tableHolder)
            return;
        this.handleScroll({ target: tableHolder });
    }
    createGroupedRows() {
        var _a;
        const rows = [];
        let globalIndex = 0;
        // Group headerın olduğu cellden sonra oluşan boşluk için genişlik hesaplaması
        const tableHolder = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.table-holder');
        if (tableHolder) {
            const tableHolderWidth = tableHolder.clientWidth;
            tableHolder.style.setProperty('--table-holder-width', `${tableHolderWidth}px`);
        }
        // For pagination, we need to determine which groups and rows to show
        let startIndex = 0;
        let endIndex = this.renderData.length;
        if (this.paginationMethod === 'client') {
            startIndex = (this.currentPage - 1) * this.internalRowsPerPage;
            endIndex = startIndex + this.internalRowsPerPage;
        }
        let currentRowIndex = 0;
        this.groupedData.forEach((group, groupIndex) => {
            // Check if this group contains any rows in the current page range
            const groupStartIndex = currentRowIndex;
            const groupEndIndex = currentRowIndex + group.rows.length;
            const isGroupExpanded = this.expandedGroups.includes(group.groupValue);
            // If this group overlaps with the visible range, show it
            if (groupEndIndex > startIndex && groupStartIndex < endIndex) {
                // Create group header row
                const totalColumns = this.columns.length + (this.selectionMode ? 1 : 0);
                const groupHeaderRow = (h("tr", { class: classNames('tk-table-group-header', { 'tk-table-collapsible-group-header': this.collapsibleGroups }, { 'tk-table-group-collapsed': this.collapsibleGroups && !isGroupExpanded }), "data-testid": getDataTestId(this.dataTestid, 'group-header', groupIndex.toString()), onClick: () => this.toggleExpandGroup(group.groupValue) }, h("td", { colSpan: totalColumns, class: "tk-table-group-header-cell", "data-testid": getDataTestId(this.dataTestid, 'group-header-cell', groupIndex.toString()) }, h("div", { class: "tk-table-group-header-content", "data-testid": getDataTestId(this.dataTestid, 'group-header-content', groupIndex.toString()) }, this.collapsibleGroups && (h("tk-icon", { icon: isGroupExpanded ? 'keyboard_arrow_down' : 'keyboard_arrow_right', variant: "neutral", dataTestid: getDataTestId(this.dataTestid, 'group-header-toggle-icon', groupIndex.toString()) })), h("span", { class: "tk-table-group-value", "data-testid": getDataTestId(this.dataTestid, 'group-header-value', groupIndex.toString()) }, group.groupValue), h("span", { class: "tk-table-group-count", "data-testid": getDataTestId(this.dataTestid, 'group-header-count', groupIndex.toString()) }, "(", group.groupCount, ")")))));
                rows.push(groupHeaderRow);
                if (isGroupExpanded) {
                    // Show only the visible rows from this group
                    group.rows.forEach((row, groupRowIndex) => {
                        const absoluteRowIndex = currentRowIndex + groupRowIndex;
                        // Only show rows that are in the visible range
                        if (absoluteRowIndex >= startIndex && absoluteRowIndex < endIndex) {
                            const dataRow = this.createDataRow(row, globalIndex + groupRowIndex);
                            rows.push(dataRow);
                        }
                    });
                }
            }
            currentRowIndex += group.rows.length;
            globalIndex += group.rows.length;
        });
        return rows;
    }
    createDataRow(row, index) {
        var _a;
        const rowKey = String((_a = row === null || row === void 0 ? void 0 : row[this.dataKey]) !== null && _a !== void 0 ? _a : index);
        let styleRowObject;
        const leftColumns = this.columns.filter(c => c.fixed === 'left');
        if (typeof this.rowStyle == 'function') {
            const stylesRow = this.rowStyle(row, index);
            if (stylesRow !== undefined)
                styleRowObject = Object.assign({}, stylesRow);
        }
        let isRowDisabled = false;
        if (this.selectionRowDisabled) {
            isRowDisabled = this.selectionRowDisabled(row);
        }
        let selectionTd;
        if (this.selectionMode === 'checkbox') {
            selectionTd = (h("td", { class: classNames('non-text', 'tk-table-left-sticky', 'tk-table-sticky-first', { 'tk-table-sticky-shadow-right': leftColumns.length === 0 }), style: this.getSelectionStickyStyle(index), "data-testid": getDataTestId(this.dataTestid, 'body-row-selection-cell', rowKey) }, h("tk-checkbox", { id: this.el.id ? `${this.el.id}-checkbox-${index}` : undefined, value: some(this.selection, itemValue => isEqual(itemValue, row)), disabled: isRowDisabled, "onTk-change": e => this.handleCheckboxSelectChange(e.detail, row), onClick: e => e.stopPropagation(), "data-testid": getDataTestId(this.dataTestid, 'body-row-selection-checkbox', rowKey) })));
        }
        else if (this.selectionMode === 'radio') {
            selectionTd = (h("td", { class: classNames('non-text', 'tk-table-left-sticky', 'tk-table-sticky-first', { 'tk-table-sticky-shadow-right': leftColumns.length === 0 }), style: this.getSelectionStickyStyle(index), "data-testid": getDataTestId(this.dataTestid, 'body-row-selection-cell', rowKey) }, h("tk-radio", { id: this.el.id ? `${this.el.id}-radio-${index}` : undefined, value: row, name: this.el.id ? `${this.el.id}-selection` : 'selection', checked: isEqual(this.selection, row), disabled: isRowDisabled, "onTk-change": () => this.handleRadioSelectChange(row), onClick: e => e.stopPropagation(), "data-testid": getDataTestId(this.dataTestid, 'body-row-selection-radio', rowKey) })));
        }
        const isSelected = this.selectionMode === 'checkbox' ? some(this.selection, itemValue => isEqual(itemValue, row)) : this.selectionMode === 'radio' ? isEqual(this.selection, row) : false;
        return (h(Fragment, null, h("tr", { class: isSelected ? 'selected' : '', onClick: e => this.handleRowClick(e, row), "aria-disabled": isRowDisabled, "data-testid": getDataTestId(this.dataTestid, 'body-row', rowKey) }, selectionTd, this.columns.map(col => {
            var _a, _b, _c;
            let tdExpanderButtonRef;
            let styleCellObject;
            const leftColumns = this.columns.filter(c => c.fixed === 'left');
            const rightColumns = this.columns.filter(c => c.fixed === 'right');
            const isFirstLeft = col.fixed === 'left' && ((_a = leftColumns[0]) === null || _a === void 0 ? void 0 : _a.field) === col.field;
            const isLastRight = col.fixed === 'right' && ((_b = rightColumns[rightColumns.length - 1]) === null || _b === void 0 ? void 0 : _b.field) === col.field;
            if (typeof this.cellStyle == 'function') {
                const stylesCell = this.cellStyle(row, col);
                if (stylesCell !== undefined)
                    styleCellObject = Object.assign({}, stylesCell);
            }
            if (col.expander) {
                return (h("td", { class: classNames(this.getStickyColumnClasses(col, isFirstLeft, isLastRight)), style: Object.assign(Object.assign(Object.assign({}, this.getStickyColumnStyle(col, index)), styleRowObject), styleCellObject), "data-testid": getDataTestId(this.dataTestid, 'body-cell-expander', rowKey, String(col.field)) }, h("tk-button", { ref: el => (tdExpanderButtonRef = el), variant: "info", icon: "keyboard_arrow_down", iconPosition: "right", type: "text", size: "small", "onTk-click": () => this.toggleExpandRow(row, tdExpanderButtonRef), dataTestid: getDataTestId(this.dataTestid, 'body-expander-button', rowKey, String(col.field)) })));
            }
            else if (typeof (col === null || col === void 0 ? void 0 : col.html) == 'function') {
                // Reuse cached custom cell elements whenever possible
                let usedElement = null;
                let computedElement = null;
                const cacheKey = `${(_c = row === null || row === void 0 ? void 0 : row[this.dataKey]) !== null && _c !== void 0 ? _c : index}::${col.field}`;
                const cached = this.customCellCache.get(cacheKey);
                if (cached) {
                    usedElement = cached;
                }
                else {
                    computedElement = col === null || col === void 0 ? void 0 : col.html(row, index);
                }
                const effectiveElement = usedElement !== null && usedElement !== void 0 ? usedElement : computedElement;
                if (typeof effectiveElement == 'string') {
                    return (h("td", { class: classNames('non-text', this.getStickyColumnClasses(col, isFirstLeft, isLastRight)), innerHTML: effectiveElement, "data-testid": getDataTestId(this.dataTestid, 'body-cell-html', rowKey, String(col.field)), style: Object.assign(Object.assign(Object.assign({}, this.getStickyColumnStyle(col, index)), styleRowObject), styleCellObject) }));
                }
                else if (typeof effectiveElement == 'object') {
                    // Update cache if we computed a new element in this render
                    if (!usedElement) {
                        this.customCellCache.set(cacheKey, effectiveElement);
                    }
                    return (h("td", { ref: el => el === null || el === void 0 ? void 0 : el.replaceChildren(effectiveElement), class: classNames('non-text', this.getStickyColumnClasses(col, isFirstLeft, isLastRight)), "data-testid": getDataTestId(this.dataTestid, 'body-cell-custom', rowKey, String(col.field)), style: Object.assign(Object.assign(Object.assign({}, this.getStickyColumnStyle(col, index)), styleRowObject), styleCellObject) }));
                }
            }
            else if (col.editable) {
                let editableInputRef;
                return (h("td", { class: classNames('non-text editable', this.getStickyColumnClasses(col, isFirstLeft, isLastRight)), "data-testid": getDataTestId(this.dataTestid, 'body-cell-editable', rowKey, String(col.field)), style: Object.assign(Object.assign(Object.assign({}, this.getStickyColumnStyle(col, index)), styleRowObject), styleCellObject) }, h("input", { ref: el => (editableInputRef = el), value: getNestedValue(row, col.field), type: "text", onKeyDown: e => handleInputKeydown(e, this.el), onBlur: () => this.handleInputBlur(row, index, col.field, editableInputRef), "data-testid": getDataTestId(this.dataTestid, 'body-cell-editable-input', rowKey, String(col.field)) })));
            }
            else {
                return (h("td", { class: classNames(this.getStickyColumnClasses(col, isFirstLeft, isLastRight)), "data-testid": getDataTestId(this.dataTestid, 'body-cell', rowKey, String(col.field)), style: Object.assign(Object.assign(Object.assign({}, this.getStickyColumnStyle(col, index)), styleRowObject), styleCellObject), ref: el => { var _a; return el === null || el === void 0 ? void 0 : el.replaceChildren((_a = getNestedValue(row, col.field)) !== null && _a !== void 0 ? _a : ''); } }));
            }
        })), this.expandedRows.length > 0 && this.expandedRows.findIndex(item => item[this.dataKey] == row[this.dataKey]) > -1 && (h("tr", { "data-testid": getDataTestId(this.dataTestid, 'body-expanded-row', rowKey) }, h("td", { colSpan: 100, style: typeof this.expandedRowStyle == 'function' ? this.expandedRowStyle(row) : {}, "data-testid": getDataTestId(this.dataTestid, 'body-expanded-row-cell', rowKey) }, h("slot", { name: `expand-content-${row[this.dataKey]}` }))))));
    }
    getStickyColumnClasses(col, isFirst = false, isLast = false) {
        var _a;
        const classes = [];
        if (col.fixed === 'left') {
            classes.push('tk-table-left-sticky');
            if (isFirst)
                classes.push('tk-table-sticky-first');
            // En sağdaki left sticky kolonda shadow göster
            const leftColumns = this.columns.filter(c => c.fixed === 'left');
            const isLastLeft = ((_a = leftColumns[leftColumns.length - 1]) === null || _a === void 0 ? void 0 : _a.field) === col.field;
            if (isLastLeft)
                classes.push('tk-table-sticky-shadow-right');
        }
        else if (col.fixed === 'right') {
            classes.push('tk-table-right-sticky');
            if (isLast)
                classes.push('tk-table-sticky-last');
            // En soldaki right sticky kolonda shadow göster (en büyük offset'li)
            const rightColumns = this.columns.filter(c => c.fixed === 'right');
            if (rightColumns.length > 0) {
                // Right kolonlarda en soldaki = en büyük offset'e sahip olan
                const leftmostRightCol = rightColumns.reduce((leftmost, current) => {
                    var _a, _b;
                    const currentOffset = ((_a = this.stickyOffsets.right) === null || _a === void 0 ? void 0 : _a[current.field]) || 0;
                    const leftmostOffset = ((_b = this.stickyOffsets.right) === null || _b === void 0 ? void 0 : _b[leftmost.field]) || 0;
                    return currentOffset > leftmostOffset ? current : leftmost;
                }, rightColumns[0]);
                const isLeftmostRight = (leftmostRightCol === null || leftmostRightCol === void 0 ? void 0 : leftmostRightCol.field) === col.field;
                if (isLeftmostRight)
                    classes.push('tk-table-sticky-shadow-left');
            }
        }
        return classes.join(' ');
    }
    getStickyColumnStyle(col, rowIndex) {
        const style = {
            width: this.columnWidths[col.field] || col.width,
            minWidth: this.columnWidths[col.field] || col.width,
            maxWidth: this.columnWidths[col.field] || col.width,
        };
        if (col.fixed === 'left' && this.stickyOffsets.left[col.field] !== undefined) {
            const leftColumns = this.columns.filter(c => c.fixed === 'left');
            const columnIndex = leftColumns.findIndex(c => c.field === col.field);
            style['--tk-table-sticky-left-offset'] = `${this.stickyOffsets.left[col.field]}px`;
            style['left'] = `${this.stickyOffsets.left[col.field]}px`;
            // Z-index: hem columnIndex hem rowIndex'e göre azalt
            // En soldaki kolon en yüksek z-index'e sahip olmalı (en üstte durmalı)
            style['zIndex'] = Math.max(99 - columnIndex - (rowIndex !== null && rowIndex !== void 0 ? rowIndex : 0), 0);
        }
        else if (col.fixed === 'right' && this.stickyOffsets.right[col.field] !== undefined) {
            style['--tk-table-sticky-right-offset'] = `${this.stickyOffsets.right[col.field]}px`;
            style['right'] = `${this.stickyOffsets.right[col.field]}px`;
            // Right sticky z-index mantığı: En sağdaki (offset=0) en yüksek z-index
            const rightOffset = this.stickyOffsets.right[col.field];
            // Right kolonlar için z-index: offset'e göre hesaplıyoruz
            // offset=0 (en sağdaki) -> z-index=30
            // offset arttıkça z-index azalır
            const offsetLevel = Math.floor(rightOffset / 50); // Her 50px'te bir level
            const zIndex = 99 - offsetLevel;
            style['zIndex'] = Math.max(zIndex - (rowIndex !== null && rowIndex !== void 0 ? rowIndex : 0), 0); // Satır aşağı indikçe z-index azalır
        }
        return style;
    }
    getSelectionStickyStyle(rowIndex) {
        const style = {};
        // Selection column is always the left-most sticky column
        style['--tk-table-sticky-left-offset'] = `0px`;
        style['left'] = `0px`;
        if (rowIndex !== undefined) {
            style['zIndex'] = Math.max(99 - (rowIndex !== null && rowIndex !== void 0 ? rowIndex : 0), 0);
        }
        return style;
    }
    createHead() {
        const theadClasses = classNames(this.headerType);
        let selectionTh;
        if (this.selectionMode === 'checkbox') {
            const leftColumns = this.columns.filter(c => c.fixed === 'left');
            selectionTh = (h("th", { style: Object.assign({ width: '20px', maxWidth: '20px' }, this.getSelectionStickyStyle()), class: classNames('non-text', 'tk-table-left-sticky', 'tk-table-sticky-first', { 'tk-table-sticky-shadow-right': leftColumns.length === 0 }), "data-testid": getDataTestId(this.dataTestid, 'head-selection-cell') }, h("tk-checkbox", { id: this.el.id ? `${this.el.id}-checkbox-all` : undefined, value: this.isAllRowsSelected(), disabled: !(this.renderData.length > 0), ref: el => (this.refSelectAll = el), indeterminate: Array.isArray(this.selection) && this.selection.length > 0 && !this.isAllRowsSelected(), "onTk-change": e => this.handleSelectAll(e.detail), "data-testid": getDataTestId(this.dataTestid, 'head-selection-checkbox') })));
        }
        else if (this.selectionMode === 'radio') {
            const leftColumns = this.columns.filter(c => c.fixed === 'left');
            selectionTh = (h("th", { style: Object.assign({ width: '20px', maxWidth: '20px' }, this.getSelectionStickyStyle()), class: classNames('non-text', 'tk-table-left-sticky', 'tk-table-sticky-first', { 'tk-table-sticky-shadow-right': leftColumns.length === 0 }), "data-testid": getDataTestId(this.dataTestid, 'head-selection-cell') }));
        }
        return (h("thead", { class: theadClasses, "data-testid": getDataTestId(this.dataTestid, 'head') }, h("tr", { "data-testid": getDataTestId(this.dataTestid, 'head-row') }, selectionTh, this.columns.map(col => {
            var _a, _b, _c, _d, _e, _f, _g, _h;
            let refSortIcon;
            let refSearchIcon;
            let _sortIcon;
            let _searchIcon;
            let _headerStructure;
            let _customHeader;
            if (typeof (col === null || col === void 0 ? void 0 : col.headerHtml) == 'function') {
                _customHeader = col.headerHtml();
            }
            _headerStructure = (h("div", { class: "header-container", "data-testid": getDataTestId(this.dataTestid, 'head-cell-header-container', String(col.field)), ref: el => {
                    if (_customHeader) {
                        if (typeof _customHeader === 'string') {
                            el === null || el === void 0 ? void 0 : el.setHTMLUnsafe(_customHeader);
                        }
                        else if (typeof _customHeader == 'object') {
                            el === null || el === void 0 ? void 0 : el.replaceChildren(_customHeader);
                        }
                    }
                    else {
                        const _header = document.createElement('div');
                        _header.classList.add('header');
                        _header.title = col.header;
                        _header.textContent = col.header;
                        let _subHeader = null;
                        if (!col.subHeader) {
                            el === null || el === void 0 ? void 0 : el.replaceChildren(_header);
                        }
                        else {
                            _subHeader = document.createElement('div');
                            _subHeader.classList.add('sub-header');
                            _subHeader.title = col.subHeader;
                            _subHeader.textContent = col.subHeader;
                            el === null || el === void 0 ? void 0 : el.replaceChildren(_header, _subHeader);
                        }
                    }
                } }));
            // generate expander th
            if (col.expander) {
                const leftColumns = this.columns.filter(c => c.fixed === 'left');
                const rightColumns = this.columns.filter(c => c.fixed === 'right');
                const isFirstLeft = col.fixed === 'left' && ((_a = leftColumns[0]) === null || _a === void 0 ? void 0 : _a.field) === col.field;
                const isLastRight = col.fixed === 'right' && ((_b = rightColumns[rightColumns.length - 1]) === null || _b === void 0 ? void 0 : _b.field) === col.field;
                return (h("th", { class: classNames(this.getStickyColumnClasses(col, isFirstLeft, isLastRight)), style: Object.assign(Object.assign({}, this.getStickyColumnStyle(col)), col === null || col === void 0 ? void 0 : col.style), "data-testid": getDataTestId(this.dataTestid, 'head-cell-expander') }));
            }
            // generate head sort and search icons
            const sortIndex = this.sorts.findIndex(s => s.field === col.field);
            const sortObj = this.sorts.find(s => s.field === col.field);
            const iconType = sortObj ? (sortObj.order === 'asc' ? 'arrow_drop_up' : sortObj.order === 'desc' ? 'arrow_drop_down' : 'swap_vert') : 'swap_vert';
            const showBadge = sortIndex > -1 && this.sorts.length > 0 && this.multiSort;
            _sortIcon = showBadge ? (h("tk-badge", { count: sortIndex + 1, type: "text", rounded: true, size: "small", dataTestid: getDataTestId(this.dataTestid, 'head-cell-sort-badge', String(col.field)) }, h("tk-icon", Object.assign({}, getIconElementProps(iconType, {
                class: classNames('sort-icon'),
                color: this.headerType === 'basic' ? 'var(--icon-darkest)' : 'var(--icon-lightest)',
                ref: (el) => (refSortIcon = el),
                onClick: () => { var _a; return ((_a = this.renderData) === null || _a === void 0 ? void 0 : _a.length) > 0 && this.handleSortIconClick(refSortIcon, col); },
            }), { dataTestid: getDataTestId(this.dataTestid, 'head-cell-sort-icon', String(col.field)) })))) : (h("tk-icon", Object.assign({}, getIconElementProps('swap_vert', {
                class: classNames('sort-icon'),
                color: this.headerType === 'basic' ? 'var(--icon-darkest)' : 'var(--icon-lightest)',
                ref: (el) => (refSortIcon = el),
                onClick: () => { var _a; return ((_a = this.renderData) === null || _a === void 0 ? void 0 : _a.length) > 0 && this.handleSortIconClick(refSortIcon, col); },
            }), { dataTestid: getDataTestId(this.dataTestid, 'head-cell-sort-icon', String(col.field)) })));
            const currentFilter = this.filters.find(item => item.field == col.field);
            const hasFilter = currentFilter && ((Array.isArray(currentFilter.value) && currentFilter.value.length > 0) || (!Array.isArray(currentFilter.value) && currentFilter.value !== ''));
            const filterPanelOpen = (_c = this.elFilterPanelElement) === null || _c === void 0 ? void 0 : _c.classList.contains(`${col.field}-filter-panel`);
            const singleSorted = this.sortField === col.field && this.sortOrder;
            const multiSorted = !!(sortObj === null || sortObj === void 0 ? void 0 : sortObj.order);
            const isSorted = this.multiSort ? multiSorted : singleSorted;
            const noSortAndFilter = !isSorted && !hasFilter;
            if (col.searchable) {
                _searchIcon = (h("tk-icon", Object.assign({}, getIconElementProps(((_d = col === null || col === void 0 ? void 0 : col.filterElements) === null || _d === void 0 ? void 0 : _d.icon) || 'search', {
                    class: classNames('filter-icon'),
                    color: this.headerType === 'basic' ? 'var(--icon-darkest)' : 'var(--icon-lightest)',
                    ref: (el) => (refSearchIcon = el),
                    onClick: () => this.handleSearchIconClick(refSearchIcon, col.field),
                }), { dataTestid: getDataTestId(this.dataTestid, 'head-cell-search-icon', String(col.field)) })));
                // filtrelenmiş ise badge ile göster
                if (hasFilter) {
                    _searchIcon = (h("tk-badge", { dot: true, dataTestid: getDataTestId(this.dataTestid, 'head-cell-search-badge', String(col.field)) }, _searchIcon));
                }
            }
            const leftColumns = this.columns.filter(c => c.fixed === 'left');
            const rightColumns = this.columns.filter(c => c.fixed === 'right');
            const isFirstLeft = col.fixed === 'left' && ((_e = leftColumns[0]) === null || _e === void 0 ? void 0 : _e.field) === col.field;
            const isLastRight = col.fixed === 'right' && ((_f = rightColumns[rightColumns.length - 1]) === null || _f === void 0 ? void 0 : _f.field) === col.field;
            const buttonDirection = (_h = (_g = col === null || col === void 0 ? void 0 : col.headerActionsOptions) === null || _g === void 0 ? void 0 : _g.direction) !== null && _h !== void 0 ? _h : 'horizontal';
            return (h("th", { class: classNames(this.getStickyColumnClasses(col, isFirstLeft, isLastRight)), style: Object.assign(Object.assign({}, this.getStickyColumnStyle(col)), col === null || col === void 0 ? void 0 : col.style), "data-field": col.field, "data-testid": getDataTestId(this.dataTestid, 'head-cell', String(col.field)) }, h("div", { class: "tk-table-head-cell", "data-testid": getDataTestId(this.dataTestid, 'head-cell-content', String(col.field)) }, _headerStructure, (col.sortable || col.searchable) && (h("div", { class: classNames('icons', { 'show-icon-on-hover': col.showIconsOnHover && noSortAndFilter && !filterPanelOpen }, buttonDirection), "data-testid": getDataTestId(this.dataTestid, 'head-cell-icons', String(col.field)) }, col.sortable && _sortIcon, _searchIcon)), h("div", { class: "tk-table-resize-handle", onMouseDown: e => this.handleMouseDown(e, this.columns.indexOf(col)), "data-testid": getDataTestId(this.dataTestid, 'head-cell-resize-handle', String(col.field)) }))));
        }))));
    }
    createBody() {
        var _a;
        if (!this.isResizing && !this.isSelectionUpdating) {
            this.customCellCache.clear();
        }
        if (this.renderData.length > 0) {
            return (h("tbody", { "data-testid": getDataTestId(this.dataTestid, 'body') }, h("slot", { name: "body-header" }), this.groupByColumnField ? this.createGroupedRows() : this.renderData.map((row, index) => this.createDataRow(row, index)), h("slot", { name: "body-footer" })));
        }
        else if (this.hasEmptyFilterSlot && ((_a = this.filters) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            return (h("tbody", { "data-testid": getDataTestId(this.dataTestid, 'body-empty-filter') }, h("tr", { "data-testid": getDataTestId(this.dataTestid, 'body-empty-filter-row') }, h("td", { colSpan: 100, "data-testid": getDataTestId(this.dataTestid, 'body-empty-filter-cell') }, h("slot", { name: "empty-filter" })))));
        }
        else if (this.hasEmptyDataSlot) {
            return (h("tbody", { "data-testid": getDataTestId(this.dataTestid, 'body-empty-data') }, h("tr", { "data-testid": getDataTestId(this.dataTestid, 'body-empty-data-row') }, h("td", { colSpan: 100, "data-testid": getDataTestId(this.dataTestid, 'body-empty-data-cell') }, h("slot", { name: "empty-data" })))));
        }
    }
    renderHeader() {
        if (this.cardTitle || this.hasHeaderRightSlot)
            return (h("div", { class: "tk-table-header", "data-testid": getDataTestId(this.dataTestid, 'header') }, h("div", { "data-testid": getDataTestId(this.dataTestid, 'header-title') }, this.cardTitle), h("div", { "data-testid": getDataTestId(this.dataTestid, 'header-right') }, h("slot", { name: "header-right" }))));
    }
    renderTable() {
        return (h("div", { class: "table-holder", "data-testid": getDataTestId(this.dataTestid, 'table-holder') }, h("table", { "data-testid": getDataTestId(this.dataTestid, 'content-table') }, this.createHead(), this.loading ? (h("tbody", { class: "loading-holder", "data-testid": getDataTestId(this.dataTestid, 'loading-holder') }, h("tr", { "data-testid": getDataTestId(this.dataTestid, 'loading-row') }, h("td", { colSpan: this.columns.length + 1, "data-testid": getDataTestId(this.dataTestid, 'loading-cell') }, h("div", { class: "loading", "data-testid": getDataTestId(this.dataTestid, 'loading-content') }, h("svg", { width: "97", height: "97", viewBox: "0 0 97 97", fill: "none", xmlns: "http://www.w3.org/2000/svg", "data-testid": getDataTestId(this.dataTestid, 'loading-icon') }, h("path", { d: "M88.8501 0.5H48.8501C35.5901 0.5 23.5901 5.87 14.9101 14.56C6.2201 23.25 0.850098 35.25 0.850098 48.5C0.850098 75.01 22.3401 96.5 48.8501 96.5C75.3601 96.5 96.8501 75.01 96.8501 48.5V8.5C96.8501 4.09 93.2601 0.5 88.8501 0.5ZM71.4801 71.13C65.6901 76.92 57.6901 80.5 48.8501 80.5C31.1701 80.5 16.8501 66.18 16.8501 48.5C16.8501 39.67 20.4301 31.67 26.2201 25.87C32.0101 20.08 40.0101 16.5 48.8501 16.5C57.6901 16.5 65.6901 20.08 71.4801 25.87C77.2701 31.67 80.8501 39.67 80.8501 48.5C80.8501 57.33 77.2701 65.34 71.4801 71.13Z", fill: "#C92027" }))))))) : (this.createBody()))));
    }
    renderPagination() {
        if (['client', 'server'].includes(this.paginationMethod))
            return (h("tk-pagination", { type: this.paginationType, totalItems: this.totalItems, rowsPerPage: this.internalRowsPerPage, rowsPerPageOptions: this.rowsPerPageOptions, currentPage: this.currentPage, pageReportTemplate: this.pageReportTemplate, itemsReportTemplate: this.itemsReportTemplate, "data-testid": getDataTestId(this.dataTestid, 'pagination'), "onTk-page-change": e => this.handlePageChange(e), "onTk-rows-per-page-change": e => {
                    var _a;
                    this.internalRowsPerPage = e.detail;
                    const tmpData = this.getTableViewData();
                    this.generateRenderData(tmpData, 1);
                    if (this.refSelectAll)
                        this.refSelectAll.value = false;
                    if (((_a = this.selection) === null || _a === void 0 ? void 0 : _a.length) > 0)
                        this.handleSelectAll(false);
                } }));
    }
    render() {
        const rootClasses = classNames('tk-table-container', this.size, {
            striped: this.striped,
        });
        return (h("div", { key: '7bef05efde71fe937cbf8d696be386a7e586c2d6', class: rootClasses, style: this.containerStyle, "data-testid": getDataTestId(this.dataTestid, 'container') }, this.renderHeader(), this.renderTable(), this.renderPagination()));
    }
    static get is() { return "tk-table"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-table.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-table.css"]
        };
    }
    static get properties() {
        return {
            "columns": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ITableColumn[]",
                    "resolved": "ITableColumn[]",
                    "references": {
                        "ITableColumn": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableColumn",
                            "referenceLocation": "ITableColumn"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The column definitions (Array of Objects)"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of container element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "selectionMode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'radio' | 'checkbox'",
                    "resolved": "\"checkbox\" | \"radio\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines how rows can be selected, either with radio buttons (single selection) or checkboxes (multiple selection)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "selection-mode"
            },
            "selection": {
                "type": "any",
                "mutable": true,
                "complexType": {
                    "original": "any[] | any",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "List of the selected"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "selection",
                "defaultValue": "[]"
            },
            "selectionRowDisabled": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Function",
                    "resolved": "Function",
                    "references": {
                        "Function": {
                            "location": "global",
                            "id": "global::Function"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "A function that returns true if the row should be disabled"
                },
                "getter": false,
                "setter": false
            },
            "headerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'dark' | 'primary'",
                    "resolved": "\"basic\" | \"dark\" | \"primary\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Style to apply to header of table"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header-type",
                "defaultValue": "'basic'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'xsmall' | 'small' | 'base'",
                    "resolved": "\"base\" | \"small\" | \"xsmall\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "dataKey": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Property of each row that defines the unique key of each row"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "data-key"
            },
            "data": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "any[]",
                    "resolved": "any[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Rows of data to display"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "cardTitle": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "card-title",
                "defaultValue": "''"
            },
            "striped": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Enables or disables alternating row background colors for easier readability."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "striped",
                "defaultValue": "false"
            },
            "rowsPerPage": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Number of items per page."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rows-per-page",
                "defaultValue": "6"
            },
            "rowsPerPageOptions": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "number[]",
                    "resolved": "number[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Number of rows per page options"
                },
                "getter": false,
                "setter": false
            },
            "totalItems": {
                "type": "number",
                "mutable": true,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Number of total items."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "total-items"
            },
            "paginationMethod": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines whether pagination is handled on the client or server side."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "pagination-method"
            },
            "paginationType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'outlined' | 'text' | 'grouped'",
                    "resolved": "\"grouped\" | \"outlined\" | \"text\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'outlined'"
                        }],
                    "text": "The type of the pagination"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "pagination-type",
                "defaultValue": "'outlined'"
            },
            "pageReportTemplate": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'page: {currentPage} of {totalPages}'"
                        }],
                    "text": "Template string for current page report in pagination.\nAvailable placeholders: {currentPage}, {totalPages}"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "page-report-template",
                "defaultValue": "'page: {currentPage} of {totalPages}'"
            },
            "itemsReportTemplate": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'item: {startItem}-{endItem} of {totalItems}'"
                        }],
                    "text": "Template string for items report in pagination.\nAvailable placeholders: {startItem}, {endItem}, {totalItems}"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "items-report-template",
                "defaultValue": "'item: {startItem}-{endItem} of {totalItems}'"
            },
            "loading": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a loading indicator while data is being fetched or processed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "loading"
            },
            "expandedRows": {
                "type": "unknown",
                "mutable": true,
                "complexType": {
                    "original": "any[]",
                    "resolved": "any[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Specifies which rows are expanded to show additional content."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "multiSort": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Enables multi-column sorting."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "multi-sort",
                "defaultValue": "false"
            },
            "cellStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "(row: any, column: ITableColumn) => any",
                    "resolved": "(row: any, column: ITableColumn) => any",
                    "references": {
                        "ITableColumn": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableColumn",
                            "referenceLocation": "ITableColumn"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provides a function to customize cell styles.\nThis function takes the row and column information and returns the style object for a specific cell."
                },
                "getter": false,
                "setter": false
            },
            "rowStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "(row: any, index?: number) => any",
                    "resolved": "(row: any, index?: number) => any",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provides a function to customize row styles.\nThis function takes row information and row index, and returns the style object for a specific  row."
                },
                "getter": false,
                "setter": false
            },
            "expandedRowStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "(row: any) => any",
                    "resolved": "(row: any) => any",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provides a function to customize expanded row styles.\nThis function takes row information and returns the style object for the expanded row content."
                },
                "getter": false,
                "setter": false
            },
            "groupBy": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "example",
                            "text": "groupBy=\"status\" // Groups by the 'status' column"
                        }],
                    "text": "Column field name to group the table data by.\nWhen specified, the table will automatically group rows by unique values in this column.\nSet to null or undefined to disable grouping.\nThis makes the component controlled - changes should be handled via tkGroupByChange event."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "group-by"
            },
            "collapsibleGroups": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If true, group headers will have an expand/collapse button to show/hide the rows in that group."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "collapsible-groups",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "elActiveSearchIcon": {},
            "selectColumn": {},
            "sortField": {},
            "sortOrder": {},
            "filters": {},
            "currentPage": {},
            "renderData": {},
            "hasHeaderRightSlot": {},
            "hasEmptyDataSlot": {},
            "hasEmptyFilterSlot": {},
            "isFilterOpen": {},
            "columnWidths": {},
            "stickyOffsets": {},
            "sorts": {},
            "groupedData": {},
            "groupByColumnField": {},
            "isControlledGrouping": {},
            "expandedGroups": {},
            "internalRowsPerPage": {}
        };
    }
    static get events() {
        return [{
                "method": "tkSelectionChange",
                "name": "tk-selection-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "any[] | any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "tkRequest",
                "name": "tk-request",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a request needs to be made to the server."
                },
                "complexType": {
                    "original": "ITableRequest",
                    "resolved": "ITableRequest",
                    "references": {
                        "ITableRequest": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableRequest",
                            "referenceLocation": "ITableRequest"
                        }
                    }
                }
            }, {
                "method": "tkExpandedRowsChange",
                "name": "tk-expanded-rows-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the expanded rows change."
                },
                "complexType": {
                    "original": "any[]",
                    "resolved": "any[]",
                    "references": {}
                }
            }, {
                "method": "tkCellEdit",
                "name": "tk-cell-edit",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a cell is edited."
                },
                "complexType": {
                    "original": "ITableCellEdit",
                    "resolved": "ITableCellEdit",
                    "references": {
                        "ITableCellEdit": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableCellEdit",
                            "referenceLocation": "ITableCellEdit"
                        }
                    }
                }
            }, {
                "method": "tkRowClick",
                "name": "tk-row-click",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "param",
                            "text": "row The row data that was clicked"
                        }],
                    "text": "Emitted when a row is clicked."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "tkGroupByChange",
                "name": "tk-group-by-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "param",
                            "text": "groupBy The new groupBy field name (null if grouping is cleared)"
                        }],
                    "text": "Emitted when the groupBy value changes.\nAlways emitted for both controlled and uncontrolled components.\nFor controlled components, handle this event to update the groupBy prop."
                },
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "serverRequest": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "ITableRequest": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableRequest",
                            "referenceLocation": "ITableRequest"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Allows tk-request event to be triggered manually",
                    "tags": []
                }
            },
            "exportFile": {
                "complexType": {
                    "signature": "(options: ITableExportOptions) => Promise<void>",
                    "parameters": [{
                            "name": "options",
                            "type": "ITableExportOptions",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "ITableExportOptions": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableExportOptions",
                            "referenceLocation": "ITableExportOptions"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Exports the table data to a file",
                    "tags": [{
                            "name": "param",
                            "text": "options"
                        }]
                }
            },
            "clearFilters": {
                "complexType": {
                    "signature": "(columns?: string[]) => Promise<void>",
                    "parameters": [{
                            "name": "columns",
                            "type": "string[]",
                            "docs": "Optional array of column field names to clear filters for. If not provided, all filters are cleared."
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Clears all filters or specific column filters",
                    "tags": [{
                            "name": "param",
                            "text": "columns Optional array of column field names to clear filters for. If not provided, all filters are cleared."
                        }]
                }
            },
            "clearSorting": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLTkIconElement": {
                            "location": "global",
                            "id": "global::HTMLTkIconElement"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Clears all sorting for server side pagination",
                    "tags": []
                }
            },
            "getFilters": {
                "complexType": {
                    "signature": "() => Promise<ITableFilter[]>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "ITableFilter": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableFilter",
                            "referenceLocation": "ITableFilter"
                        }
                    },
                    "return": "Promise<ITableFilter[]>"
                },
                "docs": {
                    "text": "Returns the current filters",
                    "tags": [{
                            "name": "returns",
                            "text": "The current filters"
                        }]
                }
            },
            "getSorting": {
                "complexType": {
                    "signature": "() => Promise<ITableSort[] | { field: string; order: \"desc\" | \"asc\"; }>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "ITableSort": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableSort",
                            "referenceLocation": "ITableSort"
                        }
                    },
                    "return": "Promise<ITableSort[] | { field: string; order: \"desc\" | \"asc\"; }>"
                },
                "docs": {
                    "text": "Returns the current sorting settings",
                    "tags": [{
                            "name": "returns",
                            "text": "The current sorting settings"
                        }]
                }
            },
            "setCurrentPage": {
                "complexType": {
                    "signature": "(page: number) => Promise<void>",
                    "parameters": [{
                            "name": "page",
                            "type": "number",
                            "docs": "The page number to set (1-based index)"
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Sets the current page for pagination",
                    "tags": [{
                            "name": "param",
                            "text": "page The page number to set (1-based index)"
                        }]
                }
            },
            "setFilters": {
                "complexType": {
                    "signature": "(filters: ITableFilter[]) => Promise<void>",
                    "parameters": [{
                            "name": "filters",
                            "type": "ITableFilter[]",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "ITableFilter": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableFilter",
                            "referenceLocation": "ITableFilter"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Sets the current filter settings",
                    "tags": []
                }
            },
            "setSorting": {
                "complexType": {
                    "signature": "(sorts: ITableSort[] | { field: string; order: \"asc\" | \"desc\"; }) => Promise<void>",
                    "parameters": [{
                            "name": "sorts",
                            "type": "ITableSort[] | { field: string; order: \"desc\" | \"asc\"; }",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "ITableSort": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-table/types.ts::ITableSort",
                            "referenceLocation": "ITableSort"
                        },
                        "HTMLTkIconElement": {
                            "location": "global",
                            "id": "global::HTMLTkIconElement"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Sets the current sorting settings",
                    "tags": []
                }
            },
            "runFilters": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Applies the current filters to the data for client side pagination",
                    "tags": []
                }
            },
            "groupByColumn": {
                "complexType": {
                    "signature": "(columnField: string) => Promise<void>",
                    "parameters": [{
                            "name": "columnField",
                            "type": "string",
                            "docs": "The field name to group by (e.g., 'status', 'category', 'department')"
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Groups table data by the specified column field\nCreates group header rows that display the unique value and count of items in that group.\nFor example, if you have a 'status' column with values 'Open' and 'Closed',\nthis will create group headers like \"Open (5)\" and \"Closed (3)\".\n\nAlways emits tkGroupByChange event. For uncontrolled components, also updates internal state.",
                    "tags": [{
                            "name": "param",
                            "text": "columnField The field name to group by (e.g., 'status', 'category', 'department')"
                        }, {
                            "name": "example",
                            "text": "// Controlled usage (with groupBy prop)\nawait tableRef.groupByColumn('status'); // Emits tkGroupByChange event\n\n// Uncontrolled usage (no groupBy prop)\nawait tableRef.groupByColumn('status'); // Updates internal state and emits tkGroupByChange event"
                        }]
                }
            },
            "clearGrouping": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Clears the current grouping and returns to normal table view\n\nAlways emits tkGroupByChange event with null value. For uncontrolled components, also clears internal state.",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "columns",
                "methodName": "columnsChanged"
            }, {
                "propName": "data",
                "methodName": "dataChanged"
            }, {
                "propName": "rowsPerPage",
                "methodName": "rowsPerPageChanged"
            }, {
                "propName": "groupBy",
                "methodName": "groupByChanged"
            }];
    }
    static get listeners() {
        return [{
                "name": "click",
                "method": "checkForClickOutside",
                "target": "window",
                "capture": false,
                "passive": false
            }];
    }
}
