// @takeoff-ui/core@0.11.9/dist/collection/components/tk-timeline/tk-timeline.js
import { h, Fragment } from "@stencil/core";
import classNames from "classnames";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkTimeline` is a component that displays a vertical or horizontal timeline of events.
 * The `TkTimelineItem` is a helper component used to create customized content within the `TkTimeline` component.
 * @react `import { TkTimeline, TkTimelineItem } from '@takeoff-ui/react'`
 * @vue `import { TkTimeline, TkTimelineItem } from '@takeoff-ui/vue'`
 * @angular `import { TkTimeline, TkTimelineItem } from '@takeoff-ui/angular'`
 */
export class TkTimeline {
    constructor() {
        this.slottedItemsCount = 0;
        /**
         * An array of objects representing the items to display on the timeline.
         * Each object should have at least a `title`. `description` and `date` are optional.
         */
        this.items = [];
        /**
         * The orientation of the timeline.
         */
        this.orientation = 'horizontal';
        /**
         * Whether to alternate the position of timeline items relative to the line.
         */
        this.alternate = true;
    }
    connectedCallback() {
        // MutationObserver ile DOM değişikliklerini izle
        if (typeof window !== 'undefined') {
            this.mutationObserver = new MutationObserver(() => {
                this.updateSlottedItemsCount();
            });
            this.mutationObserver.observe(this.el, {
                childList: true,
                subtree: true,
            });
        }
    }
    componentDidLoad() {
        this.updateSlottedItemsCount();
    }
    disconnectedCallback() {
        if (this.mutationObserver) {
            this.mutationObserver.disconnect();
        }
    }
    updateSlottedItemsCount() {
        const newCount = this.el.querySelectorAll('tk-timeline-item').length;
        if (newCount !== this.slottedItemsCount) {
            this.slottedItemsCount = newCount;
        }
    }
    determineContentPlacement(index) {
        if (!this.alternate) {
            return 'end';
        }
        const isEvenItem = index % 2 === 0;
        return isEvenItem ? 'end' : 'start';
    }
    createItemContent(item, index) {
        return (h(Fragment, null, h("div", { class: "tk-timeline-item-content-inner", "data-testid": getDataTestId(this.dataTestid, 'item-content', index.toString()) }, item.title && (h("div", { class: "tk-timeline-item-title", "data-testid": getDataTestId(this.dataTestid, 'item-title', index.toString()) }, item.title)), item.description && (h("div", { class: "tk-timeline-item-description", "data-testid": getDataTestId(this.dataTestid, 'item-description', index.toString()) }, item.description))), item.date && (h("div", { class: "tk-timeline-item-date", "data-testid": getDataTestId(this.dataTestid, 'item-date', index.toString()) }, item.date))));
    }
    renderTimelineItem(item, index) {
        const contentPlacement = this.determineContentPlacement(index);
        const isFirst = index === 0;
        const isLast = index === this.items.length - 1;
        const itemClasses = classNames('tk-timeline-item', `tk-timeline-item-placement-${contentPlacement}`, {
            'tk-timeline-item-first': isFirst,
            'tk-timeline-item-last': isLast,
            'tk-timeline-item-alternate': this.alternate,
        });
        return (h("li", { class: itemClasses, "data-testid": getDataTestId(this.dataTestid, 'item', index.toString()) }, this.alternate && (h("div", { class: "tk-timeline-item-content tk-timeline-item-content-start", "data-testid": getDataTestId(this.dataTestid, 'item-start', index.toString()) }, contentPlacement === 'start' && this.createItemContent(item, index))), h("div", { class: "tk-timeline-item-separator", "data-testid": getDataTestId(this.dataTestid, 'item-separator', index.toString()) }, h("div", { class: "tk-timeline-item-point", "data-testid": getDataTestId(this.dataTestid, 'item-point', index.toString()) }), h("div", { class: "tk-timeline-item-connector", "data-testid": getDataTestId(this.dataTestid, 'item-connector', index.toString()) })), h("div", { class: "tk-timeline-item-content tk-timeline-item-content-end", "data-testid": getDataTestId(this.dataTestid, 'item-end', index.toString()) }, contentPlacement === 'end' && this.createItemContent(item, index))));
    }
    renderSlottedTimelineItems() {
        const timelineItems = Array.from(this.el.querySelectorAll('tk-timeline-item'));
        return timelineItems.map((item, index) => {
            const contentPlacement = this.determineContentPlacement(index);
            const isFirst = index === 0;
            const isLast = index === timelineItems.length - 1;
            const itemClasses = classNames('tk-timeline-item', `tk-timeline-item-placement-${contentPlacement}`, {
                'tk-timeline-item-first': isFirst,
                'tk-timeline-item-last': isLast,
                'tk-timeline-item-alternate': this.alternate,
            });
            // Set attributes on the timeline item for styling purposes
            item.setAttribute('data-class', `tk-timeline-item-${index}`);
            item.setAttribute('data-index', index.toString());
            return (h("li", { class: itemClasses, "data-testid": getDataTestId(this.dataTestid, 'item', index.toString()) }, this.alternate && (h("div", { class: "tk-timeline-item-content tk-timeline-item-content-start", "data-testid": getDataTestId(this.dataTestid, 'item-start', index.toString()) }, contentPlacement === 'start' && h("slot", { name: `item-${index}` }))), h("div", { class: "tk-timeline-item-separator", "data-testid": getDataTestId(this.dataTestid, 'item-separator', index.toString()) }, h("div", { class: "tk-timeline-item-point", "data-testid": getDataTestId(this.dataTestid, 'item-point', index.toString()) }), h("div", { class: "tk-timeline-item-connector", "data-testid": getDataTestId(this.dataTestid, 'item-connector', index.toString()) })), h("div", { class: "tk-timeline-item-content tk-timeline-item-content-end", "data-testid": getDataTestId(this.dataTestid, 'item-end', index.toString()) }, contentPlacement === 'end' && h("slot", { name: `item-${index}` }))));
        });
    }
    render() {
        const hostClasses = classNames('tk-timeline', `tk-timeline-${this.orientation}`);
        const hasSlottedItems = this.slottedItemsCount > 0;
        return (h("div", { key: 'e06fbcdfa4a9bd0d2231dbe618e3df8f0842811a', class: hostClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("ul", { key: '89eba65c61f411ae349ee9f9a29a300e6a717dc9', class: "tk-timeline-items", "data-testid": getDataTestId(this.dataTestid, 'items') }, hasSlottedItems ? h(Fragment, null, this.renderSlottedTimelineItems()) : this.items.map((item, index) => this.renderTimelineItem(item, index)))));
    }
    static get is() { return "tk-timeline"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-timeline.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-timeline.css"]
        };
    }
    static get properties() {
        return {
            "items": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "TimelineItem[]",
                    "resolved": "TimelineItem[]",
                    "references": {
                        "TimelineItem": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-timeline/types.ts::TimelineItem",
                            "referenceLocation": "TimelineItem"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "An array of objects representing the items to display on the timeline.\nEach object should have at least a `title`. `description` and `date` are optional."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "orientation": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'horizontal' | 'vertical'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The orientation of the timeline."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "orientation",
                "defaultValue": "'horizontal'"
            },
            "alternate": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether to alternate the position of timeline items relative to the line."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "alternate",
                "defaultValue": "true"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "slottedItemsCount": {}
        };
    }
    static get elementRef() { return "el"; }
}
