// @takeoff-ui/core@0.11.9/dist/collection/utils/position-utils.js
import { computePosition, offset, flip, shift, arrow, autoUpdate, size, hide } from "@floating-ui/dom";
import { applyStyles } from "./style-utils";
function positionFloatingElement(triggerElement, floatingElement, arrowElement, options) {
    const { placement, offset: off = 8, size: sizeOptions } = options || {};
    if (arrowElement) {
        applyStyles(arrowElement, {
            position: 'absolute',
            width: '9px',
            height: '9px',
            background: 'inherit',
            border: '1px solid',
            borderColor: 'inherit',
            transform: 'rotate(45deg)',
            zIndex: '1300',
        });
    }
    const middleware = [
        offset(off),
        flip(),
        shift({ padding: 5 }),
        ...(arrowElement ? [arrow({ element: arrowElement, padding: 8 })] : []),
        ...(sizeOptions ? [size(sizeOptions)] : []),
        hide(),
    ];
    return computePosition(triggerElement, floatingElement, {
        strategy: 'fixed',
        placement,
        middleware,
    }).then(({ x, y, middlewareData, placement }) => {
        floatingElement.style.left = `${x}px`;
        floatingElement.style.top = `${y}px`;
        const side = placement.split('-')[0];
        const reverseSide = { top: 'bottom', right: 'left', bottom: 'top', left: 'right' }[side];
        if (arrowElement && middlewareData.arrow) {
            const { x: ax, y: ay } = middlewareData.arrow;
            arrowElement.dataset.side = side;
            applyStyles(arrowElement, {
                left: ax != null ? `${ax}px` : '',
                top: ay != null ? `${ay}px` : '',
                right: '',
                bottom: '',
                [reverseSide]: '-5.7px',
            });
            const borderStyles = {
                top: { borderTop: 'none', borderLeft: 'none' },
                bottom: { borderBottom: 'none', borderRight: 'none' },
                left: { borderLeft: 'none', borderBottom: 'none' },
                right: { borderRight: 'none', borderTop: 'none' },
            };
            applyStyles(arrowElement, borderStyles[side]);
        }
        if (middlewareData.hide) {
            if (middlewareData.hide.referenceHidden) {
                floatingElement.classList.add('floating-hidden');
            }
            else {
                floatingElement.classList.remove('floating-hidden');
            }
        }
        return placement;
    });
}
export function floatingElementAutoUpdate(triggerElement, floatingElement, arrowElement, options, handlePlacement) {
    return autoUpdate(triggerElement, floatingElement, () => {
        positionFloatingElement(triggerElement, floatingElement, arrowElement, options).then(position => handlePlacement === null || handlePlacement === void 0 ? void 0 : handlePlacement(position));
    }, { animationFrame: true });
}
