// @takeoff-ui/core@0.11.9/dist/collection/components/tk-datepicker/tk-datepicker.js
import { h, Fragment } from "@stencil/core";
import { format, parse, isValid } from "date-fns";
import { v4 as uuidv4 } from "uuid";
import classNames from "classnames";
import { ClickOutsideMixin } from "../../utils/clickoutside-mixin";
import { floatingElementAutoUpdate } from "../../utils/position-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkDatepicker` component is a versatile and customizable date picker that supports `single` date and date `range` selection. It offers various display modes, localization, and customizable date formatting.
 * @slot footer - Custom footer template.
 * @slot footer-actions - Custom actions template to default footer.
 * @react `import { TkDatepicker } from '@takeoff-ui/react'`
 * @vue `import { TkDatepicker } from '@takeoff-ui/vue'`
 * @angular `import { TkDatepicker } from '@takeoff-ui/angular'`
 */
export class TkDatePicker {
    constructor() {
        this.today = new Date();
        this.uniqueId = uuidv4();
        this.isUpdatingAmPm = false;
        this.weeksLength = 0;
        this.currentMonth = new Date();
        this.internalSelectedDates = { start: null, end: null };
        this.inputValue = '';
        this.internalStartTime = null;
        this.internalEndTime = null;
        this.internalAmPm = 'AM';
        this.hoverDate = null;
        this.currentView = 'days';
        this.isInvalid = false;
        this.isOpen = false;
        this.concealUntilMeasured = false;
        /**
         * Defines the size for the label
         * @defaultValue base
         */
        this.size = 'base';
        /**
         * Whether the datepicker is disabled
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * Whether the datepicker is in an invalid state
         */
        this.invalid = false;
        /**
         * Whether the datepicker is read-only
         */
        this.readonly = false;
        /**
         * Indicates whether the input of datepicker can be cleared
         * @defaultValue false
         */
        this.clearable = false;
        /**
         * Represents whether the datepicker is in a loading state.
         * If true, renders a spinner inside the input and panel.
         * @defaultValue false
         */
        this.loading = false;
        /**
         * Specifies a material icon name to be displayed.
         */
        this.icon = 'calendar_month';
        /**
         * Defines the position of the icon.
         * @defaultValue left
         */
        this.iconPosition = 'left';
        /**
         * Minimum selectable date
         */
        this.minDate = '';
        /**
         * Maximum selectable date
         */
        this.maxDate = '';
        /**
         * Array of dates that are allowed to be selected. All other dates will be disabled.
         *
         * Note: Format should match dateFormat prop
         */
        this.allowedDates = [];
        /**
         * Array of dates that are disabled for selection.
         *
         * Format should match dateFormat prop
         */
        this.disabledDates = [];
        /**
         * Whether to display inline panel
         * @defaultValue false
         */
        this.inline = false;
        /**
         * Whether to require manual confirmation (Apply button) before committing changes.
         * If true, changes are only applied when apply() is called.
         * @defaultValue false
         */
        this.allowApplyButton = false;
        /**
         * The selection mode of the date picker: 'single' for single date selection, 'range' for date range selection.
         * @defaultValue single
         */
        this.mode = 'single';
        /**
         * Locale for date formatting
         * @defaultValue en
         */
        this.locale = 'en';
        /**
         * The visual variant of the footer: 'basic', 'divided', or 'light'.
         * @defaultValue basic
         */
        this.footerType = 'basic';
        /**
         * Date format pattern
         * @defaultValue yyyy-MM-dd
         */
        this.dateFormat = 'yyyy-MM-dd';
        /**
         * Whether to disable input mask
         * @defaultValue false
         */
        this.disableMask = false;
        /**
         * Header visual variant
         * @defaultValue basic
         */
        this.headerType = 'basic';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * Disabled week days (0-6, where 0 is Sunday and 6 is Saturday)
         * Example: [0,6] will disable Sunday and Saturday
         */
        this.disabledWeekDays = [];
        /**
         * Whether to show the timepicker panel next to the calendar.
         * @defaultValue false
         */
        this.showTimePicker = false;
        /**
         * Enables time-only mode. In this mode, no date selection is required and the input shows a time mask.
         * When enabled, the panel renders only the time picker and `tk-change` emits a time string (e.g. HH:mm or hh:mm a).
         * @defaultValue false
         */
        this.timeOnly = false;
        /**
         * Hour increment step.
         * @defaultValue 1
         */
        this.hourStep = 1;
        /**
         * Minute increment step.
         * @defaultValue 1
         */
        this.minuteStep = 1;
        /**
         * Time format: '12' or '24'.
         * @defaultValue '24'
         */
        this.timeFormat = '24';
        this.remeasureCalendarOnNextFrame = () => {
            if (!(this.showTimePicker && (this.inline || this.isOpen)))
                return;
            requestAnimationFrame(() => {
                const h = this.measureCalendarTableHeight();
                if (h > 0)
                    this.calendarTableHeightPx = h;
            });
        };
        /**
         * Click outside handler implementation - called by the mixin
         */
        this.closeHandler = () => {
            if (this.inline)
                return;
            this.isOpen = false;
        };
        this.handleApply = () => {
            const { start, end } = this.internalSelectedDates;
            let emitValue = null;
            if (this.timeOnly) {
                if (this.internalStartTime) {
                    const tempDate = new Date();
                    tempDate.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
                    emitValue = format(tempDate, this.getOnlyTimeFormat());
                }
            }
            else {
                const formattedStart = start ? this.formatDateOrDateTime(start, 'start') : null;
                if (this.mode === 'range') {
                    emitValue = {
                        start: formattedStart || '',
                        end: end ? this.formatDateOrDateTime(end, 'end') : undefined,
                    };
                }
                else {
                    emitValue = formattedStart;
                }
            }
            this.inputValue = this.formatInputValue();
            this.value = emitValue;
            this.tkChange.emit(emitValue);
            const isEndSelected = this.mode === 'range' ? !!this.internalSelectedDates.end : true;
            if (!this.inline && this.isOpen && !this.showTimePicker && isEndSelected) {
                this.isOpen = false;
            }
            this.isInvalid = false;
            this.currentView = 'days';
        };
        this.handleIncreaseHour = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (this.timeFormat === '12') {
                const hoursList = Array.from({ length: Math.ceil(12 / this.hourStep) }, (_, i) => i * this.hourStep + 1);
                let displayHour = targetTimeState.time.hour % 12;
                displayHour = displayHour === 0 ? 12 : displayHour;
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - displayHour) < Math.abs(prev - displayHour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const nextIdx = Math.min(idx + 1, hoursList.length - 1);
                const newDisplayHour = hoursList[nextIdx];
                // Convert back to 24-hour format
                let newHour24 = newDisplayHour === 12 ? 0 : newDisplayHour;
                if (this.internalAmPm === 'PM') {
                    newHour24 += 12;
                }
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
            }
            else {
                // 24h mode - move to next hour in step list
                const hoursList = Array.from({ length: Math.ceil(24 / this.hourStep) }, (_, i) => i * this.hourStep);
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.hour) < Math.abs(prev - targetTimeState.time.hour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const nextIdx = Math.min(idx + 1, hoursList.length - 1);
                const hour = hoursList[nextIdx];
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
            }
            this.emitTimeChange();
        };
        this.handleDecreaseHour = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (this.timeFormat === '12') {
                const hoursList = Array.from({ length: Math.ceil(12 / this.hourStep) }, (_, i) => i * this.hourStep + 1);
                let displayHour = targetTimeState.time.hour % 12;
                displayHour = displayHour === 0 ? 12 : displayHour;
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - displayHour) < Math.abs(prev - displayHour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const prevIdx = Math.max(idx - 1, 0);
                const newDisplayHour = hoursList[prevIdx];
                // Convert back to 24-hour format
                let newHour24 = newDisplayHour === 12 ? 0 : newDisplayHour;
                if (this.internalAmPm === 'PM') {
                    newHour24 += 12;
                }
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: newHour24 });
                }
            }
            else {
                // 24h mode - move to previous hour in step list
                const hoursList = Array.from({ length: Math.ceil(24 / this.hourStep) }, (_, i) => i * this.hourStep);
                // Find closest hour in step list
                const closestHour = hoursList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.hour) < Math.abs(prev - targetTimeState.time.hour) ? curr : prev));
                const idx = hoursList.indexOf(closestHour);
                const prevIdx = Math.max(idx - 1, 0);
                const hour = hoursList[prevIdx];
                if (targetTimeState.type === 'start') {
                    this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
                else {
                    this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
                }
            }
            this.emitTimeChange();
        };
        this.handleHourClick = (hour) => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: hour });
            }
            this.emitTimeChange();
        };
        this.handleIncreaseMinute = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            const minutesList = Array.from({ length: Math.ceil(60 / this.minuteStep) }, (_, i) => i * this.minuteStep);
            // Find closest minute in step list
            const closestMinute = minutesList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.minute) < Math.abs(prev - targetTimeState.time.minute) ? curr : prev));
            const idx = minutesList.indexOf(closestMinute);
            const nextIdx = Math.min(idx + 1, minutesList.length - 1);
            const minute = minutesList[nextIdx];
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            this.emitTimeChange();
        };
        this.handleDecreaseMinute = () => {
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            const minutesList = Array.from({ length: Math.ceil(60 / this.minuteStep) }, (_, i) => i * this.minuteStep);
            // Find closest minute in step list
            const closestMinute = minutesList.reduce((prev, curr) => (Math.abs(curr - targetTimeState.time.minute) < Math.abs(prev - targetTimeState.time.minute) ? curr : prev));
            const idx = minutesList.indexOf(closestMinute);
            const prevIdx = Math.max(idx - 1, 0);
            const minute = minutesList[prevIdx];
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: minute });
            }
            this.emitTimeChange();
        };
        this.handleMinuteClick = (min) => {
            if (this.disabled || this.readonly)
                return;
            const targetTimeState = this.getTimeStateToModify();
            if (!targetTimeState)
                return;
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: min });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { minute: min });
            }
            this.emitTimeChange();
        };
        this.handleAmPmToggle = (event) => {
            // Prevent inner toggle group's tk-change from leaking outside and overwriting consumer value
            if (this.disabled || this.readonly)
                return;
            event.stopPropagation();
            if (this.timeFormat !== '12')
                return;
            const newAmPm = event.detail;
            // Just update the state - the watcher will handle the time conversion
            this.internalAmPm = newAmPm;
        };
        this.handleInputBlur = () => {
            if ((this.timeOnly || this.showTimePicker) && this.timeFormat === '12' && this.internalStartTime) {
                this.internalAmPm = this.internalStartTime.hour >= 12 ? 'PM' : 'AM';
                this.inputValue = this.formatInputValue();
            }
        };
        this.handleInputClick = (e) => {
            if (this.disabled || this.readonly) {
                e.preventDefault();
                return;
            }
            this.isOpen = !this.isOpen;
            this.remeasureCalendarOnNextFrame();
        };
        this.handleDateClick = (date) => {
            if (this.disabled || this.readonly || this.isDateDisabled(date))
                return;
            const normalizedDate = this.normalizeDate(date);
            const defaultTime = this.getDefaultTime();
            if (this.mode === 'single') {
                this.internalSelectedDates = { start: normalizedDate, end: null };
                if (this.showTimePicker) {
                    if (!this.internalStartTime) {
                        this.internalStartTime = defaultTime;
                        if (this.timeFormat === '12')
                            this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
                    }
                }
                else {
                    this.internalStartTime = null;
                }
                this.internalEndTime = this.internalStartTime;
                if (!this.inline && !this.showTimePicker && !this.allowApplyButton) {
                    this.isOpen = false;
                }
            }
            else if (this.mode === 'range') {
                const { start, end } = this.internalSelectedDates;
                if (!start || (start && end)) {
                    this.internalSelectedDates = { start: normalizedDate, end: null };
                    if (this.showTimePicker) {
                        if (!this.internalStartTime) {
                            this.internalStartTime = defaultTime;
                            if (this.timeFormat === '12')
                                this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
                        }
                        this.internalEndTime = null;
                    }
                    else {
                        this.internalStartTime = null;
                        this.internalEndTime = null;
                    }
                    this.hoverDate = null;
                }
                else {
                    let newStart;
                    let newEnd;
                    if (normalizedDate < start) {
                        newStart = normalizedDate;
                        newEnd = start;
                        if (this.showTimePicker) {
                            const tempTime = this.internalStartTime;
                            this.internalStartTime = this.internalEndTime || this.internalStartTime || defaultTime;
                            this.internalEndTime = tempTime || defaultTime;
                        }
                    }
                    else {
                        newStart = start;
                        newEnd = normalizedDate;
                        if (this.showTimePicker && !this.internalEndTime) {
                            this.internalEndTime = this.internalStartTime || defaultTime;
                        }
                    }
                    this.internalSelectedDates = { start: newStart, end: newEnd };
                    this.hoverDate = null;
                    if (!this.inline && !this.showTimePicker && !this.allowApplyButton) {
                        this.isOpen = false;
                    }
                }
            }
            // if swap occurred, bail
            if (!this.ensureRangeOrder()) {
                return;
            }
            this.remeasureCalendarOnNextFrame();
            if (!this.allowApplyButton) {
                this.handleApply();
            }
            this.isInvalid = false;
        };
        this.handleInputKeyDown = (event) => {
            if (this.timeOnly)
                return; // let time mask/type
            if (this.disableMask || this.mode === 'range') {
                event.preventDefault();
            }
        };
        this.handleInputChange = (event) => {
            if (this.disableMask || this.mode === 'range') {
                event.preventDefault();
                return;
            }
            this.remeasureCalendarOnNextFrame();
            this.inputValue = event.detail;
            this.tkInputChange.emit(this.inputValue);
        };
        this.handleInputClearClick = () => {
            if (this.clearable) {
                this.inputValue = '';
                this.internalSelectedDates = { start: null, end: null };
                this.internalStartTime = null;
                this.internalEndTime = null;
                this.tkChange.emit(null);
            }
        };
        this.handleInputInput = () => {
            if (this.disableMask || this.mode === 'range')
                return;
            clearTimeout(this.debounceTimer);
            this.debounceTimer = window.setTimeout(() => {
                if (this.inputValue) {
                    if (this.timeOnly) {
                        const parsedTime = this.parseTimeString(this.inputValue);
                        if (parsedTime) {
                            // min/maxTime validation
                            if (this.isTimeDisabled(parsedTime.getHours(), parsedTime.getMinutes())) {
                                this.isInvalid = true;
                                this.tkChange.emit(undefined);
                                return;
                            }
                            this.internalStartTime = { hour: parsedTime.getHours(), minute: parsedTime.getMinutes() };
                            this.internalEndTime = this.internalStartTime;
                            this.isInvalid = false;
                            if (!this.allowApplyButton) {
                                this.tkChange.emit(format(parsedTime, this.getOnlyTimeFormat()));
                            }
                        }
                        else if (!this.allowApplyButton) {
                            this.isInvalid = true;
                            this.tkInvalid.emit({ message: 'Invalid time format', value: this.inputValue });
                            this.tkChange.emit(undefined);
                        }
                    }
                    else {
                        const parser = this.showTimePicker ? this.parseFullDateTime.bind(this) : this.parseInputDate.bind(this);
                        const parsedDate = parser(this.inputValue);
                        if (parsedDate && !this.isDateDisabled(parsedDate)) {
                            // min/maxTime validation for timepicker mode
                            if (this.showTimePicker && this.isTimeDisabled(parsedDate.getHours(), parsedDate.getMinutes())) {
                                this.isInvalid = true;
                                this.tkChange.emit(undefined);
                                return;
                            }
                            const normalized = this.normalizeDate(parsedDate);
                            this.internalSelectedDates = {
                                start: normalized,
                                end: null,
                            };
                            let formattedValue;
                            if (this.showTimePicker) {
                                const time = { hour: parsedDate.getHours(), minute: parsedDate.getMinutes() };
                                this.internalStartTime = time;
                                this.internalEndTime = time;
                                formattedValue = format(parsedDate, this.getFullDateTimeFormat());
                            }
                            else {
                                this.internalStartTime = null;
                                this.internalEndTime = null;
                                formattedValue = this.formatDateOrDateTime(parsedDate, 'start');
                            }
                            this.isInvalid = false;
                            this.tkChange.emit(formattedValue);
                        }
                        else if (!this.allowApplyButton) {
                            this.isInvalid = true;
                            this.tkInvalid.emit({ message: 'Invalid date format', value: this.inputValue });
                        }
                    }
                }
                else {
                    this.isInvalid = false;
                    this.internalSelectedDates = { start: null, end: null };
                    this.internalStartTime = null;
                    this.internalEndTime = null;
                    if (!this.allowApplyButton) {
                        this.tkChange.emit(undefined);
                    }
                }
            }, 300);
        };
        this.handleDateHover = (date) => {
            if (this.mode === 'range' && this.internalSelectedDates.start && !this.internalSelectedDates.end) {
                this.hoverDate = date;
            }
            this.remeasureCalendarOnNextFrame();
        };
        this.handleMonthChange = (increment) => {
            this.currentMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() + increment);
            this.remeasureCalendarOnNextFrame();
        };
        this.handleYearChange = (increment) => {
            if (this.currentView === 'years') {
                this.currentMonth = new Date(this.currentMonth.getFullYear() + increment * 12, this.currentMonth.getMonth());
                return;
            }
            this.currentMonth = new Date(this.currentMonth.getFullYear() + increment, this.currentMonth.getMonth());
            this.remeasureCalendarOnNextFrame();
        };
        this.handleViewChange = (e, view) => {
            e.stopPropagation();
            this.currentView = view;
            this.remeasureCalendarOnNextFrame();
        };
    }
    internalAmPmChanged(newValue) {
        this.updateTimeBasedOnAmPm(newValue);
    }
    isOpenChanged(newValue) {
        if (!this.inline) {
            if (newValue) {
                if (this.internalSelectedDates.start) {
                    this.currentMonth = new Date(this.internalSelectedDates.start.getFullYear(), this.internalSelectedDates.start.getMonth());
                }
                if (this.showTimePicker) {
                    this.concealUntilMeasured = true;
                    requestAnimationFrame(() => {
                        const h = this.measureCalendarTableHeight();
                        if (h > 0)
                            this.calendarTableHeightPx = h;
                        this.concealUntilMeasured = false;
                    });
                }
            }
            else {
                // Kullanıcı değişikliklerini uygulamadan paneli kapatırsa, geçici seçimleri temizle
                if (this.allowApplyButton) {
                    this.processDateValue(this.value, true);
                }
                this.currentView = 'days';
                this.calendarTableHeightPx = undefined;
                this.concealUntilMeasured = false;
            }
        }
    }
    valueChanged(newValue, oldValue) {
        if (JSON.stringify(newValue) === JSON.stringify(oldValue)) {
            return;
        }
        this.processDateValue(newValue, true);
        this.remeasureCalendarOnNextFrame();
    }
    dateFormatChanged() {
        if (this.timeOnly)
            return;
        this.updateMaskOptions();
    }
    showTimePickerChanged() {
        this.updateMaskOptions();
    }
    timeOnlyChanged() {
        // Update mask according to the active mode
        this.updateMaskOptions();
    }
    timeFormatChanged() {
        if (this.timeOnly || this.showTimePicker) {
            this.updateMaskOptions();
            // Sync AM/PM with the current hour when switching to 12h
            if (this.timeFormat === '12' && this.internalStartTime) {
                this.internalAmPm = this.internalStartTime.hour >= 12 ? 'PM' : 'AM';
            }
        }
    }
    componentWillLoad() {
        this.updateMaskOptions();
        if (this.allowedDates) {
            this.allowedDates = this.allowedDates.filter(date => {
                const parsed = this.parseInputDate(date);
                return parsed !== null;
            });
        }
        if (this.disabledDates) {
            this.disabledDates = this.disabledDates.filter(date => {
                const parsed = this.parseInputDate(date);
                return parsed !== null;
            });
        }
        this.initializeDates();
        this.hasFooterSlot = !!this.el.querySelector('[slot="footer"]');
        this.hasFooterActionsSlot = !!this.el.querySelector('[slot="footer-actions"]');
    }
    componentDidLoad() {
        var _a, _b, _c;
        (_b = (_a = this.internals) === null || _a === void 0 ? void 0 : _a.form) === null || _b === void 0 ? void 0 : _b.addEventListener('reset', () => {
            this.handleFormReset();
        });
        // Initialize click outside mixin only if not inline mode
        if (!this.inline) {
            const tkInputArea = (_c = this.inputRef) === null || _c === void 0 ? void 0 : _c.querySelector('.tk-input');
            this.clickOutsideMixin = new ClickOutsideMixin({
                referenceElement: tkInputArea !== null && tkInputArea !== void 0 ? tkInputArea : this.el,
                handler: this.closeHandler,
                disabled: this.disabled || this.readonly || !this.isOpen,
            });
        }
        if (this.inline && this.showTimePicker) {
            requestAnimationFrame(() => {
                const h = this.measureCalendarTableHeight();
                if (h > 0)
                    this.calendarTableHeightPx = h;
            });
        }
    }
    componentDidUpdate() {
        var _a, _b, _c;
        // Update click outside mixin configuration based on current state
        (_a = this.clickOutsideMixin) === null || _a === void 0 ? void 0 : _a.updateConfig({
            disabled: this.disabled || this.readonly || this.inline || !this.isOpen,
            ignoredElements: this.panelRef ? [this.panelRef] : [],
        });
        if (this.isOpen) {
            if (this.inputRef && this.panelRef) {
                // Clean up old floating UI listeners before setting up new ones
                (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
                this.updatePosition();
            }
        }
        else {
            // Remove floating UI listeners when datepicker closes
            (_c = this.cleanup) === null || _c === void 0 ? void 0 : _c.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
        }
    }
    disconnectedCallback() {
        var _a, _b, _c, _d;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        (_c = (_b = this.internals) === null || _b === void 0 ? void 0 : _b.form) === null || _c === void 0 ? void 0 : _c.removeEventListener('reset', this.handleFormReset);
        // Call mixin's disconnectedCallback for cleanup
        (_d = this.clickOutsideMixin) === null || _d === void 0 ? void 0 : _d.disconnectedCallback();
    }
    formResetCallback() {
        this.handleFormReset();
    }
    /**
     * Sets the date to today
     */
    async setToday() {
        const today = this.normalizeDate(new Date());
        this.currentMonth = today;
        this.internalSelectedDates = { start: today, end: null };
        if (this.showTimePicker) {
            const defaultTime = this.getDefaultTime();
            this.internalStartTime = defaultTime;
            this.internalEndTime = null;
            if (this.timeFormat === '12')
                this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
        }
        else {
            this.internalStartTime = null;
            this.internalEndTime = null;
        }
    }
    /**
     * Applies the current internal selection and emits tk-change
     */
    async apply() {
        this.handleApply();
    }
    /**
     * Closes the datepicker panel if it is open.
     */
    async closePanel() {
        if (this.isOpen) {
            this.isOpen = false;
        }
    }
    updatePosition() {
        const inputEl = this.inputRef.querySelector('.tk-input');
        this.cleanup = floatingElementAutoUpdate(inputEl, this.panelRef, undefined, {
            placement: 'bottom-start',
        });
    }
    isTimeDisabled(hour, minute) {
        const toTotalMinutes = (h, m) => h * 60 + m;
        let testHour = hour;
        if (this.timeFormat === '12') {
            // 12h format - convert to 24h for comparison
            testHour = (hour === 12 ? 0 : hour) + (this.internalAmPm === 'PM' ? 12 : 0);
        }
        const total = toTotalMinutes(testHour, minute);
        const parseBound = (val) => {
            if (!val)
                return null;
            // Try parsing with current timeFormat and AM/PM context
            let parsed = null;
            if (this.timeFormat === '12' && /am|pm|AM|PM/i.test(val)) {
                // If AM/PM present, parse as 12h
                parsed = this.parseTimeString(val);
            }
            else if (this.timeFormat === '12') {
                // If not present, append current AM/PM
                parsed = this.parseTimeString(val + ' ' + this.internalAmPm);
            }
            else {
                // 24h mode
                parsed = this.parseTimeString(val);
            }
            if (!parsed)
                return null;
            return toTotalMinutes(parsed.getHours(), parsed.getMinutes());
        };
        const minBound = parseBound(this.minTime);
        const maxBound = parseBound(this.maxTime);
        if (minBound !== null && total < minBound)
            return true;
        if (maxBound !== null && total > maxBound)
            return true;
        return false;
    }
    updateTimeBasedOnAmPm(newAmPm) {
        if (this.timeFormat !== '12' || this.isUpdatingAmPm) {
            return;
        }
        this.isUpdatingAmPm = true;
        const targetTimeState = this.getTimeStateToModify();
        if (!targetTimeState) {
            this.isUpdatingAmPm = false;
            return;
        }
        // Convert current hour based on AM/PM change
        let currentHour = targetTimeState.time.hour;
        let needsUpdate = false;
        if (newAmPm === 'PM' && currentHour < 12) {
            currentHour += 12;
            needsUpdate = true;
        }
        else if (newAmPm === 'AM' && currentHour >= 12) {
            currentHour -= 12;
            needsUpdate = true;
        }
        // Only update if the hour actually changed
        if (needsUpdate) {
            if (targetTimeState.type === 'start') {
                this.internalStartTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: currentHour });
            }
            else {
                this.internalEndTime = Object.assign(Object.assign({}, targetTimeState.time), { hour: currentHour });
            }
            this.emitTimeChange();
        }
        this.isUpdatingAmPm = false;
    }
    getResolvedFirstDayIndex() {
        if (this.firstDayOfWeekIndex !== undefined && this.firstDayOfWeekIndex !== null) {
            if (this.firstDayOfWeekIndex >= 0 && this.firstDayOfWeekIndex <= 6) {
                return this.firstDayOfWeekIndex;
            }
            else {
                console.warn(`Invalid firstDayOfWeekIndex value: ${this.firstDayOfWeekIndex}. Defaulting to 0 (Monday).`);
                return 0;
            }
        }
        try {
            const resolvedLocale = this.locale || 'en';
            if (typeof Intl !== 'undefined' && Intl.Locale && typeof Intl.Locale === 'function' && Intl.Locale.prototype.hasOwnProperty('weekInfo')) {
                // @ts-ignore: Accessing weekInfo from potentially unknown Intl.Locale type
                const localeInfo = new Intl.Locale(resolvedLocale).getWeekInfo();
                if (localeInfo && localeInfo.firstDay !== undefined) {
                    // Formula: (IntlDay + 6) % 7 maps 1(Mon) to 0, 7(Sun) to 6
                    return (localeInfo.firstDay + 6) % 7;
                }
            }
        }
        catch (e) {
            console.warn(`Could not determine first day of week from locale '${this.locale}'. Error or unsupported: ${e}`);
        }
        return 0;
    }
    getFullDateTimeFormat() {
        const timePattern = this.timeFormat === '12' ? 'hh:mm a' : 'HH:mm';
        return /[hH]/.test(this.dateFormat) ? this.dateFormat : `${this.dateFormat} ${timePattern}`;
    }
    getOnlyTimeFormat() {
        return this.timeFormat === '12' ? 'hh:mm a' : 'HH:mm';
    }
    getDefaultTime() {
        const now = new Date();
        let hour = now.getHours();
        let minute = now.getMinutes();
        // Align hour to the configured hourStep (floor to nearest step)
        const hourStep = Math.max(1, this.hourStep || 1);
        hour = Math.floor(hour / hourStep) * hourStep;
        // Align minutes to the configured step (floor to nearest step)
        const minuteStep = Math.max(1, this.minuteStep || 1);
        minute = Math.floor(minute / minuteStep) * minuteStep;
        if (!this.isTimeDisabled(hour, minute)) {
            return { hour, minute };
        }
        const hours = this.timeFormat === '12'
            ? Array.from({ length: Math.ceil(12 / hourStep) }, (_, i) => Math.min(i * hourStep + 1, 12))
            : Array.from({ length: Math.ceil(24 / hourStep) }, (_, i) => i * hourStep);
        const minutes = Array.from({ length: Math.ceil(60 / minuteStep) }, (_, i) => i * minuteStep);
        for (const h of hours) {
            for (const m of minutes) {
                if (!this.isTimeDisabled(h, m)) {
                    return { hour: h, minute: m };
                }
            }
        }
        // No valid time found, log warning
        console.warn('TkDatepicker: No valid time slots available for the current configuration.');
        return { hour, minute };
    }
    getDateWithTime(date, type) {
        if (!date)
            return null;
        const newDate = new Date(date);
        const timeToApply = type === 'start' ? this.internalStartTime : this.internalEndTime;
        if (this.showTimePicker && timeToApply) {
            newDate.setHours(timeToApply.hour, timeToApply.minute);
        }
        return newDate;
    }
    getMaskOptionsFromDateFormat(format) {
        var _a;
        const delimiter = ((_a = format.match(/[^a-zA-Z]/)) === null || _a === void 0 ? void 0 : _a[0]) || '';
        const datePattern = [];
        const blockSizes = [];
        const parts = format.split(/[^a-zA-Z]/);
        parts.forEach(part => {
            switch (part.toLowerCase()) {
                case 'yyyy':
                    datePattern.push('Y');
                    blockSizes.push(4);
                    break;
                case 'yy':
                    datePattern.push('Y');
                    blockSizes.push(2);
                    break;
                case 'mm':
                case 'm':
                    datePattern.push('m');
                    blockSizes.push(2);
                    break;
                case 'dd':
                case 'd':
                    datePattern.push('d');
                    blockSizes.push(2);
                    break;
            }
        });
        const dateDelimiters = Array(blockSizes.length - 1).fill(delimiter);
        if (this.showTimePicker) {
            this.maskOptions = {
                blocks: [...blockSizes, 2, 2], // date blocks + HH:MM
                delimiters: [...dateDelimiters, ' ', ':'],
                numericOnly: true,
            };
        }
        else {
            this.maskOptions = {
                date: true,
                delimiter,
                datePattern,
            };
        }
    }
    processDateValue(value, updateCurrentMonth = false) {
        if (this.timeOnly) {
            // In time-only mode, value is expected to be a time string (e.g., HH:mm or hh:mm a)
            let startTime = null;
            if (typeof value === 'string' && value) {
                const parsed = this.parseTimeString(value);
                if (parsed) {
                    startTime = { hour: parsed.getHours(), minute: parsed.getMinutes() };
                }
            }
            // For initial load, set AM/PM based on actual time, but respect user changes after that
            if (this.timeFormat === '12' && startTime) {
                // Only set if this is the initial default state, otherwise respect user choice
                if (this.internalAmPm === 'AM' && startTime.hour >= 12) {
                    this.internalAmPm = 'PM';
                }
            }
            this.internalStartTime = startTime;
            this.internalEndTime = startTime;
            // Do not set any dates in time-only mode
            this.internalSelectedDates = { start: null, end: null };
            this.inputValue = this.formatInputValue();
            return;
        }
        let startDate = null;
        let endDate = null;
        let startTime = null;
        let endTime = null;
        const now = new Date();
        const defaultTime = this.getDefaultTime();
        if (value) {
            let startString = null;
            let endString = null;
            if (typeof value === 'string') {
                startString = value;
            }
            else if (typeof value === 'object' && value.start) {
                startString = value.start;
                endString = value.end || null;
            }
            if (startString) {
                const parsedStartDateTime = this.showTimePicker ? this.parseFullDateTime(startString) : this.parseInputDate(startString);
                if (parsedStartDateTime && !this.isDateDisabled(parsedStartDateTime)) {
                    startDate = this.normalizeDate(parsedStartDateTime);
                    if (this.showTimePicker) {
                        const expectedFullFormat = this.formatFullDateTime(parsedStartDateTime);
                        if (startString === expectedFullFormat || startString.includes(' ')) {
                            startTime = { hour: parsedStartDateTime.getHours(), minute: parsedStartDateTime.getMinutes() };
                        }
                        else {
                            startTime = defaultTime;
                        }
                    }
                }
                else {
                    startDate = null;
                    startTime = null;
                }
            }
            if (this.mode === 'range' && endString && startDate) {
                const parsedEndDateTime = this.showTimePicker ? this.parseFullDateTime(endString) : this.parseInputDate(endString);
                if (parsedEndDateTime && !this.isDateDisabled(parsedEndDateTime)) {
                    endDate = this.normalizeDate(parsedEndDateTime);
                    if (this.showTimePicker) {
                        const expectedFullFormat = this.formatFullDateTime(parsedEndDateTime);
                        if (endString === expectedFullFormat || endString.includes(' ')) {
                            endTime = { hour: parsedEndDateTime.getHours(), minute: parsedEndDateTime.getMinutes() };
                        }
                        else {
                            endTime = startTime || defaultTime;
                        }
                    }
                }
                else {
                    endDate = null;
                    endTime = null;
                }
            }
            else if (this.mode === 'range') {
                endDate = null;
                endTime = null;
            }
            else {
                endDate = null;
                endTime = null;
            }
        }
        else {
            if (this.showTimePicker) {
                startTime = defaultTime;
                endTime = this.mode === 'range' ? null : defaultTime;
            }
        }
        this.internalSelectedDates = { start: startDate, end: endDate };
        this.internalStartTime = startTime;
        this.internalEndTime = this.mode === 'range' ? endTime : startTime;
        if (updateCurrentMonth && startDate) {
            this.currentMonth = new Date(startDate.getFullYear(), startDate.getMonth());
        }
        else if (!startDate) {
            this.currentMonth = new Date(this.currentMonth || now);
        }
        this.inputValue = this.formatInputValue();
    }
    initializeDates() {
        var _a;
        // Initialize currentMonth with priority: defaultDate > first allowedDate month > first non-disabled month > now
        const now = new Date();
        const parsedDefaultDate = this.parseDefaultDate(this.defaultDate);
        if (parsedDefaultDate) {
            this.currentMonth = parsedDefaultDate;
        }
        else if (this.allowedDates && this.allowedDates.length > 0) {
            this.currentMonth = (_a = this.getFirstAvailableAllowedMonth()) !== null && _a !== void 0 ? _a : now;
        }
        else {
            let candidateMonth = now;
            if (this.disabledDates && this.disabledDates.length > 0) {
                let limit = 12; // Prevent infinite loop by limiting to 12 months of searching
                while (this.isMonthFullyDisabled(candidateMonth.getFullYear(), candidateMonth.getMonth()) && limit > 0) {
                    candidateMonth = new Date(candidateMonth.getFullYear(), candidateMonth.getMonth() + 1, 1);
                    limit--;
                }
            }
            this.currentMonth = candidateMonth;
        }
        this.processDateValue(this.value, true);
        if (this.showTimePicker && !this.internalStartTime && this.internalSelectedDates.start) {
            const defaultTime = this.getDefaultTime();
            this.internalStartTime = defaultTime;
            if (this.timeFormat === '12')
                this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
            if (this.mode === 'range' && !this.internalEndTime && this.internalSelectedDates.end) {
                this.internalEndTime = this.internalStartTime;
            }
            else if (this.mode === 'single') {
                this.internalEndTime = this.internalStartTime;
            }
        }
        this.inputValue = this.formatInputValue();
    }
    measureCalendarTableHeight() {
        var _a;
        const table = (_a = this.panelRef) === null || _a === void 0 ? void 0 : _a.querySelector('.tk-datepicker-table');
        return table ? Math.ceil(table.getBoundingClientRect().height) : 0;
    }
    normalizeDate(date) {
        return new Date(date.getFullYear(), date.getMonth(), date.getDate());
    }
    formatDate(date) {
        return format(date, this.dateFormat);
    }
    formatFullDateTime(date) {
        if (!date)
            return '';
        return format(date, this.getFullDateTimeFormat());
    }
    formatDateOrDateTime(date, type) {
        if (this.timeOnly) {
            if (!this.internalStartTime)
                return '';
            const temp = new Date();
            temp.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
            return format(temp, this.getOnlyTimeFormat());
        }
        if (this.showTimePicker && date && type) {
            const dateWithCorrectTime = this.getDateWithTime(date, type);
            return dateWithCorrectTime ? this.formatFullDateTime(dateWithCorrectTime) : '';
        }
        else if (date) {
            return this.formatDate(date);
        }
        return '';
    }
    parseInputDate(dateString) {
        const parsedDate = parse(dateString, this.dateFormat, new Date());
        if (isValid(parsedDate) && this.formatDate(parsedDate) === dateString) {
            return parsedDate;
        }
        return null;
    }
    parseFullDateTime(dateTimeString) {
        // Try full date+time format first
        const formatString = this.getFullDateTimeFormat();
        const parsedDate = parse(dateTimeString, formatString, new Date());
        if (isValid(parsedDate) && format(parsedDate, formatString) === dateTimeString) {
            return parsedDate;
        }
        // Fallback: allow typing HH:mm without AM/PM for 12h mode
        if (this.timeFormat === '12') {
            const altFormat = `${this.dateFormat} HH:mm`;
            const parsedAlt = parse(dateTimeString, altFormat, new Date());
            if (isValid(parsedAlt) && format(parsedAlt, altFormat) === dateTimeString) {
                return parsedAlt;
            }
        }
        return null;
    }
    parseTimeString(timeString) {
        const base = new Date();
        const primaryFmt = this.getOnlyTimeFormat();
        let parsed = parse(timeString, primaryFmt, base);
        if (isValid(parsed) && format(parsed, primaryFmt) === timeString) {
            return parsed;
        }
        // Fallbacks
        // 1) In 12h mode, allow typing time without AM/PM (e.g., "11:00").
        //    We interpret it using the current AM/PM selection.
        if (this.timeFormat === '12') {
            const simpleMatch = timeString.match(/^(\d{1,2}):(\d{2})$/);
            if (simpleMatch) {
                let hour = parseInt(simpleMatch[1], 10);
                const minute = parseInt(simpleMatch[2], 10);
                if (Number.isFinite(hour) && Number.isFinite(minute) && minute >= 0 && minute <= 59) {
                    if (this.internalAmPm === 'PM' && hour < 12)
                        hour += 12;
                    if (this.internalAmPm === 'AM' && hour === 12)
                        hour = 0;
                    const t = new Date(base);
                    t.setHours(Math.max(0, Math.min(23, hour)), minute, 0, 0);
                    return t;
                }
            }
            // Also accept explicit 24h input (HH:mm) and convert to a Date directly
            const alt24 = parse(timeString, 'HH:mm', base);
            if (isValid(alt24) && format(alt24, 'HH:mm') === timeString) {
                return alt24;
            }
        }
        return null;
    }
    isDateDisabled(date) {
        const normalizedDate = this.normalizeDate(date);
        const dateStr = this.formatDate(normalizedDate);
        const minDate = this.minDate ? this.parseInputDate(this.minDate) : null;
        const maxDate = this.maxDate ? this.parseInputDate(this.maxDate) : null;
        if (minDate) {
            minDate.setHours(0, 0, 0, 0);
        }
        if (maxDate) {
            maxDate.setHours(23, 59, 59, 999);
        }
        const isOutOfRange = (minDate && date < minDate) || (maxDate && date > maxDate);
        if (isOutOfRange)
            return true;
        if (this.allowedDates && this.allowedDates.length > 0) {
            const isAllowed = this.allowedDates.some(allowedDate => {
                const parsedAllowedDate = this.parseInputDate(allowedDate);
                return parsedAllowedDate && this.formatDate(this.normalizeDate(parsedAllowedDate)) === dateStr;
            });
            return !isAllowed;
        }
        if (this.disabledDates && this.disabledDates.length > 0) {
            return this.disabledDates.some(disabledDate => {
                const parsedDisabledDate = this.parseInputDate(disabledDate);
                return parsedDisabledDate && this.formatDate(this.normalizeDate(parsedDisabledDate)) === dateStr;
            });
        }
        if (this.disabledWeekDays && this.disabledWeekDays.length > 0) {
            const dayOfWeek = date.getDay();
            return this.disabledWeekDays.includes(dayOfWeek);
        }
        return false;
    }
    isMonthFullyDisabled(year, month) {
        // Check if the entire month has all days disabled
        const daysInMonth = new Date(year, month + 1, 0).getDate();
        for (let day = 1; day <= daysInMonth; day++) {
            const date = new Date(year, month, day);
            if (!this.isDateDisabled(date)) {
                return false; // Found at least one enabled date
            }
        }
        return true; // All days in the month are disabled
    }
    parseDefaultDate(value) {
        if (!value)
            return null;
        const now = new Date();
        const delimiters = ['.', '/', '_', '-'];
        const delimiterRegex = new RegExp(`[${delimiters.join('')}]`);
        const parts = value.trim().split(delimiterRegex);
        if (parts.length === 2) {
            const [first, second] = parts;
            if (/^\d{4}$/.test(first) && /^\d{2}$/.test(second)) {
                // yyyy-MM
                const year = Number.parseInt(first, 10);
                const month = Number.parseInt(second, 10);
                if (month >= 1 && month <= 12)
                    return new Date(year, month - 1, 1);
            }
            else if (/^\d{2}$/.test(first) && /^\d{4}$/.test(second)) {
                // MM-yyyy
                const month = Number.parseInt(first, 10);
                const year = Number.parseInt(second, 10);
                if (month >= 1 && month <= 12)
                    return new Date(year, month - 1, 1);
            }
            return null;
        }
        // (yyyy)
        if (parts.length === 1 && /^\d{4}$/.test(parts[0])) {
            const year = Number.parseInt(parts[0], 10);
            return new Date(year, now.getMonth(), 1);
        }
        //  (MM)
        if (parts.length === 1 && /^\d{2}$/.test(parts[0])) {
            const month = Number.parseInt(parts[0], 10);
            if (month >= 1 && month <= 12)
                return new Date(now.getFullYear(), month - 1, 1);
        }
        return null;
    }
    getFirstAvailableAllowedMonth() {
        var _a;
        if (!((_a = this.allowedDates) === null || _a === void 0 ? void 0 : _a.length))
            return null;
        const parsedAllowedDates = this.allowedDates.map(date => this.parseInputDate(date)).filter((date) => date !== null);
        if (!parsedAllowedDates.length)
            return null;
        const minDate = parsedAllowedDates.reduce((min, curr) => (curr.getTime() < min.getTime() ? curr : min), parsedAllowedDates[0]);
        return new Date(minDate.getFullYear(), minDate.getMonth(), 1);
    }
    formatInputValue() {
        if (this.timeOnly) {
            if (this.internalStartTime) {
                const temp = new Date();
                temp.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
                return format(temp, this.getOnlyTimeFormat());
            }
            return '';
        }
        const { start, end } = this.internalSelectedDates;
        if (start) {
            const formattedStart = this.formatDateOrDateTime(start, 'start');
            if (this.mode === 'range' && end) {
                const formattedEnd = this.formatDateOrDateTime(end, 'end');
                return `${formattedStart} - ${formattedEnd}`;
            }
            return formattedStart;
        }
        return '';
    }
    isToday(date) {
        return date.getDate() === this.today.getDate() && date.getMonth() === this.today.getMonth() && date.getFullYear() === this.today.getFullYear();
    }
    ensureDateTimeInitialized(type) {
        if (this.timeOnly) {
            // In time-only mode, prefer parsing the current input value instead of resetting to default
            if (!this.internalStartTime) {
                let parsedFromInput = null;
                if (this.inputValue) {
                    parsedFromInput = this.parseTimeString(this.inputValue);
                }
                const resolvedTime = parsedFromInput ? { hour: parsedFromInput.getHours(), minute: parsedFromInput.getMinutes() } : this.getDefaultTime();
                this.internalStartTime = resolvedTime;
                this.internalEndTime = resolvedTime;
                if (this.timeFormat === '12')
                    this.internalAmPm = resolvedTime.hour >= 12 ? 'PM' : 'AM';
            }
            return;
        }
        if (!this.internalSelectedDates.start && type === 'start') {
            const today = this.normalizeDate(new Date());
            this.internalSelectedDates = Object.assign(Object.assign({}, this.internalSelectedDates), { start: today });
            this.currentMonth = new Date(today);
        }
        if (!this.internalSelectedDates.end && type === 'end' && this.mode === 'range') {
            return; // Don't initialize end date if not set
        }
        if (this.showTimePicker) {
            const defaultTime = this.getDefaultTime();
            if (type === 'start' && !this.internalStartTime) {
                this.internalStartTime = defaultTime;
                if (this.timeFormat === '12')
                    this.internalAmPm = defaultTime.hour >= 12 ? 'PM' : 'AM';
            }
            if (type === 'end' && this.mode === 'range' && this.internalSelectedDates.end && !this.internalEndTime) {
                this.internalEndTime = this.internalStartTime || defaultTime;
            }
        }
    }
    ensureRangeOrder() {
        var _a, _b;
        if (this.mode !== 'range' || !this.showTimePicker)
            return true;
        const { start, end } = this.internalSelectedDates;
        if (!start || !end)
            return true;
        const startTs = (_a = this.getDateWithTime(start, 'start')) === null || _a === void 0 ? void 0 : _a.getTime();
        const endTs = (_b = this.getDateWithTime(end, 'end')) === null || _b === void 0 ? void 0 : _b.getTime();
        if (startTs !== undefined && endTs !== undefined && endTs < startTs) {
            this.internalSelectedDates = { start: end, end: start };
            [this.internalStartTime, this.internalEndTime] = [this.internalEndTime, this.internalStartTime];
            this.inputValue = this.formatInputValue();
            this.tkChange.emit({
                start: this.formatDateOrDateTime(end, 'start'),
                end: this.formatDateOrDateTime(start, 'end'),
            });
            return false;
        }
        return true;
    }
    emitTimeChange() {
        // Apply buttona basılmadıkça anlık değişiklikleri emit etmesin diye
        if (this.allowApplyButton)
            return;
        if (this.timeOnly) {
            if (!this.internalStartTime)
                return;
            const temp = new Date();
            temp.setHours(this.internalStartTime.hour, this.internalStartTime.minute, 0, 0);
            const value = format(temp, this.getOnlyTimeFormat());
            this.tkChange.emit(value);
            this.inputValue = this.formatInputValue();
            return;
        }
        if (!this.showTimePicker || !this.internalSelectedDates.start)
            return;
        // if swap occurred, bail
        if (!this.ensureRangeOrder()) {
            return;
        }
        const { start, end } = this.internalSelectedDates;
        let emitValue;
        const formattedStart = this.formatDateOrDateTime(start, 'start');
        if (this.mode === 'range') {
            const formattedEnd = end ? this.formatDateOrDateTime(end, 'end') : null;
            emitValue = {
                start: formattedStart,
                end: formattedEnd,
            };
        }
        else {
            emitValue = formattedStart;
        }
        this.tkChange.emit(emitValue);
        this.inputValue = this.formatInputValue();
    }
    updateMaskOptions() {
        if (this.timeOnly) {
            // timeformat'a göre mask options'ları belirle
            this.maskOptions = {
                time: true,
                timePattern: ['h', 'm'],
                timeFormat: this.timeFormat,
            };
        }
        else {
            this.getMaskOptionsFromDateFormat(this.dateFormat);
        }
    }
    getTimeStateToModify() {
        // Allow time changes if any time UI is active: showTimePicker or timeOnly mode
        if (!(this.showTimePicker || this.timeOnly))
            return null;
        let targetType = 'start';
        if (this.mode === 'range' && this.internalSelectedDates.end) {
            targetType = 'end';
        }
        this.ensureDateTimeInitialized(targetType);
        const timeState = targetType === 'start' ? this.internalStartTime : this.internalEndTime;
        if (!timeState)
            return null;
        return { time: timeState, type: targetType };
    }
    handleYearSelect(e, year) {
        e.stopPropagation();
        this.currentMonth = new Date(year, this.currentMonth.getMonth());
        this.currentView = 'months';
        this.remeasureCalendarOnNextFrame();
    }
    handleFormReset() {
        const initialValueAttr = this.el.getAttribute('value');
        let initialValue = null;
        if (initialValueAttr) {
            try {
                initialValue = JSON.parse(initialValueAttr);
            }
            catch (e) {
                initialValue = initialValueAttr;
            }
        }
        this.processDateValue(initialValue, true);
        this.tkChange.emit(this.value);
    }
    createDayCell(date, isAdjacentMonth) {
        const { start = null, end = null } = this.internalSelectedDates;
        const dateTime = date.getTime();
        const isSelectedStart = start && dateTime === start.getTime();
        const isSelectedEnd = start && end && dateTime === end.getTime();
        const isSelected = isSelectedStart || isSelectedEnd;
        let isInRange = false;
        const rangeEnd = end || this.hoverDate;
        let visualRangeStart = null;
        let visualRangeEnd = null;
        if (start && rangeEnd) {
            if (start.getTime() <= (rangeEnd === null || rangeEnd === void 0 ? void 0 : rangeEnd.getTime())) {
                visualRangeStart = start;
                visualRangeEnd = rangeEnd;
            }
            else {
                // Reverse selection: user selected end date first, then hovering/selecting earlier date
                visualRangeStart = rangeEnd;
                visualRangeEnd = start;
            }
            isInRange = dateTime > (visualRangeStart === null || visualRangeStart === void 0 ? void 0 : visualRangeStart.getTime()) && dateTime < (visualRangeEnd === null || visualRangeEnd === void 0 ? void 0 : visualRangeEnd.getTime());
        }
        const isVisualRangeStart = visualRangeStart && dateTime === (visualRangeStart === null || visualRangeStart === void 0 ? void 0 : visualRangeStart.getTime());
        const isVisualRangeEnd = visualRangeEnd && dateTime === (visualRangeEnd === null || visualRangeEnd === void 0 ? void 0 : visualRangeEnd.getTime());
        const isDisabled = this.isDateDisabled(date);
        const isToday = this.isToday(date);
        const dateKey = `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`;
        return (h("td", { class: classNames('tk-datepicker-day-cell', {
                'selected': isSelected,
                'in-range': isInRange,
                'range-start': isVisualRangeStart && this.mode === 'range',
                'range-end': isVisualRangeEnd && this.mode === 'range',
                'today': isToday && !isSelected && !isInRange,
                'disabled': isDisabled || this.disabled,
                'readonly': this.readonly,
                'adjacent-month': isAdjacentMonth,
            }), onClick: () => !isDisabled && this.handleDateClick(date), onMouseEnter: () => this.handleDateHover(date), "data-testid": getDataTestId(this.dataTestid, 'day-cell', dateKey) }, h("span", { class: "tk-datepicker-day", "data-testid": getDataTestId(this.dataTestid, 'day', dateKey) }, date.getDate())));
    }
    createWeekDays() {
        const firstDayOfMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth(), 1);
        const lastDayOfMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() + 1, 0);
        const daysInPreviousMonth = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth(), 0).getDate();
        const daysInCurrentMonth = lastDayOfMonth.getDate();
        const weeks = [];
        let days = [];
        const resolvedStartOfWeekIndex = this.getResolvedFirstDayIndex();
        const startOfWeekForGetDay = (resolvedStartOfWeekIndex + 1) % 7;
        const firstDayOfWeek = firstDayOfMonth.getDay();
        let emptyCells = (firstDayOfWeek - startOfWeekForGetDay + 7) % 7;
        // Previous month's days
        for (let i = emptyCells - 1; i >= 0; i--) {
            const date = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() - 1, daysInPreviousMonth - i);
            days.push(this.createDayCell(date, true));
        }
        // Current month's days
        for (let i = 1; i <= daysInCurrentMonth; i++) {
            const date = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth(), i);
            days.push(this.createDayCell(date, false));
            if (days.length === 7) {
                weeks.push(h("tr", { "data-testid": getDataTestId(this.dataTestid, 'week', weeks.length.toString()) }, days));
                days = [];
            }
        }
        // Next month's days
        if (days.length > 0) {
            let nextMonthDay = 1;
            while (days.length < 7) {
                const date = new Date(this.currentMonth.getFullYear(), this.currentMonth.getMonth() + 1, nextMonthDay);
                days.push(this.createDayCell(date, true));
                nextMonthDay++;
            }
            weeks.push(h("tr", { "data-testid": getDataTestId(this.dataTestid, 'week', weeks.length.toString()) }, days));
        }
        this.weeksLength = weeks.length;
        return (h("tbody", { class: "tk-datepicker-days", "data-testid": getDataTestId(this.dataTestid, 'days') }, weeks));
    }
    createWeekDayNames() {
        const startOfWeekIndex = this.getResolvedFirstDayIndex();
        // Use a known Monday as a base to generate other days reliably
        const baseMonday = new Date(2023, 0, 2);
        const weekdays = [...Array(7)].map((_, i) => {
            const dayOffset = (i + startOfWeekIndex) % 7;
            const dateForDay = new Date(baseMonday);
            dateForDay.setDate(baseMonday.getDate() + dayOffset);
            return dateForDay.toLocaleString(this.locale, { weekday: 'short' });
        });
        return (h("thead", { "data-testid": getDataTestId(this.dataTestid, 'week-days-header') }, h("tr", { class: "tk-datepicker-week-days", "data-testid": getDataTestId(this.dataTestid, 'week-days-row') }, weekdays.map((day, index) => (h("th", { class: "tk-datepicker-week-day-cell", "data-testid": getDataTestId(this.dataTestid, 'week-cell', index.toString()) }, h("span", { class: "tk-datepicker-week-day", "data-testid": getDataTestId(this.dataTestid, 'week-day', index.toString()) }, day)))))));
    }
    createMonths() {
        const months = [...Array(12)].map((_, i) => new Date(2023, i).toLocaleString(this.locale, { month: 'short' }));
        const rows = [];
        for (let i = 0; i < 12; i += 4) {
            const monthGroup = months.slice(i, i + 4);
            rows.push(h("tr", { class: "tk-datepicker-month-row", "data-testid": getDataTestId(this.dataTestid, 'month', (i / 4).toString()) }, monthGroup.map((month, index) => {
                const monthIndex = i + index;
                const isSelected = this.currentMonth.getMonth() === monthIndex;
                return (h("td", { class: classNames('tk-datepicker-month', {
                        selected: isSelected,
                    }), "data-testid": getDataTestId(this.dataTestid, 'month-option', monthIndex.toString()), onClick: (e) => {
                        e.stopPropagation();
                        this.currentMonth = new Date(this.currentMonth.getFullYear(), monthIndex);
                        this.currentView = 'days';
                    } }, month));
            })));
        }
        return (h("tbody", { class: "tk-datepicker-months", "data-testid": getDataTestId(this.dataTestid, 'months') }, rows));
    }
    createYears() {
        const currentYear = this.currentMonth.getFullYear();
        const startYear = Math.floor(currentYear / 12) * 12;
        const years = Array.from({ length: 12 }, (_, i) => startYear + i);
        const rows = [];
        for (let i = 0; i < years.length; i += 4) {
            const yearGroup = years.slice(i, i + 4);
            rows.push(h("tr", { class: "tk-datepicker-year-row", "data-testid": getDataTestId(this.dataTestid, 'year', (i / 4).toString()) }, yearGroup.map(year => (h("td", { class: classNames('tk-datepicker-year', {
                    selected: year === currentYear,
                }), "data-testid": getDataTestId(this.dataTestid, 'year', year.toString()), onClick: e => this.handleYearSelect(e, year) }, year)))));
        }
        return (h("tbody", { class: "tk-datepicker-years", "data-testid": getDataTestId(this.dataTestid, 'years') }, rows));
    }
    createHeader() {
        const monthName = this.currentMonth.toLocaleString(this.locale, {
            month: 'long',
        });
        const year = this.currentMonth.getFullYear().toString();
        const headerClasses = classNames('tk-datepicker-header', `tk-datepicker-header-${this.headerType}`);
        return (h("div", { class: headerClasses, "data-testid": getDataTestId(this.dataTestid, 'header') }, h("div", { class: "tk-datepicker-header-content", "data-testid": getDataTestId(this.dataTestid, 'header-content') }, h("div", { class: "tk-datepicker-header-content-start", "data-testid": getDataTestId(this.dataTestid, 'header-start') }, h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "keyboard_double_arrow_left", "onTk-click": () => this.handleYearChange(-1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: getDataTestId(this.dataTestid, 'prev-year-button') }), h("span", { class: "tk-datepicker-divider", "data-testid": getDataTestId(this.dataTestid, 'header-start-divider') }), h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "chevron_left", "onTk-click": () => this.handleMonthChange(-1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: getDataTestId(this.dataTestid, 'prev-month-button') })), h("div", { class: "tk-datepicker-select-container", "data-testid": getDataTestId(this.dataTestid, 'header-select') }, h("div", { class: classNames('tk-datepicker-select-month', { disabled: this.readonly || this.disabled || this.loading }), onClick: e => !this.loading && this.handleViewChange(e, 'months'), "data-testid": getDataTestId(this.dataTestid, 'header-month') }, monthName), h("div", { class: classNames('tk-datepicker-select-year', { disabled: this.readonly || this.disabled || this.loading }), onClick: e => !this.loading && this.handleViewChange(e, 'years'), "data-testid": getDataTestId(this.dataTestid, 'header-year') }, year)), h("div", { class: "tk-datepicker-header-content-end", "data-testid": getDataTestId(this.dataTestid, 'header-end') }, h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "chevron_right", "onTk-click": () => this.handleMonthChange(1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: getDataTestId(this.dataTestid, 'next-month-button') }), h("span", { class: "tk-datepicker-divider", "data-testid": getDataTestId(this.dataTestid, 'header-end-divider') }), h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', icon: "keyboard_double_arrow_right", "onTk-click": () => this.handleYearChange(1), type: "text", disabled: this.readonly || this.disabled || this.loading, dataTestid: getDataTestId(this.dataTestid, 'next-year-button') })))));
    }
    createFooter() {
        if (this.hasFooterSlot) {
            return h("slot", { name: "footer" });
        }
        else if (this.hasFooterActionsSlot) {
            const footerClass = classNames('tk-datepicker-footer', `tk-datepicker-footer-${this.footerType}`);
            return (h("div", { class: footerClass, "data-testid": getDataTestId(this.dataTestid, 'footer') }, h("slot", { name: "footer-actions" })));
        }
        return null;
    }
    createTimePicker() {
        var _a, _b;
        let timeToDisplay = null;
        if (this.mode === 'single') {
            timeToDisplay = this.internalStartTime;
        }
        else {
            if (this.internalSelectedDates.end) {
                timeToDisplay = this.internalEndTime;
            }
            else if (this.internalSelectedDates.start) {
                timeToDisplay = this.internalStartTime;
            }
        }
        const defaultTime = this.getDefaultTime();
        const currentTime = timeToDisplay || defaultTime;
        let displayHour = currentTime.hour;
        if (this.timeFormat === '12') {
            displayHour = displayHour % 12 === 0 ? 12 : displayHour % 12;
        }
        const displayMinute = currentTime.minute;
        const hours = this.timeFormat === '12'
            ? Array.from({ length: Math.ceil(12 / this.hourStep) }, (_, i) => i * this.hourStep + 1)
            : Array.from({ length: Math.ceil(24 / this.hourStep) }, (_, i) => i * this.hourStep);
        const minutes = Array.from({ length: Math.ceil(60 / this.minuteStep) }, (_, i) => i * this.minuteStep);
        const isHourDisabled = (hour) => this.isTimeDisabled(hour, displayMinute);
        const isMinuteDisabled = (minute) => this.isTimeDisabled(displayHour, minute);
        // Find closest hour in the hours array
        const findClosestInArray = (value, arr) => {
            if (arr.includes(value))
                return value;
            // Find the closest value in the array
            return arr.reduce((prev, curr) => (Math.abs(curr - value) < Math.abs(prev - value) ? curr : prev));
        };
        const currentHour = findClosestInArray(displayHour, hours);
        const currentMinute = findClosestInArray(displayMinute, minutes);
        const sliceRange = (options, selected) => {
            const idx = options.indexOf(selected);
            return Array.from({ length: this.weeksLength === 6 ? 5 : this.weeksLength === 4 ? 3 : 4 }, (_, i) => {
                const optIndex = idx - (this.weeksLength === 4 ? 1 : 2) + i;
                return optIndex >= 0 && optIndex < options.length ? options[optIndex] : null;
            });
        };
        const isDisabled = this.disabled || this.readonly;
        const visibleHours = sliceRange(hours, currentHour);
        const visibleMinutes = sliceRange(minutes, currentMinute);
        const isMinHour = currentHour === hours[0];
        const isMaxHour = currentHour === hours[hours.length - 1];
        const isMinMinute = currentMinute === minutes[0];
        const isMaxMinute = currentMinute === minutes[minutes.length - 1];
        const nextHourDisabled = isHourDisabled(hours[hours.indexOf(currentHour) + 1]);
        const prevHourDisabled = isHourDisabled(hours[hours.indexOf(currentHour) - 1]);
        const nextMinuteDisabled = isMinuteDisabled(minutes[minutes.indexOf(currentMinute) + 1]);
        const prevMinuteDisabled = isMinuteDisabled(minutes[minutes.indexOf(currentMinute) - 1]);
        let AMDisabled = false;
        let PMDisabled = false;
        if (this.timeFormat === '12') {
            if ((_a = this.minTime) === null || _a === void 0 ? void 0 : _a.toLowerCase().includes('pm')) {
                this.internalAmPm = 'PM';
                AMDisabled = true;
            }
            if ((_b = this.maxTime) === null || _b === void 0 ? void 0 : _b.toLowerCase().includes('am')) {
                this.internalAmPm = 'AM';
                PMDisabled = true;
            }
        }
        return (h("div", { class: classNames('tk-datepicker-timepicker-panel', this.timeOnly && 'tk-datepicker-timepicker-panel-only'), "data-testid": getDataTestId(this.dataTestid, 'timepicker') }, h("div", { class: classNames('tk-datepicker-timepicker-header', `tk-datepicker-timepicker-header-${this.headerType}`, this.timeOnly && 'tk-datepicker-timepicker-header-only'), "data-testid": getDataTestId(this.dataTestid, 'timepicker-header') }, this.timeFormat === '12' && (h("tk-toggle-button-group", { value: this.internalAmPm, type: this.headerType === 'basic' ? 'basic' : this.headerType === 'light' ? 'light' : 'divided', "onTk-change": e => this.handleAmPmToggle(e), class: "tk-datepicker-ampm-toggle", dataTestid: getDataTestId(this.dataTestid, 'timepicker-toggle') }, h("tk-toggle-button", { key: "AM", type: this.headerType === 'primary' || this.headerType === 'dark' ? 'filled' : 'text', variant: this.headerType === 'dark' ? 'neutral' : 'primary', value: "AM", label: "AM", size: "small", disabled: isDisabled || AMDisabled, dataTestid: getDataTestId(this.dataTestid, 'timepicker-am') }), h("tk-toggle-button", { key: "PM", type: this.headerType === 'primary' || this.headerType === 'dark' ? 'filled' : 'text', variant: this.headerType === 'dark' ? 'neutral' : 'primary', value: "PM", label: "PM", size: "small", disabled: isDisabled || PMDisabled, dataTestid: getDataTestId(this.dataTestid, 'timepicker-pm') })))), h("div", { class: classNames('tk-datepicker-timepicker-body', ['primary', 'light', 'dark'].includes(this.headerType) && `tk-datepicker-timepicker-body-${this.headerType}`, this.weeksLength === 4 && 'tk-datepicker-timepicker-body-4-weeks', this.timeOnly && 'tk-datepicker-timepicker-body-only'), style: {
                height: this.calendarTableHeightPx ? `${this.calendarTableHeightPx}px` : undefined,
            }, "data-testid": getDataTestId(this.dataTestid, 'timepicker-body') }, h("div", { class: "tk-datepicker-timepicker-col", "data-testid": getDataTestId(this.dataTestid, 'timepicker-hours') }, h("div", null, h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_less", "onTk-click": this.handleDecreaseHour, disabled: isMinHour || prevHourDisabled || isDisabled, dataTestid: getDataTestId(this.dataTestid, 'timepicker-hours-decrease') }), h("div", { class: classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-hours-separator-top') })), visibleHours.map((hour, hourIndex) => hour === null ? (h("div", { class: classNames('tk-datepicker-timepicker-value tk-datepicker-timepicker-value-empty', { disabled: isDisabled }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-hours-value-empty', hourIndex.toString()) })) : (h("div", { class: classNames('tk-datepicker-timepicker-value', {
                'selected': hour === currentHour,
                'tk-datepicker-timepicker-value-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-value-primary': this.headerType === 'primary',
                'disabled': isDisabled || isHourDisabled(hour),
            }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-hour', hour.toString()), onClick: () => {
                if (isDisabled || isHourDisabled(hour)) {
                    return;
                }
                else if (this.timeFormat === '12') {
                    const hour24 = (hour === 12 ? 0 : hour) + (this.internalAmPm === 'PM' ? 12 : 0);
                    this.handleHourClick(hour24);
                }
                else {
                    this.handleHourClick(hour);
                }
            } }, hour.toString().padStart(2, '0')))), h("div", null, h("div", { class: classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-hours-separator-bottom') }), h("tk-button", { variant: this.headerType === 'primary' || this.headerType === 'dark' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_more", "onTk-click": this.handleIncreaseHour, disabled: isMaxHour || nextHourDisabled || isDisabled, dataTestid: getDataTestId(this.dataTestid, 'timepicker-hours-increase') }))), h("div", { class: "tk-datepicker-timepicker-col", "data-testid": getDataTestId(this.dataTestid, 'timepicker-minutes') }, h("div", null, h("tk-button", { variant: this.headerType === 'dark' || this.headerType === 'primary' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_less", "onTk-click": this.handleDecreaseMinute, disabled: isMinMinute || prevMinuteDisabled || isDisabled, dataTestid: getDataTestId(this.dataTestid, 'timepicker-minutes-decrease') }), h("div", { class: classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-minutes-separator-top') })), visibleMinutes.map((m, minuteIndex) => m === null ? (h("div", { class: classNames('tk-datepicker-timepicker-value tk-datepicker-timepicker-value-empty', { disabled: isDisabled }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-minutes-value-empty', minuteIndex.toString()) })) : (h("div", { class: classNames('tk-datepicker-timepicker-value', {
                'selected': m === currentMinute,
                'tk-datepicker-timepicker-value-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-value-primary': this.headerType === 'primary',
                'disabled': isDisabled || isMinuteDisabled(m),
            }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-minute', m.toString()), onClick: () => {
                if (isDisabled || isMinuteDisabled(m))
                    return;
                this.handleMinuteClick(m);
            } }, String(m).padStart(2, '0')))), h("div", null, h("div", { class: classNames('tk-datepicker-timepicker-separator', {
                'tk-datepicker-timepicker-separator-dark': this.headerType === 'dark',
                'tk-datepicker-timepicker-separator-primary': this.headerType === 'primary',
            }), "data-testid": getDataTestId(this.dataTestid, 'timepicker-minutes-separator-bottom') }), h("tk-button", { variant: this.headerType === 'dark' || this.headerType === 'primary' ? 'white' : 'neutral', type: "text", size: "base", icon: "expand_more", "onTk-click": this.handleIncreaseMinute, disabled: isMaxMinute || nextMinuteDisabled || isDisabled, dataTestid: getDataTestId(this.dataTestid, 'timepicker-minutes-increase') }))))));
    }
    createLoading() {
        return (h("div", { class: "tk-datepicker-loading", "data-testid": getDataTestId(this.dataTestid, 'loading') }, h("tk-spinner", { size: this.size, dataTestid: getDataTestId(this.dataTestid, 'loading-spinner') })));
    }
    renderInput() {
        if (this.inline)
            return null;
        const shouldUseMask = !this.disableMask && this.mode !== 'range';
        const maskOptionsToPass = shouldUseMask ? this.maskOptions : undefined;
        return (h("tk-input", { ref: el => (this.inputRef = el), label: this.label, size: this.size, mode: "text", icon: this.icon, iconPosition: this.iconPosition, class: classNames('tk-datepicker-input', { 'tk-table-input': this.el.classList.contains('tk-table-datepicker') }), name: this.name, hint: this.hint, clearable: this.clearable, loading: this.loading, disabled: this.disabled, invalid: this.invalid || this.isInvalid, readonly: this.readonly, error: this.error, placeholder: this.placeholder || (this.timeOnly ? this.getOnlyTimeFormat() : this.showTimePicker ? this.getFullDateTimeFormat() : this.dateFormat).toUpperCase(), value: this.inputValue, maskOptions: maskOptionsToPass, "onTk-change": this.handleInputChange, "onTk-clear-click": this.handleInputClearClick, onInput: this.handleInputInput, onKeyDown: this.handleInputKeyDown, onClick: this.handleInputClick, "onTk-blur": this.handleInputBlur, "aria-expanded": !!this.isOpen, "aria-haspopup": "true", "data-tk-datepicker-id": this.uniqueId, showAsterisk: this.showAsterisk, dataTestid: getDataTestId(this.dataTestid, 'input') }));
    }
    renderPanel() {
        if (!this.isOpen && !this.inline)
            return null;
        const panelClasses = classNames('tk-datepicker-panel', {
            'tk-datepicker-panel-inline': this.inline,
        });
        const bodyClasses = classNames('tk-datepicker-body', {
            'tk-datepicker-months-view': this.currentView === 'months',
            'tk-datepicker-years-view': this.currentView === 'years',
        });
        // Time-only mode: render only time picker
        if (this.timeOnly) {
            return (h("div", { class: panelClasses, ref: el => (this.panelRef = el), role: !this.inline ? 'dialog' : undefined, "aria-modal": "true", "data-tk-datepicker-id": this.uniqueId, "data-testid": getDataTestId(this.dataTestid, 'panel') }, h("div", { class: "tk-datepicker-panel-inner", "data-testid": getDataTestId(this.dataTestid, 'panel-inner') }, this.createTimePicker()), this.createFooter()));
        }
        // Default: date (calendar), optionally alongside time picker
        return (h("div", { class: panelClasses, ref: el => (this.panelRef = el), role: !this.inline ? 'dialog' : undefined, "aria-modal": "true", "data-tk-datepicker-id": this.uniqueId, style: { visibility: this.concealUntilMeasured ? 'hidden' : undefined }, "data-testid": getDataTestId(this.dataTestid, 'panel') }, h("div", { class: "tk-datepicker-panel-inner", "data-testid": getDataTestId(this.dataTestid, 'panel-inner') }, h("div", { class: "tk-datepicker-calendar-container", "data-testid": getDataTestId(this.dataTestid, 'calendar') }, this.createHeader(), h("div", { class: bodyClasses, "data-testid": getDataTestId(this.dataTestid, 'body') }, this.loading ? (this.createLoading()) : (h("table", { class: "tk-datepicker-table", "data-testid": getDataTestId(this.dataTestid, 'table') }, this.currentView === 'days' && (h(Fragment, null, this.createWeekDayNames(), this.createWeekDays())), this.currentView === 'months' && this.createMonths(), this.currentView === 'years' && this.createYears())))), this.showTimePicker && this.createTimePicker()), this.createFooter()));
    }
    render() {
        return (h("div", { key: '80b177ebfad0d28bc21c37aaa1991158c83cfc26', class: "tk-datepicker-container", "data-testid": getDataTestId(this.dataTestid, 'container') }, this.renderInput(), this.renderPanel()));
    }
    static get is() { return "tk-datepicker"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-datepicker.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-datepicker.css"]
        };
    }
    static get properties() {
        return {
            "value": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IDateSelection",
                    "resolved": "IDateSelection | string",
                    "references": {
                        "IDateSelection": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-datepicker/types.ts::IDateSelection",
                            "referenceLocation": "IDateSelection"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The value representing the selected date(s)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the label for the input"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "base"
                        }],
                    "text": "Defines the size for the label"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the datepicker is disabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the datepicker is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "readonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the datepicker is read-only"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "readonly",
                "defaultValue": "false"
            },
            "clearable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input of datepicker can be cleared"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "clearable",
                "defaultValue": "false"
            },
            "loading": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Represents whether the datepicker is in a loading state.\nIf true, renders a spinner inside the input and panel."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "loading",
                "defaultValue": "false"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Error message to display"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Hint text to display"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon name to be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon",
                "defaultValue": "'calendar_month'"
            },
            "iconPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "left"
                        }],
                    "text": "Defines the position of the icon."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon-position",
                "defaultValue": "'left'"
            },
            "minDate": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Minimum selectable date"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "min-date",
                "defaultValue": "''"
            },
            "maxDate": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Maximum selectable date"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max-date",
                "defaultValue": "''"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The name of the control."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name"
            },
            "allowedDates": {
                "type": "unknown",
                "mutable": true,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Array of dates that are allowed to be selected. All other dates will be disabled.\n\nNote: Format should match dateFormat prop"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "disabledDates": {
                "type": "unknown",
                "mutable": true,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Array of dates that are disabled for selection.\n\nFormat should match dateFormat prop"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "inline": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether to display inline panel"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "inline",
                "defaultValue": "false"
            },
            "allowApplyButton": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether to require manual confirmation (Apply button) before committing changes.\nIf true, changes are only applied when apply() is called."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "allow-apply-button",
                "defaultValue": "false"
            },
            "mode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'single' | 'range'",
                    "resolved": "\"range\" | \"single\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "single"
                        }],
                    "text": "The selection mode of the date picker: 'single' for single date selection, 'range' for date range selection."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mode",
                "defaultValue": "'single'"
            },
            "locale": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "en"
                        }],
                    "text": "Locale for date formatting"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "locale",
                "defaultValue": "'en'"
            },
            "footerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light'",
                    "resolved": "\"basic\" | \"divided\" | \"light\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "basic"
                        }],
                    "text": "The visual variant of the footer: 'basic', 'divided', or 'light'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "footer-type",
                "defaultValue": "'basic'"
            },
            "dateFormat": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "yyyy-MM-dd"
                        }],
                    "text": "Date format pattern"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "date-format",
                "defaultValue": "'yyyy-MM-dd'"
            },
            "disableMask": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether to disable input mask"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disable-mask",
                "defaultValue": "false"
            },
            "headerType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light' | 'primary' | 'dark'",
                    "resolved": "\"basic\" | \"dark\" | \"divided\" | \"light\" | \"primary\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "basic"
                        }],
                    "text": "Header visual variant"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header-type",
                "defaultValue": "'basic'"
            },
            "placeholder": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Input placeholder text"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "placeholder"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "disabledWeekDays": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "number[]",
                    "resolved": "number[]",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Disabled week days (0-6, where 0 is Sunday and 6 is Saturday)\nExample: [0,6] will disable Sunday and Saturday"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "showTimePicker": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether to show the timepicker panel next to the calendar."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-time-picker",
                "defaultValue": "false"
            },
            "timeOnly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Enables time-only mode. In this mode, no date selection is required and the input shows a time mask.\nWhen enabled, the panel renders only the time picker and `tk-change` emits a time string (e.g. HH:mm or hh:mm a)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "time-only",
                "defaultValue": "false"
            },
            "minTime": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Minimum selectable time (HH:mm, HH:mm AA format)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "min-time"
            },
            "maxTime": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Maximum selectable time (HH:mm, HH:mm AA format)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max-time"
            },
            "hourStep": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "1"
                        }],
                    "text": "Hour increment step."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hour-step",
                "defaultValue": "1"
            },
            "minuteStep": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "1"
                        }],
                    "text": "Minute increment step."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "minute-step",
                "defaultValue": "1"
            },
            "timeFormat": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'12' | '24'",
                    "resolved": "\"12\" | \"24\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'24'"
                        }],
                    "text": "Time format: '12' or '24'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "time-format",
                "defaultValue": "'24'"
            },
            "firstDayOfWeekIndex": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Defines the first day of the week. 0 for Monday, 1 for Tuesday, ..., 6 for Sunday.\nIf not provided, the first day of the week is determined by the `locale` prop.\nIf neither `firstDayOfWeekIndex` nor `locale` provide a value, defaults to Monday (0).\nProviding this prop overrides the locale setting for the start of the week."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "first-day-of-week-index"
            },
            "defaultDate": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Defines the default year/month to display when the panel is opened.\nAccepted formats:\n- yyyy-MM\n- MM-yyyy\n- yyyy\n- MM"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "default-date"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hasFooterSlot": {},
            "hasFooterActionsSlot": {},
            "currentMonth": {},
            "internalSelectedDates": {},
            "inputValue": {},
            "internalStartTime": {},
            "internalEndTime": {},
            "internalAmPm": {},
            "hoverDate": {},
            "currentView": {},
            "maskOptions": {},
            "isInvalid": {},
            "isOpen": {},
            "concealUntilMeasured": {},
            "calendarTableHeightPx": {}
        };
    }
    static get events() {
        return [{
                "method": "tkInputChange",
                "name": "tk-input-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted on input value changes"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }, {
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted on date selection changes"
                },
                "complexType": {
                    "original": "IDateSelection | string | null",
                    "resolved": "IDateSelection | string",
                    "references": {
                        "IDateSelection": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-datepicker/types.ts::IDateSelection",
                            "referenceLocation": "IDateSelection"
                        }
                    }
                }
            }, {
                "method": "tkInvalid",
                "name": "tk-invalid",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the input is invalid"
                },
                "complexType": {
                    "original": "{ message: string; value: string }",
                    "resolved": "{ message: string; value: string; }",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "setToday": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Sets the date to today",
                    "tags": []
                }
            },
            "apply": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Applies the current internal selection and emits tk-change",
                    "tags": []
                }
            },
            "closePanel": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Closes the datepicker panel if it is open.",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "internalAmPm",
                "methodName": "internalAmPmChanged"
            }, {
                "propName": "isOpen",
                "methodName": "isOpenChanged"
            }, {
                "propName": "value",
                "methodName": "valueChanged"
            }, {
                "propName": "dateFormat",
                "methodName": "dateFormatChanged"
            }, {
                "propName": "showTimePicker",
                "methodName": "showTimePickerChanged"
            }, {
                "propName": "timeOnly",
                "methodName": "timeOnlyChanged"
            }, {
                "propName": "timeFormat",
                "methodName": "timeFormatChanged"
            }];
    }
    static get attachInternalsMemberName() { return "internals"; }
}
