// @takeoff-ui/core@0.11.9/dist/collection/components/tk-phone-input/tk-phone-input.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { v4 as uuidv4 } from "uuid";
import { INTERNAL_COUNTRIES } from "./constants";
import { getIconElementProps } from "../../utils/icon-utils";
import { floatingElementAutoUpdate } from "../../utils/position-utils";
import { applyStyles } from "../../utils/style-utils";
import { renderHint } from "../../utils/hint-utils";
/**
 * The TkPhoneInput component allows users to input phone numbers with country selection and validation.
 * @react `import { TkPhoneInput } from '@takeoff-ui/react'`
 * @vue `import { TkPhoneInput } from '@takeoff-ui/vue'`
 * @angular `import { TkPhoneInput } from '@takeoff-ui/angular'`
 */
export class TkPhoneInput {
    constructor() {
        this.uniqueId = uuidv4();
        /**
         * The list of countries to display in the dropdown.
         */
        this.countries = [];
        /**
         * The current input value with mask applied.
         */
        this.inputValue = '';
        /**
         * Whether the input has been touched.
         */
        this.hasFocus = false;
        /**
         * Whether the country dropdown is open.
         */
        this.isDropdownOpen = false;
        /**
         * The current search term for filtering countries.
         */
        this.searchTerm = '';
        /**
         * Whether the input is disabled.
         * * @defaultValue false
         */
        this.disabled = false;
        /**
         * If `true`, the user cannot modify the value.
         * @defaultValue false
         */
        this.readonly = false;
        /**
         * Indicates whether the input is in an invalid state
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * The default country to select (ISO country code).
         */
        this.defaultCountry = 'TR';
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * Sets size for the component.
         */
        this.size = 'base';
        /**
         * Hides the phone flag icon in the dropdown button and list.
         * @default false
         */
        this.hideFlag = false;
        /**
         * Determines the width of the dropdown. Accepts values like 'match-parent', 'auto', or a specific width in '300px'.
         * @defaultValue match-parent
         */
        this.dropdownWidthMode = 'match-parent';
        /**
         * Toggle the country dropdown.
         */
        this.toggleDropdown = () => {
            var _a;
            if (!this.disabled) {
                this.isDropdownOpen = !this.isDropdownOpen;
                if (this.isDropdownOpen) {
                    (_a = this.searchInput) === null || _a === void 0 ? void 0 : _a.focus();
                }
            }
        };
        /**
         * Handle country selection from the dropdown.
         */
        this.handleCountrySelect = (country) => {
            var _a;
            this.setSelectedCountry(country.id);
            this.value = {
                rawValue: '',
                maskedValue: '',
                country: {
                    id: this.selectedCountry.id,
                    label: this.selectedCountry.label,
                    dialCode: this.selectedCountry.dialCode,
                    mask: this.selectedCountry.mask,
                    placeholder: this.selectedCountry.placeholder,
                },
            };
            this.tkChange.emit(this.value);
            this.inputValue = '';
            this.isDropdownOpen = false;
            this.searchTerm = '';
            (_a = this.inputRef) === null || _a === void 0 ? void 0 : _a.focus();
        };
        /**
         * Event handler for clicks outside the component to close dropdown.
         */
        this.handleClickOutside = (event) => {
            if (!this.el.contains(event.target)) {
                this.isDropdownOpen = false;
            }
        };
        /**
         * Handle changes to the country search input.
         */
        this.handleSearchChange = (event) => {
            this.searchTerm = event.target.value;
        };
        /**
         * Handle changes to the phone number input.
         */
        this.handleInput = (event) => {
            var _a;
            const inputElement = event.target;
            const rawValue = inputElement.value.replace(/\D/g, '');
            if (!((_a = this.selectedCountry.mask) === null || _a === void 0 ? void 0 : _a.length)) {
                this.inputValue = rawValue;
                this.inputRef.value = this.inputValue;
                // reassign the value object and emit the change event before returning. Masked Value consists of raw value
                this.value = {
                    rawValue,
                    maskedValue: rawValue,
                    country: {
                        id: this.selectedCountry.id,
                        label: this.selectedCountry.label,
                        dialCode: this.selectedCountry.dialCode,
                        placeholder: this.selectedCountry.placeholder,
                    },
                };
                this.tkChange.emit(this.value);
                return;
            }
            const currentMask = this.selectedCountry.mask;
            const maxDigits = (currentMask.match(/9/g) || []).length;
            const hasInvalidCharacters = /[^\d() -]/.test(inputElement.value);
            if (rawValue.length > maxDigits || hasInvalidCharacters)
                return (this.inputRef.value = this.inputValue);
            this.hasFocus = false;
            this.inputValue = this.applyMask(rawValue, currentMask);
            this.inputRef.value = this.inputValue;
            this.value = {
                rawValue,
                maskedValue: this.inputValue,
                country: {
                    id: this.selectedCountry.id,
                    label: this.selectedCountry.label,
                    dialCode: this.selectedCountry.dialCode,
                    mask: currentMask,
                    placeholder: this.selectedCountry.placeholder,
                },
            };
            this.tkChange.emit(this.value);
        };
        this.handleInputBlur = () => {
            this.hasFocus = false;
            this.tkBlur.emit();
        };
        this.handleInputFocus = () => {
            this.hasFocus = true;
            this.tkFocus.emit();
        };
        this.handleFormReset = () => {
            var _a;
            this.value = {
                rawValue: '',
                maskedValue: '',
                country: {
                    id: this.selectedCountry.id,
                    label: this.selectedCountry.label,
                    dialCode: this.selectedCountry.dialCode,
                    mask: this.selectedCountry.mask,
                    placeholder: this.selectedCountry.placeholder,
                },
            };
            this.tkChange.emit(this.value);
            this.isDropdownOpen = false;
            this.inputValue = '';
            this.inputRef.value = '';
            this.searchTerm = '';
            (_a = this.inputRef) === null || _a === void 0 ? void 0 : _a.focus();
        };
    }
    valueChanged(newValue) {
        var _a, _b;
        if (!newValue || (typeof newValue === 'object' && Object.keys(newValue).length === 0)) {
            this.handleFormReset();
        }
        if (!newValue.rawValue && !newValue.maskedValue) {
            this.inputValue = '';
        }
        if (newValue) {
            if ((_a = newValue === null || newValue === void 0 ? void 0 : newValue.country) === null || _a === void 0 ? void 0 : _a.id) {
                this.setSelectedCountry((_b = newValue === null || newValue === void 0 ? void 0 : newValue.country) === null || _b === void 0 ? void 0 : _b.id);
            }
            this.inputValue = this.applyMask(newValue.rawValue, this.selectedCountry.mask);
        }
    }
    disabledChanged(newValue) {
        if (newValue)
            this.isDropdownOpen = false;
    }
    /**
     * Add event listeners when the component is connected to the DOM.
     */
    connectedCallback() {
        document.addEventListener('click', this.handleClickOutside);
    }
    /**
     * Remove event listeners when the component is disconnected from the DOM.
     */
    disconnectedCallback() {
        var _a;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
        document.removeEventListener('click', this.handleClickOutside);
    }
    /**
     * Initialize the component before it is rendered.
     */
    componentWillLoad() {
        var _a, _b, _c, _d;
        this.initializeCountries();
        if ((this === null || this === void 0 ? void 0 : this.value) && ((_a = Object.keys(this === null || this === void 0 ? void 0 : this.value)) === null || _a === void 0 ? void 0 : _a.length)) {
            if ((_c = (_b = this.value) === null || _b === void 0 ? void 0 : _b.country) === null || _c === void 0 ? void 0 : _c.id) {
                this.setSelectedCountry(this.value.country.id);
            }
            this.inputValue = this.applyMask((_d = this === null || this === void 0 ? void 0 : this.value) === null || _d === void 0 ? void 0 : _d.rawValue, this === null || this === void 0 ? void 0 : this.selectedCountry.mask);
        }
        else {
            this.setSelectedCountry(this.defaultCountry);
        }
    }
    /**
     * Update the component when properties change.
     */
    componentDidUpdate() {
        var _a, _b, _c;
        if (this.isDropdownOpen) {
            // Clean up old floating UI listeners before setting up new ones
            (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
            this.updatePosition();
        }
        else {
            // Remove floating UI listeners when dropdown closes
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
            (_c = this.panelRef) === null || _c === void 0 ? void 0 : _c.remove();
        }
    }
    formResetCallback() {
        this.handleFormReset();
    }
    updatePosition() {
        const dropdownWidthMode = this.dropdownWidthMode;
        const tkInputRootEl = this.el.querySelector('.tk-phone-input-wrapper');
        this.cleanup = floatingElementAutoUpdate(tkInputRootEl, this.panelRef, undefined, {
            placement: 'bottom-start',
            size: {
                apply({ rects, elements }) {
                    if (dropdownWidthMode === 'match-parent') {
                        applyStyles(elements.floating, {
                            width: `${rects.reference.width}px`,
                        });
                    }
                    else if (dropdownWidthMode !== 'auto' && dropdownWidthMode.length > 0) {
                        applyStyles(elements.floating, {
                            width: dropdownWidthMode,
                        });
                    }
                },
            },
        });
    }
    /**
     * Initialize the list of countries from the provided prop or fallback to internal list.
     */
    initializeCountries() {
        if (this.countryList) {
            this.countries = this.countryList;
        }
        else {
            this.countries = INTERNAL_COUNTRIES;
        }
    }
    /**
     * Set the selected country by its ID.
     */
    setSelectedCountry(countryId) {
        const country = this.countries.find(c => c.id.toUpperCase() === countryId.toUpperCase());
        this.selectedCountry = country || this.countries[0];
    }
    /**
     * Apply the mask to a raw phone number.
     */
    applyMask(rawValue, mask) {
        if (!(mask === null || mask === void 0 ? void 0 : mask.length) || !(rawValue === null || rawValue === void 0 ? void 0 : rawValue.length))
            return rawValue;
        let maskedValue = '';
        let digitIndex = 0;
        for (let i = 0; i < mask.length && digitIndex < rawValue.length; i++) {
            if (mask[i] === '9') {
                maskedValue += rawValue[digitIndex++];
            }
            else {
                maskedValue += mask[i];
            }
        }
        return maskedValue;
    }
    /**
     * Get the filtered list of countries based on the search term.
     */
    getFilteredCountries() {
        const term = this.searchTerm.toLowerCase();
        if (!term)
            return this.countries;
        return this.countries.filter(country => { var _a; return country.label.toLowerCase().includes(term) || ((_a = country.dialCode) === null || _a === void 0 ? void 0 : _a.includes(term)); });
    }
    /**
     * Get the flag class based on the country object.
     */
    getFlagClass(country) {
        return classNames('flag', (country === null || country === void 0 ? void 0 : country.dialCode) ? `flag-${country === null || country === void 0 ? void 0 : country.id.toLowerCase()}` : 'flag-none');
    }
    renderLabel() {
        var _a;
        let label;
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            const asterisk = h("span", { class: "tk-phone-input-asterisk" }, "*");
            label = (h("label", { htmlFor: this.uniqueId, class: "tk-phone-input-label" }, this.label, this.showAsterisk ? asterisk : ''));
        }
        return label;
    }
    /**
     * Create the country selector dropdown.
     */
    renderCountrySelector() {
        return (h("div", { class: "tk-phone-input-dropdown" }, this.createDropdownButton(), this.isDropdownOpen && (h("div", { class: "tk-phone-input-dropdown-menu", role: "listbox", ref: el => (this.panelRef = el) }, this.createDropdownSearch(), this.createDropdownList()))));
    }
    /**
     * Create the dropdown button for selecting a country.
     */
    createDropdownButton() {
        const selectedClass = classNames('tk-phone-input-dropdown-button-selected', {
            'tk-phone-input-no-dial-code': !this.selectedCountry.dialCode,
        });
        return (h("button", { class: "tk-phone-input-dropdown-button", onClick: this.toggleDropdown, type: "button" }, h("div", { class: selectedClass }, h("tk-icon", Object.assign({}, getIconElementProps('stat_minus_1', { variant: null, size: 'large' }, 'rounded', 'span'))), !this.hideFlag && this.renderFlag(this.selectedCountry), this.selectedCountry.dialCode && h("span", { class: "tk-phone-input-dial-code" }, this.selectedCountry.dialCode))));
    }
    /**
     * Create the search input for filtering countries in the dropdown.
     */
    createDropdownSearch() {
        return (h("tk-input", { class: "tk-phone-input-dropdown-menu-search", size: this.size, placeholder: "Search", value: this.searchTerm, "onTk-change": this.handleSearchChange, ref: el => (this.searchInput = el), icon: "search", iconPosition: "right", onClick: (e) => e.stopPropagation() }));
    }
    /**
     * Create the dropdown list of countries.
     */
    createDropdownList() {
        return (h("ul", { class: "tk-phone-input-menu" }, this.getFilteredCountries().map(country => (h("li", { class: "tk-phone-input-menu-item", onClick: () => this.handleCountrySelect(country), key: country.id, role: "option", "aria-selected": country.id === this.selectedCountry.id }, !this.hideFlag && this.renderFlag(country), country.label && h("span", { class: "tk-phone-input-menu-country-label" }, country.label), country.dialCode && h("span", { class: "tk-phone-input-menu-dial-id" }, country.dialCode))))));
    }
    /**
     * Create the phone number input field.
     */
    renderPhoneInput() {
        var _a, _b;
        return (h("input", { id: this.uniqueId, type: "tel", class: "tk-phone-input-input", autoComplete: "off", placeholder: ((_a = this.selectedCountry) === null || _a === void 0 ? void 0 : _a.placeholder) || this.placeholder || ((_b = this.selectedCountry.mask) === null || _b === void 0 ? void 0 : _b.replace(/9/g, '9')), value: this.inputValue, onInput: this.handleInput, onBlur: this.handleInputBlur, onFocus: this.handleInputFocus, disabled: this.disabled, ref: el => (this.inputRef = el) }));
    }
    /**
     * Render the flag element for a country. Shows a close icon for countries without a dial code.
     */
    renderFlag(country) {
        const flagClass = this.getFlagClass(country);
        return country.dialCode ? (h("div", { class: flagClass, "aria-label": `${country.label} flag` })) : (h("div", { class: flagClass }, h("tk-icon", Object.assign({}, getIconElementProps('close', { color: 'var(--static-white)' })))));
    }
    /**
     * Render the component.
     */
    render() {
        return (h("div", { key: '7622740fa83a9e25aca50bed39634dfe9f12b578', class: classNames('tk-phone-input-container', `tk-phone-input-container-${this.size}`), "aria-invalid": this.invalid, "aria-disabled": this.disabled, "aria-readonly": this.readonly }, this.renderLabel(), h("div", { key: '86a1c6d319cea43d53824732efb1095244c247bb', class: "tk-phone-input-wrapper" }, this.renderCountrySelector(), this.renderPhoneInput()), renderHint(this.hint, this.error, this.invalid)));
    }
    static get is() { return "tk-phone-input"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-phone-input.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-phone-input.css"]
        };
    }
    static get properties() {
        return {
            "value": {
                "type": "any",
                "mutable": true,
                "complexType": {
                    "original": "IPhoneInputValue | any",
                    "resolved": "any",
                    "references": {
                        "IPhoneInputValue": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-phone-input/types.ts::IPhoneInputValue",
                            "referenceLocation": "IPhoneInputValue"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the phone input.\nThis is a list of phone input data objects.\nIt can be mutable to allow two-way binding."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The label for the phone input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "countryList": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ICountry[]",
                    "resolved": "ICountry[]",
                    "references": {
                        "ICountry": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-phone-input/types.ts::ICountry",
                            "referenceLocation": "ICountry"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The list of countries to display in the dropdown.\nCan be provided as an array of Country objects or as a JSON string."
                },
                "getter": false,
                "setter": false
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the input is disabled.\n*"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "readonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If `true`, the user cannot modify the value."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "readonly",
                "defaultValue": "false"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the input is in an invalid state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "defaultCountry": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The default country to select (ISO country code)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "default-country",
                "defaultValue": "'TR'"
            },
            "placeholder": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Placeholder text for the phone input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "placeholder"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Provided a hint or additional information about the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This is the error message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "hideFlag": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "false"
                        }],
                    "text": "Hides the phone flag icon in the dropdown button and list."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-flag",
                "defaultValue": "false"
            },
            "dropdownWidthMode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'match-parent' | 'auto' | string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "match-parent"
                        }],
                    "text": "Determines the width of the dropdown. Accepts values like 'match-parent', 'auto', or a specific width in '300px'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "dropdown-width-mode",
                "defaultValue": "'match-parent'"
            }
        };
    }
    static get states() {
        return {
            "countries": {},
            "selectedCountry": {},
            "inputValue": {},
            "hasFocus": {},
            "isDropdownOpen": {},
            "searchTerm": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": false,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the value has changed."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "tkBlur",
                "name": "tk-blur",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the input loses focus."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "tkFocus",
                "name": "tk-focus",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the input has focus."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "valueChanged"
            }, {
                "propName": "disabled",
                "methodName": "disabledChanged"
            }];
    }
}
