// @takeoff-ui/core@0.11.9/dist/collection/components/tk-spinner/tk-spinner.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkSpinner` component description.
 * @react `import { TkSpinner } from @takeoff-ui/react`
 * @vue `import { TkSpinner } from @takeoff-ui/vue`
 * @angular `import { TkSpinner } from @takeoff-ui/angular`
 */
export class TkSpinner {
    constructor() {
        this.isButton = false;
        /**
         * Controls the orientation of the spinner component.
         * @defaultValue 'horizontal'
         */
        this.orientation = 'vertical';
        /**
         * Controls the size of the spinner component.
         * @defaultValue 'base'
         */
        this.size = 'base';
        /**
         * Sets the style of spinner component.
         * @defaultValue 'rounded'
         */
        this.type = 'rounded';
        /**
         * Sets the color variant of spinner component.
         * @defaultValue 'neutral'
         */
        this.variant = 'neutral';
    }
    componentWillLoad() {
        var _a;
        this.isButton = (_a = this.el.closest('button')) === null || _a === void 0 ? void 0 : _a.classList.contains('tk-button');
    }
    renderSpinner() {
        switch (this.type) {
            case 'rounded':
                if (this.isButton) {
                    const button = this.el.closest('button');
                    const borderColor = window.getComputedStyle(button).color;
                    return h("div", { style: { borderColor: borderColor, borderTopColor: 'transparent' }, class: "spinner-rounded", "data-testid": getDataTestId(this.dataTestid, 'rounded') });
                }
                return h("div", { class: "spinner-rounded", "data-testid": getDataTestId(this.dataTestid, 'rounded') });
            case 'dots':
                return (h("div", { class: "spinner-dots", "data-testid": getDataTestId(this.dataTestid, 'dots') }, h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-0') }), h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-1') }), h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-2') }), h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-3') }), h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-4') }), h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-5') }), h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-6') }), h("div", { class: "dot", "data-testid": getDataTestId(this.dataTestid, 'dot-7') })));
            case 'lines':
                return (h("div", { class: "spinner-lines", "data-testid": getDataTestId(this.dataTestid, 'lines') }, h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-0') }), h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-1') }), h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-2') }), h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-3') }), h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-4') }), h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-5') }), h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-6') }), h("div", { class: "line", "data-testid": getDataTestId(this.dataTestid, 'line-7') })));
            case 'three-dots':
                return (h("div", { class: "spinner-three-dots", "data-testid": getDataTestId(this.dataTestid, 'three-dots') }, h("span", { class: "dot1", "data-testid": getDataTestId(this.dataTestid, 'dot-0') }), h("span", { class: "dot2", "data-testid": getDataTestId(this.dataTestid, 'dot-1') }), h("span", { class: "dot3", "data-testid": getDataTestId(this.dataTestid, 'dot-2') })));
            case 'pulse':
                return h("div", { class: "spinner-pulse", "data-testid": getDataTestId(this.dataTestid, 'pulse') });
            case 'loader':
                return h("div", { class: "spinner-loader", "data-testid": getDataTestId(this.dataTestid, 'loader') });
            case 'logo':
                return (h("div", { class: "spinner-logo", "data-testid": getDataTestId(this.dataTestid, 'logo') }, h("svg", { viewBox: "0 0 48 48", fill: "none", xmlns: "http://www.w3.org/2000/svg", "data-testid": getDataTestId(this.dataTestid, 'logo-icon') }, h("path", { "fill-rule": "evenodd", "clip-rule": "evenodd", d: "M24 0C10.7452 0 0 10.7452 0 24C0 37.2548 10.7452 48 24 48C37.2548 48 48 37.2548 48 24V4C48 1.79086 46.2091 0 44 0H24ZM24 8C15.1634 8 8 15.1634 8 24C8 32.8366 15.1634 40 24 40C32.8366 40 40 32.8366 40 24C40 15.1634 32.8366 8 24 8Z" }))));
            default:
                return null;
        }
    }
    render() {
        return (h("div", { key: '13590c7e676f5573f94dbc287ee9ba433800ca8b', role: "progressbar", class: classNames('tk-spin-container', {
                [this.orientation]: true,
                [this.size]: true,
                [this.variant]: true,
            }), "data-testid": getDataTestId(this.dataTestid, 'container') }, this.renderSpinner(), this.label && (h("div", { key: '98a9ab07791bc77380c982f33d56449b10907d3c', class: "tk-spinner-label", "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label))));
    }
    static get is() { return "tk-spinner"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-spinner.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-spinner.css"]
        };
    }
    static get properties() {
        return {
            "orientation": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'horizontal' | 'vertical'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'horizontal'"
                        }],
                    "text": "Controls the orientation of the spinner component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "orientation",
                "defaultValue": "'vertical'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'xxsmall' | 'xsmall' | 'small' | 'base' | 'large' | 'xlarge'",
                    "resolved": "\"base\" | \"large\" | \"small\" | \"xlarge\" | \"xsmall\" | \"xxsmall\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Controls the size of the spinner component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'rounded' | 'dots' | 'lines' | 'pulse' | 'three-dots' | 'loader' | 'logo'",
                    "resolved": "\"dots\" | \"lines\" | \"loader\" | \"logo\" | \"pulse\" | \"rounded\" | \"three-dots\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'rounded'"
                        }],
                    "text": "Sets the style of spinner component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'rounded'"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'neutral' | 'info' | 'success' | 'warning' | 'danger'",
                    "resolved": "\"danger\" | \"info\" | \"neutral\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'neutral'"
                        }],
                    "text": "Sets the color variant of spinner component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'neutral'"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Sets the label of the spinner component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get elementRef() { return "el"; }
}
