// @takeoff-ui/core@0.11.9/dist/collection/components/tk-upload/tk-upload.js
import { h, Fragment } from "@stencil/core";
import classNames from "classnames";
import mime from "mime";
import { filesize } from "filesize";
import { getIconElementProps } from "../../utils/icon-utils";
import { renderHint } from "../../utils/hint-utils";
/**
 * The TkUpload component is an interface element that allows users to select and upload files from their devices to a server or a target location. It typically includes a "Choose File" button and a field displaying the selected file's name. This component simplifies the process of file selection and uploading.
 * @react `import { TkUpload } from '@takeoff-ui/react'`
 * @vue `import { TkUpload } from '@takeoff-ui/vue'`
 * @angular `import { TkUpload } from '@takeoff-ui/angular'`
 */
export class TkUpload {
    constructor() {
        this.errorMessages = [];
        this.isDragOver = false;
        /**
         * If `autoUpload` is set to `true`, the upload button will be hidden,
         * and the `tkUpload` event will be automatically triggered for each newly added file.
         */
        this.autoUpload = false;
        /**
         * Enables drag and drop functionality for file uploads.
         * @defaultValue true
         */
        this.dragDrop = true;
        /**
         * Displays a red asterisk (*) next to the label for visual emphasis.
         */
        this.showAsterisk = false;
        /**
         * Displays the uploaded files.
         * @defaultValue true
         */
        this.showFiles = true;
        /**
         * Label text displayed inside the choose button.
         */
        this.chooseButtonLabel = 'Choose File';
        /**
         * Label text displayed inside the upload button.
         */
        this.uploadButtonLabel = 'Upload';
        /**
         * Acceptable file types for upload. Use MIME types or extensions separated by commas.
         */
        this.accept = '*';
        /**
         * Disables the input, preventing user interaction.
         */
        this.disabled = false;
        /**
         * Allows multiple file selection.
         */
        this.multiple = false;
        /**
         * Title displayed in the upload component.
         */
        this.title = 'Choose a file or drag & drop it here.';
        /**
         * Title displayed in the upload component when drag and drop is active.
         */
        this.dragDropTitle = 'Drop files here';
        /**
         * The file value of the upload.
         */
        this.value = [];
        /**
         * Description displayed under the title.
         */
        this.description = 'JPEG, PNG, PDF and MP4 formats, up to 50 MB.';
        /**
         * Description displayed under the title when drag and drop is active.
         */
        this.dragDropDescription = 'Release to upload files';
        /**
         * Indicates whether the upload is in an invalid state, uploads will fail eventually
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * Indicates whether the upload is in an loading state
         * @defaultValue false
         */
        this.loading = false;
        /**
         * Type of the file upload area.
         */
        this.type = 'default';
        /**
         * Indicates whether the files can be downloaded. When true, a download button will be displayed next to each file.
         */
        this.downloadable = false;
        this.handleDragEnter = (e) => {
            if (!this.dragDrop || this.disabled)
                return;
            e.preventDefault();
            e.stopPropagation();
            this.isDragOver = true;
        };
        this.handleDragLeave = (e) => {
            if (!this.dragDrop || this.disabled)
                return;
            e.preventDefault();
            e.stopPropagation();
            // Only set isDragOver to false if we're leaving the dropzone entirely
            const rect = e.currentTarget.getBoundingClientRect();
            const x = e.clientX;
            const y = e.clientY;
            if (x < rect.left || x > rect.right || y < rect.top || y > rect.bottom) {
                this.isDragOver = false;
            }
        };
        this.handleDragOver = (e) => {
            if (!this.dragDrop || this.disabled)
                return;
            e.preventDefault();
            e.stopPropagation();
            this.isDragOver = true;
        };
        this.handleDrop = (e) => {
            var _a;
            if (!this.dragDrop || this.disabled)
                return;
            e.preventDefault();
            e.stopPropagation();
            this.isDragOver = false;
            const files = (_a = e.dataTransfer) === null || _a === void 0 ? void 0 : _a.files;
            if (files && files.length > 0) {
                this.handleProcessFiles(files);
            }
        };
    }
    valueChanged(newValue) {
        if (!newValue || (newValue === null || newValue === void 0 ? void 0 : newValue.length) <= 0)
            this.inputRef.value = null;
    }
    isImageFile(file) {
        const IMAGE_TYPES = ['image/jpeg', 'image/png', 'image/gif', 'image/webp'];
        return IMAGE_TYPES.includes(file.type);
    }
    handleProcessFiles(files) {
        const rejectedFiles = [];
        const isFileCountRejected = this.maxFileCount && this.value.length + files.length > this.maxFileCount;
        const acceptedFiles = [];
        // Validate file count
        if (isFileCountRejected) {
            rejectedFiles.push({
                reason: `Max file count exceeded, max ${this.maxFileCount} files allowed.`,
                file: files,
            });
        }
        else {
            Array.from(files).forEach(file => {
                // Validate file size
                if (file.size > this.maxFileSize) {
                    rejectedFiles.push({
                        reason: `File size exceeds the maximum allowed size of ${this.maxFileSize / (1024 * 1024)} MB.`,
                        file,
                    });
                    return;
                }
                // Validate file type
                if (this.accept !== '*' && !this.accept.includes(file.type)) {
                    rejectedFiles.push({
                        reason: `Invalid file type. Accepted types are: ${this.accept}.`,
                        file,
                    });
                    return;
                }
                // If valid, add to accepted files
                if (this.multiple)
                    this.value.push(file);
                else
                    this.value = [file];
                acceptedFiles.push(file);
            });
        }
        // Update state and emit events
        this.errorMessages = rejectedFiles.map(item => (item.file instanceof File ? `${item.file.name}: ${item.reason}` : item.reason));
        if (!isFileCountRejected && acceptedFiles.length > 0) {
            this.tkChange.emit(this.value);
            if (this.autoUpload)
                this.handleUploadButtonClick();
        }
        if (rejectedFiles.length > 0) {
            this.tkFilesRejected.emit(rejectedFiles);
        }
    }
    handleInputChange(e) {
        const input = e.target;
        const files = input.files;
        if (files && files.length > 0) {
            this.handleProcessFiles(files);
        }
    }
    handleDownloadButtonClick(file) {
        this.tkDownloadFile.emit(file);
    }
    handleRemoveButtonClick(file) {
        const removeIndex = this.value.findIndex(item => file.lastModified == item.lastModified && file.name == item.name);
        const removedFile = this.value.splice(removeIndex, 1);
        this.value = [...this.value];
        this.inputRef.value = null;
        this.tkChange.emit(this.value);
        this.tkRemovedFile.emit(removedFile[0]);
    }
    handleUploadButtonClick() {
        this.tkUpload.emit(this.value);
    }
    renderFilePreview(file) {
        if (this.isImageFile(file)) {
            return h("img", { src: URL.createObjectURL(file), width: 50, height: 50 });
        }
        else {
            let ext = mime.getExtension(file.type);
            if (!ext) {
                const tmpArr = file.name.split('.');
                ext = tmpArr[tmpArr.length - 1];
            }
            return (h(Fragment, null, h("svg", { width: "32", height: "32", viewBox: "0 0 32 32", fill: "none", xmlns: "http://www.w3.org/2000/svg" }, h("path", { d: "M23 1V7.66667C23 8.40333 23.5967 9 24.3333 9H31", stroke: "#CACFD8", "stroke-width": "1.5" }), h("path", { d: "M21.5147 0.75C22.9071 0.75 24.2425 1.30312 25.227 2.28769L29.7123 6.77297C30.6969 7.75754 31.25 9.09289 31.25 10.4853V26C31.25 28.8995 28.8995 31.25 26 31.25H12C9.1005 31.25 6.75 28.8995 6.75 26V6C6.75 3.10051 9.10051 0.75 12 0.75H21.5147Z", stroke: "#CACFD8", "stroke-width": "1.5" })), h("div", { class: classNames('extension-text', ext) }, ext === null || ext === void 0 ? void 0 : ext.toUpperCase())));
        }
    }
    renderState() {
        let iconName = 'check_circle';
        let state = 'added';
        let iconVariant = 'success';
        if (this.loading) {
            state = 'loading';
            iconName = 'progress_activity';
        }
        else if (this.invalid) {
            state = 'failed';
            iconName = 'dangerous';
            iconVariant = 'danger';
        }
        return (h("label", { class: classNames('tk-upload-state', state) }, state == 'loading' ? (h("tk-spinner", { type: "three-dots", size: "xsmall" })) : (h("tk-icon", Object.assign({}, getIconElementProps(iconName, { variant: iconVariant }, 'rounded', 'span')))), state));
    }
    renderDropzone() {
        var _a;
        const dropzoneClasses = classNames('tk-upload-dropzone', this.type, {
            'drag-over': this.isDragOver,
            'disabled': this.disabled,
        });
        const dropzoneProps = this.dragDrop && !this.disabled
            ? {
                onDragEnter: this.handleDragEnter,
                onDragLeave: this.handleDragLeave,
                onDragOver: this.handleDragOver,
                onDrop: this.handleDrop,
            }
            : {};
        return (h("div", Object.assign({ class: dropzoneClasses }, dropzoneProps), h("div", { class: "tk-upload-icon" }, h("tk-icon", Object.assign({}, getIconElementProps('file_upload', { size: 'xxlarge' }, 'rounded', 'span')))), h("div", { class: "tk-upload-content" }, h("div", { class: "tk-upload-text-holder" }, h("div", { class: "tk-upload-title" }, this.isDragOver ? this.dragDropTitle : this.title), h("div", { class: "tk-upload-description" }, this.isDragOver ? this.dragDropDescription : this.description)), h("div", { class: "tk-upload-input" }, h("tk-button", { label: this.chooseButtonLabel, variant: "neutral", type: "outlined", icon: "folder", disabled: this.disabled, "onTk-click": () => this.inputRef.click() }), !this.autoUpload && (h("tk-button", { label: this.uploadButtonLabel, variant: "neutral", type: "outlined", icon: "file_upload", loading: this.loading, disabled: !(((_a = this.value) === null || _a === void 0 ? void 0 : _a.length) > 0) || this.loading, "onTk-click": this.handleUploadButtonClick.bind(this) })), h("input", { ref: el => (this.inputRef = el), type: "file", accept: this.accept, multiple: this.multiple ? true : undefined, onChange: e => this.handleInputChange(e), style: { display: 'none' } })))));
    }
    renderFileholder() {
        var _a;
        return (h("div", { class: "tk-upload-file-holder" }, (_a = this.value) === null || _a === void 0 ? void 0 : _a.map((item, index) => (h("div", { class: "tk-upload-file-item", key: index }, h("div", { class: "tk-upload-file-preview" }, this.renderFilePreview(item)), h("div", { class: "tk-upload-file-content" }, h("div", { class: "tk-upload-file-info" }, h("div", { class: "tk-upload-file-name" }, item.name), h("div", { class: "tk-upload-file-size-state" }, h("span", { class: "tk-upload-size" }, filesize(item.size, { standard: 'jedec' })), this.renderState())), h("div", { class: "tk-upload-file-buttons" }, this.downloadable && (h("tk-button", { variant: "neutral", type: "outlined", icon: "file_download", size: "small", "onTk-click": () => this.handleDownloadButtonClick(item) })), h("tk-button", { variant: "neutral", type: "outlined", icon: "close", size: "small", "onTk-click": () => this.handleRemoveButtonClick(item) }))))))));
    }
    render() {
        var _a;
        let label;
        const rootClasses = classNames('tk-upload-container', {
            'drag-drop-enabled': this.dragDrop,
        });
        if (((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            const asterisk = h("span", { key: '436cb31d453cd1ad0e44722e7d8319ed8713cc48', class: "asterisk" }, "*");
            label = (h("label", { key: '5d7a430cd34f15d1e9e8d99aaab1334fc5570dfc', class: "label" }, this.label, this.showAsterisk ? asterisk : ''));
        }
        return (h("div", { key: '7322de6d8a90959527084f8b30a098bacdcca318', class: rootClasses, "aria-disabled": this.disabled, "aria-invalid": this.invalid }, label, this.renderDropzone(), renderHint(this.hint, this.error, this.invalid), this.showFiles && this.renderFileholder()));
    }
    static get is() { return "tk-upload"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-upload.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-upload.css"]
        };
    }
    static get properties() {
        return {
            "autoUpload": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If `autoUpload` is set to `true`, the upload button will be hidden,\nand the `tkUpload` event will be automatically triggered for each newly added file."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "auto-upload",
                "defaultValue": "false"
            },
            "dragDrop": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Enables drag and drop functionality for file uploads."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "drag-drop",
                "defaultValue": "true"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Defines the label of the upload area"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Provided a hint or additional information about the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Provided a error about the upload."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "showAsterisk": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Displays a red asterisk (*) next to the label for visual emphasis."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-asterisk",
                "defaultValue": "false"
            },
            "showFiles": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Displays the uploaded files."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-files",
                "defaultValue": "true"
            },
            "chooseButtonLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label text displayed inside the choose button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "choose-button-label",
                "defaultValue": "'Choose File'"
            },
            "uploadButtonLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label text displayed inside the upload button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "upload-button-label",
                "defaultValue": "'Upload'"
            },
            "accept": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Acceptable file types for upload. Use MIME types or extensions separated by commas."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "accept",
                "defaultValue": "'*'"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disables the input, preventing user interaction."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "multiple": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Allows multiple file selection."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "multiple",
                "defaultValue": "false"
            },
            "maxFileCount": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Maximum allowed file count"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max-file-count"
            },
            "maxFileSize": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Maximum allowed file size in bytes."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "max-file-size"
            },
            "title": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Title displayed in the upload component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "title",
                "defaultValue": "'Choose a file or drag & drop it here.'"
            },
            "dragDropTitle": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Title displayed in the upload component when drag and drop is active."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "drag-drop-title",
                "defaultValue": "'Drop files here'"
            },
            "value": {
                "type": "unknown",
                "mutable": true,
                "complexType": {
                    "original": "File[]",
                    "resolved": "File[]",
                    "references": {
                        "File": {
                            "location": "global",
                            "id": "global::File"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The file value of the upload."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "description": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Description displayed under the title."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "description",
                "defaultValue": "'JPEG, PNG, PDF and MP4 formats, up to 50 MB.'"
            },
            "dragDropDescription": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Description displayed under the title when drag and drop is active."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "drag-drop-description",
                "defaultValue": "'Release to upload files'"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the upload is in an invalid state, uploads will fail eventually"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "loading": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Indicates whether the upload is in an loading state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "loading",
                "defaultValue": "false"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'default' | 'centered'",
                    "resolved": "\"centered\" | \"default\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Type of the file upload area."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'default'"
            },
            "downloadable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Indicates whether the files can be downloaded. When true, a download button will be displayed next to each file."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "downloadable",
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "errorMessages": {},
            "isDragOver": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when one or more files pass validation."
                },
                "complexType": {
                    "original": "File[]",
                    "resolved": "File[]",
                    "references": {
                        "File": {
                            "location": "global",
                            "id": "global::File"
                        }
                    }
                }
            }, {
                "method": "tkFilesRejected",
                "name": "tk-files-rejected",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when one or more files fail validation, with reasons."
                },
                "complexType": {
                    "original": "{ reason: string; file: File | FileList }[]",
                    "resolved": "{ reason: string; file: File | FileList; }[]",
                    "references": {
                        "File": {
                            "location": "global",
                            "id": "global::File"
                        },
                        "FileList": {
                            "location": "global",
                            "id": "global::FileList"
                        }
                    }
                }
            }, {
                "method": "tkUpload",
                "name": "tk-upload",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the user initiates file upload."
                },
                "complexType": {
                    "original": "File[]",
                    "resolved": "File[]",
                    "references": {
                        "File": {
                            "location": "global",
                            "id": "global::File"
                        }
                    }
                }
            }, {
                "method": "tkRemovedFile",
                "name": "tk-removed-file",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a file is removed from the accepted list."
                },
                "complexType": {
                    "original": "File",
                    "resolved": "File",
                    "references": {
                        "File": {
                            "location": "global",
                            "id": "global::File"
                        }
                    }
                }
            }, {
                "method": "tkDownloadFile",
                "name": "tk-download-file",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a file is download button is clicked."
                },
                "complexType": {
                    "original": "File",
                    "resolved": "File",
                    "references": {
                        "File": {
                            "location": "global",
                            "id": "global::File"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "valueChanged"
            }];
    }
}
