// @takeoff-ui/core@0.11.9/dist/collection/components/tk-breadcrumb/tk-breadcrumb-item/tk-breadcrumb-item.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { renderIcons } from "../../../utils/icon-utils";
import { getDataTestId } from "../../../utils/test-id-utils";
export class TkBreadcrumbItem {
    constructor() {
        /**
         * Whether the item is an external url
         * @defaultValue false
         */
        this.isExternal = false;
        /**
         * Indicates if the item is the last one
         */
        this.isCurrent = false;
    }
    render() {
        const rootClasses = classNames('tk-breadcrumb-item', {
            'tk-breadcrumb-item-current': this.isCurrent,
            'tk-breadcrumb-item-icon-only': !this.label,
        });
        const linkProps = Object.assign({ href: this.href }, (this.isExternal && { target: '_blank', rel: 'noopener noreferrer' }));
        const { leftIcon, rightIcon } = renderIcons(this.icon, {
            additionalProps: { color: this.isCurrent ? 'var(--icon-dark)' : 'var(--icon-sub-base)' },
            iconTag: 'span',
            dataTestid: this.dataTestid,
        });
        return (h("li", { key: '4da263827e8c472d3c8702d68de3608c37f1990e', class: rootClasses, "aria-current": this.isCurrent ? 'page' : null, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("a", Object.assign({ key: 'd58ce370ecff1602ab16fd621163d76d0b0c363b', class: "tk-breadcrumb-link" }, linkProps, { tabindex: this.isCurrent ? -1 : 0, "data-testid": getDataTestId(this.dataTestid, 'link') }), leftIcon, this.label && (h("span", { key: '0e6106d9cdb71663e1b9b9bb55a635f0b6d1cb36', class: "tk-breadcrumb-item-label", "data-testid": getDataTestId(this.dataTestid, 'label') }, h("slot", { key: '1eed244b3eb0bcc8a55d9356b9714fe463a3957f' }, this.label))), rightIcon)));
    }
    static get is() { return "tk-breadcrumb-item"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-breadcrumb-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-breadcrumb-item.css"]
        };
    }
    static get properties() {
        return {
            "href": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "URL for the item"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "href"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions",
                    "resolved": "IIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon to display alongside the label"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label text for the breadcrumb item"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "isExternal": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the item is an external url"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "is-external",
                "defaultValue": "false"
            },
            "isCurrent": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the item is the last one"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "is-current",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get elementRef() { return "el"; }
}
