// @takeoff-ui/core@0.11.9/dist/collection/components/tk-tooltip/tk-tooltip.js
import { h, Fragment } from "@stencil/core";
import { renderIcons } from "../../utils/icon-utils";
import { floatingElementAutoUpdate } from "../../utils/position-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The TkTooltip is used to display additional information when element is hovered over.
 * @slot trigger - The trigger slot defines the element that will trigger the tooltip
 * @slot content - Define custom HTML content for the tooltip, which replaces the default header, description and icon elements
 * @react `import { TkTooltip } from '@takeoff-ui/react'`
 * @vue `import { TkTooltip } from '@takeoff-ui/vue'`
 * @angular `import { TkTooltip } from '@takeoff-ui/angular'`
 */
export class TkTooltip {
    constructor() {
        this.isOpen = false;
        /**
         * Controls if tooltip has custom content.
         * @defaultValue false
         */
        this.hasContentSlot = false;
        /**
         * Sets the position of the tooltip.
         * @defaultValue 'right'
         */
        this.position = 'right';
        /**
         * Sets the color variant of the tooltip.
         * @defaultValue 'neutral'
         */
        this.variant = 'neutral';
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        this.handleMouseEnter = () => {
            this.isOpen = true;
        };
        this.handleMouseLeave = () => {
            this.isOpen = false;
        };
    }
    positionChanged() {
        if (this.tooltipElement) {
            this.updatePosition();
        }
    }
    componentWillLoad() {
        this.hasContentSlot = !!this.el.querySelector('[slot="content"]');
    }
    componentDidLoad() {
        var _a, _b;
        this.triggerElement = this.el.querySelector('[slot="trigger"]');
        (_a = this.triggerElement) === null || _a === void 0 ? void 0 : _a.addEventListener('mouseenter', this.handleMouseEnter);
        (_b = this.triggerElement) === null || _b === void 0 ? void 0 : _b.addEventListener('mouseleave', this.handleMouseLeave);
    }
    componentDidUpdate() {
        var _a, _b;
        if (this.isOpen) {
            // Clean up old floating UI listeners before setting up new ones
            (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
            this.updatePosition();
        }
        else {
            // Remove floating UI listeners when tooltip closes
            (_b = this.cleanup) === null || _b === void 0 ? void 0 : _b.call(this);
            // Clear reference to allow garbage collection
            this.cleanup = null;
        }
    }
    disconnectedCallback() {
        var _a;
        // Clean up floating UI listeners on component unmount
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        // Clear reference to allow garbage collection
        this.cleanup = null;
    }
    updatePosition() {
        this.cleanup = floatingElementAutoUpdate(this.triggerElement, this.tooltipElement, this.arrowElement, { placement: this.position });
    }
    render() {
        let _leftIcon;
        let _rightIcon;
        if (this.icon) {
            const { leftIcon, rightIcon } = renderIcons(this.icon, {
                variant: this.variant === 'dark' ? 'neutral' : this.variant,
                sign: true,
                size: 'small',
                dataTestid: this.dataTestid,
            });
            _leftIcon = leftIcon;
            _rightIcon = rightIcon;
        }
        return (h("div", { key: '10236d0ac062c5574174974cec6da65b70489d5d', class: "tk-tooltip", "data-testid": getDataTestId(this.dataTestid, 'container') }, h("slot", { key: '2069b22ceb94133347e13dfc54d9b32c00410787', name: "trigger" }), this.isOpen && (h("div", { key: 'cde814f061c1f107d1ca9d9e96d09ac417515281', ref: el => (this.tooltipElement = el), class: {
                'tk-tooltip-content': true,
                [`tk-tooltip-${this.variant}`]: !!this.variant,
            }, style: Object.assign({}, this.containerStyle), role: "tooltip", "data-testid": getDataTestId(this.dataTestid, 'content') }, this.hasContentSlot ? (h("slot", { name: "content" })) : (h(Fragment, null, _leftIcon, h("div", null, h("div", { class: "tk-tooltip-header", "data-testid": getDataTestId(this.dataTestid, 'header') }, this.header), h("div", { class: "tk-tooltip-description", "data-testid": getDataTestId(this.dataTestid, 'description') }, this.description)), _rightIcon)), h("div", { key: 'db79b0ab1c3d5f13c081cb62d1b5646725f74349', ref: el => (this.arrowElement = el), class: "tk-tooltip-arrow", "data-testid": getDataTestId(this.dataTestid, 'arrow') })))));
    }
    static get is() { return "tk-tooltip"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-tooltip.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-tooltip.css"]
        };
    }
    static get properties() {
        return {
            "header": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets header text for the tooltip."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "header"
            },
            "description": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets description text for the tooltip."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "description"
            },
            "position": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'top' | 'top-start' | 'top-end' | 'bottom' | 'bottom-start' | 'bottom-end' | 'left' | 'left-start' | 'left-end' | 'right' | 'right-start' | 'right-end'",
                    "resolved": "\"bottom\" | \"bottom-end\" | \"bottom-start\" | \"left\" | \"left-end\" | \"left-start\" | \"right\" | \"right-end\" | \"right-start\" | \"top\" | \"top-end\" | \"top-start\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'right'"
                        }],
                    "text": "Sets the position of the tooltip."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "position",
                "defaultValue": "'right'"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'dark' | 'white' | 'info' | 'success' | 'warning' | 'danger' | 'neutral'",
                    "resolved": "\"danger\" | \"dark\" | \"info\" | \"neutral\" | \"success\" | \"warning\" | \"white\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'neutral'"
                        }],
                    "text": "Sets the color variant of the tooltip."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'neutral'"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon name to be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of container element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "isOpen": {},
            "hasContentSlot": {}
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "position",
                "methodName": "positionChanged"
            }];
    }
}
