// @takeoff-ui/core@0.11.9/dist/collection/components/tk-treeview/tk-tree-view.js
import { h } from "@stencil/core";
import classNames from "classnames";
/**
 * The `TkTreeview` component displays hierarchical data in a tree structure with expandable/collapsible nodes.
 * Uses array-based data structure for better performance and easier data management.
 * @react `import { TkTreeView } from '@takeoff-ui/react'`
 * @vue `import { TkTreeView } from '@takeoff-ui/vue'`
 * @angular `import { TkTreeView } from '@takeoff-ui/angular'`
 */
export class TkTreeView {
    constructor() {
        this.isAllSelected = false;
        this.expandedPaths = new Set();
        this.selectedPath = null;
        /**
         * başlangıçta expanded keys lerin seçili gibi görünmesini ve herhangi bir item'a tıklandığında
         * bu state'in false'a çekilerek expandedKeys'lerin sürekli seçili görünmesini engellemek için kurgulanmıştır.s
         */
        this.isInitialLoad = true;
        /**
         * Array of tree items data. This is the primary way to provide data to the tree view.
         */
        this.items = [];
        /**
         * Tree view mode: 'basic' or 'stepper'.
         */
        this.mode = 'basic';
        /**
         * Tree view type: 'basic', 'divided', or 'light'.
         */
        this.type = 'basic';
        /**
         * Tree view size: 'large', 'base' or 'small'.
         */
        this.size = 'base';
        /**
         * If true, disables all interaction with the tree view.
         */
        this.disabled = false;
        /**
         * Icon for branch items (items with children). When empty, no icon is shown.
         */
        this.branchIcon = '';
        /**
         * Icon for leaf items (items without children). When empty, no icon is shown.
         */
        this.leafIcon = '';
        /**
         * Show/hide the badge for children count on directories.
         */
        this.showBadge = true;
        /**
         * Show/hide badges with zero count. Default is true.
         * When false, badges with 0 count will be hidden (works for both selected count and children count).
         */
        this.showZeroCountBadges = true;
        /**
         * If true, enables checkbox selection for tree items.
         */
        this.selectable = false;
        /**
         * Show/hide the pointer icon for selected items.
         */
        this.showPointer = true;
        /**
         * Selection strategy for checkboxes:
         * <br />
         * **all:** selecting a node selects the node itself and all descendants
         * <br />
         * **leaf:** selecting a node selects only leaf descendants (and leaf itself if it is a leaf)
         */
        this.selectionStrategy = 'all';
        /**
         * The style attribute of container element
         */
        this.containerStyle = null;
        /**
         * The style attribute of column element for stepper mode
         */
        this.stepStyle = null;
        /**
         * If true, shows a "Select All" checkbox above the tree items. Only effective when `selectable` is true.
         */
        this.selectAll = false;
        /**
         * Label for the "Select All" checkbox row.
         * @defaultValue 'Select All'
         */
        this.selectAllLabel = 'Select All';
        /**
         * If true, expands all nodes in basic mode.
         * <br />
         * **Note:** This prop is ignored when expandedKeys is provided.
         *
         */
        this.expandAll = false;
        this.handleToggleUnified = (pathStr, item) => {
            const descendants = this.collectDescendantPaths(pathStr, item);
            if (this.mode === 'stepper') {
                // Generate ancestor paths inline: "0-1-2" -> ["0", "0-1", "0-1-2"]
                const pathParts = pathStr.split('-');
                const ancestors = [];
                for (let i = 1; i <= pathParts.length; i++) {
                    ancestors.push(pathParts.slice(0, i).join('-'));
                }
                const isExpanded = this.expandedPaths.has(pathStr);
                if (isExpanded) {
                    // Stepper: Close self and descendants
                    if (!this.isControlled()) {
                        this.expandedPaths.delete(pathStr);
                        descendants.forEach(descPath => this.expandedPaths.delete(descPath));
                    }
                    if (pathStr === this.selectedPath || descendants.includes(this.selectedPath)) {
                        this.selectedPath = null;
                    }
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const newPaths = new Set(this.expandedPaths);
                        newPaths.delete(pathStr);
                        descendants.forEach(descPath => newPaths.delete(descPath));
                        const deepestPaths = this.getDeepestPaths(newPaths);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                }
                else {
                    // Stepper: Open path to this item
                    if (!this.isControlled()) {
                        this.expandedPaths = new Set(ancestors);
                    }
                    this.handleSelect(pathStr, item);
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const ancestorSet = new Set(ancestors);
                        const deepestPaths = this.getDeepestPaths(ancestorSet);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                    return;
                }
            }
            else {
                if (this.expandedPaths.has(pathStr)) {
                    // Basic: Close self and descendants
                    if (!this.isControlled()) {
                        this.expandedPaths.delete(pathStr);
                        descendants.forEach(descPath => this.expandedPaths.delete(descPath));
                    }
                    if (this.selectedPath === pathStr || descendants.includes(this.selectedPath)) {
                        this.selectedPath = null;
                    }
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const newPaths = new Set(this.expandedPaths);
                        newPaths.delete(pathStr);
                        descendants.forEach(descPath => newPaths.delete(descPath));
                        const deepestPaths = this.getDeepestPaths(newPaths);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                }
                else {
                    // Basic: Open only self
                    if (!this.isControlled()) {
                        this.expandedPaths.add(pathStr);
                    }
                    this.handleSelect(pathStr, item);
                    // Emit new state for controlled mode
                    if (this.isControlled()) {
                        const newPaths = new Set(this.expandedPaths);
                        newPaths.add(pathStr);
                        const deepestPaths = this.getDeepestPaths(newPaths);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                }
            }
        };
        this.handleSelect = (pathStr, item) => {
            this.selectedPath = pathStr;
            this.isInitialLoad = false;
            this.tkItemClick.emit(item);
        };
        /**
         * Get all keys (including branches and leaves) from a list of items recursively.
         */
        this.getAllKeys = (items) => {
            const keys = [];
            for (const item of items) {
                if (item.key)
                    keys.push(item.key);
                if (item.children && item.children.length > 0) {
                    keys.push(...this.getAllKeys(item.children));
                }
            }
            return keys;
        };
        /**
         * Get all child keys recursively
         */
        this.getAllChildKeys = (item) => {
            const keys = [];
            if (item.children && item.children.length > 0) {
                item.children.forEach(child => {
                    keys.push(child.key);
                    keys.push(...this.getAllChildKeys(child));
                });
            }
            return keys;
        };
        /**
         * Get all leaf descendant keys including the item's own key if it is a leaf
         */
        this.getLeafKeys = (item) => {
            const keys = [];
            const isLeaf = !item.children || item.children.length === 0;
            if (isLeaf) {
                if (item.key)
                    keys.push(item.key);
                return keys;
            }
            for (const child of item.children) {
                keys.push(...this.getLeafKeys(child));
            }
            return keys;
        };
        /**
         * Handle checkbox change events.
         */
        this.handleCheckboxChange = (checked, item) => {
            var _a;
            if (this.selectionStrategy === 'all') {
                const childKeys = this.getAllChildKeys(item);
                if (checked) {
                    // Add current item and all children
                    this.value = [...(this.value || []), item.key, ...childKeys];
                }
                else {
                    // Remove current item and all children
                    this.value = (_a = this.value) === null || _a === void 0 ? void 0 : _a.filter(key => key !== item.key && !childKeys.includes(key));
                }
            }
            else {
                // 'leaf' strategy
                const leafKeys = this.getLeafKeys(item);
                if (checked) {
                    this.value = [...(this.value || []), ...leafKeys];
                }
                else {
                    this.value = (this.value || []).filter(key => !leafKeys.includes(key));
                }
                // Ensure we don't keep non-leaf keys selected under leaf strategy
                const allLeafsUnderRoot = this.items.flatMap(root => this.getLeafKeys(root));
                this.value = (this.value || []).filter(key => allLeafsUnderRoot.includes(key));
            }
            // Deduplicate and clean undefined/null
            this.value = Array.from(new Set((this.value || []).filter((k) => !!k)));
            this.tkChange.emit(this.value);
        };
        /**
         * Check if all child nodes are selected and return checkbox state
         */
        this.getCheckboxState = (item) => {
            var _a;
            const isLeaf = !item.children || item.children.length === 0;
            const isDirectlySelected = ((_a = this.value) === null || _a === void 0 ? void 0 : _a.includes(item.key)) || false;
            // Leaf nodes: direct selection matters in both strategies
            if (isLeaf) {
                return { checked: isDirectlySelected, indeterminate: false };
            }
            // Directory nodes: compute based on children
            const childStates = item.children.map(child => this.getCheckboxState(child));
            const checkedChildren = childStates.filter(state => state.checked || state.indeterminate);
            const allChecked = childStates.every(state => state.checked);
            if (this.selectionStrategy === 'all') {
                if (allChecked && checkedChildren.length === item.children.length) {
                    return { checked: true, indeterminate: false };
                }
                else if (checkedChildren.length > 0) {
                    return { checked: false, indeterminate: true };
                }
                else {
                    return { checked: isDirectlySelected, indeterminate: false };
                }
            }
            else {
                // 'leaf' strategy: ignore isDirectlySelected for directories
                if (allChecked && checkedChildren.length === item.children.length) {
                    return { checked: true, indeterminate: false };
                }
                else if (checkedChildren.length > 0) {
                    return { checked: false, indeterminate: true };
                }
                else {
                    return { checked: false, indeterminate: false };
                }
            }
        };
        /**
         * Count selected items within the subtree rooted at the given item
         */
        this.getSelectedCount = (item) => {
            var _a;
            let count = 0;
            if (item.key && ((_a = this.value) === null || _a === void 0 ? void 0 : _a.includes(item.key))) {
                count += 1;
            }
            if (item.children && item.children.length > 0) {
                for (const child of item.children) {
                    count += this.getSelectedCount(child);
                }
            }
            return count;
        };
        /**
         * Handle item click events.
         */
        this.handleItemClick = (pathStr, item, isDisabled, isDirectory) => {
            if (this.disabled || isDisabled)
                return;
            if (isDirectory) {
                this.handleToggleUnified(pathStr, item);
            }
            else {
                // In stepper mode, when a file (leaf) is clicked, collapse any expanded
                // directory at the same level by keeping only the ancestors of the file's parent
                if (this.mode === 'stepper') {
                    const parentPath = pathStr.includes('-') ? pathStr.split('-').slice(0, -1).join('-') : '';
                    const ancestors = [];
                    if (parentPath) {
                        const parentParts = parentPath.split('-');
                        for (let i = 1; i <= parentParts.length; i++) {
                            ancestors.push(parentParts.slice(0, i).join('-'));
                        }
                    }
                    if (this.isControlled()) {
                        const ancestorSet = new Set(ancestors);
                        const deepestPaths = this.getDeepestPaths(ancestorSet);
                        const keyPaths = deepestPaths.map(indexPath => this.indexPathToKey(indexPath)).filter((key) => key !== null);
                        this.tkExpandChange.emit(keyPaths);
                    }
                    else {
                        this.expandedPaths = new Set(ancestors);
                    }
                }
                this.handleSelect(pathStr, item);
            }
        };
        this.handleSelectAllClick = () => {
            if (this.disabled)
                return;
            if (this.isAllSelected) {
                this.value = [];
            }
            else {
                this.value = this.selectionStrategy === 'leaf' ? this.items.flatMap(item => this.getLeafKeys(item)) : this.getAllKeys(this.items);
            }
            this.tkChange.emit(this.value);
        };
        this.renderItem = (item, basePath = '', index, depth = 0) => {
            var _a;
            const pathStr = basePath ? `${basePath}-${index}` : `${index}`;
            const isDirectory = !!(item.children && item.children.length > 0);
            const isExpanded = this.expandedPaths.has(pathStr);
            const isSelected = this.selectedPath === pathStr || (this.isInitialLoad && ((_a = this.expandedKeys) === null || _a === void 0 ? void 0 : _a.includes(item.key)));
            const isDisabled = this.disabled || item.disabled;
            const nodeClass = classNames('tk-tree-view', 'node', {
                directory: isDirectory,
                file: !isDirectory,
                expanded: isExpanded,
                selected: isSelected,
                disabled: isDisabled,
            });
            const selectedCount = this.getSelectedCount(item);
            return (h("div", { class: nodeClass }, this.showPointer && (isExpanded || isSelected) && h("span", { class: classNames('tk-tree-view', 'pointer', this.size) }), h("div", { class: classNames('tk-tree-view', 'label', {
                    selected: isSelected,
                    disabled: isDisabled,
                }, this.size), onClick: () => {
                    this.handleItemClick(pathStr, item, isDisabled, isDirectory);
                } }, isDirectory && this.mode === 'basic' && h("tk-icon", { variant: isSelected ? 'primary' : 'neutral', icon: isExpanded ? 'arrow_drop_down' : 'arrow_right', size: this.size }), this.selectable && (h("tk-checkbox", { onClick: e => {
                    e.stopPropagation();
                }, value: this.getCheckboxState(item).checked, indeterminate: this.getCheckboxState(item).indeterminate, disabled: isDisabled, "onTk-change": e => {
                    e.stopPropagation();
                    this.handleCheckboxChange(e.detail, item);
                } })), isDirectory && this.branchIcon && h("tk-icon", { icon: this.branchIcon, variant: isSelected ? 'primary' : 'neutral', size: this.size }), !isDirectory && this.leafIcon && h("tk-icon", { icon: this.leafIcon, variant: isSelected ? 'primary' : 'neutral', size: this.size }), h("div", { class: classNames('tk-tree-view', 'text-container', this.size) }, h("span", { class: classNames('tk-tree-view', 'text', this.size) }, item.label), (() => {
                var _a, _b, _c, _d, _e, _f, _g, _h, _j;
                const badgeCount = this.selectable ? selectedCount : (_a = item.children) === null || _a === void 0 ? void 0 : _a.length;
                const shouldShowBadge = isDirectory && this.showBadge && (this.showZeroCountBadges || badgeCount > 0);
                return (shouldShowBadge && (h("tk-badge", { count: badgeCount, size: this.size, type: (_c = (_b = this.badgeOptions) === null || _b === void 0 ? void 0 : _b.type) !== null && _c !== void 0 ? _c : 'filledlight', variant: (_e = (_d = this.badgeOptions) === null || _d === void 0 ? void 0 : _d.variant) !== null && _e !== void 0 ? _e : 'neutral', rounded: (_g = (_f = this.badgeOptions) === null || _f === void 0 ? void 0 : _f.rounded) !== null && _g !== void 0 ? _g : true, icon: (_h = this.badgeOptions) === null || _h === void 0 ? void 0 : _h.icon, iconPosition: (_j = this.badgeOptions) === null || _j === void 0 ? void 0 : _j.iconPosition })));
            })()), this.mode === 'stepper' && isDirectory && item.children && item.children.length > 0 && (h("tk-icon", { variant: isSelected ? 'primary' : 'neutral', icon: !isExpanded ? 'keyboard_arrow_down' : 'keyboard_arrow_right', size: this.size }))), this.mode === 'basic' && isDirectory && isExpanded && item.children && item.children.length > 0 && (h("div", { class: classNames('tk-tree-view', 'children') }, item.children.map((child, childIndex) => this.renderItem(child, pathStr, childIndex, depth + 1))))));
        };
    }
    itemsChanged() {
        this.initializeExpandedPaths();
    }
    modeChanged() {
        this.initializeExpandedPaths();
    }
    expandAllChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            this.initializeExpandedPaths();
        }
    }
    expandedKeysChanged(newValue) {
        if (Array.isArray(newValue)) {
            // For each value, find its index path in the tree
            const indexPaths = [];
            const invalidKeys = [];
            newValue.forEach((key, index) => {
                // Find the index path for this key
                if (this.mode === 'stepper' && index !== 0) {
                    return;
                }
                const indexPath = this.findIndexPath(key);
                if (indexPath) {
                    indexPaths.push(indexPath);
                }
                else {
                    invalidKeys.push(key);
                }
            });
            // Log invalid keys
            if (invalidKeys.length > 0) {
                console.error('Invalid keys given to expandedKeys prop:', invalidKeys);
            }
            this.expandedPaths = this.expandKeysWithAncestors(indexPaths);
            // In controlled mode, clear selection if the selected node is no longer expanded
            if (this.selectedPath !== null) {
                // Check if the selected path or any of its ancestors is collapsed
                const selectedPathParts = this.selectedPath.split('-');
                let isSelectedPathVisible = true;
                // Check each ancestor up to and including the selected path
                for (let i = 1; i <= selectedPathParts.length; i++) {
                    const ancestorPath = selectedPathParts.slice(0, i).join('-');
                    // For directories, they need to be expanded to show their children
                    if (i < selectedPathParts.length && !this.expandedPaths.has(ancestorPath)) {
                        isSelectedPathVisible = false;
                        break;
                    }
                }
                // Clear selection if not visible
                if (!isSelectedPathVisible) {
                    this.selectedPath = null;
                }
            }
        }
    }
    componentWillLoad() {
        this.initializeExpandedPaths();
    }
    componentWillRender() {
        this.isAllSelected = this.computeIsAllSelected();
    }
    /**
     * Check if the component is in controlled mode.
     * Controlled mode is when the parent component manages expansion state via expandedKeys prop.
     * We use Array.isArray() instead of checking undefined to handle edge cases like null.
     */
    isControlled() {
        return Array.isArray(this.expandedKeys);
    }
    /**
     * Find the full index-based path to a given key by searching the tree recursively
     * Example: key "4" -> "0-1-0" (index path from root to target)
     * Returns null if the key is not found
     */
    findIndexPath(targetKey, items = this.items, currentPath = []) {
        for (let i = 0; i < items.length; i++) {
            const item = items[i];
            const newPath = [...currentPath, i];
            if (item.key === targetKey) {
                return newPath.join('-');
            }
            if (item.children && item.children.length > 0) {
                const found = this.findIndexPath(targetKey, item.children, newPath);
                if (found) {
                    return found;
                }
            }
        }
        return null;
    }
    /**
     * Convert an index-based path to its corresponding key (last item in path)
     * Example: "0-1-2" -> "4" (returns only the key of the last item, not the full path)
     * Returns null if the path is invalid
     */
    indexPathToKey(indexPath) {
        const indices = indexPath.split('-').map(Number);
        let currentLevel = this.items;
        // Navigate to the target item
        for (let i = 0; i < indices.length; i++) {
            const index = indices[i];
            if (!currentLevel || index < 0 || index >= currentLevel.length) {
                return null;
            }
            const item = currentLevel[index];
            // If this is the last index, return its key
            if (i === indices.length - 1) {
                return item.key || null;
            }
            // Otherwise, navigate to children
            currentLevel = item.children;
        }
        return null;
    }
    /**
     * Get the deepest (longest) paths from a set of expanded paths
     * Example: Set(["0", "0-1", "0-1-2", "3"]) -> ["0-1-2", "3"]
     * Returns only the paths with maximum depth in each branch
     */
    getDeepestPaths(paths) {
        const allPaths = Array.from(paths);
        // Find paths that don't have any child path in the set
        return allPaths.filter(path => {
            // A path is deepest if no other path starts with it followed by a hyphen
            return !allPaths.some(otherPath => otherPath.startsWith(path + '-'));
        });
    }
    /**
     * Expand keys and automatically include all ancestor paths
     * Example: ["0-1-2"] -> Set(["0", "0-1", "0-1-2"])
     * Note: Expects index-based paths. Validation should be done before calling this method.
     */
    expandKeysWithAncestors(keys) {
        const expandedSet = new Set();
        keys.forEach(key => {
            // Add the key itself
            expandedSet.add(key);
            // Add all ancestor paths
            const parts = key.split('-');
            for (let i = 1; i < parts.length; i++) {
                const ancestorPath = parts.slice(0, i).join('-');
                expandedSet.add(ancestorPath);
            }
        });
        return expandedSet;
    }
    /**
     * Collect all descendant paths recursively
     */
    collectDescendantPaths(pathStr, item) {
        const paths = [];
        if (item.children && item.children.length > 0) {
            item.children.forEach((child, index) => {
                const childPath = `${pathStr}-${index}`;
                paths.push(childPath);
                paths.push(...this.collectDescendantPaths(childPath, child));
            });
        }
        return paths;
    }
    /**
     * Initialize expanded paths based on mode and expandAll configuration
     */
    initializeExpandedPaths() {
        if (!this.items || this.items.length === 0) {
            this.expandedPaths = new Set();
            return;
        }
        // Skip initialization if expandedKeys is being used for control
        if (this.isControlled()) {
            const indexPaths = [];
            const invalidKeys = [];
            this.expandedKeys.forEach((key, index) => {
                // Find the index path for this key
                if (this.mode === 'stepper' && index !== 0) {
                    return;
                }
                const indexPath = this.findIndexPath(key);
                if (indexPath) {
                    indexPaths.push(indexPath);
                }
                else {
                    invalidKeys.push(key);
                }
            });
            if (invalidKeys.length > 0) {
                console.error('Invalid keys given to expandedKeys prop:', invalidKeys);
            }
            this.expandedPaths = this.expandKeysWithAncestors(indexPaths);
            return;
        }
        if (!this.expandAll)
            return;
        const expanded = new Set();
        if (this.mode === 'basic') {
            // Expand all directory nodes
            const traverse = (nodes, base = '') => {
                nodes.forEach((node, idx) => {
                    const path = base ? `${base}-${idx}` : `${idx}`;
                    if (node.children && node.children.length > 0) {
                        expanded.add(path);
                        traverse(node.children, path);
                    }
                });
            };
            traverse(this.items);
        }
        else if (this.mode === 'stepper') {
            // Expand the first directory all the way down
            let level = this.items;
            let base = '';
            while (level && level.length > 0) {
                const idx = level.findIndex(n => n.children && n.children.length > 0);
                if (idx === -1)
                    break;
                const path = base ? `${base}-${idx}` : `${idx}`;
                expanded.add(path);
                base = path;
                level = level[idx].children;
            }
        }
        this.expandedPaths = expanded;
    }
    /**
     * Get stepper columns for stepper mode display.
     */
    getStepperColumns() {
        const columns = [];
        let currentLevel = this.items;
        let currentBasePath = '';
        while (currentLevel && currentLevel.length > 0) {
            columns.push({ items: currentLevel, basePath: currentBasePath });
            const expandedItem = currentLevel.find((_, index) => {
                const itemPath = currentBasePath ? `${currentBasePath}-${index}` : `${index}`;
                return this.expandedPaths.has(itemPath);
            });
            if (expandedItem && expandedItem.children && expandedItem.children.length > 0) {
                const expandedIndex = currentLevel.indexOf(expandedItem);
                currentBasePath = currentBasePath ? `${currentBasePath}-${expandedIndex}` : `${expandedIndex}`;
                currentLevel = expandedItem.children;
            }
            else {
                break;
            }
        }
        return columns;
    }
    computeIsAllSelected() {
        const allKeys = this.selectionStrategy === 'leaf' ? this.items.flatMap(item => this.getLeafKeys(item)) : this.getAllKeys(this.items);
        return allKeys.length > 0 && allKeys.every(k => { var _a; return (_a = this.value) === null || _a === void 0 ? void 0 : _a.includes(k); });
    }
    renderSelectAll() {
        var _a, _b, _c, _d;
        if (!this.selectAll || !this.selectable)
            return null;
        const isIndeterminate = !this.isAllSelected && ((_b = (_a = this.value) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) > 0;
        const selectedCount = (_d = (_c = this.value) === null || _c === void 0 ? void 0 : _c.length) !== null && _d !== void 0 ? _d : 0;
        return (h("div", { class: classNames('select-all', this.size, { disabled: this.disabled }), onClick: this.handleSelectAllClick }, h("div", { class: classNames('select-all-content') }, h("tk-checkbox", { value: this.isAllSelected, indeterminate: isIndeterminate, disabled: this.disabled, onClick: e => e.stopPropagation(), "onTk-change": e => {
                e.stopPropagation();
                this.handleSelectAllClick();
            } }), h("div", null, this.selectAllLabel)), selectedCount > 0 && h("tk-badge", { count: selectedCount, size: this.size, type: "filledlight", variant: "neutral", rounded: true })));
    }
    render() {
        if (!this.items || !this.items.length)
            return null;
        if (this.mode === 'stepper') {
            const columns = this.getStepperColumns();
            return (h("div", { class: classNames('tk-tree-view', 'stepper'), style: this.containerStyle }, this.renderSelectAll(), columns.map((column, idx) => (h("div", { class: classNames('tk-tree-view', 'column', 'divided'), key: idx, style: this.stepStyle }, column.items.map((item, itemIndex) => this.renderItem(item, column.basePath, itemIndex)))))));
        }
        else {
            const rootClasses = classNames('tk-tree-view', this.type, { disabled: this.disabled });
            return (h("div", { class: rootClasses, style: this.containerStyle }, this.renderSelectAll(), h("div", { class: classNames('tk-tree-view', 'content') }, this.items.map((item, index) => this.renderItem(item, '', index)))));
        }
    }
    static get is() { return "tk-tree-view"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-tree-view.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-tree-view.css"]
        };
    }
    static get properties() {
        return {
            "items": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ITreeItem[]",
                    "resolved": "ITreeItem[]",
                    "references": {
                        "ITreeItem": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-treeview/types.ts::ITreeItem",
                            "referenceLocation": "ITreeItem"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Array of tree items data. This is the primary way to provide data to the tree view."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "mode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'stepper'",
                    "resolved": "\"basic\" | \"stepper\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Tree view mode: 'basic' or 'stepper'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "mode",
                "defaultValue": "'basic'"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'basic' | 'divided' | 'light'",
                    "resolved": "\"basic\" | \"divided\" | \"light\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Tree view type: 'basic', 'divided', or 'light'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'basic'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Tree view size: 'large', 'base' or 'small'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, disables all interaction with the tree view."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "branchIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Icon for branch items (items with children). When empty, no icon is shown."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "branch-icon",
                "defaultValue": "''"
            },
            "leafIcon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Icon for leaf items (items without children). When empty, no icon is shown."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "leaf-icon",
                "defaultValue": "''"
            },
            "showBadge": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show/hide the badge for children count on directories."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-badge",
                "defaultValue": "true"
            },
            "showZeroCountBadges": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show/hide badges with zero count. Default is true.\nWhen false, badges with 0 count will be hidden (works for both selected count and children count)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-zero-count-badges",
                "defaultValue": "true"
            },
            "badgeOptions": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "IBadgeOptions",
                    "resolved": "IBadgeOptions",
                    "references": {
                        "IBadgeOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IBadgeOptions",
                            "id": "src/global/interfaces/IBadgeOptions.ts::IBadgeOptions",
                            "referenceLocation": "IBadgeOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Badge customization options for children count display."
                },
                "getter": false,
                "setter": false
            },
            "value": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the selected tree item."
                },
                "getter": false,
                "setter": false
            },
            "selectable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, enables checkbox selection for tree items."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "selectable",
                "defaultValue": "false"
            },
            "showPointer": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show/hide the pointer icon for selected items."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-pointer",
                "defaultValue": "true"
            },
            "selectionStrategy": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'all' | 'leaf'",
                    "resolved": "\"all\" | \"leaf\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Selection strategy for checkboxes:\n<br />\n**all:** selecting a node selects the node itself and all descendants\n<br />\n**leaf:** selecting a node selects only leaf descendants (and leaf itself if it is a leaf)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "selection-strategy",
                "defaultValue": "'all'"
            },
            "containerStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of container element"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "stepStyle": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "CSSStyleProperties",
                    "resolved": "{ [x: number]: string; accentColor: string; alignContent: string; alignItems: string; alignSelf: string; alignmentBaseline: string; all: string; animation: string; animationComposition: string; animationDelay: string; animationDirection: string; animationDuration: string; animationFillMode: string; animationIterationCount: string; animationName: string; animationPlayState: string; animationTimingFunction: string; appearance: string; aspectRatio: string; backdropFilter: string; backfaceVisibility: string; background: string; backgroundAttachment: string; backgroundBlendMode: string; backgroundClip: string; backgroundColor: string; backgroundImage: string; backgroundOrigin: string; backgroundPosition: string; backgroundPositionX: string; backgroundPositionY: string; backgroundRepeat: string; backgroundSize: string; baselineShift: string; baselineSource: string; blockSize: string; border: string; borderBlock: string; borderBlockColor: string; borderBlockEnd: string; borderBlockEndColor: string; borderBlockEndStyle: string; borderBlockEndWidth: string; borderBlockStart: string; borderBlockStartColor: string; borderBlockStartStyle: string; borderBlockStartWidth: string; borderBlockStyle: string; borderBlockWidth: string; borderBottom: string; borderBottomColor: string; borderBottomLeftRadius: string; borderBottomRightRadius: string; borderBottomStyle: string; borderBottomWidth: string; borderCollapse: string; borderColor: string; borderEndEndRadius: string; borderEndStartRadius: string; borderImage: string; borderImageOutset: string; borderImageRepeat: string; borderImageSlice: string; borderImageSource: string; borderImageWidth: string; borderInline: string; borderInlineColor: string; borderInlineEnd: string; borderInlineEndColor: string; borderInlineEndStyle: string; borderInlineEndWidth: string; borderInlineStart: string; borderInlineStartColor: string; borderInlineStartStyle: string; borderInlineStartWidth: string; borderInlineStyle: string; borderInlineWidth: string; borderLeft: string; borderLeftColor: string; borderLeftStyle: string; borderLeftWidth: string; borderRadius: string; borderRight: string; borderRightColor: string; borderRightStyle: string; borderRightWidth: string; borderSpacing: string; borderStartEndRadius: string; borderStartStartRadius: string; borderStyle: string; borderTop: string; borderTopColor: string; borderTopLeftRadius: string; borderTopRightRadius: string; borderTopStyle: string; borderTopWidth: string; borderWidth: string; bottom: string; boxDecorationBreak: string; boxShadow: string; boxSizing: string; breakAfter: string; breakBefore: string; breakInside: string; captionSide: string; caretColor: string; clear: string; clip: string; clipPath: string; clipRule: string; color: string; colorInterpolation: string; colorInterpolationFilters: string; colorScheme: string; columnCount: string; columnFill: string; columnGap: string; columnRule: string; columnRuleColor: string; columnRuleStyle: string; columnRuleWidth: string; columnSpan: string; columnWidth: string; columns: string; contain: string; containIntrinsicBlockSize: string; containIntrinsicHeight: string; containIntrinsicInlineSize: string; containIntrinsicSize: string; containIntrinsicWidth: string; container: string; containerName: string; containerType: string; content: string; contentVisibility: string; counterIncrement: string; counterReset: string; counterSet: string; cssFloat: string; cssText: string; cursor: string; cx: string; cy: string; d: string; direction: string; display: string; dominantBaseline: string; emptyCells: string; fill: string; fillOpacity: string; fillRule: string; filter: string; flex: string; flexBasis: string; flexDirection: string; flexFlow: string; flexGrow: string; flexShrink: string; flexWrap: string; float: string; floodColor: string; floodOpacity: string; font: string; fontFamily: string; fontFeatureSettings: string; fontKerning: string; fontOpticalSizing: string; fontPalette: string; fontSize: string; fontSizeAdjust: string; fontStretch: string; fontStyle: string; fontSynthesis: string; fontSynthesisSmallCaps: string; fontSynthesisStyle: string; fontSynthesisWeight: string; fontVariant: string; fontVariantAlternates: string; fontVariantCaps: string; fontVariantEastAsian: string; fontVariantLigatures: string; fontVariantNumeric: string; fontVariantPosition: string; fontVariationSettings: string; fontWeight: string; forcedColorAdjust: string; gap: string; grid: string; gridArea: string; gridAutoColumns: string; gridAutoFlow: string; gridAutoRows: string; gridColumn: string; gridColumnEnd: string; gridColumnGap: string; gridColumnStart: string; gridGap: string; gridRow: string; gridRowEnd: string; gridRowGap: string; gridRowStart: string; gridTemplate: string; gridTemplateAreas: string; gridTemplateColumns: string; gridTemplateRows: string; height: string; hyphenateCharacter: string; hyphens: string; imageOrientation: string; imageRendering: string; inlineSize: string; inset: string; insetBlock: string; insetBlockEnd: string; insetBlockStart: string; insetInline: string; insetInlineEnd: string; insetInlineStart: string; isolation: string; justifyContent: string; justifyItems: string; justifySelf: string; left: string; letterSpacing: string; lightingColor: string; lineBreak: string; lineHeight: string; listStyle: string; listStyleImage: string; listStylePosition: string; listStyleType: string; margin: string; marginBlock: string; marginBlockEnd: string; marginBlockStart: string; marginBottom: string; marginInline: string; marginInlineEnd: string; marginInlineStart: string; marginLeft: string; marginRight: string; marginTop: string; marker: string; markerEnd: string; markerMid: string; markerStart: string; mask: string; maskClip: string; maskComposite: string; maskImage: string; maskMode: string; maskOrigin: string; maskPosition: string; maskRepeat: string; maskSize: string; maskType: string; mathDepth: string; mathStyle: string; maxBlockSize: string; maxHeight: string; maxInlineSize: string; maxWidth: string; minBlockSize: string; minHeight: string; minInlineSize: string; minWidth: string; mixBlendMode: string; objectFit: string; objectPosition: string; offset: string; offsetAnchor: string; offsetDistance: string; offsetPath: string; offsetPosition: string; offsetRotate: string; opacity: string; order: string; orphans: string; outline: string; outlineColor: string; outlineOffset: string; outlineStyle: string; outlineWidth: string; overflow: string; overflowAnchor: string; overflowClipMargin: string; overflowWrap: string; overflowX: string; overflowY: string; overscrollBehavior: string; overscrollBehaviorBlock: string; overscrollBehaviorInline: string; overscrollBehaviorX: string; overscrollBehaviorY: string; padding: string; paddingBlock: string; paddingBlockEnd: string; paddingBlockStart: string; paddingBottom: string; paddingInline: string; paddingInlineEnd: string; paddingInlineStart: string; paddingLeft: string; paddingRight: string; paddingTop: string; page: string; pageBreakAfter: string; pageBreakBefore: string; pageBreakInside: string; paintOrder: string; perspective: string; perspectiveOrigin: string; placeContent: string; placeItems: string; placeSelf: string; pointerEvents: string; position: string; printColorAdjust: string; quotes: string; r: string; resize: string; right: string; rotate: string; rowGap: string; rubyAlign: string; rubyPosition: string; rx: string; ry: string; scale: string; scrollBehavior: string; scrollMargin: string; scrollMarginBlock: string; scrollMarginBlockEnd: string; scrollMarginBlockStart: string; scrollMarginBottom: string; scrollMarginInline: string; scrollMarginInlineEnd: string; scrollMarginInlineStart: string; scrollMarginLeft: string; scrollMarginRight: string; scrollMarginTop: string; scrollPadding: string; scrollPaddingBlock: string; scrollPaddingBlockEnd: string; scrollPaddingBlockStart: string; scrollPaddingBottom: string; scrollPaddingInline: string; scrollPaddingInlineEnd: string; scrollPaddingInlineStart: string; scrollPaddingLeft: string; scrollPaddingRight: string; scrollPaddingTop: string; scrollSnapAlign: string; scrollSnapStop: string; scrollSnapType: string; scrollbarColor: string; scrollbarGutter: string; scrollbarWidth: string; shapeImageThreshold: string; shapeMargin: string; shapeOutside: string; shapeRendering: string; stopColor: string; stopOpacity: string; stroke: string; strokeDasharray: string; strokeDashoffset: string; strokeLinecap: string; strokeLinejoin: string; strokeMiterlimit: string; strokeOpacity: string; strokeWidth: string; tabSize: string; tableLayout: string; textAlign: string; textAlignLast: string; textAnchor: string; textBox: string; textBoxEdge: string; textBoxTrim: string; textCombineUpright: string; textDecoration: string; textDecorationColor: string; textDecorationLine: string; textDecorationSkipInk: string; textDecorationStyle: string; textDecorationThickness: string; textEmphasis: string; textEmphasisColor: string; textEmphasisPosition: string; textEmphasisStyle: string; textIndent: string; textOrientation: string; textOverflow: string; textRendering: string; textShadow: string; textTransform: string; textUnderlineOffset: string; textUnderlinePosition: string; textWrap: string; textWrapMode: string; textWrapStyle: string; top: string; touchAction: string; transform: string; transformBox: string; transformOrigin: string; transformStyle: string; transition: string; transitionBehavior: string; transitionDelay: string; transitionDuration: string; transitionProperty: string; transitionTimingFunction: string; translate: string; unicodeBidi: string; userSelect: string; vectorEffect: string; verticalAlign: string; viewTransitionClass: string; viewTransitionName: string; visibility: string; webkitAlignContent: string; webkitAlignItems: string; webkitAlignSelf: string; webkitAnimation: string; webkitAnimationDelay: string; webkitAnimationDirection: string; webkitAnimationDuration: string; webkitAnimationFillMode: string; webkitAnimationIterationCount: string; webkitAnimationName: string; webkitAnimationPlayState: string; webkitAnimationTimingFunction: string; webkitAppearance: string; webkitBackfaceVisibility: string; webkitBackgroundClip: string; webkitBackgroundOrigin: string; webkitBackgroundSize: string; webkitBorderBottomLeftRadius: string; webkitBorderBottomRightRadius: string; webkitBorderRadius: string; webkitBorderTopLeftRadius: string; webkitBorderTopRightRadius: string; webkitBoxAlign: string; webkitBoxFlex: string; webkitBoxOrdinalGroup: string; webkitBoxOrient: string; webkitBoxPack: string; webkitBoxShadow: string; webkitBoxSizing: string; webkitFilter: string; webkitFlex: string; webkitFlexBasis: string; webkitFlexDirection: string; webkitFlexFlow: string; webkitFlexGrow: string; webkitFlexShrink: string; webkitFlexWrap: string; webkitJustifyContent: string; webkitLineClamp: string; webkitMask: string; webkitMaskBoxImage: string; webkitMaskBoxImageOutset: string; webkitMaskBoxImageRepeat: string; webkitMaskBoxImageSlice: string; webkitMaskBoxImageSource: string; webkitMaskBoxImageWidth: string; webkitMaskClip: string; webkitMaskComposite: string; webkitMaskImage: string; webkitMaskOrigin: string; webkitMaskPosition: string; webkitMaskRepeat: string; webkitMaskSize: string; webkitOrder: string; webkitPerspective: string; webkitPerspectiveOrigin: string; webkitTextFillColor: string; webkitTextSizeAdjust: string; webkitTextStroke: string; webkitTextStrokeColor: string; webkitTextStrokeWidth: string; webkitTransform: string; webkitTransformOrigin: string; webkitTransformStyle: string; webkitTransition: string; webkitTransitionDelay: string; webkitTransitionDuration: string; webkitTransitionProperty: string; webkitTransitionTimingFunction: string; webkitUserSelect: string; whiteSpace: string; whiteSpaceCollapse: string; widows: string; width: string; willChange: string; wordBreak: string; wordSpacing: string; wordWrap: string; writingMode: string; x: string; y: string; zIndex: string; zoom: string; }",
                    "references": {
                        "CSSStyleProperties": {
                            "location": "import",
                            "path": "../../global/types",
                            "id": "src/global/types.ts::CSSStyleProperties",
                            "referenceLocation": "CSSStyleProperties"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The style attribute of column element for stepper mode"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "null"
            },
            "selectAll": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, shows a \"Select All\" checkbox above the tree items. Only effective when `selectable` is true."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "select-all",
                "defaultValue": "false"
            },
            "selectAllLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'Select All'"
                        }],
                    "text": "Label for the \"Select All\" checkbox row."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "select-all-label",
                "defaultValue": "'Select All'"
            },
            "expandAll": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, expands all nodes in basic mode.\n<br />\n**Note:** This prop is ignored when expandedKeys is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "expand-all",
                "defaultValue": "false"
            },
            "expandedKeys": {
                "type": "unknown",
                "mutable": true,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Array of keys that should be expanded.\n\n<br /> **Usage:**\nProvide an array of item keys: `[\"atakan\", \"mehmet\", \"4\"]`\n<br />\nEach key must be unique in the tree structure"
                },
                "getter": false,
                "setter": false
            }
        };
    }
    static get states() {
        return {
            "expandedPaths": {},
            "selectedPath": {},
            "isInitialLoad": {}
        };
    }
    static get events() {
        return [{
                "method": "tkItemClick",
                "name": "tk-item-click",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when a tree item is clicked."
                },
                "complexType": {
                    "original": "ITreeItem",
                    "resolved": "ITreeItem",
                    "references": {
                        "ITreeItem": {
                            "location": "import",
                            "path": "./types",
                            "id": "src/components/tk-treeview/types.ts::ITreeItem",
                            "referenceLocation": "ITreeItem"
                        }
                    }
                }
            }, {
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the selected value changes."
                },
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                }
            }, {
                "method": "tkExpandChange",
                "name": "tk-expand-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the expanded paths change in controlled mode.\nEmits an array of keys (e.g., [\"4\", \"13\"]) representing the expanded items.\nOnly the keys of expanded items are emitted, not full paths."
                },
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "items",
                "methodName": "itemsChanged"
            }, {
                "propName": "mode",
                "methodName": "modeChanged"
            }, {
                "propName": "expandAll",
                "methodName": "expandAllChanged"
            }, {
                "propName": "expandedKeys",
                "methodName": "expandedKeysChanged"
            }];
    }
}
