// @takeoff-ui/core@0.11.9/dist/collection/components/tk-toggle/tk-toggle.js
import { h } from "@stencil/core";
import { v4 as uuidv4 } from "uuid";
import classNames from "classnames";
import { getIconElementProps } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
import { renderHint } from "../../utils/hint-utils";
/**
 * The TkToggle component is another basic element for user input. You can use this for turning settings, features or true/false inputs on and off.
 * @react `import { TkToggle } from '@takeoff-ui/react'`
 * @vue `import { TkToggle } from '@takeoff-ui/vue'`
 * @angular `import { TkToggle } from '@takeoff-ui/angular'`
 */
export class TkToggle {
    constructor() {
        this.uniqueId = uuidv4();
        /**
         * The internal checked state of the toggle.
         * @defaultValue false
         */
        this.checked = false;
        /**
         * The aria-labelledby attribute of the toggle.
         * @defaultValue null
         */
        this.ariaLabelledby = null;
        /**
         * Whether the toggle is disabled.
         * @defaultValue false
         */
        this.disabled = false;
        /**
         * Specifies a material icon name to be displayed.
         * @defaultValue check
         */
        this.icon = 'check';
        /**
         * The name attribute of the toggle.
         * @defaultValue null
         */
        this.name = null;
        /**
         * Whether to show the icon in the toggle.
         * @defaultValue true
         */
        this.showIcon = true;
        /**
         * Sets size for the component.
         * @defaultValue base
         */
        this.size = 'base';
        /**
         * The variant of the toggle.
         * @defaultValue info
         */
        this.variant = 'info';
        /**
         * Whether the toggle is in an invalid state.
         * @defaultValue false
         */
        this.invalid = false;
        /**
         * The current state of the toggle.
         * @defaultValue false
         */
        this.value = false;
        this.handleInputChange = (event) => {
            if (!this.disabled) {
                const target = event.target;
                this.value = target.checked;
                this.checked = target.checked;
                this.tkChange.emit(this.checked);
            }
        };
    }
    valueChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            this.checked = newValue;
        }
    }
    componentWillLoad() {
        this.checked = this.value;
        // If the tk-toggle has a tabindex attribute we get the value
        // and pass it down to the native input, then remove it from the
        // tk-toggle to avoid causing tabbing twice on the same element
        if (this.el.hasAttribute('tabindex')) {
            const tabindex = Number(this.el.getAttribute('tabindex'));
            this.tabindex = tabindex !== null ? tabindex : undefined;
            this.el.removeAttribute('tabindex');
        }
        this.hasDefaultSlot = Array.from(this.el.childNodes).some(node => { var _a; return (node.nodeType === Node.ELEMENT_NODE && !node.hasAttribute('slot')) || (node.nodeType === Node.TEXT_NODE && ((_a = node.textContent) === null || _a === void 0 ? void 0 : _a.trim().length) > 0); });
    }
    formResetCallback() {
        this.handleFormReset();
    }
    /**
     * Returns the native `<input>` element used under the hood.
     */
    getInputElement() {
        return Promise.resolve(this.nativeInput);
    }
    getActiveIcon() {
        return this.checked ? (this.invalid ? 'close' : this.icon) : '';
    }
    handleFormReset() {
        this.value = false;
        this.checked = false;
        this.invalid = false;
        this.tkChange.emit(this.value);
    }
    renderInput() {
        return (h("input", { id: this.uniqueId, ref: input => (this.nativeInput = input), class: "tk-toggle-input", type: "checkbox", role: "switch", disabled: this.disabled, checked: this.checked, onChange: this.handleInputChange, tabIndex: this.tabindex, "aria-checked": this.checked.toString(), "aria-disabled": this.disabled, "aria-invalid": this.invalid || undefined, "aria-label": this.name, "aria-labelledby": this.ariaLabelledby, "data-testid": getDataTestId(this.dataTestid, 'native-input') }));
    }
    render() {
        const rootClasses = classNames('tk-toggle', `tk-toggle-${this.size}`, `tk-toggle-${this.variant}`, {
            'tk-toggle-disabled': this.disabled,
            'tk-toggle-invalid': this.invalid,
        });
        return (h("div", { key: '94489c42774b04248d77f1ec81e0127baa6d0b2c', class: rootClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("label", { key: '32921838a47b29b512162608f63fb5f76a358a57', htmlFor: this.uniqueId, "data-testid": getDataTestId(this.dataTestid, 'label-container') }, h("div", { key: '72dec5f2f583488b31bedc790c850f5efc2740be', class: "tk-toggle-input-container", "data-testid": getDataTestId(this.dataTestid, 'control') }, this.renderInput(), h("span", { key: 'b27e148954fa9fcfa2baecf47e433e44a57431ff', class: "tk-toggle-thumb", "data-testid": getDataTestId(this.dataTestid, 'thumb') }, this.showIcon && (h("tk-icon", Object.assign({ key: '28c633b63b2d92b0269c8288826a6d96490cd43c' }, getIconElementProps(this.getActiveIcon(), {
            variant: this.invalid ? 'danger' : this.variant,
            size: this.size === 'large' || this.size === 'xlarge' ? 'medium' : this.size,
            color: this.disabled ? 'var(--icon-lightest)' : '',
        }), { dataTestid: getDataTestId(this.dataTestid, 'thumb-icon') }))))), this.hasDefaultSlot ? (h("slot", null)) : (this.label && (h("span", { class: "tk-toggle-label", "data-testid": getDataTestId(this.dataTestid, 'label') }, this.label)))), renderHint(this.hint, this.error, this.invalid, this.dataTestid)));
    }
    static get is() { return "tk-toggle"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-toggle.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-toggle.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabelledby": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "null"
                        }],
                    "text": "The aria-labelledby attribute of the toggle."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "aria-labelledby",
                "defaultValue": "null"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the toggle is disabled."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "disabled",
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "check"
                        }],
                    "text": "Specifies a material icon name to be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon",
                "defaultValue": "'check'"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "null"
                        }],
                    "text": "The name attribute of the toggle."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name",
                "defaultValue": "null"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The label for the toggle."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "hint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Provides a hint or additional information about the toggle."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hint"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "This is the error message that will be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "error"
            },
            "showIcon": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Whether to show the icon in the toggle."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-icon",
                "defaultValue": "true"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'xlarge' | 'large' | 'base' | 'small' | 'xsmall'",
                    "resolved": "\"base\" | \"large\" | \"small\" | \"xlarge\" | \"xsmall\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "base"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'secondary' | 'neutral' | 'info' | 'success' | 'warning' | 'danger' | 'verified' | 'purple' | 'cyan' | 'business' | 'teal' | 'dark' | 'white'",
                    "resolved": "\"business\" | \"cyan\" | \"danger\" | \"dark\" | \"info\" | \"neutral\" | \"primary\" | \"purple\" | \"secondary\" | \"success\" | \"teal\" | \"verified\" | \"warning\" | \"white\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "info"
                        }],
                    "text": "The variant of the toggle."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'info'"
            },
            "invalid": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the toggle is in an invalid state."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "invalid",
                "defaultValue": "false"
            },
            "value": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "The current state of the toggle."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "value",
                "defaultValue": "false"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hasDefaultSlot": {},
            "checked": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "event",
                            "text": "onChange"
                        }],
                    "text": "Event emitted when the toggle value changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "getInputElement": {
                "complexType": {
                    "signature": "() => Promise<HTMLInputElement>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLInputElement": {
                            "location": "global",
                            "id": "global::HTMLInputElement"
                        }
                    },
                    "return": "Promise<HTMLInputElement>"
                },
                "docs": {
                    "text": "Returns the native `<input>` element used under the hood.",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "valueChanged"
            }];
    }
    static get attachInternalsMemberName() { return "internals"; }
}
