// @takeoff-ui/core@0.11.9/dist/collection/components/tk-badge/tk-badge.js
import { h, Host } from "@stencil/core";
import classNames from "classnames";
import { renderIcons, isMultiIconOptions } from "../../utils/icon-utils";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The TkBadge component allows you to create a small badge for adding information like contextual data that needs to stand out and get noticed. It is also often useful in combination with other elements like a user avatar to show a number of new messages.
 * @react `import { TkBadge } from '@takeoff-ui/react'`
 * @vue `import { TkBadge } from '@takeoff-ui/vue'`
 * @angular `import { TkBadge } from '@takeoff-ui/angular'`
 */
export class TkBadge {
    constructor() {
        /**
         * State to check if slot content exists.
         * @defaultValue false
         */
        this.hasSlot = false;
        /**
         * If true, shows a small dot on the badge.
         * @defaultValue false
         */
        this.dot = false;
        /**
         * This field specifies the design type of the component.
         * @defaultValue filled
         */
        this.type = 'filled';
        /**
         * Defines the position of the icon.
         * @defaultValue 'left'
         */
        this.iconPosition = 'left';
        /**
         * Makes the badge corners rounded.
         * @defaultValue false
         */
        this.rounded = false;
        /**
         * Sets size for the component.
         * @defaultValue 'base'
         */
        this.size = 'base';
        /**
         * If true, the badge will take the full width of its container.
         * @defaultValue false
         */
        this.fullWidth = false;
        /**
         * Determines the badge's variant for different styles.
         * @defaultValue 'primary'
         */
        this.variant = 'primary';
    }
    componentDidLoad() {
        var _a, _b;
        this.hasSlot = ((_b = (_a = this.el.shadowRoot.querySelector('slot')) === null || _a === void 0 ? void 0 : _a.assignedNodes) === null || _b === void 0 ? void 0 : _b.call(_a).length) > 0;
    }
    isValidCount() {
        if (this.count === undefined || this.count === null) {
            return false;
        }
        if (this.count === 0 || this.count === '0') {
            return true;
        }
        // Handle other numeric strings and numbers
        const numericValue = Number(this.count);
        return !isNaN(numericValue);
    }
    renderContent() {
        let content;
        if (!this.dot) {
            const contentConfig = {
                label: {
                    value: this.label,
                    class: 'label',
                },
                count: {
                    value: this.count,
                    class: 'count',
                },
            };
            const type = this.label ? 'label' : this.isValidCount() ? 'count' : null;
            const { value = '', class: contentClass = '' } = contentConfig[type] || {};
            content = (h("span", { class: contentClass, "data-testid": getDataTestId(this.dataTestid, contentClass) }, value));
        }
        return content;
    }
    render() {
        const isCountOnly = this.isValidCount() && !this.label && !this.dot && !this.icon;
        const hasMultipleIcons = isMultiIconOptions(this.icon);
        const rootClasses = classNames('tk-badge-container', {
            'has-slot': this.hasSlot,
        });
        const badgeClasses = classNames('tk-badge', this.variant, this.size, this.type, {
            'rounded': this.rounded,
            'icon-only': this.icon && !this.label && !this.dot && !hasMultipleIcons,
            'dot': this.dot,
            'count': isCountOnly,
            [`icon-${this.iconPosition}`]: !!this.icon && !this.dot,
        });
        // Handle icon rendering using utility function
        let _leftIcon;
        let _rightIcon;
        if (this.icon && !this.dot) {
            const { leftIcon, rightIcon } = renderIcons(this.icon, {
                variant: this.variant,
                size: this.size,
                additionalProps: Object.assign({}, (this.type === 'filled' ? { color: 'var(--static-white)' } : {})),
                dataTestid: this.dataTestid,
            }, this.iconPosition);
            _leftIcon = leftIcon;
            _rightIcon = rightIcon;
        }
        return (h(Host, { key: 'c13841f4c7b1ede2f6802ba33928c30f27cb36e6', class: { 'full-width': this.fullWidth && !this.hasSlot } }, h("div", { key: '63a058643c9492f7e5a82ae44e6d79582ee7a6f7', class: rootClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("slot", { key: '9d100bb10f3f06eb1a08806d12ab283585c7c1a3' }), h("span", { key: '4992a146aff12490943720a2ebd1679a05edcaf8', class: badgeClasses, "data-testid": getDataTestId(this.dataTestid, 'content') }, _leftIcon, this.renderContent(), _rightIcon))));
    }
    static get is() { return "tk-badge"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-badge.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-badge.css"]
        };
    }
    static get properties() {
        return {
            "dot": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If true, shows a small dot on the badge."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "dot",
                "defaultValue": "false"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'filled' | 'filledlight' | 'outlined' | 'text'",
                    "resolved": "\"filled\" | \"filledlight\" | \"outlined\" | \"text\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "filled"
                        }],
                    "text": "This field specifies the design type of the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "type",
                "defaultValue": "'filled'"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | IIconOptions | IMultiIconOptions",
                    "resolved": "IIconOptions | IMultiIconOptions | string",
                    "references": {
                        "IIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IIconOptions",
                            "referenceLocation": "IIconOptions"
                        },
                        "IMultiIconOptions": {
                            "location": "import",
                            "path": "../../global/interfaces/IIconOptions",
                            "id": "src/global/interfaces/IIconOptions.ts::IMultiIconOptions",
                            "referenceLocation": "IMultiIconOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies a material icon name to be displayed."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon"
            },
            "iconPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'right'",
                    "resolved": "\"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'left'"
                        }],
                    "text": "Defines the position of the icon."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "icon-position",
                "defaultValue": "'left'"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the label for the element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label"
            },
            "count": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "number | string",
                    "resolved": "number | string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Defines the number for the element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "count"
            },
            "rounded": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Makes the badge corners rounded."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rounded",
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'large' | 'base' | 'small'",
                    "resolved": "\"base\" | \"large\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Sets size for the component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            },
            "fullWidth": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "If true, the badge will take the full width of its container."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "full-width",
                "defaultValue": "false"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'secondary' | 'neutral' | 'info' | 'success' | 'danger' | 'warning' | 'verified' | 'purple' | 'cyan' | 'business' | 'teal'",
                    "resolved": "\"business\" | \"cyan\" | \"danger\" | \"info\" | \"neutral\" | \"primary\" | \"purple\" | \"secondary\" | \"success\" | \"teal\" | \"verified\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'primary'"
                        }],
                    "text": "Determines the badge's variant for different styles."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'primary'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "hasSlot": {}
        };
    }
    static get elementRef() { return "el"; }
}
