// @takeoff-ui/core@0.11.9/dist/collection/components/tk-avatar/tk-avatar.js
import { h } from "@stencil/core";
import classNames from "classnames";
/**
 * The `TkAvatar` represents a user, labels, and display the images or a brand.
 * @react `import { TkAvatar, TkAvatarGroup } from '@takeoff-ui/react'`
 * @vue `import { TkAvatar, TkAvatarGroup } from '@takeoff-ui/vue'`
 * @angular `import { TkAvatar, TkAvatarGroup } from '@takeoff-ui/angular'`
 */
export class TkAvatar {
    constructor() {
        /**
         * ID of the element that labels the avatar - accessibility
         */
        this.ariaLabelledby = null;
        /**
         * Background style of the avatar
         * @defaultValue 'brand'
         */
        this.background = 'brand';
        /**
         * The badge to be displayed in the avatar
         * @defaultValue false
         */
        this.badge = false;
        /**
         * The status of the badge
         * @defaultValue success
         */
        this.badgeStatus = 'success';
        /**
         * URL of the image to be displayed in the avatar
         */
        this.image = null;
        /**
         * Text label to be displayed in the avatar
         */
        this.label = null;
        /**
         * Name associated with the avatar - accessibility
         */
        this.name = null;
        /**
         * Whether the avatar should have rounded corners
         * @defaultValue false
         */
        this.rounded = false;
        /**
         * Whether to hide the shadow effect on the avatar
         * @defaultValue false
         */
        this.hideShadow = false;
        /**
         * Appearance of the avatar
         * @defaultValue 'primary'
         */
        this.variant = 'primary';
        /**
         * Size of the avatar
         * @defaultValue 'base'
         */
        this.size = 'base';
    }
    renderDefaultSvg() {
        return (h("svg", { xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 64 52", fill: "currentColor" }, h("circle", { cx: "32.3068", cy: "12.0212", r: "12.0212", fill: "currentColor" }), h("path", { d: "M64 54.0952C64 46.9218 60.6286 40.0421 54.6274 34.9697C48.6263 29.8973 40.4869 27.0476 32 27.0476C23.5131 27.0476 15.3737 29.8973 9.37259 34.9697C3.37142 40.0421 1.28149e-06 46.9217 0 54.0952L32 54.0952H64Z", fill: "currentColor" })));
    }
    render() {
        const containerClasses = classNames('tk-avatar-container', `tk-avatar-${this.size}`);
        const rootClasses = classNames('tk-avatar', `tk-avatar-${this.variant}`, `tk-avatar-${this.background}`, {
            'tk-avatar-rounded': this.rounded,
            'tk-avatar-hide-shadow': this.hideShadow,
        });
        return (h("div", { key: '6b947c7376691db139a3abc2ce4f32353ea2310a', class: containerClasses, part: `avatar ${this.size}` }, h("div", { key: '2447b3629ca07864974a05e226b1b900927da43d', class: rootClasses, part: "avatar", "aria-label": this.name, "aria-labelledby": this.ariaLabelledby }, this.label && h("span", { key: '30d5f6725d7d2d31f8ee7e5f070003ad6dbd8f81', class: "tk-avatar-label" }, this.label), this.image && h("img", { key: 'bb35b2231286e76c300377eb147b6d79c0321db4', class: "tk-avatar-image", src: this.image, alt: this.name }), !this.label && !this.image && this.renderDefaultSvg()), this.badge && this.rounded && h("span", { key: '1624479704d3c89f61a485e75be64f0fc6c39099', part: "badge", class: classNames('tk-avatar-badge', `tk-avatar-badge-status-${this.badgeStatus}`) })));
    }
    static get is() { return "tk-avatar"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-avatar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-avatar.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabelledby": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ID of the element that labels the avatar - accessibility"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "aria-labelledby",
                "defaultValue": "null"
            },
            "background": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'brand' | 'solid'",
                    "resolved": "\"brand\" | \"solid\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'brand'"
                        }],
                    "text": "Background style of the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "background",
                "defaultValue": "'brand'"
            },
            "badge": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "The badge to be displayed in the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "badge",
                "defaultValue": "false"
            },
            "badgeStatus": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'success' | 'warning' | 'info' | 'danger'",
                    "resolved": "\"danger\" | \"info\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "success"
                        }],
                    "text": "The status of the badge"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "badge-status",
                "defaultValue": "'success'"
            },
            "image": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "URL of the image to be displayed in the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "image",
                "defaultValue": "null"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Text label to be displayed in the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "label",
                "defaultValue": "null"
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Name associated with the avatar - accessibility"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "name",
                "defaultValue": "null"
            },
            "rounded": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the avatar should have rounded corners"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "rounded",
                "defaultValue": "false"
            },
            "hideShadow": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether to hide the shadow effect on the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "hide-shadow",
                "defaultValue": "false"
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'primary' | 'light' | 'success' | 'info' | 'warning' | 'danger'",
                    "resolved": "\"danger\" | \"info\" | \"light\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'primary'"
                        }],
                    "text": "Appearance of the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "variant",
                "defaultValue": "'primary'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'xsmall' | 'small' | 'base' | 'large' | 'xlarge'",
                    "resolved": "\"base\" | \"large\" | \"small\" | \"xlarge\" | \"xsmall\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'base'"
                        }],
                    "text": "Size of the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "size",
                "defaultValue": "'base'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
