// @takeoff-ui/core@0.11.9/dist/collection/components/tk-avatar/tk-avatar-group.js
import { h } from "@stencil/core";
import classNames from "classnames";
export class TkAvatarGroup {
    constructor() {
        /**
         * Whether the avatars should have a shared background
         * @defaultValue false
         */
        this.compact = false;
    }
    render() {
        return (h("div", { key: '5484f708b0fb8d47487903a903371d56791af90d', class: classNames('tk-avatar-group', { 'tk-avatar-group-compact': this.compact }) }, h("slot", { key: '13945b1308fa374be87f317d4f22b5c387951378' })));
    }
    static get is() { return "tk-avatar-group"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-avatar-group.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-avatar-group.css"]
        };
    }
    static get properties() {
        return {
            "compact": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Whether the avatars should have a shared background"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "compact",
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "el"; }
}
