// @takeoff-ui/core@0.11.9/dist/collection/components/tk-carousel/tk-carousel.js
import { h } from "@stencil/core";
import classNames from "classnames";
import { getDataTestId } from "../../utils/test-id-utils";
/**
 * The `TkCarousel` is a content slider component with various options.
 * @react `import { TkCarousel } from '@takeoff-ui/react'`
 * @vue `import { TkCarousel } from '@takeoff-ui/vue'`
 * @angular `import { TkCarousel } from '@takeoff-ui/angular'`
 */
export class TkCarousel {
    constructor() {
        this.items = [];
        this.activeIndex = 0;
        this.totalItems = 0;
        /**
         * Controls whether the carousel indicators are shown
         * @defaultValue true
         */
        this.showIndicators = true;
        /**
         * Controls whether the navigation arrows are shown
         * @defaultValue true
         */
        this.showArrows = true;
        /**
         * Controls whether the carousel should autoplay
         * @defaultValue false
         */
        this.autoplay = false;
        /**
         * Controls the interval of the autoplay in milliseconds
         * @defaultValue 3000
         */
        this.autoplayDelay = 3000;
        /**
         * Controls whether it should loop back to the start after reaching the end
         * @defaultValue true
         */
        this.circular = true;
        /**
         * Placement of the navigation indicators
         * @defaultValue 'inside'
         */
        this.navigationPlacement = 'inside';
        /**
         * Position of the navigation indicators.
         * For horizontal orientation: 'distributed' | 'top' | 'bottom'
         * For vertical orientation: 'left' | 'right'
         * @defaultValue 'distributed'
         */
        this.navigationPosition = 'distributed';
        /**
         * Controls whether the pause/play button is shown
         * @defaultValue false
         */
        this.showPlayerButton = false;
        /**
         * Number of items to show per view
         * @defaultValue 1
         */
        this.itemsPerView = 1;
        /**
         * Orientation of the carousel
         * @defaultValue 'horizontal'
         */
        this.orientation = 'horizontal';
        /**
         * Height of the carousel when orientation is vertical
         * @defaultValue '300px'
         */
        this.verticalViewHeight = '300px';
        this.startAutoplay = () => {
            if (!this.autoplay || this.totalItems <= this.itemsPerView)
                return;
            this.stopAutoplay();
            this.autoplayTimer = window.setInterval(() => {
                const lastStartingItem = this.totalItems - this.itemsPerView;
                if (this.activeIndex >= lastStartingItem) {
                    if (this.circular) {
                        this.changeItem(0);
                    }
                    else {
                        this.stopAutoplay();
                    }
                }
                else {
                    this.changeItem(this.activeIndex + 1);
                }
            }, this.autoplayDelay);
        };
        this.stopAutoplay = () => {
            if (this.autoplayTimer) {
                clearInterval(this.autoplayTimer);
                this.autoplayTimer = undefined;
            }
        };
        this.handlePrevClick = () => {
            this.changeItem(this.activeIndex - 1);
            this.tkChange.emit(this.activeIndex);
        };
        this.handleNextClick = () => {
            this.changeItem(this.activeIndex + 1);
            this.tkChange.emit(this.activeIndex);
        };
        this.handleIndicatorClick = (index) => {
            this.changeItem(index);
        };
        this.handleKeyDown = (e) => {
            if (this.orientation === 'vertical') {
                if (e.key === 'ArrowUp') {
                    e.preventDefault();
                    this.handlePrevClick();
                }
                else if (e.key === 'ArrowDown') {
                    e.preventDefault();
                    this.handleNextClick();
                }
            }
            else {
                if (e.key === 'ArrowLeft') {
                    e.preventDefault();
                    this.handlePrevClick();
                }
                else if (e.key === 'ArrowRight') {
                    e.preventDefault();
                    this.handleNextClick();
                }
            }
        };
    }
    componentDidLoad() {
        var _a, _b;
        if (!this.slotElement) {
            this.slotElement = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('slot');
        }
        this.itemsContainer = (_b = this.el.shadowRoot) === null || _b === void 0 ? void 0 : _b.querySelector('.tk-carousel-items-container');
        this.el.tabIndex = 0;
        this.updateItemPosition();
        this.updateItems();
        this.startAutoplay();
        this.el.addEventListener('keydown', this.handleKeyDown);
    }
    disconnectedCallback() {
        this.stopAutoplay();
        this.el.removeEventListener('keydown', this.handleKeyDown);
    }
    /**
     * Returns a valid navigation position based on orientation.
     * Horizontal: 'distributed' | 'top' | 'bottom'
     * Vertical: 'left' | 'right'
     */
    get validNavigationPosition() {
        const horizontalPositions = ['distributed', 'top', 'bottom'];
        const verticalPositions = ['left', 'right'];
        if (this.orientation === 'vertical') {
            return verticalPositions.includes(this.navigationPosition) ? this.navigationPosition : 'right';
        }
        return horizontalPositions.includes(this.navigationPosition) ? this.navigationPosition : 'distributed';
    }
    updateItems() {
        var _a;
        const assigned = ((_a = this.slotElement) === null || _a === void 0 ? void 0 : _a.assignedElements()) || [];
        this.items = assigned.filter(el => !el.hasAttribute('slot'));
        this.totalItems = this.items.length;
    }
    updateItemPosition() {
        var _a;
        if (!this.itemsContainer)
            return;
        const itemsContainerElement = (_a = this.el.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('.tk-carousel-overlay');
        if (!itemsContainerElement)
            return;
        const gapSize = 16;
        const totalGaps = Math.max(0, this.itemsPerView - 1);
        const totalGapSpace = totalGaps * gapSize;
        let containerSize;
        if (this.orientation === 'horizontal') {
            containerSize = itemsContainerElement.getBoundingClientRect().width;
        }
        else {
            this.el.style.height = String(this.verticalViewHeight);
            containerSize = Number.parseFloat(this.verticalViewHeight);
        }
        const itemSize = (containerSize - totalGapSpace) / this.itemsPerView;
        const translateValue = this.activeIndex * (itemSize + gapSize);
        if (this.orientation === 'vertical') {
            this.itemsContainer.style.transform = `translateY(-${translateValue}px)`;
        }
        else {
            this.itemsContainer.style.transform = `translateX(-${translateValue}px)`;
        }
    }
    changeItem(index) {
        if (!this.totalItems)
            return;
        const lastStartingItem = Math.max(0, this.totalItems - this.itemsPerView);
        if (this.circular) {
            if (index < 0) {
                index = lastStartingItem;
            }
            else if (index > lastStartingItem) {
                index = 0;
            }
        }
        else {
            if (index < 0)
                index = 0;
            if (index > lastStartingItem)
                index = lastStartingItem;
        }
        this.activeIndex = index;
        this.updateItemPosition();
        this.tkChange.emit(this.activeIndex);
    }
    createPrevButton() {
        if (!this.showArrows || this.items.length <= this.itemsPerView)
            return null;
        if (!this.circular && this.activeIndex === 0)
            return null;
        return (h("tk-button", { size: "small", class: "prev-button", icon: this.orientation === 'vertical' ? 'keyboard_arrow_up' : 'chevron_left', "onTk-click": this.handlePrevClick, dataTestid: getDataTestId(this.dataTestid, 'prev-button') }));
    }
    createNextButton() {
        const lastStartingItem = Math.max(0, this.totalItems - this.itemsPerView);
        if (!this.showArrows || this.items.length <= this.itemsPerView)
            return null;
        if (!this.circular && this.activeIndex >= lastStartingItem)
            return null;
        return (h("tk-button", { size: "small", class: "next-button", icon: this.orientation === 'vertical' ? 'keyboard_arrow_down' : 'chevron_right', "onTk-click": this.handleNextClick, dataTestid: getDataTestId(this.dataTestid, 'next-button') }));
    }
    createPlayerButtons() {
        if (!this.showPlayerButton || !this.autoplay)
            return null;
        const iconProps = {
            color: 'var(--background-light)',
            icon: this.autoplayTimer ? 'pause_circle' : 'play_circle',
            onClick: this.autoplayTimer ? this.stopAutoplay : this.startAutoplay,
        };
        return h("tk-icon", Object.assign({ class: "player-button", size: "small" }, iconProps, { dataTestid: getDataTestId(this.dataTestid, 'player-icon') }));
    }
    createIndicators() {
        if (!this.showIndicators || this.items.length <= this.itemsPerView)
            return null;
        const indicatorCount = Math.max(1, this.totalItems - this.itemsPerView + 1);
        return (h("div", { class: "tk-carousel-indicators", "data-testid": getDataTestId(this.dataTestid, 'indicators') }, this.createPlayerButtons(), Array.from({ length: indicatorCount }).map((_, index) => {
            const indicatorDataTestid = getDataTestId(this.dataTestid, 'indicator', index.toString());
            return (h("div", { class: "tk-carousel-indicator", onClick: () => this.handleIndicatorClick(index), "data-testid": getDataTestId(indicatorDataTestid) }, h("div", { class: classNames('tk-carousel-indicator-dot', { active: index === this.activeIndex }), "data-testid": getDataTestId(indicatorDataTestid, 'dot') })));
        })));
    }
    renderNavigationElement() {
        const navPosition = this.validNavigationPosition;
        return (h("div", { class: classNames('tk-carousel-navigation-holder', { 'tk-carousel-navigation': navPosition !== 'distributed' }, { 'vertical-navigation': navPosition === 'left' || navPosition === 'right' }), "data-testid": getDataTestId(this.dataTestid, 'navigation') }, this.createPrevButton(), this.createIndicators(), this.createNextButton()));
    }
    render() {
        const navPosition = this.validNavigationPosition;
        const rootClasses = classNames('tk-carousel', this.navigationPlacement, navPosition, { vertical: this.orientation === 'vertical' }, { 'show-arrows': this.showArrows }, { 'show-indicators': this.showIndicators });
        this.el.style.setProperty('--items-per-view', String(this.itemsPerView));
        return (h("div", { key: '77f46c60f3210b667dd41a7824f2193ffbff8f14', class: rootClasses, "data-testid": getDataTestId(this.dataTestid, 'container') }, h("div", { key: '2cd9391d7a28317ce9f6fdbd4a5c53ce001b9c9b', class: "tk-carousel-overlay", "data-testid": getDataTestId(this.dataTestid, 'overlay') }, h("div", { key: 'c34ff3931692c586dbfdad306f11528fa33ece41', class: "tk-carousel-items-container", "data-testid": getDataTestId(this.dataTestid, 'items') }, h("slot", { key: '28b8ab6ad330463aa8a47161de039af37ae8aa3d', onSlotchange: () => {
                this.updateItems();
                this.updateItemPosition();
            } }))), this.renderNavigationElement()));
    }
    static get is() { return "tk-carousel"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tk-carousel.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tk-carousel.css"]
        };
    }
    static get properties() {
        return {
            "showIndicators": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Controls whether the carousel indicators are shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-indicators",
                "defaultValue": "true"
            },
            "showArrows": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Controls whether the navigation arrows are shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-arrows",
                "defaultValue": "true"
            },
            "autoplay": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the carousel should autoplay"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "autoplay",
                "defaultValue": "false"
            },
            "autoplayDelay": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "3000"
                        }],
                    "text": "Controls the interval of the autoplay in milliseconds"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "autoplay-delay",
                "defaultValue": "3000"
            },
            "circular": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "true"
                        }],
                    "text": "Controls whether it should loop back to the start after reaching the end"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "circular",
                "defaultValue": "true"
            },
            "navigationPlacement": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'inside' | 'outside'",
                    "resolved": "\"inside\" | \"outside\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'inside'"
                        }],
                    "text": "Placement of the navigation indicators"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "navigation-placement",
                "defaultValue": "'inside'"
            },
            "navigationPosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'distributed' | 'top' | 'bottom' | 'left' | 'right'",
                    "resolved": "\"bottom\" | \"distributed\" | \"left\" | \"right\" | \"top\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'distributed'"
                        }],
                    "text": "Position of the navigation indicators.\nFor horizontal orientation: 'distributed' | 'top' | 'bottom'\nFor vertical orientation: 'left' | 'right'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "navigation-position",
                "defaultValue": "'distributed'"
            },
            "showPlayerButton": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "false"
                        }],
                    "text": "Controls whether the pause/play button is shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "show-player-button",
                "defaultValue": "false"
            },
            "itemsPerView": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "1"
                        }],
                    "text": "Number of items to show per view"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "items-per-view",
                "defaultValue": "1"
            },
            "orientation": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'horizontal' | 'vertical'",
                    "resolved": "\"horizontal\" | \"vertical\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'horizontal'"
                        }],
                    "text": "Orientation of the carousel"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "orientation",
                "defaultValue": "'horizontal'"
            },
            "verticalViewHeight": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "defaultValue",
                            "text": "'300px'"
                        }],
                    "text": "Height of the carousel when orientation is vertical"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "attribute": "vertical-view-height",
                "defaultValue": "'300px'"
            },
            "dataTestid": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Sets the data-testid attribute on the root container element."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "attribute": "data-testid"
            }
        };
    }
    static get states() {
        return {
            "autoplayTimer": {},
            "activeIndex": {},
            "totalItems": {}
        };
    }
    static get events() {
        return [{
                "method": "tkChange",
                "name": "tk-change",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when item is changed"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
}
